// @cdktf/provider-aws/lib/worklink/worklink-fleet.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorklinkFleet = exports.WorklinkFleetNetworkOutputReference = exports.worklinkFleetNetworkToTerraform = exports.WorklinkFleetIdentityProviderOutputReference = exports.worklinkFleetIdentityProviderToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function worklinkFleetIdentityProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        saml_metadata: cdktf.stringToTerraform(struct.samlMetadata),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.worklinkFleetIdentityProviderToTerraform = worklinkFleetIdentityProviderToTerraform;
/**
 * @stability stable
 */
class WorklinkFleetIdentityProviderOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._samlMetadata) {
            hasAnyValues = true;
            internalValueResult.samlMetadata = this._samlMetadata;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._samlMetadata = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._samlMetadata = value.samlMetadata;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get samlMetadata() {
        return this.getStringAttribute('saml_metadata');
    }
    /**
     * @stability stable
     */
    set samlMetadata(value) {
        this._samlMetadata = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get samlMetadataInput() {
        return this._samlMetadata;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.WorklinkFleetIdentityProviderOutputReference = WorklinkFleetIdentityProviderOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WorklinkFleetIdentityProviderOutputReference[_a] = { fqn: "@cdktf/provider-aws.worklink.WorklinkFleetIdentityProviderOutputReference", version: "3.0.1" };
function worklinkFleetNetworkToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnetIds),
        vpc_id: cdktf.stringToTerraform(struct.vpcId),
    };
}
exports.worklinkFleetNetworkToTerraform = worklinkFleetNetworkToTerraform;
/**
 * @stability stable
 */
class WorklinkFleetNetworkOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        if (this._vpcId) {
            hasAnyValues = true;
            internalValueResult.vpcId = this._vpcId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
            this._vpcId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
            this._vpcId = value.vpcId;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
}
exports.WorklinkFleetNetworkOutputReference = WorklinkFleetNetworkOutputReference;
_b = JSII_RTTI_SYMBOL_1;
WorklinkFleetNetworkOutputReference[_b] = { fqn: "@cdktf/provider-aws.worklink.WorklinkFleetNetworkOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/worklink_fleet.html aws_worklink_fleet}.
 *
 * @stability stable
 */
class WorklinkFleet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/worklink_fleet.html aws_worklink_fleet} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_worklink_fleet',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // identity_provider - computed: false, optional: true, required: false
        this._identityProvider = new WorklinkFleetIdentityProviderOutputReference(this, "identity_provider", true);
        // network - computed: false, optional: true, required: false
        this._network = new WorklinkFleetNetworkOutputReference(this, "network", true);
        this._auditStreamArn = config.auditStreamArn;
        this._deviceCaCertificate = config.deviceCaCertificate;
        this._displayName = config.displayName;
        this._name = config.name;
        this._optimizeForEndUserLocation = config.optimizeForEndUserLocation;
        this._identityProvider.internalValue = config.identityProvider;
        this._network.internalValue = config.network;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get auditStreamArn() {
        return this.getStringAttribute('audit_stream_arn');
    }
    /**
     * @stability stable
     */
    set auditStreamArn(value) {
        this._auditStreamArn = value;
    }
    /**
     * @stability stable
     */
    resetAuditStreamArn() {
        this._auditStreamArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get auditStreamArnInput() {
        return this._auditStreamArn;
    }
    // company_code - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get companyCode() {
        return this.getStringAttribute('company_code');
    }
    // created_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    /**
     * @stability stable
     */
    get deviceCaCertificate() {
        return this.getStringAttribute('device_ca_certificate');
    }
    /**
     * @stability stable
     */
    set deviceCaCertificate(value) {
        this._deviceCaCertificate = value;
    }
    /**
     * @stability stable
     */
    resetDeviceCaCertificate() {
        this._deviceCaCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deviceCaCertificateInput() {
        return this._deviceCaCertificate;
    }
    /**
     * @stability stable
     */
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    /**
     * @stability stable
     */
    set displayName(value) {
        this._displayName = value;
    }
    /**
     * @stability stable
     */
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get displayNameInput() {
        return this._displayName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdatedTime() {
        return this.getStringAttribute('last_updated_time');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get optimizeForEndUserLocation() {
        return this.getBooleanAttribute('optimize_for_end_user_location');
    }
    /**
     * @stability stable
     */
    set optimizeForEndUserLocation(value) {
        this._optimizeForEndUserLocation = value;
    }
    /**
     * @stability stable
     */
    resetOptimizeForEndUserLocation() {
        this._optimizeForEndUserLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get optimizeForEndUserLocationInput() {
        return this._optimizeForEndUserLocation;
    }
    /**
     * @stability stable
     */
    get identityProvider() {
        return this._identityProvider;
    }
    /**
     * @stability stable
     */
    putIdentityProvider(value) {
        this._identityProvider.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIdentityProvider() {
        this._identityProvider.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get identityProviderInput() {
        return this._identityProvider.internalValue;
    }
    /**
     * @stability stable
     */
    get network() {
        return this._network;
    }
    /**
     * @stability stable
     */
    putNetwork(value) {
        this._network.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNetwork() {
        this._network.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkInput() {
        return this._network.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            audit_stream_arn: cdktf.stringToTerraform(this._auditStreamArn),
            device_ca_certificate: cdktf.stringToTerraform(this._deviceCaCertificate),
            display_name: cdktf.stringToTerraform(this._displayName),
            name: cdktf.stringToTerraform(this._name),
            optimize_for_end_user_location: cdktf.booleanToTerraform(this._optimizeForEndUserLocation),
            identity_provider: worklinkFleetIdentityProviderToTerraform(this._identityProvider.internalValue),
            network: worklinkFleetNetworkToTerraform(this._network.internalValue),
        };
    }
}
exports.WorklinkFleet = WorklinkFleet;
_c = JSII_RTTI_SYMBOL_1;
WorklinkFleet[_c] = { fqn: "@cdktf/provider-aws.worklink.WorklinkFleet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WorklinkFleet.tfResourceType = "aws_worklink_fleet";
