// @cdktf/provider-aws/lib/route53/route53-resolver-firewall-config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53ResolverFirewallConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_firewall_config.html aws_route53_resolver_firewall_config}.
 *
 * @stability stable
 */
class Route53ResolverFirewallConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_firewall_config.html aws_route53_resolver_firewall_config} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_resolver_firewall_config',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._firewallFailOpen = config.firewallFailOpen;
        this._resourceId = config.resourceId;
    }
    /**
     * @stability stable
     */
    get firewallFailOpen() {
        return this.getStringAttribute('firewall_fail_open');
    }
    /**
     * @stability stable
     */
    set firewallFailOpen(value) {
        this._firewallFailOpen = value;
    }
    /**
     * @stability stable
     */
    resetFirewallFailOpen() {
        this._firewallFailOpen = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get firewallFailOpenInput() {
        return this._firewallFailOpen;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    /**
     * @stability stable
     */
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceIdInput() {
        return this._resourceId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            firewall_fail_open: cdktf.stringToTerraform(this._firewallFailOpen),
            resource_id: cdktf.stringToTerraform(this._resourceId),
        };
    }
}
exports.Route53ResolverFirewallConfig = Route53ResolverFirewallConfig;
_a = JSII_RTTI_SYMBOL_1;
Route53ResolverFirewallConfig[_a] = { fqn: "@cdktf/provider-aws.route53.Route53ResolverFirewallConfig", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Route53ResolverFirewallConfig.tfResourceType = "aws_route53_resolver_firewall_config";
