// @cdktf/provider-aws/lib/route53/route53-zone-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53ZoneAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_zone_association.html aws_route53_zone_association}.
 *
 * @stability stable
 */
class Route53ZoneAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_zone_association.html aws_route53_zone_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_zone_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._vpcId = config.vpcId;
        this._vpcRegion = config.vpcRegion;
        this._zoneId = config.zoneId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // owning_account - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get owningAccount() {
        return this.getStringAttribute('owning_account');
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
    /**
     * @stability stable
     */
    get vpcRegion() {
        return this.getStringAttribute('vpc_region');
    }
    /**
     * @stability stable
     */
    set vpcRegion(value) {
        this._vpcRegion = value;
    }
    /**
     * @stability stable
     */
    resetVpcRegion() {
        this._vpcRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcRegionInput() {
        return this._vpcRegion;
    }
    /**
     * @stability stable
     */
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    /**
     * @stability stable
     */
    set zoneId(value) {
        this._zoneId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get zoneIdInput() {
        return this._zoneId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            vpc_region: cdktf.stringToTerraform(this._vpcRegion),
            zone_id: cdktf.stringToTerraform(this._zoneId),
        };
    }
}
exports.Route53ZoneAssociation = Route53ZoneAssociation;
_a = JSII_RTTI_SYMBOL_1;
Route53ZoneAssociation[_a] = { fqn: "@cdktf/provider-aws.route53.Route53ZoneAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Route53ZoneAssociation.tfResourceType = "aws_route53_zone_association";
