// @cdktf/provider-aws/lib/appmesh/appmesh-virtual-gateway.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppmeshVirtualGateway = exports.AppmeshVirtualGatewaySpecOutputReference = exports.appmeshVirtualGatewaySpecToTerraform = exports.AppmeshVirtualGatewaySpecLoggingOutputReference = exports.appmeshVirtualGatewaySpecLoggingToTerraform = exports.AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference = exports.appmeshVirtualGatewaySpecLoggingAccessLogToTerraform = exports.AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference = exports.appmeshVirtualGatewaySpecLoggingAccessLogFileToTerraform = exports.AppmeshVirtualGatewaySpecListenerOutputReference = exports.appmeshVirtualGatewaySpecListenerToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustSdsToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustFileToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsCertificateToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsCertificateSdsToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsCertificateFileToTerraform = exports.AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference = exports.appmeshVirtualGatewaySpecListenerTlsCertificateAcmToTerraform = exports.AppmeshVirtualGatewaySpecListenerPortMappingOutputReference = exports.appmeshVirtualGatewaySpecListenerPortMappingToTerraform = exports.AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference = exports.appmeshVirtualGatewaySpecListenerHealthCheckToTerraform = exports.AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference = exports.appmeshVirtualGatewaySpecListenerConnectionPoolToTerraform = exports.AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference = exports.appmeshVirtualGatewaySpecListenerConnectionPoolHttp2ToTerraform = exports.AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference = exports.appmeshVirtualGatewaySpecListenerConnectionPoolHttpToTerraform = exports.AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference = exports.appmeshVirtualGatewaySpecListenerConnectionPoolGrpcToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsToTerraform = exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference = exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_chain: cdktf.stringToTerraform(struct.certificateChain),
        private_key: cdktf.stringToTerraform(struct.privateKey),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateChain) {
            hasAnyValues = true;
            internalValueResult.certificateChain = this._certificateChain;
        }
        if (this._privateKey) {
            hasAnyValues = true;
            internalValueResult.privateKey = this._privateKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateChain = undefined;
            this._privateKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateChain = value.certificateChain;
            this._privateKey = value.privateKey;
        }
    }
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    /**
     * @stability stable
     */
    set certificateChain(value) {
        this._certificateChain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateChainInput() {
        return this._certificateChain;
    }
    /**
     * @stability stable
     */
    get privateKey() {
        return this.getStringAttribute('private_key');
    }
    /**
     * @stability stable
     */
    set privateKey(value) {
        this._privateKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateKeyInput() {
        return this._privateKey;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference[_a] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_name: cdktf.stringToTerraform(struct.secretName),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretName) {
            hasAnyValues = true;
            internalValueResult.secretName = this._secretName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretName = value.secretName;
        }
    }
    /**
     * @stability stable
     */
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
    /**
     * @stability stable
     */
    set secretName(value) {
        this._secretName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretNameInput() {
        return this._secretName;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference[_b] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileToTerraform(struct.file),
        sds: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsToTerraform(struct.sds),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // file - computed: false, optional: true, required: false
        this._file = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateFileOutputReference(this, "file", true);
        // sds - computed: false, optional: true, required: false
        this._sds = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateSdsOutputReference(this, "sds", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._file) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.file = (_13 = this._file) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._sds) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.sds = (_15 = this._sds) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._file.internalValue = undefined;
            this._sds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._file.internalValue = value.file;
            this._sds.internalValue = value.sds;
        }
    }
    /**
     * @stability stable
     */
    get file() {
        return this._file;
    }
    /**
     * @stability stable
     */
    putFile(value) {
        this._file.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFile() {
        this._file.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileInput() {
        return this._file.internalValue;
    }
    /**
     * @stability stable
     */
    get sds() {
        return this._sds;
    }
    /**
     * @stability stable
     */
    putSds(value) {
        this._sds.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSds() {
        this._sds.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sdsInput() {
        return this._sds.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference[_c] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.listMapper(cdktf.stringToTerraform)(struct.exact),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
        }
    }
    /**
     * @stability stable
     */
    get exact() {
        return this.getListAttribute('exact');
    }
    /**
     * @stability stable
     */
    set exact(value) {
        this._exact = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get exactInput() {
        return this._exact;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference[_d] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        match: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchToTerraform(struct.match),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesMatchOutputReference(this, "match", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._match) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.match = (_13 = this._match) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._match.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._match.internalValue = value.match;
        }
    }
    /**
     * @stability stable
     */
    get match() {
        return this._match;
    }
    /**
     * @stability stable
     */
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference[_e] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_authority_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.certificateAuthorityArns),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateAuthorityArns) {
            hasAnyValues = true;
            internalValueResult.certificateAuthorityArns = this._certificateAuthorityArns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateAuthorityArns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateAuthorityArns = value.certificateAuthorityArns;
        }
    }
    /**
     * @stability stable
     */
    get certificateAuthorityArns() {
        return this.getListAttribute('certificate_authority_arns');
    }
    /**
     * @stability stable
     */
    set certificateAuthorityArns(value) {
        this._certificateAuthorityArns = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateAuthorityArnsInput() {
        return this._certificateAuthorityArns;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference[_f] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_chain: cdktf.stringToTerraform(struct.certificateChain),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateChain) {
            hasAnyValues = true;
            internalValueResult.certificateChain = this._certificateChain;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateChain = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateChain = value.certificateChain;
        }
    }
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    /**
     * @stability stable
     */
    set certificateChain(value) {
        this._certificateChain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateChainInput() {
        return this._certificateChain;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference;
_g = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference[_g] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_name: cdktf.stringToTerraform(struct.secretName),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretName) {
            hasAnyValues = true;
            internalValueResult.secretName = this._secretName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretName = value.secretName;
        }
    }
    /**
     * @stability stable
     */
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
    /**
     * @stability stable
     */
    set secretName(value) {
        this._secretName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretNameInput() {
        return this._secretName;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference[_h] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        acm: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmToTerraform(struct.acm),
        file: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileToTerraform(struct.file),
        sds: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsToTerraform(struct.sds),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // acm - computed: false, optional: true, required: false
        this._acm = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustAcmOutputReference(this, "acm", true);
        // file - computed: false, optional: true, required: false
        this._file = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustFileOutputReference(this, "file", true);
        // sds - computed: false, optional: true, required: false
        this._sds = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustSdsOutputReference(this, "sds", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15, _16, _17;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._acm) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.acm = (_13 = this._acm) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._file) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.file = (_15 = this._file) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        if ((_16 = this._sds) === null || _16 === void 0 ? void 0 : _16.internalValue) {
            hasAnyValues = true;
            internalValueResult.sds = (_17 = this._sds) === null || _17 === void 0 ? void 0 : _17.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._acm.internalValue = undefined;
            this._file.internalValue = undefined;
            this._sds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._acm.internalValue = value.acm;
            this._file.internalValue = value.file;
            this._sds.internalValue = value.sds;
        }
    }
    /**
     * @stability stable
     */
    get acm() {
        return this._acm;
    }
    /**
     * @stability stable
     */
    putAcm(value) {
        this._acm.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAcm() {
        this._acm.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acmInput() {
        return this._acm.internalValue;
    }
    /**
     * @stability stable
     */
    get file() {
        return this._file;
    }
    /**
     * @stability stable
     */
    putFile(value) {
        this._file.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFile() {
        this._file.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileInput() {
        return this._file.internalValue;
    }
    /**
     * @stability stable
     */
    get sds() {
        return this._sds;
    }
    /**
     * @stability stable
     */
    putSds(value) {
        this._sds.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSds() {
        this._sds.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sdsInput() {
        return this._sds.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference;
_j = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference[_j] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subject_alternative_names: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesToTerraform(struct.subjectAlternativeNames),
        trust: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustToTerraform(struct.trust),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // subject_alternative_names - computed: false, optional: true, required: false
        this._subjectAlternativeNames = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationSubjectAlternativeNamesOutputReference(this, "subject_alternative_names", true);
        // trust - computed: false, optional: false, required: true
        this._trust = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationTrustOutputReference(this, "trust", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._subjectAlternativeNames) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.subjectAlternativeNames = (_13 = this._subjectAlternativeNames) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._trust) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.trust = (_15 = this._trust) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._subjectAlternativeNames.internalValue = undefined;
            this._trust.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._subjectAlternativeNames.internalValue = value.subjectAlternativeNames;
            this._trust.internalValue = value.trust;
        }
    }
    /**
     * @stability stable
     */
    get subjectAlternativeNames() {
        return this._subjectAlternativeNames;
    }
    /**
     * @stability stable
     */
    putSubjectAlternativeNames(value) {
        this._subjectAlternativeNames.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSubjectAlternativeNames() {
        this._subjectAlternativeNames.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subjectAlternativeNamesInput() {
        return this._subjectAlternativeNames.internalValue;
    }
    /**
     * @stability stable
     */
    get trust() {
        return this._trust;
    }
    /**
     * @stability stable
     */
    putTrust(value) {
        this._trust.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get trustInput() {
        return this._trust.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference;
_k = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference[_k] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enforce: cdktf.booleanToTerraform(struct.enforce),
        ports: cdktf.listMapper(cdktf.numberToTerraform)(struct.ports),
        certificate: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateToTerraform(struct.certificate),
        validation: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationToTerraform(struct.validation),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // certificate - computed: false, optional: true, required: false
        this._certificate = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsCertificateOutputReference(this, "certificate", true);
        // validation - computed: false, optional: false, required: true
        this._validation = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsValidationOutputReference(this, "validation", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enforce) {
            hasAnyValues = true;
            internalValueResult.enforce = this._enforce;
        }
        if (this._ports) {
            hasAnyValues = true;
            internalValueResult.ports = this._ports;
        }
        if ((_12 = this._certificate) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.certificate = (_13 = this._certificate) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._validation) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.validation = (_15 = this._validation) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enforce = undefined;
            this._ports = undefined;
            this._certificate.internalValue = undefined;
            this._validation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enforce = value.enforce;
            this._ports = value.ports;
            this._certificate.internalValue = value.certificate;
            this._validation.internalValue = value.validation;
        }
    }
    /**
     * @stability stable
     */
    get enforce() {
        return this.getBooleanAttribute('enforce');
    }
    /**
     * @stability stable
     */
    set enforce(value) {
        this._enforce = value;
    }
    /**
     * @stability stable
     */
    resetEnforce() {
        this._enforce = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enforceInput() {
        return this._enforce;
    }
    /**
     * @stability stable
     */
    get ports() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ports');
    }
    /**
     * @stability stable
     */
    set ports(value) {
        this._ports = value;
    }
    /**
     * @stability stable
     */
    resetPorts() {
        this._ports = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portsInput() {
        return this._ports;
    }
    /**
     * @stability stable
     */
    get certificate() {
        return this._certificate;
    }
    /**
     * @stability stable
     */
    putCertificate(value) {
        this._certificate.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCertificate() {
        this._certificate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateInput() {
        return this._certificate.internalValue;
    }
    /**
     * @stability stable
     */
    get validation() {
        return this._validation;
    }
    /**
     * @stability stable
     */
    putValidation(value) {
        this._validation.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get validationInput() {
        return this._validation.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference[_l] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsClientPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        tls: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsToTerraform(struct.tls),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsClientPolicyToTerraform = appmeshVirtualGatewaySpecBackendDefaultsClientPolicyToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // tls - computed: false, optional: true, required: false
        this._tls = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyTlsOutputReference(this, "tls", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._tls) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.tls = (_13 = this._tls) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tls.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tls.internalValue = value.tls;
        }
    }
    /**
     * @stability stable
     */
    get tls() {
        return this._tls;
    }
    /**
     * @stability stable
     */
    putTls(value) {
        this._tls.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTls() {
        this._tls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsInput() {
        return this._tls.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference;
_m = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference[_m] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecBackendDefaultsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_policy: appmeshVirtualGatewaySpecBackendDefaultsClientPolicyToTerraform(struct.clientPolicy),
    };
}
exports.appmeshVirtualGatewaySpecBackendDefaultsToTerraform = appmeshVirtualGatewaySpecBackendDefaultsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecBackendDefaultsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // client_policy - computed: false, optional: true, required: false
        this._clientPolicy = new AppmeshVirtualGatewaySpecBackendDefaultsClientPolicyOutputReference(this, "client_policy", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._clientPolicy) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.clientPolicy = (_13 = this._clientPolicy) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientPolicy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientPolicy.internalValue = value.clientPolicy;
        }
    }
    /**
     * @stability stable
     */
    get clientPolicy() {
        return this._clientPolicy;
    }
    /**
     * @stability stable
     */
    putClientPolicy(value) {
        this._clientPolicy.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetClientPolicy() {
        this._clientPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientPolicyInput() {
        return this._clientPolicy.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecBackendDefaultsOutputReference = AppmeshVirtualGatewaySpecBackendDefaultsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecBackendDefaultsOutputReference[_o] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecBackendDefaultsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerConnectionPoolGrpcToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_requests: cdktf.numberToTerraform(struct.maxRequests),
    };
}
exports.appmeshVirtualGatewaySpecListenerConnectionPoolGrpcToTerraform = appmeshVirtualGatewaySpecListenerConnectionPoolGrpcToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxRequests) {
            hasAnyValues = true;
            internalValueResult.maxRequests = this._maxRequests;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxRequests = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxRequests = value.maxRequests;
        }
    }
    /**
     * @stability stable
     */
    get maxRequests() {
        return this.getNumberAttribute('max_requests');
    }
    /**
     * @stability stable
     */
    set maxRequests(value) {
        this._maxRequests = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxRequestsInput() {
        return this._maxRequests;
    }
}
exports.AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference = AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference;
_p = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference[_p] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerConnectionPoolHttpToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_connections: cdktf.numberToTerraform(struct.maxConnections),
        max_pending_requests: cdktf.numberToTerraform(struct.maxPendingRequests),
    };
}
exports.appmeshVirtualGatewaySpecListenerConnectionPoolHttpToTerraform = appmeshVirtualGatewaySpecListenerConnectionPoolHttpToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConnections) {
            hasAnyValues = true;
            internalValueResult.maxConnections = this._maxConnections;
        }
        if (this._maxPendingRequests) {
            hasAnyValues = true;
            internalValueResult.maxPendingRequests = this._maxPendingRequests;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConnections = undefined;
            this._maxPendingRequests = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConnections = value.maxConnections;
            this._maxPendingRequests = value.maxPendingRequests;
        }
    }
    /**
     * @stability stable
     */
    get maxConnections() {
        return this.getNumberAttribute('max_connections');
    }
    /**
     * @stability stable
     */
    set maxConnections(value) {
        this._maxConnections = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConnectionsInput() {
        return this._maxConnections;
    }
    /**
     * @stability stable
     */
    get maxPendingRequests() {
        return this.getNumberAttribute('max_pending_requests');
    }
    /**
     * @stability stable
     */
    set maxPendingRequests(value) {
        this._maxPendingRequests = value;
    }
    /**
     * @stability stable
     */
    resetMaxPendingRequests() {
        this._maxPendingRequests = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxPendingRequestsInput() {
        return this._maxPendingRequests;
    }
}
exports.AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference = AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference;
_q = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference[_q] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerConnectionPoolHttp2ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_requests: cdktf.numberToTerraform(struct.maxRequests),
    };
}
exports.appmeshVirtualGatewaySpecListenerConnectionPoolHttp2ToTerraform = appmeshVirtualGatewaySpecListenerConnectionPoolHttp2ToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxRequests) {
            hasAnyValues = true;
            internalValueResult.maxRequests = this._maxRequests;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxRequests = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxRequests = value.maxRequests;
        }
    }
    /**
     * @stability stable
     */
    get maxRequests() {
        return this.getNumberAttribute('max_requests');
    }
    /**
     * @stability stable
     */
    set maxRequests(value) {
        this._maxRequests = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxRequestsInput() {
        return this._maxRequests;
    }
}
exports.AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference = AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference;
_r = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference[_r] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerConnectionPoolToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grpc: appmeshVirtualGatewaySpecListenerConnectionPoolGrpcToTerraform(struct.grpc),
        http: appmeshVirtualGatewaySpecListenerConnectionPoolHttpToTerraform(struct.http),
        http2: appmeshVirtualGatewaySpecListenerConnectionPoolHttp2ToTerraform(struct.http2),
    };
}
exports.appmeshVirtualGatewaySpecListenerConnectionPoolToTerraform = appmeshVirtualGatewaySpecListenerConnectionPoolToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // grpc - computed: false, optional: true, required: false
        this._grpc = new AppmeshVirtualGatewaySpecListenerConnectionPoolGrpcOutputReference(this, "grpc", true);
        // http - computed: false, optional: true, required: false
        this._http = new AppmeshVirtualGatewaySpecListenerConnectionPoolHttpOutputReference(this, "http", true);
        // http2 - computed: false, optional: true, required: false
        this._http2 = new AppmeshVirtualGatewaySpecListenerConnectionPoolHttp2OutputReference(this, "http2", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15, _16, _17;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._grpc) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.grpc = (_13 = this._grpc) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._http) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.http = (_15 = this._http) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        if ((_16 = this._http2) === null || _16 === void 0 ? void 0 : _16.internalValue) {
            hasAnyValues = true;
            internalValueResult.http2 = (_17 = this._http2) === null || _17 === void 0 ? void 0 : _17.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._grpc.internalValue = undefined;
            this._http.internalValue = undefined;
            this._http2.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._grpc.internalValue = value.grpc;
            this._http.internalValue = value.http;
            this._http2.internalValue = value.http2;
        }
    }
    /**
     * @stability stable
     */
    get grpc() {
        return this._grpc;
    }
    /**
     * @stability stable
     */
    putGrpc(value) {
        this._grpc.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetGrpc() {
        this._grpc.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get grpcInput() {
        return this._grpc.internalValue;
    }
    /**
     * @stability stable
     */
    get http() {
        return this._http;
    }
    /**
     * @stability stable
     */
    putHttp(value) {
        this._http.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetHttp() {
        this._http.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpInput() {
        return this._http.internalValue;
    }
    /**
     * @stability stable
     */
    get http2() {
        return this._http2;
    }
    /**
     * @stability stable
     */
    putHttp2(value) {
        this._http2.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetHttp2() {
        this._http2.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get http2Input() {
        return this._http2.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference = AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference;
_s = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference[_s] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        healthy_threshold: cdktf.numberToTerraform(struct.healthyThreshold),
        interval_millis: cdktf.numberToTerraform(struct.intervalMillis),
        path: cdktf.stringToTerraform(struct.path),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        timeout_millis: cdktf.numberToTerraform(struct.timeoutMillis),
        unhealthy_threshold: cdktf.numberToTerraform(struct.unhealthyThreshold),
    };
}
exports.appmeshVirtualGatewaySpecListenerHealthCheckToTerraform = appmeshVirtualGatewaySpecListenerHealthCheckToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthyThreshold) {
            hasAnyValues = true;
            internalValueResult.healthyThreshold = this._healthyThreshold;
        }
        if (this._intervalMillis) {
            hasAnyValues = true;
            internalValueResult.intervalMillis = this._intervalMillis;
        }
        if (this._path) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._port) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._timeoutMillis) {
            hasAnyValues = true;
            internalValueResult.timeoutMillis = this._timeoutMillis;
        }
        if (this._unhealthyThreshold) {
            hasAnyValues = true;
            internalValueResult.unhealthyThreshold = this._unhealthyThreshold;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._healthyThreshold = undefined;
            this._intervalMillis = undefined;
            this._path = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._timeoutMillis = undefined;
            this._unhealthyThreshold = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._healthyThreshold = value.healthyThreshold;
            this._intervalMillis = value.intervalMillis;
            this._path = value.path;
            this._port = value.port;
            this._protocol = value.protocol;
            this._timeoutMillis = value.timeoutMillis;
            this._unhealthyThreshold = value.unhealthyThreshold;
        }
    }
    /**
     * @stability stable
     */
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    /**
     * @stability stable
     */
    set healthyThreshold(value) {
        this._healthyThreshold = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthyThresholdInput() {
        return this._healthyThreshold;
    }
    /**
     * @stability stable
     */
    get intervalMillis() {
        return this.getNumberAttribute('interval_millis');
    }
    /**
     * @stability stable
     */
    set intervalMillis(value) {
        this._intervalMillis = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get intervalMillisInput() {
        return this._intervalMillis;
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    /**
     * @stability stable
     */
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    /**
     * @stability stable
     */
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    /**
     * @stability stable
     */
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    /**
     * @stability stable
     */
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get protocolInput() {
        return this._protocol;
    }
    /**
     * @stability stable
     */
    get timeoutMillis() {
        return this.getNumberAttribute('timeout_millis');
    }
    /**
     * @stability stable
     */
    set timeoutMillis(value) {
        this._timeoutMillis = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutMillisInput() {
        return this._timeoutMillis;
    }
    /**
     * @stability stable
     */
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
    /**
     * @stability stable
     */
    set unhealthyThreshold(value) {
        this._unhealthyThreshold = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get unhealthyThresholdInput() {
        return this._unhealthyThreshold;
    }
}
exports.AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference = AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference;
_t = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference[_t] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerPortMappingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
    };
}
exports.appmeshVirtualGatewaySpecListenerPortMappingToTerraform = appmeshVirtualGatewaySpecListenerPortMappingToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerPortMappingOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
            this._protocol = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
            this._protocol = value.protocol;
        }
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    /**
     * @stability stable
     */
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    /**
     * @stability stable
     */
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get protocolInput() {
        return this._protocol;
    }
}
exports.AppmeshVirtualGatewaySpecListenerPortMappingOutputReference = AppmeshVirtualGatewaySpecListenerPortMappingOutputReference;
_u = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerPortMappingOutputReference[_u] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerPortMappingOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsCertificateAcmToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_arn: cdktf.stringToTerraform(struct.certificateArn),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsCertificateAcmToTerraform = appmeshVirtualGatewaySpecListenerTlsCertificateAcmToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateArn) {
            hasAnyValues = true;
            internalValueResult.certificateArn = this._certificateArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateArn = value.certificateArn;
        }
    }
    /**
     * @stability stable
     */
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    /**
     * @stability stable
     */
    set certificateArn(value) {
        this._certificateArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateArnInput() {
        return this._certificateArn;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference = AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference;
_v = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference[_v] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsCertificateFileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_chain: cdktf.stringToTerraform(struct.certificateChain),
        private_key: cdktf.stringToTerraform(struct.privateKey),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsCertificateFileToTerraform = appmeshVirtualGatewaySpecListenerTlsCertificateFileToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateChain) {
            hasAnyValues = true;
            internalValueResult.certificateChain = this._certificateChain;
        }
        if (this._privateKey) {
            hasAnyValues = true;
            internalValueResult.privateKey = this._privateKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateChain = undefined;
            this._privateKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateChain = value.certificateChain;
            this._privateKey = value.privateKey;
        }
    }
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    /**
     * @stability stable
     */
    set certificateChain(value) {
        this._certificateChain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateChainInput() {
        return this._certificateChain;
    }
    /**
     * @stability stable
     */
    get privateKey() {
        return this.getStringAttribute('private_key');
    }
    /**
     * @stability stable
     */
    set privateKey(value) {
        this._privateKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateKeyInput() {
        return this._privateKey;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference = AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference;
_w = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference[_w] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsCertificateSdsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_name: cdktf.stringToTerraform(struct.secretName),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsCertificateSdsToTerraform = appmeshVirtualGatewaySpecListenerTlsCertificateSdsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretName) {
            hasAnyValues = true;
            internalValueResult.secretName = this._secretName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretName = value.secretName;
        }
    }
    /**
     * @stability stable
     */
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
    /**
     * @stability stable
     */
    set secretName(value) {
        this._secretName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretNameInput() {
        return this._secretName;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference = AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference[_x] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsCertificateToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        acm: appmeshVirtualGatewaySpecListenerTlsCertificateAcmToTerraform(struct.acm),
        file: appmeshVirtualGatewaySpecListenerTlsCertificateFileToTerraform(struct.file),
        sds: appmeshVirtualGatewaySpecListenerTlsCertificateSdsToTerraform(struct.sds),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsCertificateToTerraform = appmeshVirtualGatewaySpecListenerTlsCertificateToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // acm - computed: false, optional: true, required: false
        this._acm = new AppmeshVirtualGatewaySpecListenerTlsCertificateAcmOutputReference(this, "acm", true);
        // file - computed: false, optional: true, required: false
        this._file = new AppmeshVirtualGatewaySpecListenerTlsCertificateFileOutputReference(this, "file", true);
        // sds - computed: false, optional: true, required: false
        this._sds = new AppmeshVirtualGatewaySpecListenerTlsCertificateSdsOutputReference(this, "sds", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15, _16, _17;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._acm) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.acm = (_13 = this._acm) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._file) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.file = (_15 = this._file) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        if ((_16 = this._sds) === null || _16 === void 0 ? void 0 : _16.internalValue) {
            hasAnyValues = true;
            internalValueResult.sds = (_17 = this._sds) === null || _17 === void 0 ? void 0 : _17.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._acm.internalValue = undefined;
            this._file.internalValue = undefined;
            this._sds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._acm.internalValue = value.acm;
            this._file.internalValue = value.file;
            this._sds.internalValue = value.sds;
        }
    }
    /**
     * @stability stable
     */
    get acm() {
        return this._acm;
    }
    /**
     * @stability stable
     */
    putAcm(value) {
        this._acm.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAcm() {
        this._acm.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acmInput() {
        return this._acm.internalValue;
    }
    /**
     * @stability stable
     */
    get file() {
        return this._file;
    }
    /**
     * @stability stable
     */
    putFile(value) {
        this._file.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFile() {
        this._file.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileInput() {
        return this._file.internalValue;
    }
    /**
     * @stability stable
     */
    get sds() {
        return this._sds;
    }
    /**
     * @stability stable
     */
    putSds(value) {
        this._sds.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSds() {
        this._sds.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sdsInput() {
        return this._sds.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference = AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference;
_y = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference[_y] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.listMapper(cdktf.stringToTerraform)(struct.exact),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
        }
    }
    /**
     * @stability stable
     */
    get exact() {
        return this.getListAttribute('exact');
    }
    /**
     * @stability stable
     */
    set exact(value) {
        this._exact = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get exactInput() {
        return this._exact;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference;
_z = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference[_z] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        match: appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchToTerraform(struct.match),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesMatchOutputReference(this, "match", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._match) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.match = (_13 = this._match) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._match.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._match.internalValue = value.match;
        }
    }
    /**
     * @stability stable
     */
    get match() {
        return this._match;
    }
    /**
     * @stability stable
     */
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference[_0] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationTrustFileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_chain: cdktf.stringToTerraform(struct.certificateChain),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustFileToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationTrustFileToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateChain) {
            hasAnyValues = true;
            internalValueResult.certificateChain = this._certificateChain;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateChain = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateChain = value.certificateChain;
        }
    }
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    /**
     * @stability stable
     */
    set certificateChain(value) {
        this._certificateChain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateChainInput() {
        return this._certificateChain;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference[_1] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationTrustSdsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_name: cdktf.stringToTerraform(struct.secretName),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustSdsToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationTrustSdsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretName) {
            hasAnyValues = true;
            internalValueResult.secretName = this._secretName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretName = value.secretName;
        }
    }
    /**
     * @stability stable
     */
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
    /**
     * @stability stable
     */
    set secretName(value) {
        this._secretName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretNameInput() {
        return this._secretName;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference[_2] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationTrustToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file: appmeshVirtualGatewaySpecListenerTlsValidationTrustFileToTerraform(struct.file),
        sds: appmeshVirtualGatewaySpecListenerTlsValidationTrustSdsToTerraform(struct.sds),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationTrustToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationTrustToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // file - computed: false, optional: true, required: false
        this._file = new AppmeshVirtualGatewaySpecListenerTlsValidationTrustFileOutputReference(this, "file", true);
        // sds - computed: false, optional: true, required: false
        this._sds = new AppmeshVirtualGatewaySpecListenerTlsValidationTrustSdsOutputReference(this, "sds", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._file) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.file = (_13 = this._file) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._sds) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.sds = (_15 = this._sds) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._file.internalValue = undefined;
            this._sds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._file.internalValue = value.file;
            this._sds.internalValue = value.sds;
        }
    }
    /**
     * @stability stable
     */
    get file() {
        return this._file;
    }
    /**
     * @stability stable
     */
    putFile(value) {
        this._file.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFile() {
        this._file.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileInput() {
        return this._file.internalValue;
    }
    /**
     * @stability stable
     */
    get sds() {
        return this._sds;
    }
    /**
     * @stability stable
     */
    putSds(value) {
        this._sds.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSds() {
        this._sds.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sdsInput() {
        return this._sds.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference[_3] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsValidationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subject_alternative_names: appmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesToTerraform(struct.subjectAlternativeNames),
        trust: appmeshVirtualGatewaySpecListenerTlsValidationTrustToTerraform(struct.trust),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsValidationToTerraform = appmeshVirtualGatewaySpecListenerTlsValidationToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // subject_alternative_names - computed: false, optional: true, required: false
        this._subjectAlternativeNames = new AppmeshVirtualGatewaySpecListenerTlsValidationSubjectAlternativeNamesOutputReference(this, "subject_alternative_names", true);
        // trust - computed: false, optional: false, required: true
        this._trust = new AppmeshVirtualGatewaySpecListenerTlsValidationTrustOutputReference(this, "trust", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._subjectAlternativeNames) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.subjectAlternativeNames = (_13 = this._subjectAlternativeNames) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._trust) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.trust = (_15 = this._trust) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._subjectAlternativeNames.internalValue = undefined;
            this._trust.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._subjectAlternativeNames.internalValue = value.subjectAlternativeNames;
            this._trust.internalValue = value.trust;
        }
    }
    /**
     * @stability stable
     */
    get subjectAlternativeNames() {
        return this._subjectAlternativeNames;
    }
    /**
     * @stability stable
     */
    putSubjectAlternativeNames(value) {
        this._subjectAlternativeNames.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSubjectAlternativeNames() {
        this._subjectAlternativeNames.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subjectAlternativeNamesInput() {
        return this._subjectAlternativeNames.internalValue;
    }
    /**
     * @stability stable
     */
    get trust() {
        return this._trust;
    }
    /**
     * @stability stable
     */
    putTrust(value) {
        this._trust.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get trustInput() {
        return this._trust.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference = AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference[_4] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerTlsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
        certificate: appmeshVirtualGatewaySpecListenerTlsCertificateToTerraform(struct.certificate),
        validation: appmeshVirtualGatewaySpecListenerTlsValidationToTerraform(struct.validation),
    };
}
exports.appmeshVirtualGatewaySpecListenerTlsToTerraform = appmeshVirtualGatewaySpecListenerTlsToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerTlsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // certificate - computed: false, optional: false, required: true
        this._certificate = new AppmeshVirtualGatewaySpecListenerTlsCertificateOutputReference(this, "certificate", true);
        // validation - computed: false, optional: true, required: false
        this._validation = new AppmeshVirtualGatewaySpecListenerTlsValidationOutputReference(this, "validation", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if ((_12 = this._certificate) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.certificate = (_13 = this._certificate) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._validation) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.validation = (_15 = this._validation) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
            this._certificate.internalValue = undefined;
            this._validation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
            this._certificate.internalValue = value.certificate;
            this._validation.internalValue = value.validation;
        }
    }
    /**
     * @stability stable
     */
    get mode() {
        return this.getStringAttribute('mode');
    }
    /**
     * @stability stable
     */
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modeInput() {
        return this._mode;
    }
    /**
     * @stability stable
     */
    get certificate() {
        return this._certificate;
    }
    /**
     * @stability stable
     */
    putCertificate(value) {
        this._certificate.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateInput() {
        return this._certificate.internalValue;
    }
    /**
     * @stability stable
     */
    get validation() {
        return this._validation;
    }
    /**
     * @stability stable
     */
    putValidation(value) {
        this._validation.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetValidation() {
        this._validation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get validationInput() {
        return this._validation.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerTlsOutputReference = AppmeshVirtualGatewaySpecListenerTlsOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerTlsOutputReference[_5] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerTlsOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecListenerToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_pool: appmeshVirtualGatewaySpecListenerConnectionPoolToTerraform(struct.connectionPool),
        health_check: appmeshVirtualGatewaySpecListenerHealthCheckToTerraform(struct.healthCheck),
        port_mapping: appmeshVirtualGatewaySpecListenerPortMappingToTerraform(struct.portMapping),
        tls: appmeshVirtualGatewaySpecListenerTlsToTerraform(struct.tls),
    };
}
exports.appmeshVirtualGatewaySpecListenerToTerraform = appmeshVirtualGatewaySpecListenerToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecListenerOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // connection_pool - computed: false, optional: true, required: false
        this._connectionPool = new AppmeshVirtualGatewaySpecListenerConnectionPoolOutputReference(this, "connection_pool", true);
        // health_check - computed: false, optional: true, required: false
        this._healthCheck = new AppmeshVirtualGatewaySpecListenerHealthCheckOutputReference(this, "health_check", true);
        // port_mapping - computed: false, optional: false, required: true
        this._portMapping = new AppmeshVirtualGatewaySpecListenerPortMappingOutputReference(this, "port_mapping", true);
        // tls - computed: false, optional: true, required: false
        this._tls = new AppmeshVirtualGatewaySpecListenerTlsOutputReference(this, "tls", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15, _16, _17, _18, _19;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._connectionPool) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.connectionPool = (_13 = this._connectionPool) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._healthCheck) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.healthCheck = (_15 = this._healthCheck) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        if ((_16 = this._portMapping) === null || _16 === void 0 ? void 0 : _16.internalValue) {
            hasAnyValues = true;
            internalValueResult.portMapping = (_17 = this._portMapping) === null || _17 === void 0 ? void 0 : _17.internalValue;
        }
        if ((_18 = this._tls) === null || _18 === void 0 ? void 0 : _18.internalValue) {
            hasAnyValues = true;
            internalValueResult.tls = (_19 = this._tls) === null || _19 === void 0 ? void 0 : _19.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connectionPool.internalValue = undefined;
            this._healthCheck.internalValue = undefined;
            this._portMapping.internalValue = undefined;
            this._tls.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connectionPool.internalValue = value.connectionPool;
            this._healthCheck.internalValue = value.healthCheck;
            this._portMapping.internalValue = value.portMapping;
            this._tls.internalValue = value.tls;
        }
    }
    /**
     * @stability stable
     */
    get connectionPool() {
        return this._connectionPool;
    }
    /**
     * @stability stable
     */
    putConnectionPool(value) {
        this._connectionPool.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConnectionPool() {
        this._connectionPool.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionPoolInput() {
        return this._connectionPool.internalValue;
    }
    /**
     * @stability stable
     */
    get healthCheck() {
        return this._healthCheck;
    }
    /**
     * @stability stable
     */
    putHealthCheck(value) {
        this._healthCheck.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheck() {
        this._healthCheck.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckInput() {
        return this._healthCheck.internalValue;
    }
    /**
     * @stability stable
     */
    get portMapping() {
        return this._portMapping;
    }
    /**
     * @stability stable
     */
    putPortMapping(value) {
        this._portMapping.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portMappingInput() {
        return this._portMapping.internalValue;
    }
    /**
     * @stability stable
     */
    get tls() {
        return this._tls;
    }
    /**
     * @stability stable
     */
    putTls(value) {
        this._tls.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTls() {
        this._tls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsInput() {
        return this._tls.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecListenerOutputReference = AppmeshVirtualGatewaySpecListenerOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecListenerOutputReference[_6] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecListenerOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecLoggingAccessLogFileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.appmeshVirtualGatewaySpecLoggingAccessLogFileToTerraform = appmeshVirtualGatewaySpecLoggingAccessLogFileToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
        }
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
}
exports.AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference = AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference[_7] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecLoggingAccessLogToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file: appmeshVirtualGatewaySpecLoggingAccessLogFileToTerraform(struct.file),
    };
}
exports.appmeshVirtualGatewaySpecLoggingAccessLogToTerraform = appmeshVirtualGatewaySpecLoggingAccessLogToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // file - computed: false, optional: true, required: false
        this._file = new AppmeshVirtualGatewaySpecLoggingAccessLogFileOutputReference(this, "file", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._file) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.file = (_13 = this._file) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._file.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._file.internalValue = value.file;
        }
    }
    /**
     * @stability stable
     */
    get file() {
        return this._file;
    }
    /**
     * @stability stable
     */
    putFile(value) {
        this._file.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFile() {
        this._file.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileInput() {
        return this._file.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference = AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference[_8] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecLoggingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_log: appmeshVirtualGatewaySpecLoggingAccessLogToTerraform(struct.accessLog),
    };
}
exports.appmeshVirtualGatewaySpecLoggingToTerraform = appmeshVirtualGatewaySpecLoggingToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecLoggingOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // access_log - computed: false, optional: true, required: false
        this._accessLog = new AppmeshVirtualGatewaySpecLoggingAccessLogOutputReference(this, "access_log", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._accessLog) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.accessLog = (_13 = this._accessLog) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accessLog.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accessLog.internalValue = value.accessLog;
        }
    }
    /**
     * @stability stable
     */
    get accessLog() {
        return this._accessLog;
    }
    /**
     * @stability stable
     */
    putAccessLog(value) {
        this._accessLog.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAccessLog() {
        this._accessLog.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessLogInput() {
        return this._accessLog.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecLoggingOutputReference = AppmeshVirtualGatewaySpecLoggingOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecLoggingOutputReference[_9] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecLoggingOutputReference", version: "3.0.1" };
function appmeshVirtualGatewaySpecToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backend_defaults: appmeshVirtualGatewaySpecBackendDefaultsToTerraform(struct.backendDefaults),
        listener: appmeshVirtualGatewaySpecListenerToTerraform(struct.listener),
        logging: appmeshVirtualGatewaySpecLoggingToTerraform(struct.logging),
    };
}
exports.appmeshVirtualGatewaySpecToTerraform = appmeshVirtualGatewaySpecToTerraform;
/**
 * @stability stable
 */
class AppmeshVirtualGatewaySpecOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // backend_defaults - computed: false, optional: true, required: false
        this._backendDefaults = new AppmeshVirtualGatewaySpecBackendDefaultsOutputReference(this, "backend_defaults", true);
        // listener - computed: false, optional: false, required: true
        this._listener = new AppmeshVirtualGatewaySpecListenerOutputReference(this, "listener", true);
        // logging - computed: false, optional: true, required: false
        this._logging = new AppmeshVirtualGatewaySpecLoggingOutputReference(this, "logging", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _12, _13, _14, _15, _16, _17;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_12 = this._backendDefaults) === null || _12 === void 0 ? void 0 : _12.internalValue) {
            hasAnyValues = true;
            internalValueResult.backendDefaults = (_13 = this._backendDefaults) === null || _13 === void 0 ? void 0 : _13.internalValue;
        }
        if ((_14 = this._listener) === null || _14 === void 0 ? void 0 : _14.internalValue) {
            hasAnyValues = true;
            internalValueResult.listener = (_15 = this._listener) === null || _15 === void 0 ? void 0 : _15.internalValue;
        }
        if ((_16 = this._logging) === null || _16 === void 0 ? void 0 : _16.internalValue) {
            hasAnyValues = true;
            internalValueResult.logging = (_17 = this._logging) === null || _17 === void 0 ? void 0 : _17.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._backendDefaults.internalValue = undefined;
            this._listener.internalValue = undefined;
            this._logging.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._backendDefaults.internalValue = value.backendDefaults;
            this._listener.internalValue = value.listener;
            this._logging.internalValue = value.logging;
        }
    }
    /**
     * @stability stable
     */
    get backendDefaults() {
        return this._backendDefaults;
    }
    /**
     * @stability stable
     */
    putBackendDefaults(value) {
        this._backendDefaults.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetBackendDefaults() {
        this._backendDefaults.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get backendDefaultsInput() {
        return this._backendDefaults.internalValue;
    }
    /**
     * @stability stable
     */
    get listener() {
        return this._listener;
    }
    /**
     * @stability stable
     */
    putListener(value) {
        this._listener.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get listenerInput() {
        return this._listener.internalValue;
    }
    /**
     * @stability stable
     */
    get logging() {
        return this._logging;
    }
    /**
     * @stability stable
     */
    putLogging(value) {
        this._logging.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLogging() {
        this._logging.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingInput() {
        return this._logging.internalValue;
    }
}
exports.AppmeshVirtualGatewaySpecOutputReference = AppmeshVirtualGatewaySpecOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGatewaySpecOutputReference[_10] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGatewaySpecOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/appmesh_virtual_gateway.html aws_appmesh_virtual_gateway}.
 *
 * @stability stable
 */
class AppmeshVirtualGateway extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appmesh_virtual_gateway.html aws_appmesh_virtual_gateway} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appmesh_virtual_gateway',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // spec - computed: false, optional: false, required: true
        this._spec = new AppmeshVirtualGatewaySpecOutputReference(this, "spec", true);
        this._meshName = config.meshName;
        this._meshOwner = config.meshOwner;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._spec.internalValue = config.spec;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    /**
     * @stability stable
     */
    get meshName() {
        return this.getStringAttribute('mesh_name');
    }
    /**
     * @stability stable
     */
    set meshName(value) {
        this._meshName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get meshNameInput() {
        return this._meshName;
    }
    /**
     * @stability stable
     */
    get meshOwner() {
        return this.getStringAttribute('mesh_owner');
    }
    /**
     * @stability stable
     */
    set meshOwner(value) {
        this._meshOwner = value;
    }
    /**
     * @stability stable
     */
    resetMeshOwner() {
        this._meshOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get meshOwnerInput() {
        return this._meshOwner;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // resource_owner - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get resourceOwner() {
        return this.getStringAttribute('resource_owner');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get spec() {
        return this._spec;
    }
    /**
     * @stability stable
     */
    putSpec(value) {
        this._spec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get specInput() {
        return this._spec.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            mesh_name: cdktf.stringToTerraform(this._meshName),
            mesh_owner: cdktf.stringToTerraform(this._meshOwner),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            spec: appmeshVirtualGatewaySpecToTerraform(this._spec.internalValue),
        };
    }
}
exports.AppmeshVirtualGateway = AppmeshVirtualGateway;
_11 = JSII_RTTI_SYMBOL_1;
AppmeshVirtualGateway[_11] = { fqn: "@cdktf/provider-aws.appmesh.AppmeshVirtualGateway", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AppmeshVirtualGateway.tfResourceType = "aws_appmesh_virtual_gateway";
