// @cdktf/provider-aws/lib/sagemaker/sagemaker-code-repository.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerCodeRepository = exports.SagemakerCodeRepositoryGitConfigOutputReference = exports.sagemakerCodeRepositoryGitConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerCodeRepositoryGitConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch: cdktf.stringToTerraform(struct.branch),
        repository_url: cdktf.stringToTerraform(struct.repositoryUrl),
        secret_arn: cdktf.stringToTerraform(struct.secretArn),
    };
}
exports.sagemakerCodeRepositoryGitConfigToTerraform = sagemakerCodeRepositoryGitConfigToTerraform;
/**
 * @stability stable
 */
class SagemakerCodeRepositoryGitConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._branch) {
            hasAnyValues = true;
            internalValueResult.branch = this._branch;
        }
        if (this._repositoryUrl) {
            hasAnyValues = true;
            internalValueResult.repositoryUrl = this._repositoryUrl;
        }
        if (this._secretArn) {
            hasAnyValues = true;
            internalValueResult.secretArn = this._secretArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._branch = undefined;
            this._repositoryUrl = undefined;
            this._secretArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._branch = value.branch;
            this._repositoryUrl = value.repositoryUrl;
            this._secretArn = value.secretArn;
        }
    }
    /**
     * @stability stable
     */
    get branch() {
        return this.getStringAttribute('branch');
    }
    /**
     * @stability stable
     */
    set branch(value) {
        this._branch = value;
    }
    /**
     * @stability stable
     */
    resetBranch() {
        this._branch = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get branchInput() {
        return this._branch;
    }
    /**
     * @stability stable
     */
    get repositoryUrl() {
        return this.getStringAttribute('repository_url');
    }
    /**
     * @stability stable
     */
    set repositoryUrl(value) {
        this._repositoryUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get repositoryUrlInput() {
        return this._repositoryUrl;
    }
    /**
     * @stability stable
     */
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    /**
     * @stability stable
     */
    set secretArn(value) {
        this._secretArn = value;
    }
    /**
     * @stability stable
     */
    resetSecretArn() {
        this._secretArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretArnInput() {
        return this._secretArn;
    }
}
exports.SagemakerCodeRepositoryGitConfigOutputReference = SagemakerCodeRepositoryGitConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerCodeRepositoryGitConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerCodeRepositoryGitConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_code_repository.html aws_sagemaker_code_repository}.
 *
 * @stability stable
 */
class SagemakerCodeRepository extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_code_repository.html aws_sagemaker_code_repository} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_code_repository',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // git_config - computed: false, optional: false, required: true
        this._gitConfig = new SagemakerCodeRepositoryGitConfigOutputReference(this, "git_config", true);
        this._codeRepositoryName = config.codeRepositoryName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._gitConfig.internalValue = config.gitConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get codeRepositoryName() {
        return this.getStringAttribute('code_repository_name');
    }
    /**
     * @stability stable
     */
    set codeRepositoryName(value) {
        this._codeRepositoryName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get codeRepositoryNameInput() {
        return this._codeRepositoryName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get gitConfig() {
        return this._gitConfig;
    }
    /**
     * @stability stable
     */
    putGitConfig(value) {
        this._gitConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gitConfigInput() {
        return this._gitConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            code_repository_name: cdktf.stringToTerraform(this._codeRepositoryName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            git_config: sagemakerCodeRepositoryGitConfigToTerraform(this._gitConfig.internalValue),
        };
    }
}
exports.SagemakerCodeRepository = SagemakerCodeRepository;
_b = JSII_RTTI_SYMBOL_1;
SagemakerCodeRepository[_b] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerCodeRepository", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SagemakerCodeRepository.tfResourceType = "aws_sagemaker_code_repository";
