// @cdktf/provider-aws/lib/cur/data-aws-cur-report-definition.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCurReportDefinition = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/cur_report_definition.html aws_cur_report_definition}.
 *
 * @stability stable
 */
class DataAwsCurReportDefinition extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cur_report_definition.html aws_cur_report_definition} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cur_report_definition',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._reportName = config.reportName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // additional_artifacts - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get additionalArtifacts() {
        return this.getListAttribute('additional_artifacts');
    }
    // additional_schema_elements - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get additionalSchemaElements() {
        return this.getListAttribute('additional_schema_elements');
    }
    // compression - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get compression() {
        return this.getStringAttribute('compression');
    }
    // format - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get format() {
        return this.getStringAttribute('format');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // refresh_closed_reports - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get refreshClosedReports() {
        return this.getBooleanAttribute('refresh_closed_reports');
    }
    /**
     * @stability stable
     */
    get reportName() {
        return this.getStringAttribute('report_name');
    }
    /**
     * @stability stable
     */
    set reportName(value) {
        this._reportName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportNameInput() {
        return this._reportName;
    }
    // report_versioning - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get reportVersioning() {
        return this.getStringAttribute('report_versioning');
    }
    // s3_bucket - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3Bucket() {
        return this.getStringAttribute('s3_bucket');
    }
    // s3_prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3Prefix() {
        return this.getStringAttribute('s3_prefix');
    }
    // s3_region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3Region() {
        return this.getStringAttribute('s3_region');
    }
    // time_unit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get timeUnit() {
        return this.getStringAttribute('time_unit');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            report_name: cdktf.stringToTerraform(this._reportName),
        };
    }
}
exports.DataAwsCurReportDefinition = DataAwsCurReportDefinition;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCurReportDefinition[_a] = { fqn: "@cdktf/provider-aws.cur.DataAwsCurReportDefinition", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCurReportDefinition.tfResourceType = "aws_cur_report_definition";
