// @cdktf/provider-aws/lib/globalaccelerator/data-aws-globalaccelerator-accelerator.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsGlobalacceleratorAccelerator = exports.DataAwsGlobalacceleratorAcceleratorIpSets = exports.DataAwsGlobalacceleratorAcceleratorAttributes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsGlobalacceleratorAcceleratorAttributes extends cdktf.ComplexComputedList {
    // flow_logs_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get flowLogsEnabled() {
        return this.getBooleanAttribute('flow_logs_enabled');
    }
    // flow_logs_s3_bucket - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get flowLogsS3Bucket() {
        return this.getStringAttribute('flow_logs_s3_bucket');
    }
    // flow_logs_s3_prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get flowLogsS3Prefix() {
        return this.getStringAttribute('flow_logs_s3_prefix');
    }
}
exports.DataAwsGlobalacceleratorAcceleratorAttributes = DataAwsGlobalacceleratorAcceleratorAttributes;
_a = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorAttributes[_a] = { fqn: "@cdktf/provider-aws.globalaccelerator.DataAwsGlobalacceleratorAcceleratorAttributes", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsGlobalacceleratorAcceleratorIpSets extends cdktf.ComplexComputedList {
    // ip_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddresses() {
        return this.getListAttribute('ip_addresses');
    }
    // ip_family - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
}
exports.DataAwsGlobalacceleratorAcceleratorIpSets = DataAwsGlobalacceleratorAcceleratorIpSets;
_b = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorIpSets[_b] = { fqn: "@cdktf/provider-aws.globalaccelerator.DataAwsGlobalacceleratorAcceleratorIpSets", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/globalaccelerator_accelerator.html aws_globalaccelerator_accelerator}.
 *
 * @stability stable
 */
class DataAwsGlobalacceleratorAccelerator extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/globalaccelerator_accelerator.html aws_globalaccelerator_accelerator} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_accelerator',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attributes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    attributes(index) {
        return new DataAwsGlobalacceleratorAcceleratorAttributes(this, 'attributes', index);
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // ip_address_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    // ip_sets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ipSets(index) {
        return new DataAwsGlobalacceleratorAcceleratorIpSets(this, 'ip_sets', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsGlobalacceleratorAccelerator = DataAwsGlobalacceleratorAccelerator;
_c = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAccelerator[_c] = { fqn: "@cdktf/provider-aws.globalaccelerator.DataAwsGlobalacceleratorAccelerator", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsGlobalacceleratorAccelerator.tfResourceType = "aws_globalaccelerator_accelerator";
