// @cdktf/provider-aws/lib/globalaccelerator/globalaccelerator-endpoint-group.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalacceleratorEndpointGroup = exports.GlobalacceleratorEndpointGroupTimeoutsOutputReference = exports.globalacceleratorEndpointGroupTimeoutsToTerraform = exports.globalacceleratorEndpointGroupPortOverrideToTerraform = exports.globalacceleratorEndpointGroupEndpointConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function globalacceleratorEndpointGroupEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_ip_preservation_enabled: cdktf.booleanToTerraform(struct.clientIpPreservationEnabled),
        endpoint_id: cdktf.stringToTerraform(struct.endpointId),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.globalacceleratorEndpointGroupEndpointConfigurationToTerraform = globalacceleratorEndpointGroupEndpointConfigurationToTerraform;
function globalacceleratorEndpointGroupPortOverrideToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_port: cdktf.numberToTerraform(struct.endpointPort),
        listener_port: cdktf.numberToTerraform(struct.listenerPort),
    };
}
exports.globalacceleratorEndpointGroupPortOverrideToTerraform = globalacceleratorEndpointGroupPortOverrideToTerraform;
function globalacceleratorEndpointGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.globalacceleratorEndpointGroupTimeoutsToTerraform = globalacceleratorEndpointGroupTimeoutsToTerraform;
/**
 * @stability stable
 */
class GlobalacceleratorEndpointGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.GlobalacceleratorEndpointGroupTimeoutsOutputReference = GlobalacceleratorEndpointGroupTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroupTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorEndpointGroupTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_endpoint_group.html aws_globalaccelerator_endpoint_group}.
 *
 * @stability stable
 */
class GlobalacceleratorEndpointGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_endpoint_group.html aws_globalaccelerator_endpoint_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_endpoint_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GlobalacceleratorEndpointGroupTimeoutsOutputReference(this, "timeouts", true);
        this._endpointGroupRegion = config.endpointGroupRegion;
        this._healthCheckIntervalSeconds = config.healthCheckIntervalSeconds;
        this._healthCheckPath = config.healthCheckPath;
        this._healthCheckPort = config.healthCheckPort;
        this._healthCheckProtocol = config.healthCheckProtocol;
        this._listenerArn = config.listenerArn;
        this._thresholdCount = config.thresholdCount;
        this._trafficDialPercentage = config.trafficDialPercentage;
        this._endpointConfiguration = config.endpointConfiguration;
        this._portOverride = config.portOverride;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get endpointGroupRegion() {
        return this.getStringAttribute('endpoint_group_region');
    }
    /**
     * @stability stable
     */
    set endpointGroupRegion(value) {
        this._endpointGroupRegion = value;
    }
    /**
     * @stability stable
     */
    resetEndpointGroupRegion() {
        this._endpointGroupRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointGroupRegionInput() {
        return this._endpointGroupRegion;
    }
    /**
     * @stability stable
     */
    get healthCheckIntervalSeconds() {
        return this.getNumberAttribute('health_check_interval_seconds');
    }
    /**
     * @stability stable
     */
    set healthCheckIntervalSeconds(value) {
        this._healthCheckIntervalSeconds = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheckIntervalSeconds() {
        this._healthCheckIntervalSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckIntervalSecondsInput() {
        return this._healthCheckIntervalSeconds;
    }
    /**
     * @stability stable
     */
    get healthCheckPath() {
        return this.getStringAttribute('health_check_path');
    }
    /**
     * @stability stable
     */
    set healthCheckPath(value) {
        this._healthCheckPath = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheckPath() {
        this._healthCheckPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckPathInput() {
        return this._healthCheckPath;
    }
    /**
     * @stability stable
     */
    get healthCheckPort() {
        return this.getNumberAttribute('health_check_port');
    }
    /**
     * @stability stable
     */
    set healthCheckPort(value) {
        this._healthCheckPort = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheckPort() {
        this._healthCheckPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckPortInput() {
        return this._healthCheckPort;
    }
    /**
     * @stability stable
     */
    get healthCheckProtocol() {
        return this.getStringAttribute('health_check_protocol');
    }
    /**
     * @stability stable
     */
    set healthCheckProtocol(value) {
        this._healthCheckProtocol = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheckProtocol() {
        this._healthCheckProtocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckProtocolInput() {
        return this._healthCheckProtocol;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get listenerArn() {
        return this.getStringAttribute('listener_arn');
    }
    /**
     * @stability stable
     */
    set listenerArn(value) {
        this._listenerArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get listenerArnInput() {
        return this._listenerArn;
    }
    /**
     * @stability stable
     */
    get thresholdCount() {
        return this.getNumberAttribute('threshold_count');
    }
    /**
     * @stability stable
     */
    set thresholdCount(value) {
        this._thresholdCount = value;
    }
    /**
     * @stability stable
     */
    resetThresholdCount() {
        this._thresholdCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thresholdCountInput() {
        return this._thresholdCount;
    }
    /**
     * @stability stable
     */
    get trafficDialPercentage() {
        return this.getNumberAttribute('traffic_dial_percentage');
    }
    /**
     * @stability stable
     */
    set trafficDialPercentage(value) {
        this._trafficDialPercentage = value;
    }
    /**
     * @stability stable
     */
    resetTrafficDialPercentage() {
        this._trafficDialPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get trafficDialPercentageInput() {
        return this._trafficDialPercentage;
    }
    /**
     * @stability stable
     */
    get endpointConfiguration() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('endpoint_configuration');
    }
    /**
     * @stability stable
     */
    set endpointConfiguration(value) {
        this._endpointConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetEndpointConfiguration() {
        this._endpointConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointConfigurationInput() {
        return this._endpointConfiguration;
    }
    /**
     * @stability stable
     */
    get portOverride() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('port_override');
    }
    /**
     * @stability stable
     */
    set portOverride(value) {
        this._portOverride = value;
    }
    /**
     * @stability stable
     */
    resetPortOverride() {
        this._portOverride = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portOverrideInput() {
        return this._portOverride;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            endpoint_group_region: cdktf.stringToTerraform(this._endpointGroupRegion),
            health_check_interval_seconds: cdktf.numberToTerraform(this._healthCheckIntervalSeconds),
            health_check_path: cdktf.stringToTerraform(this._healthCheckPath),
            health_check_port: cdktf.numberToTerraform(this._healthCheckPort),
            health_check_protocol: cdktf.stringToTerraform(this._healthCheckProtocol),
            listener_arn: cdktf.stringToTerraform(this._listenerArn),
            threshold_count: cdktf.numberToTerraform(this._thresholdCount),
            traffic_dial_percentage: cdktf.numberToTerraform(this._trafficDialPercentage),
            endpoint_configuration: cdktf.listMapper(globalacceleratorEndpointGroupEndpointConfigurationToTerraform)(this._endpointConfiguration),
            port_override: cdktf.listMapper(globalacceleratorEndpointGroupPortOverrideToTerraform)(this._portOverride),
            timeouts: globalacceleratorEndpointGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GlobalacceleratorEndpointGroup = GlobalacceleratorEndpointGroup;
_b = JSII_RTTI_SYMBOL_1;
GlobalacceleratorEndpointGroup[_b] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorEndpointGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlobalacceleratorEndpointGroup.tfResourceType = "aws_globalaccelerator_endpoint_group";
