// @cdktf/provider-aws/lib/globalaccelerator/globalaccelerator-accelerator.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalacceleratorAccelerator = exports.GlobalacceleratorAcceleratorTimeoutsOutputReference = exports.globalacceleratorAcceleratorTimeoutsToTerraform = exports.GlobalacceleratorAcceleratorAttributesOutputReference = exports.globalacceleratorAcceleratorAttributesToTerraform = exports.GlobalacceleratorAcceleratorIpSets = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class GlobalacceleratorAcceleratorIpSets extends cdktf.ComplexComputedList {
    // ip_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddresses() {
        return this.getListAttribute('ip_addresses');
    }
    // ip_family - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
}
exports.GlobalacceleratorAcceleratorIpSets = GlobalacceleratorAcceleratorIpSets;
_a = JSII_RTTI_SYMBOL_1;
GlobalacceleratorAcceleratorIpSets[_a] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorAcceleratorIpSets", version: "3.0.1" };
function globalacceleratorAcceleratorAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        flow_logs_enabled: cdktf.booleanToTerraform(struct.flowLogsEnabled),
        flow_logs_s3_bucket: cdktf.stringToTerraform(struct.flowLogsS3Bucket),
        flow_logs_s3_prefix: cdktf.stringToTerraform(struct.flowLogsS3Prefix),
    };
}
exports.globalacceleratorAcceleratorAttributesToTerraform = globalacceleratorAcceleratorAttributesToTerraform;
/**
 * @stability stable
 */
class GlobalacceleratorAcceleratorAttributesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._flowLogsEnabled) {
            hasAnyValues = true;
            internalValueResult.flowLogsEnabled = this._flowLogsEnabled;
        }
        if (this._flowLogsS3Bucket) {
            hasAnyValues = true;
            internalValueResult.flowLogsS3Bucket = this._flowLogsS3Bucket;
        }
        if (this._flowLogsS3Prefix) {
            hasAnyValues = true;
            internalValueResult.flowLogsS3Prefix = this._flowLogsS3Prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._flowLogsEnabled = undefined;
            this._flowLogsS3Bucket = undefined;
            this._flowLogsS3Prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._flowLogsEnabled = value.flowLogsEnabled;
            this._flowLogsS3Bucket = value.flowLogsS3Bucket;
            this._flowLogsS3Prefix = value.flowLogsS3Prefix;
        }
    }
    /**
     * @stability stable
     */
    get flowLogsEnabled() {
        return this.getBooleanAttribute('flow_logs_enabled');
    }
    /**
     * @stability stable
     */
    set flowLogsEnabled(value) {
        this._flowLogsEnabled = value;
    }
    /**
     * @stability stable
     */
    resetFlowLogsEnabled() {
        this._flowLogsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get flowLogsEnabledInput() {
        return this._flowLogsEnabled;
    }
    /**
     * @stability stable
     */
    get flowLogsS3Bucket() {
        return this.getStringAttribute('flow_logs_s3_bucket');
    }
    /**
     * @stability stable
     */
    set flowLogsS3Bucket(value) {
        this._flowLogsS3Bucket = value;
    }
    /**
     * @stability stable
     */
    resetFlowLogsS3Bucket() {
        this._flowLogsS3Bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get flowLogsS3BucketInput() {
        return this._flowLogsS3Bucket;
    }
    /**
     * @stability stable
     */
    get flowLogsS3Prefix() {
        return this.getStringAttribute('flow_logs_s3_prefix');
    }
    /**
     * @stability stable
     */
    set flowLogsS3Prefix(value) {
        this._flowLogsS3Prefix = value;
    }
    /**
     * @stability stable
     */
    resetFlowLogsS3Prefix() {
        this._flowLogsS3Prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get flowLogsS3PrefixInput() {
        return this._flowLogsS3Prefix;
    }
}
exports.GlobalacceleratorAcceleratorAttributesOutputReference = GlobalacceleratorAcceleratorAttributesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlobalacceleratorAcceleratorAttributesOutputReference[_b] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorAcceleratorAttributesOutputReference", version: "3.0.1" };
function globalacceleratorAcceleratorTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.globalacceleratorAcceleratorTimeoutsToTerraform = globalacceleratorAcceleratorTimeoutsToTerraform;
/**
 * @stability stable
 */
class GlobalacceleratorAcceleratorTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.GlobalacceleratorAcceleratorTimeoutsOutputReference = GlobalacceleratorAcceleratorTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlobalacceleratorAcceleratorTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorAcceleratorTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_accelerator.html aws_globalaccelerator_accelerator}.
 *
 * @stability stable
 */
class GlobalacceleratorAccelerator extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_accelerator.html aws_globalaccelerator_accelerator} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_accelerator',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // attributes - computed: false, optional: true, required: false
        this._attributes = new GlobalacceleratorAcceleratorAttributesOutputReference(this, "attributes", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GlobalacceleratorAcceleratorTimeoutsOutputReference(this, "timeouts", true);
        this._enabled = config.enabled;
        this._ipAddressType = config.ipAddressType;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._attributes.internalValue = config.attributes;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    /**
     * @stability stable
     */
    set ipAddressType(value) {
        this._ipAddressType = value;
    }
    /**
     * @stability stable
     */
    resetIpAddressType() {
        this._ipAddressType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipAddressTypeInput() {
        return this._ipAddressType;
    }
    // ip_sets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ipSets(index) {
        return new GlobalacceleratorAcceleratorIpSets(this, 'ip_sets', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get attributes() {
        return this._attributes;
    }
    /**
     * @stability stable
     */
    putAttributes(value) {
        this._attributes.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAttributes() {
        this._attributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get attributesInput() {
        return this._attributes.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enabled: cdktf.booleanToTerraform(this._enabled),
            ip_address_type: cdktf.stringToTerraform(this._ipAddressType),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            attributes: globalacceleratorAcceleratorAttributesToTerraform(this._attributes.internalValue),
            timeouts: globalacceleratorAcceleratorTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GlobalacceleratorAccelerator = GlobalacceleratorAccelerator;
_d = JSII_RTTI_SYMBOL_1;
GlobalacceleratorAccelerator[_d] = { fqn: "@cdktf/provider-aws.globalaccelerator.GlobalacceleratorAccelerator", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlobalacceleratorAccelerator.tfResourceType = "aws_globalaccelerator_accelerator";
