// @cdktf/provider-aws/lib/elasticache/data-aws-elasticache-cluster.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElasticacheCluster = exports.DataAwsElasticacheClusterCacheNodes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsElasticacheClusterCacheNodes extends cdktf.ComplexComputedList {
    // address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get address() {
        return this.getStringAttribute('address');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataAwsElasticacheClusterCacheNodes = DataAwsElasticacheClusterCacheNodes;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheClusterCacheNodes[_a] = { fqn: "@cdktf/provider-aws.elasticache.DataAwsElasticacheClusterCacheNodes", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/elasticache_cluster.html aws_elasticache_cluster}.
 *
 * @stability stable
 */
class DataAwsElasticacheCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/elasticache_cluster.html aws_elasticache_cluster} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterId = config.clusterId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // cache_nodes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    cacheNodes(index) {
        return new DataAwsElasticacheClusterCacheNodes(this, 'cache_nodes', index);
    }
    // cluster_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterAddress() {
        return this.getStringAttribute('cluster_address');
    }
    /**
     * @stability stable
     */
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    /**
     * @stability stable
     */
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterIdInput() {
        return this._clusterId;
    }
    // configuration_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get configurationEndpoint() {
        return this.getStringAttribute('configuration_endpoint');
    }
    // engine - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engine() {
        return this.getStringAttribute('engine');
    }
    // engine_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // maintenance_window - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get maintenanceWindow() {
        return this.getStringAttribute('maintenance_window');
    }
    // node_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    // notification_topic_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get notificationTopicArn() {
        return this.getStringAttribute('notification_topic_arn');
    }
    // num_cache_nodes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get numCacheNodes() {
        return this.getNumberAttribute('num_cache_nodes');
    }
    // parameter_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get parameterGroupName() {
        return this.getStringAttribute('parameter_group_name');
    }
    // port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    // replication_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get replicationGroupId() {
        return this.getStringAttribute('replication_group_id');
    }
    // security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    // security_group_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupNames() {
        return this.getListAttribute('security_group_names');
    }
    // snapshot_retention_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotRetentionLimit() {
        return this.getNumberAttribute('snapshot_retention_limit');
    }
    // snapshot_window - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotWindow() {
        return this.getStringAttribute('snapshot_window');
    }
    // subnet_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetGroupName() {
        return this.getStringAttribute('subnet_group_name');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsElasticacheCluster = DataAwsElasticacheCluster;
_b = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheCluster[_b] = { fqn: "@cdktf/provider-aws.elasticache.DataAwsElasticacheCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsElasticacheCluster.tfResourceType = "aws_elasticache_cluster";
