// @cdktf/provider-aws/lib/cloudfront/cloudfront-distribution.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudfrontDistribution = exports.CloudfrontDistributionViewerCertificateOutputReference = exports.cloudfrontDistributionViewerCertificateToTerraform = exports.CloudfrontDistributionRestrictionsOutputReference = exports.cloudfrontDistributionRestrictionsToTerraform = exports.CloudfrontDistributionRestrictionsGeoRestrictionOutputReference = exports.cloudfrontDistributionRestrictionsGeoRestrictionToTerraform = exports.cloudfrontDistributionOriginGroupToTerraform = exports.cloudfrontDistributionOriginGroupMemberToTerraform = exports.CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference = exports.cloudfrontDistributionOriginGroupFailoverCriteriaToTerraform = exports.cloudfrontDistributionOriginToTerraform = exports.CloudfrontDistributionOriginS3OriginConfigOutputReference = exports.cloudfrontDistributionOriginS3OriginConfigToTerraform = exports.CloudfrontDistributionOriginOriginShieldOutputReference = exports.cloudfrontDistributionOriginOriginShieldToTerraform = exports.CloudfrontDistributionOriginCustomOriginConfigOutputReference = exports.cloudfrontDistributionOriginCustomOriginConfigToTerraform = exports.cloudfrontDistributionOriginCustomHeaderToTerraform = exports.cloudfrontDistributionOrderedCacheBehaviorToTerraform = exports.cloudfrontDistributionOrderedCacheBehaviorLambdaFunctionAssociationToTerraform = exports.cloudfrontDistributionOrderedCacheBehaviorFunctionAssociationToTerraform = exports.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference = exports.cloudfrontDistributionOrderedCacheBehaviorForwardedValuesToTerraform = exports.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference = exports.cloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesToTerraform = exports.CloudfrontDistributionLoggingConfigOutputReference = exports.cloudfrontDistributionLoggingConfigToTerraform = exports.CloudfrontDistributionDefaultCacheBehaviorOutputReference = exports.cloudfrontDistributionDefaultCacheBehaviorToTerraform = exports.cloudfrontDistributionDefaultCacheBehaviorLambdaFunctionAssociationToTerraform = exports.cloudfrontDistributionDefaultCacheBehaviorFunctionAssociationToTerraform = exports.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference = exports.cloudfrontDistributionDefaultCacheBehaviorForwardedValuesToTerraform = exports.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference = exports.cloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesToTerraform = exports.cloudfrontDistributionCustomErrorResponseToTerraform = exports.CloudfrontDistributionTrustedSigners = exports.CloudfrontDistributionTrustedSignersItems = exports.CloudfrontDistributionTrustedKeyGroups = exports.CloudfrontDistributionTrustedKeyGroupsItems = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class CloudfrontDistributionTrustedKeyGroupsItems extends cdktf.ComplexComputedList {
    // key_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyGroupId() {
        return this.getStringAttribute('key_group_id');
    }
    // key_pair_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyPairIds() {
        return this.getListAttribute('key_pair_ids');
    }
}
exports.CloudfrontDistributionTrustedKeyGroupsItems = CloudfrontDistributionTrustedKeyGroupsItems;
_a = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionTrustedKeyGroupsItems[_a] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionTrustedKeyGroupsItems", version: "3.0.1" };
/**
 * @stability stable
 */
class CloudfrontDistributionTrustedKeyGroups extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('items');
    }
}
exports.CloudfrontDistributionTrustedKeyGroups = CloudfrontDistributionTrustedKeyGroups;
_b = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionTrustedKeyGroups[_b] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionTrustedKeyGroups", version: "3.0.1" };
/**
 * @stability stable
 */
class CloudfrontDistributionTrustedSignersItems extends cdktf.ComplexComputedList {
    // aws_account_number - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get awsAccountNumber() {
        return this.getStringAttribute('aws_account_number');
    }
    // key_pair_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyPairIds() {
        return this.getListAttribute('key_pair_ids');
    }
}
exports.CloudfrontDistributionTrustedSignersItems = CloudfrontDistributionTrustedSignersItems;
_c = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionTrustedSignersItems[_c] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionTrustedSignersItems", version: "3.0.1" };
/**
 * @stability stable
 */
class CloudfrontDistributionTrustedSigners extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('items');
    }
}
exports.CloudfrontDistributionTrustedSigners = CloudfrontDistributionTrustedSigners;
_d = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionTrustedSigners[_d] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionTrustedSigners", version: "3.0.1" };
function cloudfrontDistributionCustomErrorResponseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        error_caching_min_ttl: cdktf.numberToTerraform(struct.errorCachingMinTtl),
        error_code: cdktf.numberToTerraform(struct.errorCode),
        response_code: cdktf.numberToTerraform(struct.responseCode),
        response_page_path: cdktf.stringToTerraform(struct.responsePagePath),
    };
}
exports.cloudfrontDistributionCustomErrorResponseToTerraform = cloudfrontDistributionCustomErrorResponseToTerraform;
function cloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        forward: cdktf.stringToTerraform(struct.forward),
        whitelisted_names: cdktf.listMapper(cdktf.stringToTerraform)(struct.whitelistedNames),
    };
}
exports.cloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesToTerraform = cloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._forward) {
            hasAnyValues = true;
            internalValueResult.forward = this._forward;
        }
        if (this._whitelistedNames) {
            hasAnyValues = true;
            internalValueResult.whitelistedNames = this._whitelistedNames;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._forward = undefined;
            this._whitelistedNames = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._forward = value.forward;
            this._whitelistedNames = value.whitelistedNames;
        }
    }
    /**
     * @stability stable
     */
    get forward() {
        return this.getStringAttribute('forward');
    }
    /**
     * @stability stable
     */
    set forward(value) {
        this._forward = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forwardInput() {
        return this._forward;
    }
    /**
     * @stability stable
     */
    get whitelistedNames() {
        return this.getListAttribute('whitelisted_names');
    }
    /**
     * @stability stable
     */
    set whitelistedNames(value) {
        this._whitelistedNames = value;
    }
    /**
     * @stability stable
     */
    resetWhitelistedNames() {
        this._whitelistedNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get whitelistedNamesInput() {
        return this._whitelistedNames;
    }
}
exports.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference = CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference", version: "3.0.1" };
function cloudfrontDistributionDefaultCacheBehaviorForwardedValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        headers: cdktf.listMapper(cdktf.stringToTerraform)(struct.headers),
        query_string: cdktf.booleanToTerraform(struct.queryString),
        query_string_cache_keys: cdktf.listMapper(cdktf.stringToTerraform)(struct.queryStringCacheKeys),
        cookies: cloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesToTerraform(struct.cookies),
    };
}
exports.cloudfrontDistributionDefaultCacheBehaviorForwardedValuesToTerraform = cloudfrontDistributionDefaultCacheBehaviorForwardedValuesToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cookies - computed: false, optional: false, required: true
        this._cookies = new CloudfrontDistributionDefaultCacheBehaviorForwardedValuesCookiesOutputReference(this, "cookies", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headers) {
            hasAnyValues = true;
            internalValueResult.headers = this._headers;
        }
        if (this._queryString) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString;
        }
        if (this._queryStringCacheKeys) {
            hasAnyValues = true;
            internalValueResult.queryStringCacheKeys = this._queryStringCacheKeys;
        }
        if ((_u = this._cookies) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.cookies = (_v = this._cookies) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._headers = undefined;
            this._queryString = undefined;
            this._queryStringCacheKeys = undefined;
            this._cookies.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._headers = value.headers;
            this._queryString = value.queryString;
            this._queryStringCacheKeys = value.queryStringCacheKeys;
            this._cookies.internalValue = value.cookies;
        }
    }
    /**
     * @stability stable
     */
    get headers() {
        return this.getListAttribute('headers');
    }
    /**
     * @stability stable
     */
    set headers(value) {
        this._headers = value;
    }
    /**
     * @stability stable
     */
    resetHeaders() {
        this._headers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headersInput() {
        return this._headers;
    }
    /**
     * @stability stable
     */
    get queryString() {
        return this.getBooleanAttribute('query_string');
    }
    /**
     * @stability stable
     */
    set queryString(value) {
        this._queryString = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringInput() {
        return this._queryString;
    }
    /**
     * @stability stable
     */
    get queryStringCacheKeys() {
        return this.getListAttribute('query_string_cache_keys');
    }
    /**
     * @stability stable
     */
    set queryStringCacheKeys(value) {
        this._queryStringCacheKeys = value;
    }
    /**
     * @stability stable
     */
    resetQueryStringCacheKeys() {
        this._queryStringCacheKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringCacheKeysInput() {
        return this._queryStringCacheKeys;
    }
    /**
     * @stability stable
     */
    get cookies() {
        return this._cookies;
    }
    /**
     * @stability stable
     */
    putCookies(value) {
        this._cookies.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookiesInput() {
        return this._cookies.internalValue;
    }
}
exports.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference = CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference[_f] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference", version: "3.0.1" };
function cloudfrontDistributionDefaultCacheBehaviorFunctionAssociationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_type: cdktf.stringToTerraform(struct.eventType),
        function_arn: cdktf.stringToTerraform(struct.functionArn),
    };
}
exports.cloudfrontDistributionDefaultCacheBehaviorFunctionAssociationToTerraform = cloudfrontDistributionDefaultCacheBehaviorFunctionAssociationToTerraform;
function cloudfrontDistributionDefaultCacheBehaviorLambdaFunctionAssociationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_type: cdktf.stringToTerraform(struct.eventType),
        include_body: cdktf.booleanToTerraform(struct.includeBody),
        lambda_arn: cdktf.stringToTerraform(struct.lambdaArn),
    };
}
exports.cloudfrontDistributionDefaultCacheBehaviorLambdaFunctionAssociationToTerraform = cloudfrontDistributionDefaultCacheBehaviorLambdaFunctionAssociationToTerraform;
function cloudfrontDistributionDefaultCacheBehaviorToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allowed_methods: cdktf.listMapper(cdktf.stringToTerraform)(struct.allowedMethods),
        cache_policy_id: cdktf.stringToTerraform(struct.cachePolicyId),
        cached_methods: cdktf.listMapper(cdktf.stringToTerraform)(struct.cachedMethods),
        compress: cdktf.booleanToTerraform(struct.compress),
        default_ttl: cdktf.numberToTerraform(struct.defaultTtl),
        field_level_encryption_id: cdktf.stringToTerraform(struct.fieldLevelEncryptionId),
        max_ttl: cdktf.numberToTerraform(struct.maxTtl),
        min_ttl: cdktf.numberToTerraform(struct.minTtl),
        origin_request_policy_id: cdktf.stringToTerraform(struct.originRequestPolicyId),
        realtime_log_config_arn: cdktf.stringToTerraform(struct.realtimeLogConfigArn),
        response_headers_policy_id: cdktf.stringToTerraform(struct.responseHeadersPolicyId),
        smooth_streaming: cdktf.booleanToTerraform(struct.smoothStreaming),
        target_origin_id: cdktf.stringToTerraform(struct.targetOriginId),
        trusted_key_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.trustedKeyGroups),
        trusted_signers: cdktf.listMapper(cdktf.stringToTerraform)(struct.trustedSigners),
        viewer_protocol_policy: cdktf.stringToTerraform(struct.viewerProtocolPolicy),
        forwarded_values: cloudfrontDistributionDefaultCacheBehaviorForwardedValuesToTerraform(struct.forwardedValues),
        function_association: cdktf.listMapper(cloudfrontDistributionDefaultCacheBehaviorFunctionAssociationToTerraform)(struct.functionAssociation),
        lambda_function_association: cdktf.listMapper(cloudfrontDistributionDefaultCacheBehaviorLambdaFunctionAssociationToTerraform)(struct.lambdaFunctionAssociation),
    };
}
exports.cloudfrontDistributionDefaultCacheBehaviorToTerraform = cloudfrontDistributionDefaultCacheBehaviorToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionDefaultCacheBehaviorOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // forwarded_values - computed: false, optional: true, required: false
        this._forwardedValues = new CloudfrontDistributionDefaultCacheBehaviorForwardedValuesOutputReference(this, "forwarded_values", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowedMethods) {
            hasAnyValues = true;
            internalValueResult.allowedMethods = this._allowedMethods;
        }
        if (this._cachePolicyId) {
            hasAnyValues = true;
            internalValueResult.cachePolicyId = this._cachePolicyId;
        }
        if (this._cachedMethods) {
            hasAnyValues = true;
            internalValueResult.cachedMethods = this._cachedMethods;
        }
        if (this._compress) {
            hasAnyValues = true;
            internalValueResult.compress = this._compress;
        }
        if (this._defaultTtl) {
            hasAnyValues = true;
            internalValueResult.defaultTtl = this._defaultTtl;
        }
        if (this._fieldLevelEncryptionId) {
            hasAnyValues = true;
            internalValueResult.fieldLevelEncryptionId = this._fieldLevelEncryptionId;
        }
        if (this._maxTtl) {
            hasAnyValues = true;
            internalValueResult.maxTtl = this._maxTtl;
        }
        if (this._minTtl) {
            hasAnyValues = true;
            internalValueResult.minTtl = this._minTtl;
        }
        if (this._originRequestPolicyId) {
            hasAnyValues = true;
            internalValueResult.originRequestPolicyId = this._originRequestPolicyId;
        }
        if (this._realtimeLogConfigArn) {
            hasAnyValues = true;
            internalValueResult.realtimeLogConfigArn = this._realtimeLogConfigArn;
        }
        if (this._responseHeadersPolicyId) {
            hasAnyValues = true;
            internalValueResult.responseHeadersPolicyId = this._responseHeadersPolicyId;
        }
        if (this._smoothStreaming) {
            hasAnyValues = true;
            internalValueResult.smoothStreaming = this._smoothStreaming;
        }
        if (this._targetOriginId) {
            hasAnyValues = true;
            internalValueResult.targetOriginId = this._targetOriginId;
        }
        if (this._trustedKeyGroups) {
            hasAnyValues = true;
            internalValueResult.trustedKeyGroups = this._trustedKeyGroups;
        }
        if (this._trustedSigners) {
            hasAnyValues = true;
            internalValueResult.trustedSigners = this._trustedSigners;
        }
        if (this._viewerProtocolPolicy) {
            hasAnyValues = true;
            internalValueResult.viewerProtocolPolicy = this._viewerProtocolPolicy;
        }
        if ((_u = this._forwardedValues) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.forwardedValues = (_v = this._forwardedValues) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if (this._functionAssociation) {
            hasAnyValues = true;
            internalValueResult.functionAssociation = this._functionAssociation;
        }
        if (this._lambdaFunctionAssociation) {
            hasAnyValues = true;
            internalValueResult.lambdaFunctionAssociation = this._lambdaFunctionAssociation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowedMethods = undefined;
            this._cachePolicyId = undefined;
            this._cachedMethods = undefined;
            this._compress = undefined;
            this._defaultTtl = undefined;
            this._fieldLevelEncryptionId = undefined;
            this._maxTtl = undefined;
            this._minTtl = undefined;
            this._originRequestPolicyId = undefined;
            this._realtimeLogConfigArn = undefined;
            this._responseHeadersPolicyId = undefined;
            this._smoothStreaming = undefined;
            this._targetOriginId = undefined;
            this._trustedKeyGroups = undefined;
            this._trustedSigners = undefined;
            this._viewerProtocolPolicy = undefined;
            this._forwardedValues.internalValue = undefined;
            this._functionAssociation = undefined;
            this._lambdaFunctionAssociation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowedMethods = value.allowedMethods;
            this._cachePolicyId = value.cachePolicyId;
            this._cachedMethods = value.cachedMethods;
            this._compress = value.compress;
            this._defaultTtl = value.defaultTtl;
            this._fieldLevelEncryptionId = value.fieldLevelEncryptionId;
            this._maxTtl = value.maxTtl;
            this._minTtl = value.minTtl;
            this._originRequestPolicyId = value.originRequestPolicyId;
            this._realtimeLogConfigArn = value.realtimeLogConfigArn;
            this._responseHeadersPolicyId = value.responseHeadersPolicyId;
            this._smoothStreaming = value.smoothStreaming;
            this._targetOriginId = value.targetOriginId;
            this._trustedKeyGroups = value.trustedKeyGroups;
            this._trustedSigners = value.trustedSigners;
            this._viewerProtocolPolicy = value.viewerProtocolPolicy;
            this._forwardedValues.internalValue = value.forwardedValues;
            this._functionAssociation = value.functionAssociation;
            this._lambdaFunctionAssociation = value.lambdaFunctionAssociation;
        }
    }
    /**
     * @stability stable
     */
    get allowedMethods() {
        return this.getListAttribute('allowed_methods');
    }
    /**
     * @stability stable
     */
    set allowedMethods(value) {
        this._allowedMethods = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowedMethodsInput() {
        return this._allowedMethods;
    }
    /**
     * @stability stable
     */
    get cachePolicyId() {
        return this.getStringAttribute('cache_policy_id');
    }
    /**
     * @stability stable
     */
    set cachePolicyId(value) {
        this._cachePolicyId = value;
    }
    /**
     * @stability stable
     */
    resetCachePolicyId() {
        this._cachePolicyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cachePolicyIdInput() {
        return this._cachePolicyId;
    }
    /**
     * @stability stable
     */
    get cachedMethods() {
        return this.getListAttribute('cached_methods');
    }
    /**
     * @stability stable
     */
    set cachedMethods(value) {
        this._cachedMethods = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cachedMethodsInput() {
        return this._cachedMethods;
    }
    /**
     * @stability stable
     */
    get compress() {
        return this.getBooleanAttribute('compress');
    }
    /**
     * @stability stable
     */
    set compress(value) {
        this._compress = value;
    }
    /**
     * @stability stable
     */
    resetCompress() {
        this._compress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get compressInput() {
        return this._compress;
    }
    /**
     * @stability stable
     */
    get defaultTtl() {
        return this.getNumberAttribute('default_ttl');
    }
    /**
     * @stability stable
     */
    set defaultTtl(value) {
        this._defaultTtl = value;
    }
    /**
     * @stability stable
     */
    resetDefaultTtl() {
        this._defaultTtl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultTtlInput() {
        return this._defaultTtl;
    }
    /**
     * @stability stable
     */
    get fieldLevelEncryptionId() {
        return this.getStringAttribute('field_level_encryption_id');
    }
    /**
     * @stability stable
     */
    set fieldLevelEncryptionId(value) {
        this._fieldLevelEncryptionId = value;
    }
    /**
     * @stability stable
     */
    resetFieldLevelEncryptionId() {
        this._fieldLevelEncryptionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fieldLevelEncryptionIdInput() {
        return this._fieldLevelEncryptionId;
    }
    /**
     * @stability stable
     */
    get maxTtl() {
        return this.getNumberAttribute('max_ttl');
    }
    /**
     * @stability stable
     */
    set maxTtl(value) {
        this._maxTtl = value;
    }
    /**
     * @stability stable
     */
    resetMaxTtl() {
        this._maxTtl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxTtlInput() {
        return this._maxTtl;
    }
    /**
     * @stability stable
     */
    get minTtl() {
        return this.getNumberAttribute('min_ttl');
    }
    /**
     * @stability stable
     */
    set minTtl(value) {
        this._minTtl = value;
    }
    /**
     * @stability stable
     */
    resetMinTtl() {
        this._minTtl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minTtlInput() {
        return this._minTtl;
    }
    /**
     * @stability stable
     */
    get originRequestPolicyId() {
        return this.getStringAttribute('origin_request_policy_id');
    }
    /**
     * @stability stable
     */
    set originRequestPolicyId(value) {
        this._originRequestPolicyId = value;
    }
    /**
     * @stability stable
     */
    resetOriginRequestPolicyId() {
        this._originRequestPolicyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originRequestPolicyIdInput() {
        return this._originRequestPolicyId;
    }
    /**
     * @stability stable
     */
    get realtimeLogConfigArn() {
        return this.getStringAttribute('realtime_log_config_arn');
    }
    /**
     * @stability stable
     */
    set realtimeLogConfigArn(value) {
        this._realtimeLogConfigArn = value;
    }
    /**
     * @stability stable
     */
    resetRealtimeLogConfigArn() {
        this._realtimeLogConfigArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get realtimeLogConfigArnInput() {
        return this._realtimeLogConfigArn;
    }
    /**
     * @stability stable
     */
    get responseHeadersPolicyId() {
        return this.getStringAttribute('response_headers_policy_id');
    }
    /**
     * @stability stable
     */
    set responseHeadersPolicyId(value) {
        this._responseHeadersPolicyId = value;
    }
    /**
     * @stability stable
     */
    resetResponseHeadersPolicyId() {
        this._responseHeadersPolicyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseHeadersPolicyIdInput() {
        return this._responseHeadersPolicyId;
    }
    /**
     * @stability stable
     */
    get smoothStreaming() {
        return this.getBooleanAttribute('smooth_streaming');
    }
    /**
     * @stability stable
     */
    set smoothStreaming(value) {
        this._smoothStreaming = value;
    }
    /**
     * @stability stable
     */
    resetSmoothStreaming() {
        this._smoothStreaming = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get smoothStreamingInput() {
        return this._smoothStreaming;
    }
    /**
     * @stability stable
     */
    get targetOriginId() {
        return this.getStringAttribute('target_origin_id');
    }
    /**
     * @stability stable
     */
    set targetOriginId(value) {
        this._targetOriginId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetOriginIdInput() {
        return this._targetOriginId;
    }
    /**
     * @stability stable
     */
    get trustedKeyGroups() {
        return this.getListAttribute('trusted_key_groups');
    }
    /**
     * @stability stable
     */
    set trustedKeyGroups(value) {
        this._trustedKeyGroups = value;
    }
    /**
     * @stability stable
     */
    resetTrustedKeyGroups() {
        this._trustedKeyGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get trustedKeyGroupsInput() {
        return this._trustedKeyGroups;
    }
    /**
     * @stability stable
     */
    get trustedSigners() {
        return this.getListAttribute('trusted_signers');
    }
    /**
     * @stability stable
     */
    set trustedSigners(value) {
        this._trustedSigners = value;
    }
    /**
     * @stability stable
     */
    resetTrustedSigners() {
        this._trustedSigners = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get trustedSignersInput() {
        return this._trustedSigners;
    }
    /**
     * @stability stable
     */
    get viewerProtocolPolicy() {
        return this.getStringAttribute('viewer_protocol_policy');
    }
    /**
     * @stability stable
     */
    set viewerProtocolPolicy(value) {
        this._viewerProtocolPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get viewerProtocolPolicyInput() {
        return this._viewerProtocolPolicy;
    }
    /**
     * @stability stable
     */
    get forwardedValues() {
        return this._forwardedValues;
    }
    /**
     * @stability stable
     */
    putForwardedValues(value) {
        this._forwardedValues.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetForwardedValues() {
        this._forwardedValues.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forwardedValuesInput() {
        return this._forwardedValues.internalValue;
    }
    /**
     * @stability stable
     */
    get functionAssociation() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('function_association');
    }
    /**
     * @stability stable
     */
    set functionAssociation(value) {
        this._functionAssociation = value;
    }
    /**
     * @stability stable
     */
    resetFunctionAssociation() {
        this._functionAssociation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get functionAssociationInput() {
        return this._functionAssociation;
    }
    /**
     * @stability stable
     */
    get lambdaFunctionAssociation() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('lambda_function_association');
    }
    /**
     * @stability stable
     */
    set lambdaFunctionAssociation(value) {
        this._lambdaFunctionAssociation = value;
    }
    /**
     * @stability stable
     */
    resetLambdaFunctionAssociation() {
        this._lambdaFunctionAssociation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaFunctionAssociationInput() {
        return this._lambdaFunctionAssociation;
    }
}
exports.CloudfrontDistributionDefaultCacheBehaviorOutputReference = CloudfrontDistributionDefaultCacheBehaviorOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionDefaultCacheBehaviorOutputReference[_g] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionDefaultCacheBehaviorOutputReference", version: "3.0.1" };
function cloudfrontDistributionLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        include_cookies: cdktf.booleanToTerraform(struct.includeCookies),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.cloudfrontDistributionLoggingConfigToTerraform = cloudfrontDistributionLoggingConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._includeCookies) {
            hasAnyValues = true;
            internalValueResult.includeCookies = this._includeCookies;
        }
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._includeCookies = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._includeCookies = value.includeCookies;
            this._prefix = value.prefix;
        }
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get includeCookies() {
        return this.getBooleanAttribute('include_cookies');
    }
    /**
     * @stability stable
     */
    set includeCookies(value) {
        this._includeCookies = value;
    }
    /**
     * @stability stable
     */
    resetIncludeCookies() {
        this._includeCookies = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeCookiesInput() {
        return this._includeCookies;
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
}
exports.CloudfrontDistributionLoggingConfigOutputReference = CloudfrontDistributionLoggingConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionLoggingConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionLoggingConfigOutputReference", version: "3.0.1" };
function cloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        forward: cdktf.stringToTerraform(struct.forward),
        whitelisted_names: cdktf.listMapper(cdktf.stringToTerraform)(struct.whitelistedNames),
    };
}
exports.cloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesToTerraform = cloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._forward) {
            hasAnyValues = true;
            internalValueResult.forward = this._forward;
        }
        if (this._whitelistedNames) {
            hasAnyValues = true;
            internalValueResult.whitelistedNames = this._whitelistedNames;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._forward = undefined;
            this._whitelistedNames = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._forward = value.forward;
            this._whitelistedNames = value.whitelistedNames;
        }
    }
    /**
     * @stability stable
     */
    get forward() {
        return this.getStringAttribute('forward');
    }
    /**
     * @stability stable
     */
    set forward(value) {
        this._forward = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forwardInput() {
        return this._forward;
    }
    /**
     * @stability stable
     */
    get whitelistedNames() {
        return this.getListAttribute('whitelisted_names');
    }
    /**
     * @stability stable
     */
    set whitelistedNames(value) {
        this._whitelistedNames = value;
    }
    /**
     * @stability stable
     */
    resetWhitelistedNames() {
        this._whitelistedNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get whitelistedNamesInput() {
        return this._whitelistedNames;
    }
}
exports.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference = CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference[_j] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference", version: "3.0.1" };
function cloudfrontDistributionOrderedCacheBehaviorForwardedValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        headers: cdktf.listMapper(cdktf.stringToTerraform)(struct.headers),
        query_string: cdktf.booleanToTerraform(struct.queryString),
        query_string_cache_keys: cdktf.listMapper(cdktf.stringToTerraform)(struct.queryStringCacheKeys),
        cookies: cloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesToTerraform(struct.cookies),
    };
}
exports.cloudfrontDistributionOrderedCacheBehaviorForwardedValuesToTerraform = cloudfrontDistributionOrderedCacheBehaviorForwardedValuesToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cookies - computed: false, optional: false, required: true
        this._cookies = new CloudfrontDistributionOrderedCacheBehaviorForwardedValuesCookiesOutputReference(this, "cookies", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headers) {
            hasAnyValues = true;
            internalValueResult.headers = this._headers;
        }
        if (this._queryString) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString;
        }
        if (this._queryStringCacheKeys) {
            hasAnyValues = true;
            internalValueResult.queryStringCacheKeys = this._queryStringCacheKeys;
        }
        if ((_u = this._cookies) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.cookies = (_v = this._cookies) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._headers = undefined;
            this._queryString = undefined;
            this._queryStringCacheKeys = undefined;
            this._cookies.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._headers = value.headers;
            this._queryString = value.queryString;
            this._queryStringCacheKeys = value.queryStringCacheKeys;
            this._cookies.internalValue = value.cookies;
        }
    }
    /**
     * @stability stable
     */
    get headers() {
        return this.getListAttribute('headers');
    }
    /**
     * @stability stable
     */
    set headers(value) {
        this._headers = value;
    }
    /**
     * @stability stable
     */
    resetHeaders() {
        this._headers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headersInput() {
        return this._headers;
    }
    /**
     * @stability stable
     */
    get queryString() {
        return this.getBooleanAttribute('query_string');
    }
    /**
     * @stability stable
     */
    set queryString(value) {
        this._queryString = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringInput() {
        return this._queryString;
    }
    /**
     * @stability stable
     */
    get queryStringCacheKeys() {
        return this.getListAttribute('query_string_cache_keys');
    }
    /**
     * @stability stable
     */
    set queryStringCacheKeys(value) {
        this._queryStringCacheKeys = value;
    }
    /**
     * @stability stable
     */
    resetQueryStringCacheKeys() {
        this._queryStringCacheKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringCacheKeysInput() {
        return this._queryStringCacheKeys;
    }
    /**
     * @stability stable
     */
    get cookies() {
        return this._cookies;
    }
    /**
     * @stability stable
     */
    putCookies(value) {
        this._cookies.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookiesInput() {
        return this._cookies.internalValue;
    }
}
exports.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference = CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference[_k] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOrderedCacheBehaviorForwardedValuesOutputReference", version: "3.0.1" };
function cloudfrontDistributionOrderedCacheBehaviorFunctionAssociationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_type: cdktf.stringToTerraform(struct.eventType),
        function_arn: cdktf.stringToTerraform(struct.functionArn),
    };
}
exports.cloudfrontDistributionOrderedCacheBehaviorFunctionAssociationToTerraform = cloudfrontDistributionOrderedCacheBehaviorFunctionAssociationToTerraform;
function cloudfrontDistributionOrderedCacheBehaviorLambdaFunctionAssociationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_type: cdktf.stringToTerraform(struct.eventType),
        include_body: cdktf.booleanToTerraform(struct.includeBody),
        lambda_arn: cdktf.stringToTerraform(struct.lambdaArn),
    };
}
exports.cloudfrontDistributionOrderedCacheBehaviorLambdaFunctionAssociationToTerraform = cloudfrontDistributionOrderedCacheBehaviorLambdaFunctionAssociationToTerraform;
function cloudfrontDistributionOrderedCacheBehaviorToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allowed_methods: cdktf.listMapper(cdktf.stringToTerraform)(struct.allowedMethods),
        cache_policy_id: cdktf.stringToTerraform(struct.cachePolicyId),
        cached_methods: cdktf.listMapper(cdktf.stringToTerraform)(struct.cachedMethods),
        compress: cdktf.booleanToTerraform(struct.compress),
        default_ttl: cdktf.numberToTerraform(struct.defaultTtl),
        field_level_encryption_id: cdktf.stringToTerraform(struct.fieldLevelEncryptionId),
        max_ttl: cdktf.numberToTerraform(struct.maxTtl),
        min_ttl: cdktf.numberToTerraform(struct.minTtl),
        origin_request_policy_id: cdktf.stringToTerraform(struct.originRequestPolicyId),
        path_pattern: cdktf.stringToTerraform(struct.pathPattern),
        realtime_log_config_arn: cdktf.stringToTerraform(struct.realtimeLogConfigArn),
        response_headers_policy_id: cdktf.stringToTerraform(struct.responseHeadersPolicyId),
        smooth_streaming: cdktf.booleanToTerraform(struct.smoothStreaming),
        target_origin_id: cdktf.stringToTerraform(struct.targetOriginId),
        trusted_key_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.trustedKeyGroups),
        trusted_signers: cdktf.listMapper(cdktf.stringToTerraform)(struct.trustedSigners),
        viewer_protocol_policy: cdktf.stringToTerraform(struct.viewerProtocolPolicy),
        forwarded_values: cloudfrontDistributionOrderedCacheBehaviorForwardedValuesToTerraform(struct.forwardedValues),
        function_association: cdktf.listMapper(cloudfrontDistributionOrderedCacheBehaviorFunctionAssociationToTerraform)(struct.functionAssociation),
        lambda_function_association: cdktf.listMapper(cloudfrontDistributionOrderedCacheBehaviorLambdaFunctionAssociationToTerraform)(struct.lambdaFunctionAssociation),
    };
}
exports.cloudfrontDistributionOrderedCacheBehaviorToTerraform = cloudfrontDistributionOrderedCacheBehaviorToTerraform;
function cloudfrontDistributionOriginCustomHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudfrontDistributionOriginCustomHeaderToTerraform = cloudfrontDistributionOriginCustomHeaderToTerraform;
function cloudfrontDistributionOriginCustomOriginConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        http_port: cdktf.numberToTerraform(struct.httpPort),
        https_port: cdktf.numberToTerraform(struct.httpsPort),
        origin_keepalive_timeout: cdktf.numberToTerraform(struct.originKeepaliveTimeout),
        origin_protocol_policy: cdktf.stringToTerraform(struct.originProtocolPolicy),
        origin_read_timeout: cdktf.numberToTerraform(struct.originReadTimeout),
        origin_ssl_protocols: cdktf.listMapper(cdktf.stringToTerraform)(struct.originSslProtocols),
    };
}
exports.cloudfrontDistributionOriginCustomOriginConfigToTerraform = cloudfrontDistributionOriginCustomOriginConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOriginCustomOriginConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpPort) {
            hasAnyValues = true;
            internalValueResult.httpPort = this._httpPort;
        }
        if (this._httpsPort) {
            hasAnyValues = true;
            internalValueResult.httpsPort = this._httpsPort;
        }
        if (this._originKeepaliveTimeout) {
            hasAnyValues = true;
            internalValueResult.originKeepaliveTimeout = this._originKeepaliveTimeout;
        }
        if (this._originProtocolPolicy) {
            hasAnyValues = true;
            internalValueResult.originProtocolPolicy = this._originProtocolPolicy;
        }
        if (this._originReadTimeout) {
            hasAnyValues = true;
            internalValueResult.originReadTimeout = this._originReadTimeout;
        }
        if (this._originSslProtocols) {
            hasAnyValues = true;
            internalValueResult.originSslProtocols = this._originSslProtocols;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._httpPort = undefined;
            this._httpsPort = undefined;
            this._originKeepaliveTimeout = undefined;
            this._originProtocolPolicy = undefined;
            this._originReadTimeout = undefined;
            this._originSslProtocols = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._httpPort = value.httpPort;
            this._httpsPort = value.httpsPort;
            this._originKeepaliveTimeout = value.originKeepaliveTimeout;
            this._originProtocolPolicy = value.originProtocolPolicy;
            this._originReadTimeout = value.originReadTimeout;
            this._originSslProtocols = value.originSslProtocols;
        }
    }
    /**
     * @stability stable
     */
    get httpPort() {
        return this.getNumberAttribute('http_port');
    }
    /**
     * @stability stable
     */
    set httpPort(value) {
        this._httpPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpPortInput() {
        return this._httpPort;
    }
    /**
     * @stability stable
     */
    get httpsPort() {
        return this.getNumberAttribute('https_port');
    }
    /**
     * @stability stable
     */
    set httpsPort(value) {
        this._httpsPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpsPortInput() {
        return this._httpsPort;
    }
    /**
     * @stability stable
     */
    get originKeepaliveTimeout() {
        return this.getNumberAttribute('origin_keepalive_timeout');
    }
    /**
     * @stability stable
     */
    set originKeepaliveTimeout(value) {
        this._originKeepaliveTimeout = value;
    }
    /**
     * @stability stable
     */
    resetOriginKeepaliveTimeout() {
        this._originKeepaliveTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originKeepaliveTimeoutInput() {
        return this._originKeepaliveTimeout;
    }
    /**
     * @stability stable
     */
    get originProtocolPolicy() {
        return this.getStringAttribute('origin_protocol_policy');
    }
    /**
     * @stability stable
     */
    set originProtocolPolicy(value) {
        this._originProtocolPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originProtocolPolicyInput() {
        return this._originProtocolPolicy;
    }
    /**
     * @stability stable
     */
    get originReadTimeout() {
        return this.getNumberAttribute('origin_read_timeout');
    }
    /**
     * @stability stable
     */
    set originReadTimeout(value) {
        this._originReadTimeout = value;
    }
    /**
     * @stability stable
     */
    resetOriginReadTimeout() {
        this._originReadTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originReadTimeoutInput() {
        return this._originReadTimeout;
    }
    /**
     * @stability stable
     */
    get originSslProtocols() {
        return this.getListAttribute('origin_ssl_protocols');
    }
    /**
     * @stability stable
     */
    set originSslProtocols(value) {
        this._originSslProtocols = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originSslProtocolsInput() {
        return this._originSslProtocols;
    }
}
exports.CloudfrontDistributionOriginCustomOriginConfigOutputReference = CloudfrontDistributionOriginCustomOriginConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOriginCustomOriginConfigOutputReference[_l] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOriginCustomOriginConfigOutputReference", version: "3.0.1" };
function cloudfrontDistributionOriginOriginShieldToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        origin_shield_region: cdktf.stringToTerraform(struct.originShieldRegion),
    };
}
exports.cloudfrontDistributionOriginOriginShieldToTerraform = cloudfrontDistributionOriginOriginShieldToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOriginOriginShieldOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._originShieldRegion) {
            hasAnyValues = true;
            internalValueResult.originShieldRegion = this._originShieldRegion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._originShieldRegion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._originShieldRegion = value.originShieldRegion;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get originShieldRegion() {
        return this.getStringAttribute('origin_shield_region');
    }
    /**
     * @stability stable
     */
    set originShieldRegion(value) {
        this._originShieldRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originShieldRegionInput() {
        return this._originShieldRegion;
    }
}
exports.CloudfrontDistributionOriginOriginShieldOutputReference = CloudfrontDistributionOriginOriginShieldOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOriginOriginShieldOutputReference[_m] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOriginOriginShieldOutputReference", version: "3.0.1" };
function cloudfrontDistributionOriginS3OriginConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        origin_access_identity: cdktf.stringToTerraform(struct.originAccessIdentity),
    };
}
exports.cloudfrontDistributionOriginS3OriginConfigToTerraform = cloudfrontDistributionOriginS3OriginConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOriginS3OriginConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._originAccessIdentity) {
            hasAnyValues = true;
            internalValueResult.originAccessIdentity = this._originAccessIdentity;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._originAccessIdentity = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._originAccessIdentity = value.originAccessIdentity;
        }
    }
    /**
     * @stability stable
     */
    get originAccessIdentity() {
        return this.getStringAttribute('origin_access_identity');
    }
    /**
     * @stability stable
     */
    set originAccessIdentity(value) {
        this._originAccessIdentity = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originAccessIdentityInput() {
        return this._originAccessIdentity;
    }
}
exports.CloudfrontDistributionOriginS3OriginConfigOutputReference = CloudfrontDistributionOriginS3OriginConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOriginS3OriginConfigOutputReference[_o] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOriginS3OriginConfigOutputReference", version: "3.0.1" };
function cloudfrontDistributionOriginToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_attempts: cdktf.numberToTerraform(struct.connectionAttempts),
        connection_timeout: cdktf.numberToTerraform(struct.connectionTimeout),
        domain_name: cdktf.stringToTerraform(struct.domainName),
        origin_id: cdktf.stringToTerraform(struct.originId),
        origin_path: cdktf.stringToTerraform(struct.originPath),
        custom_header: cdktf.listMapper(cloudfrontDistributionOriginCustomHeaderToTerraform)(struct.customHeader),
        custom_origin_config: cloudfrontDistributionOriginCustomOriginConfigToTerraform(struct.customOriginConfig),
        origin_shield: cloudfrontDistributionOriginOriginShieldToTerraform(struct.originShield),
        s3_origin_config: cloudfrontDistributionOriginS3OriginConfigToTerraform(struct.s3OriginConfig),
    };
}
exports.cloudfrontDistributionOriginToTerraform = cloudfrontDistributionOriginToTerraform;
function cloudfrontDistributionOriginGroupFailoverCriteriaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status_codes: cdktf.listMapper(cdktf.numberToTerraform)(struct.statusCodes),
    };
}
exports.cloudfrontDistributionOriginGroupFailoverCriteriaToTerraform = cloudfrontDistributionOriginGroupFailoverCriteriaToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._statusCodes) {
            hasAnyValues = true;
            internalValueResult.statusCodes = this._statusCodes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._statusCodes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._statusCodes = value.statusCodes;
        }
    }
    /**
     * @stability stable
     */
    get statusCodes() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('status_codes');
    }
    /**
     * @stability stable
     */
    set statusCodes(value) {
        this._statusCodes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusCodesInput() {
        return this._statusCodes;
    }
}
exports.CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference = CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference[_p] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionOriginGroupFailoverCriteriaOutputReference", version: "3.0.1" };
function cloudfrontDistributionOriginGroupMemberToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        origin_id: cdktf.stringToTerraform(struct.originId),
    };
}
exports.cloudfrontDistributionOriginGroupMemberToTerraform = cloudfrontDistributionOriginGroupMemberToTerraform;
function cloudfrontDistributionOriginGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        origin_id: cdktf.stringToTerraform(struct.originId),
        failover_criteria: cloudfrontDistributionOriginGroupFailoverCriteriaToTerraform(struct.failoverCriteria),
        member: cdktf.listMapper(cloudfrontDistributionOriginGroupMemberToTerraform)(struct.member),
    };
}
exports.cloudfrontDistributionOriginGroupToTerraform = cloudfrontDistributionOriginGroupToTerraform;
function cloudfrontDistributionRestrictionsGeoRestrictionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        locations: cdktf.listMapper(cdktf.stringToTerraform)(struct.locations),
        restriction_type: cdktf.stringToTerraform(struct.restrictionType),
    };
}
exports.cloudfrontDistributionRestrictionsGeoRestrictionToTerraform = cloudfrontDistributionRestrictionsGeoRestrictionToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionRestrictionsGeoRestrictionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._locations) {
            hasAnyValues = true;
            internalValueResult.locations = this._locations;
        }
        if (this._restrictionType) {
            hasAnyValues = true;
            internalValueResult.restrictionType = this._restrictionType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._locations = undefined;
            this._restrictionType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._locations = value.locations;
            this._restrictionType = value.restrictionType;
        }
    }
    /**
     * @stability stable
     */
    get locations() {
        return this.getListAttribute('locations');
    }
    /**
     * @stability stable
     */
    set locations(value) {
        this._locations = value;
    }
    /**
     * @stability stable
     */
    resetLocations() {
        this._locations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationsInput() {
        return this._locations;
    }
    /**
     * @stability stable
     */
    get restrictionType() {
        return this.getStringAttribute('restriction_type');
    }
    /**
     * @stability stable
     */
    set restrictionType(value) {
        this._restrictionType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictionTypeInput() {
        return this._restrictionType;
    }
}
exports.CloudfrontDistributionRestrictionsGeoRestrictionOutputReference = CloudfrontDistributionRestrictionsGeoRestrictionOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionRestrictionsGeoRestrictionOutputReference[_q] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionRestrictionsGeoRestrictionOutputReference", version: "3.0.1" };
function cloudfrontDistributionRestrictionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        geo_restriction: cloudfrontDistributionRestrictionsGeoRestrictionToTerraform(struct.geoRestriction),
    };
}
exports.cloudfrontDistributionRestrictionsToTerraform = cloudfrontDistributionRestrictionsToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionRestrictionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // geo_restriction - computed: false, optional: false, required: true
        this._geoRestriction = new CloudfrontDistributionRestrictionsGeoRestrictionOutputReference(this, "geo_restriction", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._geoRestriction) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.geoRestriction = (_v = this._geoRestriction) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._geoRestriction.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._geoRestriction.internalValue = value.geoRestriction;
        }
    }
    /**
     * @stability stable
     */
    get geoRestriction() {
        return this._geoRestriction;
    }
    /**
     * @stability stable
     */
    putGeoRestriction(value) {
        this._geoRestriction.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get geoRestrictionInput() {
        return this._geoRestriction.internalValue;
    }
}
exports.CloudfrontDistributionRestrictionsOutputReference = CloudfrontDistributionRestrictionsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionRestrictionsOutputReference[_r] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionRestrictionsOutputReference", version: "3.0.1" };
function cloudfrontDistributionViewerCertificateToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        acm_certificate_arn: cdktf.stringToTerraform(struct.acmCertificateArn),
        cloudfront_default_certificate: cdktf.booleanToTerraform(struct.cloudfrontDefaultCertificate),
        iam_certificate_id: cdktf.stringToTerraform(struct.iamCertificateId),
        minimum_protocol_version: cdktf.stringToTerraform(struct.minimumProtocolVersion),
        ssl_support_method: cdktf.stringToTerraform(struct.sslSupportMethod),
    };
}
exports.cloudfrontDistributionViewerCertificateToTerraform = cloudfrontDistributionViewerCertificateToTerraform;
/**
 * @stability stable
 */
class CloudfrontDistributionViewerCertificateOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._acmCertificateArn) {
            hasAnyValues = true;
            internalValueResult.acmCertificateArn = this._acmCertificateArn;
        }
        if (this._cloudfrontDefaultCertificate) {
            hasAnyValues = true;
            internalValueResult.cloudfrontDefaultCertificate = this._cloudfrontDefaultCertificate;
        }
        if (this._iamCertificateId) {
            hasAnyValues = true;
            internalValueResult.iamCertificateId = this._iamCertificateId;
        }
        if (this._minimumProtocolVersion) {
            hasAnyValues = true;
            internalValueResult.minimumProtocolVersion = this._minimumProtocolVersion;
        }
        if (this._sslSupportMethod) {
            hasAnyValues = true;
            internalValueResult.sslSupportMethod = this._sslSupportMethod;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._acmCertificateArn = undefined;
            this._cloudfrontDefaultCertificate = undefined;
            this._iamCertificateId = undefined;
            this._minimumProtocolVersion = undefined;
            this._sslSupportMethod = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._acmCertificateArn = value.acmCertificateArn;
            this._cloudfrontDefaultCertificate = value.cloudfrontDefaultCertificate;
            this._iamCertificateId = value.iamCertificateId;
            this._minimumProtocolVersion = value.minimumProtocolVersion;
            this._sslSupportMethod = value.sslSupportMethod;
        }
    }
    /**
     * @stability stable
     */
    get acmCertificateArn() {
        return this.getStringAttribute('acm_certificate_arn');
    }
    /**
     * @stability stable
     */
    set acmCertificateArn(value) {
        this._acmCertificateArn = value;
    }
    /**
     * @stability stable
     */
    resetAcmCertificateArn() {
        this._acmCertificateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acmCertificateArnInput() {
        return this._acmCertificateArn;
    }
    /**
     * @stability stable
     */
    get cloudfrontDefaultCertificate() {
        return this.getBooleanAttribute('cloudfront_default_certificate');
    }
    /**
     * @stability stable
     */
    set cloudfrontDefaultCertificate(value) {
        this._cloudfrontDefaultCertificate = value;
    }
    /**
     * @stability stable
     */
    resetCloudfrontDefaultCertificate() {
        this._cloudfrontDefaultCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudfrontDefaultCertificateInput() {
        return this._cloudfrontDefaultCertificate;
    }
    /**
     * @stability stable
     */
    get iamCertificateId() {
        return this.getStringAttribute('iam_certificate_id');
    }
    /**
     * @stability stable
     */
    set iamCertificateId(value) {
        this._iamCertificateId = value;
    }
    /**
     * @stability stable
     */
    resetIamCertificateId() {
        this._iamCertificateId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamCertificateIdInput() {
        return this._iamCertificateId;
    }
    /**
     * @stability stable
     */
    get minimumProtocolVersion() {
        return this.getStringAttribute('minimum_protocol_version');
    }
    /**
     * @stability stable
     */
    set minimumProtocolVersion(value) {
        this._minimumProtocolVersion = value;
    }
    /**
     * @stability stable
     */
    resetMinimumProtocolVersion() {
        this._minimumProtocolVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minimumProtocolVersionInput() {
        return this._minimumProtocolVersion;
    }
    /**
     * @stability stable
     */
    get sslSupportMethod() {
        return this.getStringAttribute('ssl_support_method');
    }
    /**
     * @stability stable
     */
    set sslSupportMethod(value) {
        this._sslSupportMethod = value;
    }
    /**
     * @stability stable
     */
    resetSslSupportMethod() {
        this._sslSupportMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sslSupportMethodInput() {
        return this._sslSupportMethod;
    }
}
exports.CloudfrontDistributionViewerCertificateOutputReference = CloudfrontDistributionViewerCertificateOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudfrontDistributionViewerCertificateOutputReference[_s] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistributionViewerCertificateOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudfront_distribution.html aws_cloudfront_distribution}.
 *
 * @stability stable
 */
class CloudfrontDistribution extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudfront_distribution.html aws_cloudfront_distribution} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_distribution',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // default_cache_behavior - computed: false, optional: false, required: true
        this._defaultCacheBehavior = new CloudfrontDistributionDefaultCacheBehaviorOutputReference(this, "default_cache_behavior", true);
        // logging_config - computed: false, optional: true, required: false
        this._loggingConfig = new CloudfrontDistributionLoggingConfigOutputReference(this, "logging_config", true);
        // restrictions - computed: false, optional: false, required: true
        this._restrictions = new CloudfrontDistributionRestrictionsOutputReference(this, "restrictions", true);
        // viewer_certificate - computed: false, optional: false, required: true
        this._viewerCertificate = new CloudfrontDistributionViewerCertificateOutputReference(this, "viewer_certificate", true);
        this._aliases = config.aliases;
        this._comment = config.comment;
        this._defaultRootObject = config.defaultRootObject;
        this._enabled = config.enabled;
        this._httpVersion = config.httpVersion;
        this._isIpv6Enabled = config.isIpv6Enabled;
        this._priceClass = config.priceClass;
        this._retainOnDelete = config.retainOnDelete;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._waitForDeployment = config.waitForDeployment;
        this._webAclId = config.webAclId;
        this._customErrorResponse = config.customErrorResponse;
        this._defaultCacheBehavior.internalValue = config.defaultCacheBehavior;
        this._loggingConfig.internalValue = config.loggingConfig;
        this._orderedCacheBehavior = config.orderedCacheBehavior;
        this._origin = config.origin;
        this._originGroup = config.originGroup;
        this._restrictions.internalValue = config.restrictions;
        this._viewerCertificate.internalValue = config.viewerCertificate;
    }
    /**
     * @stability stable
     */
    get aliases() {
        return this.getListAttribute('aliases');
    }
    /**
     * @stability stable
     */
    set aliases(value) {
        this._aliases = value;
    }
    /**
     * @stability stable
     */
    resetAliases() {
        this._aliases = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get aliasesInput() {
        return this._aliases;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // caller_reference - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get callerReference() {
        return this.getStringAttribute('caller_reference');
    }
    /**
     * @stability stable
     */
    get comment() {
        return this.getStringAttribute('comment');
    }
    /**
     * @stability stable
     */
    set comment(value) {
        this._comment = value;
    }
    /**
     * @stability stable
     */
    resetComment() {
        this._comment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commentInput() {
        return this._comment;
    }
    /**
     * @stability stable
     */
    get defaultRootObject() {
        return this.getStringAttribute('default_root_object');
    }
    /**
     * @stability stable
     */
    set defaultRootObject(value) {
        this._defaultRootObject = value;
    }
    /**
     * @stability stable
     */
    resetDefaultRootObject() {
        this._defaultRootObject = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultRootObjectInput() {
        return this._defaultRootObject;
    }
    // domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // etag - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get etag() {
        return this.getStringAttribute('etag');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    /**
     * @stability stable
     */
    get httpVersion() {
        return this.getStringAttribute('http_version');
    }
    /**
     * @stability stable
     */
    set httpVersion(value) {
        this._httpVersion = value;
    }
    /**
     * @stability stable
     */
    resetHttpVersion() {
        this._httpVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpVersionInput() {
        return this._httpVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // in_progress_validation_batches - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get inProgressValidationBatches() {
        return this.getNumberAttribute('in_progress_validation_batches');
    }
    /**
     * @stability stable
     */
    get isIpv6Enabled() {
        return this.getBooleanAttribute('is_ipv6_enabled');
    }
    /**
     * @stability stable
     */
    set isIpv6Enabled(value) {
        this._isIpv6Enabled = value;
    }
    /**
     * @stability stable
     */
    resetIsIpv6Enabled() {
        this._isIpv6Enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get isIpv6EnabledInput() {
        return this._isIpv6Enabled;
    }
    // last_modified_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModifiedTime() {
        return this.getStringAttribute('last_modified_time');
    }
    /**
     * @stability stable
     */
    get priceClass() {
        return this.getStringAttribute('price_class');
    }
    /**
     * @stability stable
     */
    set priceClass(value) {
        this._priceClass = value;
    }
    /**
     * @stability stable
     */
    resetPriceClass() {
        this._priceClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get priceClassInput() {
        return this._priceClass;
    }
    /**
     * @stability stable
     */
    get retainOnDelete() {
        return this.getBooleanAttribute('retain_on_delete');
    }
    /**
     * @stability stable
     */
    set retainOnDelete(value) {
        this._retainOnDelete = value;
    }
    /**
     * @stability stable
     */
    resetRetainOnDelete() {
        this._retainOnDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get retainOnDeleteInput() {
        return this._retainOnDelete;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // trusted_key_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    trustedKeyGroups(index) {
        return new CloudfrontDistributionTrustedKeyGroups(this, 'trusted_key_groups', index);
    }
    // trusted_signers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    trustedSigners(index) {
        return new CloudfrontDistributionTrustedSigners(this, 'trusted_signers', index);
    }
    /**
     * @stability stable
     */
    get waitForDeployment() {
        return this.getBooleanAttribute('wait_for_deployment');
    }
    /**
     * @stability stable
     */
    set waitForDeployment(value) {
        this._waitForDeployment = value;
    }
    /**
     * @stability stable
     */
    resetWaitForDeployment() {
        this._waitForDeployment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get waitForDeploymentInput() {
        return this._waitForDeployment;
    }
    /**
     * @stability stable
     */
    get webAclId() {
        return this.getStringAttribute('web_acl_id');
    }
    /**
     * @stability stable
     */
    set webAclId(value) {
        this._webAclId = value;
    }
    /**
     * @stability stable
     */
    resetWebAclId() {
        this._webAclId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get webAclIdInput() {
        return this._webAclId;
    }
    /**
     * @stability stable
     */
    get customErrorResponse() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('custom_error_response');
    }
    /**
     * @stability stable
     */
    set customErrorResponse(value) {
        this._customErrorResponse = value;
    }
    /**
     * @stability stable
     */
    resetCustomErrorResponse() {
        this._customErrorResponse = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customErrorResponseInput() {
        return this._customErrorResponse;
    }
    /**
     * @stability stable
     */
    get defaultCacheBehavior() {
        return this._defaultCacheBehavior;
    }
    /**
     * @stability stable
     */
    putDefaultCacheBehavior(value) {
        this._defaultCacheBehavior.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultCacheBehaviorInput() {
        return this._defaultCacheBehavior.internalValue;
    }
    /**
     * @stability stable
     */
    get loggingConfig() {
        return this._loggingConfig;
    }
    /**
     * @stability stable
     */
    putLoggingConfig(value) {
        this._loggingConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLoggingConfig() {
        this._loggingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingConfigInput() {
        return this._loggingConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get orderedCacheBehavior() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ordered_cache_behavior');
    }
    /**
     * @stability stable
     */
    set orderedCacheBehavior(value) {
        this._orderedCacheBehavior = value;
    }
    /**
     * @stability stable
     */
    resetOrderedCacheBehavior() {
        this._orderedCacheBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get orderedCacheBehaviorInput() {
        return this._orderedCacheBehavior;
    }
    /**
     * @stability stable
     */
    get origin() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('origin');
    }
    /**
     * @stability stable
     */
    set origin(value) {
        this._origin = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originInput() {
        return this._origin;
    }
    /**
     * @stability stable
     */
    get originGroup() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('origin_group');
    }
    /**
     * @stability stable
     */
    set originGroup(value) {
        this._originGroup = value;
    }
    /**
     * @stability stable
     */
    resetOriginGroup() {
        this._originGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get originGroupInput() {
        return this._originGroup;
    }
    /**
     * @stability stable
     */
    get restrictions() {
        return this._restrictions;
    }
    /**
     * @stability stable
     */
    putRestrictions(value) {
        this._restrictions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictionsInput() {
        return this._restrictions.internalValue;
    }
    /**
     * @stability stable
     */
    get viewerCertificate() {
        return this._viewerCertificate;
    }
    /**
     * @stability stable
     */
    putViewerCertificate(value) {
        this._viewerCertificate.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get viewerCertificateInput() {
        return this._viewerCertificate.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            aliases: cdktf.listMapper(cdktf.stringToTerraform)(this._aliases),
            comment: cdktf.stringToTerraform(this._comment),
            default_root_object: cdktf.stringToTerraform(this._defaultRootObject),
            enabled: cdktf.booleanToTerraform(this._enabled),
            http_version: cdktf.stringToTerraform(this._httpVersion),
            is_ipv6_enabled: cdktf.booleanToTerraform(this._isIpv6Enabled),
            price_class: cdktf.stringToTerraform(this._priceClass),
            retain_on_delete: cdktf.booleanToTerraform(this._retainOnDelete),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            wait_for_deployment: cdktf.booleanToTerraform(this._waitForDeployment),
            web_acl_id: cdktf.stringToTerraform(this._webAclId),
            custom_error_response: cdktf.listMapper(cloudfrontDistributionCustomErrorResponseToTerraform)(this._customErrorResponse),
            default_cache_behavior: cloudfrontDistributionDefaultCacheBehaviorToTerraform(this._defaultCacheBehavior.internalValue),
            logging_config: cloudfrontDistributionLoggingConfigToTerraform(this._loggingConfig.internalValue),
            ordered_cache_behavior: cdktf.listMapper(cloudfrontDistributionOrderedCacheBehaviorToTerraform)(this._orderedCacheBehavior),
            origin: cdktf.listMapper(cloudfrontDistributionOriginToTerraform)(this._origin),
            origin_group: cdktf.listMapper(cloudfrontDistributionOriginGroupToTerraform)(this._originGroup),
            restrictions: cloudfrontDistributionRestrictionsToTerraform(this._restrictions.internalValue),
            viewer_certificate: cloudfrontDistributionViewerCertificateToTerraform(this._viewerCertificate.internalValue),
        };
    }
}
exports.CloudfrontDistribution = CloudfrontDistribution;
_t = JSII_RTTI_SYMBOL_1;
CloudfrontDistribution[_t] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontDistribution", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudfrontDistribution.tfResourceType = "aws_cloudfront_distribution";
