// @cdktf/provider-aws/lib/ecr/data-aws-ecr-authorization-token.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcrAuthorizationToken = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecr_authorization_token.html aws_ecr_authorization_token}.
 *
 * @stability stable
 */
class DataAwsEcrAuthorizationToken extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecr_authorization_token.html aws_ecr_authorization_token} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_authorization_token',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._registryId = config.registryId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // authorization_token - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authorizationToken() {
        return this.getStringAttribute('authorization_token');
    }
    // expires_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get expiresAt() {
        return this.getStringAttribute('expires_at');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // password - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get password() {
        return this.getStringAttribute('password');
    }
    // proxy_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get proxyEndpoint() {
        return this.getStringAttribute('proxy_endpoint');
    }
    /**
     * @stability stable
     */
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    /**
     * @stability stable
     */
    set registryId(value) {
        this._registryId = value;
    }
    /**
     * @stability stable
     */
    resetRegistryId() {
        this._registryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registryIdInput() {
        return this._registryId;
    }
    // user_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userName() {
        return this.getStringAttribute('user_name');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            registry_id: cdktf.stringToTerraform(this._registryId),
        };
    }
}
exports.DataAwsEcrAuthorizationToken = DataAwsEcrAuthorizationToken;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcrAuthorizationToken[_a] = { fqn: "@cdktf/provider-aws.ecr.DataAwsEcrAuthorizationToken", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEcrAuthorizationToken.tfResourceType = "aws_ecr_authorization_token";
