// @cdktf/provider-aws/lib/batch/data-aws-batch-job-queue.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBatchJobQueue = exports.DataAwsBatchJobQueueComputeEnvironmentOrder = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsBatchJobQueueComputeEnvironmentOrder extends cdktf.ComplexComputedList {
    // compute_environment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get computeEnvironment() {
        return this.getStringAttribute('compute_environment');
    }
    // order - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get order() {
        return this.getNumberAttribute('order');
    }
}
exports.DataAwsBatchJobQueueComputeEnvironmentOrder = DataAwsBatchJobQueueComputeEnvironmentOrder;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBatchJobQueueComputeEnvironmentOrder[_a] = { fqn: "@cdktf/provider-aws.batch.DataAwsBatchJobQueueComputeEnvironmentOrder", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/batch_job_queue.html aws_batch_job_queue}.
 *
 * @stability stable
 */
class DataAwsBatchJobQueue extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/batch_job_queue.html aws_batch_job_queue} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_batch_job_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // compute_environment_order - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    computeEnvironmentOrder(index) {
        return new DataAwsBatchJobQueueComputeEnvironmentOrder(this, 'compute_environment_order', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // priority - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // status_reason - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get statusReason() {
        return this.getStringAttribute('status_reason');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsBatchJobQueue = DataAwsBatchJobQueue;
_b = JSII_RTTI_SYMBOL_1;
DataAwsBatchJobQueue[_b] = { fqn: "@cdktf/provider-aws.batch.DataAwsBatchJobQueue", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsBatchJobQueue.tfResourceType = "aws_batch_job_queue";
