// @cdktf/provider-aws/lib/ecr/data-aws-ecr-repository.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcrRepository = exports.DataAwsEcrRepositoryImageScanningConfiguration = exports.DataAwsEcrRepositoryEncryptionConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsEcrRepositoryEncryptionConfiguration extends cdktf.ComplexComputedList {
    // encryption_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encryptionType() {
        return this.getStringAttribute('encryption_type');
    }
    // kms_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
}
exports.DataAwsEcrRepositoryEncryptionConfiguration = DataAwsEcrRepositoryEncryptionConfiguration;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcrRepositoryEncryptionConfiguration[_a] = { fqn: "@cdktf/provider-aws.ecr.DataAwsEcrRepositoryEncryptionConfiguration", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsEcrRepositoryImageScanningConfiguration extends cdktf.ComplexComputedList {
    // scan_on_push - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get scanOnPush() {
        return this.getBooleanAttribute('scan_on_push');
    }
}
exports.DataAwsEcrRepositoryImageScanningConfiguration = DataAwsEcrRepositoryImageScanningConfiguration;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEcrRepositoryImageScanningConfiguration[_b] = { fqn: "@cdktf/provider-aws.ecr.DataAwsEcrRepositoryImageScanningConfiguration", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecr_repository.html aws_ecr_repository}.
 *
 * @stability stable
 */
class DataAwsEcrRepository extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecr_repository.html aws_ecr_repository} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_repository',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._registryId = config.registryId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // encryption_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    encryptionConfiguration(index) {
        return new DataAwsEcrRepositoryEncryptionConfiguration(this, 'encryption_configuration', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image_scanning_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    imageScanningConfiguration(index) {
        return new DataAwsEcrRepositoryImageScanningConfiguration(this, 'image_scanning_configuration', index);
    }
    // image_tag_mutability - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageTagMutability() {
        return this.getStringAttribute('image_tag_mutability');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    /**
     * @stability stable
     */
    set registryId(value) {
        this._registryId = value;
    }
    /**
     * @stability stable
     */
    resetRegistryId() {
        this._registryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registryIdInput() {
        return this._registryId;
    }
    // repository_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get repositoryUrl() {
        return this.getStringAttribute('repository_url');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            registry_id: cdktf.stringToTerraform(this._registryId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEcrRepository = DataAwsEcrRepository;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEcrRepository[_c] = { fqn: "@cdktf/provider-aws.ecr.DataAwsEcrRepository", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEcrRepository.tfResourceType = "aws_ecr_repository";
