// @cdktf/provider-aws/lib/iot/iot-thing-group.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThingGroup = exports.IotThingGroupPropertiesOutputReference = exports.iotThingGroupPropertiesToTerraform = exports.IotThingGroupPropertiesAttributePayloadOutputReference = exports.iotThingGroupPropertiesAttributePayloadToTerraform = exports.IotThingGroupMetadata = exports.IotThingGroupMetadataRootToParentGroups = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class IotThingGroupMetadataRootToParentGroups extends cdktf.ComplexComputedList {
    // group_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get groupArn() {
        return this.getStringAttribute('group_arn');
    }
    // group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get groupName() {
        return this.getStringAttribute('group_name');
    }
}
exports.IotThingGroupMetadataRootToParentGroups = IotThingGroupMetadataRootToParentGroups;
_a = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadataRootToParentGroups[_a] = { fqn: "@cdktf/provider-aws.iot.IotThingGroupMetadataRootToParentGroups", version: "3.0.1" };
/**
 * @stability stable
 */
class IotThingGroupMetadata extends cdktf.ComplexComputedList {
    // creation_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    // parent_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get parentGroupName() {
        return this.getStringAttribute('parent_group_name');
    }
    // root_to_parent_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rootToParentGroups() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('root_to_parent_groups');
    }
}
exports.IotThingGroupMetadata = IotThingGroupMetadata;
_b = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadata[_b] = { fqn: "@cdktf/provider-aws.iot.IotThingGroupMetadata", version: "3.0.1" };
function iotThingGroupPropertiesAttributePayloadToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        attributes: cdktf.hashMapper(cdktf.anyToTerraform)(struct.attributes),
    };
}
exports.iotThingGroupPropertiesAttributePayloadToTerraform = iotThingGroupPropertiesAttributePayloadToTerraform;
/**
 * @stability stable
 */
class IotThingGroupPropertiesAttributePayloadOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._attributes) {
            hasAnyValues = true;
            internalValueResult.attributes = this._attributes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._attributes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._attributes = value.attributes;
        }
    }
    /**
     * @stability stable
     */
    get attributes() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('attributes');
    }
    /**
     * @stability stable
     */
    set attributes(value) {
        this._attributes = value;
    }
    /**
     * @stability stable
     */
    resetAttributes() {
        this._attributes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get attributesInput() {
        return this._attributes;
    }
}
exports.IotThingGroupPropertiesAttributePayloadOutputReference = IotThingGroupPropertiesAttributePayloadOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IotThingGroupPropertiesAttributePayloadOutputReference[_c] = { fqn: "@cdktf/provider-aws.iot.IotThingGroupPropertiesAttributePayloadOutputReference", version: "3.0.1" };
function iotThingGroupPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        attribute_payload: iotThingGroupPropertiesAttributePayloadToTerraform(struct.attributePayload),
    };
}
exports.iotThingGroupPropertiesToTerraform = iotThingGroupPropertiesToTerraform;
/**
 * @stability stable
 */
class IotThingGroupPropertiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // attribute_payload - computed: false, optional: true, required: false
        this._attributePayload = new IotThingGroupPropertiesAttributePayloadOutputReference(this, "attribute_payload", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _f, _g;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if ((_f = this._attributePayload) === null || _f === void 0 ? void 0 : _f.internalValue) {
            hasAnyValues = true;
            internalValueResult.attributePayload = (_g = this._attributePayload) === null || _g === void 0 ? void 0 : _g.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._attributePayload.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._attributePayload.internalValue = value.attributePayload;
        }
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get attributePayload() {
        return this._attributePayload;
    }
    /**
     * @stability stable
     */
    putAttributePayload(value) {
        this._attributePayload.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAttributePayload() {
        this._attributePayload.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get attributePayloadInput() {
        return this._attributePayload.internalValue;
    }
}
exports.IotThingGroupPropertiesOutputReference = IotThingGroupPropertiesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IotThingGroupPropertiesOutputReference[_d] = { fqn: "@cdktf/provider-aws.iot.IotThingGroupPropertiesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group.html aws_iot_thing_group}.
 *
 * @stability stable
 */
class IotThingGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group.html aws_iot_thing_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // properties - computed: false, optional: true, required: false
        this._properties = new IotThingGroupPropertiesOutputReference(this, "properties", true);
        this._name = config.name;
        this._parentGroupName = config.parentGroupName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._properties.internalValue = config.properties;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // metadata - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    metadata(index) {
        return new IotThingGroupMetadata(this, 'metadata', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get parentGroupName() {
        return this.getStringAttribute('parent_group_name');
    }
    /**
     * @stability stable
     */
    set parentGroupName(value) {
        this._parentGroupName = value;
    }
    /**
     * @stability stable
     */
    resetParentGroupName() {
        this._parentGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parentGroupNameInput() {
        return this._parentGroupName;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getNumberAttribute('version');
    }
    /**
     * @stability stable
     */
    get properties() {
        return this._properties;
    }
    /**
     * @stability stable
     */
    putProperties(value) {
        this._properties.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetProperties() {
        this._properties.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get propertiesInput() {
        return this._properties.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            parent_group_name: cdktf.stringToTerraform(this._parentGroupName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            properties: iotThingGroupPropertiesToTerraform(this._properties.internalValue),
        };
    }
}
exports.IotThingGroup = IotThingGroup;
_e = JSII_RTTI_SYMBOL_1;
IotThingGroup[_e] = { fqn: "@cdktf/provider-aws.iot.IotThingGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotThingGroup.tfResourceType = "aws_iot_thing_group";
