// @cdktf/provider-aws/lib/synthetics/synthetics-canary.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsCanary = exports.SyntheticsCanaryVpcConfigOutputReference = exports.syntheticsCanaryVpcConfigToTerraform = exports.SyntheticsCanaryScheduleOutputReference = exports.syntheticsCanaryScheduleToTerraform = exports.SyntheticsCanaryRunConfigOutputReference = exports.syntheticsCanaryRunConfigToTerraform = exports.SyntheticsCanaryArtifactConfigOutputReference = exports.syntheticsCanaryArtifactConfigToTerraform = exports.SyntheticsCanaryArtifactConfigS3EncryptionOutputReference = exports.syntheticsCanaryArtifactConfigS3EncryptionToTerraform = exports.SyntheticsCanaryTimeline = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class SyntheticsCanaryTimeline extends cdktf.ComplexComputedList {
    // created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get created() {
        return this.getStringAttribute('created');
    }
    // last_modified - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    // last_started - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastStarted() {
        return this.getStringAttribute('last_started');
    }
    // last_stopped - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastStopped() {
        return this.getStringAttribute('last_stopped');
    }
}
exports.SyntheticsCanaryTimeline = SyntheticsCanaryTimeline;
_a = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryTimeline[_a] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryTimeline", version: "3.0.1" };
function syntheticsCanaryArtifactConfigS3EncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_mode: cdktf.stringToTerraform(struct.encryptionMode),
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
    };
}
exports.syntheticsCanaryArtifactConfigS3EncryptionToTerraform = syntheticsCanaryArtifactConfigS3EncryptionToTerraform;
/**
 * @stability stable
 */
class SyntheticsCanaryArtifactConfigS3EncryptionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionMode) {
            hasAnyValues = true;
            internalValueResult.encryptionMode = this._encryptionMode;
        }
        if (this._kmsKeyArn) {
            hasAnyValues = true;
            internalValueResult.kmsKeyArn = this._kmsKeyArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionMode = undefined;
            this._kmsKeyArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionMode = value.encryptionMode;
            this._kmsKeyArn = value.kmsKeyArn;
        }
    }
    /**
     * @stability stable
     */
    get encryptionMode() {
        return this.getStringAttribute('encryption_mode');
    }
    /**
     * @stability stable
     */
    set encryptionMode(value) {
        this._encryptionMode = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionMode() {
        this._encryptionMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionModeInput() {
        return this._encryptionMode;
    }
    /**
     * @stability stable
     */
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    /**
     * @stability stable
     */
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
}
exports.SyntheticsCanaryArtifactConfigS3EncryptionOutputReference = SyntheticsCanaryArtifactConfigS3EncryptionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryArtifactConfigS3EncryptionOutputReference[_b] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryArtifactConfigS3EncryptionOutputReference", version: "3.0.1" };
function syntheticsCanaryArtifactConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_encryption: syntheticsCanaryArtifactConfigS3EncryptionToTerraform(struct.s3Encryption),
    };
}
exports.syntheticsCanaryArtifactConfigToTerraform = syntheticsCanaryArtifactConfigToTerraform;
/**
 * @stability stable
 */
class SyntheticsCanaryArtifactConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // s3_encryption - computed: false, optional: true, required: false
        this._s3Encryption = new SyntheticsCanaryArtifactConfigS3EncryptionOutputReference(this, "s3_encryption", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _h, _j;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_h = this._s3Encryption) === null || _h === void 0 ? void 0 : _h.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3Encryption = (_j = this._s3Encryption) === null || _j === void 0 ? void 0 : _j.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3Encryption.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3Encryption.internalValue = value.s3Encryption;
        }
    }
    /**
     * @stability stable
     */
    get s3Encryption() {
        return this._s3Encryption;
    }
    /**
     * @stability stable
     */
    putS3Encryption(value) {
        this._s3Encryption.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetS3Encryption() {
        this._s3Encryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3EncryptionInput() {
        return this._s3Encryption.internalValue;
    }
}
exports.SyntheticsCanaryArtifactConfigOutputReference = SyntheticsCanaryArtifactConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryArtifactConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryArtifactConfigOutputReference", version: "3.0.1" };
function syntheticsCanaryRunConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        active_tracing: cdktf.booleanToTerraform(struct.activeTracing),
        memory_in_mb: cdktf.numberToTerraform(struct.memoryInMb),
        timeout_in_seconds: cdktf.numberToTerraform(struct.timeoutInSeconds),
    };
}
exports.syntheticsCanaryRunConfigToTerraform = syntheticsCanaryRunConfigToTerraform;
/**
 * @stability stable
 */
class SyntheticsCanaryRunConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._activeTracing) {
            hasAnyValues = true;
            internalValueResult.activeTracing = this._activeTracing;
        }
        if (this._memoryInMb) {
            hasAnyValues = true;
            internalValueResult.memoryInMb = this._memoryInMb;
        }
        if (this._timeoutInSeconds) {
            hasAnyValues = true;
            internalValueResult.timeoutInSeconds = this._timeoutInSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._activeTracing = undefined;
            this._memoryInMb = undefined;
            this._timeoutInSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._activeTracing = value.activeTracing;
            this._memoryInMb = value.memoryInMb;
            this._timeoutInSeconds = value.timeoutInSeconds;
        }
    }
    /**
     * @stability stable
     */
    get activeTracing() {
        return this.getBooleanAttribute('active_tracing');
    }
    /**
     * @stability stable
     */
    set activeTracing(value) {
        this._activeTracing = value;
    }
    /**
     * @stability stable
     */
    resetActiveTracing() {
        this._activeTracing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get activeTracingInput() {
        return this._activeTracing;
    }
    /**
     * @stability stable
     */
    get memoryInMb() {
        return this.getNumberAttribute('memory_in_mb');
    }
    /**
     * @stability stable
     */
    set memoryInMb(value) {
        this._memoryInMb = value;
    }
    /**
     * @stability stable
     */
    resetMemoryInMb() {
        this._memoryInMb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get memoryInMbInput() {
        return this._memoryInMb;
    }
    /**
     * @stability stable
     */
    get timeoutInSeconds() {
        return this.getNumberAttribute('timeout_in_seconds');
    }
    /**
     * @stability stable
     */
    set timeoutInSeconds(value) {
        this._timeoutInSeconds = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutInSeconds() {
        this._timeoutInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutInSecondsInput() {
        return this._timeoutInSeconds;
    }
}
exports.SyntheticsCanaryRunConfigOutputReference = SyntheticsCanaryRunConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryRunConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryRunConfigOutputReference", version: "3.0.1" };
function syntheticsCanaryScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        duration_in_seconds: cdktf.numberToTerraform(struct.durationInSeconds),
        expression: cdktf.stringToTerraform(struct.expression),
    };
}
exports.syntheticsCanaryScheduleToTerraform = syntheticsCanaryScheduleToTerraform;
/**
 * @stability stable
 */
class SyntheticsCanaryScheduleOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._durationInSeconds) {
            hasAnyValues = true;
            internalValueResult.durationInSeconds = this._durationInSeconds;
        }
        if (this._expression) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._durationInSeconds = undefined;
            this._expression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._durationInSeconds = value.durationInSeconds;
            this._expression = value.expression;
        }
    }
    /**
     * @stability stable
     */
    get durationInSeconds() {
        return this.getNumberAttribute('duration_in_seconds');
    }
    /**
     * @stability stable
     */
    set durationInSeconds(value) {
        this._durationInSeconds = value;
    }
    /**
     * @stability stable
     */
    resetDurationInSeconds() {
        this._durationInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get durationInSecondsInput() {
        return this._durationInSeconds;
    }
    /**
     * @stability stable
     */
    get expression() {
        return this.getStringAttribute('expression');
    }
    /**
     * @stability stable
     */
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get expressionInput() {
        return this._expression;
    }
}
exports.SyntheticsCanaryScheduleOutputReference = SyntheticsCanaryScheduleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryScheduleOutputReference[_e] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryScheduleOutputReference", version: "3.0.1" };
function syntheticsCanaryVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnetIds),
    };
}
exports.syntheticsCanaryVpcConfigToTerraform = syntheticsCanaryVpcConfigToTerraform;
/**
 * @stability stable
 */
class SyntheticsCanaryVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    /**
     * @stability stable
     */
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.SyntheticsCanaryVpcConfigOutputReference = SyntheticsCanaryVpcConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryVpcConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanaryVpcConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/synthetics_canary.html aws_synthetics_canary}.
 *
 * @stability stable
 */
class SyntheticsCanary extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/synthetics_canary.html aws_synthetics_canary} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_synthetics_canary',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // artifact_config - computed: false, optional: true, required: false
        this._artifactConfig = new SyntheticsCanaryArtifactConfigOutputReference(this, "artifact_config", true);
        // run_config - computed: false, optional: true, required: false
        this._runConfig = new SyntheticsCanaryRunConfigOutputReference(this, "run_config", true);
        // schedule - computed: false, optional: false, required: true
        this._schedule = new SyntheticsCanaryScheduleOutputReference(this, "schedule", true);
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new SyntheticsCanaryVpcConfigOutputReference(this, "vpc_config", true);
        this._artifactS3Location = config.artifactS3Location;
        this._executionRoleArn = config.executionRoleArn;
        this._failureRetentionPeriod = config.failureRetentionPeriod;
        this._handler = config.handler;
        this._name = config.name;
        this._runtimeVersion = config.runtimeVersion;
        this._s3Bucket = config.s3Bucket;
        this._s3Key = config.s3Key;
        this._s3Version = config.s3Version;
        this._startCanary = config.startCanary;
        this._successRetentionPeriod = config.successRetentionPeriod;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._zipFile = config.zipFile;
        this._artifactConfig.internalValue = config.artifactConfig;
        this._runConfig.internalValue = config.runConfig;
        this._schedule.internalValue = config.schedule;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get artifactS3Location() {
        return this.getStringAttribute('artifact_s3_location');
    }
    /**
     * @stability stable
     */
    set artifactS3Location(value) {
        this._artifactS3Location = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get artifactS3LocationInput() {
        return this._artifactS3Location;
    }
    // engine_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engineArn() {
        return this.getStringAttribute('engine_arn');
    }
    /**
     * @stability stable
     */
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    /**
     * @stability stable
     */
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    /**
     * @stability stable
     */
    get failureRetentionPeriod() {
        return this.getNumberAttribute('failure_retention_period');
    }
    /**
     * @stability stable
     */
    set failureRetentionPeriod(value) {
        this._failureRetentionPeriod = value;
    }
    /**
     * @stability stable
     */
    resetFailureRetentionPeriod() {
        this._failureRetentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get failureRetentionPeriodInput() {
        return this._failureRetentionPeriod;
    }
    /**
     * @stability stable
     */
    get handler() {
        return this.getStringAttribute('handler');
    }
    /**
     * @stability stable
     */
    set handler(value) {
        this._handler = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get handlerInput() {
        return this._handler;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get runtimeVersion() {
        return this.getStringAttribute('runtime_version');
    }
    /**
     * @stability stable
     */
    set runtimeVersion(value) {
        this._runtimeVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runtimeVersionInput() {
        return this._runtimeVersion;
    }
    /**
     * @stability stable
     */
    get s3Bucket() {
        return this.getStringAttribute('s3_bucket');
    }
    /**
     * @stability stable
     */
    set s3Bucket(value) {
        this._s3Bucket = value;
    }
    /**
     * @stability stable
     */
    resetS3Bucket() {
        this._s3Bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketInput() {
        return this._s3Bucket;
    }
    /**
     * @stability stable
     */
    get s3Key() {
        return this.getStringAttribute('s3_key');
    }
    /**
     * @stability stable
     */
    set s3Key(value) {
        this._s3Key = value;
    }
    /**
     * @stability stable
     */
    resetS3Key() {
        this._s3Key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KeyInput() {
        return this._s3Key;
    }
    /**
     * @stability stable
     */
    get s3Version() {
        return this.getStringAttribute('s3_version');
    }
    /**
     * @stability stable
     */
    set s3Version(value) {
        this._s3Version = value;
    }
    /**
     * @stability stable
     */
    resetS3Version() {
        this._s3Version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3VersionInput() {
        return this._s3Version;
    }
    // source_location_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceLocationArn() {
        return this.getStringAttribute('source_location_arn');
    }
    /**
     * @stability stable
     */
    get startCanary() {
        return this.getBooleanAttribute('start_canary');
    }
    /**
     * @stability stable
     */
    set startCanary(value) {
        this._startCanary = value;
    }
    /**
     * @stability stable
     */
    resetStartCanary() {
        this._startCanary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startCanaryInput() {
        return this._startCanary;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get successRetentionPeriod() {
        return this.getNumberAttribute('success_retention_period');
    }
    /**
     * @stability stable
     */
    set successRetentionPeriod(value) {
        this._successRetentionPeriod = value;
    }
    /**
     * @stability stable
     */
    resetSuccessRetentionPeriod() {
        this._successRetentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get successRetentionPeriodInput() {
        return this._successRetentionPeriod;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // timeline - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    timeline(index) {
        return new SyntheticsCanaryTimeline(this, 'timeline', index);
    }
    /**
     * @stability stable
     */
    get zipFile() {
        return this.getStringAttribute('zip_file');
    }
    /**
     * @stability stable
     */
    set zipFile(value) {
        this._zipFile = value;
    }
    /**
     * @stability stable
     */
    resetZipFile() {
        this._zipFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get zipFileInput() {
        return this._zipFile;
    }
    /**
     * @stability stable
     */
    get artifactConfig() {
        return this._artifactConfig;
    }
    /**
     * @stability stable
     */
    putArtifactConfig(value) {
        this._artifactConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetArtifactConfig() {
        this._artifactConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get artifactConfigInput() {
        return this._artifactConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get runConfig() {
        return this._runConfig;
    }
    /**
     * @stability stable
     */
    putRunConfig(value) {
        this._runConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRunConfig() {
        this._runConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runConfigInput() {
        return this._runConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get schedule() {
        return this._schedule;
    }
    /**
     * @stability stable
     */
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    /**
     * @stability stable
     */
    get vpcConfig() {
        return this._vpcConfig;
    }
    /**
     * @stability stable
     */
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            artifact_s3_location: cdktf.stringToTerraform(this._artifactS3Location),
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            failure_retention_period: cdktf.numberToTerraform(this._failureRetentionPeriod),
            handler: cdktf.stringToTerraform(this._handler),
            name: cdktf.stringToTerraform(this._name),
            runtime_version: cdktf.stringToTerraform(this._runtimeVersion),
            s3_bucket: cdktf.stringToTerraform(this._s3Bucket),
            s3_key: cdktf.stringToTerraform(this._s3Key),
            s3_version: cdktf.stringToTerraform(this._s3Version),
            start_canary: cdktf.booleanToTerraform(this._startCanary),
            success_retention_period: cdktf.numberToTerraform(this._successRetentionPeriod),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            zip_file: cdktf.stringToTerraform(this._zipFile),
            artifact_config: syntheticsCanaryArtifactConfigToTerraform(this._artifactConfig.internalValue),
            run_config: syntheticsCanaryRunConfigToTerraform(this._runConfig.internalValue),
            schedule: syntheticsCanaryScheduleToTerraform(this._schedule.internalValue),
            vpc_config: syntheticsCanaryVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
}
exports.SyntheticsCanary = SyntheticsCanary;
_g = JSII_RTTI_SYMBOL_1;
SyntheticsCanary[_g] = { fqn: "@cdktf/provider-aws.synthetics.SyntheticsCanary", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SyntheticsCanary.tfResourceType = "aws_synthetics_canary";
