// @cdktf/provider-aws/lib/ecs/data-aws-ecs-cluster.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcsCluster = exports.DataAwsEcsClusterSetting = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsEcsClusterSetting extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsEcsClusterSetting = DataAwsEcsClusterSetting;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcsClusterSetting[_a] = { fqn: "@cdktf/provider-aws.ecs.DataAwsEcsClusterSetting", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecs_cluster.html aws_ecs_cluster}.
 *
 * @stability stable
 */
class DataAwsEcsCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecs_cluster.html aws_ecs_cluster} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterName = config.clusterName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // pending_tasks_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get pendingTasksCount() {
        return this.getNumberAttribute('pending_tasks_count');
    }
    // registered_container_instances_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get registeredContainerInstancesCount() {
        return this.getNumberAttribute('registered_container_instances_count');
    }
    // running_tasks_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get runningTasksCount() {
        return this.getNumberAttribute('running_tasks_count');
    }
    // setting - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    setting(index) {
        return new DataAwsEcsClusterSetting(this, 'setting', index);
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
        };
    }
}
exports.DataAwsEcsCluster = DataAwsEcsCluster;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEcsCluster[_b] = { fqn: "@cdktf/provider-aws.ecs.DataAwsEcsCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEcsCluster.tfResourceType = "aws_ecs_cluster";
