// @cdktf/provider-aws/lib/ecs/data-aws-ecs-container-definition.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcsContainerDefinition = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecs_container_definition.html aws_ecs_container_definition}.
 *
 * @stability stable
 */
class DataAwsEcsContainerDefinition extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecs_container_definition.html aws_ecs_container_definition} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_container_definition',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._containerName = config.containerName;
        this._taskDefinition = config.taskDefinition;
    }
    /**
     * @stability stable
     */
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    /**
     * @stability stable
     */
    set containerName(value) {
        this._containerName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get containerNameInput() {
        return this._containerName;
    }
    // cpu - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cpu() {
        return this.getNumberAttribute('cpu');
    }
    // disable_networking - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get disableNetworking() {
        return this.getBooleanAttribute('disable_networking');
    }
    // docker_labels - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    dockerLabels(key) {
        return new cdktf.StringMap(this, 'docker_labels').lookup(key);
    }
    // environment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    environment(key) {
        return new cdktf.StringMap(this, 'environment').lookup(key);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get image() {
        return this.getStringAttribute('image');
    }
    // image_digest - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageDigest() {
        return this.getStringAttribute('image_digest');
    }
    // memory - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get memory() {
        return this.getNumberAttribute('memory');
    }
    // memory_reservation - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get memoryReservation() {
        return this.getNumberAttribute('memory_reservation');
    }
    /**
     * @stability stable
     */
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    /**
     * @stability stable
     */
    set taskDefinition(value) {
        this._taskDefinition = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskDefinitionInput() {
        return this._taskDefinition;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            container_name: cdktf.stringToTerraform(this._containerName),
            task_definition: cdktf.stringToTerraform(this._taskDefinition),
        };
    }
}
exports.DataAwsEcsContainerDefinition = DataAwsEcsContainerDefinition;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcsContainerDefinition[_a] = { fqn: "@cdktf/provider-aws.ecs.DataAwsEcsContainerDefinition", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEcsContainerDefinition.tfResourceType = "aws_ecs_container_definition";
