// @cdktf/provider-aws/lib/wafv2/wafv2-web-acl-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wafv2WebAclAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafv2_web_acl_association.html aws_wafv2_web_acl_association}.
 *
 * @stability stable
 */
class Wafv2WebAclAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafv2_web_acl_association.html aws_wafv2_web_acl_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafv2_web_acl_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._resourceArn = config.resourceArn;
        this._webAclArn = config.webAclArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get webAclArn() {
        return this.getStringAttribute('web_acl_arn');
    }
    /**
     * @stability stable
     */
    set webAclArn(value) {
        this._webAclArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get webAclArnInput() {
        return this._webAclArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
            web_acl_arn: cdktf.stringToTerraform(this._webAclArn),
        };
    }
}
exports.Wafv2WebAclAssociation = Wafv2WebAclAssociation;
_a = JSII_RTTI_SYMBOL_1;
Wafv2WebAclAssociation[_a] = { fqn: "@cdktf/provider-aws.wafv2.Wafv2WebAclAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Wafv2WebAclAssociation.tfResourceType = "aws_wafv2_web_acl_association";
