// @cdktf/provider-aws/lib/lex/lex-intent.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LexIntent = exports.LexIntentTimeoutsOutputReference = exports.lexIntentTimeoutsToTerraform = exports.lexIntentSlotToTerraform = exports.LexIntentSlotValueElicitationPromptOutputReference = exports.lexIntentSlotValueElicitationPromptToTerraform = exports.lexIntentSlotValueElicitationPromptMessageToTerraform = exports.LexIntentRejectionStatementOutputReference = exports.lexIntentRejectionStatementToTerraform = exports.lexIntentRejectionStatementMessageToTerraform = exports.LexIntentFulfillmentActivityOutputReference = exports.lexIntentFulfillmentActivityToTerraform = exports.LexIntentFulfillmentActivityCodeHookOutputReference = exports.lexIntentFulfillmentActivityCodeHookToTerraform = exports.LexIntentFollowUpPromptOutputReference = exports.lexIntentFollowUpPromptToTerraform = exports.LexIntentFollowUpPromptRejectionStatementOutputReference = exports.lexIntentFollowUpPromptRejectionStatementToTerraform = exports.lexIntentFollowUpPromptRejectionStatementMessageToTerraform = exports.LexIntentFollowUpPromptPromptOutputReference = exports.lexIntentFollowUpPromptPromptToTerraform = exports.lexIntentFollowUpPromptPromptMessageToTerraform = exports.LexIntentDialogCodeHookOutputReference = exports.lexIntentDialogCodeHookToTerraform = exports.LexIntentConfirmationPromptOutputReference = exports.lexIntentConfirmationPromptToTerraform = exports.lexIntentConfirmationPromptMessageToTerraform = exports.LexIntentConclusionStatementOutputReference = exports.lexIntentConclusionStatementToTerraform = exports.lexIntentConclusionStatementMessageToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lexIntentConclusionStatementMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentConclusionStatementMessageToTerraform = lexIntentConclusionStatementMessageToTerraform;
function lexIntentConclusionStatementToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentConclusionStatementMessageToTerraform)(struct.message),
    };
}
exports.lexIntentConclusionStatementToTerraform = lexIntentConclusionStatementToTerraform;
/**
 * @stability stable
 */
class LexIntentConclusionStatementOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentConclusionStatementOutputReference = LexIntentConclusionStatementOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LexIntentConclusionStatementOutputReference[_a] = { fqn: "@cdktf/provider-aws.lex.LexIntentConclusionStatementOutputReference", version: "3.0.1" };
function lexIntentConfirmationPromptMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentConfirmationPromptMessageToTerraform = lexIntentConfirmationPromptMessageToTerraform;
function lexIntentConfirmationPromptToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_attempts: cdktf.numberToTerraform(struct.maxAttempts),
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentConfirmationPromptMessageToTerraform)(struct.message),
    };
}
exports.lexIntentConfirmationPromptToTerraform = lexIntentConfirmationPromptToTerraform;
/**
 * @stability stable
 */
class LexIntentConfirmationPromptOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxAttempts) {
            hasAnyValues = true;
            internalValueResult.maxAttempts = this._maxAttempts;
        }
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxAttempts = undefined;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxAttempts = value.maxAttempts;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get maxAttempts() {
        return this.getNumberAttribute('max_attempts');
    }
    /**
     * @stability stable
     */
    set maxAttempts(value) {
        this._maxAttempts = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxAttemptsInput() {
        return this._maxAttempts;
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentConfirmationPromptOutputReference = LexIntentConfirmationPromptOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LexIntentConfirmationPromptOutputReference[_b] = { fqn: "@cdktf/provider-aws.lex.LexIntentConfirmationPromptOutputReference", version: "3.0.1" };
function lexIntentDialogCodeHookToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_version: cdktf.stringToTerraform(struct.messageVersion),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.lexIntentDialogCodeHookToTerraform = lexIntentDialogCodeHookToTerraform;
/**
 * @stability stable
 */
class LexIntentDialogCodeHookOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageVersion) {
            hasAnyValues = true;
            internalValueResult.messageVersion = this._messageVersion;
        }
        if (this._uri) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._messageVersion = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._messageVersion = value.messageVersion;
            this._uri = value.uri;
        }
    }
    /**
     * @stability stable
     */
    get messageVersion() {
        return this.getStringAttribute('message_version');
    }
    /**
     * @stability stable
     */
    set messageVersion(value) {
        this._messageVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageVersionInput() {
        return this._messageVersion;
    }
    /**
     * @stability stable
     */
    get uri() {
        return this.getStringAttribute('uri');
    }
    /**
     * @stability stable
     */
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get uriInput() {
        return this._uri;
    }
}
exports.LexIntentDialogCodeHookOutputReference = LexIntentDialogCodeHookOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LexIntentDialogCodeHookOutputReference[_c] = { fqn: "@cdktf/provider-aws.lex.LexIntentDialogCodeHookOutputReference", version: "3.0.1" };
function lexIntentFollowUpPromptPromptMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentFollowUpPromptPromptMessageToTerraform = lexIntentFollowUpPromptPromptMessageToTerraform;
function lexIntentFollowUpPromptPromptToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_attempts: cdktf.numberToTerraform(struct.maxAttempts),
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentFollowUpPromptPromptMessageToTerraform)(struct.message),
    };
}
exports.lexIntentFollowUpPromptPromptToTerraform = lexIntentFollowUpPromptPromptToTerraform;
/**
 * @stability stable
 */
class LexIntentFollowUpPromptPromptOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxAttempts) {
            hasAnyValues = true;
            internalValueResult.maxAttempts = this._maxAttempts;
        }
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxAttempts = undefined;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxAttempts = value.maxAttempts;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get maxAttempts() {
        return this.getNumberAttribute('max_attempts');
    }
    /**
     * @stability stable
     */
    set maxAttempts(value) {
        this._maxAttempts = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxAttemptsInput() {
        return this._maxAttempts;
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentFollowUpPromptPromptOutputReference = LexIntentFollowUpPromptPromptOutputReference;
_d = JSII_RTTI_SYMBOL_1;
LexIntentFollowUpPromptPromptOutputReference[_d] = { fqn: "@cdktf/provider-aws.lex.LexIntentFollowUpPromptPromptOutputReference", version: "3.0.1" };
function lexIntentFollowUpPromptRejectionStatementMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentFollowUpPromptRejectionStatementMessageToTerraform = lexIntentFollowUpPromptRejectionStatementMessageToTerraform;
function lexIntentFollowUpPromptRejectionStatementToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentFollowUpPromptRejectionStatementMessageToTerraform)(struct.message),
    };
}
exports.lexIntentFollowUpPromptRejectionStatementToTerraform = lexIntentFollowUpPromptRejectionStatementToTerraform;
/**
 * @stability stable
 */
class LexIntentFollowUpPromptRejectionStatementOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentFollowUpPromptRejectionStatementOutputReference = LexIntentFollowUpPromptRejectionStatementOutputReference;
_e = JSII_RTTI_SYMBOL_1;
LexIntentFollowUpPromptRejectionStatementOutputReference[_e] = { fqn: "@cdktf/provider-aws.lex.LexIntentFollowUpPromptRejectionStatementOutputReference", version: "3.0.1" };
function lexIntentFollowUpPromptToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prompt: lexIntentFollowUpPromptPromptToTerraform(struct.prompt),
        rejection_statement: lexIntentFollowUpPromptRejectionStatementToTerraform(struct.rejectionStatement),
    };
}
exports.lexIntentFollowUpPromptToTerraform = lexIntentFollowUpPromptToTerraform;
/**
 * @stability stable
 */
class LexIntentFollowUpPromptOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // prompt - computed: false, optional: false, required: true
        this._prompt = new LexIntentFollowUpPromptPromptOutputReference(this, "prompt", true);
        // rejection_statement - computed: false, optional: false, required: true
        this._rejectionStatement = new LexIntentFollowUpPromptRejectionStatementOutputReference(this, "rejection_statement", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _o, _p, _q, _r;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_o = this._prompt) === null || _o === void 0 ? void 0 : _o.internalValue) {
            hasAnyValues = true;
            internalValueResult.prompt = (_p = this._prompt) === null || _p === void 0 ? void 0 : _p.internalValue;
        }
        if ((_q = this._rejectionStatement) === null || _q === void 0 ? void 0 : _q.internalValue) {
            hasAnyValues = true;
            internalValueResult.rejectionStatement = (_r = this._rejectionStatement) === null || _r === void 0 ? void 0 : _r.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prompt.internalValue = undefined;
            this._rejectionStatement.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prompt.internalValue = value.prompt;
            this._rejectionStatement.internalValue = value.rejectionStatement;
        }
    }
    /**
     * @stability stable
     */
    get prompt() {
        return this._prompt;
    }
    /**
     * @stability stable
     */
    putPrompt(value) {
        this._prompt.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get promptInput() {
        return this._prompt.internalValue;
    }
    /**
     * @stability stable
     */
    get rejectionStatement() {
        return this._rejectionStatement;
    }
    /**
     * @stability stable
     */
    putRejectionStatement(value) {
        this._rejectionStatement.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rejectionStatementInput() {
        return this._rejectionStatement.internalValue;
    }
}
exports.LexIntentFollowUpPromptOutputReference = LexIntentFollowUpPromptOutputReference;
_f = JSII_RTTI_SYMBOL_1;
LexIntentFollowUpPromptOutputReference[_f] = { fqn: "@cdktf/provider-aws.lex.LexIntentFollowUpPromptOutputReference", version: "3.0.1" };
function lexIntentFulfillmentActivityCodeHookToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_version: cdktf.stringToTerraform(struct.messageVersion),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.lexIntentFulfillmentActivityCodeHookToTerraform = lexIntentFulfillmentActivityCodeHookToTerraform;
/**
 * @stability stable
 */
class LexIntentFulfillmentActivityCodeHookOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageVersion) {
            hasAnyValues = true;
            internalValueResult.messageVersion = this._messageVersion;
        }
        if (this._uri) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._messageVersion = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._messageVersion = value.messageVersion;
            this._uri = value.uri;
        }
    }
    /**
     * @stability stable
     */
    get messageVersion() {
        return this.getStringAttribute('message_version');
    }
    /**
     * @stability stable
     */
    set messageVersion(value) {
        this._messageVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageVersionInput() {
        return this._messageVersion;
    }
    /**
     * @stability stable
     */
    get uri() {
        return this.getStringAttribute('uri');
    }
    /**
     * @stability stable
     */
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get uriInput() {
        return this._uri;
    }
}
exports.LexIntentFulfillmentActivityCodeHookOutputReference = LexIntentFulfillmentActivityCodeHookOutputReference;
_g = JSII_RTTI_SYMBOL_1;
LexIntentFulfillmentActivityCodeHookOutputReference[_g] = { fqn: "@cdktf/provider-aws.lex.LexIntentFulfillmentActivityCodeHookOutputReference", version: "3.0.1" };
function lexIntentFulfillmentActivityToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        code_hook: lexIntentFulfillmentActivityCodeHookToTerraform(struct.codeHook),
    };
}
exports.lexIntentFulfillmentActivityToTerraform = lexIntentFulfillmentActivityToTerraform;
/**
 * @stability stable
 */
class LexIntentFulfillmentActivityOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // code_hook - computed: false, optional: true, required: false
        this._codeHook = new LexIntentFulfillmentActivityCodeHookOutputReference(this, "code_hook", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _o, _p;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if ((_o = this._codeHook) === null || _o === void 0 ? void 0 : _o.internalValue) {
            hasAnyValues = true;
            internalValueResult.codeHook = (_p = this._codeHook) === null || _p === void 0 ? void 0 : _p.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._codeHook.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._codeHook.internalValue = value.codeHook;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get codeHook() {
        return this._codeHook;
    }
    /**
     * @stability stable
     */
    putCodeHook(value) {
        this._codeHook.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCodeHook() {
        this._codeHook.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get codeHookInput() {
        return this._codeHook.internalValue;
    }
}
exports.LexIntentFulfillmentActivityOutputReference = LexIntentFulfillmentActivityOutputReference;
_h = JSII_RTTI_SYMBOL_1;
LexIntentFulfillmentActivityOutputReference[_h] = { fqn: "@cdktf/provider-aws.lex.LexIntentFulfillmentActivityOutputReference", version: "3.0.1" };
function lexIntentRejectionStatementMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentRejectionStatementMessageToTerraform = lexIntentRejectionStatementMessageToTerraform;
function lexIntentRejectionStatementToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentRejectionStatementMessageToTerraform)(struct.message),
    };
}
exports.lexIntentRejectionStatementToTerraform = lexIntentRejectionStatementToTerraform;
/**
 * @stability stable
 */
class LexIntentRejectionStatementOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentRejectionStatementOutputReference = LexIntentRejectionStatementOutputReference;
_j = JSII_RTTI_SYMBOL_1;
LexIntentRejectionStatementOutputReference[_j] = { fqn: "@cdktf/provider-aws.lex.LexIntentRejectionStatementOutputReference", version: "3.0.1" };
function lexIntentSlotValueElicitationPromptMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content: cdktf.stringToTerraform(struct.content),
        content_type: cdktf.stringToTerraform(struct.contentType),
        group_number: cdktf.numberToTerraform(struct.groupNumber),
    };
}
exports.lexIntentSlotValueElicitationPromptMessageToTerraform = lexIntentSlotValueElicitationPromptMessageToTerraform;
function lexIntentSlotValueElicitationPromptToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_attempts: cdktf.numberToTerraform(struct.maxAttempts),
        response_card: cdktf.stringToTerraform(struct.responseCard),
        message: cdktf.listMapper(lexIntentSlotValueElicitationPromptMessageToTerraform)(struct.message),
    };
}
exports.lexIntentSlotValueElicitationPromptToTerraform = lexIntentSlotValueElicitationPromptToTerraform;
/**
 * @stability stable
 */
class LexIntentSlotValueElicitationPromptOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxAttempts) {
            hasAnyValues = true;
            internalValueResult.maxAttempts = this._maxAttempts;
        }
        if (this._responseCard) {
            hasAnyValues = true;
            internalValueResult.responseCard = this._responseCard;
        }
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxAttempts = undefined;
            this._responseCard = undefined;
            this._message = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxAttempts = value.maxAttempts;
            this._responseCard = value.responseCard;
            this._message = value.message;
        }
    }
    /**
     * @stability stable
     */
    get maxAttempts() {
        return this.getNumberAttribute('max_attempts');
    }
    /**
     * @stability stable
     */
    set maxAttempts(value) {
        this._maxAttempts = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxAttemptsInput() {
        return this._maxAttempts;
    }
    /**
     * @stability stable
     */
    get responseCard() {
        return this.getStringAttribute('response_card');
    }
    /**
     * @stability stable
     */
    set responseCard(value) {
        this._responseCard = value;
    }
    /**
     * @stability stable
     */
    resetResponseCard() {
        this._responseCard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseCardInput() {
        return this._responseCard;
    }
    /**
     * @stability stable
     */
    get message() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
}
exports.LexIntentSlotValueElicitationPromptOutputReference = LexIntentSlotValueElicitationPromptOutputReference;
_k = JSII_RTTI_SYMBOL_1;
LexIntentSlotValueElicitationPromptOutputReference[_k] = { fqn: "@cdktf/provider-aws.lex.LexIntentSlotValueElicitationPromptOutputReference", version: "3.0.1" };
function lexIntentSlotToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        name: cdktf.stringToTerraform(struct.name),
        priority: cdktf.numberToTerraform(struct.priority),
        response_card: cdktf.stringToTerraform(struct.responseCard),
        sample_utterances: cdktf.listMapper(cdktf.stringToTerraform)(struct.sampleUtterances),
        slot_constraint: cdktf.stringToTerraform(struct.slotConstraint),
        slot_type: cdktf.stringToTerraform(struct.slotType),
        slot_type_version: cdktf.stringToTerraform(struct.slotTypeVersion),
        value_elicitation_prompt: lexIntentSlotValueElicitationPromptToTerraform(struct.valueElicitationPrompt),
    };
}
exports.lexIntentSlotToTerraform = lexIntentSlotToTerraform;
function lexIntentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lexIntentTimeoutsToTerraform = lexIntentTimeoutsToTerraform;
/**
 * @stability stable
 */
class LexIntentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.LexIntentTimeoutsOutputReference = LexIntentTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
LexIntentTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-aws.lex.LexIntentTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lex_intent.html aws_lex_intent}.
 *
 * @stability stable
 */
class LexIntent extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lex_intent.html aws_lex_intent} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lex_intent',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // conclusion_statement - computed: false, optional: true, required: false
        this._conclusionStatement = new LexIntentConclusionStatementOutputReference(this, "conclusion_statement", true);
        // confirmation_prompt - computed: false, optional: true, required: false
        this._confirmationPrompt = new LexIntentConfirmationPromptOutputReference(this, "confirmation_prompt", true);
        // dialog_code_hook - computed: false, optional: true, required: false
        this._dialogCodeHook = new LexIntentDialogCodeHookOutputReference(this, "dialog_code_hook", true);
        // follow_up_prompt - computed: false, optional: true, required: false
        this._followUpPrompt = new LexIntentFollowUpPromptOutputReference(this, "follow_up_prompt", true);
        // fulfillment_activity - computed: false, optional: false, required: true
        this._fulfillmentActivity = new LexIntentFulfillmentActivityOutputReference(this, "fulfillment_activity", true);
        // rejection_statement - computed: false, optional: true, required: false
        this._rejectionStatement = new LexIntentRejectionStatementOutputReference(this, "rejection_statement", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LexIntentTimeoutsOutputReference(this, "timeouts", true);
        this._createVersion = config.createVersion;
        this._description = config.description;
        this._name = config.name;
        this._parentIntentSignature = config.parentIntentSignature;
        this._sampleUtterances = config.sampleUtterances;
        this._conclusionStatement.internalValue = config.conclusionStatement;
        this._confirmationPrompt.internalValue = config.confirmationPrompt;
        this._dialogCodeHook.internalValue = config.dialogCodeHook;
        this._followUpPrompt.internalValue = config.followUpPrompt;
        this._fulfillmentActivity.internalValue = config.fulfillmentActivity;
        this._rejectionStatement.internalValue = config.rejectionStatement;
        this._slot = config.slot;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // checksum - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get checksum() {
        return this.getStringAttribute('checksum');
    }
    /**
     * @stability stable
     */
    get createVersion() {
        return this.getBooleanAttribute('create_version');
    }
    /**
     * @stability stable
     */
    set createVersion(value) {
        this._createVersion = value;
    }
    /**
     * @stability stable
     */
    resetCreateVersion() {
        this._createVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createVersionInput() {
        return this._createVersion;
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get parentIntentSignature() {
        return this.getStringAttribute('parent_intent_signature');
    }
    /**
     * @stability stable
     */
    set parentIntentSignature(value) {
        this._parentIntentSignature = value;
    }
    /**
     * @stability stable
     */
    resetParentIntentSignature() {
        this._parentIntentSignature = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parentIntentSignatureInput() {
        return this._parentIntentSignature;
    }
    /**
     * @stability stable
     */
    get sampleUtterances() {
        return this.getListAttribute('sample_utterances');
    }
    /**
     * @stability stable
     */
    set sampleUtterances(value) {
        this._sampleUtterances = value;
    }
    /**
     * @stability stable
     */
    resetSampleUtterances() {
        this._sampleUtterances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sampleUtterancesInput() {
        return this._sampleUtterances;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    get conclusionStatement() {
        return this._conclusionStatement;
    }
    /**
     * @stability stable
     */
    putConclusionStatement(value) {
        this._conclusionStatement.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConclusionStatement() {
        this._conclusionStatement.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get conclusionStatementInput() {
        return this._conclusionStatement.internalValue;
    }
    /**
     * @stability stable
     */
    get confirmationPrompt() {
        return this._confirmationPrompt;
    }
    /**
     * @stability stable
     */
    putConfirmationPrompt(value) {
        this._confirmationPrompt.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConfirmationPrompt() {
        this._confirmationPrompt.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get confirmationPromptInput() {
        return this._confirmationPrompt.internalValue;
    }
    /**
     * @stability stable
     */
    get dialogCodeHook() {
        return this._dialogCodeHook;
    }
    /**
     * @stability stable
     */
    putDialogCodeHook(value) {
        this._dialogCodeHook.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDialogCodeHook() {
        this._dialogCodeHook.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dialogCodeHookInput() {
        return this._dialogCodeHook.internalValue;
    }
    /**
     * @stability stable
     */
    get followUpPrompt() {
        return this._followUpPrompt;
    }
    /**
     * @stability stable
     */
    putFollowUpPrompt(value) {
        this._followUpPrompt.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFollowUpPrompt() {
        this._followUpPrompt.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get followUpPromptInput() {
        return this._followUpPrompt.internalValue;
    }
    /**
     * @stability stable
     */
    get fulfillmentActivity() {
        return this._fulfillmentActivity;
    }
    /**
     * @stability stable
     */
    putFulfillmentActivity(value) {
        this._fulfillmentActivity.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fulfillmentActivityInput() {
        return this._fulfillmentActivity.internalValue;
    }
    /**
     * @stability stable
     */
    get rejectionStatement() {
        return this._rejectionStatement;
    }
    /**
     * @stability stable
     */
    putRejectionStatement(value) {
        this._rejectionStatement.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRejectionStatement() {
        this._rejectionStatement.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rejectionStatementInput() {
        return this._rejectionStatement.internalValue;
    }
    /**
     * @stability stable
     */
    get slot() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('slot');
    }
    /**
     * @stability stable
     */
    set slot(value) {
        this._slot = value;
    }
    /**
     * @stability stable
     */
    resetSlot() {
        this._slot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get slotInput() {
        return this._slot;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            create_version: cdktf.booleanToTerraform(this._createVersion),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            parent_intent_signature: cdktf.stringToTerraform(this._parentIntentSignature),
            sample_utterances: cdktf.listMapper(cdktf.stringToTerraform)(this._sampleUtterances),
            conclusion_statement: lexIntentConclusionStatementToTerraform(this._conclusionStatement.internalValue),
            confirmation_prompt: lexIntentConfirmationPromptToTerraform(this._confirmationPrompt.internalValue),
            dialog_code_hook: lexIntentDialogCodeHookToTerraform(this._dialogCodeHook.internalValue),
            follow_up_prompt: lexIntentFollowUpPromptToTerraform(this._followUpPrompt.internalValue),
            fulfillment_activity: lexIntentFulfillmentActivityToTerraform(this._fulfillmentActivity.internalValue),
            rejection_statement: lexIntentRejectionStatementToTerraform(this._rejectionStatement.internalValue),
            slot: cdktf.listMapper(lexIntentSlotToTerraform)(this._slot),
            timeouts: lexIntentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.LexIntent = LexIntent;
_m = JSII_RTTI_SYMBOL_1;
LexIntent[_m] = { fqn: "@cdktf/provider-aws.lex.LexIntent", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LexIntent.tfResourceType = "aws_lex_intent";
