// @cdktf/provider-aws/lib/lex/lex-bot-alias.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LexBotAlias = exports.LexBotAliasTimeoutsOutputReference = exports.lexBotAliasTimeoutsToTerraform = exports.LexBotAliasConversationLogsOutputReference = exports.lexBotAliasConversationLogsToTerraform = exports.lexBotAliasConversationLogsLogSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lexBotAliasConversationLogsLogSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination: cdktf.stringToTerraform(struct.destination),
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
        log_type: cdktf.stringToTerraform(struct.logType),
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
    };
}
exports.lexBotAliasConversationLogsLogSettingsToTerraform = lexBotAliasConversationLogsLogSettingsToTerraform;
function lexBotAliasConversationLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_role_arn: cdktf.stringToTerraform(struct.iamRoleArn),
        log_settings: cdktf.listMapper(lexBotAliasConversationLogsLogSettingsToTerraform)(struct.logSettings),
    };
}
exports.lexBotAliasConversationLogsToTerraform = lexBotAliasConversationLogsToTerraform;
/**
 * @stability stable
 */
class LexBotAliasConversationLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iamRoleArn) {
            hasAnyValues = true;
            internalValueResult.iamRoleArn = this._iamRoleArn;
        }
        if (this._logSettings) {
            hasAnyValues = true;
            internalValueResult.logSettings = this._logSettings;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iamRoleArn = undefined;
            this._logSettings = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iamRoleArn = value.iamRoleArn;
            this._logSettings = value.logSettings;
        }
    }
    /**
     * @stability stable
     */
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    /**
     * @stability stable
     */
    set iamRoleArn(value) {
        this._iamRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamRoleArnInput() {
        return this._iamRoleArn;
    }
    /**
     * @stability stable
     */
    get logSettings() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('log_settings');
    }
    /**
     * @stability stable
     */
    set logSettings(value) {
        this._logSettings = value;
    }
    /**
     * @stability stable
     */
    resetLogSettings() {
        this._logSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logSettingsInput() {
        return this._logSettings;
    }
}
exports.LexBotAliasConversationLogsOutputReference = LexBotAliasConversationLogsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LexBotAliasConversationLogsOutputReference[_a] = { fqn: "@cdktf/provider-aws.lex.LexBotAliasConversationLogsOutputReference", version: "3.0.1" };
function lexBotAliasTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lexBotAliasTimeoutsToTerraform = lexBotAliasTimeoutsToTerraform;
/**
 * @stability stable
 */
class LexBotAliasTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.LexBotAliasTimeoutsOutputReference = LexBotAliasTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LexBotAliasTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.lex.LexBotAliasTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lex_bot_alias.html aws_lex_bot_alias}.
 *
 * @stability stable
 */
class LexBotAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lex_bot_alias.html aws_lex_bot_alias} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lex_bot_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // conversation_logs - computed: false, optional: true, required: false
        this._conversationLogs = new LexBotAliasConversationLogsOutputReference(this, "conversation_logs", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LexBotAliasTimeoutsOutputReference(this, "timeouts", true);
        this._botName = config.botName;
        this._botVersion = config.botVersion;
        this._description = config.description;
        this._name = config.name;
        this._conversationLogs.internalValue = config.conversationLogs;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get botName() {
        return this.getStringAttribute('bot_name');
    }
    /**
     * @stability stable
     */
    set botName(value) {
        this._botName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get botNameInput() {
        return this._botName;
    }
    /**
     * @stability stable
     */
    get botVersion() {
        return this.getStringAttribute('bot_version');
    }
    /**
     * @stability stable
     */
    set botVersion(value) {
        this._botVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get botVersionInput() {
        return this._botVersion;
    }
    // checksum - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get checksum() {
        return this.getStringAttribute('checksum');
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get conversationLogs() {
        return this._conversationLogs;
    }
    /**
     * @stability stable
     */
    putConversationLogs(value) {
        this._conversationLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConversationLogs() {
        this._conversationLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get conversationLogsInput() {
        return this._conversationLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bot_name: cdktf.stringToTerraform(this._botName),
            bot_version: cdktf.stringToTerraform(this._botVersion),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            conversation_logs: lexBotAliasConversationLogsToTerraform(this._conversationLogs.internalValue),
            timeouts: lexBotAliasTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.LexBotAlias = LexBotAlias;
_c = JSII_RTTI_SYMBOL_1;
LexBotAlias[_c] = { fqn: "@cdktf/provider-aws.lex.LexBotAlias", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LexBotAlias.tfResourceType = "aws_lex_bot_alias";
