// @cdktf/provider-aws/lib/secretsmanager/data-aws-secretsmanager-secret.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSecretsmanagerSecret = exports.DataAwsSecretsmanagerSecretRotationRules = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsSecretsmanagerSecretRotationRules extends cdktf.ComplexComputedList {
    // automatically_after_days - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get automaticallyAfterDays() {
        return this.getNumberAttribute('automatically_after_days');
    }
}
exports.DataAwsSecretsmanagerSecretRotationRules = DataAwsSecretsmanagerSecretRotationRules;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecretRotationRules[_a] = { fqn: "@cdktf/provider-aws.secretsmanager.DataAwsSecretsmanagerSecretRotationRules", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret.html aws_secretsmanager_secret}.
 *
 * @stability stable
 */
class DataAwsSecretsmanagerSecret extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret.html aws_secretsmanager_secret} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    // rotation_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rotationEnabled() {
        return this.getBooleanAttribute('rotation_enabled');
    }
    // rotation_lambda_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rotationLambdaArn() {
        return this.getStringAttribute('rotation_lambda_arn');
    }
    // rotation_rules - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    rotationRules(index) {
        return new DataAwsSecretsmanagerSecretRotationRules(this, 'rotation_rules', index);
    }
    // tags - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    tags(key) {
        return new cdktf.StringMap(this, 'tags').lookup(key);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsSecretsmanagerSecret = DataAwsSecretsmanagerSecret;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecret[_b] = { fqn: "@cdktf/provider-aws.secretsmanager.DataAwsSecretsmanagerSecret", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSecretsmanagerSecret.tfResourceType = "aws_secretsmanager_secret";
