// @cdktf/provider-aws/lib/secretsmanager/secretsmanager-secret-rotation.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsmanagerSecretRotation = exports.SecretsmanagerSecretRotationRotationRulesOutputReference = exports.secretsmanagerSecretRotationRotationRulesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function secretsmanagerSecretRotationRotationRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        automatically_after_days: cdktf.numberToTerraform(struct.automaticallyAfterDays),
    };
}
exports.secretsmanagerSecretRotationRotationRulesToTerraform = secretsmanagerSecretRotationRotationRulesToTerraform;
/**
 * @stability stable
 */
class SecretsmanagerSecretRotationRotationRulesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._automaticallyAfterDays) {
            hasAnyValues = true;
            internalValueResult.automaticallyAfterDays = this._automaticallyAfterDays;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._automaticallyAfterDays = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._automaticallyAfterDays = value.automaticallyAfterDays;
        }
    }
    /**
     * @stability stable
     */
    get automaticallyAfterDays() {
        return this.getNumberAttribute('automatically_after_days');
    }
    /**
     * @stability stable
     */
    set automaticallyAfterDays(value) {
        this._automaticallyAfterDays = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get automaticallyAfterDaysInput() {
        return this._automaticallyAfterDays;
    }
}
exports.SecretsmanagerSecretRotationRotationRulesOutputReference = SecretsmanagerSecretRotationRotationRulesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SecretsmanagerSecretRotationRotationRulesOutputReference[_a] = { fqn: "@cdktf/provider-aws.secretsmanager.SecretsmanagerSecretRotationRotationRulesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret_rotation.html aws_secretsmanager_secret_rotation}.
 *
 * @stability stable
 */
class SecretsmanagerSecretRotation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret_rotation.html aws_secretsmanager_secret_rotation} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_rotation',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // rotation_rules - computed: false, optional: false, required: true
        this._rotationRules = new SecretsmanagerSecretRotationRotationRulesOutputReference(this, "rotation_rules", true);
        this._rotationLambdaArn = config.rotationLambdaArn;
        this._secretId = config.secretId;
        this._tags = config.tags;
        this._rotationRules.internalValue = config.rotationRules;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // rotation_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rotationEnabled() {
        return this.getBooleanAttribute('rotation_enabled');
    }
    /**
     * @stability stable
     */
    get rotationLambdaArn() {
        return this.getStringAttribute('rotation_lambda_arn');
    }
    /**
     * @stability stable
     */
    set rotationLambdaArn(value) {
        this._rotationLambdaArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rotationLambdaArnInput() {
        return this._rotationLambdaArn;
    }
    /**
     * @stability stable
     */
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    /**
     * @stability stable
     */
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretIdInput() {
        return this._secretId;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get rotationRules() {
        return this._rotationRules;
    }
    /**
     * @stability stable
     */
    putRotationRules(value) {
        this._rotationRules.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rotationRulesInput() {
        return this._rotationRules.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            rotation_lambda_arn: cdktf.stringToTerraform(this._rotationLambdaArn),
            secret_id: cdktf.stringToTerraform(this._secretId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            rotation_rules: secretsmanagerSecretRotationRotationRulesToTerraform(this._rotationRules.internalValue),
        };
    }
}
exports.SecretsmanagerSecretRotation = SecretsmanagerSecretRotation;
_b = JSII_RTTI_SYMBOL_1;
SecretsmanagerSecretRotation[_b] = { fqn: "@cdktf/provider-aws.secretsmanager.SecretsmanagerSecretRotation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecretsmanagerSecretRotation.tfResourceType = "aws_secretsmanager_secret_rotation";
