// @cdktf/provider-aws/lib/autoscaling/autoscaling-notification.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoscalingNotification = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/autoscaling_notification.html aws_autoscaling_notification}.
 *
 * @stability stable
 */
class AutoscalingNotification extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/autoscaling_notification.html aws_autoscaling_notification} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_autoscaling_notification',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._groupNames = config.groupNames;
        this._notifications = config.notifications;
        this._topicArn = config.topicArn;
    }
    /**
     * @stability stable
     */
    get groupNames() {
        return this.getListAttribute('group_names');
    }
    /**
     * @stability stable
     */
    set groupNames(value) {
        this._groupNames = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupNamesInput() {
        return this._groupNames;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get notifications() {
        return this.getListAttribute('notifications');
    }
    /**
     * @stability stable
     */
    set notifications(value) {
        this._notifications = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationsInput() {
        return this._notifications;
    }
    /**
     * @stability stable
     */
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    /**
     * @stability stable
     */
    set topicArn(value) {
        this._topicArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get topicArnInput() {
        return this._topicArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            group_names: cdktf.listMapper(cdktf.stringToTerraform)(this._groupNames),
            notifications: cdktf.listMapper(cdktf.stringToTerraform)(this._notifications),
            topic_arn: cdktf.stringToTerraform(this._topicArn),
        };
    }
}
exports.AutoscalingNotification = AutoscalingNotification;
_a = JSII_RTTI_SYMBOL_1;
AutoscalingNotification[_a] = { fqn: "@cdktf/provider-aws.autoscaling.AutoscalingNotification", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AutoscalingNotification.tfResourceType = "aws_autoscaling_notification";
