// @cdktf/provider-aws/lib/dynamodb/dynamodb-kinesis-streaming-destination.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamodbKinesisStreamingDestination = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/dynamodb_kinesis_streaming_destination.html aws_dynamodb_kinesis_streaming_destination}.
 *
 * @stability stable
 */
class DynamodbKinesisStreamingDestination extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dynamodb_kinesis_streaming_destination.html aws_dynamodb_kinesis_streaming_destination} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_kinesis_streaming_destination',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._streamArn = config.streamArn;
        this._tableName = config.tableName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    /**
     * @stability stable
     */
    set streamArn(value) {
        this._streamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get streamArnInput() {
        return this._streamArn;
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            stream_arn: cdktf.stringToTerraform(this._streamArn),
            table_name: cdktf.stringToTerraform(this._tableName),
        };
    }
}
exports.DynamodbKinesisStreamingDestination = DynamodbKinesisStreamingDestination;
_a = JSII_RTTI_SYMBOL_1;
DynamodbKinesisStreamingDestination[_a] = { fqn: "@cdktf/provider-aws.dynamodb.DynamodbKinesisStreamingDestination", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DynamodbKinesisStreamingDestination.tfResourceType = "aws_dynamodb_kinesis_streaming_destination";
