// jsii-rosetta/lib/typescript/imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeImportDeclaration = exports.analyzeImportEquals = void 0;
const ts = require("typescript");
const jsii_utils_1 = require("../jsii/jsii-utils");
const util_1 = require("../util");
const ast_utils_1 = require("./ast-utils");
function analyzeImportEquals(node, context) {
    let moduleName = '???';
    ast_utils_1.matchAst(node.moduleReference, ast_utils_1.nodeOfType('ref', ts.SyntaxKind.ExternalModuleReference), (bindings) => {
        moduleName = ast_utils_1.stringFromLiteral(bindings.ref.expression);
    });
    return {
        node,
        packageName: moduleName,
        moduleSymbol: jsii_utils_1.lookupJsiiSymbolFromNode(context.typeChecker, node.name),
        imports: { import: 'full', alias: context.textOf(node.name) },
    };
}
exports.analyzeImportEquals = analyzeImportEquals;
function analyzeImportDeclaration(node, context) {
    var _a;
    const packageName = ast_utils_1.stringFromLiteral(node.moduleSpecifier);
    const starBindings = ast_utils_1.matchAst(node, ast_utils_1.nodeOfType(ts.SyntaxKind.ImportDeclaration, ast_utils_1.nodeOfType(ts.SyntaxKind.ImportClause, ast_utils_1.nodeOfType('namespace', ts.SyntaxKind.NamespaceImport))));
    if (starBindings) {
        return {
            node,
            packageName,
            moduleSymbol: jsii_utils_1.lookupJsiiSymbolFromNode(context.typeChecker, starBindings.namespace.name),
            imports: {
                import: 'full',
                alias: context.textOf(starBindings.namespace.name),
            },
        };
    }
    const namedBindings = ast_utils_1.matchAst(node, ast_utils_1.nodeOfType(ts.SyntaxKind.ImportDeclaration, ast_utils_1.nodeOfType(ts.SyntaxKind.ImportClause, ast_utils_1.nodeOfType(ts.SyntaxKind.NamedImports, ast_utils_1.allOfType(ts.SyntaxKind.ImportSpecifier, 'specifiers')))));
    const elements = [];
    if (namedBindings) {
        elements.push(...namedBindings.specifiers.map((spec) => {
            // regular import { name }, renamed import { propertyName, name }
            if (spec.propertyName) {
                // Renamed import
                return {
                    sourceName: context.textOf(spec.propertyName),
                    alias: context.textOf(spec.name),
                    importedSymbol: jsii_utils_1.lookupJsiiSymbolFromNode(context.typeChecker, spec.propertyName),
                };
            }
            return {
                sourceName: context.textOf(spec.name),
                importedSymbol: jsii_utils_1.lookupJsiiSymbolFromNode(context.typeChecker, spec.name),
            };
        }));
    }
    return {
        node,
        packageName,
        imports: { import: 'selective', elements },
        moduleSymbol: util_1.fmap((_a = elements === null || elements === void 0 ? void 0 : elements[0]) === null || _a === void 0 ? void 0 : _a.importedSymbol, jsii_utils_1.parentSymbol),
    };
}
exports.analyzeImportDeclaration = analyzeImportDeclaration;
