// @cdktf/provider-aws/lib/appstream/appstream-stack.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamStack = exports.appstreamStackUserSettingsToTerraform = exports.appstreamStackStorageConnectorsToTerraform = exports.AppstreamStackApplicationSettingsOutputReference = exports.appstreamStackApplicationSettingsToTerraform = exports.appstreamStackAccessEndpointsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appstreamStackAccessEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_type: cdktf.stringToTerraform(struct.endpointType),
        vpce_id: cdktf.stringToTerraform(struct.vpceId),
    };
}
exports.appstreamStackAccessEndpointsToTerraform = appstreamStackAccessEndpointsToTerraform;
function appstreamStackApplicationSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        settings_group: cdktf.stringToTerraform(struct.settingsGroup),
    };
}
exports.appstreamStackApplicationSettingsToTerraform = appstreamStackApplicationSettingsToTerraform;
/**
 * @stability stable
 */
class AppstreamStackApplicationSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._settingsGroup) {
            hasAnyValues = true;
            internalValueResult.settingsGroup = this._settingsGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._settingsGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._settingsGroup = value.settingsGroup;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get settingsGroup() {
        return this.getStringAttribute('settings_group');
    }
    /**
     * @stability stable
     */
    set settingsGroup(value) {
        this._settingsGroup = value;
    }
    /**
     * @stability stable
     */
    resetSettingsGroup() {
        this._settingsGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get settingsGroupInput() {
        return this._settingsGroup;
    }
}
exports.AppstreamStackApplicationSettingsOutputReference = AppstreamStackApplicationSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppstreamStackApplicationSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.appstream.AppstreamStackApplicationSettingsOutputReference", version: "3.0.1" };
function appstreamStackStorageConnectorsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connector_type: cdktf.stringToTerraform(struct.connectorType),
        domains: cdktf.listMapper(cdktf.stringToTerraform)(struct.domains),
        resource_identifier: cdktf.stringToTerraform(struct.resourceIdentifier),
    };
}
exports.appstreamStackStorageConnectorsToTerraform = appstreamStackStorageConnectorsToTerraform;
function appstreamStackUserSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        permission: cdktf.stringToTerraform(struct.permission),
    };
}
exports.appstreamStackUserSettingsToTerraform = appstreamStackUserSettingsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/appstream_stack.html aws_appstream_stack}.
 *
 * @stability stable
 */
class AppstreamStack extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appstream_stack.html aws_appstream_stack} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_stack',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // application_settings - computed: false, optional: true, required: false
        this._applicationSettings = new AppstreamStackApplicationSettingsOutputReference(this, "application_settings", true);
        this._description = config.description;
        this._displayName = config.displayName;
        this._embedHostDomains = config.embedHostDomains;
        this._feedbackUrl = config.feedbackUrl;
        this._name = config.name;
        this._redirectUrl = config.redirectUrl;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._accessEndpoints = config.accessEndpoints;
        this._applicationSettings.internalValue = config.applicationSettings;
        this._storageConnectors = config.storageConnectors;
        this._userSettings = config.userSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    /**
     * @stability stable
     */
    set displayName(value) {
        this._displayName = value;
    }
    /**
     * @stability stable
     */
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get displayNameInput() {
        return this._displayName;
    }
    /**
     * @stability stable
     */
    get embedHostDomains() {
        return this.getListAttribute('embed_host_domains');
    }
    /**
     * @stability stable
     */
    set embedHostDomains(value) {
        this._embedHostDomains = value;
    }
    /**
     * @stability stable
     */
    resetEmbedHostDomains() {
        this._embedHostDomains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get embedHostDomainsInput() {
        return this._embedHostDomains;
    }
    /**
     * @stability stable
     */
    get feedbackUrl() {
        return this.getStringAttribute('feedback_url');
    }
    /**
     * @stability stable
     */
    set feedbackUrl(value) {
        this._feedbackUrl = value;
    }
    /**
     * @stability stable
     */
    resetFeedbackUrl() {
        this._feedbackUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get feedbackUrlInput() {
        return this._feedbackUrl;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get redirectUrl() {
        return this.getStringAttribute('redirect_url');
    }
    /**
     * @stability stable
     */
    set redirectUrl(value) {
        this._redirectUrl = value;
    }
    /**
     * @stability stable
     */
    resetRedirectUrl() {
        this._redirectUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get redirectUrlInput() {
        return this._redirectUrl;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get accessEndpoints() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('access_endpoints');
    }
    /**
     * @stability stable
     */
    set accessEndpoints(value) {
        this._accessEndpoints = value;
    }
    /**
     * @stability stable
     */
    resetAccessEndpoints() {
        this._accessEndpoints = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessEndpointsInput() {
        return this._accessEndpoints;
    }
    /**
     * @stability stable
     */
    get applicationSettings() {
        return this._applicationSettings;
    }
    /**
     * @stability stable
     */
    putApplicationSettings(value) {
        this._applicationSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetApplicationSettings() {
        this._applicationSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applicationSettingsInput() {
        return this._applicationSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get storageConnectors() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('storage_connectors');
    }
    /**
     * @stability stable
     */
    set storageConnectors(value) {
        this._storageConnectors = value;
    }
    /**
     * @stability stable
     */
    resetStorageConnectors() {
        this._storageConnectors = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageConnectorsInput() {
        return this._storageConnectors;
    }
    /**
     * @stability stable
     */
    get userSettings() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('user_settings');
    }
    /**
     * @stability stable
     */
    set userSettings(value) {
        this._userSettings = value;
    }
    /**
     * @stability stable
     */
    resetUserSettings() {
        this._userSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userSettingsInput() {
        return this._userSettings;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            embed_host_domains: cdktf.listMapper(cdktf.stringToTerraform)(this._embedHostDomains),
            feedback_url: cdktf.stringToTerraform(this._feedbackUrl),
            name: cdktf.stringToTerraform(this._name),
            redirect_url: cdktf.stringToTerraform(this._redirectUrl),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            access_endpoints: cdktf.listMapper(appstreamStackAccessEndpointsToTerraform)(this._accessEndpoints),
            application_settings: appstreamStackApplicationSettingsToTerraform(this._applicationSettings.internalValue),
            storage_connectors: cdktf.listMapper(appstreamStackStorageConnectorsToTerraform)(this._storageConnectors),
            user_settings: cdktf.listMapper(appstreamStackUserSettingsToTerraform)(this._userSettings),
        };
    }
}
exports.AppstreamStack = AppstreamStack;
_b = JSII_RTTI_SYMBOL_1;
AppstreamStack[_b] = { fqn: "@cdktf/provider-aws.appstream.AppstreamStack", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AppstreamStack.tfResourceType = "aws_appstream_stack";
