// @cdktf/provider-aws/lib/imagebuilder/data-aws-imagebuilder-infrastructure-configuration.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsImagebuilderInfrastructureConfiguration = exports.DataAwsImagebuilderInfrastructureConfigurationLogging = exports.DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs extends cdktf.ComplexComputedList {
    // s3_bucket_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    // s3_key_prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
}
exports.DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs = DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs;
_a = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs[_a] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderInfrastructureConfigurationLoggingS3Logs", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsImagebuilderInfrastructureConfigurationLogging extends cdktf.ComplexComputedList {
    // s3_logs - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3Logs() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('s3_logs');
    }
}
exports.DataAwsImagebuilderInfrastructureConfigurationLogging = DataAwsImagebuilderInfrastructureConfigurationLogging;
_b = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderInfrastructureConfigurationLogging[_b] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderInfrastructureConfigurationLogging", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_infrastructure_configuration.html aws_imagebuilder_infrastructure_configuration}.
 *
 * @stability stable
 */
class DataAwsImagebuilderInfrastructureConfiguration extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_infrastructure_configuration.html aws_imagebuilder_infrastructure_configuration} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_infrastructure_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
        this._resourceTags = config.resourceTags;
        this._tags = config.tags;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // date_created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // date_updated - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateUpdated() {
        return this.getStringAttribute('date_updated');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // instance_profile_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceProfileName() {
        return this.getStringAttribute('instance_profile_name');
    }
    // instance_types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceTypes() {
        return this.getListAttribute('instance_types');
    }
    // key_pair - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyPair() {
        return this.getStringAttribute('key_pair');
    }
    // logging - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    logging(index) {
        return new DataAwsImagebuilderInfrastructureConfigurationLogging(this, 'logging', index);
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    get resourceTags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('resource_tags');
    }
    /**
     * @stability stable
     */
    set resourceTags(value) {
        this._resourceTags = value;
    }
    /**
     * @stability stable
     */
    resetResourceTags() {
        this._resourceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTagsInput() {
        return this._resourceTags;
    }
    // security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    // sns_topic_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snsTopicArn() {
        return this.getStringAttribute('sns_topic_arn');
    }
    // subnet_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // terminate_instance_on_failure - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get terminateInstanceOnFailure() {
        return this.getBooleanAttribute('terminate_instance_on_failure');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            resource_tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._resourceTags),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsImagebuilderInfrastructureConfiguration = DataAwsImagebuilderInfrastructureConfiguration;
_c = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderInfrastructureConfiguration[_c] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderInfrastructureConfiguration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsImagebuilderInfrastructureConfiguration.tfResourceType = "aws_imagebuilder_infrastructure_configuration";
