// @cdktf/provider-aws/lib/imagebuilder/data-aws-imagebuilder-image.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsImagebuilderImage = exports.DataAwsImagebuilderImageOutputResources = exports.DataAwsImagebuilderImageOutputResourcesAmis = exports.DataAwsImagebuilderImageImageTestsConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsImagebuilderImageImageTestsConfiguration extends cdktf.ComplexComputedList {
    // image_tests_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageTestsEnabled() {
        return this.getBooleanAttribute('image_tests_enabled');
    }
    // timeout_minutes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get timeoutMinutes() {
        return this.getNumberAttribute('timeout_minutes');
    }
}
exports.DataAwsImagebuilderImageImageTestsConfiguration = DataAwsImagebuilderImageImageTestsConfiguration;
_a = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageImageTestsConfiguration[_a] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImageImageTestsConfiguration", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsImagebuilderImageOutputResourcesAmis extends cdktf.ComplexComputedList {
    // account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // image - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get image() {
        return this.getStringAttribute('image');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsImagebuilderImageOutputResourcesAmis = DataAwsImagebuilderImageOutputResourcesAmis;
_b = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResourcesAmis[_b] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImageOutputResourcesAmis", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsImagebuilderImageOutputResources extends cdktf.ComplexComputedList {
    // amis - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get amis() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('amis');
    }
}
exports.DataAwsImagebuilderImageOutputResources = DataAwsImagebuilderImageOutputResources;
_c = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResources[_c] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImageOutputResources", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image.html aws_imagebuilder_image}.
 *
 * @stability stable
 */
class DataAwsImagebuilderImage extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image.html aws_imagebuilder_image} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_image',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
        this._tags = config.tags;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // build_version_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get buildVersionArn() {
        return this.getStringAttribute('build_version_arn');
    }
    // date_created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // distribution_configuration_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get distributionConfigurationArn() {
        return this.getStringAttribute('distribution_configuration_arn');
    }
    // enhanced_image_metadata_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enhancedImageMetadataEnabled() {
        return this.getBooleanAttribute('enhanced_image_metadata_enabled');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image_recipe_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageRecipeArn() {
        return this.getStringAttribute('image_recipe_arn');
    }
    // image_tests_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    imageTestsConfiguration(index) {
        return new DataAwsImagebuilderImageImageTestsConfiguration(this, 'image_tests_configuration', index);
    }
    // infrastructure_configuration_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get infrastructureConfigurationArn() {
        return this.getStringAttribute('infrastructure_configuration_arn');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // os_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get osVersion() {
        return this.getStringAttribute('os_version');
    }
    // output_resources - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    outputResources(index) {
        return new DataAwsImagebuilderImageOutputResources(this, 'output_resources', index);
    }
    // platform - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get platform() {
        return this.getStringAttribute('platform');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsImagebuilderImage = DataAwsImagebuilderImage;
_d = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImage[_d] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImage", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsImagebuilderImage.tfResourceType = "aws_imagebuilder_image";
