// @cdktf/provider-aws/lib/cloudformation/cloudformation-stack.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudformationStack = exports.CloudformationStackTimeoutsOutputReference = exports.cloudformationStackTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudformationStackTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudformationStackTimeoutsToTerraform = cloudformationStackTimeoutsToTerraform;
/**
 * @stability stable
 */
class CloudformationStackTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.CloudformationStackTimeoutsOutputReference = CloudformationStackTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudformationStackTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudformation.CloudformationStackTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_stack.html aws_cloudformation_stack}.
 *
 * @stability stable
 */
class CloudformationStack extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_stack.html aws_cloudformation_stack} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudformation_stack',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudformationStackTimeoutsOutputReference(this, "timeouts", true);
        this._capabilities = config.capabilities;
        this._disableRollback = config.disableRollback;
        this._iamRoleArn = config.iamRoleArn;
        this._name = config.name;
        this._notificationArns = config.notificationArns;
        this._onFailure = config.onFailure;
        this._parameters = config.parameters;
        this._policyBody = config.policyBody;
        this._policyUrl = config.policyUrl;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._templateBody = config.templateBody;
        this._templateUrl = config.templateUrl;
        this._timeoutInMinutes = config.timeoutInMinutes;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get capabilities() {
        return this.getListAttribute('capabilities');
    }
    /**
     * @stability stable
     */
    set capabilities(value) {
        this._capabilities = value;
    }
    /**
     * @stability stable
     */
    resetCapabilities() {
        this._capabilities = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capabilitiesInput() {
        return this._capabilities;
    }
    /**
     * @stability stable
     */
    get disableRollback() {
        return this.getBooleanAttribute('disable_rollback');
    }
    /**
     * @stability stable
     */
    set disableRollback(value) {
        this._disableRollback = value;
    }
    /**
     * @stability stable
     */
    resetDisableRollback() {
        this._disableRollback = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableRollbackInput() {
        return this._disableRollback;
    }
    /**
     * @stability stable
     */
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    /**
     * @stability stable
     */
    set iamRoleArn(value) {
        this._iamRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetIamRoleArn() {
        this._iamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamRoleArnInput() {
        return this._iamRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get notificationArns() {
        return this.getListAttribute('notification_arns');
    }
    /**
     * @stability stable
     */
    set notificationArns(value) {
        this._notificationArns = value;
    }
    /**
     * @stability stable
     */
    resetNotificationArns() {
        this._notificationArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationArnsInput() {
        return this._notificationArns;
    }
    /**
     * @stability stable
     */
    get onFailure() {
        return this.getStringAttribute('on_failure');
    }
    /**
     * @stability stable
     */
    set onFailure(value) {
        this._onFailure = value;
    }
    /**
     * @stability stable
     */
    resetOnFailure() {
        this._onFailure = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get onFailureInput() {
        return this._onFailure;
    }
    // outputs - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    outputs(key) {
        return new cdktf.StringMap(this, 'outputs').lookup(key);
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get policyBody() {
        return this.getStringAttribute('policy_body');
    }
    /**
     * @stability stable
     */
    set policyBody(value) {
        this._policyBody = value;
    }
    /**
     * @stability stable
     */
    resetPolicyBody() {
        this._policyBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyBodyInput() {
        return this._policyBody;
    }
    /**
     * @stability stable
     */
    get policyUrl() {
        return this.getStringAttribute('policy_url');
    }
    /**
     * @stability stable
     */
    set policyUrl(value) {
        this._policyUrl = value;
    }
    /**
     * @stability stable
     */
    resetPolicyUrl() {
        this._policyUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyUrlInput() {
        return this._policyUrl;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get templateBody() {
        return this.getStringAttribute('template_body');
    }
    /**
     * @stability stable
     */
    set templateBody(value) {
        this._templateBody = value;
    }
    /**
     * @stability stable
     */
    resetTemplateBody() {
        this._templateBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateBodyInput() {
        return this._templateBody;
    }
    /**
     * @stability stable
     */
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    /**
     * @stability stable
     */
    set templateUrl(value) {
        this._templateUrl = value;
    }
    /**
     * @stability stable
     */
    resetTemplateUrl() {
        this._templateUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateUrlInput() {
        return this._templateUrl;
    }
    /**
     * @stability stable
     */
    get timeoutInMinutes() {
        return this.getNumberAttribute('timeout_in_minutes');
    }
    /**
     * @stability stable
     */
    set timeoutInMinutes(value) {
        this._timeoutInMinutes = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutInMinutes() {
        this._timeoutInMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutInMinutesInput() {
        return this._timeoutInMinutes;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            capabilities: cdktf.listMapper(cdktf.stringToTerraform)(this._capabilities),
            disable_rollback: cdktf.booleanToTerraform(this._disableRollback),
            iam_role_arn: cdktf.stringToTerraform(this._iamRoleArn),
            name: cdktf.stringToTerraform(this._name),
            notification_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._notificationArns),
            on_failure: cdktf.stringToTerraform(this._onFailure),
            parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._parameters),
            policy_body: cdktf.stringToTerraform(this._policyBody),
            policy_url: cdktf.stringToTerraform(this._policyUrl),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            template_body: cdktf.stringToTerraform(this._templateBody),
            template_url: cdktf.stringToTerraform(this._templateUrl),
            timeout_in_minutes: cdktf.numberToTerraform(this._timeoutInMinutes),
            timeouts: cloudformationStackTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.CloudformationStack = CloudformationStack;
_b = JSII_RTTI_SYMBOL_1;
CloudformationStack[_b] = { fqn: "@cdktf/provider-aws.cloudformation.CloudformationStack", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudformationStack.tfResourceType = "aws_cloudformation_stack";
