// @cdktf/provider-aws/lib/cloudformation/data-aws-cloudformation-type.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudformationType = exports.DataAwsCloudformationTypeLoggingConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsCloudformationTypeLoggingConfig extends cdktf.ComplexComputedList {
    // log_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get logGroupName() {
        return this.getStringAttribute('log_group_name');
    }
    // log_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get logRoleArn() {
        return this.getStringAttribute('log_role_arn');
    }
}
exports.DataAwsCloudformationTypeLoggingConfig = DataAwsCloudformationTypeLoggingConfig;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudformationTypeLoggingConfig[_a] = { fqn: "@cdktf/provider-aws.cloudformation.DataAwsCloudformationTypeLoggingConfig", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudformation_type.html aws_cloudformation_type}.
 *
 * @stability stable
 */
class DataAwsCloudformationType extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudformation_type.html aws_cloudformation_type} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudformation_type',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._type = config.type;
        this._typeName = config.typeName;
        this._versionId = config.versionId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_version_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultVersionId() {
        return this.getStringAttribute('default_version_id');
    }
    // deprecated_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deprecatedStatus() {
        return this.getStringAttribute('deprecated_status');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // documentation_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get documentationUrl() {
        return this.getStringAttribute('documentation_url');
    }
    // execution_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // is_default_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get isDefaultVersion() {
        return this.getBooleanAttribute('is_default_version');
    }
    // logging_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    loggingConfig(index) {
        return new DataAwsCloudformationTypeLoggingConfig(this, 'logging_config', index);
    }
    // provisioning_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get provisioningType() {
        return this.getStringAttribute('provisioning_type');
    }
    // schema - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get schema() {
        return this.getStringAttribute('schema');
    }
    // source_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceUrl() {
        return this.getStringAttribute('source_url');
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    // type_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get typeArn() {
        return this.getStringAttribute('type_arn');
    }
    /**
     * @stability stable
     */
    get typeName() {
        return this.getStringAttribute('type_name');
    }
    /**
     * @stability stable
     */
    set typeName(value) {
        this._typeName = value;
    }
    /**
     * @stability stable
     */
    resetTypeName() {
        this._typeName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeNameInput() {
        return this._typeName;
    }
    /**
     * @stability stable
     */
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    /**
     * @stability stable
     */
    set versionId(value) {
        this._versionId = value;
    }
    /**
     * @stability stable
     */
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionIdInput() {
        return this._versionId;
    }
    // visibility - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get visibility() {
        return this.getStringAttribute('visibility');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            type: cdktf.stringToTerraform(this._type),
            type_name: cdktf.stringToTerraform(this._typeName),
            version_id: cdktf.stringToTerraform(this._versionId),
        };
    }
}
exports.DataAwsCloudformationType = DataAwsCloudformationType;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCloudformationType[_b] = { fqn: "@cdktf/provider-aws.cloudformation.DataAwsCloudformationType", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCloudformationType.tfResourceType = "aws_cloudformation_type";
