// @cdktf/provider-aws/lib/signer/signer-signing-profile-permission.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignerSigningProfilePermission = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/signer_signing_profile_permission.html aws_signer_signing_profile_permission}.
 *
 * @stability stable
 */
class SignerSigningProfilePermission extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/signer_signing_profile_permission.html aws_signer_signing_profile_permission} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_signer_signing_profile_permission',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._action = config.action;
        this._principal = config.principal;
        this._profileName = config.profileName;
        this._profileVersion = config.profileVersion;
        this._statementId = config.statementId;
        this._statementIdPrefix = config.statementIdPrefix;
    }
    /**
     * @stability stable
     */
    get action() {
        return this.getStringAttribute('action');
    }
    /**
     * @stability stable
     */
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionInput() {
        return this._action;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get principal() {
        return this.getStringAttribute('principal');
    }
    /**
     * @stability stable
     */
    set principal(value) {
        this._principal = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalInput() {
        return this._principal;
    }
    /**
     * @stability stable
     */
    get profileName() {
        return this.getStringAttribute('profile_name');
    }
    /**
     * @stability stable
     */
    set profileName(value) {
        this._profileName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get profileNameInput() {
        return this._profileName;
    }
    /**
     * @stability stable
     */
    get profileVersion() {
        return this.getStringAttribute('profile_version');
    }
    /**
     * @stability stable
     */
    set profileVersion(value) {
        this._profileVersion = value;
    }
    /**
     * @stability stable
     */
    resetProfileVersion() {
        this._profileVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get profileVersionInput() {
        return this._profileVersion;
    }
    /**
     * @stability stable
     */
    get statementId() {
        return this.getStringAttribute('statement_id');
    }
    /**
     * @stability stable
     */
    set statementId(value) {
        this._statementId = value;
    }
    /**
     * @stability stable
     */
    resetStatementId() {
        this._statementId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statementIdInput() {
        return this._statementId;
    }
    /**
     * @stability stable
     */
    get statementIdPrefix() {
        return this.getStringAttribute('statement_id_prefix');
    }
    /**
     * @stability stable
     */
    set statementIdPrefix(value) {
        this._statementIdPrefix = value;
    }
    /**
     * @stability stable
     */
    resetStatementIdPrefix() {
        this._statementIdPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statementIdPrefixInput() {
        return this._statementIdPrefix;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            action: cdktf.stringToTerraform(this._action),
            principal: cdktf.stringToTerraform(this._principal),
            profile_name: cdktf.stringToTerraform(this._profileName),
            profile_version: cdktf.stringToTerraform(this._profileVersion),
            statement_id: cdktf.stringToTerraform(this._statementId),
            statement_id_prefix: cdktf.stringToTerraform(this._statementIdPrefix),
        };
    }
}
exports.SignerSigningProfilePermission = SignerSigningProfilePermission;
_a = JSII_RTTI_SYMBOL_1;
SignerSigningProfilePermission[_a] = { fqn: "@cdktf/provider-aws.signer.SignerSigningProfilePermission", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SignerSigningProfilePermission.tfResourceType = "aws_signer_signing_profile_permission";
