// jsii-rosetta/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.indexBy = exports.isDefined = exports.groupBy = exports.sortBy = exports.mapValues = exports.fmap = exports.mkDict = exports.setExtend = exports.partition = exports.flat = exports.divideEvenly = exports.hasStrictBranding = exports.annotateStrictDiagnostic = exports.StrictBrand = exports.printDiagnostics = exports.startsWithUppercase = void 0;
function startsWithUppercase(x) {
    return /^[A-Z]/.exec(x) != null;
}
exports.startsWithUppercase = startsWithUppercase;
function printDiagnostics(diags, stream) {
    // Don't print too much, at some point it just clogs up the log
    const maxDiags = 50;
    for (const diag of diags.slice(0, maxDiags)) {
        stream.write(diag.formattedMessage);
    }
    if (diags.length > maxDiags) {
        stream.write(`(...and ${diags.length - maxDiags} more diagnostics not shown)`);
    }
}
exports.printDiagnostics = printDiagnostics;
exports.StrictBrand = 'jsii.strict';
/**
 * Annotate a diagnostic with a magic property to indicate it's a strict diagnostic
 */
function annotateStrictDiagnostic(diag) {
    Object.defineProperty(diag, exports.StrictBrand, {
        configurable: false,
        enumerable: true,
        value: true,
        writable: false,
    });
}
exports.annotateStrictDiagnostic = annotateStrictDiagnostic;
/**
 * Return whether or not the given diagnostic was annotated with the magic strict property
 */
function hasStrictBranding(diag) {
    return !!diag[exports.StrictBrand];
}
exports.hasStrictBranding = hasStrictBranding;
/**
 * Chunk an array of elements into approximately equal groups
 */
function divideEvenly(groups, xs) {
    const chunkSize = Math.ceil(xs.length / groups);
    const ret = [];
    for (let i = 0; i < groups; i++) {
        ret.push(xs.slice(i * chunkSize, (i + 1) * chunkSize));
    }
    return ret;
}
exports.divideEvenly = divideEvenly;
function flat(xs) {
    return Array.prototype.concat.apply([], xs);
}
exports.flat = flat;
/**
 * Partition a list in twain using a predicate
 *
 * Returns [elements-matching-predicate, elements-not-matching-predicate];
 */
function partition(xs, pred) {
    const truthy = new Array();
    const falsy = new Array();
    for (const x of xs) {
        if (pred(x)) {
            truthy.push(x);
        }
        else {
            falsy.push(x);
        }
    }
    return [truthy, falsy];
}
exports.partition = partition;
function setExtend(xs, els) {
    for (const el of els) {
        xs.add(el);
    }
}
exports.setExtend = setExtend;
function mkDict(xs) {
    const ret = {};
    for (const [key, value] of xs) {
        ret[key] = value;
    }
    return ret;
}
exports.mkDict = mkDict;
function fmap(value, fn) {
    if (value == null) {
        return undefined;
    }
    return fn(value);
}
exports.fmap = fmap;
function mapValues(xs, fn) {
    const ret = {};
    for (const [key, value] of Object.entries(xs)) {
        ret[key] = fn(value);
    }
    return ret;
}
exports.mapValues = mapValues;
/**
 * Sort an array by a key function.
 *
 * Instead of having to write your own comparators for your types any time you
 * want to sort, you supply a function that maps a value to a compound sort key
 * consisting of numbers or strings. The sorting will happen by that sort key
 * instead.
 */
function sortBy(xs, keyFn) {
    return xs.sort((a, b) => {
        const aKey = keyFn(a);
        const bKey = keyFn(b);
        for (let i = 0; i < Math.min(aKey.length, bKey.length); i++) {
            // Compare aKey[i] to bKey[i]
            const av = aKey[i];
            const bv = bKey[i];
            if (av === bv) {
                continue;
            }
            if (typeof av !== typeof bv) {
                throw new Error(`Type of sort key ${JSON.stringify(aKey)} not same as ${JSON.stringify(bKey)}`);
            }
            if (typeof av === 'number' && typeof bv === 'number') {
                return av - bv;
            }
            if (typeof av === 'string' && typeof bv === 'string') {
                return av.localeCompare(bv);
            }
        }
        return aKey.length - bKey.length;
    });
}
exports.sortBy = sortBy;
/**
 * Group elements by a key
 *
 * Supply a function that maps each element to a key string.
 *
 * Returns a map of the key to the list of elements that map to that key.
 */
function groupBy(xs, keyFn) {
    const ret = {};
    for (const x of xs) {
        const key = keyFn(x);
        if (ret[key]) {
            ret[key].push(x);
        }
        else {
            ret[key] = [x];
        }
    }
    return ret;
}
exports.groupBy = groupBy;
function isDefined(x) {
    return x !== undefined;
}
exports.isDefined = isDefined;
function indexBy(xs, fn) {
    return mkDict(xs.map((x) => [fn(x), x]));
}
exports.indexBy = indexBy;
