// @cdktf/provider-generator/lib/get/generator/models/struct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigStruct = exports.Struct = void 0;
const util_1 = require("../../../util");
class Struct {
    constructor(name, attributes, isClass = false, isAnonymous = false, isSingleItem = false) {
        this.name = name;
        this.attributes = attributes;
        this.isClass = isClass;
        this.isAnonymous = isAnonymous;
        this.isSingleItem = isSingleItem;
    }
    get assignableAttributes() {
        const attributes = this.isAnonymous
            ? this.attributes
            : this.attributes.filter((attribute) => attribute.isAssignable);
        return this.filterIgnoredAttributes(attributes);
    }
    get optionalAttributes() {
        return this.attributes.filter((attribute) => attribute.optional);
    }
    get allOptional() {
        return (this.attributes.filter((attribute) => !attribute.optional && !attribute.computed).length == 0);
    }
    get attributeType() {
        return `${this.name}${this.allOptional ? " = {}" : ""}`;
    }
    filterIgnoredAttributes(attributes) {
        return attributes;
    }
    get extends() {
        return "";
    }
    get attributeTypeNames() {
        return this.attributes.map((a) => a.type.typeName);
    }
    get mapperName() {
        return `${util_1.downcaseFirst(this.name)}ToTerraform`;
    }
    get outputReferenceName() {
        return `${this.name}OutputReference`;
    }
    get isProvider() {
        return this.attributes.some((att) => att.isProvider);
    }
    get attributeTypeNamesFromClasses() {
        return this.attributes
            .filter((a) => { var _a; return (_a = a.type.struct) === null || _a === void 0 ? void 0 : _a.isClass; })
            .map((a) => a.type.typeName);
    }
}
exports.Struct = Struct;
class ConfigStruct extends Struct {
    filterIgnoredAttributes(attributes) {
        return attributes.filter((attribute) => !attribute.isConfigIgnored);
    }
    get extends() {
        return ` extends cdktf.TerraformMetaArguments`;
    }
}
exports.ConfigStruct = ConfigStruct;
