// @cdktf/provider-aws/lib/s3/s3-account-public-access-block.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3AccountPublicAccessBlock = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_account_public_access_block.html aws_s3_account_public_access_block}.
 *
 * @stability stable
 */
class S3AccountPublicAccessBlock extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_account_public_access_block.html aws_s3_account_public_access_block} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_account_public_access_block',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accountId = config.accountId;
        this._blockPublicAcls = config.blockPublicAcls;
        this._blockPublicPolicy = config.blockPublicPolicy;
        this._ignorePublicAcls = config.ignorePublicAcls;
        this._restrictPublicBuckets = config.restrictPublicBuckets;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    /**
     * @stability stable
     */
    get blockPublicAcls() {
        return this.getBooleanAttribute('block_public_acls');
    }
    /**
     * @stability stable
     */
    set blockPublicAcls(value) {
        this._blockPublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicAcls() {
        this._blockPublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicAclsInput() {
        return this._blockPublicAcls;
    }
    /**
     * @stability stable
     */
    get blockPublicPolicy() {
        return this.getBooleanAttribute('block_public_policy');
    }
    /**
     * @stability stable
     */
    set blockPublicPolicy(value) {
        this._blockPublicPolicy = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicPolicy() {
        this._blockPublicPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicPolicyInput() {
        return this._blockPublicPolicy;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ignorePublicAcls() {
        return this.getBooleanAttribute('ignore_public_acls');
    }
    /**
     * @stability stable
     */
    set ignorePublicAcls(value) {
        this._ignorePublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetIgnorePublicAcls() {
        this._ignorePublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ignorePublicAclsInput() {
        return this._ignorePublicAcls;
    }
    /**
     * @stability stable
     */
    get restrictPublicBuckets() {
        return this.getBooleanAttribute('restrict_public_buckets');
    }
    /**
     * @stability stable
     */
    set restrictPublicBuckets(value) {
        this._restrictPublicBuckets = value;
    }
    /**
     * @stability stable
     */
    resetRestrictPublicBuckets() {
        this._restrictPublicBuckets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictPublicBucketsInput() {
        return this._restrictPublicBuckets;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            block_public_acls: cdktf.booleanToTerraform(this._blockPublicAcls),
            block_public_policy: cdktf.booleanToTerraform(this._blockPublicPolicy),
            ignore_public_acls: cdktf.booleanToTerraform(this._ignorePublicAcls),
            restrict_public_buckets: cdktf.booleanToTerraform(this._restrictPublicBuckets),
        };
    }
}
exports.S3AccountPublicAccessBlock = S3AccountPublicAccessBlock;
_a = JSII_RTTI_SYMBOL_1;
S3AccountPublicAccessBlock[_a] = { fqn: "@cdktf/provider-aws.s3.S3AccountPublicAccessBlock", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
S3AccountPublicAccessBlock.tfResourceType = "aws_s3_account_public_access_block";
