// @cdktf/provider-aws/lib/s3/s3control-multi-region-access-point.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3ControlMultiRegionAccessPoint = exports.S3ControlMultiRegionAccessPointTimeoutsOutputReference = exports.s3ControlMultiRegionAccessPointTimeoutsToTerraform = exports.S3ControlMultiRegionAccessPointDetailsOutputReference = exports.s3ControlMultiRegionAccessPointDetailsToTerraform = exports.s3ControlMultiRegionAccessPointDetailsRegionToTerraform = exports.S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference = exports.s3ControlMultiRegionAccessPointDetailsPublicAccessBlockToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3ControlMultiRegionAccessPointDetailsPublicAccessBlockToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        block_public_acls: cdktf.booleanToTerraform(struct.blockPublicAcls),
        block_public_policy: cdktf.booleanToTerraform(struct.blockPublicPolicy),
        ignore_public_acls: cdktf.booleanToTerraform(struct.ignorePublicAcls),
        restrict_public_buckets: cdktf.booleanToTerraform(struct.restrictPublicBuckets),
    };
}
exports.s3ControlMultiRegionAccessPointDetailsPublicAccessBlockToTerraform = s3ControlMultiRegionAccessPointDetailsPublicAccessBlockToTerraform;
/**
 * @stability stable
 */
class S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._blockPublicAcls) {
            hasAnyValues = true;
            internalValueResult.blockPublicAcls = this._blockPublicAcls;
        }
        if (this._blockPublicPolicy) {
            hasAnyValues = true;
            internalValueResult.blockPublicPolicy = this._blockPublicPolicy;
        }
        if (this._ignorePublicAcls) {
            hasAnyValues = true;
            internalValueResult.ignorePublicAcls = this._ignorePublicAcls;
        }
        if (this._restrictPublicBuckets) {
            hasAnyValues = true;
            internalValueResult.restrictPublicBuckets = this._restrictPublicBuckets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._blockPublicAcls = undefined;
            this._blockPublicPolicy = undefined;
            this._ignorePublicAcls = undefined;
            this._restrictPublicBuckets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._blockPublicAcls = value.blockPublicAcls;
            this._blockPublicPolicy = value.blockPublicPolicy;
            this._ignorePublicAcls = value.ignorePublicAcls;
            this._restrictPublicBuckets = value.restrictPublicBuckets;
        }
    }
    /**
     * @stability stable
     */
    get blockPublicAcls() {
        return this.getBooleanAttribute('block_public_acls');
    }
    /**
     * @stability stable
     */
    set blockPublicAcls(value) {
        this._blockPublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicAcls() {
        this._blockPublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicAclsInput() {
        return this._blockPublicAcls;
    }
    /**
     * @stability stable
     */
    get blockPublicPolicy() {
        return this.getBooleanAttribute('block_public_policy');
    }
    /**
     * @stability stable
     */
    set blockPublicPolicy(value) {
        this._blockPublicPolicy = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicPolicy() {
        this._blockPublicPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicPolicyInput() {
        return this._blockPublicPolicy;
    }
    /**
     * @stability stable
     */
    get ignorePublicAcls() {
        return this.getBooleanAttribute('ignore_public_acls');
    }
    /**
     * @stability stable
     */
    set ignorePublicAcls(value) {
        this._ignorePublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetIgnorePublicAcls() {
        this._ignorePublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ignorePublicAclsInput() {
        return this._ignorePublicAcls;
    }
    /**
     * @stability stable
     */
    get restrictPublicBuckets() {
        return this.getBooleanAttribute('restrict_public_buckets');
    }
    /**
     * @stability stable
     */
    set restrictPublicBuckets(value) {
        this._restrictPublicBuckets = value;
    }
    /**
     * @stability stable
     */
    resetRestrictPublicBuckets() {
        this._restrictPublicBuckets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictPublicBucketsInput() {
        return this._restrictPublicBuckets;
    }
}
exports.S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference = S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3.S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference", version: "3.0.1" };
function s3ControlMultiRegionAccessPointDetailsRegionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
    };
}
exports.s3ControlMultiRegionAccessPointDetailsRegionToTerraform = s3ControlMultiRegionAccessPointDetailsRegionToTerraform;
function s3ControlMultiRegionAccessPointDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        public_access_block: s3ControlMultiRegionAccessPointDetailsPublicAccessBlockToTerraform(struct.publicAccessBlock),
        region: cdktf.listMapper(s3ControlMultiRegionAccessPointDetailsRegionToTerraform)(struct.region),
    };
}
exports.s3ControlMultiRegionAccessPointDetailsToTerraform = s3ControlMultiRegionAccessPointDetailsToTerraform;
/**
 * @stability stable
 */
class S3ControlMultiRegionAccessPointDetailsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // public_access_block - computed: false, optional: true, required: false
        this._publicAccessBlock = new S3ControlMultiRegionAccessPointDetailsPublicAccessBlockOutputReference(this, "public_access_block", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if ((_e = this._publicAccessBlock) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.publicAccessBlock = (_f = this._publicAccessBlock) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        if (this._region) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._publicAccessBlock.internalValue = undefined;
            this._region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._publicAccessBlock.internalValue = value.publicAccessBlock;
            this._region = value.region;
        }
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get publicAccessBlock() {
        return this._publicAccessBlock;
    }
    /**
     * @stability stable
     */
    putPublicAccessBlock(value) {
        this._publicAccessBlock.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPublicAccessBlock() {
        this._publicAccessBlock.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicAccessBlockInput() {
        return this._publicAccessBlock.internalValue;
    }
    /**
     * @stability stable
     */
    get region() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('region');
    }
    /**
     * @stability stable
     */
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionInput() {
        return this._region;
    }
}
exports.S3ControlMultiRegionAccessPointDetailsOutputReference = S3ControlMultiRegionAccessPointDetailsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3ControlMultiRegionAccessPointDetailsOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3.S3ControlMultiRegionAccessPointDetailsOutputReference", version: "3.0.1" };
function s3ControlMultiRegionAccessPointTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.s3ControlMultiRegionAccessPointTimeoutsToTerraform = s3ControlMultiRegionAccessPointTimeoutsToTerraform;
/**
 * @stability stable
 */
class S3ControlMultiRegionAccessPointTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.S3ControlMultiRegionAccessPointTimeoutsOutputReference = S3ControlMultiRegionAccessPointTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3ControlMultiRegionAccessPointTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3.S3ControlMultiRegionAccessPointTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3control_multi_region_access_point.html aws_s3control_multi_region_access_point}.
 *
 * @stability stable
 */
class S3ControlMultiRegionAccessPoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3control_multi_region_access_point.html aws_s3control_multi_region_access_point} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3control_multi_region_access_point',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // details - computed: false, optional: false, required: true
        this._details = new S3ControlMultiRegionAccessPointDetailsOutputReference(this, "details", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new S3ControlMultiRegionAccessPointTimeoutsOutputReference(this, "timeouts", true);
        this._accountId = config.accountId;
        this._details.internalValue = config.details;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // alias - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get alias() {
        return this.getStringAttribute('alias');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get details() {
        return this._details;
    }
    /**
     * @stability stable
     */
    putDetails(value) {
        this._details.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get detailsInput() {
        return this._details.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            details: s3ControlMultiRegionAccessPointDetailsToTerraform(this._details.internalValue),
            timeouts: s3ControlMultiRegionAccessPointTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.S3ControlMultiRegionAccessPoint = S3ControlMultiRegionAccessPoint;
_d = JSII_RTTI_SYMBOL_1;
S3ControlMultiRegionAccessPoint[_d] = { fqn: "@cdktf/provider-aws.s3.S3ControlMultiRegionAccessPoint", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
S3ControlMultiRegionAccessPoint.tfResourceType = "aws_s3control_multi_region_access_point";
