// @cdktf/provider-aws/lib/directoryservice/directory-service-conditional-forwarder.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectoryServiceConditionalForwarder = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/directory_service_conditional_forwarder.html aws_directory_service_conditional_forwarder}.
 *
 * @stability stable
 */
class DirectoryServiceConditionalForwarder extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/directory_service_conditional_forwarder.html aws_directory_service_conditional_forwarder} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_directory_service_conditional_forwarder',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._directoryId = config.directoryId;
        this._dnsIps = config.dnsIps;
        this._remoteDomainName = config.remoteDomainName;
    }
    /**
     * @stability stable
     */
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    /**
     * @stability stable
     */
    set directoryId(value) {
        this._directoryId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get directoryIdInput() {
        return this._directoryId;
    }
    /**
     * @stability stable
     */
    get dnsIps() {
        return this.getListAttribute('dns_ips');
    }
    /**
     * @stability stable
     */
    set dnsIps(value) {
        this._dnsIps = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dnsIpsInput() {
        return this._dnsIps;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get remoteDomainName() {
        return this.getStringAttribute('remote_domain_name');
    }
    /**
     * @stability stable
     */
    set remoteDomainName(value) {
        this._remoteDomainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get remoteDomainNameInput() {
        return this._remoteDomainName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            directory_id: cdktf.stringToTerraform(this._directoryId),
            dns_ips: cdktf.listMapper(cdktf.stringToTerraform)(this._dnsIps),
            remote_domain_name: cdktf.stringToTerraform(this._remoteDomainName),
        };
    }
}
exports.DirectoryServiceConditionalForwarder = DirectoryServiceConditionalForwarder;
_a = JSII_RTTI_SYMBOL_1;
DirectoryServiceConditionalForwarder[_a] = { fqn: "@cdktf/provider-aws.directoryservice.DirectoryServiceConditionalForwarder", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DirectoryServiceConditionalForwarder.tfResourceType = "aws_directory_service_conditional_forwarder";
