// @cdktf/provider-aws/lib/ec2/data-aws-instance.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsInstance = exports.dataAwsInstanceFilterToTerraform = exports.DataAwsInstanceRootBlockDevice = exports.DataAwsInstanceMetadataOptions = exports.DataAwsInstanceEphemeralBlockDevice = exports.DataAwsInstanceEnclaveOptions = exports.DataAwsInstanceEbsBlockDevice = exports.DataAwsInstanceCreditSpecification = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsInstanceCreditSpecification extends cdktf.ComplexComputedList {
    // cpu_credits - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cpuCredits() {
        return this.getStringAttribute('cpu_credits');
    }
}
exports.DataAwsInstanceCreditSpecification = DataAwsInstanceCreditSpecification;
_a = JSII_RTTI_SYMBOL_1;
DataAwsInstanceCreditSpecification[_a] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceCreditSpecification", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsInstanceEbsBlockDevice extends cdktf.ComplexComputedList {
    // delete_on_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    // device_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // encrypted - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // snapshot_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // tags - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    // throughput - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeId() {
        return this.getStringAttribute('volume_id');
    }
    // volume_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsInstanceEbsBlockDevice = DataAwsInstanceEbsBlockDevice;
_b = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEbsBlockDevice[_b] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceEbsBlockDevice", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsInstanceEnclaveOptions extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsInstanceEnclaveOptions = DataAwsInstanceEnclaveOptions;
_c = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEnclaveOptions[_c] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceEnclaveOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsInstanceEphemeralBlockDevice extends cdktf.ComplexComputedList {
    // device_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // no_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get noDevice() {
        return this.getBooleanAttribute('no_device');
    }
    // virtual_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualName() {
        return this.getStringAttribute('virtual_name');
    }
}
exports.DataAwsInstanceEphemeralBlockDevice = DataAwsInstanceEphemeralBlockDevice;
_d = JSII_RTTI_SYMBOL_1;
DataAwsInstanceEphemeralBlockDevice[_d] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceEphemeralBlockDevice", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsInstanceMetadataOptions extends cdktf.ComplexComputedList {
    // http_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpEndpoint() {
        return this.getStringAttribute('http_endpoint');
    }
    // http_put_response_hop_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpPutResponseHopLimit() {
        return this.getNumberAttribute('http_put_response_hop_limit');
    }
    // http_tokens - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpTokens() {
        return this.getStringAttribute('http_tokens');
    }
}
exports.DataAwsInstanceMetadataOptions = DataAwsInstanceMetadataOptions;
_e = JSII_RTTI_SYMBOL_1;
DataAwsInstanceMetadataOptions[_e] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceMetadataOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsInstanceRootBlockDevice extends cdktf.ComplexComputedList {
    // delete_on_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    // device_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // encrypted - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // tags - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    // throughput - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeId() {
        return this.getStringAttribute('volume_id');
    }
    // volume_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsInstanceRootBlockDevice = DataAwsInstanceRootBlockDevice;
_f = JSII_RTTI_SYMBOL_1;
DataAwsInstanceRootBlockDevice[_f] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstanceRootBlockDevice", version: "3.0.1" };
function dataAwsInstanceFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsInstanceFilterToTerraform = dataAwsInstanceFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/instance.html aws_instance}.
 *
 * @stability stable
 */
class DataAwsInstance extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/instance.html aws_instance} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._getPasswordData = config.fetchPasswordData;
        this._getUserData = config.fetchUserData;
        this._instanceId = config.instanceId;
        this._instanceTags = config.instanceTags;
        this._tags = config.tags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ami - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ami() {
        return this.getStringAttribute('ami');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // associate_public_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associatePublicIpAddress() {
        return this.getBooleanAttribute('associate_public_ip_address');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // credit_specification - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    creditSpecification(index) {
        return new DataAwsInstanceCreditSpecification(this, 'credit_specification', index);
    }
    // disable_api_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get disableApiTermination() {
        return this.getBooleanAttribute('disable_api_termination');
    }
    // ebs_block_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ebsBlockDevice(index) {
        return new DataAwsInstanceEbsBlockDevice(this, 'ebs_block_device', index);
    }
    // ebs_optimized - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    // enclave_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    enclaveOptions(index) {
        return new DataAwsInstanceEnclaveOptions(this, 'enclave_options', index);
    }
    // ephemeral_block_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ephemeralBlockDevice(index) {
        return new DataAwsInstanceEphemeralBlockDevice(this, 'ephemeral_block_device', index);
    }
    /**
     * @stability stable
     */
    get fetchPasswordData() {
        return this.getBooleanAttribute('get_password_data');
    }
    /**
     * @stability stable
     */
    set fetchPasswordData(value) {
        this._getPasswordData = value;
    }
    /**
     * @stability stable
     */
    resetFetchPasswordData() {
        this._getPasswordData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fetchPasswordDataInput() {
        return this._getPasswordData;
    }
    /**
     * @stability stable
     */
    get fetchUserData() {
        return this.getBooleanAttribute('get_user_data');
    }
    /**
     * @stability stable
     */
    set fetchUserData(value) {
        this._getUserData = value;
    }
    /**
     * @stability stable
     */
    resetFetchUserData() {
        this._getUserData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fetchUserDataInput() {
        return this._getUserData;
    }
    // host_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostId() {
        return this.getStringAttribute('host_id');
    }
    // iam_instance_profile - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iamInstanceProfile() {
        return this.getStringAttribute('iam_instance_profile');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    /**
     * @stability stable
     */
    set instanceId(value) {
        this._instanceId = value;
    }
    /**
     * @stability stable
     */
    resetInstanceId() {
        this._instanceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceIdInput() {
        return this._instanceId;
    }
    // instance_state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceState() {
        return this.getStringAttribute('instance_state');
    }
    /**
     * @stability stable
     */
    get instanceTags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('instance_tags');
    }
    /**
     * @stability stable
     */
    set instanceTags(value) {
        this._instanceTags = value;
    }
    /**
     * @stability stable
     */
    resetInstanceTags() {
        this._instanceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTagsInput() {
        return this._instanceTags;
    }
    // instance_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // ipv6_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6Addresses() {
        return this.getListAttribute('ipv6_addresses');
    }
    // key_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    // metadata_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    metadataOptions(index) {
        return new DataAwsInstanceMetadataOptions(this, 'metadata_options', index);
    }
    // monitoring - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get monitoring() {
        return this.getBooleanAttribute('monitoring');
    }
    // network_interface_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    // outpost_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get outpostArn() {
        return this.getStringAttribute('outpost_arn');
    }
    // password_data - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get passwordData() {
        return this.getStringAttribute('password_data');
    }
    // placement_group - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get placementGroup() {
        return this.getStringAttribute('placement_group');
    }
    // placement_partition_number - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get placementPartitionNumber() {
        return this.getNumberAttribute('placement_partition_number');
    }
    // private_dns - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateDns() {
        return this.getStringAttribute('private_dns');
    }
    // private_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIp() {
        return this.getStringAttribute('private_ip');
    }
    // public_dns - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicDns() {
        return this.getStringAttribute('public_dns');
    }
    // public_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicIp() {
        return this.getStringAttribute('public_ip');
    }
    // root_block_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    rootBlockDevice(index) {
        return new DataAwsInstanceRootBlockDevice(this, 'root_block_device', index);
    }
    // secondary_private_ips - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secondaryPrivateIps() {
        return this.getListAttribute('secondary_private_ips');
    }
    // security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    // source_dest_check - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceDestCheck() {
        return this.getBooleanAttribute('source_dest_check');
    }
    // subnet_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // tenancy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
    // user_data - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userData() {
        return this.getStringAttribute('user_data');
    }
    // user_data_base64 - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userDataBase64() {
        return this.getStringAttribute('user_data_base64');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            get_password_data: cdktf.booleanToTerraform(this._getPasswordData),
            get_user_data: cdktf.booleanToTerraform(this._getUserData),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            instance_tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._instanceTags),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsInstanceFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsInstance = DataAwsInstance;
_g = JSII_RTTI_SYMBOL_1;
DataAwsInstance[_g] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstance", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsInstance.tfResourceType = "aws_instance";
