// @cdktf/provider-aws/lib/ec2/ec2-transit-gateway-peering-attachment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2TransitGatewayPeeringAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway_peering_attachment.html aws_ec2_transit_gateway_peering_attachment}.
 *
 * @stability stable
 */
class Ec2TransitGatewayPeeringAttachment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway_peering_attachment.html aws_ec2_transit_gateway_peering_attachment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_transit_gateway_peering_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._peerAccountId = config.peerAccountId;
        this._peerRegion = config.peerRegion;
        this._peerTransitGatewayId = config.peerTransitGatewayId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._transitGatewayId = config.transitGatewayId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get peerAccountId() {
        return this.getStringAttribute('peer_account_id');
    }
    /**
     * @stability stable
     */
    set peerAccountId(value) {
        this._peerAccountId = value;
    }
    /**
     * @stability stable
     */
    resetPeerAccountId() {
        this._peerAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerAccountIdInput() {
        return this._peerAccountId;
    }
    /**
     * @stability stable
     */
    get peerRegion() {
        return this.getStringAttribute('peer_region');
    }
    /**
     * @stability stable
     */
    set peerRegion(value) {
        this._peerRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerRegionInput() {
        return this._peerRegion;
    }
    /**
     * @stability stable
     */
    get peerTransitGatewayId() {
        return this.getStringAttribute('peer_transit_gateway_id');
    }
    /**
     * @stability stable
     */
    set peerTransitGatewayId(value) {
        this._peerTransitGatewayId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerTransitGatewayIdInput() {
        return this._peerTransitGatewayId;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get transitGatewayId() {
        return this.getStringAttribute('transit_gateway_id');
    }
    /**
     * @stability stable
     */
    set transitGatewayId(value) {
        this._transitGatewayId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get transitGatewayIdInput() {
        return this._transitGatewayId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            peer_account_id: cdktf.stringToTerraform(this._peerAccountId),
            peer_region: cdktf.stringToTerraform(this._peerRegion),
            peer_transit_gateway_id: cdktf.stringToTerraform(this._peerTransitGatewayId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            transit_gateway_id: cdktf.stringToTerraform(this._transitGatewayId),
        };
    }
}
exports.Ec2TransitGatewayPeeringAttachment = Ec2TransitGatewayPeeringAttachment;
_a = JSII_RTTI_SYMBOL_1;
Ec2TransitGatewayPeeringAttachment[_a] = { fqn: "@cdktf/provider-aws.ec2.Ec2TransitGatewayPeeringAttachment", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Ec2TransitGatewayPeeringAttachment.tfResourceType = "aws_ec2_transit_gateway_peering_attachment";
