// @cdktf/provider-aws/lib/ec2/launch-template.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaunchTemplate = exports.launchTemplateTagSpecificationsToTerraform = exports.LaunchTemplatePlacementOutputReference = exports.launchTemplatePlacementToTerraform = exports.launchTemplateNetworkInterfacesToTerraform = exports.LaunchTemplateMonitoringOutputReference = exports.launchTemplateMonitoringToTerraform = exports.LaunchTemplateMetadataOptionsOutputReference = exports.launchTemplateMetadataOptionsToTerraform = exports.launchTemplateLicenseSpecificationToTerraform = exports.LaunchTemplateInstanceMarketOptionsOutputReference = exports.launchTemplateInstanceMarketOptionsToTerraform = exports.LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference = exports.launchTemplateInstanceMarketOptionsSpotOptionsToTerraform = exports.LaunchTemplateIamInstanceProfileOutputReference = exports.launchTemplateIamInstanceProfileToTerraform = exports.LaunchTemplateHibernationOptionsOutputReference = exports.launchTemplateHibernationOptionsToTerraform = exports.LaunchTemplateEnclaveOptionsOutputReference = exports.launchTemplateEnclaveOptionsToTerraform = exports.LaunchTemplateElasticInferenceAcceleratorOutputReference = exports.launchTemplateElasticInferenceAcceleratorToTerraform = exports.launchTemplateElasticGpuSpecificationsToTerraform = exports.LaunchTemplateCreditSpecificationOutputReference = exports.launchTemplateCreditSpecificationToTerraform = exports.LaunchTemplateCpuOptionsOutputReference = exports.launchTemplateCpuOptionsToTerraform = exports.LaunchTemplateCapacityReservationSpecificationOutputReference = exports.launchTemplateCapacityReservationSpecificationToTerraform = exports.LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference = exports.launchTemplateCapacityReservationSpecificationCapacityReservationTargetToTerraform = exports.launchTemplateBlockDeviceMappingsToTerraform = exports.LaunchTemplateBlockDeviceMappingsEbsOutputReference = exports.launchTemplateBlockDeviceMappingsEbsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function launchTemplateBlockDeviceMappingsEbsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_on_termination: cdktf.stringToTerraform(struct.deleteOnTermination),
        encrypted: cdktf.stringToTerraform(struct.encrypted),
        iops: cdktf.numberToTerraform(struct.iops),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        snapshot_id: cdktf.stringToTerraform(struct.snapshotId),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.launchTemplateBlockDeviceMappingsEbsToTerraform = launchTemplateBlockDeviceMappingsEbsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateBlockDeviceMappingsEbsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteOnTermination) {
            hasAnyValues = true;
            internalValueResult.deleteOnTermination = this._deleteOnTermination;
        }
        if (this._encrypted) {
            hasAnyValues = true;
            internalValueResult.encrypted = this._encrypted;
        }
        if (this._iops) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._kmsKeyId) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._snapshotId) {
            hasAnyValues = true;
            internalValueResult.snapshotId = this._snapshotId;
        }
        if (this._throughput) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._volumeSize) {
            hasAnyValues = true;
            internalValueResult.volumeSize = this._volumeSize;
        }
        if (this._volumeType) {
            hasAnyValues = true;
            internalValueResult.volumeType = this._volumeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteOnTermination = undefined;
            this._encrypted = undefined;
            this._iops = undefined;
            this._kmsKeyId = undefined;
            this._snapshotId = undefined;
            this._throughput = undefined;
            this._volumeSize = undefined;
            this._volumeType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteOnTermination = value.deleteOnTermination;
            this._encrypted = value.encrypted;
            this._iops = value.iops;
            this._kmsKeyId = value.kmsKeyId;
            this._snapshotId = value.snapshotId;
            this._throughput = value.throughput;
            this._volumeSize = value.volumeSize;
            this._volumeType = value.volumeType;
        }
    }
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getStringAttribute('delete_on_termination');
    }
    /**
     * @stability stable
     */
    set deleteOnTermination(value) {
        this._deleteOnTermination = value;
    }
    /**
     * @stability stable
     */
    resetDeleteOnTermination() {
        this._deleteOnTermination = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteOnTerminationInput() {
        return this._deleteOnTermination;
    }
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getStringAttribute('encrypted');
    }
    /**
     * @stability stable
     */
    set encrypted(value) {
        this._encrypted = value;
    }
    /**
     * @stability stable
     */
    resetEncrypted() {
        this._encrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptedInput() {
        return this._encrypted;
    }
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    /**
     * @stability stable
     */
    set iops(value) {
        this._iops = value;
    }
    /**
     * @stability stable
     */
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iopsInput() {
        return this._iops;
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    /**
     * @stability stable
     */
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    /**
     * @stability stable
     */
    set snapshotId(value) {
        this._snapshotId = value;
    }
    /**
     * @stability stable
     */
    resetSnapshotId() {
        this._snapshotId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snapshotIdInput() {
        return this._snapshotId;
    }
    /**
     * @stability stable
     */
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    /**
     * @stability stable
     */
    set throughput(value) {
        this._throughput = value;
    }
    /**
     * @stability stable
     */
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throughputInput() {
        return this._throughput;
    }
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    /**
     * @stability stable
     */
    set volumeSize(value) {
        this._volumeSize = value;
    }
    /**
     * @stability stable
     */
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get volumeSizeInput() {
        return this._volumeSize;
    }
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    /**
     * @stability stable
     */
    set volumeType(value) {
        this._volumeType = value;
    }
    /**
     * @stability stable
     */
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get volumeTypeInput() {
        return this._volumeType;
    }
}
exports.LaunchTemplateBlockDeviceMappingsEbsOutputReference = LaunchTemplateBlockDeviceMappingsEbsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LaunchTemplateBlockDeviceMappingsEbsOutputReference[_a] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateBlockDeviceMappingsEbsOutputReference", version: "3.0.1" };
function launchTemplateBlockDeviceMappingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        device_name: cdktf.stringToTerraform(struct.deviceName),
        no_device: cdktf.stringToTerraform(struct.noDevice),
        virtual_name: cdktf.stringToTerraform(struct.virtualName),
        ebs: launchTemplateBlockDeviceMappingsEbsToTerraform(struct.ebs),
    };
}
exports.launchTemplateBlockDeviceMappingsToTerraform = launchTemplateBlockDeviceMappingsToTerraform;
function launchTemplateCapacityReservationSpecificationCapacityReservationTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        capacity_reservation_id: cdktf.stringToTerraform(struct.capacityReservationId),
    };
}
exports.launchTemplateCapacityReservationSpecificationCapacityReservationTargetToTerraform = launchTemplateCapacityReservationSpecificationCapacityReservationTargetToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._capacityReservationId) {
            hasAnyValues = true;
            internalValueResult.capacityReservationId = this._capacityReservationId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._capacityReservationId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._capacityReservationId = value.capacityReservationId;
        }
    }
    /**
     * @stability stable
     */
    get capacityReservationId() {
        return this.getStringAttribute('capacity_reservation_id');
    }
    /**
     * @stability stable
     */
    set capacityReservationId(value) {
        this._capacityReservationId = value;
    }
    /**
     * @stability stable
     */
    resetCapacityReservationId() {
        this._capacityReservationId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityReservationIdInput() {
        return this._capacityReservationId;
    }
}
exports.LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference = LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference[_b] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference", version: "3.0.1" };
function launchTemplateCapacityReservationSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        capacity_reservation_preference: cdktf.stringToTerraform(struct.capacityReservationPreference),
        capacity_reservation_target: launchTemplateCapacityReservationSpecificationCapacityReservationTargetToTerraform(struct.capacityReservationTarget),
    };
}
exports.launchTemplateCapacityReservationSpecificationToTerraform = launchTemplateCapacityReservationSpecificationToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateCapacityReservationSpecificationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // capacity_reservation_target - computed: false, optional: true, required: false
        this._capacityReservationTarget = new LaunchTemplateCapacityReservationSpecificationCapacityReservationTargetOutputReference(this, "capacity_reservation_target", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _r, _s;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._capacityReservationPreference) {
            hasAnyValues = true;
            internalValueResult.capacityReservationPreference = this._capacityReservationPreference;
        }
        if ((_r = this._capacityReservationTarget) === null || _r === void 0 ? void 0 : _r.internalValue) {
            hasAnyValues = true;
            internalValueResult.capacityReservationTarget = (_s = this._capacityReservationTarget) === null || _s === void 0 ? void 0 : _s.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._capacityReservationPreference = undefined;
            this._capacityReservationTarget.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._capacityReservationPreference = value.capacityReservationPreference;
            this._capacityReservationTarget.internalValue = value.capacityReservationTarget;
        }
    }
    /**
     * @stability stable
     */
    get capacityReservationPreference() {
        return this.getStringAttribute('capacity_reservation_preference');
    }
    /**
     * @stability stable
     */
    set capacityReservationPreference(value) {
        this._capacityReservationPreference = value;
    }
    /**
     * @stability stable
     */
    resetCapacityReservationPreference() {
        this._capacityReservationPreference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityReservationPreferenceInput() {
        return this._capacityReservationPreference;
    }
    /**
     * @stability stable
     */
    get capacityReservationTarget() {
        return this._capacityReservationTarget;
    }
    /**
     * @stability stable
     */
    putCapacityReservationTarget(value) {
        this._capacityReservationTarget.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCapacityReservationTarget() {
        this._capacityReservationTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityReservationTargetInput() {
        return this._capacityReservationTarget.internalValue;
    }
}
exports.LaunchTemplateCapacityReservationSpecificationOutputReference = LaunchTemplateCapacityReservationSpecificationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LaunchTemplateCapacityReservationSpecificationOutputReference[_c] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateCapacityReservationSpecificationOutputReference", version: "3.0.1" };
function launchTemplateCpuOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        core_count: cdktf.numberToTerraform(struct.coreCount),
        threads_per_core: cdktf.numberToTerraform(struct.threadsPerCore),
    };
}
exports.launchTemplateCpuOptionsToTerraform = launchTemplateCpuOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateCpuOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coreCount) {
            hasAnyValues = true;
            internalValueResult.coreCount = this._coreCount;
        }
        if (this._threadsPerCore) {
            hasAnyValues = true;
            internalValueResult.threadsPerCore = this._threadsPerCore;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coreCount = undefined;
            this._threadsPerCore = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coreCount = value.coreCount;
            this._threadsPerCore = value.threadsPerCore;
        }
    }
    /**
     * @stability stable
     */
    get coreCount() {
        return this.getNumberAttribute('core_count');
    }
    /**
     * @stability stable
     */
    set coreCount(value) {
        this._coreCount = value;
    }
    /**
     * @stability stable
     */
    resetCoreCount() {
        this._coreCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get coreCountInput() {
        return this._coreCount;
    }
    /**
     * @stability stable
     */
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
    /**
     * @stability stable
     */
    set threadsPerCore(value) {
        this._threadsPerCore = value;
    }
    /**
     * @stability stable
     */
    resetThreadsPerCore() {
        this._threadsPerCore = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get threadsPerCoreInput() {
        return this._threadsPerCore;
    }
}
exports.LaunchTemplateCpuOptionsOutputReference = LaunchTemplateCpuOptionsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
LaunchTemplateCpuOptionsOutputReference[_d] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateCpuOptionsOutputReference", version: "3.0.1" };
function launchTemplateCreditSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_credits: cdktf.stringToTerraform(struct.cpuCredits),
    };
}
exports.launchTemplateCreditSpecificationToTerraform = launchTemplateCreditSpecificationToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateCreditSpecificationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuCredits) {
            hasAnyValues = true;
            internalValueResult.cpuCredits = this._cpuCredits;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuCredits = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuCredits = value.cpuCredits;
        }
    }
    /**
     * @stability stable
     */
    get cpuCredits() {
        return this.getStringAttribute('cpu_credits');
    }
    /**
     * @stability stable
     */
    set cpuCredits(value) {
        this._cpuCredits = value;
    }
    /**
     * @stability stable
     */
    resetCpuCredits() {
        this._cpuCredits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cpuCreditsInput() {
        return this._cpuCredits;
    }
}
exports.LaunchTemplateCreditSpecificationOutputReference = LaunchTemplateCreditSpecificationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
LaunchTemplateCreditSpecificationOutputReference[_e] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateCreditSpecificationOutputReference", version: "3.0.1" };
function launchTemplateElasticGpuSpecificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.launchTemplateElasticGpuSpecificationsToTerraform = launchTemplateElasticGpuSpecificationsToTerraform;
function launchTemplateElasticInferenceAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.launchTemplateElasticInferenceAcceleratorToTerraform = launchTemplateElasticInferenceAcceleratorToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateElasticInferenceAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.LaunchTemplateElasticInferenceAcceleratorOutputReference = LaunchTemplateElasticInferenceAcceleratorOutputReference;
_f = JSII_RTTI_SYMBOL_1;
LaunchTemplateElasticInferenceAcceleratorOutputReference[_f] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateElasticInferenceAcceleratorOutputReference", version: "3.0.1" };
function launchTemplateEnclaveOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.launchTemplateEnclaveOptionsToTerraform = launchTemplateEnclaveOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateEnclaveOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
}
exports.LaunchTemplateEnclaveOptionsOutputReference = LaunchTemplateEnclaveOptionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
LaunchTemplateEnclaveOptionsOutputReference[_g] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateEnclaveOptionsOutputReference", version: "3.0.1" };
function launchTemplateHibernationOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        configured: cdktf.booleanToTerraform(struct.configured),
    };
}
exports.launchTemplateHibernationOptionsToTerraform = launchTemplateHibernationOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateHibernationOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._configured) {
            hasAnyValues = true;
            internalValueResult.configured = this._configured;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._configured = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._configured = value.configured;
        }
    }
    /**
     * @stability stable
     */
    get configured() {
        return this.getBooleanAttribute('configured');
    }
    /**
     * @stability stable
     */
    set configured(value) {
        this._configured = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get configuredInput() {
        return this._configured;
    }
}
exports.LaunchTemplateHibernationOptionsOutputReference = LaunchTemplateHibernationOptionsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
LaunchTemplateHibernationOptionsOutputReference[_h] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateHibernationOptionsOutputReference", version: "3.0.1" };
function launchTemplateIamInstanceProfileToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.launchTemplateIamInstanceProfileToTerraform = launchTemplateIamInstanceProfileToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateIamInstanceProfileOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._name = value.name;
        }
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    /**
     * @stability stable
     */
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
}
exports.LaunchTemplateIamInstanceProfileOutputReference = LaunchTemplateIamInstanceProfileOutputReference;
_j = JSII_RTTI_SYMBOL_1;
LaunchTemplateIamInstanceProfileOutputReference[_j] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateIamInstanceProfileOutputReference", version: "3.0.1" };
function launchTemplateInstanceMarketOptionsSpotOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        block_duration_minutes: cdktf.numberToTerraform(struct.blockDurationMinutes),
        instance_interruption_behavior: cdktf.stringToTerraform(struct.instanceInterruptionBehavior),
        max_price: cdktf.stringToTerraform(struct.maxPrice),
        spot_instance_type: cdktf.stringToTerraform(struct.spotInstanceType),
        valid_until: cdktf.stringToTerraform(struct.validUntil),
    };
}
exports.launchTemplateInstanceMarketOptionsSpotOptionsToTerraform = launchTemplateInstanceMarketOptionsSpotOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._blockDurationMinutes) {
            hasAnyValues = true;
            internalValueResult.blockDurationMinutes = this._blockDurationMinutes;
        }
        if (this._instanceInterruptionBehavior) {
            hasAnyValues = true;
            internalValueResult.instanceInterruptionBehavior = this._instanceInterruptionBehavior;
        }
        if (this._maxPrice) {
            hasAnyValues = true;
            internalValueResult.maxPrice = this._maxPrice;
        }
        if (this._spotInstanceType) {
            hasAnyValues = true;
            internalValueResult.spotInstanceType = this._spotInstanceType;
        }
        if (this._validUntil) {
            hasAnyValues = true;
            internalValueResult.validUntil = this._validUntil;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._blockDurationMinutes = undefined;
            this._instanceInterruptionBehavior = undefined;
            this._maxPrice = undefined;
            this._spotInstanceType = undefined;
            this._validUntil = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._blockDurationMinutes = value.blockDurationMinutes;
            this._instanceInterruptionBehavior = value.instanceInterruptionBehavior;
            this._maxPrice = value.maxPrice;
            this._spotInstanceType = value.spotInstanceType;
            this._validUntil = value.validUntil;
        }
    }
    /**
     * @stability stable
     */
    get blockDurationMinutes() {
        return this.getNumberAttribute('block_duration_minutes');
    }
    /**
     * @stability stable
     */
    set blockDurationMinutes(value) {
        this._blockDurationMinutes = value;
    }
    /**
     * @stability stable
     */
    resetBlockDurationMinutes() {
        this._blockDurationMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockDurationMinutesInput() {
        return this._blockDurationMinutes;
    }
    /**
     * @stability stable
     */
    get instanceInterruptionBehavior() {
        return this.getStringAttribute('instance_interruption_behavior');
    }
    /**
     * @stability stable
     */
    set instanceInterruptionBehavior(value) {
        this._instanceInterruptionBehavior = value;
    }
    /**
     * @stability stable
     */
    resetInstanceInterruptionBehavior() {
        this._instanceInterruptionBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceInterruptionBehaviorInput() {
        return this._instanceInterruptionBehavior;
    }
    /**
     * @stability stable
     */
    get maxPrice() {
        return this.getStringAttribute('max_price');
    }
    /**
     * @stability stable
     */
    set maxPrice(value) {
        this._maxPrice = value;
    }
    /**
     * @stability stable
     */
    resetMaxPrice() {
        this._maxPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxPriceInput() {
        return this._maxPrice;
    }
    /**
     * @stability stable
     */
    get spotInstanceType() {
        return this.getStringAttribute('spot_instance_type');
    }
    /**
     * @stability stable
     */
    set spotInstanceType(value) {
        this._spotInstanceType = value;
    }
    /**
     * @stability stable
     */
    resetSpotInstanceType() {
        this._spotInstanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spotInstanceTypeInput() {
        return this._spotInstanceType;
    }
    /**
     * @stability stable
     */
    get validUntil() {
        return this.getStringAttribute('valid_until');
    }
    /**
     * @stability stable
     */
    set validUntil(value) {
        this._validUntil = value;
    }
    /**
     * @stability stable
     */
    resetValidUntil() {
        this._validUntil = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get validUntilInput() {
        return this._validUntil;
    }
}
exports.LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference = LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference[_k] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference", version: "3.0.1" };
function launchTemplateInstanceMarketOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        market_type: cdktf.stringToTerraform(struct.marketType),
        spot_options: launchTemplateInstanceMarketOptionsSpotOptionsToTerraform(struct.spotOptions),
    };
}
exports.launchTemplateInstanceMarketOptionsToTerraform = launchTemplateInstanceMarketOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateInstanceMarketOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // spot_options - computed: false, optional: true, required: false
        this._spotOptions = new LaunchTemplateInstanceMarketOptionsSpotOptionsOutputReference(this, "spot_options", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _r, _s;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._marketType) {
            hasAnyValues = true;
            internalValueResult.marketType = this._marketType;
        }
        if ((_r = this._spotOptions) === null || _r === void 0 ? void 0 : _r.internalValue) {
            hasAnyValues = true;
            internalValueResult.spotOptions = (_s = this._spotOptions) === null || _s === void 0 ? void 0 : _s.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._marketType = undefined;
            this._spotOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._marketType = value.marketType;
            this._spotOptions.internalValue = value.spotOptions;
        }
    }
    /**
     * @stability stable
     */
    get marketType() {
        return this.getStringAttribute('market_type');
    }
    /**
     * @stability stable
     */
    set marketType(value) {
        this._marketType = value;
    }
    /**
     * @stability stable
     */
    resetMarketType() {
        this._marketType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get marketTypeInput() {
        return this._marketType;
    }
    /**
     * @stability stable
     */
    get spotOptions() {
        return this._spotOptions;
    }
    /**
     * @stability stable
     */
    putSpotOptions(value) {
        this._spotOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSpotOptions() {
        this._spotOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spotOptionsInput() {
        return this._spotOptions.internalValue;
    }
}
exports.LaunchTemplateInstanceMarketOptionsOutputReference = LaunchTemplateInstanceMarketOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
LaunchTemplateInstanceMarketOptionsOutputReference[_l] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateInstanceMarketOptionsOutputReference", version: "3.0.1" };
function launchTemplateLicenseSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        license_configuration_arn: cdktf.stringToTerraform(struct.licenseConfigurationArn),
    };
}
exports.launchTemplateLicenseSpecificationToTerraform = launchTemplateLicenseSpecificationToTerraform;
function launchTemplateMetadataOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        http_endpoint: cdktf.stringToTerraform(struct.httpEndpoint),
        http_protocol_ipv6: cdktf.stringToTerraform(struct.httpProtocolIpv6),
        http_put_response_hop_limit: cdktf.numberToTerraform(struct.httpPutResponseHopLimit),
        http_tokens: cdktf.stringToTerraform(struct.httpTokens),
    };
}
exports.launchTemplateMetadataOptionsToTerraform = launchTemplateMetadataOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateMetadataOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpEndpoint) {
            hasAnyValues = true;
            internalValueResult.httpEndpoint = this._httpEndpoint;
        }
        if (this._httpProtocolIpv6) {
            hasAnyValues = true;
            internalValueResult.httpProtocolIpv6 = this._httpProtocolIpv6;
        }
        if (this._httpPutResponseHopLimit) {
            hasAnyValues = true;
            internalValueResult.httpPutResponseHopLimit = this._httpPutResponseHopLimit;
        }
        if (this._httpTokens) {
            hasAnyValues = true;
            internalValueResult.httpTokens = this._httpTokens;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._httpEndpoint = undefined;
            this._httpProtocolIpv6 = undefined;
            this._httpPutResponseHopLimit = undefined;
            this._httpTokens = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._httpEndpoint = value.httpEndpoint;
            this._httpProtocolIpv6 = value.httpProtocolIpv6;
            this._httpPutResponseHopLimit = value.httpPutResponseHopLimit;
            this._httpTokens = value.httpTokens;
        }
    }
    /**
     * @stability stable
     */
    get httpEndpoint() {
        return this.getStringAttribute('http_endpoint');
    }
    /**
     * @stability stable
     */
    set httpEndpoint(value) {
        this._httpEndpoint = value;
    }
    /**
     * @stability stable
     */
    resetHttpEndpoint() {
        this._httpEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpEndpointInput() {
        return this._httpEndpoint;
    }
    /**
     * @stability stable
     */
    get httpProtocolIpv6() {
        return this.getStringAttribute('http_protocol_ipv6');
    }
    /**
     * @stability stable
     */
    set httpProtocolIpv6(value) {
        this._httpProtocolIpv6 = value;
    }
    /**
     * @stability stable
     */
    resetHttpProtocolIpv6() {
        this._httpProtocolIpv6 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpProtocolIpv6Input() {
        return this._httpProtocolIpv6;
    }
    /**
     * @stability stable
     */
    get httpPutResponseHopLimit() {
        return this.getNumberAttribute('http_put_response_hop_limit');
    }
    /**
     * @stability stable
     */
    set httpPutResponseHopLimit(value) {
        this._httpPutResponseHopLimit = value;
    }
    /**
     * @stability stable
     */
    resetHttpPutResponseHopLimit() {
        this._httpPutResponseHopLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpPutResponseHopLimitInput() {
        return this._httpPutResponseHopLimit;
    }
    /**
     * @stability stable
     */
    get httpTokens() {
        return this.getStringAttribute('http_tokens');
    }
    /**
     * @stability stable
     */
    set httpTokens(value) {
        this._httpTokens = value;
    }
    /**
     * @stability stable
     */
    resetHttpTokens() {
        this._httpTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpTokensInput() {
        return this._httpTokens;
    }
}
exports.LaunchTemplateMetadataOptionsOutputReference = LaunchTemplateMetadataOptionsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
LaunchTemplateMetadataOptionsOutputReference[_m] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateMetadataOptionsOutputReference", version: "3.0.1" };
function launchTemplateMonitoringToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.launchTemplateMonitoringToTerraform = launchTemplateMonitoringToTerraform;
/**
 * @stability stable
 */
class LaunchTemplateMonitoringOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
}
exports.LaunchTemplateMonitoringOutputReference = LaunchTemplateMonitoringOutputReference;
_o = JSII_RTTI_SYMBOL_1;
LaunchTemplateMonitoringOutputReference[_o] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplateMonitoringOutputReference", version: "3.0.1" };
function launchTemplateNetworkInterfacesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        associate_carrier_ip_address: cdktf.stringToTerraform(struct.associateCarrierIpAddress),
        associate_public_ip_address: cdktf.stringToTerraform(struct.associatePublicIpAddress),
        delete_on_termination: cdktf.stringToTerraform(struct.deleteOnTermination),
        description: cdktf.stringToTerraform(struct.description),
        device_index: cdktf.numberToTerraform(struct.deviceIndex),
        interface_type: cdktf.stringToTerraform(struct.interfaceType),
        ipv4_address_count: cdktf.numberToTerraform(struct.ipv4AddressCount),
        ipv4_addresses: cdktf.listMapper(cdktf.stringToTerraform)(struct.ipv4Addresses),
        ipv6_address_count: cdktf.numberToTerraform(struct.ipv6AddressCount),
        ipv6_addresses: cdktf.listMapper(cdktf.stringToTerraform)(struct.ipv6Addresses),
        network_card_index: cdktf.numberToTerraform(struct.networkCardIndex),
        network_interface_id: cdktf.stringToTerraform(struct.networkInterfaceId),
        private_ip_address: cdktf.stringToTerraform(struct.privateIpAddress),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroups),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
    };
}
exports.launchTemplateNetworkInterfacesToTerraform = launchTemplateNetworkInterfacesToTerraform;
function launchTemplatePlacementToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        affinity: cdktf.stringToTerraform(struct.affinity),
        availability_zone: cdktf.stringToTerraform(struct.availabilityZone),
        group_name: cdktf.stringToTerraform(struct.groupName),
        host_id: cdktf.stringToTerraform(struct.hostId),
        host_resource_group_arn: cdktf.stringToTerraform(struct.hostResourceGroupArn),
        partition_number: cdktf.numberToTerraform(struct.partitionNumber),
        spread_domain: cdktf.stringToTerraform(struct.spreadDomain),
        tenancy: cdktf.stringToTerraform(struct.tenancy),
    };
}
exports.launchTemplatePlacementToTerraform = launchTemplatePlacementToTerraform;
/**
 * @stability stable
 */
class LaunchTemplatePlacementOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._affinity) {
            hasAnyValues = true;
            internalValueResult.affinity = this._affinity;
        }
        if (this._availabilityZone) {
            hasAnyValues = true;
            internalValueResult.availabilityZone = this._availabilityZone;
        }
        if (this._groupName) {
            hasAnyValues = true;
            internalValueResult.groupName = this._groupName;
        }
        if (this._hostId) {
            hasAnyValues = true;
            internalValueResult.hostId = this._hostId;
        }
        if (this._hostResourceGroupArn) {
            hasAnyValues = true;
            internalValueResult.hostResourceGroupArn = this._hostResourceGroupArn;
        }
        if (this._partitionNumber) {
            hasAnyValues = true;
            internalValueResult.partitionNumber = this._partitionNumber;
        }
        if (this._spreadDomain) {
            hasAnyValues = true;
            internalValueResult.spreadDomain = this._spreadDomain;
        }
        if (this._tenancy) {
            hasAnyValues = true;
            internalValueResult.tenancy = this._tenancy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._affinity = undefined;
            this._availabilityZone = undefined;
            this._groupName = undefined;
            this._hostId = undefined;
            this._hostResourceGroupArn = undefined;
            this._partitionNumber = undefined;
            this._spreadDomain = undefined;
            this._tenancy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._affinity = value.affinity;
            this._availabilityZone = value.availabilityZone;
            this._groupName = value.groupName;
            this._hostId = value.hostId;
            this._hostResourceGroupArn = value.hostResourceGroupArn;
            this._partitionNumber = value.partitionNumber;
            this._spreadDomain = value.spreadDomain;
            this._tenancy = value.tenancy;
        }
    }
    /**
     * @stability stable
     */
    get affinity() {
        return this.getStringAttribute('affinity');
    }
    /**
     * @stability stable
     */
    set affinity(value) {
        this._affinity = value;
    }
    /**
     * @stability stable
     */
    resetAffinity() {
        this._affinity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get affinityInput() {
        return this._affinity;
    }
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    /**
     * @stability stable
     */
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    /**
     * @stability stable
     */
    resetAvailabilityZone() {
        this._availabilityZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    /**
     * @stability stable
     */
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    /**
     * @stability stable
     */
    set groupName(value) {
        this._groupName = value;
    }
    /**
     * @stability stable
     */
    resetGroupName() {
        this._groupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupNameInput() {
        return this._groupName;
    }
    /**
     * @stability stable
     */
    get hostId() {
        return this.getStringAttribute('host_id');
    }
    /**
     * @stability stable
     */
    set hostId(value) {
        this._hostId = value;
    }
    /**
     * @stability stable
     */
    resetHostId() {
        this._hostId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostIdInput() {
        return this._hostId;
    }
    /**
     * @stability stable
     */
    get hostResourceGroupArn() {
        return this.getStringAttribute('host_resource_group_arn');
    }
    /**
     * @stability stable
     */
    set hostResourceGroupArn(value) {
        this._hostResourceGroupArn = value;
    }
    /**
     * @stability stable
     */
    resetHostResourceGroupArn() {
        this._hostResourceGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostResourceGroupArnInput() {
        return this._hostResourceGroupArn;
    }
    /**
     * @stability stable
     */
    get partitionNumber() {
        return this.getNumberAttribute('partition_number');
    }
    /**
     * @stability stable
     */
    set partitionNumber(value) {
        this._partitionNumber = value;
    }
    /**
     * @stability stable
     */
    resetPartitionNumber() {
        this._partitionNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get partitionNumberInput() {
        return this._partitionNumber;
    }
    /**
     * @stability stable
     */
    get spreadDomain() {
        return this.getStringAttribute('spread_domain');
    }
    /**
     * @stability stable
     */
    set spreadDomain(value) {
        this._spreadDomain = value;
    }
    /**
     * @stability stable
     */
    resetSpreadDomain() {
        this._spreadDomain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spreadDomainInput() {
        return this._spreadDomain;
    }
    /**
     * @stability stable
     */
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
    /**
     * @stability stable
     */
    set tenancy(value) {
        this._tenancy = value;
    }
    /**
     * @stability stable
     */
    resetTenancy() {
        this._tenancy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tenancyInput() {
        return this._tenancy;
    }
}
exports.LaunchTemplatePlacementOutputReference = LaunchTemplatePlacementOutputReference;
_p = JSII_RTTI_SYMBOL_1;
LaunchTemplatePlacementOutputReference[_p] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplatePlacementOutputReference", version: "3.0.1" };
function launchTemplateTagSpecificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_type: cdktf.stringToTerraform(struct.resourceType),
        tags: cdktf.hashMapper(cdktf.anyToTerraform)(struct.tags),
    };
}
exports.launchTemplateTagSpecificationsToTerraform = launchTemplateTagSpecificationsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/launch_template.html aws_launch_template}.
 *
 * @stability stable
 */
class LaunchTemplate extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/launch_template.html aws_launch_template} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_launch_template',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // capacity_reservation_specification - computed: false, optional: true, required: false
        this._capacityReservationSpecification = new LaunchTemplateCapacityReservationSpecificationOutputReference(this, "capacity_reservation_specification", true);
        // cpu_options - computed: false, optional: true, required: false
        this._cpuOptions = new LaunchTemplateCpuOptionsOutputReference(this, "cpu_options", true);
        // credit_specification - computed: false, optional: true, required: false
        this._creditSpecification = new LaunchTemplateCreditSpecificationOutputReference(this, "credit_specification", true);
        // elastic_inference_accelerator - computed: false, optional: true, required: false
        this._elasticInferenceAccelerator = new LaunchTemplateElasticInferenceAcceleratorOutputReference(this, "elastic_inference_accelerator", true);
        // enclave_options - computed: false, optional: true, required: false
        this._enclaveOptions = new LaunchTemplateEnclaveOptionsOutputReference(this, "enclave_options", true);
        // hibernation_options - computed: false, optional: true, required: false
        this._hibernationOptions = new LaunchTemplateHibernationOptionsOutputReference(this, "hibernation_options", true);
        // iam_instance_profile - computed: false, optional: true, required: false
        this._iamInstanceProfile = new LaunchTemplateIamInstanceProfileOutputReference(this, "iam_instance_profile", true);
        // instance_market_options - computed: false, optional: true, required: false
        this._instanceMarketOptions = new LaunchTemplateInstanceMarketOptionsOutputReference(this, "instance_market_options", true);
        // metadata_options - computed: false, optional: true, required: false
        this._metadataOptions = new LaunchTemplateMetadataOptionsOutputReference(this, "metadata_options", true);
        // monitoring - computed: false, optional: true, required: false
        this._monitoring = new LaunchTemplateMonitoringOutputReference(this, "monitoring", true);
        // placement - computed: false, optional: true, required: false
        this._placement = new LaunchTemplatePlacementOutputReference(this, "placement", true);
        this._defaultVersion = config.defaultVersion;
        this._description = config.description;
        this._disableApiTermination = config.disableApiTermination;
        this._ebsOptimized = config.ebsOptimized;
        this._imageId = config.imageId;
        this._instanceInitiatedShutdownBehavior = config.instanceInitiatedShutdownBehavior;
        this._instanceType = config.instanceType;
        this._kernelId = config.kernelId;
        this._keyName = config.keyName;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._ramDiskId = config.ramDiskId;
        this._securityGroupNames = config.securityGroupNames;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._updateDefaultVersion = config.updateDefaultVersion;
        this._userData = config.userData;
        this._vpcSecurityGroupIds = config.vpcSecurityGroupIds;
        this._blockDeviceMappings = config.blockDeviceMappings;
        this._capacityReservationSpecification.internalValue = config.capacityReservationSpecification;
        this._cpuOptions.internalValue = config.cpuOptions;
        this._creditSpecification.internalValue = config.creditSpecification;
        this._elasticGpuSpecifications = config.elasticGpuSpecifications;
        this._elasticInferenceAccelerator.internalValue = config.elasticInferenceAccelerator;
        this._enclaveOptions.internalValue = config.enclaveOptions;
        this._hibernationOptions.internalValue = config.hibernationOptions;
        this._iamInstanceProfile.internalValue = config.iamInstanceProfile;
        this._instanceMarketOptions.internalValue = config.instanceMarketOptions;
        this._licenseSpecification = config.licenseSpecification;
        this._metadataOptions.internalValue = config.metadataOptions;
        this._monitoring.internalValue = config.monitoring;
        this._networkInterfaces = config.networkInterfaces;
        this._placement.internalValue = config.placement;
        this._tagSpecifications = config.tagSpecifications;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get defaultVersion() {
        return this.getNumberAttribute('default_version');
    }
    /**
     * @stability stable
     */
    set defaultVersion(value) {
        this._defaultVersion = value;
    }
    /**
     * @stability stable
     */
    resetDefaultVersion() {
        this._defaultVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultVersionInput() {
        return this._defaultVersion;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get disableApiTermination() {
        return this.getBooleanAttribute('disable_api_termination');
    }
    /**
     * @stability stable
     */
    set disableApiTermination(value) {
        this._disableApiTermination = value;
    }
    /**
     * @stability stable
     */
    resetDisableApiTermination() {
        this._disableApiTermination = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableApiTerminationInput() {
        return this._disableApiTermination;
    }
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getStringAttribute('ebs_optimized');
    }
    /**
     * @stability stable
     */
    set ebsOptimized(value) {
        this._ebsOptimized = value;
    }
    /**
     * @stability stable
     */
    resetEbsOptimized() {
        this._ebsOptimized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsOptimizedInput() {
        return this._ebsOptimized;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    /**
     * @stability stable
     */
    set imageId(value) {
        this._imageId = value;
    }
    /**
     * @stability stable
     */
    resetImageId() {
        this._imageId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get imageIdInput() {
        return this._imageId;
    }
    /**
     * @stability stable
     */
    get instanceInitiatedShutdownBehavior() {
        return this.getStringAttribute('instance_initiated_shutdown_behavior');
    }
    /**
     * @stability stable
     */
    set instanceInitiatedShutdownBehavior(value) {
        this._instanceInitiatedShutdownBehavior = value;
    }
    /**
     * @stability stable
     */
    resetInstanceInitiatedShutdownBehavior() {
        this._instanceInitiatedShutdownBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceInitiatedShutdownBehaviorInput() {
        return this._instanceInitiatedShutdownBehavior;
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get kernelId() {
        return this.getStringAttribute('kernel_id');
    }
    /**
     * @stability stable
     */
    set kernelId(value) {
        this._kernelId = value;
    }
    /**
     * @stability stable
     */
    resetKernelId() {
        this._kernelId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kernelIdInput() {
        return this._kernelId;
    }
    /**
     * @stability stable
     */
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    /**
     * @stability stable
     */
    set keyName(value) {
        this._keyName = value;
    }
    /**
     * @stability stable
     */
    resetKeyName() {
        this._keyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyNameInput() {
        return this._keyName;
    }
    // latest_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get latestVersion() {
        return this.getNumberAttribute('latest_version');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get ramDiskId() {
        return this.getStringAttribute('ram_disk_id');
    }
    /**
     * @stability stable
     */
    set ramDiskId(value) {
        this._ramDiskId = value;
    }
    /**
     * @stability stable
     */
    resetRamDiskId() {
        this._ramDiskId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ramDiskIdInput() {
        return this._ramDiskId;
    }
    /**
     * @stability stable
     */
    get securityGroupNames() {
        return this.getListAttribute('security_group_names');
    }
    /**
     * @stability stable
     */
    set securityGroupNames(value) {
        this._securityGroupNames = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupNames() {
        this._securityGroupNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupNamesInput() {
        return this._securityGroupNames;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get updateDefaultVersion() {
        return this.getBooleanAttribute('update_default_version');
    }
    /**
     * @stability stable
     */
    set updateDefaultVersion(value) {
        this._updateDefaultVersion = value;
    }
    /**
     * @stability stable
     */
    resetUpdateDefaultVersion() {
        this._updateDefaultVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateDefaultVersionInput() {
        return this._updateDefaultVersion;
    }
    /**
     * @stability stable
     */
    get userData() {
        return this.getStringAttribute('user_data');
    }
    /**
     * @stability stable
     */
    set userData(value) {
        this._userData = value;
    }
    /**
     * @stability stable
     */
    resetUserData() {
        this._userData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userDataInput() {
        return this._userData;
    }
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    /**
     * @stability stable
     */
    set vpcSecurityGroupIds(value) {
        this._vpcSecurityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetVpcSecurityGroupIds() {
        this._vpcSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcSecurityGroupIdsInput() {
        return this._vpcSecurityGroupIds;
    }
    /**
     * @stability stable
     */
    get blockDeviceMappings() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('block_device_mappings');
    }
    /**
     * @stability stable
     */
    set blockDeviceMappings(value) {
        this._blockDeviceMappings = value;
    }
    /**
     * @stability stable
     */
    resetBlockDeviceMappings() {
        this._blockDeviceMappings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockDeviceMappingsInput() {
        return this._blockDeviceMappings;
    }
    /**
     * @stability stable
     */
    get capacityReservationSpecification() {
        return this._capacityReservationSpecification;
    }
    /**
     * @stability stable
     */
    putCapacityReservationSpecification(value) {
        this._capacityReservationSpecification.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCapacityReservationSpecification() {
        this._capacityReservationSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityReservationSpecificationInput() {
        return this._capacityReservationSpecification.internalValue;
    }
    /**
     * @stability stable
     */
    get cpuOptions() {
        return this._cpuOptions;
    }
    /**
     * @stability stable
     */
    putCpuOptions(value) {
        this._cpuOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCpuOptions() {
        this._cpuOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cpuOptionsInput() {
        return this._cpuOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get creditSpecification() {
        return this._creditSpecification;
    }
    /**
     * @stability stable
     */
    putCreditSpecification(value) {
        this._creditSpecification.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCreditSpecification() {
        this._creditSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get creditSpecificationInput() {
        return this._creditSpecification.internalValue;
    }
    /**
     * @stability stable
     */
    get elasticGpuSpecifications() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('elastic_gpu_specifications');
    }
    /**
     * @stability stable
     */
    set elasticGpuSpecifications(value) {
        this._elasticGpuSpecifications = value;
    }
    /**
     * @stability stable
     */
    resetElasticGpuSpecifications() {
        this._elasticGpuSpecifications = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticGpuSpecificationsInput() {
        return this._elasticGpuSpecifications;
    }
    /**
     * @stability stable
     */
    get elasticInferenceAccelerator() {
        return this._elasticInferenceAccelerator;
    }
    /**
     * @stability stable
     */
    putElasticInferenceAccelerator(value) {
        this._elasticInferenceAccelerator.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetElasticInferenceAccelerator() {
        this._elasticInferenceAccelerator.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticInferenceAcceleratorInput() {
        return this._elasticInferenceAccelerator.internalValue;
    }
    /**
     * @stability stable
     */
    get enclaveOptions() {
        return this._enclaveOptions;
    }
    /**
     * @stability stable
     */
    putEnclaveOptions(value) {
        this._enclaveOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEnclaveOptions() {
        this._enclaveOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enclaveOptionsInput() {
        return this._enclaveOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get hibernationOptions() {
        return this._hibernationOptions;
    }
    /**
     * @stability stable
     */
    putHibernationOptions(value) {
        this._hibernationOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetHibernationOptions() {
        this._hibernationOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hibernationOptionsInput() {
        return this._hibernationOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get iamInstanceProfile() {
        return this._iamInstanceProfile;
    }
    /**
     * @stability stable
     */
    putIamInstanceProfile(value) {
        this._iamInstanceProfile.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIamInstanceProfile() {
        this._iamInstanceProfile.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamInstanceProfileInput() {
        return this._iamInstanceProfile.internalValue;
    }
    /**
     * @stability stable
     */
    get instanceMarketOptions() {
        return this._instanceMarketOptions;
    }
    /**
     * @stability stable
     */
    putInstanceMarketOptions(value) {
        this._instanceMarketOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetInstanceMarketOptions() {
        this._instanceMarketOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceMarketOptionsInput() {
        return this._instanceMarketOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get licenseSpecification() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('license_specification');
    }
    /**
     * @stability stable
     */
    set licenseSpecification(value) {
        this._licenseSpecification = value;
    }
    /**
     * @stability stable
     */
    resetLicenseSpecification() {
        this._licenseSpecification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get licenseSpecificationInput() {
        return this._licenseSpecification;
    }
    /**
     * @stability stable
     */
    get metadataOptions() {
        return this._metadataOptions;
    }
    /**
     * @stability stable
     */
    putMetadataOptions(value) {
        this._metadataOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMetadataOptions() {
        this._metadataOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metadataOptionsInput() {
        return this._metadataOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get monitoring() {
        return this._monitoring;
    }
    /**
     * @stability stable
     */
    putMonitoring(value) {
        this._monitoring.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMonitoring() {
        this._monitoring.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get monitoringInput() {
        return this._monitoring.internalValue;
    }
    /**
     * @stability stable
     */
    get networkInterfaces() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('network_interfaces');
    }
    /**
     * @stability stable
     */
    set networkInterfaces(value) {
        this._networkInterfaces = value;
    }
    /**
     * @stability stable
     */
    resetNetworkInterfaces() {
        this._networkInterfaces = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkInterfacesInput() {
        return this._networkInterfaces;
    }
    /**
     * @stability stable
     */
    get placement() {
        return this._placement;
    }
    /**
     * @stability stable
     */
    putPlacement(value) {
        this._placement.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPlacement() {
        this._placement.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get placementInput() {
        return this._placement.internalValue;
    }
    /**
     * @stability stable
     */
    get tagSpecifications() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tag_specifications');
    }
    /**
     * @stability stable
     */
    set tagSpecifications(value) {
        this._tagSpecifications = value;
    }
    /**
     * @stability stable
     */
    resetTagSpecifications() {
        this._tagSpecifications = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagSpecificationsInput() {
        return this._tagSpecifications;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            default_version: cdktf.numberToTerraform(this._defaultVersion),
            description: cdktf.stringToTerraform(this._description),
            disable_api_termination: cdktf.booleanToTerraform(this._disableApiTermination),
            ebs_optimized: cdktf.stringToTerraform(this._ebsOptimized),
            image_id: cdktf.stringToTerraform(this._imageId),
            instance_initiated_shutdown_behavior: cdktf.stringToTerraform(this._instanceInitiatedShutdownBehavior),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            kernel_id: cdktf.stringToTerraform(this._kernelId),
            key_name: cdktf.stringToTerraform(this._keyName),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            ram_disk_id: cdktf.stringToTerraform(this._ramDiskId),
            security_group_names: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroupNames),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            update_default_version: cdktf.booleanToTerraform(this._updateDefaultVersion),
            user_data: cdktf.stringToTerraform(this._userData),
            vpc_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._vpcSecurityGroupIds),
            block_device_mappings: cdktf.listMapper(launchTemplateBlockDeviceMappingsToTerraform)(this._blockDeviceMappings),
            capacity_reservation_specification: launchTemplateCapacityReservationSpecificationToTerraform(this._capacityReservationSpecification.internalValue),
            cpu_options: launchTemplateCpuOptionsToTerraform(this._cpuOptions.internalValue),
            credit_specification: launchTemplateCreditSpecificationToTerraform(this._creditSpecification.internalValue),
            elastic_gpu_specifications: cdktf.listMapper(launchTemplateElasticGpuSpecificationsToTerraform)(this._elasticGpuSpecifications),
            elastic_inference_accelerator: launchTemplateElasticInferenceAcceleratorToTerraform(this._elasticInferenceAccelerator.internalValue),
            enclave_options: launchTemplateEnclaveOptionsToTerraform(this._enclaveOptions.internalValue),
            hibernation_options: launchTemplateHibernationOptionsToTerraform(this._hibernationOptions.internalValue),
            iam_instance_profile: launchTemplateIamInstanceProfileToTerraform(this._iamInstanceProfile.internalValue),
            instance_market_options: launchTemplateInstanceMarketOptionsToTerraform(this._instanceMarketOptions.internalValue),
            license_specification: cdktf.listMapper(launchTemplateLicenseSpecificationToTerraform)(this._licenseSpecification),
            metadata_options: launchTemplateMetadataOptionsToTerraform(this._metadataOptions.internalValue),
            monitoring: launchTemplateMonitoringToTerraform(this._monitoring.internalValue),
            network_interfaces: cdktf.listMapper(launchTemplateNetworkInterfacesToTerraform)(this._networkInterfaces),
            placement: launchTemplatePlacementToTerraform(this._placement.internalValue),
            tag_specifications: cdktf.listMapper(launchTemplateTagSpecificationsToTerraform)(this._tagSpecifications),
        };
    }
}
exports.LaunchTemplate = LaunchTemplate;
_q = JSII_RTTI_SYMBOL_1;
LaunchTemplate[_q] = { fqn: "@cdktf/provider-aws.ec2.LaunchTemplate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LaunchTemplate.tfResourceType = "aws_launch_template";
