// @cdktf/provider-aws/lib/servicecatalog/data-aws-servicecatalog-launch-paths.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServicecatalogLaunchPaths = exports.DataAwsServicecatalogLaunchPathsSummaries = exports.DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries extends cdktf.ComplexComputedList {
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries = DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.DataAwsServicecatalogLaunchPathsSummariesConstraintSummaries", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsServicecatalogLaunchPathsSummaries extends cdktf.ComplexComputedList {
    // constraint_summaries - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get constraintSummaries() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('constraint_summaries');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // path_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get pathId() {
        return this.getStringAttribute('path_id');
    }
    // tags - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
}
exports.DataAwsServicecatalogLaunchPathsSummaries = DataAwsServicecatalogLaunchPathsSummaries;
_b = JSII_RTTI_SYMBOL_1;
DataAwsServicecatalogLaunchPathsSummaries[_b] = { fqn: "@cdktf/provider-aws.servicecatalog.DataAwsServicecatalogLaunchPathsSummaries", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/servicecatalog_launch_paths.html aws_servicecatalog_launch_paths}.
 *
 * @stability stable
 */
class DataAwsServicecatalogLaunchPaths extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/servicecatalog_launch_paths.html aws_servicecatalog_launch_paths} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_launch_paths',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._acceptLanguage = config.acceptLanguage;
        this._productId = config.productId;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get productId() {
        return this.getStringAttribute('product_id');
    }
    /**
     * @stability stable
     */
    set productId(value) {
        this._productId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productIdInput() {
        return this._productId;
    }
    // summaries - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    summaries(index) {
        return new DataAwsServicecatalogLaunchPathsSummaries(this, 'summaries', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            product_id: cdktf.stringToTerraform(this._productId),
        };
    }
}
exports.DataAwsServicecatalogLaunchPaths = DataAwsServicecatalogLaunchPaths;
_c = JSII_RTTI_SYMBOL_1;
DataAwsServicecatalogLaunchPaths[_c] = { fqn: "@cdktf/provider-aws.servicecatalog.DataAwsServicecatalogLaunchPaths", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsServicecatalogLaunchPaths.tfResourceType = "aws_servicecatalog_launch_paths";
