// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-product.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogProduct = exports.ServicecatalogProductProvisioningArtifactParametersOutputReference = exports.servicecatalogProductProvisioningArtifactParametersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function servicecatalogProductProvisioningArtifactParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        disable_template_validation: cdktf.booleanToTerraform(struct.disableTemplateValidation),
        name: cdktf.stringToTerraform(struct.name),
        template_physical_id: cdktf.stringToTerraform(struct.templatePhysicalId),
        template_url: cdktf.stringToTerraform(struct.templateUrl),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.servicecatalogProductProvisioningArtifactParametersToTerraform = servicecatalogProductProvisioningArtifactParametersToTerraform;
/**
 * @stability stable
 */
class ServicecatalogProductProvisioningArtifactParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._disableTemplateValidation) {
            hasAnyValues = true;
            internalValueResult.disableTemplateValidation = this._disableTemplateValidation;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._templatePhysicalId) {
            hasAnyValues = true;
            internalValueResult.templatePhysicalId = this._templatePhysicalId;
        }
        if (this._templateUrl) {
            hasAnyValues = true;
            internalValueResult.templateUrl = this._templateUrl;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._disableTemplateValidation = undefined;
            this._name = undefined;
            this._templatePhysicalId = undefined;
            this._templateUrl = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._disableTemplateValidation = value.disableTemplateValidation;
            this._name = value.name;
            this._templatePhysicalId = value.templatePhysicalId;
            this._templateUrl = value.templateUrl;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get disableTemplateValidation() {
        return this.getBooleanAttribute('disable_template_validation');
    }
    /**
     * @stability stable
     */
    set disableTemplateValidation(value) {
        this._disableTemplateValidation = value;
    }
    /**
     * @stability stable
     */
    resetDisableTemplateValidation() {
        this._disableTemplateValidation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableTemplateValidationInput() {
        return this._disableTemplateValidation;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get templatePhysicalId() {
        return this.getStringAttribute('template_physical_id');
    }
    /**
     * @stability stable
     */
    set templatePhysicalId(value) {
        this._templatePhysicalId = value;
    }
    /**
     * @stability stable
     */
    resetTemplatePhysicalId() {
        this._templatePhysicalId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templatePhysicalIdInput() {
        return this._templatePhysicalId;
    }
    /**
     * @stability stable
     */
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    /**
     * @stability stable
     */
    set templateUrl(value) {
        this._templateUrl = value;
    }
    /**
     * @stability stable
     */
    resetTemplateUrl() {
        this._templateUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateUrlInput() {
        return this._templateUrl;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.ServicecatalogProductProvisioningArtifactParametersOutputReference = ServicecatalogProductProvisioningArtifactParametersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogProductProvisioningArtifactParametersOutputReference[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProductProvisioningArtifactParametersOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_product.html aws_servicecatalog_product}.
 *
 * @stability stable
 */
class ServicecatalogProduct extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_product.html aws_servicecatalog_product} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_product',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // provisioning_artifact_parameters - computed: false, optional: false, required: true
        this._provisioningArtifactParameters = new ServicecatalogProductProvisioningArtifactParametersOutputReference(this, "provisioning_artifact_parameters", true);
        this._acceptLanguage = config.acceptLanguage;
        this._description = config.description;
        this._distributor = config.distributor;
        this._name = config.name;
        this._owner = config.owner;
        this._supportDescription = config.supportDescription;
        this._supportEmail = config.supportEmail;
        this._supportUrl = config.supportUrl;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._provisioningArtifactParameters.internalValue = config.provisioningArtifactParameters;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get distributor() {
        return this.getStringAttribute('distributor');
    }
    /**
     * @stability stable
     */
    set distributor(value) {
        this._distributor = value;
    }
    /**
     * @stability stable
     */
    resetDistributor() {
        this._distributor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get distributorInput() {
        return this._distributor;
    }
    // has_default_path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hasDefaultPath() {
        return this.getBooleanAttribute('has_default_path');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get owner() {
        return this.getStringAttribute('owner');
    }
    /**
     * @stability stable
     */
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerInput() {
        return this._owner;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get supportDescription() {
        return this.getStringAttribute('support_description');
    }
    /**
     * @stability stable
     */
    set supportDescription(value) {
        this._supportDescription = value;
    }
    /**
     * @stability stable
     */
    resetSupportDescription() {
        this._supportDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get supportDescriptionInput() {
        return this._supportDescription;
    }
    /**
     * @stability stable
     */
    get supportEmail() {
        return this.getStringAttribute('support_email');
    }
    /**
     * @stability stable
     */
    set supportEmail(value) {
        this._supportEmail = value;
    }
    /**
     * @stability stable
     */
    resetSupportEmail() {
        this._supportEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get supportEmailInput() {
        return this._supportEmail;
    }
    /**
     * @stability stable
     */
    get supportUrl() {
        return this.getStringAttribute('support_url');
    }
    /**
     * @stability stable
     */
    set supportUrl(value) {
        this._supportUrl = value;
    }
    /**
     * @stability stable
     */
    resetSupportUrl() {
        this._supportUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get supportUrlInput() {
        return this._supportUrl;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get provisioningArtifactParameters() {
        return this._provisioningArtifactParameters;
    }
    /**
     * @stability stable
     */
    putProvisioningArtifactParameters(value) {
        this._provisioningArtifactParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get provisioningArtifactParametersInput() {
        return this._provisioningArtifactParameters.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            description: cdktf.stringToTerraform(this._description),
            distributor: cdktf.stringToTerraform(this._distributor),
            name: cdktf.stringToTerraform(this._name),
            owner: cdktf.stringToTerraform(this._owner),
            support_description: cdktf.stringToTerraform(this._supportDescription),
            support_email: cdktf.stringToTerraform(this._supportEmail),
            support_url: cdktf.stringToTerraform(this._supportUrl),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            provisioning_artifact_parameters: servicecatalogProductProvisioningArtifactParametersToTerraform(this._provisioningArtifactParameters.internalValue),
        };
    }
}
exports.ServicecatalogProduct = ServicecatalogProduct;
_b = JSII_RTTI_SYMBOL_1;
ServicecatalogProduct[_b] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProduct", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogProduct.tfResourceType = "aws_servicecatalog_product";
