// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-budget-resource-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogBudgetResourceAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_budget_resource_association.html aws_servicecatalog_budget_resource_association}.
 *
 * @stability stable
 */
class ServicecatalogBudgetResourceAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_budget_resource_association.html aws_servicecatalog_budget_resource_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_budget_resource_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._budgetName = config.budgetName;
        this._resourceId = config.resourceId;
    }
    /**
     * @stability stable
     */
    get budgetName() {
        return this.getStringAttribute('budget_name');
    }
    /**
     * @stability stable
     */
    set budgetName(value) {
        this._budgetName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get budgetNameInput() {
        return this._budgetName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    /**
     * @stability stable
     */
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceIdInput() {
        return this._resourceId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            budget_name: cdktf.stringToTerraform(this._budgetName),
            resource_id: cdktf.stringToTerraform(this._resourceId),
        };
    }
}
exports.ServicecatalogBudgetResourceAssociation = ServicecatalogBudgetResourceAssociation;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogBudgetResourceAssociation[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogBudgetResourceAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogBudgetResourceAssociation.tfResourceType = "aws_servicecatalog_budget_resource_association";
