// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-provisioning-artifact.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogProvisioningArtifact = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_provisioning_artifact.html aws_servicecatalog_provisioning_artifact}.
 *
 * @stability stable
 */
class ServicecatalogProvisioningArtifact extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_provisioning_artifact.html aws_servicecatalog_provisioning_artifact} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_provisioning_artifact',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._acceptLanguage = config.acceptLanguage;
        this._active = config.active;
        this._description = config.description;
        this._disableTemplateValidation = config.disableTemplateValidation;
        this._guidance = config.guidance;
        this._name = config.name;
        this._productId = config.productId;
        this._templatePhysicalId = config.templatePhysicalId;
        this._templateUrl = config.templateUrl;
        this._type = config.type;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    /**
     * @stability stable
     */
    get active() {
        return this.getBooleanAttribute('active');
    }
    /**
     * @stability stable
     */
    set active(value) {
        this._active = value;
    }
    /**
     * @stability stable
     */
    resetActive() {
        this._active = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get activeInput() {
        return this._active;
    }
    // created_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get disableTemplateValidation() {
        return this.getBooleanAttribute('disable_template_validation');
    }
    /**
     * @stability stable
     */
    set disableTemplateValidation(value) {
        this._disableTemplateValidation = value;
    }
    /**
     * @stability stable
     */
    resetDisableTemplateValidation() {
        this._disableTemplateValidation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableTemplateValidationInput() {
        return this._disableTemplateValidation;
    }
    /**
     * @stability stable
     */
    get guidance() {
        return this.getStringAttribute('guidance');
    }
    /**
     * @stability stable
     */
    set guidance(value) {
        this._guidance = value;
    }
    /**
     * @stability stable
     */
    resetGuidance() {
        this._guidance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get guidanceInput() {
        return this._guidance;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get productId() {
        return this.getStringAttribute('product_id');
    }
    /**
     * @stability stable
     */
    set productId(value) {
        this._productId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productIdInput() {
        return this._productId;
    }
    /**
     * @stability stable
     */
    get templatePhysicalId() {
        return this.getStringAttribute('template_physical_id');
    }
    /**
     * @stability stable
     */
    set templatePhysicalId(value) {
        this._templatePhysicalId = value;
    }
    /**
     * @stability stable
     */
    resetTemplatePhysicalId() {
        this._templatePhysicalId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templatePhysicalIdInput() {
        return this._templatePhysicalId;
    }
    /**
     * @stability stable
     */
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    /**
     * @stability stable
     */
    set templateUrl(value) {
        this._templateUrl = value;
    }
    /**
     * @stability stable
     */
    resetTemplateUrl() {
        this._templateUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateUrlInput() {
        return this._templateUrl;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            active: cdktf.booleanToTerraform(this._active),
            description: cdktf.stringToTerraform(this._description),
            disable_template_validation: cdktf.booleanToTerraform(this._disableTemplateValidation),
            guidance: cdktf.stringToTerraform(this._guidance),
            name: cdktf.stringToTerraform(this._name),
            product_id: cdktf.stringToTerraform(this._productId),
            template_physical_id: cdktf.stringToTerraform(this._templatePhysicalId),
            template_url: cdktf.stringToTerraform(this._templateUrl),
            type: cdktf.stringToTerraform(this._type),
        };
    }
}
exports.ServicecatalogProvisioningArtifact = ServicecatalogProvisioningArtifact;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogProvisioningArtifact[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProvisioningArtifact", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogProvisioningArtifact.tfResourceType = "aws_servicecatalog_provisioning_artifact";
