// @cdktf/provider-aws/lib/imagebuilder/imagebuilder-distribution-configuration.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImagebuilderDistributionConfiguration = exports.imagebuilderDistributionConfigurationDistributionToTerraform = exports.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference = exports.imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationToTerraform = exports.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference = exports.imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        user_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.userGroups),
        user_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.userIds),
    };
}
exports.imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionToTerraform = imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionToTerraform;
/**
 * @stability stable
 */
class ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._userGroups) {
            hasAnyValues = true;
            internalValueResult.userGroups = this._userGroups;
        }
        if (this._userIds) {
            hasAnyValues = true;
            internalValueResult.userIds = this._userIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._userGroups = undefined;
            this._userIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._userGroups = value.userGroups;
            this._userIds = value.userIds;
        }
    }
    /**
     * @stability stable
     */
    get userGroups() {
        return this.getListAttribute('user_groups');
    }
    /**
     * @stability stable
     */
    set userGroups(value) {
        this._userGroups = value;
    }
    /**
     * @stability stable
     */
    resetUserGroups() {
        this._userGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userGroupsInput() {
        return this._userGroups;
    }
    /**
     * @stability stable
     */
    get userIds() {
        return this.getListAttribute('user_ids');
    }
    /**
     * @stability stable
     */
    set userIds(value) {
        this._userIds = value;
    }
    /**
     * @stability stable
     */
    resetUserIds() {
        this._userIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userIdsInput() {
        return this._userIds;
    }
}
exports.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference = ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference[_a] = { fqn: "@cdktf/provider-aws.imagebuilder.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference", version: "3.0.1" };
function imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ami_tags: cdktf.hashMapper(cdktf.anyToTerraform)(struct.amiTags),
        description: cdktf.stringToTerraform(struct.description),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        name: cdktf.stringToTerraform(struct.name),
        target_account_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.targetAccountIds),
        launch_permission: imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionToTerraform(struct.launchPermission),
    };
}
exports.imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationToTerraform = imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationToTerraform;
/**
 * @stability stable
 */
class ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // launch_permission - computed: false, optional: true, required: false
        this._launchPermission = new ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationLaunchPermissionOutputReference(this, "launch_permission", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _d, _e;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._amiTags) {
            hasAnyValues = true;
            internalValueResult.amiTags = this._amiTags;
        }
        if (this._description) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._kmsKeyId) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._targetAccountIds) {
            hasAnyValues = true;
            internalValueResult.targetAccountIds = this._targetAccountIds;
        }
        if ((_d = this._launchPermission) === null || _d === void 0 ? void 0 : _d.internalValue) {
            hasAnyValues = true;
            internalValueResult.launchPermission = (_e = this._launchPermission) === null || _e === void 0 ? void 0 : _e.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._amiTags = undefined;
            this._description = undefined;
            this._kmsKeyId = undefined;
            this._name = undefined;
            this._targetAccountIds = undefined;
            this._launchPermission.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._amiTags = value.amiTags;
            this._description = value.description;
            this._kmsKeyId = value.kmsKeyId;
            this._name = value.name;
            this._targetAccountIds = value.targetAccountIds;
            this._launchPermission.internalValue = value.launchPermission;
        }
    }
    /**
     * @stability stable
     */
    get amiTags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ami_tags');
    }
    /**
     * @stability stable
     */
    set amiTags(value) {
        this._amiTags = value;
    }
    /**
     * @stability stable
     */
    resetAmiTags() {
        this._amiTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get amiTagsInput() {
        return this._amiTags;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get targetAccountIds() {
        return this.getListAttribute('target_account_ids');
    }
    /**
     * @stability stable
     */
    set targetAccountIds(value) {
        this._targetAccountIds = value;
    }
    /**
     * @stability stable
     */
    resetTargetAccountIds() {
        this._targetAccountIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetAccountIdsInput() {
        return this._targetAccountIds;
    }
    /**
     * @stability stable
     */
    get launchPermission() {
        return this._launchPermission;
    }
    /**
     * @stability stable
     */
    putLaunchPermission(value) {
        this._launchPermission.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLaunchPermission() {
        this._launchPermission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchPermissionInput() {
        return this._launchPermission.internalValue;
    }
}
exports.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference = ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.imagebuilder.ImagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationOutputReference", version: "3.0.1" };
function imagebuilderDistributionConfigurationDistributionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        license_configuration_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.licenseConfigurationArns),
        region: cdktf.stringToTerraform(struct.region),
        ami_distribution_configuration: imagebuilderDistributionConfigurationDistributionAmiDistributionConfigurationToTerraform(struct.amiDistributionConfiguration),
    };
}
exports.imagebuilderDistributionConfigurationDistributionToTerraform = imagebuilderDistributionConfigurationDistributionToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/imagebuilder_distribution_configuration.html aws_imagebuilder_distribution_configuration}.
 *
 * @stability stable
 */
class ImagebuilderDistributionConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/imagebuilder_distribution_configuration.html aws_imagebuilder_distribution_configuration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_distribution_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._description = config.description;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._distribution = config.distribution;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // date_created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // date_updated - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateUpdated() {
        return this.getStringAttribute('date_updated');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get distribution() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('distribution');
    }
    /**
     * @stability stable
     */
    set distribution(value) {
        this._distribution = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get distributionInput() {
        return this._distribution;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            distribution: cdktf.listMapper(imagebuilderDistributionConfigurationDistributionToTerraform)(this._distribution),
        };
    }
}
exports.ImagebuilderDistributionConfiguration = ImagebuilderDistributionConfiguration;
_c = JSII_RTTI_SYMBOL_1;
ImagebuilderDistributionConfiguration[_c] = { fqn: "@cdktf/provider-aws.imagebuilder.ImagebuilderDistributionConfiguration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ImagebuilderDistributionConfiguration.tfResourceType = "aws_imagebuilder_distribution_configuration";
