// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-principal-portfolio-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogPrincipalPortfolioAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_principal_portfolio_association.html aws_servicecatalog_principal_portfolio_association}.
 *
 * @stability stable
 */
class ServicecatalogPrincipalPortfolioAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_principal_portfolio_association.html aws_servicecatalog_principal_portfolio_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_principal_portfolio_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._acceptLanguage = config.acceptLanguage;
        this._portfolioId = config.portfolioId;
        this._principalArn = config.principalArn;
        this._principalType = config.principalType;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get portfolioId() {
        return this.getStringAttribute('portfolio_id');
    }
    /**
     * @stability stable
     */
    set portfolioId(value) {
        this._portfolioId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portfolioIdInput() {
        return this._portfolioId;
    }
    /**
     * @stability stable
     */
    get principalArn() {
        return this.getStringAttribute('principal_arn');
    }
    /**
     * @stability stable
     */
    set principalArn(value) {
        this._principalArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalArnInput() {
        return this._principalArn;
    }
    /**
     * @stability stable
     */
    get principalType() {
        return this.getStringAttribute('principal_type');
    }
    /**
     * @stability stable
     */
    set principalType(value) {
        this._principalType = value;
    }
    /**
     * @stability stable
     */
    resetPrincipalType() {
        this._principalType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalTypeInput() {
        return this._principalType;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            portfolio_id: cdktf.stringToTerraform(this._portfolioId),
            principal_arn: cdktf.stringToTerraform(this._principalArn),
            principal_type: cdktf.stringToTerraform(this._principalType),
        };
    }
}
exports.ServicecatalogPrincipalPortfolioAssociation = ServicecatalogPrincipalPortfolioAssociation;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogPrincipalPortfolioAssociation[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogPrincipalPortfolioAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogPrincipalPortfolioAssociation.tfResourceType = "aws_servicecatalog_principal_portfolio_association";
