// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-product-portfolio-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogProductPortfolioAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_product_portfolio_association.html aws_servicecatalog_product_portfolio_association}.
 *
 * @stability stable
 */
class ServicecatalogProductPortfolioAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_product_portfolio_association.html aws_servicecatalog_product_portfolio_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_product_portfolio_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._acceptLanguage = config.acceptLanguage;
        this._portfolioId = config.portfolioId;
        this._productId = config.productId;
        this._sourcePortfolioId = config.sourcePortfolioId;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get portfolioId() {
        return this.getStringAttribute('portfolio_id');
    }
    /**
     * @stability stable
     */
    set portfolioId(value) {
        this._portfolioId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portfolioIdInput() {
        return this._portfolioId;
    }
    /**
     * @stability stable
     */
    get productId() {
        return this.getStringAttribute('product_id');
    }
    /**
     * @stability stable
     */
    set productId(value) {
        this._productId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productIdInput() {
        return this._productId;
    }
    /**
     * @stability stable
     */
    get sourcePortfolioId() {
        return this.getStringAttribute('source_portfolio_id');
    }
    /**
     * @stability stable
     */
    set sourcePortfolioId(value) {
        this._sourcePortfolioId = value;
    }
    /**
     * @stability stable
     */
    resetSourcePortfolioId() {
        this._sourcePortfolioId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourcePortfolioIdInput() {
        return this._sourcePortfolioId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            portfolio_id: cdktf.stringToTerraform(this._portfolioId),
            product_id: cdktf.stringToTerraform(this._productId),
            source_portfolio_id: cdktf.stringToTerraform(this._sourcePortfolioId),
        };
    }
}
exports.ServicecatalogProductPortfolioAssociation = ServicecatalogProductPortfolioAssociation;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogProductPortfolioAssociation[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProductPortfolioAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogProductPortfolioAssociation.tfResourceType = "aws_servicecatalog_product_portfolio_association";
