// @cdktf/provider-aws/lib/rds/data-aws-db-proxy.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDbProxy = exports.DataAwsDbProxyAuth = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsDbProxyAuth extends cdktf.ComplexComputedList {
    // auth_scheme - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authScheme() {
        return this.getStringAttribute('auth_scheme');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // iam_auth - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iamAuth() {
        return this.getStringAttribute('iam_auth');
    }
    // secret_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
}
exports.DataAwsDbProxyAuth = DataAwsDbProxyAuth;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDbProxyAuth[_a] = { fqn: "@cdktf/provider-aws.rds.DataAwsDbProxyAuth", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/db_proxy.html aws_db_proxy}.
 *
 * @stability stable
 */
class DataAwsDbProxy extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/db_proxy.html aws_db_proxy} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_proxy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // auth - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    auth(index) {
        return new DataAwsDbProxyAuth(this, 'auth', index);
    }
    // debug_logging - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get debugLogging() {
        return this.getBooleanAttribute('debug_logging');
    }
    // endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // engine_family - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engineFamily() {
        return this.getStringAttribute('engine_family');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // idle_client_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get idleClientTimeout() {
        return this.getNumberAttribute('idle_client_timeout');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // require_tls - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get requireTls() {
        return this.getBooleanAttribute('require_tls');
    }
    // role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    // vpc_subnet_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcSubnetIds() {
        return this.getListAttribute('vpc_subnet_ids');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsDbProxy = DataAwsDbProxy;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDbProxy[_b] = { fqn: "@cdktf/provider-aws.rds.DataAwsDbProxy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsDbProxy.tfResourceType = "aws_db_proxy";
