// @cdktf/provider-aws/lib/rds/db-proxy-default-target-group.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbProxyDefaultTargetGroup = exports.DbProxyDefaultTargetGroupTimeoutsOutputReference = exports.dbProxyDefaultTargetGroupTimeoutsToTerraform = exports.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference = exports.dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_borrow_timeout: cdktf.numberToTerraform(struct.connectionBorrowTimeout),
        init_query: cdktf.stringToTerraform(struct.initQuery),
        max_connections_percent: cdktf.numberToTerraform(struct.maxConnectionsPercent),
        max_idle_connections_percent: cdktf.numberToTerraform(struct.maxIdleConnectionsPercent),
        session_pinning_filters: cdktf.listMapper(cdktf.stringToTerraform)(struct.sessionPinningFilters),
    };
}
exports.dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform = dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform;
/**
 * @stability stable
 */
class DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionBorrowTimeout) {
            hasAnyValues = true;
            internalValueResult.connectionBorrowTimeout = this._connectionBorrowTimeout;
        }
        if (this._initQuery) {
            hasAnyValues = true;
            internalValueResult.initQuery = this._initQuery;
        }
        if (this._maxConnectionsPercent) {
            hasAnyValues = true;
            internalValueResult.maxConnectionsPercent = this._maxConnectionsPercent;
        }
        if (this._maxIdleConnectionsPercent) {
            hasAnyValues = true;
            internalValueResult.maxIdleConnectionsPercent = this._maxIdleConnectionsPercent;
        }
        if (this._sessionPinningFilters) {
            hasAnyValues = true;
            internalValueResult.sessionPinningFilters = this._sessionPinningFilters;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connectionBorrowTimeout = undefined;
            this._initQuery = undefined;
            this._maxConnectionsPercent = undefined;
            this._maxIdleConnectionsPercent = undefined;
            this._sessionPinningFilters = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connectionBorrowTimeout = value.connectionBorrowTimeout;
            this._initQuery = value.initQuery;
            this._maxConnectionsPercent = value.maxConnectionsPercent;
            this._maxIdleConnectionsPercent = value.maxIdleConnectionsPercent;
            this._sessionPinningFilters = value.sessionPinningFilters;
        }
    }
    /**
     * @stability stable
     */
    get connectionBorrowTimeout() {
        return this.getNumberAttribute('connection_borrow_timeout');
    }
    /**
     * @stability stable
     */
    set connectionBorrowTimeout(value) {
        this._connectionBorrowTimeout = value;
    }
    /**
     * @stability stable
     */
    resetConnectionBorrowTimeout() {
        this._connectionBorrowTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionBorrowTimeoutInput() {
        return this._connectionBorrowTimeout;
    }
    /**
     * @stability stable
     */
    get initQuery() {
        return this.getStringAttribute('init_query');
    }
    /**
     * @stability stable
     */
    set initQuery(value) {
        this._initQuery = value;
    }
    /**
     * @stability stable
     */
    resetInitQuery() {
        this._initQuery = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get initQueryInput() {
        return this._initQuery;
    }
    /**
     * @stability stable
     */
    get maxConnectionsPercent() {
        return this.getNumberAttribute('max_connections_percent');
    }
    /**
     * @stability stable
     */
    set maxConnectionsPercent(value) {
        this._maxConnectionsPercent = value;
    }
    /**
     * @stability stable
     */
    resetMaxConnectionsPercent() {
        this._maxConnectionsPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConnectionsPercentInput() {
        return this._maxConnectionsPercent;
    }
    /**
     * @stability stable
     */
    get maxIdleConnectionsPercent() {
        return this.getNumberAttribute('max_idle_connections_percent');
    }
    /**
     * @stability stable
     */
    set maxIdleConnectionsPercent(value) {
        this._maxIdleConnectionsPercent = value;
    }
    /**
     * @stability stable
     */
    resetMaxIdleConnectionsPercent() {
        this._maxIdleConnectionsPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxIdleConnectionsPercentInput() {
        return this._maxIdleConnectionsPercent;
    }
    /**
     * @stability stable
     */
    get sessionPinningFilters() {
        return this.getListAttribute('session_pinning_filters');
    }
    /**
     * @stability stable
     */
    set sessionPinningFilters(value) {
        this._sessionPinningFilters = value;
    }
    /**
     * @stability stable
     */
    resetSessionPinningFilters() {
        this._sessionPinningFilters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sessionPinningFiltersInput() {
        return this._sessionPinningFilters;
    }
}
exports.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference = DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.rds.DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference", version: "3.0.1" };
function dbProxyDefaultTargetGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dbProxyDefaultTargetGroupTimeoutsToTerraform = dbProxyDefaultTargetGroupTimeoutsToTerraform;
/**
 * @stability stable
 */
class DbProxyDefaultTargetGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.DbProxyDefaultTargetGroupTimeoutsOutputReference = DbProxyDefaultTargetGroupTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroupTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.rds.DbProxyDefaultTargetGroupTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/db_proxy_default_target_group.html aws_db_proxy_default_target_group}.
 *
 * @stability stable
 */
class DbProxyDefaultTargetGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/db_proxy_default_target_group.html aws_db_proxy_default_target_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_proxy_default_target_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // connection_pool_config - computed: false, optional: true, required: false
        this._connectionPoolConfig = new DbProxyDefaultTargetGroupConnectionPoolConfigOutputReference(this, "connection_pool_config", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbProxyDefaultTargetGroupTimeoutsOutputReference(this, "timeouts", true);
        this._dbProxyName = config.dbProxyName;
        this._connectionPoolConfig.internalValue = config.connectionPoolConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get dbProxyName() {
        return this.getStringAttribute('db_proxy_name');
    }
    /**
     * @stability stable
     */
    set dbProxyName(value) {
        this._dbProxyName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dbProxyNameInput() {
        return this._dbProxyName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    get connectionPoolConfig() {
        return this._connectionPoolConfig;
    }
    /**
     * @stability stable
     */
    putConnectionPoolConfig(value) {
        this._connectionPoolConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConnectionPoolConfig() {
        this._connectionPoolConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionPoolConfigInput() {
        return this._connectionPoolConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            db_proxy_name: cdktf.stringToTerraform(this._dbProxyName),
            connection_pool_config: dbProxyDefaultTargetGroupConnectionPoolConfigToTerraform(this._connectionPoolConfig.internalValue),
            timeouts: dbProxyDefaultTargetGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DbProxyDefaultTargetGroup = DbProxyDefaultTargetGroup;
_c = JSII_RTTI_SYMBOL_1;
DbProxyDefaultTargetGroup[_c] = { fqn: "@cdktf/provider-aws.rds.DbProxyDefaultTargetGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DbProxyDefaultTargetGroup.tfResourceType = "aws_db_proxy_default_target_group";
