// @cdktf/provider-aws/lib/waf/waf-regex-match-set.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafRegexMatchSet = exports.wafRegexMatchSetRegexMatchTupleToTerraform = exports.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference = exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data: cdktf.stringToTerraform(struct.data),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform = wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform;
/**
 * @stability stable
 */
class WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._data) {
            hasAnyValues = true;
            internalValueResult.data = this._data;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._data = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._data = value.data;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get data() {
        return this.getStringAttribute('data');
    }
    /**
     * @stability stable
     */
    set data(value) {
        this._data = value;
    }
    /**
     * @stability stable
     */
    resetData() {
        this._data = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataInput() {
        return this._data;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference = WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference[_a] = { fqn: "@cdktf/provider-aws.waf.WafRegexMatchSetRegexMatchTupleFieldToMatchOutputReference", version: "3.0.1" };
function wafRegexMatchSetRegexMatchTupleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        regex_pattern_set_id: cdktf.stringToTerraform(struct.regexPatternSetId),
        text_transformation: cdktf.stringToTerraform(struct.textTransformation),
        field_to_match: wafRegexMatchSetRegexMatchTupleFieldToMatchToTerraform(struct.fieldToMatch),
    };
}
exports.wafRegexMatchSetRegexMatchTupleToTerraform = wafRegexMatchSetRegexMatchTupleToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/waf_regex_match_set.html aws_waf_regex_match_set}.
 *
 * @stability stable
 */
class WafRegexMatchSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/waf_regex_match_set.html aws_waf_regex_match_set} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_waf_regex_match_set',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._regexMatchTuple = config.regexMatchTuple;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get regexMatchTuple() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('regex_match_tuple');
    }
    /**
     * @stability stable
     */
    set regexMatchTuple(value) {
        this._regexMatchTuple = value;
    }
    /**
     * @stability stable
     */
    resetRegexMatchTuple() {
        this._regexMatchTuple = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regexMatchTupleInput() {
        return this._regexMatchTuple;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            regex_match_tuple: cdktf.listMapper(wafRegexMatchSetRegexMatchTupleToTerraform)(this._regexMatchTuple),
        };
    }
}
exports.WafRegexMatchSet = WafRegexMatchSet;
_b = JSII_RTTI_SYMBOL_1;
WafRegexMatchSet[_b] = { fqn: "@cdktf/provider-aws.waf.WafRegexMatchSet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WafRegexMatchSet.tfResourceType = "aws_waf_regex_match_set";
