// cdktf-cli/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logFileName = exports.processLoggerError = exports.processLoggerDebug = exports.getLogger = exports.logger = void 0;
const log4js_1 = require("log4js");
Object.defineProperty(exports, "getLogger", { enumerable: true, get: function () { return log4js_1.getLogger; } });
const logger = log4js_1.getLogger();
exports.logger = logger;
logger.level = process.env.CDKTF_LOG_LEVEL || "INFO";
const logFileName = "cdktf.log";
exports.logFileName = logFileName;
if (process.env.CDKTF_DISABLE_LOGGING === "false") {
    log4js_1.configure({
        appenders: { cdktf: { type: "file", filename: "./" + logFileName } },
        categories: { default: { appenders: ["cdktf"], level: "debug" } },
    });
}
const processLoggerDebug = (chunk) => {
    logger.debug(chunk.toString());
};
exports.processLoggerDebug = processLoggerDebug;
const processLoggerError = (chunk) => {
    logger.error(chunk.toString());
};
exports.processLoggerError = processLoggerError;
