// @cdktf/provider-aws/lib/ssoadmin/ssoadmin-account-assignment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminAccountAssignment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_account_assignment.html aws_ssoadmin_account_assignment}.
 *
 * @stability stable
 */
class SsoadminAccountAssignment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_account_assignment.html aws_ssoadmin_account_assignment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_account_assignment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._instanceArn = config.instanceArn;
        this._permissionSetArn = config.permissionSetArn;
        this._principalId = config.principalId;
        this._principalType = config.principalType;
        this._targetId = config.targetId;
        this._targetType = config.targetType;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    /**
     * @stability stable
     */
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceArnInput() {
        return this._instanceArn;
    }
    /**
     * @stability stable
     */
    get permissionSetArn() {
        return this.getStringAttribute('permission_set_arn');
    }
    /**
     * @stability stable
     */
    set permissionSetArn(value) {
        this._permissionSetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionSetArnInput() {
        return this._permissionSetArn;
    }
    /**
     * @stability stable
     */
    get principalId() {
        return this.getStringAttribute('principal_id');
    }
    /**
     * @stability stable
     */
    set principalId(value) {
        this._principalId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalIdInput() {
        return this._principalId;
    }
    /**
     * @stability stable
     */
    get principalType() {
        return this.getStringAttribute('principal_type');
    }
    /**
     * @stability stable
     */
    set principalType(value) {
        this._principalType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalTypeInput() {
        return this._principalType;
    }
    /**
     * @stability stable
     */
    get targetId() {
        return this.getStringAttribute('target_id');
    }
    /**
     * @stability stable
     */
    set targetId(value) {
        this._targetId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetIdInput() {
        return this._targetId;
    }
    /**
     * @stability stable
     */
    get targetType() {
        return this.getStringAttribute('target_type');
    }
    /**
     * @stability stable
     */
    set targetType(value) {
        this._targetType = value;
    }
    /**
     * @stability stable
     */
    resetTargetType() {
        this._targetType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetTypeInput() {
        return this._targetType;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            permission_set_arn: cdktf.stringToTerraform(this._permissionSetArn),
            principal_id: cdktf.stringToTerraform(this._principalId),
            principal_type: cdktf.stringToTerraform(this._principalType),
            target_id: cdktf.stringToTerraform(this._targetId),
            target_type: cdktf.stringToTerraform(this._targetType),
        };
    }
}
exports.SsoadminAccountAssignment = SsoadminAccountAssignment;
_a = JSII_RTTI_SYMBOL_1;
SsoadminAccountAssignment[_a] = { fqn: "@cdktf/provider-aws.ssoadmin.SsoadminAccountAssignment", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsoadminAccountAssignment.tfResourceType = "aws_ssoadmin_account_assignment";
