// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-route.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Route = exports.apigatewayv2RouteRequestParameterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2RouteRequestParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        request_parameter_key: cdktf.stringToTerraform(struct.requestParameterKey),
        required: cdktf.booleanToTerraform(struct.required),
    };
}
exports.apigatewayv2RouteRequestParameterToTerraform = apigatewayv2RouteRequestParameterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route.html aws_apigatewayv2_route}.
 *
 * @stability stable
 */
class Apigatewayv2Route extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route.html aws_apigatewayv2_route} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_route',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._apiId = config.apiId;
        this._apiKeyRequired = config.apiKeyRequired;
        this._authorizationScopes = config.authorizationScopes;
        this._authorizationType = config.authorizationType;
        this._authorizerId = config.authorizerId;
        this._modelSelectionExpression = config.modelSelectionExpression;
        this._operationName = config.operationName;
        this._requestModels = config.requestModels;
        this._routeKey = config.routeKey;
        this._routeResponseSelectionExpression = config.routeResponseSelectionExpression;
        this._target = config.target;
        this._requestParameter = config.requestParameter;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    /**
     * @stability stable
     */
    get apiKeyRequired() {
        return this.getBooleanAttribute('api_key_required');
    }
    /**
     * @stability stable
     */
    set apiKeyRequired(value) {
        this._apiKeyRequired = value;
    }
    /**
     * @stability stable
     */
    resetApiKeyRequired() {
        this._apiKeyRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiKeyRequiredInput() {
        return this._apiKeyRequired;
    }
    /**
     * @stability stable
     */
    get authorizationScopes() {
        return this.getListAttribute('authorization_scopes');
    }
    /**
     * @stability stable
     */
    set authorizationScopes(value) {
        this._authorizationScopes = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizationScopes() {
        this._authorizationScopes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizationScopesInput() {
        return this._authorizationScopes;
    }
    /**
     * @stability stable
     */
    get authorizationType() {
        return this.getStringAttribute('authorization_type');
    }
    /**
     * @stability stable
     */
    set authorizationType(value) {
        this._authorizationType = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizationType() {
        this._authorizationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizationTypeInput() {
        return this._authorizationType;
    }
    /**
     * @stability stable
     */
    get authorizerId() {
        return this.getStringAttribute('authorizer_id');
    }
    /**
     * @stability stable
     */
    set authorizerId(value) {
        this._authorizerId = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizerId() {
        this._authorizerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerIdInput() {
        return this._authorizerId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get modelSelectionExpression() {
        return this.getStringAttribute('model_selection_expression');
    }
    /**
     * @stability stable
     */
    set modelSelectionExpression(value) {
        this._modelSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetModelSelectionExpression() {
        this._modelSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modelSelectionExpressionInput() {
        return this._modelSelectionExpression;
    }
    /**
     * @stability stable
     */
    get operationName() {
        return this.getStringAttribute('operation_name');
    }
    /**
     * @stability stable
     */
    set operationName(value) {
        this._operationName = value;
    }
    /**
     * @stability stable
     */
    resetOperationName() {
        this._operationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get operationNameInput() {
        return this._operationName;
    }
    /**
     * @stability stable
     */
    get requestModels() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_models');
    }
    /**
     * @stability stable
     */
    set requestModels(value) {
        this._requestModels = value;
    }
    /**
     * @stability stable
     */
    resetRequestModels() {
        this._requestModels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestModelsInput() {
        return this._requestModels;
    }
    /**
     * @stability stable
     */
    get routeKey() {
        return this.getStringAttribute('route_key');
    }
    /**
     * @stability stable
     */
    set routeKey(value) {
        this._routeKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeKeyInput() {
        return this._routeKey;
    }
    /**
     * @stability stable
     */
    get routeResponseSelectionExpression() {
        return this.getStringAttribute('route_response_selection_expression');
    }
    /**
     * @stability stable
     */
    set routeResponseSelectionExpression(value) {
        this._routeResponseSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetRouteResponseSelectionExpression() {
        this._routeResponseSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeResponseSelectionExpressionInput() {
        return this._routeResponseSelectionExpression;
    }
    /**
     * @stability stable
     */
    get target() {
        return this.getStringAttribute('target');
    }
    /**
     * @stability stable
     */
    set target(value) {
        this._target = value;
    }
    /**
     * @stability stable
     */
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetInput() {
        return this._target;
    }
    /**
     * @stability stable
     */
    get requestParameter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_parameter');
    }
    /**
     * @stability stable
     */
    set requestParameter(value) {
        this._requestParameter = value;
    }
    /**
     * @stability stable
     */
    resetRequestParameter() {
        this._requestParameter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestParameterInput() {
        return this._requestParameter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            api_key_required: cdktf.booleanToTerraform(this._apiKeyRequired),
            authorization_scopes: cdktf.listMapper(cdktf.stringToTerraform)(this._authorizationScopes),
            authorization_type: cdktf.stringToTerraform(this._authorizationType),
            authorizer_id: cdktf.stringToTerraform(this._authorizerId),
            model_selection_expression: cdktf.stringToTerraform(this._modelSelectionExpression),
            operation_name: cdktf.stringToTerraform(this._operationName),
            request_models: cdktf.hashMapper(cdktf.anyToTerraform)(this._requestModels),
            route_key: cdktf.stringToTerraform(this._routeKey),
            route_response_selection_expression: cdktf.stringToTerraform(this._routeResponseSelectionExpression),
            target: cdktf.stringToTerraform(this._target),
            request_parameter: cdktf.listMapper(apigatewayv2RouteRequestParameterToTerraform)(this._requestParameter),
        };
    }
}
exports.Apigatewayv2Route = Apigatewayv2Route;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2Route[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2Route", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2Route.tfResourceType = "aws_apigatewayv2_route";
