// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-integration-response.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2IntegrationResponse = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_integration_response.html aws_apigatewayv2_integration_response}.
 *
 * @stability stable
 */
class Apigatewayv2IntegrationResponse extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_integration_response.html aws_apigatewayv2_integration_response} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_integration_response',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._apiId = config.apiId;
        this._contentHandlingStrategy = config.contentHandlingStrategy;
        this._integrationId = config.integrationId;
        this._integrationResponseKey = config.integrationResponseKey;
        this._responseTemplates = config.responseTemplates;
        this._templateSelectionExpression = config.templateSelectionExpression;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    /**
     * @stability stable
     */
    get contentHandlingStrategy() {
        return this.getStringAttribute('content_handling_strategy');
    }
    /**
     * @stability stable
     */
    set contentHandlingStrategy(value) {
        this._contentHandlingStrategy = value;
    }
    /**
     * @stability stable
     */
    resetContentHandlingStrategy() {
        this._contentHandlingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentHandlingStrategyInput() {
        return this._contentHandlingStrategy;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get integrationId() {
        return this.getStringAttribute('integration_id');
    }
    /**
     * @stability stable
     */
    set integrationId(value) {
        this._integrationId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationIdInput() {
        return this._integrationId;
    }
    /**
     * @stability stable
     */
    get integrationResponseKey() {
        return this.getStringAttribute('integration_response_key');
    }
    /**
     * @stability stable
     */
    set integrationResponseKey(value) {
        this._integrationResponseKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationResponseKeyInput() {
        return this._integrationResponseKey;
    }
    /**
     * @stability stable
     */
    get responseTemplates() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_templates');
    }
    /**
     * @stability stable
     */
    set responseTemplates(value) {
        this._responseTemplates = value;
    }
    /**
     * @stability stable
     */
    resetResponseTemplates() {
        this._responseTemplates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseTemplatesInput() {
        return this._responseTemplates;
    }
    /**
     * @stability stable
     */
    get templateSelectionExpression() {
        return this.getStringAttribute('template_selection_expression');
    }
    /**
     * @stability stable
     */
    set templateSelectionExpression(value) {
        this._templateSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetTemplateSelectionExpression() {
        this._templateSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateSelectionExpressionInput() {
        return this._templateSelectionExpression;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            content_handling_strategy: cdktf.stringToTerraform(this._contentHandlingStrategy),
            integration_id: cdktf.stringToTerraform(this._integrationId),
            integration_response_key: cdktf.stringToTerraform(this._integrationResponseKey),
            response_templates: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseTemplates),
            template_selection_expression: cdktf.stringToTerraform(this._templateSelectionExpression),
        };
    }
}
exports.Apigatewayv2IntegrationResponse = Apigatewayv2IntegrationResponse;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2IntegrationResponse[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2IntegrationResponse", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2IntegrationResponse.tfResourceType = "aws_apigatewayv2_integration_response";
