// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-stage.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Stage = exports.apigatewayv2StageRouteSettingsToTerraform = exports.Apigatewayv2StageDefaultRouteSettingsOutputReference = exports.apigatewayv2StageDefaultRouteSettingsToTerraform = exports.Apigatewayv2StageAccessLogSettingsOutputReference = exports.apigatewayv2StageAccessLogSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2StageAccessLogSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_arn: cdktf.stringToTerraform(struct.destinationArn),
        format: cdktf.stringToTerraform(struct.format),
    };
}
exports.apigatewayv2StageAccessLogSettingsToTerraform = apigatewayv2StageAccessLogSettingsToTerraform;
/**
 * @stability stable
 */
class Apigatewayv2StageAccessLogSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationArn) {
            hasAnyValues = true;
            internalValueResult.destinationArn = this._destinationArn;
        }
        if (this._format) {
            hasAnyValues = true;
            internalValueResult.format = this._format;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationArn = undefined;
            this._format = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationArn = value.destinationArn;
            this._format = value.format;
        }
    }
    /**
     * @stability stable
     */
    get destinationArn() {
        return this.getStringAttribute('destination_arn');
    }
    /**
     * @stability stable
     */
    set destinationArn(value) {
        this._destinationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationArnInput() {
        return this._destinationArn;
    }
    /**
     * @stability stable
     */
    get format() {
        return this.getStringAttribute('format');
    }
    /**
     * @stability stable
     */
    set format(value) {
        this._format = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get formatInput() {
        return this._format;
    }
}
exports.Apigatewayv2StageAccessLogSettingsOutputReference = Apigatewayv2StageAccessLogSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2StageAccessLogSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2StageAccessLogSettingsOutputReference", version: "3.0.1" };
function apigatewayv2StageDefaultRouteSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_trace_enabled: cdktf.booleanToTerraform(struct.dataTraceEnabled),
        detailed_metrics_enabled: cdktf.booleanToTerraform(struct.detailedMetricsEnabled),
        logging_level: cdktf.stringToTerraform(struct.loggingLevel),
        throttling_burst_limit: cdktf.numberToTerraform(struct.throttlingBurstLimit),
        throttling_rate_limit: cdktf.numberToTerraform(struct.throttlingRateLimit),
    };
}
exports.apigatewayv2StageDefaultRouteSettingsToTerraform = apigatewayv2StageDefaultRouteSettingsToTerraform;
/**
 * @stability stable
 */
class Apigatewayv2StageDefaultRouteSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataTraceEnabled) {
            hasAnyValues = true;
            internalValueResult.dataTraceEnabled = this._dataTraceEnabled;
        }
        if (this._detailedMetricsEnabled) {
            hasAnyValues = true;
            internalValueResult.detailedMetricsEnabled = this._detailedMetricsEnabled;
        }
        if (this._loggingLevel) {
            hasAnyValues = true;
            internalValueResult.loggingLevel = this._loggingLevel;
        }
        if (this._throttlingBurstLimit) {
            hasAnyValues = true;
            internalValueResult.throttlingBurstLimit = this._throttlingBurstLimit;
        }
        if (this._throttlingRateLimit) {
            hasAnyValues = true;
            internalValueResult.throttlingRateLimit = this._throttlingRateLimit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataTraceEnabled = undefined;
            this._detailedMetricsEnabled = undefined;
            this._loggingLevel = undefined;
            this._throttlingBurstLimit = undefined;
            this._throttlingRateLimit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataTraceEnabled = value.dataTraceEnabled;
            this._detailedMetricsEnabled = value.detailedMetricsEnabled;
            this._loggingLevel = value.loggingLevel;
            this._throttlingBurstLimit = value.throttlingBurstLimit;
            this._throttlingRateLimit = value.throttlingRateLimit;
        }
    }
    /**
     * @stability stable
     */
    get dataTraceEnabled() {
        return this.getBooleanAttribute('data_trace_enabled');
    }
    /**
     * @stability stable
     */
    set dataTraceEnabled(value) {
        this._dataTraceEnabled = value;
    }
    /**
     * @stability stable
     */
    resetDataTraceEnabled() {
        this._dataTraceEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataTraceEnabledInput() {
        return this._dataTraceEnabled;
    }
    /**
     * @stability stable
     */
    get detailedMetricsEnabled() {
        return this.getBooleanAttribute('detailed_metrics_enabled');
    }
    /**
     * @stability stable
     */
    set detailedMetricsEnabled(value) {
        this._detailedMetricsEnabled = value;
    }
    /**
     * @stability stable
     */
    resetDetailedMetricsEnabled() {
        this._detailedMetricsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get detailedMetricsEnabledInput() {
        return this._detailedMetricsEnabled;
    }
    /**
     * @stability stable
     */
    get loggingLevel() {
        return this.getStringAttribute('logging_level');
    }
    /**
     * @stability stable
     */
    set loggingLevel(value) {
        this._loggingLevel = value;
    }
    /**
     * @stability stable
     */
    resetLoggingLevel() {
        this._loggingLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingLevelInput() {
        return this._loggingLevel;
    }
    /**
     * @stability stable
     */
    get throttlingBurstLimit() {
        return this.getNumberAttribute('throttling_burst_limit');
    }
    /**
     * @stability stable
     */
    set throttlingBurstLimit(value) {
        this._throttlingBurstLimit = value;
    }
    /**
     * @stability stable
     */
    resetThrottlingBurstLimit() {
        this._throttlingBurstLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throttlingBurstLimitInput() {
        return this._throttlingBurstLimit;
    }
    /**
     * @stability stable
     */
    get throttlingRateLimit() {
        return this.getNumberAttribute('throttling_rate_limit');
    }
    /**
     * @stability stable
     */
    set throttlingRateLimit(value) {
        this._throttlingRateLimit = value;
    }
    /**
     * @stability stable
     */
    resetThrottlingRateLimit() {
        this._throttlingRateLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throttlingRateLimitInput() {
        return this._throttlingRateLimit;
    }
}
exports.Apigatewayv2StageDefaultRouteSettingsOutputReference = Apigatewayv2StageDefaultRouteSettingsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2StageDefaultRouteSettingsOutputReference[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2StageDefaultRouteSettingsOutputReference", version: "3.0.1" };
function apigatewayv2StageRouteSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_trace_enabled: cdktf.booleanToTerraform(struct.dataTraceEnabled),
        detailed_metrics_enabled: cdktf.booleanToTerraform(struct.detailedMetricsEnabled),
        logging_level: cdktf.stringToTerraform(struct.loggingLevel),
        route_key: cdktf.stringToTerraform(struct.routeKey),
        throttling_burst_limit: cdktf.numberToTerraform(struct.throttlingBurstLimit),
        throttling_rate_limit: cdktf.numberToTerraform(struct.throttlingRateLimit),
    };
}
exports.apigatewayv2StageRouteSettingsToTerraform = apigatewayv2StageRouteSettingsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_stage.html aws_apigatewayv2_stage}.
 *
 * @stability stable
 */
class Apigatewayv2Stage extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_stage.html aws_apigatewayv2_stage} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_stage',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // access_log_settings - computed: false, optional: true, required: false
        this._accessLogSettings = new Apigatewayv2StageAccessLogSettingsOutputReference(this, "access_log_settings", true);
        // default_route_settings - computed: false, optional: true, required: false
        this._defaultRouteSettings = new Apigatewayv2StageDefaultRouteSettingsOutputReference(this, "default_route_settings", true);
        this._apiId = config.apiId;
        this._autoDeploy = config.autoDeploy;
        this._clientCertificateId = config.clientCertificateId;
        this._deploymentId = config.deploymentId;
        this._description = config.description;
        this._name = config.name;
        this._stageVariables = config.stageVariables;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._accessLogSettings.internalValue = config.accessLogSettings;
        this._defaultRouteSettings.internalValue = config.defaultRouteSettings;
        this._routeSettings = config.routeSettings;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get autoDeploy() {
        return this.getBooleanAttribute('auto_deploy');
    }
    /**
     * @stability stable
     */
    set autoDeploy(value) {
        this._autoDeploy = value;
    }
    /**
     * @stability stable
     */
    resetAutoDeploy() {
        this._autoDeploy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoDeployInput() {
        return this._autoDeploy;
    }
    /**
     * @stability stable
     */
    get clientCertificateId() {
        return this.getStringAttribute('client_certificate_id');
    }
    /**
     * @stability stable
     */
    set clientCertificateId(value) {
        this._clientCertificateId = value;
    }
    /**
     * @stability stable
     */
    resetClientCertificateId() {
        this._clientCertificateId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientCertificateIdInput() {
        return this._clientCertificateId;
    }
    /**
     * @stability stable
     */
    get deploymentId() {
        return this.getStringAttribute('deployment_id');
    }
    /**
     * @stability stable
     */
    set deploymentId(value) {
        this._deploymentId = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentId() {
        this._deploymentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentIdInput() {
        return this._deploymentId;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // execution_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get executionArn() {
        return this.getStringAttribute('execution_arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // invoke_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get invokeUrl() {
        return this.getStringAttribute('invoke_url');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get stageVariables() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('stage_variables');
    }
    /**
     * @stability stable
     */
    set stageVariables(value) {
        this._stageVariables = value;
    }
    /**
     * @stability stable
     */
    resetStageVariables() {
        this._stageVariables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stageVariablesInput() {
        return this._stageVariables;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get accessLogSettings() {
        return this._accessLogSettings;
    }
    /**
     * @stability stable
     */
    putAccessLogSettings(value) {
        this._accessLogSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAccessLogSettings() {
        this._accessLogSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessLogSettingsInput() {
        return this._accessLogSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get defaultRouteSettings() {
        return this._defaultRouteSettings;
    }
    /**
     * @stability stable
     */
    putDefaultRouteSettings(value) {
        this._defaultRouteSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDefaultRouteSettings() {
        this._defaultRouteSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultRouteSettingsInput() {
        return this._defaultRouteSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get routeSettings() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('route_settings');
    }
    /**
     * @stability stable
     */
    set routeSettings(value) {
        this._routeSettings = value;
    }
    /**
     * @stability stable
     */
    resetRouteSettings() {
        this._routeSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeSettingsInput() {
        return this._routeSettings;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            auto_deploy: cdktf.booleanToTerraform(this._autoDeploy),
            client_certificate_id: cdktf.stringToTerraform(this._clientCertificateId),
            deployment_id: cdktf.stringToTerraform(this._deploymentId),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            stage_variables: cdktf.hashMapper(cdktf.anyToTerraform)(this._stageVariables),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            access_log_settings: apigatewayv2StageAccessLogSettingsToTerraform(this._accessLogSettings.internalValue),
            default_route_settings: apigatewayv2StageDefaultRouteSettingsToTerraform(this._defaultRouteSettings.internalValue),
            route_settings: cdktf.listMapper(apigatewayv2StageRouteSettingsToTerraform)(this._routeSettings),
        };
    }
}
exports.Apigatewayv2Stage = Apigatewayv2Stage;
_c = JSII_RTTI_SYMBOL_1;
Apigatewayv2Stage[_c] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2Stage", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2Stage.tfResourceType = "aws_apigatewayv2_stage";
