// @cdktf/provider-aws/lib/apigateway/api-gateway-rest-api.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayRestApi = exports.ApiGatewayRestApiEndpointConfigurationOutputReference = exports.apiGatewayRestApiEndpointConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apiGatewayRestApiEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        types: cdktf.listMapper(cdktf.stringToTerraform)(struct.types),
        vpc_endpoint_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.vpcEndpointIds),
    };
}
exports.apiGatewayRestApiEndpointConfigurationToTerraform = apiGatewayRestApiEndpointConfigurationToTerraform;
/**
 * @stability stable
 */
class ApiGatewayRestApiEndpointConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._types) {
            hasAnyValues = true;
            internalValueResult.types = this._types;
        }
        if (this._vpcEndpointIds) {
            hasAnyValues = true;
            internalValueResult.vpcEndpointIds = this._vpcEndpointIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._types = undefined;
            this._vpcEndpointIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._types = value.types;
            this._vpcEndpointIds = value.vpcEndpointIds;
        }
    }
    /**
     * @stability stable
     */
    get types() {
        return this.getListAttribute('types');
    }
    /**
     * @stability stable
     */
    set types(value) {
        this._types = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typesInput() {
        return this._types;
    }
    /**
     * @stability stable
     */
    get vpcEndpointIds() {
        return this.getListAttribute('vpc_endpoint_ids');
    }
    /**
     * @stability stable
     */
    set vpcEndpointIds(value) {
        this._vpcEndpointIds = value;
    }
    /**
     * @stability stable
     */
    resetVpcEndpointIds() {
        this._vpcEndpointIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcEndpointIdsInput() {
        return this._vpcEndpointIds;
    }
}
exports.ApiGatewayRestApiEndpointConfigurationOutputReference = ApiGatewayRestApiEndpointConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayRestApiEndpointConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayRestApiEndpointConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_rest_api.html aws_api_gateway_rest_api}.
 *
 * @stability stable
 */
class ApiGatewayRestApi extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_rest_api.html aws_api_gateway_rest_api} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_rest_api',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // endpoint_configuration - computed: false, optional: true, required: false
        this._endpointConfiguration = new ApiGatewayRestApiEndpointConfigurationOutputReference(this, "endpoint_configuration", true);
        this._apiKeySource = config.apiKeySource;
        this._binaryMediaTypes = config.binaryMediaTypes;
        this._body = config.body;
        this._description = config.description;
        this._disableExecuteApiEndpoint = config.disableExecuteApiEndpoint;
        this._minimumCompressionSize = config.minimumCompressionSize;
        this._name = config.name;
        this._parameters = config.parameters;
        this._policy = config.policy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._endpointConfiguration.internalValue = config.endpointConfiguration;
    }
    /**
     * @stability stable
     */
    get apiKeySource() {
        return this.getStringAttribute('api_key_source');
    }
    /**
     * @stability stable
     */
    set apiKeySource(value) {
        this._apiKeySource = value;
    }
    /**
     * @stability stable
     */
    resetApiKeySource() {
        this._apiKeySource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiKeySourceInput() {
        return this._apiKeySource;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get binaryMediaTypes() {
        return this.getListAttribute('binary_media_types');
    }
    /**
     * @stability stable
     */
    set binaryMediaTypes(value) {
        this._binaryMediaTypes = value;
    }
    /**
     * @stability stable
     */
    resetBinaryMediaTypes() {
        this._binaryMediaTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get binaryMediaTypesInput() {
        return this._binaryMediaTypes;
    }
    /**
     * @stability stable
     */
    get body() {
        return this.getStringAttribute('body');
    }
    /**
     * @stability stable
     */
    set body(value) {
        this._body = value;
    }
    /**
     * @stability stable
     */
    resetBody() {
        this._body = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bodyInput() {
        return this._body;
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get disableExecuteApiEndpoint() {
        return this.getBooleanAttribute('disable_execute_api_endpoint');
    }
    /**
     * @stability stable
     */
    set disableExecuteApiEndpoint(value) {
        this._disableExecuteApiEndpoint = value;
    }
    /**
     * @stability stable
     */
    resetDisableExecuteApiEndpoint() {
        this._disableExecuteApiEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableExecuteApiEndpointInput() {
        return this._disableExecuteApiEndpoint;
    }
    // execution_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get executionArn() {
        return this.getStringAttribute('execution_arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get minimumCompressionSize() {
        return this.getNumberAttribute('minimum_compression_size');
    }
    /**
     * @stability stable
     */
    set minimumCompressionSize(value) {
        this._minimumCompressionSize = value;
    }
    /**
     * @stability stable
     */
    resetMinimumCompressionSize() {
        this._minimumCompressionSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minimumCompressionSizeInput() {
        return this._minimumCompressionSize;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    /**
     * @stability stable
     */
    set policy(value) {
        this._policy = value;
    }
    /**
     * @stability stable
     */
    resetPolicy() {
        this._policy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyInput() {
        return this._policy;
    }
    // root_resource_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rootResourceId() {
        return this.getStringAttribute('root_resource_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get endpointConfiguration() {
        return this._endpointConfiguration;
    }
    /**
     * @stability stable
     */
    putEndpointConfiguration(value) {
        this._endpointConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEndpointConfiguration() {
        this._endpointConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointConfigurationInput() {
        return this._endpointConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_key_source: cdktf.stringToTerraform(this._apiKeySource),
            binary_media_types: cdktf.listMapper(cdktf.stringToTerraform)(this._binaryMediaTypes),
            body: cdktf.stringToTerraform(this._body),
            description: cdktf.stringToTerraform(this._description),
            disable_execute_api_endpoint: cdktf.booleanToTerraform(this._disableExecuteApiEndpoint),
            minimum_compression_size: cdktf.numberToTerraform(this._minimumCompressionSize),
            name: cdktf.stringToTerraform(this._name),
            parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._parameters),
            policy: cdktf.stringToTerraform(this._policy),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            endpoint_configuration: apiGatewayRestApiEndpointConfigurationToTerraform(this._endpointConfiguration.internalValue),
        };
    }
}
exports.ApiGatewayRestApi = ApiGatewayRestApi;
_b = JSII_RTTI_SYMBOL_1;
ApiGatewayRestApi[_b] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayRestApi", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayRestApi.tfResourceType = "aws_api_gateway_rest_api";
