// @cdktf/provider-aws/lib/apigateway/api-gateway-integration-response.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayIntegrationResponse = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_integration_response.html aws_api_gateway_integration_response}.
 *
 * @stability stable
 */
class ApiGatewayIntegrationResponse extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_integration_response.html aws_api_gateway_integration_response} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_integration_response',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._contentHandling = config.contentHandling;
        this._httpMethod = config.httpMethod;
        this._resourceId = config.resourceId;
        this._responseParameters = config.responseParameters;
        this._responseTemplates = config.responseTemplates;
        this._restApiId = config.restApiId;
        this._selectionPattern = config.selectionPattern;
        this._statusCode = config.statusCode;
    }
    /**
     * @stability stable
     */
    get contentHandling() {
        return this.getStringAttribute('content_handling');
    }
    /**
     * @stability stable
     */
    set contentHandling(value) {
        this._contentHandling = value;
    }
    /**
     * @stability stable
     */
    resetContentHandling() {
        this._contentHandling = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentHandlingInput() {
        return this._contentHandling;
    }
    /**
     * @stability stable
     */
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    /**
     * @stability stable
     */
    set httpMethod(value) {
        this._httpMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpMethodInput() {
        return this._httpMethod;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    /**
     * @stability stable
     */
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceIdInput() {
        return this._resourceId;
    }
    /**
     * @stability stable
     */
    get responseParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_parameters');
    }
    /**
     * @stability stable
     */
    set responseParameters(value) {
        this._responseParameters = value;
    }
    /**
     * @stability stable
     */
    resetResponseParameters() {
        this._responseParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseParametersInput() {
        return this._responseParameters;
    }
    /**
     * @stability stable
     */
    get responseTemplates() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_templates');
    }
    /**
     * @stability stable
     */
    set responseTemplates(value) {
        this._responseTemplates = value;
    }
    /**
     * @stability stable
     */
    resetResponseTemplates() {
        this._responseTemplates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseTemplatesInput() {
        return this._responseTemplates;
    }
    /**
     * @stability stable
     */
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    /**
     * @stability stable
     */
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restApiIdInput() {
        return this._restApiId;
    }
    /**
     * @stability stable
     */
    get selectionPattern() {
        return this.getStringAttribute('selection_pattern');
    }
    /**
     * @stability stable
     */
    set selectionPattern(value) {
        this._selectionPattern = value;
    }
    /**
     * @stability stable
     */
    resetSelectionPattern() {
        this._selectionPattern = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get selectionPatternInput() {
        return this._selectionPattern;
    }
    /**
     * @stability stable
     */
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
    /**
     * @stability stable
     */
    set statusCode(value) {
        this._statusCode = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusCodeInput() {
        return this._statusCode;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            content_handling: cdktf.stringToTerraform(this._contentHandling),
            http_method: cdktf.stringToTerraform(this._httpMethod),
            resource_id: cdktf.stringToTerraform(this._resourceId),
            response_parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseParameters),
            response_templates: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseTemplates),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            selection_pattern: cdktf.stringToTerraform(this._selectionPattern),
            status_code: cdktf.stringToTerraform(this._statusCode),
        };
    }
}
exports.ApiGatewayIntegrationResponse = ApiGatewayIntegrationResponse;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayIntegrationResponse[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayIntegrationResponse", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayIntegrationResponse.tfResourceType = "aws_api_gateway_integration_response";
