// @cdktf/provider-aws/lib/apigateway/api-gateway-gateway-response.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayGatewayResponse = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_gateway_response.html aws_api_gateway_gateway_response}.
 *
 * @stability stable
 */
class ApiGatewayGatewayResponse extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_gateway_response.html aws_api_gateway_gateway_response} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_gateway_response',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._responseParameters = config.responseParameters;
        this._responseTemplates = config.responseTemplates;
        this._responseType = config.responseType;
        this._restApiId = config.restApiId;
        this._statusCode = config.statusCode;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get responseParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_parameters');
    }
    /**
     * @stability stable
     */
    set responseParameters(value) {
        this._responseParameters = value;
    }
    /**
     * @stability stable
     */
    resetResponseParameters() {
        this._responseParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseParametersInput() {
        return this._responseParameters;
    }
    /**
     * @stability stable
     */
    get responseTemplates() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_templates');
    }
    /**
     * @stability stable
     */
    set responseTemplates(value) {
        this._responseTemplates = value;
    }
    /**
     * @stability stable
     */
    resetResponseTemplates() {
        this._responseTemplates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseTemplatesInput() {
        return this._responseTemplates;
    }
    /**
     * @stability stable
     */
    get responseType() {
        return this.getStringAttribute('response_type');
    }
    /**
     * @stability stable
     */
    set responseType(value) {
        this._responseType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseTypeInput() {
        return this._responseType;
    }
    /**
     * @stability stable
     */
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    /**
     * @stability stable
     */
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restApiIdInput() {
        return this._restApiId;
    }
    /**
     * @stability stable
     */
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
    /**
     * @stability stable
     */
    set statusCode(value) {
        this._statusCode = value;
    }
    /**
     * @stability stable
     */
    resetStatusCode() {
        this._statusCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusCodeInput() {
        return this._statusCode;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            response_parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseParameters),
            response_templates: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseTemplates),
            response_type: cdktf.stringToTerraform(this._responseType),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            status_code: cdktf.stringToTerraform(this._statusCode),
        };
    }
}
exports.ApiGatewayGatewayResponse = ApiGatewayGatewayResponse;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayGatewayResponse[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayGatewayResponse", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayGatewayResponse.tfResourceType = "aws_api_gateway_gateway_response";
