// @cdktf/provider-aws/lib/apigateway/api-gateway-usage-plan.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayUsagePlan = exports.ApiGatewayUsagePlanThrottleSettingsOutputReference = exports.apiGatewayUsagePlanThrottleSettingsToTerraform = exports.ApiGatewayUsagePlanQuotaSettingsOutputReference = exports.apiGatewayUsagePlanQuotaSettingsToTerraform = exports.apiGatewayUsagePlanApiStagesToTerraform = exports.apiGatewayUsagePlanApiStagesThrottleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apiGatewayUsagePlanApiStagesThrottleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        burst_limit: cdktf.numberToTerraform(struct.burstLimit),
        path: cdktf.stringToTerraform(struct.path),
        rate_limit: cdktf.numberToTerraform(struct.rateLimit),
    };
}
exports.apiGatewayUsagePlanApiStagesThrottleToTerraform = apiGatewayUsagePlanApiStagesThrottleToTerraform;
function apiGatewayUsagePlanApiStagesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        api_id: cdktf.stringToTerraform(struct.apiId),
        stage: cdktf.stringToTerraform(struct.stage),
        throttle: cdktf.listMapper(apiGatewayUsagePlanApiStagesThrottleToTerraform)(struct.throttle),
    };
}
exports.apiGatewayUsagePlanApiStagesToTerraform = apiGatewayUsagePlanApiStagesToTerraform;
function apiGatewayUsagePlanQuotaSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        limit: cdktf.numberToTerraform(struct.limit),
        offset: cdktf.numberToTerraform(struct.offset),
        period: cdktf.stringToTerraform(struct.period),
    };
}
exports.apiGatewayUsagePlanQuotaSettingsToTerraform = apiGatewayUsagePlanQuotaSettingsToTerraform;
/**
 * @stability stable
 */
class ApiGatewayUsagePlanQuotaSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._limit) {
            hasAnyValues = true;
            internalValueResult.limit = this._limit;
        }
        if (this._offset) {
            hasAnyValues = true;
            internalValueResult.offset = this._offset;
        }
        if (this._period) {
            hasAnyValues = true;
            internalValueResult.period = this._period;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._limit = undefined;
            this._offset = undefined;
            this._period = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._limit = value.limit;
            this._offset = value.offset;
            this._period = value.period;
        }
    }
    /**
     * @stability stable
     */
    get limit() {
        return this.getNumberAttribute('limit');
    }
    /**
     * @stability stable
     */
    set limit(value) {
        this._limit = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get limitInput() {
        return this._limit;
    }
    /**
     * @stability stable
     */
    get offset() {
        return this.getNumberAttribute('offset');
    }
    /**
     * @stability stable
     */
    set offset(value) {
        this._offset = value;
    }
    /**
     * @stability stable
     */
    resetOffset() {
        this._offset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get offsetInput() {
        return this._offset;
    }
    /**
     * @stability stable
     */
    get period() {
        return this.getStringAttribute('period');
    }
    /**
     * @stability stable
     */
    set period(value) {
        this._period = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get periodInput() {
        return this._period;
    }
}
exports.ApiGatewayUsagePlanQuotaSettingsOutputReference = ApiGatewayUsagePlanQuotaSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayUsagePlanQuotaSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayUsagePlanQuotaSettingsOutputReference", version: "3.0.1" };
function apiGatewayUsagePlanThrottleSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        burst_limit: cdktf.numberToTerraform(struct.burstLimit),
        rate_limit: cdktf.numberToTerraform(struct.rateLimit),
    };
}
exports.apiGatewayUsagePlanThrottleSettingsToTerraform = apiGatewayUsagePlanThrottleSettingsToTerraform;
/**
 * @stability stable
 */
class ApiGatewayUsagePlanThrottleSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._burstLimit) {
            hasAnyValues = true;
            internalValueResult.burstLimit = this._burstLimit;
        }
        if (this._rateLimit) {
            hasAnyValues = true;
            internalValueResult.rateLimit = this._rateLimit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._burstLimit = undefined;
            this._rateLimit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._burstLimit = value.burstLimit;
            this._rateLimit = value.rateLimit;
        }
    }
    /**
     * @stability stable
     */
    get burstLimit() {
        return this.getNumberAttribute('burst_limit');
    }
    /**
     * @stability stable
     */
    set burstLimit(value) {
        this._burstLimit = value;
    }
    /**
     * @stability stable
     */
    resetBurstLimit() {
        this._burstLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get burstLimitInput() {
        return this._burstLimit;
    }
    /**
     * @stability stable
     */
    get rateLimit() {
        return this.getNumberAttribute('rate_limit');
    }
    /**
     * @stability stable
     */
    set rateLimit(value) {
        this._rateLimit = value;
    }
    /**
     * @stability stable
     */
    resetRateLimit() {
        this._rateLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rateLimitInput() {
        return this._rateLimit;
    }
}
exports.ApiGatewayUsagePlanThrottleSettingsOutputReference = ApiGatewayUsagePlanThrottleSettingsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ApiGatewayUsagePlanThrottleSettingsOutputReference[_b] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayUsagePlanThrottleSettingsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_usage_plan.html aws_api_gateway_usage_plan}.
 *
 * @stability stable
 */
class ApiGatewayUsagePlan extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_usage_plan.html aws_api_gateway_usage_plan} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_usage_plan',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // quota_settings - computed: false, optional: true, required: false
        this._quotaSettings = new ApiGatewayUsagePlanQuotaSettingsOutputReference(this, "quota_settings", true);
        // throttle_settings - computed: false, optional: true, required: false
        this._throttleSettings = new ApiGatewayUsagePlanThrottleSettingsOutputReference(this, "throttle_settings", true);
        this._description = config.description;
        this._name = config.name;
        this._productCode = config.productCode;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._apiStages = config.apiStages;
        this._quotaSettings.internalValue = config.quotaSettings;
        this._throttleSettings.internalValue = config.throttleSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get productCode() {
        return this.getStringAttribute('product_code');
    }
    /**
     * @stability stable
     */
    set productCode(value) {
        this._productCode = value;
    }
    /**
     * @stability stable
     */
    resetProductCode() {
        this._productCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productCodeInput() {
        return this._productCode;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get apiStages() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('api_stages');
    }
    /**
     * @stability stable
     */
    set apiStages(value) {
        this._apiStages = value;
    }
    /**
     * @stability stable
     */
    resetApiStages() {
        this._apiStages = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiStagesInput() {
        return this._apiStages;
    }
    /**
     * @stability stable
     */
    get quotaSettings() {
        return this._quotaSettings;
    }
    /**
     * @stability stable
     */
    putQuotaSettings(value) {
        this._quotaSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetQuotaSettings() {
        this._quotaSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get quotaSettingsInput() {
        return this._quotaSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get throttleSettings() {
        return this._throttleSettings;
    }
    /**
     * @stability stable
     */
    putThrottleSettings(value) {
        this._throttleSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetThrottleSettings() {
        this._throttleSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throttleSettingsInput() {
        return this._throttleSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            product_code: cdktf.stringToTerraform(this._productCode),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            api_stages: cdktf.listMapper(apiGatewayUsagePlanApiStagesToTerraform)(this._apiStages),
            quota_settings: apiGatewayUsagePlanQuotaSettingsToTerraform(this._quotaSettings.internalValue),
            throttle_settings: apiGatewayUsagePlanThrottleSettingsToTerraform(this._throttleSettings.internalValue),
        };
    }
}
exports.ApiGatewayUsagePlan = ApiGatewayUsagePlan;
_c = JSII_RTTI_SYMBOL_1;
ApiGatewayUsagePlan[_c] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayUsagePlan", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayUsagePlan.tfResourceType = "aws_api_gateway_usage_plan";
