// @cdktf/provider-aws/lib/apigateway/api-gateway-domain-name.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayDomainName = exports.ApiGatewayDomainNameMutualTlsAuthenticationOutputReference = exports.apiGatewayDomainNameMutualTlsAuthenticationToTerraform = exports.ApiGatewayDomainNameEndpointConfigurationOutputReference = exports.apiGatewayDomainNameEndpointConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apiGatewayDomainNameEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        types: cdktf.listMapper(cdktf.stringToTerraform)(struct.types),
    };
}
exports.apiGatewayDomainNameEndpointConfigurationToTerraform = apiGatewayDomainNameEndpointConfigurationToTerraform;
/**
 * @stability stable
 */
class ApiGatewayDomainNameEndpointConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._types) {
            hasAnyValues = true;
            internalValueResult.types = this._types;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._types = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._types = value.types;
        }
    }
    /**
     * @stability stable
     */
    get types() {
        return this.getListAttribute('types');
    }
    /**
     * @stability stable
     */
    set types(value) {
        this._types = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typesInput() {
        return this._types;
    }
}
exports.ApiGatewayDomainNameEndpointConfigurationOutputReference = ApiGatewayDomainNameEndpointConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayDomainNameEndpointConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayDomainNameEndpointConfigurationOutputReference", version: "3.0.1" };
function apiGatewayDomainNameMutualTlsAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        truststore_uri: cdktf.stringToTerraform(struct.truststoreUri),
        truststore_version: cdktf.stringToTerraform(struct.truststoreVersion),
    };
}
exports.apiGatewayDomainNameMutualTlsAuthenticationToTerraform = apiGatewayDomainNameMutualTlsAuthenticationToTerraform;
/**
 * @stability stable
 */
class ApiGatewayDomainNameMutualTlsAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._truststoreUri) {
            hasAnyValues = true;
            internalValueResult.truststoreUri = this._truststoreUri;
        }
        if (this._truststoreVersion) {
            hasAnyValues = true;
            internalValueResult.truststoreVersion = this._truststoreVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._truststoreUri = undefined;
            this._truststoreVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._truststoreUri = value.truststoreUri;
            this._truststoreVersion = value.truststoreVersion;
        }
    }
    /**
     * @stability stable
     */
    get truststoreUri() {
        return this.getStringAttribute('truststore_uri');
    }
    /**
     * @stability stable
     */
    set truststoreUri(value) {
        this._truststoreUri = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get truststoreUriInput() {
        return this._truststoreUri;
    }
    /**
     * @stability stable
     */
    get truststoreVersion() {
        return this.getStringAttribute('truststore_version');
    }
    /**
     * @stability stable
     */
    set truststoreVersion(value) {
        this._truststoreVersion = value;
    }
    /**
     * @stability stable
     */
    resetTruststoreVersion() {
        this._truststoreVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get truststoreVersionInput() {
        return this._truststoreVersion;
    }
}
exports.ApiGatewayDomainNameMutualTlsAuthenticationOutputReference = ApiGatewayDomainNameMutualTlsAuthenticationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ApiGatewayDomainNameMutualTlsAuthenticationOutputReference[_b] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayDomainNameMutualTlsAuthenticationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_domain_name.html aws_api_gateway_domain_name}.
 *
 * @stability stable
 */
class ApiGatewayDomainName extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_domain_name.html aws_api_gateway_domain_name} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_domain_name',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // endpoint_configuration - computed: false, optional: true, required: false
        this._endpointConfiguration = new ApiGatewayDomainNameEndpointConfigurationOutputReference(this, "endpoint_configuration", true);
        // mutual_tls_authentication - computed: false, optional: true, required: false
        this._mutualTlsAuthentication = new ApiGatewayDomainNameMutualTlsAuthenticationOutputReference(this, "mutual_tls_authentication", true);
        this._certificateArn = config.certificateArn;
        this._certificateBody = config.certificateBody;
        this._certificateChain = config.certificateChain;
        this._certificateName = config.certificateName;
        this._certificatePrivateKey = config.certificatePrivateKey;
        this._domainName = config.domainName;
        this._regionalCertificateArn = config.regionalCertificateArn;
        this._regionalCertificateName = config.regionalCertificateName;
        this._securityPolicy = config.securityPolicy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._endpointConfiguration.internalValue = config.endpointConfiguration;
        this._mutualTlsAuthentication.internalValue = config.mutualTlsAuthentication;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    /**
     * @stability stable
     */
    set certificateArn(value) {
        this._certificateArn = value;
    }
    /**
     * @stability stable
     */
    resetCertificateArn() {
        this._certificateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateArnInput() {
        return this._certificateArn;
    }
    /**
     * @stability stable
     */
    get certificateBody() {
        return this.getStringAttribute('certificate_body');
    }
    /**
     * @stability stable
     */
    set certificateBody(value) {
        this._certificateBody = value;
    }
    /**
     * @stability stable
     */
    resetCertificateBody() {
        this._certificateBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateBodyInput() {
        return this._certificateBody;
    }
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    /**
     * @stability stable
     */
    set certificateChain(value) {
        this._certificateChain = value;
    }
    /**
     * @stability stable
     */
    resetCertificateChain() {
        this._certificateChain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateChainInput() {
        return this._certificateChain;
    }
    /**
     * @stability stable
     */
    get certificateName() {
        return this.getStringAttribute('certificate_name');
    }
    /**
     * @stability stable
     */
    set certificateName(value) {
        this._certificateName = value;
    }
    /**
     * @stability stable
     */
    resetCertificateName() {
        this._certificateName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateNameInput() {
        return this._certificateName;
    }
    /**
     * @stability stable
     */
    get certificatePrivateKey() {
        return this.getStringAttribute('certificate_private_key');
    }
    /**
     * @stability stable
     */
    set certificatePrivateKey(value) {
        this._certificatePrivateKey = value;
    }
    /**
     * @stability stable
     */
    resetCertificatePrivateKey() {
        this._certificatePrivateKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificatePrivateKeyInput() {
        return this._certificatePrivateKey;
    }
    // certificate_upload_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateUploadDate() {
        return this.getStringAttribute('certificate_upload_date');
    }
    // cloudfront_domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudfrontDomainName() {
        return this.getStringAttribute('cloudfront_domain_name');
    }
    // cloudfront_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudfrontZoneId() {
        return this.getStringAttribute('cloudfront_zone_id');
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get regionalCertificateArn() {
        return this.getStringAttribute('regional_certificate_arn');
    }
    /**
     * @stability stable
     */
    set regionalCertificateArn(value) {
        this._regionalCertificateArn = value;
    }
    /**
     * @stability stable
     */
    resetRegionalCertificateArn() {
        this._regionalCertificateArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionalCertificateArnInput() {
        return this._regionalCertificateArn;
    }
    /**
     * @stability stable
     */
    get regionalCertificateName() {
        return this.getStringAttribute('regional_certificate_name');
    }
    /**
     * @stability stable
     */
    set regionalCertificateName(value) {
        this._regionalCertificateName = value;
    }
    /**
     * @stability stable
     */
    resetRegionalCertificateName() {
        this._regionalCertificateName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionalCertificateNameInput() {
        return this._regionalCertificateName;
    }
    // regional_domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalDomainName() {
        return this.getStringAttribute('regional_domain_name');
    }
    // regional_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalZoneId() {
        return this.getStringAttribute('regional_zone_id');
    }
    /**
     * @stability stable
     */
    get securityPolicy() {
        return this.getStringAttribute('security_policy');
    }
    /**
     * @stability stable
     */
    set securityPolicy(value) {
        this._securityPolicy = value;
    }
    /**
     * @stability stable
     */
    resetSecurityPolicy() {
        this._securityPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityPolicyInput() {
        return this._securityPolicy;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get endpointConfiguration() {
        return this._endpointConfiguration;
    }
    /**
     * @stability stable
     */
    putEndpointConfiguration(value) {
        this._endpointConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEndpointConfiguration() {
        this._endpointConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointConfigurationInput() {
        return this._endpointConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get mutualTlsAuthentication() {
        return this._mutualTlsAuthentication;
    }
    /**
     * @stability stable
     */
    putMutualTlsAuthentication(value) {
        this._mutualTlsAuthentication.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMutualTlsAuthentication() {
        this._mutualTlsAuthentication.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get mutualTlsAuthenticationInput() {
        return this._mutualTlsAuthentication.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            certificate_arn: cdktf.stringToTerraform(this._certificateArn),
            certificate_body: cdktf.stringToTerraform(this._certificateBody),
            certificate_chain: cdktf.stringToTerraform(this._certificateChain),
            certificate_name: cdktf.stringToTerraform(this._certificateName),
            certificate_private_key: cdktf.stringToTerraform(this._certificatePrivateKey),
            domain_name: cdktf.stringToTerraform(this._domainName),
            regional_certificate_arn: cdktf.stringToTerraform(this._regionalCertificateArn),
            regional_certificate_name: cdktf.stringToTerraform(this._regionalCertificateName),
            security_policy: cdktf.stringToTerraform(this._securityPolicy),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            endpoint_configuration: apiGatewayDomainNameEndpointConfigurationToTerraform(this._endpointConfiguration.internalValue),
            mutual_tls_authentication: apiGatewayDomainNameMutualTlsAuthenticationToTerraform(this._mutualTlsAuthentication.internalValue),
        };
    }
}
exports.ApiGatewayDomainName = ApiGatewayDomainName;
_c = JSII_RTTI_SYMBOL_1;
ApiGatewayDomainName[_c] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayDomainName", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayDomainName.tfResourceType = "aws_api_gateway_domain_name";
