// cdktf-cli/bin/cmds/get.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const provider_generator_1 = require("@cdktf/provider-generator");
const get_1 = require("./ui/get");
const render_ink_1 = require("./helper/render-ink");
const version_check_1 = require("./helper/version-check");
const check_directory_1 = require("./helper/check-directory");
const check_environment_1 = require("./helper/check-environment");
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "get [OPTIONS]";
        this.describe = "Generate CDK Constructs for Terraform providers and modules.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .option("output", {
            default: config.codeMakerOutput,
            type: "string",
            desc: "Output directory for generated Constructs",
            alias: "o",
        })
            .option("language", {
            default: config.language,
            required: true,
            type: "string",
            desc: "Output programming language",
            alias: "l",
            choices: provider_generator_1.LANGUAGES,
        });
    }
    async handler(argv) {
        var _a, _b;
        check_directory_1.throwIfNotProjectDirectory("get");
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("get");
        const args = argv;
        const providers = (_a = config.terraformProviders) !== null && _a !== void 0 ? _a : [];
        const modules = (_b = config.terraformModules) !== null && _b !== void 0 ? _b : [];
        const { output, language } = args;
        const constraints = [
            ...providers,
            ...modules,
        ];
        if (constraints.length === 0) {
            console.error(`ERROR: Please specify providers or modules in "cdktf.json" config file`);
            process.exit(1);
        }
        await render_ink_1.renderInk(react_1.default.createElement(get_1.Get, {
            codeMakerOutput: output,
            language: language,
            constraints,
        }));
    }
}
module.exports = new Command();
