// cdktf-cli/bin/cmds/list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const list_1 = require("./ui/list");
const provider_generator_1 = require("@cdktf/provider-generator");
const render_ink_1 = require("./helper/render-ink");
const version_check_1 = require("./helper/version-check");
const check_directory_1 = require("./helper/check-directory");
const check_environment_1 = require("./helper/check-environment");
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "list [OPTIONS]";
        this.describe = "List stacks in app.";
        this.builder = (args) => args
            .option("app", {
            default: config.app,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            desc: "Output directory",
            alias: "o",
        })
            .showHelpOnFail(true);
    }
    async handler(argv) {
        check_directory_1.throwIfNotProjectDirectory("list");
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("list");
        const command = argv.app;
        const outdir = argv.output;
        await render_ink_1.renderInk(react_1.default.createElement(list_1.List, { targetDir: outdir, synthCommand: command }));
    }
}
module.exports = new Command();
