// @cdktf/provider-aws/lib/lambdafunction/lambda-alias.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaAlias = exports.LambdaAliasRoutingConfigOutputReference = exports.lambdaAliasRoutingConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lambdaAliasRoutingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        additional_version_weights: cdktf.hashMapper(cdktf.anyToTerraform)(struct.additionalVersionWeights),
    };
}
exports.lambdaAliasRoutingConfigToTerraform = lambdaAliasRoutingConfigToTerraform;
/**
 * @stability stable
 */
class LambdaAliasRoutingConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._additionalVersionWeights) {
            hasAnyValues = true;
            internalValueResult.additionalVersionWeights = this._additionalVersionWeights;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._additionalVersionWeights = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._additionalVersionWeights = value.additionalVersionWeights;
        }
    }
    /**
     * @stability stable
     */
    get additionalVersionWeights() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('additional_version_weights');
    }
    /**
     * @stability stable
     */
    set additionalVersionWeights(value) {
        this._additionalVersionWeights = value;
    }
    /**
     * @stability stable
     */
    resetAdditionalVersionWeights() {
        this._additionalVersionWeights = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get additionalVersionWeightsInput() {
        return this._additionalVersionWeights;
    }
}
exports.LambdaAliasRoutingConfigOutputReference = LambdaAliasRoutingConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LambdaAliasRoutingConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaAliasRoutingConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lambda_alias.html aws_lambda_alias}.
 *
 * @stability stable
 */
class LambdaAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lambda_alias.html aws_lambda_alias} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // routing_config - computed: false, optional: true, required: false
        this._routingConfig = new LambdaAliasRoutingConfigOutputReference(this, "routing_config", true);
        this._description = config.description;
        this._functionName = config.functionName;
        this._functionVersion = config.functionVersion;
        this._name = config.name;
        this._routingConfig.internalValue = config.routingConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    /**
     * @stability stable
     */
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get functionNameInput() {
        return this._functionName;
    }
    /**
     * @stability stable
     */
    get functionVersion() {
        return this.getStringAttribute('function_version');
    }
    /**
     * @stability stable
     */
    set functionVersion(value) {
        this._functionVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get functionVersionInput() {
        return this._functionVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // invoke_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get invokeArn() {
        return this.getStringAttribute('invoke_arn');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get routingConfig() {
        return this._routingConfig;
    }
    /**
     * @stability stable
     */
    putRoutingConfig(value) {
        this._routingConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRoutingConfig() {
        this._routingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routingConfigInput() {
        return this._routingConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            function_name: cdktf.stringToTerraform(this._functionName),
            function_version: cdktf.stringToTerraform(this._functionVersion),
            name: cdktf.stringToTerraform(this._name),
            routing_config: lambdaAliasRoutingConfigToTerraform(this._routingConfig.internalValue),
        };
    }
}
exports.LambdaAlias = LambdaAlias;
_b = JSII_RTTI_SYMBOL_1;
LambdaAlias[_b] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaAlias", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LambdaAlias.tfResourceType = "aws_lambda_alias";
