// @cdktf/provider-aws/lib/appstream/appstream-user-stack-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamUserStackAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/appstream_user_stack_association.html aws_appstream_user_stack_association}.
 *
 * @stability stable
 */
class AppstreamUserStackAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appstream_user_stack_association.html aws_appstream_user_stack_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_user_stack_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._authenticationType = config.authenticationType;
        this._sendEmailNotification = config.sendEmailNotification;
        this._stackName = config.stackName;
        this._userName = config.userName;
    }
    /**
     * @stability stable
     */
    get authenticationType() {
        return this.getStringAttribute('authentication_type');
    }
    /**
     * @stability stable
     */
    set authenticationType(value) {
        this._authenticationType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authenticationTypeInput() {
        return this._authenticationType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get sendEmailNotification() {
        return this.getBooleanAttribute('send_email_notification');
    }
    /**
     * @stability stable
     */
    set sendEmailNotification(value) {
        this._sendEmailNotification = value;
    }
    /**
     * @stability stable
     */
    resetSendEmailNotification() {
        this._sendEmailNotification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sendEmailNotificationInput() {
        return this._sendEmailNotification;
    }
    /**
     * @stability stable
     */
    get stackName() {
        return this.getStringAttribute('stack_name');
    }
    /**
     * @stability stable
     */
    set stackName(value) {
        this._stackName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackNameInput() {
        return this._stackName;
    }
    /**
     * @stability stable
     */
    get userName() {
        return this.getStringAttribute('user_name');
    }
    /**
     * @stability stable
     */
    set userName(value) {
        this._userName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userNameInput() {
        return this._userName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            authentication_type: cdktf.stringToTerraform(this._authenticationType),
            send_email_notification: cdktf.booleanToTerraform(this._sendEmailNotification),
            stack_name: cdktf.stringToTerraform(this._stackName),
            user_name: cdktf.stringToTerraform(this._userName),
        };
    }
}
exports.AppstreamUserStackAssociation = AppstreamUserStackAssociation;
_a = JSII_RTTI_SYMBOL_1;
AppstreamUserStackAssociation[_a] = { fqn: "@cdktf/provider-aws.appstream.AppstreamUserStackAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AppstreamUserStackAssociation.tfResourceType = "aws_appstream_user_stack_association";
