// jsii/lib/helpers.js
"use strict";
/**
 * Helper routines for use with the jsii compiler
 *
 * These are mostly used for testing, but all projects that need to exercise
 * the JSII compiler to test something need to share this code, so might as
 * well put it in one reusable place.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestWorkspace = exports.compileJsiiForTest = exports.sourceToAssemblyHelper = void 0;
const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const typescript_1 = require("typescript");
const compiler_1 = require("./compiler");
const project_info_1 = require("./project-info");
const utils_1 = require("./utils");
/**
 * Compile a piece of source and return the JSII assembly for it
 *
 * Only usable for trivial cases and tests.
 *
 * @param source can either be a single `string` (the content of `index.ts`), or
 *               a map of fileName to content, which *must* include `index.ts`.
 * @param options accepts a callback for historical reasons but really expects to
 *                take an options object.
 */
async function sourceToAssemblyHelper(source, options) {
    return (await compileJsiiForTest(source, options)).assembly;
}
exports.sourceToAssemblyHelper = sourceToAssemblyHelper;
/**
 * Compile a piece of source and return the assembly and compiled sources for it
 *
 * Only usable for trivial cases and tests.
 *
 * @param source can either be a single `string` (the content of `index.ts`), or
 *               a map of fileName to content, which *must* include `index.ts`.
 * @param options accepts a callback for historical reasons but really expects to
 *                take an options object.
 */
async function compileJsiiForTest(source, options, compilerOptions) {
    if (typeof source === 'string') {
        source = { 'index.ts': source };
    }
    const inSomeLocation = isOptionsObject(options) && options.compilationDirectory
        ? inOtherDir(options.compilationDirectory)
        : inTempDir;
    // Easiest way to get the source into the compiler is to write it to disk somewhere.
    // I guess we could make an in-memory compiler host but that seems like work...
    return inSomeLocation(async () => {
        var _a;
        await Promise.all(Object.entries(source).map(async ([fileName, content]) => {
            await fs.mkdirp(path.dirname(fileName));
            return fs.writeFile(fileName, content, { encoding: 'utf-8' });
        }));
        const { projectInfo, packageJson } = await makeProjectInfo('index.ts', typeof options === 'function'
            ? options
            : (pi) => {
                var _a, _b;
                Object.assign(pi, (_b = (_a = options === null || options === void 0 ? void 0 : options.packageJson) !== null && _a !== void 0 ? _a : options === null || options === void 0 ? void 0 : options.projectInfo) !== null && _b !== void 0 ? _b : {});
            });
        const compiler = new compiler_1.Compiler({
            projectInfo,
            ...compilerOptions,
        });
        const emitResult = await compiler.emit();
        const errors = emitResult.diagnostics.filter((d) => d.category === typescript_1.DiagnosticCategory.Error);
        for (const error of errors) {
            console.error(utils_1.formatDiagnostic(error, projectInfo.projectRoot));
            // logDiagnostic() doesn't work out of the box, so console.error() it is.
        }
        if (errors.length > 0 || emitResult.emitSkipped) {
            throw new Error('There were compiler errors');
        }
        const assembly = await fs.readJSON('.jsii', { encoding: 'utf-8' });
        const files = {};
        for (const filename of Object.keys(source)) {
            let jsFile = filename.replace(/\.ts$/, '.js');
            let dtsFile = filename.replace(/\.ts$/, '.d.ts');
            if (((_a = projectInfo.tsc) === null || _a === void 0 ? void 0 : _a.outDir) && filename !== 'README.md') {
                jsFile = path.join(projectInfo.tsc.outDir, jsFile);
                dtsFile = path.join(projectInfo.tsc.outDir, dtsFile);
            }
            // eslint-disable-next-line no-await-in-loop
            files[jsFile] = await fs.readFile(jsFile, { encoding: 'utf-8' });
            // eslint-disable-next-line no-await-in-loop
            files[dtsFile] = await fs.readFile(dtsFile, { encoding: 'utf-8' });
            const warningsFileName = '.warnings.jsii.js';
            if (fs.existsSync(warningsFileName)) {
                // eslint-disable-next-line no-await-in-loop
                files[warningsFileName] = await fs.readFile(warningsFileName, {
                    encoding: 'utf-8',
                });
            }
        }
        return { assembly, files, packageJson };
    });
}
exports.compileJsiiForTest = compileJsiiForTest;
async function inTempDir(block) {
    const origDir = process.cwd();
    const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'jsii'));
    process.chdir(tmpDir);
    const ret = await block();
    process.chdir(origDir);
    await fs.remove(tmpDir);
    return ret;
}
function inOtherDir(dir) {
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
    return async (block) => {
        const origDir = process.cwd();
        process.chdir(dir);
        try {
            return await block();
        }
        finally {
            process.chdir(origDir);
        }
    };
}
/**
 * Obtain project info so we can call the compiler
 *
 * Creating this directly in-memory leads to slightly different behavior from calling
 * jsii from the command-line, and I don't want to figure out right now.
 *
 * Most consistent behavior seems to be to write a package.json to disk and
 * then calling the same functions as the CLI would.
 */
async function makeProjectInfo(types, cb) {
    const packageJson = {
        types,
        main: types.replace(/(?:\.d)?\.ts(x?)/, '.js$1'),
        name: 'testpkg',
        version: '0.0.1',
        license: 'Apache-2.0',
        author: { name: 'John Doe' },
        repository: { type: 'git', url: 'https://github.com/aws/jsii.git' },
        jsii: {},
    };
    if (cb) {
        await cb(packageJson);
    }
    await fs.writeJson('package.json', packageJson, {
        encoding: 'utf-8',
        replacer: (_, v) => v,
        spaces: 2,
    });
    const { projectInfo } = await project_info_1.loadProjectInfo(path.resolve(process.cwd(), '.'));
    return { projectInfo, packageJson };
}
function isOptionsObject(x) {
    return x ? typeof x === 'object' : false;
}
/**
 * An NPM-ready workspace where we can install test-compile dependencies and compile new assemblies
 */
class TestWorkspace {
    constructor(rootDirectory) {
        this.rootDirectory = rootDirectory;
        this.installed = new Set();
    }
    /**
     * Create a new workspace.
     *
     * Creates a temporary directory, don't forget to call cleanUp
     */
    static async create() {
        const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'jsii-testworkspace'));
        await fs.ensureDir(tmpDir);
        return new TestWorkspace(tmpDir);
    }
    /**
     * Execute a block with a temporary workspace
     */
    static async withWorkspace(block) {
        const ws = await TestWorkspace.create();
        try {
            return await block(ws);
        }
        finally {
            await ws.cleanup();
        }
    }
    /**
     * Add a test-compiled jsii assembly as a dependency
     */
    async addDependency(dependencyAssembly) {
        if (this.installed.has(dependencyAssembly.assembly.name)) {
            throw new Error(`A dependency with name '${dependencyAssembly.assembly.name}' was already installed. Give one a different name.`);
        }
        this.installed.add(dependencyAssembly.assembly.name);
        // The following is silly, however: the helper has compiled the given source to
        // an assembly, and output files, and then removed their traces from disk.
        // We need those files back on disk, so write them back out again.
        //
        // We will drop them in 'node_modules/<name>' so they can be imported
        // as if they were installed.
        const modDir = path.join(this.rootDirectory, 'node_modules', dependencyAssembly.assembly.name);
        await fs.ensureDir(modDir);
        await fs.writeJSON(path.join(modDir, '.jsii'), dependencyAssembly.assembly);
        await fs.writeJSON(path.join(modDir, 'package.json'), dependencyAssembly.packageJson);
        for (const [fileName, fileContents] of Object.entries(dependencyAssembly.files)) {
            // eslint-disable-next-line no-await-in-loop
            await fs.ensureDir(path.dirname(path.join(modDir, fileName)));
            // eslint-disable-next-line no-await-in-loop
            await fs.writeFile(path.join(modDir, fileName), fileContents);
        }
    }
    dependencyDir(name) {
        if (!this.installed.has(name)) {
            throw new Error(`No dependency with name '${name}' has been installed`);
        }
        return path.join(this.rootDirectory, 'node_modules', name);
    }
    async cleanup() {
        await fs.remove(this.rootDirectory);
    }
}
exports.TestWorkspace = TestWorkspace;
