// @cdktf/provider-aws/lib/cloudfront/cloudfront-origin-request-policy.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudfrontOriginRequestPolicy = exports.CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference = exports.cloudfrontOriginRequestPolicyQueryStringsConfigToTerraform = exports.CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference = exports.cloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsToTerraform = exports.CloudfrontOriginRequestPolicyHeadersConfigOutputReference = exports.cloudfrontOriginRequestPolicyHeadersConfigToTerraform = exports.CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference = exports.cloudfrontOriginRequestPolicyHeadersConfigHeadersToTerraform = exports.CloudfrontOriginRequestPolicyCookiesConfigOutputReference = exports.cloudfrontOriginRequestPolicyCookiesConfigToTerraform = exports.CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference = exports.cloudfrontOriginRequestPolicyCookiesConfigCookiesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudfrontOriginRequestPolicyCookiesConfigCookiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        items: cdktf.listMapper(cdktf.stringToTerraform)(struct.items),
    };
}
exports.cloudfrontOriginRequestPolicyCookiesConfigCookiesToTerraform = cloudfrontOriginRequestPolicyCookiesConfigCookiesToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._items) {
            hasAnyValues = true;
            internalValueResult.items = this._items;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._items = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._items = value.items;
        }
    }
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
    /**
     * @stability stable
     */
    set items(value) {
        this._items = value;
    }
    /**
     * @stability stable
     */
    resetItems() {
        this._items = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get itemsInput() {
        return this._items;
    }
}
exports.CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference = CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference", version: "3.0.1" };
function cloudfrontOriginRequestPolicyCookiesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cookie_behavior: cdktf.stringToTerraform(struct.cookieBehavior),
        cookies: cloudfrontOriginRequestPolicyCookiesConfigCookiesToTerraform(struct.cookies),
    };
}
exports.cloudfrontOriginRequestPolicyCookiesConfigToTerraform = cloudfrontOriginRequestPolicyCookiesConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyCookiesConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cookies - computed: false, optional: true, required: false
        this._cookies = new CloudfrontOriginRequestPolicyCookiesConfigCookiesOutputReference(this, "cookies", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _h, _j;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cookieBehavior) {
            hasAnyValues = true;
            internalValueResult.cookieBehavior = this._cookieBehavior;
        }
        if ((_h = this._cookies) === null || _h === void 0 ? void 0 : _h.internalValue) {
            hasAnyValues = true;
            internalValueResult.cookies = (_j = this._cookies) === null || _j === void 0 ? void 0 : _j.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cookieBehavior = undefined;
            this._cookies.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cookieBehavior = value.cookieBehavior;
            this._cookies.internalValue = value.cookies;
        }
    }
    /**
     * @stability stable
     */
    get cookieBehavior() {
        return this.getStringAttribute('cookie_behavior');
    }
    /**
     * @stability stable
     */
    set cookieBehavior(value) {
        this._cookieBehavior = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookieBehaviorInput() {
        return this._cookieBehavior;
    }
    /**
     * @stability stable
     */
    get cookies() {
        return this._cookies;
    }
    /**
     * @stability stable
     */
    putCookies(value) {
        this._cookies.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCookies() {
        this._cookies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookiesInput() {
        return this._cookies.internalValue;
    }
}
exports.CloudfrontOriginRequestPolicyCookiesConfigOutputReference = CloudfrontOriginRequestPolicyCookiesConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyCookiesConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyCookiesConfigOutputReference", version: "3.0.1" };
function cloudfrontOriginRequestPolicyHeadersConfigHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        items: cdktf.listMapper(cdktf.stringToTerraform)(struct.items),
    };
}
exports.cloudfrontOriginRequestPolicyHeadersConfigHeadersToTerraform = cloudfrontOriginRequestPolicyHeadersConfigHeadersToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._items) {
            hasAnyValues = true;
            internalValueResult.items = this._items;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._items = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._items = value.items;
        }
    }
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
    /**
     * @stability stable
     */
    set items(value) {
        this._items = value;
    }
    /**
     * @stability stable
     */
    resetItems() {
        this._items = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get itemsInput() {
        return this._items;
    }
}
exports.CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference = CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference", version: "3.0.1" };
function cloudfrontOriginRequestPolicyHeadersConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        header_behavior: cdktf.stringToTerraform(struct.headerBehavior),
        headers: cloudfrontOriginRequestPolicyHeadersConfigHeadersToTerraform(struct.headers),
    };
}
exports.cloudfrontOriginRequestPolicyHeadersConfigToTerraform = cloudfrontOriginRequestPolicyHeadersConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyHeadersConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // headers - computed: false, optional: true, required: false
        this._headers = new CloudfrontOriginRequestPolicyHeadersConfigHeadersOutputReference(this, "headers", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _h, _j;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headerBehavior) {
            hasAnyValues = true;
            internalValueResult.headerBehavior = this._headerBehavior;
        }
        if ((_h = this._headers) === null || _h === void 0 ? void 0 : _h.internalValue) {
            hasAnyValues = true;
            internalValueResult.headers = (_j = this._headers) === null || _j === void 0 ? void 0 : _j.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._headerBehavior = undefined;
            this._headers.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._headerBehavior = value.headerBehavior;
            this._headers.internalValue = value.headers;
        }
    }
    /**
     * @stability stable
     */
    get headerBehavior() {
        return this.getStringAttribute('header_behavior');
    }
    /**
     * @stability stable
     */
    set headerBehavior(value) {
        this._headerBehavior = value;
    }
    /**
     * @stability stable
     */
    resetHeaderBehavior() {
        this._headerBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headerBehaviorInput() {
        return this._headerBehavior;
    }
    /**
     * @stability stable
     */
    get headers() {
        return this._headers;
    }
    /**
     * @stability stable
     */
    putHeaders(value) {
        this._headers.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetHeaders() {
        this._headers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headersInput() {
        return this._headers.internalValue;
    }
}
exports.CloudfrontOriginRequestPolicyHeadersConfigOutputReference = CloudfrontOriginRequestPolicyHeadersConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyHeadersConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyHeadersConfigOutputReference", version: "3.0.1" };
function cloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        items: cdktf.listMapper(cdktf.stringToTerraform)(struct.items),
    };
}
exports.cloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsToTerraform = cloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._items) {
            hasAnyValues = true;
            internalValueResult.items = this._items;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._items = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._items = value.items;
        }
    }
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
    /**
     * @stability stable
     */
    set items(value) {
        this._items = value;
    }
    /**
     * @stability stable
     */
    resetItems() {
        this._items = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get itemsInput() {
        return this._items;
    }
}
exports.CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference = CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference", version: "3.0.1" };
function cloudfrontOriginRequestPolicyQueryStringsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        query_string_behavior: cdktf.stringToTerraform(struct.queryStringBehavior),
        query_strings: cloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsToTerraform(struct.queryStrings),
    };
}
exports.cloudfrontOriginRequestPolicyQueryStringsConfigToTerraform = cloudfrontOriginRequestPolicyQueryStringsConfigToTerraform;
/**
 * @stability stable
 */
class CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // query_strings - computed: false, optional: true, required: false
        this._queryStrings = new CloudfrontOriginRequestPolicyQueryStringsConfigQueryStringsOutputReference(this, "query_strings", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _h, _j;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._queryStringBehavior) {
            hasAnyValues = true;
            internalValueResult.queryStringBehavior = this._queryStringBehavior;
        }
        if ((_h = this._queryStrings) === null || _h === void 0 ? void 0 : _h.internalValue) {
            hasAnyValues = true;
            internalValueResult.queryStrings = (_j = this._queryStrings) === null || _j === void 0 ? void 0 : _j.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._queryStringBehavior = undefined;
            this._queryStrings.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._queryStringBehavior = value.queryStringBehavior;
            this._queryStrings.internalValue = value.queryStrings;
        }
    }
    /**
     * @stability stable
     */
    get queryStringBehavior() {
        return this.getStringAttribute('query_string_behavior');
    }
    /**
     * @stability stable
     */
    set queryStringBehavior(value) {
        this._queryStringBehavior = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringBehaviorInput() {
        return this._queryStringBehavior;
    }
    /**
     * @stability stable
     */
    get queryStrings() {
        return this._queryStrings;
    }
    /**
     * @stability stable
     */
    putQueryStrings(value) {
        this._queryStrings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetQueryStrings() {
        this._queryStrings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringsInput() {
        return this._queryStrings.internalValue;
    }
}
exports.CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference = CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudfront_origin_request_policy.html aws_cloudfront_origin_request_policy}.
 *
 * @stability stable
 */
class CloudfrontOriginRequestPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudfront_origin_request_policy.html aws_cloudfront_origin_request_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_origin_request_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // cookies_config - computed: false, optional: false, required: true
        this._cookiesConfig = new CloudfrontOriginRequestPolicyCookiesConfigOutputReference(this, "cookies_config", true);
        // headers_config - computed: false, optional: false, required: true
        this._headersConfig = new CloudfrontOriginRequestPolicyHeadersConfigOutputReference(this, "headers_config", true);
        // query_strings_config - computed: false, optional: false, required: true
        this._queryStringsConfig = new CloudfrontOriginRequestPolicyQueryStringsConfigOutputReference(this, "query_strings_config", true);
        this._comment = config.comment;
        this._name = config.name;
        this._cookiesConfig.internalValue = config.cookiesConfig;
        this._headersConfig.internalValue = config.headersConfig;
        this._queryStringsConfig.internalValue = config.queryStringsConfig;
    }
    /**
     * @stability stable
     */
    get comment() {
        return this.getStringAttribute('comment');
    }
    /**
     * @stability stable
     */
    set comment(value) {
        this._comment = value;
    }
    /**
     * @stability stable
     */
    resetComment() {
        this._comment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commentInput() {
        return this._comment;
    }
    // etag - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get etag() {
        return this.getStringAttribute('etag');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get cookiesConfig() {
        return this._cookiesConfig;
    }
    /**
     * @stability stable
     */
    putCookiesConfig(value) {
        this._cookiesConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookiesConfigInput() {
        return this._cookiesConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get headersConfig() {
        return this._headersConfig;
    }
    /**
     * @stability stable
     */
    putHeadersConfig(value) {
        this._headersConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headersConfigInput() {
        return this._headersConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get queryStringsConfig() {
        return this._queryStringsConfig;
    }
    /**
     * @stability stable
     */
    putQueryStringsConfig(value) {
        this._queryStringsConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringsConfigInput() {
        return this._queryStringsConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            comment: cdktf.stringToTerraform(this._comment),
            name: cdktf.stringToTerraform(this._name),
            cookies_config: cloudfrontOriginRequestPolicyCookiesConfigToTerraform(this._cookiesConfig.internalValue),
            headers_config: cloudfrontOriginRequestPolicyHeadersConfigToTerraform(this._headersConfig.internalValue),
            query_strings_config: cloudfrontOriginRequestPolicyQueryStringsConfigToTerraform(this._queryStringsConfig.internalValue),
        };
    }
}
exports.CloudfrontOriginRequestPolicy = CloudfrontOriginRequestPolicy;
_g = JSII_RTTI_SYMBOL_1;
CloudfrontOriginRequestPolicy[_g] = { fqn: "@cdktf/provider-aws.cloudfront.CloudfrontOriginRequestPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudfrontOriginRequestPolicy.tfResourceType = "aws_cloudfront_origin_request_policy";
