// @cdktf/provider-aws/lib/cloudfront/data-aws-cloudfront-response-headers-policy.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudfrontResponseHeadersPolicy = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions = exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy = exports.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig = exports.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems = exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfig = exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders = exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins = exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods = exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders extends cdktf.ComplexComputedList {
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders = DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders[_a] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowHeaders", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods extends cdktf.ComplexComputedList {
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods = DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods[_b] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowMethods", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins extends cdktf.ComplexComputedList {
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins = DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins[_c] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlAllowOrigins", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders extends cdktf.ComplexComputedList {
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        return this.getListAttribute('items');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders = DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders;
_d = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders[_d] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCorsConfigAccessControlExposeHeaders", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCorsConfig extends cdktf.ComplexComputedList {
    // access_control_allow_credentials - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlAllowCredentials() {
        return this.getBooleanAttribute('access_control_allow_credentials');
    }
    // access_control_allow_headers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlAllowHeaders() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('access_control_allow_headers');
    }
    // access_control_allow_methods - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlAllowMethods() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('access_control_allow_methods');
    }
    // access_control_allow_origins - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlAllowOrigins() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('access_control_allow_origins');
    }
    // access_control_expose_headers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlExposeHeaders() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('access_control_expose_headers');
    }
    // access_control_max_age_sec - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlMaxAgeSec() {
        return this.getNumberAttribute('access_control_max_age_sec');
    }
    // origin_override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get originOverride() {
        return this.getBooleanAttribute('origin_override');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCorsConfig = DataAwsCloudfrontResponseHeadersPolicyCorsConfig;
_e = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCorsConfig[_e] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCorsConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems extends cdktf.ComplexComputedList {
    // header - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get header() {
        return this.getStringAttribute('header');
    }
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
    // value - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems = DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems;
_f = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems[_f] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfigItems", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig extends cdktf.ComplexComputedList {
    // items - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get items() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('items');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig = DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig;
_g = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig[_g] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy extends cdktf.ComplexComputedList {
    // content_security_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get contentSecurityPolicy() {
        return this.getStringAttribute('content_security_policy');
    }
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy;
_h = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy[_h] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentSecurityPolicy", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions extends cdktf.ComplexComputedList {
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions;
_j = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions[_j] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigContentTypeOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions extends cdktf.ComplexComputedList {
    // frame_option - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get frameOption() {
        return this.getStringAttribute('frame_option');
    }
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions;
_k = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions[_k] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigFrameOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy extends cdktf.ComplexComputedList {
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
    // referrer_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get referrerPolicy() {
        return this.getStringAttribute('referrer_policy');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy;
_l = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy[_l] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigReferrerPolicy", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity extends cdktf.ComplexComputedList {
    // access_control_max_age_sec - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessControlMaxAgeSec() {
        return this.getNumberAttribute('access_control_max_age_sec');
    }
    // include_subdomains - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get includeSubdomains() {
        return this.getBooleanAttribute('include_subdomains');
    }
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
    // preload - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get preload() {
        return this.getBooleanAttribute('preload');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity;
_m = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity[_m] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigStrictTransportSecurity", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection extends cdktf.ComplexComputedList {
    // mode_block - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get modeBlock() {
        return this.getBooleanAttribute('mode_block');
    }
    // override - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get override() {
        return this.getBooleanAttribute('override');
    }
    // protection - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get protection() {
        return this.getBooleanAttribute('protection');
    }
    // report_uri - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get reportUri() {
        return this.getStringAttribute('report_uri');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection;
_o = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection[_o] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfigXssProtection", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig extends cdktf.ComplexComputedList {
    // content_security_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get contentSecurityPolicy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('content_security_policy');
    }
    // content_type_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get contentTypeOptions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('content_type_options');
    }
    // frame_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get frameOptions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('frame_options');
    }
    // referrer_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get referrerPolicy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('referrer_policy');
    }
    // strict_transport_security - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get strictTransportSecurity() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('strict_transport_security');
    }
    // xss_protection - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get xssProtection() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('xss_protection');
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig = DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig;
_p = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig[_p] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_response_headers_policy.html aws_cloudfront_response_headers_policy}.
 *
 * @stability stable
 */
class DataAwsCloudfrontResponseHeadersPolicy extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_response_headers_policy.html aws_cloudfront_response_headers_policy} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_response_headers_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // comment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get comment() {
        return this.getStringAttribute('comment');
    }
    // cors_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    corsConfig(index) {
        return new DataAwsCloudfrontResponseHeadersPolicyCorsConfig(this, 'cors_config', index);
    }
    // custom_headers_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    customHeadersConfig(index) {
        return new DataAwsCloudfrontResponseHeadersPolicyCustomHeadersConfig(this, 'custom_headers_config', index);
    }
    // etag - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get etag() {
        return this.getStringAttribute('etag');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // security_headers_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    securityHeadersConfig(index) {
        return new DataAwsCloudfrontResponseHeadersPolicySecurityHeadersConfig(this, 'security_headers_config', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsCloudfrontResponseHeadersPolicy = DataAwsCloudfrontResponseHeadersPolicy;
_q = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontResponseHeadersPolicy[_q] = { fqn: "@cdktf/provider-aws.cloudfront.DataAwsCloudfrontResponseHeadersPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCloudfrontResponseHeadersPolicy.tfResourceType = "aws_cloudfront_response_headers_policy";
