// @cdktf/provider-aws/lib/glue/glue-trigger.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueTrigger = exports.GlueTriggerTimeoutsOutputReference = exports.glueTriggerTimeoutsToTerraform = exports.GlueTriggerPredicateOutputReference = exports.glueTriggerPredicateToTerraform = exports.glueTriggerPredicateConditionsToTerraform = exports.glueTriggerActionsToTerraform = exports.GlueTriggerActionsNotificationPropertyOutputReference = exports.glueTriggerActionsNotificationPropertyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueTriggerActionsNotificationPropertyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        notify_delay_after: cdktf.numberToTerraform(struct.notifyDelayAfter),
    };
}
exports.glueTriggerActionsNotificationPropertyToTerraform = glueTriggerActionsNotificationPropertyToTerraform;
/**
 * @stability stable
 */
class GlueTriggerActionsNotificationPropertyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._notifyDelayAfter) {
            hasAnyValues = true;
            internalValueResult.notifyDelayAfter = this._notifyDelayAfter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._notifyDelayAfter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._notifyDelayAfter = value.notifyDelayAfter;
        }
    }
    /**
     * @stability stable
     */
    get notifyDelayAfter() {
        return this.getNumberAttribute('notify_delay_after');
    }
    /**
     * @stability stable
     */
    set notifyDelayAfter(value) {
        this._notifyDelayAfter = value;
    }
    /**
     * @stability stable
     */
    resetNotifyDelayAfter() {
        this._notifyDelayAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notifyDelayAfterInput() {
        return this._notifyDelayAfter;
    }
}
exports.GlueTriggerActionsNotificationPropertyOutputReference = GlueTriggerActionsNotificationPropertyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueTriggerActionsNotificationPropertyOutputReference[_a] = { fqn: "@cdktf/provider-aws.glue.GlueTriggerActionsNotificationPropertyOutputReference", version: "3.0.1" };
function glueTriggerActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arguments: cdktf.hashMapper(cdktf.anyToTerraform)(struct.arguments),
        crawler_name: cdktf.stringToTerraform(struct.crawlerName),
        job_name: cdktf.stringToTerraform(struct.jobName),
        security_configuration: cdktf.stringToTerraform(struct.securityConfiguration),
        timeout: cdktf.numberToTerraform(struct.timeout),
        notification_property: glueTriggerActionsNotificationPropertyToTerraform(struct.notificationProperty),
    };
}
exports.glueTriggerActionsToTerraform = glueTriggerActionsToTerraform;
function glueTriggerPredicateConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crawl_state: cdktf.stringToTerraform(struct.crawlState),
        crawler_name: cdktf.stringToTerraform(struct.crawlerName),
        job_name: cdktf.stringToTerraform(struct.jobName),
        logical_operator: cdktf.stringToTerraform(struct.logicalOperator),
        state: cdktf.stringToTerraform(struct.state),
    };
}
exports.glueTriggerPredicateConditionsToTerraform = glueTriggerPredicateConditionsToTerraform;
function glueTriggerPredicateToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        logical: cdktf.stringToTerraform(struct.logical),
        conditions: cdktf.listMapper(glueTriggerPredicateConditionsToTerraform)(struct.conditions),
    };
}
exports.glueTriggerPredicateToTerraform = glueTriggerPredicateToTerraform;
/**
 * @stability stable
 */
class GlueTriggerPredicateOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logical) {
            hasAnyValues = true;
            internalValueResult.logical = this._logical;
        }
        if (this._conditions) {
            hasAnyValues = true;
            internalValueResult.conditions = this._conditions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logical = undefined;
            this._conditions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logical = value.logical;
            this._conditions = value.conditions;
        }
    }
    /**
     * @stability stable
     */
    get logical() {
        return this.getStringAttribute('logical');
    }
    /**
     * @stability stable
     */
    set logical(value) {
        this._logical = value;
    }
    /**
     * @stability stable
     */
    resetLogical() {
        this._logical = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logicalInput() {
        return this._logical;
    }
    /**
     * @stability stable
     */
    get conditions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('conditions');
    }
    /**
     * @stability stable
     */
    set conditions(value) {
        this._conditions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get conditionsInput() {
        return this._conditions;
    }
}
exports.GlueTriggerPredicateOutputReference = GlueTriggerPredicateOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlueTriggerPredicateOutputReference[_b] = { fqn: "@cdktf/provider-aws.glue.GlueTriggerPredicateOutputReference", version: "3.0.1" };
function glueTriggerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.glueTriggerTimeoutsToTerraform = glueTriggerTimeoutsToTerraform;
/**
 * @stability stable
 */
class GlueTriggerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.GlueTriggerTimeoutsOutputReference = GlueTriggerTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlueTriggerTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.glue.GlueTriggerTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_trigger.html aws_glue_trigger}.
 *
 * @stability stable
 */
class GlueTrigger extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_trigger.html aws_glue_trigger} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_trigger',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // predicate - computed: false, optional: true, required: false
        this._predicate = new GlueTriggerPredicateOutputReference(this, "predicate", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GlueTriggerTimeoutsOutputReference(this, "timeouts", true);
        this._description = config.description;
        this._enabled = config.enabled;
        this._name = config.name;
        this._schedule = config.schedule;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._workflowName = config.workflowName;
        this._actions = config.actions;
        this._predicate.internalValue = config.predicate;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    /**
     * @stability stable
     */
    set schedule(value) {
        this._schedule = value;
    }
    /**
     * @stability stable
     */
    resetSchedule() {
        this._schedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleInput() {
        return this._schedule;
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get workflowName() {
        return this.getStringAttribute('workflow_name');
    }
    /**
     * @stability stable
     */
    set workflowName(value) {
        this._workflowName = value;
    }
    /**
     * @stability stable
     */
    resetWorkflowName() {
        this._workflowName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workflowNameInput() {
        return this._workflowName;
    }
    /**
     * @stability stable
     */
    get actions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('actions');
    }
    /**
     * @stability stable
     */
    set actions(value) {
        this._actions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionsInput() {
        return this._actions;
    }
    /**
     * @stability stable
     */
    get predicate() {
        return this._predicate;
    }
    /**
     * @stability stable
     */
    putPredicate(value) {
        this._predicate.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPredicate() {
        this._predicate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get predicateInput() {
        return this._predicate.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            enabled: cdktf.booleanToTerraform(this._enabled),
            name: cdktf.stringToTerraform(this._name),
            schedule: cdktf.stringToTerraform(this._schedule),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            workflow_name: cdktf.stringToTerraform(this._workflowName),
            actions: cdktf.listMapper(glueTriggerActionsToTerraform)(this._actions),
            predicate: glueTriggerPredicateToTerraform(this._predicate.internalValue),
            timeouts: glueTriggerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GlueTrigger = GlueTrigger;
_d = JSII_RTTI_SYMBOL_1;
GlueTrigger[_d] = { fqn: "@cdktf/provider-aws.glue.GlueTrigger", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueTrigger.tfResourceType = "aws_glue_trigger";
