// @cdktf/provider-aws/lib/glue/glue-job.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueJob = exports.GlueJobNotificationPropertyOutputReference = exports.glueJobNotificationPropertyToTerraform = exports.GlueJobExecutionPropertyOutputReference = exports.glueJobExecutionPropertyToTerraform = exports.GlueJobCommandOutputReference = exports.glueJobCommandToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueJobCommandToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        python_version: cdktf.stringToTerraform(struct.pythonVersion),
        script_location: cdktf.stringToTerraform(struct.scriptLocation),
    };
}
exports.glueJobCommandToTerraform = glueJobCommandToTerraform;
/**
 * @stability stable
 */
class GlueJobCommandOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._pythonVersion) {
            hasAnyValues = true;
            internalValueResult.pythonVersion = this._pythonVersion;
        }
        if (this._scriptLocation) {
            hasAnyValues = true;
            internalValueResult.scriptLocation = this._scriptLocation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._pythonVersion = undefined;
            this._scriptLocation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._pythonVersion = value.pythonVersion;
            this._scriptLocation = value.scriptLocation;
        }
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get pythonVersion() {
        return this.getStringAttribute('python_version');
    }
    /**
     * @stability stable
     */
    set pythonVersion(value) {
        this._pythonVersion = value;
    }
    /**
     * @stability stable
     */
    resetPythonVersion() {
        this._pythonVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pythonVersionInput() {
        return this._pythonVersion;
    }
    /**
     * @stability stable
     */
    get scriptLocation() {
        return this.getStringAttribute('script_location');
    }
    /**
     * @stability stable
     */
    set scriptLocation(value) {
        this._scriptLocation = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scriptLocationInput() {
        return this._scriptLocation;
    }
}
exports.GlueJobCommandOutputReference = GlueJobCommandOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueJobCommandOutputReference[_a] = { fqn: "@cdktf/provider-aws.glue.GlueJobCommandOutputReference", version: "3.0.1" };
function glueJobExecutionPropertyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_concurrent_runs: cdktf.numberToTerraform(struct.maxConcurrentRuns),
    };
}
exports.glueJobExecutionPropertyToTerraform = glueJobExecutionPropertyToTerraform;
/**
 * @stability stable
 */
class GlueJobExecutionPropertyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConcurrentRuns) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentRuns = this._maxConcurrentRuns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConcurrentRuns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConcurrentRuns = value.maxConcurrentRuns;
        }
    }
    /**
     * @stability stable
     */
    get maxConcurrentRuns() {
        return this.getNumberAttribute('max_concurrent_runs');
    }
    /**
     * @stability stable
     */
    set maxConcurrentRuns(value) {
        this._maxConcurrentRuns = value;
    }
    /**
     * @stability stable
     */
    resetMaxConcurrentRuns() {
        this._maxConcurrentRuns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrentRunsInput() {
        return this._maxConcurrentRuns;
    }
}
exports.GlueJobExecutionPropertyOutputReference = GlueJobExecutionPropertyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlueJobExecutionPropertyOutputReference[_b] = { fqn: "@cdktf/provider-aws.glue.GlueJobExecutionPropertyOutputReference", version: "3.0.1" };
function glueJobNotificationPropertyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        notify_delay_after: cdktf.numberToTerraform(struct.notifyDelayAfter),
    };
}
exports.glueJobNotificationPropertyToTerraform = glueJobNotificationPropertyToTerraform;
/**
 * @stability stable
 */
class GlueJobNotificationPropertyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._notifyDelayAfter) {
            hasAnyValues = true;
            internalValueResult.notifyDelayAfter = this._notifyDelayAfter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._notifyDelayAfter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._notifyDelayAfter = value.notifyDelayAfter;
        }
    }
    /**
     * @stability stable
     */
    get notifyDelayAfter() {
        return this.getNumberAttribute('notify_delay_after');
    }
    /**
     * @stability stable
     */
    set notifyDelayAfter(value) {
        this._notifyDelayAfter = value;
    }
    /**
     * @stability stable
     */
    resetNotifyDelayAfter() {
        this._notifyDelayAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notifyDelayAfterInput() {
        return this._notifyDelayAfter;
    }
}
exports.GlueJobNotificationPropertyOutputReference = GlueJobNotificationPropertyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlueJobNotificationPropertyOutputReference[_c] = { fqn: "@cdktf/provider-aws.glue.GlueJobNotificationPropertyOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_job.html aws_glue_job}.
 *
 * @stability stable
 */
class GlueJob extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_job.html aws_glue_job} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_job',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // command - computed: false, optional: false, required: true
        this._command = new GlueJobCommandOutputReference(this, "command", true);
        // execution_property - computed: false, optional: true, required: false
        this._executionProperty = new GlueJobExecutionPropertyOutputReference(this, "execution_property", true);
        // notification_property - computed: false, optional: true, required: false
        this._notificationProperty = new GlueJobNotificationPropertyOutputReference(this, "notification_property", true);
        this._connections = config.connections;
        this._defaultArguments = config.defaultArguments;
        this._description = config.description;
        this._glueVersion = config.glueVersion;
        this._maxCapacity = config.maxCapacity;
        this._maxRetries = config.maxRetries;
        this._name = config.name;
        this._nonOverridableArguments = config.nonOverridableArguments;
        this._numberOfWorkers = config.numberOfWorkers;
        this._roleArn = config.roleArn;
        this._securityConfiguration = config.securityConfiguration;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeout = config.timeout;
        this._workerType = config.workerType;
        this._command.internalValue = config.command;
        this._executionProperty.internalValue = config.executionProperty;
        this._notificationProperty.internalValue = config.notificationProperty;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get connections() {
        return this.getListAttribute('connections');
    }
    /**
     * @stability stable
     */
    set connections(value) {
        this._connections = value;
    }
    /**
     * @stability stable
     */
    resetConnections() {
        this._connections = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionsInput() {
        return this._connections;
    }
    /**
     * @stability stable
     */
    get defaultArguments() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('default_arguments');
    }
    /**
     * @stability stable
     */
    set defaultArguments(value) {
        this._defaultArguments = value;
    }
    /**
     * @stability stable
     */
    resetDefaultArguments() {
        this._defaultArguments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultArgumentsInput() {
        return this._defaultArguments;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get glueVersion() {
        return this.getStringAttribute('glue_version');
    }
    /**
     * @stability stable
     */
    set glueVersion(value) {
        this._glueVersion = value;
    }
    /**
     * @stability stable
     */
    resetGlueVersion() {
        this._glueVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get glueVersionInput() {
        return this._glueVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get maxCapacity() {
        return this.getNumberAttribute('max_capacity');
    }
    /**
     * @stability stable
     */
    set maxCapacity(value) {
        this._maxCapacity = value;
    }
    /**
     * @stability stable
     */
    resetMaxCapacity() {
        this._maxCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxCapacityInput() {
        return this._maxCapacity;
    }
    /**
     * @stability stable
     */
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    /**
     * @stability stable
     */
    set maxRetries(value) {
        this._maxRetries = value;
    }
    /**
     * @stability stable
     */
    resetMaxRetries() {
        this._maxRetries = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxRetriesInput() {
        return this._maxRetries;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get nonOverridableArguments() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('non_overridable_arguments');
    }
    /**
     * @stability stable
     */
    set nonOverridableArguments(value) {
        this._nonOverridableArguments = value;
    }
    /**
     * @stability stable
     */
    resetNonOverridableArguments() {
        this._nonOverridableArguments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nonOverridableArgumentsInput() {
        return this._nonOverridableArguments;
    }
    /**
     * @stability stable
     */
    get numberOfWorkers() {
        return this.getNumberAttribute('number_of_workers');
    }
    /**
     * @stability stable
     */
    set numberOfWorkers(value) {
        this._numberOfWorkers = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfWorkers() {
        this._numberOfWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfWorkersInput() {
        return this._numberOfWorkers;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    /**
     * @stability stable
     */
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    /**
     * @stability stable
     */
    set timeout(value) {
        this._timeout = value;
    }
    /**
     * @stability stable
     */
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutInput() {
        return this._timeout;
    }
    /**
     * @stability stable
     */
    get workerType() {
        return this.getStringAttribute('worker_type');
    }
    /**
     * @stability stable
     */
    set workerType(value) {
        this._workerType = value;
    }
    /**
     * @stability stable
     */
    resetWorkerType() {
        this._workerType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workerTypeInput() {
        return this._workerType;
    }
    /**
     * @stability stable
     */
    get command() {
        return this._command;
    }
    /**
     * @stability stable
     */
    putCommand(value) {
        this._command.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commandInput() {
        return this._command.internalValue;
    }
    /**
     * @stability stable
     */
    get executionProperty() {
        return this._executionProperty;
    }
    /**
     * @stability stable
     */
    putExecutionProperty(value) {
        this._executionProperty.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetExecutionProperty() {
        this._executionProperty.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionPropertyInput() {
        return this._executionProperty.internalValue;
    }
    /**
     * @stability stable
     */
    get notificationProperty() {
        return this._notificationProperty;
    }
    /**
     * @stability stable
     */
    putNotificationProperty(value) {
        this._notificationProperty.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNotificationProperty() {
        this._notificationProperty.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationPropertyInput() {
        return this._notificationProperty.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            connections: cdktf.listMapper(cdktf.stringToTerraform)(this._connections),
            default_arguments: cdktf.hashMapper(cdktf.anyToTerraform)(this._defaultArguments),
            description: cdktf.stringToTerraform(this._description),
            glue_version: cdktf.stringToTerraform(this._glueVersion),
            max_capacity: cdktf.numberToTerraform(this._maxCapacity),
            max_retries: cdktf.numberToTerraform(this._maxRetries),
            name: cdktf.stringToTerraform(this._name),
            non_overridable_arguments: cdktf.hashMapper(cdktf.anyToTerraform)(this._nonOverridableArguments),
            number_of_workers: cdktf.numberToTerraform(this._numberOfWorkers),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            timeout: cdktf.numberToTerraform(this._timeout),
            worker_type: cdktf.stringToTerraform(this._workerType),
            command: glueJobCommandToTerraform(this._command.internalValue),
            execution_property: glueJobExecutionPropertyToTerraform(this._executionProperty.internalValue),
            notification_property: glueJobNotificationPropertyToTerraform(this._notificationProperty.internalValue),
        };
    }
}
exports.GlueJob = GlueJob;
_d = JSII_RTTI_SYMBOL_1;
GlueJob[_d] = { fqn: "@cdktf/provider-aws.glue.GlueJob", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueJob.tfResourceType = "aws_glue_job";
