// @cdktf/provider-aws/lib/glue/glue-resource-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueResourcePolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_resource_policy.html aws_glue_resource_policy}.
 *
 * @stability stable
 */
class GlueResourcePolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_resource_policy.html aws_glue_resource_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_resource_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._enableHybrid = config.enableHybrid;
        this._policy = config.policy;
    }
    /**
     * @stability stable
     */
    get enableHybrid() {
        return this.getStringAttribute('enable_hybrid');
    }
    /**
     * @stability stable
     */
    set enableHybrid(value) {
        this._enableHybrid = value;
    }
    /**
     * @stability stable
     */
    resetEnableHybrid() {
        this._enableHybrid = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableHybridInput() {
        return this._enableHybrid;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    /**
     * @stability stable
     */
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyInput() {
        return this._policy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enable_hybrid: cdktf.stringToTerraform(this._enableHybrid),
            policy: cdktf.stringToTerraform(this._policy),
        };
    }
}
exports.GlueResourcePolicy = GlueResourcePolicy;
_a = JSII_RTTI_SYMBOL_1;
GlueResourcePolicy[_a] = { fqn: "@cdktf/provider-aws.glue.GlueResourcePolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueResourcePolicy.tfResourceType = "aws_glue_resource_policy";
