// @cdktf/provider-aws/lib/glue/glue-dev-endpoint.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueDevEndpoint = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_dev_endpoint.html aws_glue_dev_endpoint}.
 *
 * @stability stable
 */
class GlueDevEndpoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_dev_endpoint.html aws_glue_dev_endpoint} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_dev_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arguments = config.arguments;
        this._extraJarsS3Path = config.extraJarsS3Path;
        this._extraPythonLibsS3Path = config.extraPythonLibsS3Path;
        this._glueVersion = config.glueVersion;
        this._name = config.name;
        this._numberOfNodes = config.numberOfNodes;
        this._numberOfWorkers = config.numberOfWorkers;
        this._publicKey = config.publicKey;
        this._publicKeys = config.publicKeys;
        this._roleArn = config.roleArn;
        this._securityConfiguration = config.securityConfiguration;
        this._securityGroupIds = config.securityGroupIds;
        this._subnetId = config.subnetId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._workerType = config.workerType;
    }
    /**
     * @stability stable
     */
    get arguments() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('arguments');
    }
    /**
     * @stability stable
     */
    set arguments(value) {
        this._arguments = value;
    }
    /**
     * @stability stable
     */
    resetArguments() {
        this._arguments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get argumentsInput() {
        return this._arguments;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    /**
     * @stability stable
     */
    get extraJarsS3Path() {
        return this.getStringAttribute('extra_jars_s3_path');
    }
    /**
     * @stability stable
     */
    set extraJarsS3Path(value) {
        this._extraJarsS3Path = value;
    }
    /**
     * @stability stable
     */
    resetExtraJarsS3Path() {
        this._extraJarsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get extraJarsS3PathInput() {
        return this._extraJarsS3Path;
    }
    /**
     * @stability stable
     */
    get extraPythonLibsS3Path() {
        return this.getStringAttribute('extra_python_libs_s3_path');
    }
    /**
     * @stability stable
     */
    set extraPythonLibsS3Path(value) {
        this._extraPythonLibsS3Path = value;
    }
    /**
     * @stability stable
     */
    resetExtraPythonLibsS3Path() {
        this._extraPythonLibsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get extraPythonLibsS3PathInput() {
        return this._extraPythonLibsS3Path;
    }
    // failure_reason - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get failureReason() {
        return this.getStringAttribute('failure_reason');
    }
    /**
     * @stability stable
     */
    get glueVersion() {
        return this.getStringAttribute('glue_version');
    }
    /**
     * @stability stable
     */
    set glueVersion(value) {
        this._glueVersion = value;
    }
    /**
     * @stability stable
     */
    resetGlueVersion() {
        this._glueVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get glueVersionInput() {
        return this._glueVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get numberOfNodes() {
        return this.getNumberAttribute('number_of_nodes');
    }
    /**
     * @stability stable
     */
    set numberOfNodes(value) {
        this._numberOfNodes = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfNodes() {
        this._numberOfNodes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfNodesInput() {
        return this._numberOfNodes;
    }
    /**
     * @stability stable
     */
    get numberOfWorkers() {
        return this.getNumberAttribute('number_of_workers');
    }
    /**
     * @stability stable
     */
    set numberOfWorkers(value) {
        this._numberOfWorkers = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfWorkers() {
        this._numberOfWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfWorkersInput() {
        return this._numberOfWorkers;
    }
    // private_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateAddress() {
        return this.getStringAttribute('private_address');
    }
    // public_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicAddress() {
        return this.getStringAttribute('public_address');
    }
    /**
     * @stability stable
     */
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    /**
     * @stability stable
     */
    set publicKey(value) {
        this._publicKey = value;
    }
    /**
     * @stability stable
     */
    resetPublicKey() {
        this._publicKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicKeyInput() {
        return this._publicKey;
    }
    /**
     * @stability stable
     */
    get publicKeys() {
        return this.getListAttribute('public_keys');
    }
    /**
     * @stability stable
     */
    set publicKeys(value) {
        this._publicKeys = value;
    }
    /**
     * @stability stable
     */
    resetPublicKeys() {
        this._publicKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicKeysInput() {
        return this._publicKeys;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    /**
     * @stability stable
     */
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    /**
     * @stability stable
     */
    set subnetId(value) {
        this._subnetId = value;
    }
    /**
     * @stability stable
     */
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdInput() {
        return this._subnetId;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    get workerType() {
        return this.getStringAttribute('worker_type');
    }
    /**
     * @stability stable
     */
    set workerType(value) {
        this._workerType = value;
    }
    /**
     * @stability stable
     */
    resetWorkerType() {
        this._workerType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workerTypeInput() {
        return this._workerType;
    }
    // yarn_endpoint_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get yarnEndpointAddress() {
        return this.getStringAttribute('yarn_endpoint_address');
    }
    // zeppelin_remote_spark_interpreter_port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zeppelinRemoteSparkInterpreterPort() {
        return this.getNumberAttribute('zeppelin_remote_spark_interpreter_port');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arguments: cdktf.hashMapper(cdktf.anyToTerraform)(this._arguments),
            extra_jars_s3_path: cdktf.stringToTerraform(this._extraJarsS3Path),
            extra_python_libs_s3_path: cdktf.stringToTerraform(this._extraPythonLibsS3Path),
            glue_version: cdktf.stringToTerraform(this._glueVersion),
            name: cdktf.stringToTerraform(this._name),
            number_of_nodes: cdktf.numberToTerraform(this._numberOfNodes),
            number_of_workers: cdktf.numberToTerraform(this._numberOfWorkers),
            public_key: cdktf.stringToTerraform(this._publicKey),
            public_keys: cdktf.listMapper(cdktf.stringToTerraform)(this._publicKeys),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroupIds),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            worker_type: cdktf.stringToTerraform(this._workerType),
        };
    }
}
exports.GlueDevEndpoint = GlueDevEndpoint;
_a = JSII_RTTI_SYMBOL_1;
GlueDevEndpoint[_a] = { fqn: "@cdktf/provider-aws.glue.GlueDevEndpoint", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueDevEndpoint.tfResourceType = "aws_glue_dev_endpoint";
