// @cdktf/provider-aws/lib/glue/glue-ml-transform.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueMlTransform = exports.GlueMlTransformParametersOutputReference = exports.glueMlTransformParametersToTerraform = exports.GlueMlTransformParametersFindMatchesParametersOutputReference = exports.glueMlTransformParametersFindMatchesParametersToTerraform = exports.glueMlTransformInputRecordTablesToTerraform = exports.GlueMlTransformSchema = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class GlueMlTransformSchema extends cdktf.ComplexComputedList {
    // data_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dataType() {
        return this.getStringAttribute('data_type');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.GlueMlTransformSchema = GlueMlTransformSchema;
_a = JSII_RTTI_SYMBOL_1;
GlueMlTransformSchema[_a] = { fqn: "@cdktf/provider-aws.glue.GlueMlTransformSchema", version: "3.0.1" };
function glueMlTransformInputRecordTablesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.glueMlTransformInputRecordTablesToTerraform = glueMlTransformInputRecordTablesToTerraform;
function glueMlTransformParametersFindMatchesParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accuracy_cost_trade_off: cdktf.numberToTerraform(struct.accuracyCostTradeOff),
        enforce_provided_labels: cdktf.booleanToTerraform(struct.enforceProvidedLabels),
        precision_recall_trade_off: cdktf.numberToTerraform(struct.precisionRecallTradeOff),
        primary_key_column_name: cdktf.stringToTerraform(struct.primaryKeyColumnName),
    };
}
exports.glueMlTransformParametersFindMatchesParametersToTerraform = glueMlTransformParametersFindMatchesParametersToTerraform;
/**
 * @stability stable
 */
class GlueMlTransformParametersFindMatchesParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accuracyCostTradeOff) {
            hasAnyValues = true;
            internalValueResult.accuracyCostTradeOff = this._accuracyCostTradeOff;
        }
        if (this._enforceProvidedLabels) {
            hasAnyValues = true;
            internalValueResult.enforceProvidedLabels = this._enforceProvidedLabels;
        }
        if (this._precisionRecallTradeOff) {
            hasAnyValues = true;
            internalValueResult.precisionRecallTradeOff = this._precisionRecallTradeOff;
        }
        if (this._primaryKeyColumnName) {
            hasAnyValues = true;
            internalValueResult.primaryKeyColumnName = this._primaryKeyColumnName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accuracyCostTradeOff = undefined;
            this._enforceProvidedLabels = undefined;
            this._precisionRecallTradeOff = undefined;
            this._primaryKeyColumnName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accuracyCostTradeOff = value.accuracyCostTradeOff;
            this._enforceProvidedLabels = value.enforceProvidedLabels;
            this._precisionRecallTradeOff = value.precisionRecallTradeOff;
            this._primaryKeyColumnName = value.primaryKeyColumnName;
        }
    }
    /**
     * @stability stable
     */
    get accuracyCostTradeOff() {
        return this.getNumberAttribute('accuracy_cost_trade_off');
    }
    /**
     * @stability stable
     */
    set accuracyCostTradeOff(value) {
        this._accuracyCostTradeOff = value;
    }
    /**
     * @stability stable
     */
    resetAccuracyCostTradeOff() {
        this._accuracyCostTradeOff = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accuracyCostTradeOffInput() {
        return this._accuracyCostTradeOff;
    }
    /**
     * @stability stable
     */
    get enforceProvidedLabels() {
        return this.getBooleanAttribute('enforce_provided_labels');
    }
    /**
     * @stability stable
     */
    set enforceProvidedLabels(value) {
        this._enforceProvidedLabels = value;
    }
    /**
     * @stability stable
     */
    resetEnforceProvidedLabels() {
        this._enforceProvidedLabels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enforceProvidedLabelsInput() {
        return this._enforceProvidedLabels;
    }
    /**
     * @stability stable
     */
    get precisionRecallTradeOff() {
        return this.getNumberAttribute('precision_recall_trade_off');
    }
    /**
     * @stability stable
     */
    set precisionRecallTradeOff(value) {
        this._precisionRecallTradeOff = value;
    }
    /**
     * @stability stable
     */
    resetPrecisionRecallTradeOff() {
        this._precisionRecallTradeOff = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get precisionRecallTradeOffInput() {
        return this._precisionRecallTradeOff;
    }
    /**
     * @stability stable
     */
    get primaryKeyColumnName() {
        return this.getStringAttribute('primary_key_column_name');
    }
    /**
     * @stability stable
     */
    set primaryKeyColumnName(value) {
        this._primaryKeyColumnName = value;
    }
    /**
     * @stability stable
     */
    resetPrimaryKeyColumnName() {
        this._primaryKeyColumnName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get primaryKeyColumnNameInput() {
        return this._primaryKeyColumnName;
    }
}
exports.GlueMlTransformParametersFindMatchesParametersOutputReference = GlueMlTransformParametersFindMatchesParametersOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlueMlTransformParametersFindMatchesParametersOutputReference[_b] = { fqn: "@cdktf/provider-aws.glue.GlueMlTransformParametersFindMatchesParametersOutputReference", version: "3.0.1" };
function glueMlTransformParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        transform_type: cdktf.stringToTerraform(struct.transformType),
        find_matches_parameters: glueMlTransformParametersFindMatchesParametersToTerraform(struct.findMatchesParameters),
    };
}
exports.glueMlTransformParametersToTerraform = glueMlTransformParametersToTerraform;
/**
 * @stability stable
 */
class GlueMlTransformParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // find_matches_parameters - computed: false, optional: false, required: true
        this._findMatchesParameters = new GlueMlTransformParametersFindMatchesParametersOutputReference(this, "find_matches_parameters", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._transformType) {
            hasAnyValues = true;
            internalValueResult.transformType = this._transformType;
        }
        if ((_e = this._findMatchesParameters) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.findMatchesParameters = (_f = this._findMatchesParameters) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._transformType = undefined;
            this._findMatchesParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._transformType = value.transformType;
            this._findMatchesParameters.internalValue = value.findMatchesParameters;
        }
    }
    /**
     * @stability stable
     */
    get transformType() {
        return this.getStringAttribute('transform_type');
    }
    /**
     * @stability stable
     */
    set transformType(value) {
        this._transformType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get transformTypeInput() {
        return this._transformType;
    }
    /**
     * @stability stable
     */
    get findMatchesParameters() {
        return this._findMatchesParameters;
    }
    /**
     * @stability stable
     */
    putFindMatchesParameters(value) {
        this._findMatchesParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get findMatchesParametersInput() {
        return this._findMatchesParameters.internalValue;
    }
}
exports.GlueMlTransformParametersOutputReference = GlueMlTransformParametersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlueMlTransformParametersOutputReference[_c] = { fqn: "@cdktf/provider-aws.glue.GlueMlTransformParametersOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_ml_transform.html aws_glue_ml_transform}.
 *
 * @stability stable
 */
class GlueMlTransform extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_ml_transform.html aws_glue_ml_transform} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_ml_transform',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // parameters - computed: false, optional: false, required: true
        this._parameters = new GlueMlTransformParametersOutputReference(this, "parameters", true);
        this._description = config.description;
        this._glueVersion = config.glueVersion;
        this._maxCapacity = config.maxCapacity;
        this._maxRetries = config.maxRetries;
        this._name = config.name;
        this._numberOfWorkers = config.numberOfWorkers;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeout = config.timeout;
        this._workerType = config.workerType;
        this._inputRecordTables = config.inputRecordTables;
        this._parameters.internalValue = config.parameters;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get glueVersion() {
        return this.getStringAttribute('glue_version');
    }
    /**
     * @stability stable
     */
    set glueVersion(value) {
        this._glueVersion = value;
    }
    /**
     * @stability stable
     */
    resetGlueVersion() {
        this._glueVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get glueVersionInput() {
        return this._glueVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // label_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get labelCount() {
        return this.getNumberAttribute('label_count');
    }
    /**
     * @stability stable
     */
    get maxCapacity() {
        return this.getNumberAttribute('max_capacity');
    }
    /**
     * @stability stable
     */
    set maxCapacity(value) {
        this._maxCapacity = value;
    }
    /**
     * @stability stable
     */
    resetMaxCapacity() {
        this._maxCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxCapacityInput() {
        return this._maxCapacity;
    }
    /**
     * @stability stable
     */
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    /**
     * @stability stable
     */
    set maxRetries(value) {
        this._maxRetries = value;
    }
    /**
     * @stability stable
     */
    resetMaxRetries() {
        this._maxRetries = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxRetriesInput() {
        return this._maxRetries;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get numberOfWorkers() {
        return this.getNumberAttribute('number_of_workers');
    }
    /**
     * @stability stable
     */
    set numberOfWorkers(value) {
        this._numberOfWorkers = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfWorkers() {
        this._numberOfWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfWorkersInput() {
        return this._numberOfWorkers;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    // schema - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    schema(index) {
        return new GlueMlTransformSchema(this, 'schema', index);
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    /**
     * @stability stable
     */
    set timeout(value) {
        this._timeout = value;
    }
    /**
     * @stability stable
     */
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutInput() {
        return this._timeout;
    }
    /**
     * @stability stable
     */
    get workerType() {
        return this.getStringAttribute('worker_type');
    }
    /**
     * @stability stable
     */
    set workerType(value) {
        this._workerType = value;
    }
    /**
     * @stability stable
     */
    resetWorkerType() {
        this._workerType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workerTypeInput() {
        return this._workerType;
    }
    /**
     * @stability stable
     */
    get inputRecordTables() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('input_record_tables');
    }
    /**
     * @stability stable
     */
    set inputRecordTables(value) {
        this._inputRecordTables = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputRecordTablesInput() {
        return this._inputRecordTables;
    }
    /**
     * @stability stable
     */
    get parameters() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    putParameters(value) {
        this._parameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            glue_version: cdktf.stringToTerraform(this._glueVersion),
            max_capacity: cdktf.numberToTerraform(this._maxCapacity),
            max_retries: cdktf.numberToTerraform(this._maxRetries),
            name: cdktf.stringToTerraform(this._name),
            number_of_workers: cdktf.numberToTerraform(this._numberOfWorkers),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            timeout: cdktf.numberToTerraform(this._timeout),
            worker_type: cdktf.stringToTerraform(this._workerType),
            input_record_tables: cdktf.listMapper(glueMlTransformInputRecordTablesToTerraform)(this._inputRecordTables),
            parameters: glueMlTransformParametersToTerraform(this._parameters.internalValue),
        };
    }
}
exports.GlueMlTransform = GlueMlTransform;
_d = JSII_RTTI_SYMBOL_1;
GlueMlTransform[_d] = { fqn: "@cdktf/provider-aws.glue.GlueMlTransform", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueMlTransform.tfResourceType = "aws_glue_ml_transform";
