// @cdktf/provider-aws/lib/glue/glue-schema.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueSchema = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_schema.html aws_glue_schema}.
 *
 * @stability stable
 */
class GlueSchema extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_schema.html aws_glue_schema} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_schema',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._compatibility = config.compatibility;
        this._dataFormat = config.dataFormat;
        this._description = config.description;
        this._registryArn = config.registryArn;
        this._schemaDefinition = config.schemaDefinition;
        this._schemaName = config.schemaName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get compatibility() {
        return this.getStringAttribute('compatibility');
    }
    /**
     * @stability stable
     */
    set compatibility(value) {
        this._compatibility = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get compatibilityInput() {
        return this._compatibility;
    }
    /**
     * @stability stable
     */
    get dataFormat() {
        return this.getStringAttribute('data_format');
    }
    /**
     * @stability stable
     */
    set dataFormat(value) {
        this._dataFormat = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataFormatInput() {
        return this._dataFormat;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // latest_schema_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get latestSchemaVersion() {
        return this.getNumberAttribute('latest_schema_version');
    }
    // next_schema_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nextSchemaVersion() {
        return this.getNumberAttribute('next_schema_version');
    }
    /**
     * @stability stable
     */
    get registryArn() {
        return this.getStringAttribute('registry_arn');
    }
    /**
     * @stability stable
     */
    set registryArn(value) {
        this._registryArn = value;
    }
    /**
     * @stability stable
     */
    resetRegistryArn() {
        this._registryArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registryArnInput() {
        return this._registryArn;
    }
    // registry_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get registryName() {
        return this.getStringAttribute('registry_name');
    }
    // schema_checkpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get schemaCheckpoint() {
        return this.getNumberAttribute('schema_checkpoint');
    }
    /**
     * @stability stable
     */
    get schemaDefinition() {
        return this.getStringAttribute('schema_definition');
    }
    /**
     * @stability stable
     */
    set schemaDefinition(value) {
        this._schemaDefinition = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaDefinitionInput() {
        return this._schemaDefinition;
    }
    /**
     * @stability stable
     */
    get schemaName() {
        return this.getStringAttribute('schema_name');
    }
    /**
     * @stability stable
     */
    set schemaName(value) {
        this._schemaName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaNameInput() {
        return this._schemaName;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            compatibility: cdktf.stringToTerraform(this._compatibility),
            data_format: cdktf.stringToTerraform(this._dataFormat),
            description: cdktf.stringToTerraform(this._description),
            registry_arn: cdktf.stringToTerraform(this._registryArn),
            schema_definition: cdktf.stringToTerraform(this._schemaDefinition),
            schema_name: cdktf.stringToTerraform(this._schemaName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.GlueSchema = GlueSchema;
_a = JSII_RTTI_SYMBOL_1;
GlueSchema[_a] = { fqn: "@cdktf/provider-aws.glue.GlueSchema", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueSchema.tfResourceType = "aws_glue_schema";
