// @cdktf/provider-aws/lib/datasources/data-aws-availability-zones.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAvailabilityZones = exports.dataAwsAvailabilityZonesFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAvailabilityZonesFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsAvailabilityZonesFilterToTerraform = dataAwsAvailabilityZonesFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/availability_zones.html aws_availability_zones}.
 *
 * @stability stable
 */
class DataAwsAvailabilityZones extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/availability_zones.html aws_availability_zones} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_availability_zones',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._allAvailabilityZones = config.allAvailabilityZones;
        this._excludeNames = config.excludeNames;
        this._excludeZoneIds = config.excludeZoneIds;
        this._state = config.state;
        this._filter = config.filter;
    }
    /**
     * @stability stable
     */
    get allAvailabilityZones() {
        return this.getBooleanAttribute('all_availability_zones');
    }
    /**
     * @stability stable
     */
    set allAvailabilityZones(value) {
        this._allAvailabilityZones = value;
    }
    /**
     * @stability stable
     */
    resetAllAvailabilityZones() {
        this._allAvailabilityZones = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allAvailabilityZonesInput() {
        return this._allAvailabilityZones;
    }
    /**
     * @stability stable
     */
    get excludeNames() {
        return this.getListAttribute('exclude_names');
    }
    /**
     * @stability stable
     */
    set excludeNames(value) {
        this._excludeNames = value;
    }
    /**
     * @stability stable
     */
    resetExcludeNames() {
        this._excludeNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludeNamesInput() {
        return this._excludeNames;
    }
    /**
     * @stability stable
     */
    get excludeZoneIds() {
        return this.getListAttribute('exclude_zone_ids');
    }
    /**
     * @stability stable
     */
    set excludeZoneIds(value) {
        this._excludeZoneIds = value;
    }
    /**
     * @stability stable
     */
    resetExcludeZoneIds() {
        this._excludeZoneIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludeZoneIdsInput() {
        return this._excludeZoneIds;
    }
    // group_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get groupNames() {
        return this.getListAttribute('group_names');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get names() {
        return this.getListAttribute('names');
    }
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    set state(value) {
        this._state = value;
    }
    /**
     * @stability stable
     */
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateInput() {
        return this._state;
    }
    // zone_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneIds() {
        return this.getListAttribute('zone_ids');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            all_availability_zones: cdktf.booleanToTerraform(this._allAvailabilityZones),
            exclude_names: cdktf.listMapper(cdktf.stringToTerraform)(this._excludeNames),
            exclude_zone_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._excludeZoneIds),
            state: cdktf.stringToTerraform(this._state),
            filter: cdktf.listMapper(dataAwsAvailabilityZonesFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsAvailabilityZones = DataAwsAvailabilityZones;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZones[_a] = { fqn: "@cdktf/provider-aws.datasources.DataAwsAvailabilityZones", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAvailabilityZones.tfResourceType = "aws_availability_zones";
