// @cdktf/provider-aws/lib/datasources/data-aws-regions.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRegions = exports.dataAwsRegionsFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRegionsFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsRegionsFilterToTerraform = dataAwsRegionsFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/regions.html aws_regions}.
 *
 * @stability stable
 */
class DataAwsRegions extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/regions.html aws_regions} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_regions',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._allRegions = config.allRegions;
        this._filter = config.filter;
    }
    /**
     * @stability stable
     */
    get allRegions() {
        return this.getBooleanAttribute('all_regions');
    }
    /**
     * @stability stable
     */
    set allRegions(value) {
        this._allRegions = value;
    }
    /**
     * @stability stable
     */
    resetAllRegions() {
        this._allRegions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allRegionsInput() {
        return this._allRegions;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get names() {
        return this.getListAttribute('names');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            all_regions: cdktf.booleanToTerraform(this._allRegions),
            filter: cdktf.listMapper(dataAwsRegionsFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsRegions = DataAwsRegions;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRegions[_a] = { fqn: "@cdktf/provider-aws.datasources.DataAwsRegions", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsRegions.tfResourceType = "aws_regions";
