// @cdktf/provider-aws/lib/datasources/launch-configuration.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LaunchConfiguration = exports.LaunchConfigurationRootBlockDeviceOutputReference = exports.launchConfigurationRootBlockDeviceToTerraform = exports.LaunchConfigurationMetadataOptionsOutputReference = exports.launchConfigurationMetadataOptionsToTerraform = exports.launchConfigurationEphemeralBlockDeviceToTerraform = exports.launchConfigurationEbsBlockDeviceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function launchConfigurationEbsBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_on_termination: cdktf.booleanToTerraform(struct.deleteOnTermination),
        device_name: cdktf.stringToTerraform(struct.deviceName),
        encrypted: cdktf.booleanToTerraform(struct.encrypted),
        iops: cdktf.numberToTerraform(struct.iops),
        no_device: cdktf.booleanToTerraform(struct.noDevice),
        snapshot_id: cdktf.stringToTerraform(struct.snapshotId),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.launchConfigurationEbsBlockDeviceToTerraform = launchConfigurationEbsBlockDeviceToTerraform;
function launchConfigurationEphemeralBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        device_name: cdktf.stringToTerraform(struct.deviceName),
        virtual_name: cdktf.stringToTerraform(struct.virtualName),
    };
}
exports.launchConfigurationEphemeralBlockDeviceToTerraform = launchConfigurationEphemeralBlockDeviceToTerraform;
function launchConfigurationMetadataOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        http_endpoint: cdktf.stringToTerraform(struct.httpEndpoint),
        http_put_response_hop_limit: cdktf.numberToTerraform(struct.httpPutResponseHopLimit),
        http_tokens: cdktf.stringToTerraform(struct.httpTokens),
    };
}
exports.launchConfigurationMetadataOptionsToTerraform = launchConfigurationMetadataOptionsToTerraform;
/**
 * @stability stable
 */
class LaunchConfigurationMetadataOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpEndpoint) {
            hasAnyValues = true;
            internalValueResult.httpEndpoint = this._httpEndpoint;
        }
        if (this._httpPutResponseHopLimit) {
            hasAnyValues = true;
            internalValueResult.httpPutResponseHopLimit = this._httpPutResponseHopLimit;
        }
        if (this._httpTokens) {
            hasAnyValues = true;
            internalValueResult.httpTokens = this._httpTokens;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._httpEndpoint = undefined;
            this._httpPutResponseHopLimit = undefined;
            this._httpTokens = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._httpEndpoint = value.httpEndpoint;
            this._httpPutResponseHopLimit = value.httpPutResponseHopLimit;
            this._httpTokens = value.httpTokens;
        }
    }
    /**
     * @stability stable
     */
    get httpEndpoint() {
        return this.getStringAttribute('http_endpoint');
    }
    /**
     * @stability stable
     */
    set httpEndpoint(value) {
        this._httpEndpoint = value;
    }
    /**
     * @stability stable
     */
    resetHttpEndpoint() {
        this._httpEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpEndpointInput() {
        return this._httpEndpoint;
    }
    /**
     * @stability stable
     */
    get httpPutResponseHopLimit() {
        return this.getNumberAttribute('http_put_response_hop_limit');
    }
    /**
     * @stability stable
     */
    set httpPutResponseHopLimit(value) {
        this._httpPutResponseHopLimit = value;
    }
    /**
     * @stability stable
     */
    resetHttpPutResponseHopLimit() {
        this._httpPutResponseHopLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpPutResponseHopLimitInput() {
        return this._httpPutResponseHopLimit;
    }
    /**
     * @stability stable
     */
    get httpTokens() {
        return this.getStringAttribute('http_tokens');
    }
    /**
     * @stability stable
     */
    set httpTokens(value) {
        this._httpTokens = value;
    }
    /**
     * @stability stable
     */
    resetHttpTokens() {
        this._httpTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpTokensInput() {
        return this._httpTokens;
    }
}
exports.LaunchConfigurationMetadataOptionsOutputReference = LaunchConfigurationMetadataOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LaunchConfigurationMetadataOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasources.LaunchConfigurationMetadataOptionsOutputReference", version: "3.0.1" };
function launchConfigurationRootBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_on_termination: cdktf.booleanToTerraform(struct.deleteOnTermination),
        encrypted: cdktf.booleanToTerraform(struct.encrypted),
        iops: cdktf.numberToTerraform(struct.iops),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.launchConfigurationRootBlockDeviceToTerraform = launchConfigurationRootBlockDeviceToTerraform;
/**
 * @stability stable
 */
class LaunchConfigurationRootBlockDeviceOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteOnTermination) {
            hasAnyValues = true;
            internalValueResult.deleteOnTermination = this._deleteOnTermination;
        }
        if (this._encrypted) {
            hasAnyValues = true;
            internalValueResult.encrypted = this._encrypted;
        }
        if (this._iops) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._throughput) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._volumeSize) {
            hasAnyValues = true;
            internalValueResult.volumeSize = this._volumeSize;
        }
        if (this._volumeType) {
            hasAnyValues = true;
            internalValueResult.volumeType = this._volumeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteOnTermination = undefined;
            this._encrypted = undefined;
            this._iops = undefined;
            this._throughput = undefined;
            this._volumeSize = undefined;
            this._volumeType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteOnTermination = value.deleteOnTermination;
            this._encrypted = value.encrypted;
            this._iops = value.iops;
            this._throughput = value.throughput;
            this._volumeSize = value.volumeSize;
            this._volumeType = value.volumeType;
        }
    }
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    /**
     * @stability stable
     */
    set deleteOnTermination(value) {
        this._deleteOnTermination = value;
    }
    /**
     * @stability stable
     */
    resetDeleteOnTermination() {
        this._deleteOnTermination = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteOnTerminationInput() {
        return this._deleteOnTermination;
    }
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    /**
     * @stability stable
     */
    set encrypted(value) {
        this._encrypted = value;
    }
    /**
     * @stability stable
     */
    resetEncrypted() {
        this._encrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptedInput() {
        return this._encrypted;
    }
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    /**
     * @stability stable
     */
    set iops(value) {
        this._iops = value;
    }
    /**
     * @stability stable
     */
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iopsInput() {
        return this._iops;
    }
    /**
     * @stability stable
     */
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    /**
     * @stability stable
     */
    set throughput(value) {
        this._throughput = value;
    }
    /**
     * @stability stable
     */
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throughputInput() {
        return this._throughput;
    }
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    /**
     * @stability stable
     */
    set volumeSize(value) {
        this._volumeSize = value;
    }
    /**
     * @stability stable
     */
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get volumeSizeInput() {
        return this._volumeSize;
    }
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    /**
     * @stability stable
     */
    set volumeType(value) {
        this._volumeType = value;
    }
    /**
     * @stability stable
     */
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get volumeTypeInput() {
        return this._volumeType;
    }
}
exports.LaunchConfigurationRootBlockDeviceOutputReference = LaunchConfigurationRootBlockDeviceOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LaunchConfigurationRootBlockDeviceOutputReference[_b] = { fqn: "@cdktf/provider-aws.datasources.LaunchConfigurationRootBlockDeviceOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/launch_configuration.html aws_launch_configuration}.
 *
 * @stability stable
 */
class LaunchConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/launch_configuration.html aws_launch_configuration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_launch_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // metadata_options - computed: false, optional: true, required: false
        this._metadataOptions = new LaunchConfigurationMetadataOptionsOutputReference(this, "metadata_options", true);
        // root_block_device - computed: false, optional: true, required: false
        this._rootBlockDevice = new LaunchConfigurationRootBlockDeviceOutputReference(this, "root_block_device", true);
        this._associatePublicIpAddress = config.associatePublicIpAddress;
        this._ebsOptimized = config.ebsOptimized;
        this._enableMonitoring = config.enableMonitoring;
        this._iamInstanceProfile = config.iamInstanceProfile;
        this._imageId = config.imageId;
        this._instanceType = config.instanceType;
        this._keyName = config.keyName;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._placementTenancy = config.placementTenancy;
        this._securityGroups = config.securityGroups;
        this._spotPrice = config.spotPrice;
        this._userData = config.userData;
        this._userDataBase64 = config.userDataBase64;
        this._vpcClassicLinkId = config.vpcClassicLinkId;
        this._vpcClassicLinkSecurityGroups = config.vpcClassicLinkSecurityGroups;
        this._ebsBlockDevice = config.ebsBlockDevice;
        this._ephemeralBlockDevice = config.ephemeralBlockDevice;
        this._metadataOptions.internalValue = config.metadataOptions;
        this._rootBlockDevice.internalValue = config.rootBlockDevice;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get associatePublicIpAddress() {
        return this.getBooleanAttribute('associate_public_ip_address');
    }
    /**
     * @stability stable
     */
    set associatePublicIpAddress(value) {
        this._associatePublicIpAddress = value;
    }
    /**
     * @stability stable
     */
    resetAssociatePublicIpAddress() {
        this._associatePublicIpAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get associatePublicIpAddressInput() {
        return this._associatePublicIpAddress;
    }
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    /**
     * @stability stable
     */
    set ebsOptimized(value) {
        this._ebsOptimized = value;
    }
    /**
     * @stability stable
     */
    resetEbsOptimized() {
        this._ebsOptimized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsOptimizedInput() {
        return this._ebsOptimized;
    }
    /**
     * @stability stable
     */
    get enableMonitoring() {
        return this.getBooleanAttribute('enable_monitoring');
    }
    /**
     * @stability stable
     */
    set enableMonitoring(value) {
        this._enableMonitoring = value;
    }
    /**
     * @stability stable
     */
    resetEnableMonitoring() {
        this._enableMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableMonitoringInput() {
        return this._enableMonitoring;
    }
    /**
     * @stability stable
     */
    get iamInstanceProfile() {
        return this.getStringAttribute('iam_instance_profile');
    }
    /**
     * @stability stable
     */
    set iamInstanceProfile(value) {
        this._iamInstanceProfile = value;
    }
    /**
     * @stability stable
     */
    resetIamInstanceProfile() {
        this._iamInstanceProfile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamInstanceProfileInput() {
        return this._iamInstanceProfile;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    /**
     * @stability stable
     */
    set imageId(value) {
        this._imageId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get imageIdInput() {
        return this._imageId;
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    /**
     * @stability stable
     */
    set keyName(value) {
        this._keyName = value;
    }
    /**
     * @stability stable
     */
    resetKeyName() {
        this._keyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyNameInput() {
        return this._keyName;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get placementTenancy() {
        return this.getStringAttribute('placement_tenancy');
    }
    /**
     * @stability stable
     */
    set placementTenancy(value) {
        this._placementTenancy = value;
    }
    /**
     * @stability stable
     */
    resetPlacementTenancy() {
        this._placementTenancy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get placementTenancyInput() {
        return this._placementTenancy;
    }
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    /**
     * @stability stable
     */
    set securityGroups(value) {
        this._securityGroups = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupsInput() {
        return this._securityGroups;
    }
    /**
     * @stability stable
     */
    get spotPrice() {
        return this.getStringAttribute('spot_price');
    }
    /**
     * @stability stable
     */
    set spotPrice(value) {
        this._spotPrice = value;
    }
    /**
     * @stability stable
     */
    resetSpotPrice() {
        this._spotPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spotPriceInput() {
        return this._spotPrice;
    }
    /**
     * @stability stable
     */
    get userData() {
        return this.getStringAttribute('user_data');
    }
    /**
     * @stability stable
     */
    set userData(value) {
        this._userData = value;
    }
    /**
     * @stability stable
     */
    resetUserData() {
        this._userData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userDataInput() {
        return this._userData;
    }
    /**
     * @stability stable
     */
    get userDataBase64() {
        return this.getStringAttribute('user_data_base64');
    }
    /**
     * @stability stable
     */
    set userDataBase64(value) {
        this._userDataBase64 = value;
    }
    /**
     * @stability stable
     */
    resetUserDataBase64() {
        this._userDataBase64 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userDataBase64Input() {
        return this._userDataBase64;
    }
    /**
     * @stability stable
     */
    get vpcClassicLinkId() {
        return this.getStringAttribute('vpc_classic_link_id');
    }
    /**
     * @stability stable
     */
    set vpcClassicLinkId(value) {
        this._vpcClassicLinkId = value;
    }
    /**
     * @stability stable
     */
    resetVpcClassicLinkId() {
        this._vpcClassicLinkId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcClassicLinkIdInput() {
        return this._vpcClassicLinkId;
    }
    /**
     * @stability stable
     */
    get vpcClassicLinkSecurityGroups() {
        return this.getListAttribute('vpc_classic_link_security_groups');
    }
    /**
     * @stability stable
     */
    set vpcClassicLinkSecurityGroups(value) {
        this._vpcClassicLinkSecurityGroups = value;
    }
    /**
     * @stability stable
     */
    resetVpcClassicLinkSecurityGroups() {
        this._vpcClassicLinkSecurityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcClassicLinkSecurityGroupsInput() {
        return this._vpcClassicLinkSecurityGroups;
    }
    /**
     * @stability stable
     */
    get ebsBlockDevice() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ebs_block_device');
    }
    /**
     * @stability stable
     */
    set ebsBlockDevice(value) {
        this._ebsBlockDevice = value;
    }
    /**
     * @stability stable
     */
    resetEbsBlockDevice() {
        this._ebsBlockDevice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsBlockDeviceInput() {
        return this._ebsBlockDevice;
    }
    /**
     * @stability stable
     */
    get ephemeralBlockDevice() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ephemeral_block_device');
    }
    /**
     * @stability stable
     */
    set ephemeralBlockDevice(value) {
        this._ephemeralBlockDevice = value;
    }
    /**
     * @stability stable
     */
    resetEphemeralBlockDevice() {
        this._ephemeralBlockDevice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ephemeralBlockDeviceInput() {
        return this._ephemeralBlockDevice;
    }
    /**
     * @stability stable
     */
    get metadataOptions() {
        return this._metadataOptions;
    }
    /**
     * @stability stable
     */
    putMetadataOptions(value) {
        this._metadataOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMetadataOptions() {
        this._metadataOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metadataOptionsInput() {
        return this._metadataOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get rootBlockDevice() {
        return this._rootBlockDevice;
    }
    /**
     * @stability stable
     */
    putRootBlockDevice(value) {
        this._rootBlockDevice.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRootBlockDevice() {
        this._rootBlockDevice.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootBlockDeviceInput() {
        return this._rootBlockDevice.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            associate_public_ip_address: cdktf.booleanToTerraform(this._associatePublicIpAddress),
            ebs_optimized: cdktf.booleanToTerraform(this._ebsOptimized),
            enable_monitoring: cdktf.booleanToTerraform(this._enableMonitoring),
            iam_instance_profile: cdktf.stringToTerraform(this._iamInstanceProfile),
            image_id: cdktf.stringToTerraform(this._imageId),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            key_name: cdktf.stringToTerraform(this._keyName),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            placement_tenancy: cdktf.stringToTerraform(this._placementTenancy),
            security_groups: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroups),
            spot_price: cdktf.stringToTerraform(this._spotPrice),
            user_data: cdktf.stringToTerraform(this._userData),
            user_data_base64: cdktf.stringToTerraform(this._userDataBase64),
            vpc_classic_link_id: cdktf.stringToTerraform(this._vpcClassicLinkId),
            vpc_classic_link_security_groups: cdktf.listMapper(cdktf.stringToTerraform)(this._vpcClassicLinkSecurityGroups),
            ebs_block_device: cdktf.listMapper(launchConfigurationEbsBlockDeviceToTerraform)(this._ebsBlockDevice),
            ephemeral_block_device: cdktf.listMapper(launchConfigurationEphemeralBlockDeviceToTerraform)(this._ephemeralBlockDevice),
            metadata_options: launchConfigurationMetadataOptionsToTerraform(this._metadataOptions.internalValue),
            root_block_device: launchConfigurationRootBlockDeviceToTerraform(this._rootBlockDevice.internalValue),
        };
    }
}
exports.LaunchConfiguration = LaunchConfiguration;
_c = JSII_RTTI_SYMBOL_1;
LaunchConfiguration[_c] = { fqn: "@cdktf/provider-aws.datasources.LaunchConfiguration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LaunchConfiguration.tfResourceType = "aws_launch_configuration";
