// @cdktf/provider-aws/lib/sagemaker/data-aws-sagemaker-prebuilt-ecr-image.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSagemakerPrebuiltEcrImage = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/sagemaker_prebuilt_ecr_image.html aws_sagemaker_prebuilt_ecr_image}.
 *
 * @stability stable
 */
class DataAwsSagemakerPrebuiltEcrImage extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/sagemaker_prebuilt_ecr_image.html aws_sagemaker_prebuilt_ecr_image} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_prebuilt_ecr_image',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dnsSuffix = config.dnsSuffix;
        this._imageTag = config.imageTag;
        this._region = config.region;
        this._repositoryName = config.repositoryName;
    }
    /**
     * @stability stable
     */
    get dnsSuffix() {
        return this.getStringAttribute('dns_suffix');
    }
    /**
     * @stability stable
     */
    set dnsSuffix(value) {
        this._dnsSuffix = value;
    }
    /**
     * @stability stable
     */
    resetDnsSuffix() {
        this._dnsSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dnsSuffixInput() {
        return this._dnsSuffix;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get imageTag() {
        return this.getStringAttribute('image_tag');
    }
    /**
     * @stability stable
     */
    set imageTag(value) {
        this._imageTag = value;
    }
    /**
     * @stability stable
     */
    resetImageTag() {
        this._imageTag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get imageTagInput() {
        return this._imageTag;
    }
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
    /**
     * @stability stable
     */
    set region(value) {
        this._region = value;
    }
    /**
     * @stability stable
     */
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionInput() {
        return this._region;
    }
    // registry_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    // registry_path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get registryPath() {
        return this.getStringAttribute('registry_path');
    }
    /**
     * @stability stable
     */
    get repositoryName() {
        return this.getStringAttribute('repository_name');
    }
    /**
     * @stability stable
     */
    set repositoryName(value) {
        this._repositoryName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get repositoryNameInput() {
        return this._repositoryName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            dns_suffix: cdktf.stringToTerraform(this._dnsSuffix),
            image_tag: cdktf.stringToTerraform(this._imageTag),
            region: cdktf.stringToTerraform(this._region),
            repository_name: cdktf.stringToTerraform(this._repositoryName),
        };
    }
}
exports.DataAwsSagemakerPrebuiltEcrImage = DataAwsSagemakerPrebuiltEcrImage;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSagemakerPrebuiltEcrImage[_a] = { fqn: "@cdktf/provider-aws.sagemaker.DataAwsSagemakerPrebuiltEcrImage", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSagemakerPrebuiltEcrImage.tfResourceType = "aws_sagemaker_prebuilt_ecr_image";
