// @cdktf/provider-aws/lib/sagemaker/sagemaker-user-profile.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerUserProfile = exports.SagemakerUserProfileUserSettingsOutputReference = exports.sagemakerUserProfileUserSettingsToTerraform = exports.SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference = exports.sagemakerUserProfileUserSettingsTensorBoardAppSettingsToTerraform = exports.SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference = exports.sagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecToTerraform = exports.SagemakerUserProfileUserSettingsSharingSettingsOutputReference = exports.sagemakerUserProfileUserSettingsSharingSettingsToTerraform = exports.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference = exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsToTerraform = exports.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference = exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecToTerraform = exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsCustomImageToTerraform = exports.SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference = exports.sagemakerUserProfileUserSettingsJupyterServerAppSettingsToTerraform = exports.SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference = exports.sagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        sagemaker_image_arn: cdktf.stringToTerraform(struct.sagemakerImageArn),
    };
}
exports.sagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecToTerraform = sagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceType) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._sagemakerImageArn) {
            hasAnyValues = true;
            internalValueResult.sagemakerImageArn = this._sagemakerImageArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instanceType = undefined;
            this._sagemakerImageArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instanceType = value.instanceType;
            this._sagemakerImageArn = value.sagemakerImageArn;
        }
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get sagemakerImageArn() {
        return this.getStringAttribute('sagemaker_image_arn');
    }
    /**
     * @stability stable
     */
    set sagemakerImageArn(value) {
        this._sagemakerImageArn = value;
    }
    /**
     * @stability stable
     */
    resetSagemakerImageArn() {
        this._sagemakerImageArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sagemakerImageArnInput() {
        return this._sagemakerImageArn;
    }
}
exports.SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference = SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsJupyterServerAppSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lifecycle_config_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.lifecycleConfigArns),
        default_resource_spec: sagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecToTerraform(struct.defaultResourceSpec),
    };
}
exports.sagemakerUserProfileUserSettingsJupyterServerAppSettingsToTerraform = sagemakerUserProfileUserSettingsJupyterServerAppSettingsToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // default_resource_spec - computed: false, optional: false, required: true
        this._defaultResourceSpec = new SagemakerUserProfileUserSettingsJupyterServerAppSettingsDefaultResourceSpecOutputReference(this, "default_resource_spec", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lifecycleConfigArns) {
            hasAnyValues = true;
            internalValueResult.lifecycleConfigArns = this._lifecycleConfigArns;
        }
        if ((_k = this._defaultResourceSpec) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.defaultResourceSpec = (_l = this._defaultResourceSpec) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lifecycleConfigArns = undefined;
            this._defaultResourceSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lifecycleConfigArns = value.lifecycleConfigArns;
            this._defaultResourceSpec.internalValue = value.defaultResourceSpec;
        }
    }
    /**
     * @stability stable
     */
    get lifecycleConfigArns() {
        return this.getListAttribute('lifecycle_config_arns');
    }
    /**
     * @stability stable
     */
    set lifecycleConfigArns(value) {
        this._lifecycleConfigArns = value;
    }
    /**
     * @stability stable
     */
    resetLifecycleConfigArns() {
        this._lifecycleConfigArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lifecycleConfigArnsInput() {
        return this._lifecycleConfigArns;
    }
    /**
     * @stability stable
     */
    get defaultResourceSpec() {
        return this._defaultResourceSpec;
    }
    /**
     * @stability stable
     */
    putDefaultResourceSpec(value) {
        this._defaultResourceSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultResourceSpecInput() {
        return this._defaultResourceSpec.internalValue;
    }
}
exports.SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference = SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference[_b] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsKernelGatewayAppSettingsCustomImageToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        app_image_config_name: cdktf.stringToTerraform(struct.appImageConfigName),
        image_name: cdktf.stringToTerraform(struct.imageName),
        image_version_number: cdktf.numberToTerraform(struct.imageVersionNumber),
    };
}
exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsCustomImageToTerraform = sagemakerUserProfileUserSettingsKernelGatewayAppSettingsCustomImageToTerraform;
function sagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        sagemaker_image_arn: cdktf.stringToTerraform(struct.sagemakerImageArn),
    };
}
exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecToTerraform = sagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceType) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._sagemakerImageArn) {
            hasAnyValues = true;
            internalValueResult.sagemakerImageArn = this._sagemakerImageArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instanceType = undefined;
            this._sagemakerImageArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instanceType = value.instanceType;
            this._sagemakerImageArn = value.sagemakerImageArn;
        }
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get sagemakerImageArn() {
        return this.getStringAttribute('sagemaker_image_arn');
    }
    /**
     * @stability stable
     */
    set sagemakerImageArn(value) {
        this._sagemakerImageArn = value;
    }
    /**
     * @stability stable
     */
    resetSagemakerImageArn() {
        this._sagemakerImageArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sagemakerImageArnInput() {
        return this._sagemakerImageArn;
    }
}
exports.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference = SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference[_c] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsKernelGatewayAppSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lifecycle_config_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.lifecycleConfigArns),
        custom_image: cdktf.listMapper(sagemakerUserProfileUserSettingsKernelGatewayAppSettingsCustomImageToTerraform)(struct.customImage),
        default_resource_spec: sagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecToTerraform(struct.defaultResourceSpec),
    };
}
exports.sagemakerUserProfileUserSettingsKernelGatewayAppSettingsToTerraform = sagemakerUserProfileUserSettingsKernelGatewayAppSettingsToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // default_resource_spec - computed: false, optional: false, required: true
        this._defaultResourceSpec = new SagemakerUserProfileUserSettingsKernelGatewayAppSettingsDefaultResourceSpecOutputReference(this, "default_resource_spec", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lifecycleConfigArns) {
            hasAnyValues = true;
            internalValueResult.lifecycleConfigArns = this._lifecycleConfigArns;
        }
        if (this._customImage) {
            hasAnyValues = true;
            internalValueResult.customImage = this._customImage;
        }
        if ((_k = this._defaultResourceSpec) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.defaultResourceSpec = (_l = this._defaultResourceSpec) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lifecycleConfigArns = undefined;
            this._customImage = undefined;
            this._defaultResourceSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lifecycleConfigArns = value.lifecycleConfigArns;
            this._customImage = value.customImage;
            this._defaultResourceSpec.internalValue = value.defaultResourceSpec;
        }
    }
    /**
     * @stability stable
     */
    get lifecycleConfigArns() {
        return this.getListAttribute('lifecycle_config_arns');
    }
    /**
     * @stability stable
     */
    set lifecycleConfigArns(value) {
        this._lifecycleConfigArns = value;
    }
    /**
     * @stability stable
     */
    resetLifecycleConfigArns() {
        this._lifecycleConfigArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lifecycleConfigArnsInput() {
        return this._lifecycleConfigArns;
    }
    /**
     * @stability stable
     */
    get customImage() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('custom_image');
    }
    /**
     * @stability stable
     */
    set customImage(value) {
        this._customImage = value;
    }
    /**
     * @stability stable
     */
    resetCustomImage() {
        this._customImage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customImageInput() {
        return this._customImage;
    }
    /**
     * @stability stable
     */
    get defaultResourceSpec() {
        return this._defaultResourceSpec;
    }
    /**
     * @stability stable
     */
    putDefaultResourceSpec(value) {
        this._defaultResourceSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultResourceSpecInput() {
        return this._defaultResourceSpec.internalValue;
    }
}
exports.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference = SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference[_d] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsSharingSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        notebook_output_option: cdktf.stringToTerraform(struct.notebookOutputOption),
        s3_kms_key_id: cdktf.stringToTerraform(struct.s3KmsKeyId),
        s3_output_path: cdktf.stringToTerraform(struct.s3OutputPath),
    };
}
exports.sagemakerUserProfileUserSettingsSharingSettingsToTerraform = sagemakerUserProfileUserSettingsSharingSettingsToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsSharingSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._notebookOutputOption) {
            hasAnyValues = true;
            internalValueResult.notebookOutputOption = this._notebookOutputOption;
        }
        if (this._s3KmsKeyId) {
            hasAnyValues = true;
            internalValueResult.s3KmsKeyId = this._s3KmsKeyId;
        }
        if (this._s3OutputPath) {
            hasAnyValues = true;
            internalValueResult.s3OutputPath = this._s3OutputPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._notebookOutputOption = undefined;
            this._s3KmsKeyId = undefined;
            this._s3OutputPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._notebookOutputOption = value.notebookOutputOption;
            this._s3KmsKeyId = value.s3KmsKeyId;
            this._s3OutputPath = value.s3OutputPath;
        }
    }
    /**
     * @stability stable
     */
    get notebookOutputOption() {
        return this.getStringAttribute('notebook_output_option');
    }
    /**
     * @stability stable
     */
    set notebookOutputOption(value) {
        this._notebookOutputOption = value;
    }
    /**
     * @stability stable
     */
    resetNotebookOutputOption() {
        this._notebookOutputOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notebookOutputOptionInput() {
        return this._notebookOutputOption;
    }
    /**
     * @stability stable
     */
    get s3KmsKeyId() {
        return this.getStringAttribute('s3_kms_key_id');
    }
    /**
     * @stability stable
     */
    set s3KmsKeyId(value) {
        this._s3KmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetS3KmsKeyId() {
        this._s3KmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KmsKeyIdInput() {
        return this._s3KmsKeyId;
    }
    /**
     * @stability stable
     */
    get s3OutputPath() {
        return this.getStringAttribute('s3_output_path');
    }
    /**
     * @stability stable
     */
    set s3OutputPath(value) {
        this._s3OutputPath = value;
    }
    /**
     * @stability stable
     */
    resetS3OutputPath() {
        this._s3OutputPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3OutputPathInput() {
        return this._s3OutputPath;
    }
}
exports.SagemakerUserProfileUserSettingsSharingSettingsOutputReference = SagemakerUserProfileUserSettingsSharingSettingsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsSharingSettingsOutputReference[_e] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsSharingSettingsOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        sagemaker_image_arn: cdktf.stringToTerraform(struct.sagemakerImageArn),
    };
}
exports.sagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecToTerraform = sagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceType) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._sagemakerImageArn) {
            hasAnyValues = true;
            internalValueResult.sagemakerImageArn = this._sagemakerImageArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instanceType = undefined;
            this._sagemakerImageArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instanceType = value.instanceType;
            this._sagemakerImageArn = value.sagemakerImageArn;
        }
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get sagemakerImageArn() {
        return this.getStringAttribute('sagemaker_image_arn');
    }
    /**
     * @stability stable
     */
    set sagemakerImageArn(value) {
        this._sagemakerImageArn = value;
    }
    /**
     * @stability stable
     */
    resetSagemakerImageArn() {
        this._sagemakerImageArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sagemakerImageArnInput() {
        return this._sagemakerImageArn;
    }
}
exports.SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference = SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference[_f] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsTensorBoardAppSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_resource_spec: sagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecToTerraform(struct.defaultResourceSpec),
    };
}
exports.sagemakerUserProfileUserSettingsTensorBoardAppSettingsToTerraform = sagemakerUserProfileUserSettingsTensorBoardAppSettingsToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // default_resource_spec - computed: false, optional: false, required: true
        this._defaultResourceSpec = new SagemakerUserProfileUserSettingsTensorBoardAppSettingsDefaultResourceSpecOutputReference(this, "default_resource_spec", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_k = this._defaultResourceSpec) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.defaultResourceSpec = (_l = this._defaultResourceSpec) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultResourceSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultResourceSpec.internalValue = value.defaultResourceSpec;
        }
    }
    /**
     * @stability stable
     */
    get defaultResourceSpec() {
        return this._defaultResourceSpec;
    }
    /**
     * @stability stable
     */
    putDefaultResourceSpec(value) {
        this._defaultResourceSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultResourceSpecInput() {
        return this._defaultResourceSpec.internalValue;
    }
}
exports.SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference = SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference[_g] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference", version: "3.0.1" };
function sagemakerUserProfileUserSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execution_role: cdktf.stringToTerraform(struct.executionRole),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroups),
        jupyter_server_app_settings: sagemakerUserProfileUserSettingsJupyterServerAppSettingsToTerraform(struct.jupyterServerAppSettings),
        kernel_gateway_app_settings: sagemakerUserProfileUserSettingsKernelGatewayAppSettingsToTerraform(struct.kernelGatewayAppSettings),
        sharing_settings: sagemakerUserProfileUserSettingsSharingSettingsToTerraform(struct.sharingSettings),
        tensor_board_app_settings: sagemakerUserProfileUserSettingsTensorBoardAppSettingsToTerraform(struct.tensorBoardAppSettings),
    };
}
exports.sagemakerUserProfileUserSettingsToTerraform = sagemakerUserProfileUserSettingsToTerraform;
/**
 * @stability stable
 */
class SagemakerUserProfileUserSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // jupyter_server_app_settings - computed: false, optional: true, required: false
        this._jupyterServerAppSettings = new SagemakerUserProfileUserSettingsJupyterServerAppSettingsOutputReference(this, "jupyter_server_app_settings", true);
        // kernel_gateway_app_settings - computed: false, optional: true, required: false
        this._kernelGatewayAppSettings = new SagemakerUserProfileUserSettingsKernelGatewayAppSettingsOutputReference(this, "kernel_gateway_app_settings", true);
        // sharing_settings - computed: false, optional: true, required: false
        this._sharingSettings = new SagemakerUserProfileUserSettingsSharingSettingsOutputReference(this, "sharing_settings", true);
        // tensor_board_app_settings - computed: false, optional: true, required: false
        this._tensorBoardAppSettings = new SagemakerUserProfileUserSettingsTensorBoardAppSettingsOutputReference(this, "tensor_board_app_settings", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l, _m, _o, _p, _q, _r, _s;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._executionRole) {
            hasAnyValues = true;
            internalValueResult.executionRole = this._executionRole;
        }
        if (this._securityGroups) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if ((_k = this._jupyterServerAppSettings) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.jupyterServerAppSettings = (_l = this._jupyterServerAppSettings) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        if ((_m = this._kernelGatewayAppSettings) === null || _m === void 0 ? void 0 : _m.internalValue) {
            hasAnyValues = true;
            internalValueResult.kernelGatewayAppSettings = (_o = this._kernelGatewayAppSettings) === null || _o === void 0 ? void 0 : _o.internalValue;
        }
        if ((_p = this._sharingSettings) === null || _p === void 0 ? void 0 : _p.internalValue) {
            hasAnyValues = true;
            internalValueResult.sharingSettings = (_q = this._sharingSettings) === null || _q === void 0 ? void 0 : _q.internalValue;
        }
        if ((_r = this._tensorBoardAppSettings) === null || _r === void 0 ? void 0 : _r.internalValue) {
            hasAnyValues = true;
            internalValueResult.tensorBoardAppSettings = (_s = this._tensorBoardAppSettings) === null || _s === void 0 ? void 0 : _s.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._executionRole = undefined;
            this._securityGroups = undefined;
            this._jupyterServerAppSettings.internalValue = undefined;
            this._kernelGatewayAppSettings.internalValue = undefined;
            this._sharingSettings.internalValue = undefined;
            this._tensorBoardAppSettings.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._executionRole = value.executionRole;
            this._securityGroups = value.securityGroups;
            this._jupyterServerAppSettings.internalValue = value.jupyterServerAppSettings;
            this._kernelGatewayAppSettings.internalValue = value.kernelGatewayAppSettings;
            this._sharingSettings.internalValue = value.sharingSettings;
            this._tensorBoardAppSettings.internalValue = value.tensorBoardAppSettings;
        }
    }
    /**
     * @stability stable
     */
    get executionRole() {
        return this.getStringAttribute('execution_role');
    }
    /**
     * @stability stable
     */
    set executionRole(value) {
        this._executionRole = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionRoleInput() {
        return this._executionRole;
    }
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    /**
     * @stability stable
     */
    set securityGroups(value) {
        this._securityGroups = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupsInput() {
        return this._securityGroups;
    }
    /**
     * @stability stable
     */
    get jupyterServerAppSettings() {
        return this._jupyterServerAppSettings;
    }
    /**
     * @stability stable
     */
    putJupyterServerAppSettings(value) {
        this._jupyterServerAppSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetJupyterServerAppSettings() {
        this._jupyterServerAppSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jupyterServerAppSettingsInput() {
        return this._jupyterServerAppSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get kernelGatewayAppSettings() {
        return this._kernelGatewayAppSettings;
    }
    /**
     * @stability stable
     */
    putKernelGatewayAppSettings(value) {
        this._kernelGatewayAppSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetKernelGatewayAppSettings() {
        this._kernelGatewayAppSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kernelGatewayAppSettingsInput() {
        return this._kernelGatewayAppSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get sharingSettings() {
        return this._sharingSettings;
    }
    /**
     * @stability stable
     */
    putSharingSettings(value) {
        this._sharingSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSharingSettings() {
        this._sharingSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sharingSettingsInput() {
        return this._sharingSettings.internalValue;
    }
    /**
     * @stability stable
     */
    get tensorBoardAppSettings() {
        return this._tensorBoardAppSettings;
    }
    /**
     * @stability stable
     */
    putTensorBoardAppSettings(value) {
        this._tensorBoardAppSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTensorBoardAppSettings() {
        this._tensorBoardAppSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tensorBoardAppSettingsInput() {
        return this._tensorBoardAppSettings.internalValue;
    }
}
exports.SagemakerUserProfileUserSettingsOutputReference = SagemakerUserProfileUserSettingsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SagemakerUserProfileUserSettingsOutputReference[_h] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfileUserSettingsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_user_profile.html aws_sagemaker_user_profile}.
 *
 * @stability stable
 */
class SagemakerUserProfile extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_user_profile.html aws_sagemaker_user_profile} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_user_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // user_settings - computed: false, optional: true, required: false
        this._userSettings = new SagemakerUserProfileUserSettingsOutputReference(this, "user_settings", true);
        this._domainId = config.domainId;
        this._singleSignOnUserIdentifier = config.singleSignOnUserIdentifier;
        this._singleSignOnUserValue = config.singleSignOnUserValue;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._userProfileName = config.userProfileName;
        this._userSettings.internalValue = config.userSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    /**
     * @stability stable
     */
    set domainId(value) {
        this._domainId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainIdInput() {
        return this._domainId;
    }
    // home_efs_file_system_uid - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get homeEfsFileSystemUid() {
        return this.getStringAttribute('home_efs_file_system_uid');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get singleSignOnUserIdentifier() {
        return this.getStringAttribute('single_sign_on_user_identifier');
    }
    /**
     * @stability stable
     */
    set singleSignOnUserIdentifier(value) {
        this._singleSignOnUserIdentifier = value;
    }
    /**
     * @stability stable
     */
    resetSingleSignOnUserIdentifier() {
        this._singleSignOnUserIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get singleSignOnUserIdentifierInput() {
        return this._singleSignOnUserIdentifier;
    }
    /**
     * @stability stable
     */
    get singleSignOnUserValue() {
        return this.getStringAttribute('single_sign_on_user_value');
    }
    /**
     * @stability stable
     */
    set singleSignOnUserValue(value) {
        this._singleSignOnUserValue = value;
    }
    /**
     * @stability stable
     */
    resetSingleSignOnUserValue() {
        this._singleSignOnUserValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get singleSignOnUserValueInput() {
        return this._singleSignOnUserValue;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get userProfileName() {
        return this.getStringAttribute('user_profile_name');
    }
    /**
     * @stability stable
     */
    set userProfileName(value) {
        this._userProfileName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userProfileNameInput() {
        return this._userProfileName;
    }
    /**
     * @stability stable
     */
    get userSettings() {
        return this._userSettings;
    }
    /**
     * @stability stable
     */
    putUserSettings(value) {
        this._userSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetUserSettings() {
        this._userSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userSettingsInput() {
        return this._userSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            domain_id: cdktf.stringToTerraform(this._domainId),
            single_sign_on_user_identifier: cdktf.stringToTerraform(this._singleSignOnUserIdentifier),
            single_sign_on_user_value: cdktf.stringToTerraform(this._singleSignOnUserValue),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            user_profile_name: cdktf.stringToTerraform(this._userProfileName),
            user_settings: sagemakerUserProfileUserSettingsToTerraform(this._userSettings.internalValue),
        };
    }
}
exports.SagemakerUserProfile = SagemakerUserProfile;
_j = JSII_RTTI_SYMBOL_1;
SagemakerUserProfile[_j] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerUserProfile", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SagemakerUserProfile.tfResourceType = "aws_sagemaker_user_profile";
