// @cdktf/provider-aws/lib/elastictranscoder/elastictranscoder-pipeline.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElastictranscoderPipeline = exports.elastictranscoderPipelineThumbnailConfigPermissionsToTerraform = exports.ElastictranscoderPipelineThumbnailConfigOutputReference = exports.elastictranscoderPipelineThumbnailConfigToTerraform = exports.ElastictranscoderPipelineNotificationsOutputReference = exports.elastictranscoderPipelineNotificationsToTerraform = exports.elastictranscoderPipelineContentConfigPermissionsToTerraform = exports.ElastictranscoderPipelineContentConfigOutputReference = exports.elastictranscoderPipelineContentConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elastictranscoderPipelineContentConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
    };
}
exports.elastictranscoderPipelineContentConfigToTerraform = elastictranscoderPipelineContentConfigToTerraform;
/**
 * @stability stable
 */
class ElastictranscoderPipelineContentConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._storageClass = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
        }
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    /**
     * @stability stable
     */
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    /**
     * @stability stable
     */
    set storageClass(value) {
        this._storageClass = value;
    }
    /**
     * @stability stable
     */
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageClassInput() {
        return this._storageClass;
    }
}
exports.ElastictranscoderPipelineContentConfigOutputReference = ElastictranscoderPipelineContentConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineContentConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.elastictranscoder.ElastictranscoderPipelineContentConfigOutputReference", version: "3.0.1" };
function elastictranscoderPipelineContentConfigPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access: cdktf.listMapper(cdktf.stringToTerraform)(struct.access),
        grantee: cdktf.stringToTerraform(struct.grantee),
        grantee_type: cdktf.stringToTerraform(struct.granteeType),
    };
}
exports.elastictranscoderPipelineContentConfigPermissionsToTerraform = elastictranscoderPipelineContentConfigPermissionsToTerraform;
function elastictranscoderPipelineNotificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        completed: cdktf.stringToTerraform(struct.completed),
        error: cdktf.stringToTerraform(struct.error),
        progressing: cdktf.stringToTerraform(struct.progressing),
        warning: cdktf.stringToTerraform(struct.warning),
    };
}
exports.elastictranscoderPipelineNotificationsToTerraform = elastictranscoderPipelineNotificationsToTerraform;
/**
 * @stability stable
 */
class ElastictranscoderPipelineNotificationsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._completed) {
            hasAnyValues = true;
            internalValueResult.completed = this._completed;
        }
        if (this._error) {
            hasAnyValues = true;
            internalValueResult.error = this._error;
        }
        if (this._progressing) {
            hasAnyValues = true;
            internalValueResult.progressing = this._progressing;
        }
        if (this._warning) {
            hasAnyValues = true;
            internalValueResult.warning = this._warning;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._completed = undefined;
            this._error = undefined;
            this._progressing = undefined;
            this._warning = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._completed = value.completed;
            this._error = value.error;
            this._progressing = value.progressing;
            this._warning = value.warning;
        }
    }
    /**
     * @stability stable
     */
    get completed() {
        return this.getStringAttribute('completed');
    }
    /**
     * @stability stable
     */
    set completed(value) {
        this._completed = value;
    }
    /**
     * @stability stable
     */
    resetCompleted() {
        this._completed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get completedInput() {
        return this._completed;
    }
    /**
     * @stability stable
     */
    get error() {
        return this.getStringAttribute('error');
    }
    /**
     * @stability stable
     */
    set error(value) {
        this._error = value;
    }
    /**
     * @stability stable
     */
    resetError() {
        this._error = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get errorInput() {
        return this._error;
    }
    /**
     * @stability stable
     */
    get progressing() {
        return this.getStringAttribute('progressing');
    }
    /**
     * @stability stable
     */
    set progressing(value) {
        this._progressing = value;
    }
    /**
     * @stability stable
     */
    resetProgressing() {
        this._progressing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get progressingInput() {
        return this._progressing;
    }
    /**
     * @stability stable
     */
    get warning() {
        return this.getStringAttribute('warning');
    }
    /**
     * @stability stable
     */
    set warning(value) {
        this._warning = value;
    }
    /**
     * @stability stable
     */
    resetWarning() {
        this._warning = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get warningInput() {
        return this._warning;
    }
}
exports.ElastictranscoderPipelineNotificationsOutputReference = ElastictranscoderPipelineNotificationsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineNotificationsOutputReference[_b] = { fqn: "@cdktf/provider-aws.elastictranscoder.ElastictranscoderPipelineNotificationsOutputReference", version: "3.0.1" };
function elastictranscoderPipelineThumbnailConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
    };
}
exports.elastictranscoderPipelineThumbnailConfigToTerraform = elastictranscoderPipelineThumbnailConfigToTerraform;
/**
 * @stability stable
 */
class ElastictranscoderPipelineThumbnailConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._storageClass = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
        }
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    /**
     * @stability stable
     */
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    /**
     * @stability stable
     */
    set storageClass(value) {
        this._storageClass = value;
    }
    /**
     * @stability stable
     */
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageClassInput() {
        return this._storageClass;
    }
}
exports.ElastictranscoderPipelineThumbnailConfigOutputReference = ElastictranscoderPipelineThumbnailConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineThumbnailConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.elastictranscoder.ElastictranscoderPipelineThumbnailConfigOutputReference", version: "3.0.1" };
function elastictranscoderPipelineThumbnailConfigPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access: cdktf.listMapper(cdktf.stringToTerraform)(struct.access),
        grantee: cdktf.stringToTerraform(struct.grantee),
        grantee_type: cdktf.stringToTerraform(struct.granteeType),
    };
}
exports.elastictranscoderPipelineThumbnailConfigPermissionsToTerraform = elastictranscoderPipelineThumbnailConfigPermissionsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/elastictranscoder_pipeline.html aws_elastictranscoder_pipeline}.
 *
 * @stability stable
 */
class ElastictranscoderPipeline extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elastictranscoder_pipeline.html aws_elastictranscoder_pipeline} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastictranscoder_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // content_config - computed: false, optional: true, required: false
        this._contentConfig = new ElastictranscoderPipelineContentConfigOutputReference(this, "content_config", true);
        // notifications - computed: false, optional: true, required: false
        this._notifications = new ElastictranscoderPipelineNotificationsOutputReference(this, "notifications", true);
        // thumbnail_config - computed: false, optional: true, required: false
        this._thumbnailConfig = new ElastictranscoderPipelineThumbnailConfigOutputReference(this, "thumbnail_config", true);
        this._awsKmsKeyArn = config.awsKmsKeyArn;
        this._inputBucket = config.inputBucket;
        this._name = config.name;
        this._outputBucket = config.outputBucket;
        this._role = config.role;
        this._contentConfig.internalValue = config.contentConfig;
        this._contentConfigPermissions = config.contentConfigPermissions;
        this._notifications.internalValue = config.notifications;
        this._thumbnailConfig.internalValue = config.thumbnailConfig;
        this._thumbnailConfigPermissions = config.thumbnailConfigPermissions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get awsKmsKeyArn() {
        return this.getStringAttribute('aws_kms_key_arn');
    }
    /**
     * @stability stable
     */
    set awsKmsKeyArn(value) {
        this._awsKmsKeyArn = value;
    }
    /**
     * @stability stable
     */
    resetAwsKmsKeyArn() {
        this._awsKmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get awsKmsKeyArnInput() {
        return this._awsKmsKeyArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get inputBucket() {
        return this.getStringAttribute('input_bucket');
    }
    /**
     * @stability stable
     */
    set inputBucket(value) {
        this._inputBucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputBucketInput() {
        return this._inputBucket;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get outputBucket() {
        return this.getStringAttribute('output_bucket');
    }
    /**
     * @stability stable
     */
    set outputBucket(value) {
        this._outputBucket = value;
    }
    /**
     * @stability stable
     */
    resetOutputBucket() {
        this._outputBucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputBucketInput() {
        return this._outputBucket;
    }
    /**
     * @stability stable
     */
    get role() {
        return this.getStringAttribute('role');
    }
    /**
     * @stability stable
     */
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleInput() {
        return this._role;
    }
    /**
     * @stability stable
     */
    get contentConfig() {
        return this._contentConfig;
    }
    /**
     * @stability stable
     */
    putContentConfig(value) {
        this._contentConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetContentConfig() {
        this._contentConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentConfigInput() {
        return this._contentConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get contentConfigPermissions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('content_config_permissions');
    }
    /**
     * @stability stable
     */
    set contentConfigPermissions(value) {
        this._contentConfigPermissions = value;
    }
    /**
     * @stability stable
     */
    resetContentConfigPermissions() {
        this._contentConfigPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentConfigPermissionsInput() {
        return this._contentConfigPermissions;
    }
    /**
     * @stability stable
     */
    get notifications() {
        return this._notifications;
    }
    /**
     * @stability stable
     */
    putNotifications(value) {
        this._notifications.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNotifications() {
        this._notifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationsInput() {
        return this._notifications.internalValue;
    }
    /**
     * @stability stable
     */
    get thumbnailConfig() {
        return this._thumbnailConfig;
    }
    /**
     * @stability stable
     */
    putThumbnailConfig(value) {
        this._thumbnailConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetThumbnailConfig() {
        this._thumbnailConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thumbnailConfigInput() {
        return this._thumbnailConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get thumbnailConfigPermissions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('thumbnail_config_permissions');
    }
    /**
     * @stability stable
     */
    set thumbnailConfigPermissions(value) {
        this._thumbnailConfigPermissions = value;
    }
    /**
     * @stability stable
     */
    resetThumbnailConfigPermissions() {
        this._thumbnailConfigPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thumbnailConfigPermissionsInput() {
        return this._thumbnailConfigPermissions;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            aws_kms_key_arn: cdktf.stringToTerraform(this._awsKmsKeyArn),
            input_bucket: cdktf.stringToTerraform(this._inputBucket),
            name: cdktf.stringToTerraform(this._name),
            output_bucket: cdktf.stringToTerraform(this._outputBucket),
            role: cdktf.stringToTerraform(this._role),
            content_config: elastictranscoderPipelineContentConfigToTerraform(this._contentConfig.internalValue),
            content_config_permissions: cdktf.listMapper(elastictranscoderPipelineContentConfigPermissionsToTerraform)(this._contentConfigPermissions),
            notifications: elastictranscoderPipelineNotificationsToTerraform(this._notifications.internalValue),
            thumbnail_config: elastictranscoderPipelineThumbnailConfigToTerraform(this._thumbnailConfig.internalValue),
            thumbnail_config_permissions: cdktf.listMapper(elastictranscoderPipelineThumbnailConfigPermissionsToTerraform)(this._thumbnailConfigPermissions),
        };
    }
}
exports.ElastictranscoderPipeline = ElastictranscoderPipeline;
_d = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipeline[_d] = { fqn: "@cdktf/provider-aws.elastictranscoder.ElastictranscoderPipeline", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ElastictranscoderPipeline.tfResourceType = "aws_elastictranscoder_pipeline";
