// @cdktf/provider-aws/lib/eventbridge/cloudwatch-event-bus-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventBusPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_bus_policy.html aws_cloudwatch_event_bus_policy}.
 *
 * @stability stable
 */
class CloudwatchEventBusPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_bus_policy.html aws_cloudwatch_event_bus_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_bus_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._eventBusName = config.eventBusName;
        this._policy = config.policy;
    }
    /**
     * @stability stable
     */
    get eventBusName() {
        return this.getStringAttribute('event_bus_name');
    }
    /**
     * @stability stable
     */
    set eventBusName(value) {
        this._eventBusName = value;
    }
    /**
     * @stability stable
     */
    resetEventBusName() {
        this._eventBusName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get eventBusNameInput() {
        return this._eventBusName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    /**
     * @stability stable
     */
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyInput() {
        return this._policy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            event_bus_name: cdktf.stringToTerraform(this._eventBusName),
            policy: cdktf.stringToTerraform(this._policy),
        };
    }
}
exports.CloudwatchEventBusPolicy = CloudwatchEventBusPolicy;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventBusPolicy[_a] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventBusPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudwatchEventBusPolicy.tfResourceType = "aws_cloudwatch_event_bus_policy";
