// @cdktf/provider-aws/lib/elasticbeanstalk/elastic-beanstalk-configuration-template.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticBeanstalkConfigurationTemplate = exports.elasticBeanstalkConfigurationTemplateSettingToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elasticBeanstalkConfigurationTemplateSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        namespace: cdktf.stringToTerraform(struct.namespace),
        resource: cdktf.stringToTerraform(struct.resource),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.elasticBeanstalkConfigurationTemplateSettingToTerraform = elasticBeanstalkConfigurationTemplateSettingToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/elastic_beanstalk_configuration_template.html aws_elastic_beanstalk_configuration_template}.
 *
 * @stability stable
 */
class ElasticBeanstalkConfigurationTemplate extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elastic_beanstalk_configuration_template.html aws_elastic_beanstalk_configuration_template} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastic_beanstalk_configuration_template',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._application = config.application;
        this._description = config.description;
        this._environmentId = config.environmentId;
        this._name = config.name;
        this._solutionStackName = config.solutionStackName;
        this._setting = config.setting;
    }
    /**
     * @stability stable
     */
    get application() {
        return this.getStringAttribute('application');
    }
    /**
     * @stability stable
     */
    set application(value) {
        this._application = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applicationInput() {
        return this._application;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get environmentId() {
        return this.getStringAttribute('environment_id');
    }
    /**
     * @stability stable
     */
    set environmentId(value) {
        this._environmentId = value;
    }
    /**
     * @stability stable
     */
    resetEnvironmentId() {
        this._environmentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentIdInput() {
        return this._environmentId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get solutionStackName() {
        return this.getStringAttribute('solution_stack_name');
    }
    /**
     * @stability stable
     */
    set solutionStackName(value) {
        this._solutionStackName = value;
    }
    /**
     * @stability stable
     */
    resetSolutionStackName() {
        this._solutionStackName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get solutionStackNameInput() {
        return this._solutionStackName;
    }
    /**
     * @stability stable
     */
    get setting() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('setting');
    }
    /**
     * @stability stable
     */
    set setting(value) {
        this._setting = value;
    }
    /**
     * @stability stable
     */
    resetSetting() {
        this._setting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get settingInput() {
        return this._setting;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            application: cdktf.stringToTerraform(this._application),
            description: cdktf.stringToTerraform(this._description),
            environment_id: cdktf.stringToTerraform(this._environmentId),
            name: cdktf.stringToTerraform(this._name),
            solution_stack_name: cdktf.stringToTerraform(this._solutionStackName),
            setting: cdktf.listMapper(elasticBeanstalkConfigurationTemplateSettingToTerraform)(this._setting),
        };
    }
}
exports.ElasticBeanstalkConfigurationTemplate = ElasticBeanstalkConfigurationTemplate;
_a = JSII_RTTI_SYMBOL_1;
ElasticBeanstalkConfigurationTemplate[_a] = { fqn: "@cdktf/provider-aws.elasticbeanstalk.ElasticBeanstalkConfigurationTemplate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ElasticBeanstalkConfigurationTemplate.tfResourceType = "aws_elastic_beanstalk_configuration_template";
