// @cdktf/provider-aws/lib/ec2/ec2-client-vpn-route.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2ClientVpnRoute = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_client_vpn_route.html aws_ec2_client_vpn_route}.
 *
 * @stability stable
 */
class Ec2ClientVpnRoute extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_client_vpn_route.html aws_ec2_client_vpn_route} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_client_vpn_route',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clientVpnEndpointId = config.clientVpnEndpointId;
        this._description = config.description;
        this._destinationCidrBlock = config.destinationCidrBlock;
        this._targetVpcSubnetId = config.targetVpcSubnetId;
    }
    /**
     * @stability stable
     */
    get clientVpnEndpointId() {
        return this.getStringAttribute('client_vpn_endpoint_id');
    }
    /**
     * @stability stable
     */
    set clientVpnEndpointId(value) {
        this._clientVpnEndpointId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientVpnEndpointIdInput() {
        return this._clientVpnEndpointId;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get destinationCidrBlock() {
        return this.getStringAttribute('destination_cidr_block');
    }
    /**
     * @stability stable
     */
    set destinationCidrBlock(value) {
        this._destinationCidrBlock = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationCidrBlockInput() {
        return this._destinationCidrBlock;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // origin - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get origin() {
        return this.getStringAttribute('origin');
    }
    /**
     * @stability stable
     */
    get targetVpcSubnetId() {
        return this.getStringAttribute('target_vpc_subnet_id');
    }
    /**
     * @stability stable
     */
    set targetVpcSubnetId(value) {
        this._targetVpcSubnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetVpcSubnetIdInput() {
        return this._targetVpcSubnetId;
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            client_vpn_endpoint_id: cdktf.stringToTerraform(this._clientVpnEndpointId),
            description: cdktf.stringToTerraform(this._description),
            destination_cidr_block: cdktf.stringToTerraform(this._destinationCidrBlock),
            target_vpc_subnet_id: cdktf.stringToTerraform(this._targetVpcSubnetId),
        };
    }
}
exports.Ec2ClientVpnRoute = Ec2ClientVpnRoute;
_a = JSII_RTTI_SYMBOL_1;
Ec2ClientVpnRoute[_a] = { fqn: "@cdktf/provider-aws.ec2.Ec2ClientVpnRoute", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Ec2ClientVpnRoute.tfResourceType = "aws_ec2_client_vpn_route";
