// @cdktf/provider-aws/lib/ec2/ec2-transit-gateway-prefix-list-reference.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2TransitGatewayPrefixListReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway_prefix_list_reference.html aws_ec2_transit_gateway_prefix_list_reference}.
 *
 * @stability stable
 */
class Ec2TransitGatewayPrefixListReference extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway_prefix_list_reference.html aws_ec2_transit_gateway_prefix_list_reference} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_transit_gateway_prefix_list_reference',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._blackhole = config.blackhole;
        this._prefixListId = config.prefixListId;
        this._transitGatewayAttachmentId = config.transitGatewayAttachmentId;
        this._transitGatewayRouteTableId = config.transitGatewayRouteTableId;
    }
    /**
     * @stability stable
     */
    get blackhole() {
        return this.getBooleanAttribute('blackhole');
    }
    /**
     * @stability stable
     */
    set blackhole(value) {
        this._blackhole = value;
    }
    /**
     * @stability stable
     */
    resetBlackhole() {
        this._blackhole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blackholeInput() {
        return this._blackhole;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get prefixListId() {
        return this.getStringAttribute('prefix_list_id');
    }
    /**
     * @stability stable
     */
    set prefixListId(value) {
        this._prefixListId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixListIdInput() {
        return this._prefixListId;
    }
    // prefix_list_owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get prefixListOwnerId() {
        return this.getStringAttribute('prefix_list_owner_id');
    }
    /**
     * @stability stable
     */
    get transitGatewayAttachmentId() {
        return this.getStringAttribute('transit_gateway_attachment_id');
    }
    /**
     * @stability stable
     */
    set transitGatewayAttachmentId(value) {
        this._transitGatewayAttachmentId = value;
    }
    /**
     * @stability stable
     */
    resetTransitGatewayAttachmentId() {
        this._transitGatewayAttachmentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get transitGatewayAttachmentIdInput() {
        return this._transitGatewayAttachmentId;
    }
    /**
     * @stability stable
     */
    get transitGatewayRouteTableId() {
        return this.getStringAttribute('transit_gateway_route_table_id');
    }
    /**
     * @stability stable
     */
    set transitGatewayRouteTableId(value) {
        this._transitGatewayRouteTableId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get transitGatewayRouteTableIdInput() {
        return this._transitGatewayRouteTableId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            blackhole: cdktf.booleanToTerraform(this._blackhole),
            prefix_list_id: cdktf.stringToTerraform(this._prefixListId),
            transit_gateway_attachment_id: cdktf.stringToTerraform(this._transitGatewayAttachmentId),
            transit_gateway_route_table_id: cdktf.stringToTerraform(this._transitGatewayRouteTableId),
        };
    }
}
exports.Ec2TransitGatewayPrefixListReference = Ec2TransitGatewayPrefixListReference;
_a = JSII_RTTI_SYMBOL_1;
Ec2TransitGatewayPrefixListReference[_a] = { fqn: "@cdktf/provider-aws.ec2.Ec2TransitGatewayPrefixListReference", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Ec2TransitGatewayPrefixListReference.tfResourceType = "aws_ec2_transit_gateway_prefix_list_reference";
