// @cdktf/provider-aws/lib/ec2/ec2-fleet.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2Fleet = exports.Ec2FleetTimeoutsOutputReference = exports.ec2FleetTimeoutsToTerraform = exports.Ec2FleetTargetCapacitySpecificationOutputReference = exports.ec2FleetTargetCapacitySpecificationToTerraform = exports.Ec2FleetSpotOptionsOutputReference = exports.ec2FleetSpotOptionsToTerraform = exports.Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference = exports.ec2FleetSpotOptionsMaintenanceStrategiesToTerraform = exports.Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference = exports.ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceToTerraform = exports.Ec2FleetOnDemandOptionsOutputReference = exports.ec2FleetOnDemandOptionsToTerraform = exports.Ec2FleetLaunchTemplateConfigOutputReference = exports.ec2FleetLaunchTemplateConfigToTerraform = exports.ec2FleetLaunchTemplateConfigOverrideToTerraform = exports.Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference = exports.ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        launch_template_id: cdktf.stringToTerraform(struct.launchTemplateId),
        launch_template_name: cdktf.stringToTerraform(struct.launchTemplateName),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationToTerraform = ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationToTerraform;
/**
 * @stability stable
 */
class Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._launchTemplateId) {
            hasAnyValues = true;
            internalValueResult.launchTemplateId = this._launchTemplateId;
        }
        if (this._launchTemplateName) {
            hasAnyValues = true;
            internalValueResult.launchTemplateName = this._launchTemplateName;
        }
        if (this._version) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._launchTemplateId = undefined;
            this._launchTemplateName = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._launchTemplateId = value.launchTemplateId;
            this._launchTemplateName = value.launchTemplateName;
            this._version = value.version;
        }
    }
    /**
     * @stability stable
     */
    get launchTemplateId() {
        return this.getStringAttribute('launch_template_id');
    }
    /**
     * @stability stable
     */
    set launchTemplateId(value) {
        this._launchTemplateId = value;
    }
    /**
     * @stability stable
     */
    resetLaunchTemplateId() {
        this._launchTemplateId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTemplateIdInput() {
        return this._launchTemplateId;
    }
    /**
     * @stability stable
     */
    get launchTemplateName() {
        return this.getStringAttribute('launch_template_name');
    }
    /**
     * @stability stable
     */
    set launchTemplateName(value) {
        this._launchTemplateName = value;
    }
    /**
     * @stability stable
     */
    resetLaunchTemplateName() {
        this._launchTemplateName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTemplateNameInput() {
        return this._launchTemplateName;
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
}
exports.Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference = Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference[_a] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference", version: "3.0.1" };
function ec2FleetLaunchTemplateConfigOverrideToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_zone: cdktf.stringToTerraform(struct.availabilityZone),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        max_price: cdktf.stringToTerraform(struct.maxPrice),
        priority: cdktf.numberToTerraform(struct.priority),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        weighted_capacity: cdktf.numberToTerraform(struct.weightedCapacity),
    };
}
exports.ec2FleetLaunchTemplateConfigOverrideToTerraform = ec2FleetLaunchTemplateConfigOverrideToTerraform;
function ec2FleetLaunchTemplateConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        launch_template_specification: ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationToTerraform(struct.launchTemplateSpecification),
        override: cdktf.listMapper(ec2FleetLaunchTemplateConfigOverrideToTerraform)(struct.override),
    };
}
exports.ec2FleetLaunchTemplateConfigToTerraform = ec2FleetLaunchTemplateConfigToTerraform;
/**
 * @stability stable
 */
class Ec2FleetLaunchTemplateConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // launch_template_specification - computed: false, optional: false, required: true
        this._launchTemplateSpecification = new Ec2FleetLaunchTemplateConfigLaunchTemplateSpecificationOutputReference(this, "launch_template_specification", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_k = this._launchTemplateSpecification) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.launchTemplateSpecification = (_l = this._launchTemplateSpecification) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        if (this._override) {
            hasAnyValues = true;
            internalValueResult.override = this._override;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._launchTemplateSpecification.internalValue = undefined;
            this._override = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._launchTemplateSpecification.internalValue = value.launchTemplateSpecification;
            this._override = value.override;
        }
    }
    /**
     * @stability stable
     */
    get launchTemplateSpecification() {
        return this._launchTemplateSpecification;
    }
    /**
     * @stability stable
     */
    putLaunchTemplateSpecification(value) {
        this._launchTemplateSpecification.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTemplateSpecificationInput() {
        return this._launchTemplateSpecification.internalValue;
    }
    /**
     * @stability stable
     */
    get override() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('override');
    }
    /**
     * @stability stable
     */
    set override(value) {
        this._override = value;
    }
    /**
     * @stability stable
     */
    resetOverride() {
        this._override = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get overrideInput() {
        return this._override;
    }
}
exports.Ec2FleetLaunchTemplateConfigOutputReference = Ec2FleetLaunchTemplateConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Ec2FleetLaunchTemplateConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetLaunchTemplateConfigOutputReference", version: "3.0.1" };
function ec2FleetOnDemandOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
    };
}
exports.ec2FleetOnDemandOptionsToTerraform = ec2FleetOnDemandOptionsToTerraform;
/**
 * @stability stable
 */
class Ec2FleetOnDemandOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allocationStrategy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allocationStrategy = value.allocationStrategy;
        }
    }
    /**
     * @stability stable
     */
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    /**
     * @stability stable
     */
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    /**
     * @stability stable
     */
    resetAllocationStrategy() {
        this._allocationStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
}
exports.Ec2FleetOnDemandOptionsOutputReference = Ec2FleetOnDemandOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Ec2FleetOnDemandOptionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetOnDemandOptionsOutputReference", version: "3.0.1" };
function ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        replacement_strategy: cdktf.stringToTerraform(struct.replacementStrategy),
    };
}
exports.ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceToTerraform = ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceToTerraform;
/**
 * @stability stable
 */
class Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._replacementStrategy) {
            hasAnyValues = true;
            internalValueResult.replacementStrategy = this._replacementStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._replacementStrategy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._replacementStrategy = value.replacementStrategy;
        }
    }
    /**
     * @stability stable
     */
    get replacementStrategy() {
        return this.getStringAttribute('replacement_strategy');
    }
    /**
     * @stability stable
     */
    set replacementStrategy(value) {
        this._replacementStrategy = value;
    }
    /**
     * @stability stable
     */
    resetReplacementStrategy() {
        this._replacementStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replacementStrategyInput() {
        return this._replacementStrategy;
    }
}
exports.Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference = Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference;
_d = JSII_RTTI_SYMBOL_1;
Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference[_d] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference", version: "3.0.1" };
function ec2FleetSpotOptionsMaintenanceStrategiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        capacity_rebalance: ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceToTerraform(struct.capacityRebalance),
    };
}
exports.ec2FleetSpotOptionsMaintenanceStrategiesToTerraform = ec2FleetSpotOptionsMaintenanceStrategiesToTerraform;
/**
 * @stability stable
 */
class Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // capacity_rebalance - computed: false, optional: true, required: false
        this._capacityRebalance = new Ec2FleetSpotOptionsMaintenanceStrategiesCapacityRebalanceOutputReference(this, "capacity_rebalance", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_k = this._capacityRebalance) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.capacityRebalance = (_l = this._capacityRebalance) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._capacityRebalance.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._capacityRebalance.internalValue = value.capacityRebalance;
        }
    }
    /**
     * @stability stable
     */
    get capacityRebalance() {
        return this._capacityRebalance;
    }
    /**
     * @stability stable
     */
    putCapacityRebalance(value) {
        this._capacityRebalance.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCapacityRebalance() {
        this._capacityRebalance.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityRebalanceInput() {
        return this._capacityRebalance.internalValue;
    }
}
exports.Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference = Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference[_e] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference", version: "3.0.1" };
function ec2FleetSpotOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
        instance_interruption_behavior: cdktf.stringToTerraform(struct.instanceInterruptionBehavior),
        instance_pools_to_use_count: cdktf.numberToTerraform(struct.instancePoolsToUseCount),
        maintenance_strategies: ec2FleetSpotOptionsMaintenanceStrategiesToTerraform(struct.maintenanceStrategies),
    };
}
exports.ec2FleetSpotOptionsToTerraform = ec2FleetSpotOptionsToTerraform;
/**
 * @stability stable
 */
class Ec2FleetSpotOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // maintenance_strategies - computed: false, optional: true, required: false
        this._maintenanceStrategies = new Ec2FleetSpotOptionsMaintenanceStrategiesOutputReference(this, "maintenance_strategies", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        if (this._instanceInterruptionBehavior) {
            hasAnyValues = true;
            internalValueResult.instanceInterruptionBehavior = this._instanceInterruptionBehavior;
        }
        if (this._instancePoolsToUseCount) {
            hasAnyValues = true;
            internalValueResult.instancePoolsToUseCount = this._instancePoolsToUseCount;
        }
        if ((_k = this._maintenanceStrategies) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.maintenanceStrategies = (_l = this._maintenanceStrategies) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allocationStrategy = undefined;
            this._instanceInterruptionBehavior = undefined;
            this._instancePoolsToUseCount = undefined;
            this._maintenanceStrategies.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allocationStrategy = value.allocationStrategy;
            this._instanceInterruptionBehavior = value.instanceInterruptionBehavior;
            this._instancePoolsToUseCount = value.instancePoolsToUseCount;
            this._maintenanceStrategies.internalValue = value.maintenanceStrategies;
        }
    }
    /**
     * @stability stable
     */
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    /**
     * @stability stable
     */
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    /**
     * @stability stable
     */
    resetAllocationStrategy() {
        this._allocationStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
    /**
     * @stability stable
     */
    get instanceInterruptionBehavior() {
        return this.getStringAttribute('instance_interruption_behavior');
    }
    /**
     * @stability stable
     */
    set instanceInterruptionBehavior(value) {
        this._instanceInterruptionBehavior = value;
    }
    /**
     * @stability stable
     */
    resetInstanceInterruptionBehavior() {
        this._instanceInterruptionBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceInterruptionBehaviorInput() {
        return this._instanceInterruptionBehavior;
    }
    /**
     * @stability stable
     */
    get instancePoolsToUseCount() {
        return this.getNumberAttribute('instance_pools_to_use_count');
    }
    /**
     * @stability stable
     */
    set instancePoolsToUseCount(value) {
        this._instancePoolsToUseCount = value;
    }
    /**
     * @stability stable
     */
    resetInstancePoolsToUseCount() {
        this._instancePoolsToUseCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instancePoolsToUseCountInput() {
        return this._instancePoolsToUseCount;
    }
    /**
     * @stability stable
     */
    get maintenanceStrategies() {
        return this._maintenanceStrategies;
    }
    /**
     * @stability stable
     */
    putMaintenanceStrategies(value) {
        this._maintenanceStrategies.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMaintenanceStrategies() {
        this._maintenanceStrategies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maintenanceStrategiesInput() {
        return this._maintenanceStrategies.internalValue;
    }
}
exports.Ec2FleetSpotOptionsOutputReference = Ec2FleetSpotOptionsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
Ec2FleetSpotOptionsOutputReference[_f] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetSpotOptionsOutputReference", version: "3.0.1" };
function ec2FleetTargetCapacitySpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_target_capacity_type: cdktf.stringToTerraform(struct.defaultTargetCapacityType),
        on_demand_target_capacity: cdktf.numberToTerraform(struct.onDemandTargetCapacity),
        spot_target_capacity: cdktf.numberToTerraform(struct.spotTargetCapacity),
        total_target_capacity: cdktf.numberToTerraform(struct.totalTargetCapacity),
    };
}
exports.ec2FleetTargetCapacitySpecificationToTerraform = ec2FleetTargetCapacitySpecificationToTerraform;
/**
 * @stability stable
 */
class Ec2FleetTargetCapacitySpecificationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultTargetCapacityType) {
            hasAnyValues = true;
            internalValueResult.defaultTargetCapacityType = this._defaultTargetCapacityType;
        }
        if (this._onDemandTargetCapacity) {
            hasAnyValues = true;
            internalValueResult.onDemandTargetCapacity = this._onDemandTargetCapacity;
        }
        if (this._spotTargetCapacity) {
            hasAnyValues = true;
            internalValueResult.spotTargetCapacity = this._spotTargetCapacity;
        }
        if (this._totalTargetCapacity) {
            hasAnyValues = true;
            internalValueResult.totalTargetCapacity = this._totalTargetCapacity;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultTargetCapacityType = undefined;
            this._onDemandTargetCapacity = undefined;
            this._spotTargetCapacity = undefined;
            this._totalTargetCapacity = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultTargetCapacityType = value.defaultTargetCapacityType;
            this._onDemandTargetCapacity = value.onDemandTargetCapacity;
            this._spotTargetCapacity = value.spotTargetCapacity;
            this._totalTargetCapacity = value.totalTargetCapacity;
        }
    }
    /**
     * @stability stable
     */
    get defaultTargetCapacityType() {
        return this.getStringAttribute('default_target_capacity_type');
    }
    /**
     * @stability stable
     */
    set defaultTargetCapacityType(value) {
        this._defaultTargetCapacityType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultTargetCapacityTypeInput() {
        return this._defaultTargetCapacityType;
    }
    /**
     * @stability stable
     */
    get onDemandTargetCapacity() {
        return this.getNumberAttribute('on_demand_target_capacity');
    }
    /**
     * @stability stable
     */
    set onDemandTargetCapacity(value) {
        this._onDemandTargetCapacity = value;
    }
    /**
     * @stability stable
     */
    resetOnDemandTargetCapacity() {
        this._onDemandTargetCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get onDemandTargetCapacityInput() {
        return this._onDemandTargetCapacity;
    }
    /**
     * @stability stable
     */
    get spotTargetCapacity() {
        return this.getNumberAttribute('spot_target_capacity');
    }
    /**
     * @stability stable
     */
    set spotTargetCapacity(value) {
        this._spotTargetCapacity = value;
    }
    /**
     * @stability stable
     */
    resetSpotTargetCapacity() {
        this._spotTargetCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spotTargetCapacityInput() {
        return this._spotTargetCapacity;
    }
    /**
     * @stability stable
     */
    get totalTargetCapacity() {
        return this.getNumberAttribute('total_target_capacity');
    }
    /**
     * @stability stable
     */
    set totalTargetCapacity(value) {
        this._totalTargetCapacity = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get totalTargetCapacityInput() {
        return this._totalTargetCapacity;
    }
}
exports.Ec2FleetTargetCapacitySpecificationOutputReference = Ec2FleetTargetCapacitySpecificationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
Ec2FleetTargetCapacitySpecificationOutputReference[_g] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetTargetCapacitySpecificationOutputReference", version: "3.0.1" };
function ec2FleetTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.ec2FleetTimeoutsToTerraform = ec2FleetTimeoutsToTerraform;
/**
 * @stability stable
 */
class Ec2FleetTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.Ec2FleetTimeoutsOutputReference = Ec2FleetTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
Ec2FleetTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-aws.ec2.Ec2FleetTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_fleet.html aws_ec2_fleet}.
 *
 * @stability stable
 */
class Ec2Fleet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_fleet.html aws_ec2_fleet} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_fleet',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // launch_template_config - computed: false, optional: false, required: true
        this._launchTemplateConfig = new Ec2FleetLaunchTemplateConfigOutputReference(this, "launch_template_config", true);
        // on_demand_options - computed: false, optional: true, required: false
        this._onDemandOptions = new Ec2FleetOnDemandOptionsOutputReference(this, "on_demand_options", true);
        // spot_options - computed: false, optional: true, required: false
        this._spotOptions = new Ec2FleetSpotOptionsOutputReference(this, "spot_options", true);
        // target_capacity_specification - computed: false, optional: false, required: true
        this._targetCapacitySpecification = new Ec2FleetTargetCapacitySpecificationOutputReference(this, "target_capacity_specification", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Ec2FleetTimeoutsOutputReference(this, "timeouts", true);
        this._excessCapacityTerminationPolicy = config.excessCapacityTerminationPolicy;
        this._replaceUnhealthyInstances = config.replaceUnhealthyInstances;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._terminateInstances = config.terminateInstances;
        this._terminateInstancesWithExpiration = config.terminateInstancesWithExpiration;
        this._type = config.type;
        this._launchTemplateConfig.internalValue = config.launchTemplateConfig;
        this._onDemandOptions.internalValue = config.onDemandOptions;
        this._spotOptions.internalValue = config.spotOptions;
        this._targetCapacitySpecification.internalValue = config.targetCapacitySpecification;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get excessCapacityTerminationPolicy() {
        return this.getStringAttribute('excess_capacity_termination_policy');
    }
    /**
     * @stability stable
     */
    set excessCapacityTerminationPolicy(value) {
        this._excessCapacityTerminationPolicy = value;
    }
    /**
     * @stability stable
     */
    resetExcessCapacityTerminationPolicy() {
        this._excessCapacityTerminationPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excessCapacityTerminationPolicyInput() {
        return this._excessCapacityTerminationPolicy;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get replaceUnhealthyInstances() {
        return this.getBooleanAttribute('replace_unhealthy_instances');
    }
    /**
     * @stability stable
     */
    set replaceUnhealthyInstances(value) {
        this._replaceUnhealthyInstances = value;
    }
    /**
     * @stability stable
     */
    resetReplaceUnhealthyInstances() {
        this._replaceUnhealthyInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replaceUnhealthyInstancesInput() {
        return this._replaceUnhealthyInstances;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get terminateInstances() {
        return this.getBooleanAttribute('terminate_instances');
    }
    /**
     * @stability stable
     */
    set terminateInstances(value) {
        this._terminateInstances = value;
    }
    /**
     * @stability stable
     */
    resetTerminateInstances() {
        this._terminateInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get terminateInstancesInput() {
        return this._terminateInstances;
    }
    /**
     * @stability stable
     */
    get terminateInstancesWithExpiration() {
        return this.getBooleanAttribute('terminate_instances_with_expiration');
    }
    /**
     * @stability stable
     */
    set terminateInstancesWithExpiration(value) {
        this._terminateInstancesWithExpiration = value;
    }
    /**
     * @stability stable
     */
    resetTerminateInstancesWithExpiration() {
        this._terminateInstancesWithExpiration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get terminateInstancesWithExpirationInput() {
        return this._terminateInstancesWithExpiration;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get launchTemplateConfig() {
        return this._launchTemplateConfig;
    }
    /**
     * @stability stable
     */
    putLaunchTemplateConfig(value) {
        this._launchTemplateConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTemplateConfigInput() {
        return this._launchTemplateConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get onDemandOptions() {
        return this._onDemandOptions;
    }
    /**
     * @stability stable
     */
    putOnDemandOptions(value) {
        this._onDemandOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetOnDemandOptions() {
        this._onDemandOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get onDemandOptionsInput() {
        return this._onDemandOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get spotOptions() {
        return this._spotOptions;
    }
    /**
     * @stability stable
     */
    putSpotOptions(value) {
        this._spotOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSpotOptions() {
        this._spotOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get spotOptionsInput() {
        return this._spotOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get targetCapacitySpecification() {
        return this._targetCapacitySpecification;
    }
    /**
     * @stability stable
     */
    putTargetCapacitySpecification(value) {
        this._targetCapacitySpecification.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetCapacitySpecificationInput() {
        return this._targetCapacitySpecification.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            excess_capacity_termination_policy: cdktf.stringToTerraform(this._excessCapacityTerminationPolicy),
            replace_unhealthy_instances: cdktf.booleanToTerraform(this._replaceUnhealthyInstances),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            terminate_instances: cdktf.booleanToTerraform(this._terminateInstances),
            terminate_instances_with_expiration: cdktf.booleanToTerraform(this._terminateInstancesWithExpiration),
            type: cdktf.stringToTerraform(this._type),
            launch_template_config: ec2FleetLaunchTemplateConfigToTerraform(this._launchTemplateConfig.internalValue),
            on_demand_options: ec2FleetOnDemandOptionsToTerraform(this._onDemandOptions.internalValue),
            spot_options: ec2FleetSpotOptionsToTerraform(this._spotOptions.internalValue),
            target_capacity_specification: ec2FleetTargetCapacitySpecificationToTerraform(this._targetCapacitySpecification.internalValue),
            timeouts: ec2FleetTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.Ec2Fleet = Ec2Fleet;
_j = JSII_RTTI_SYMBOL_1;
Ec2Fleet[_j] = { fqn: "@cdktf/provider-aws.ec2.Ec2Fleet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Ec2Fleet.tfResourceType = "aws_ec2_fleet";
