// @cdktf/provider-aws/lib/iam/data-aws-iam-server-certificate.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIamServerCertificate = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/iam_server_certificate.html aws_iam_server_certificate}.
 *
 * @stability stable
 */
class DataAwsIamServerCertificate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/iam_server_certificate.html aws_iam_server_certificate} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_server_certificate',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._latest = config.latest;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._pathPrefix = config.pathPrefix;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_body - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateBody() {
        return this.getStringAttribute('certificate_body');
    }
    // certificate_chain - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    // expiration_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get expirationDate() {
        return this.getStringAttribute('expiration_date');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get latest() {
        return this.getBooleanAttribute('latest');
    }
    /**
     * @stability stable
     */
    set latest(value) {
        this._latest = value;
    }
    /**
     * @stability stable
     */
    resetLatest() {
        this._latest = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get latestInput() {
        return this._latest;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    // path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    get pathPrefix() {
        return this.getStringAttribute('path_prefix');
    }
    /**
     * @stability stable
     */
    set pathPrefix(value) {
        this._pathPrefix = value;
    }
    /**
     * @stability stable
     */
    resetPathPrefix() {
        this._pathPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathPrefixInput() {
        return this._pathPrefix;
    }
    // upload_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get uploadDate() {
        return this.getStringAttribute('upload_date');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            latest: cdktf.booleanToTerraform(this._latest),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            path_prefix: cdktf.stringToTerraform(this._pathPrefix),
        };
    }
}
exports.DataAwsIamServerCertificate = DataAwsIamServerCertificate;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIamServerCertificate[_a] = { fqn: "@cdktf/provider-aws.iam.DataAwsIamServerCertificate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsIamServerCertificate.tfResourceType = "aws_iam_server_certificate";
