// @cdktf/provider-aws/lib/iam/iam-account-alias.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamAccountAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_account_alias.html aws_iam_account_alias}.
 *
 * @stability stable
 */
class IamAccountAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_account_alias.html aws_iam_account_alias} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_account_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accountAlias = config.accountAlias;
    }
    /**
     * @stability stable
     */
    get accountAlias() {
        return this.getStringAttribute('account_alias');
    }
    /**
     * @stability stable
     */
    set accountAlias(value) {
        this._accountAlias = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountAliasInput() {
        return this._accountAlias;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_alias: cdktf.stringToTerraform(this._accountAlias),
        };
    }
}
exports.IamAccountAlias = IamAccountAlias;
_a = JSII_RTTI_SYMBOL_1;
IamAccountAlias[_a] = { fqn: "@cdktf/provider-aws.iam.IamAccountAlias", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IamAccountAlias.tfResourceType = "aws_iam_account_alias";
