// @cdktf/provider-aws/lib/opsworks/opsworks-instance.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksInstance = exports.OpsworksInstanceTimeoutsOutputReference = exports.opsworksInstanceTimeoutsToTerraform = exports.opsworksInstanceRootBlockDeviceToTerraform = exports.opsworksInstanceEphemeralBlockDeviceToTerraform = exports.opsworksInstanceEbsBlockDeviceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function opsworksInstanceEbsBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_on_termination: cdktf.booleanToTerraform(struct.deleteOnTermination),
        device_name: cdktf.stringToTerraform(struct.deviceName),
        iops: cdktf.numberToTerraform(struct.iops),
        snapshot_id: cdktf.stringToTerraform(struct.snapshotId),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.opsworksInstanceEbsBlockDeviceToTerraform = opsworksInstanceEbsBlockDeviceToTerraform;
function opsworksInstanceEphemeralBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        device_name: cdktf.stringToTerraform(struct.deviceName),
        virtual_name: cdktf.stringToTerraform(struct.virtualName),
    };
}
exports.opsworksInstanceEphemeralBlockDeviceToTerraform = opsworksInstanceEphemeralBlockDeviceToTerraform;
function opsworksInstanceRootBlockDeviceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_on_termination: cdktf.booleanToTerraform(struct.deleteOnTermination),
        iops: cdktf.numberToTerraform(struct.iops),
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.opsworksInstanceRootBlockDeviceToTerraform = opsworksInstanceRootBlockDeviceToTerraform;
function opsworksInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.opsworksInstanceTimeoutsToTerraform = opsworksInstanceTimeoutsToTerraform;
/**
 * @stability stable
 */
class OpsworksInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.OpsworksInstanceTimeoutsOutputReference = OpsworksInstanceTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
OpsworksInstanceTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.opsworks.OpsworksInstanceTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/opsworks_instance.html aws_opsworks_instance}.
 *
 * @stability stable
 */
class OpsworksInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/opsworks_instance.html aws_opsworks_instance} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new OpsworksInstanceTimeoutsOutputReference(this, "timeouts", true);
        this._agentVersion = config.agentVersion;
        this._amiId = config.amiId;
        this._architecture = config.architecture;
        this._autoScalingType = config.autoScalingType;
        this._availabilityZone = config.availabilityZone;
        this._createdAt = config.createdAt;
        this._deleteEbs = config.deleteEbs;
        this._deleteEip = config.deleteEip;
        this._ebsOptimized = config.ebsOptimized;
        this._ecsClusterArn = config.ecsClusterArn;
        this._elasticIp = config.elasticIp;
        this._hostname = config.hostname;
        this._infrastructureClass = config.infrastructureClass;
        this._installUpdatesOnBoot = config.installUpdatesOnBoot;
        this._instanceProfileArn = config.instanceProfileArn;
        this._instanceType = config.instanceType;
        this._lastServiceErrorId = config.lastServiceErrorId;
        this._layerIds = config.layerIds;
        this._os = config.os;
        this._platform = config.platform;
        this._privateDns = config.privateDns;
        this._privateIp = config.privateIp;
        this._publicDns = config.publicDns;
        this._publicIp = config.publicIp;
        this._registeredBy = config.registeredBy;
        this._reportedAgentVersion = config.reportedAgentVersion;
        this._reportedOsFamily = config.reportedOsFamily;
        this._reportedOsName = config.reportedOsName;
        this._reportedOsVersion = config.reportedOsVersion;
        this._rootDeviceType = config.rootDeviceType;
        this._rootDeviceVolumeId = config.rootDeviceVolumeId;
        this._securityGroupIds = config.securityGroupIds;
        this._sshHostDsaKeyFingerprint = config.sshHostDsaKeyFingerprint;
        this._sshHostRsaKeyFingerprint = config.sshHostRsaKeyFingerprint;
        this._sshKeyName = config.sshKeyName;
        this._stackId = config.stackId;
        this._state = config.state;
        this._status = config.status;
        this._subnetId = config.subnetId;
        this._tenancy = config.tenancy;
        this._virtualizationType = config.virtualizationType;
        this._ebsBlockDevice = config.ebsBlockDevice;
        this._ephemeralBlockDevice = config.ephemeralBlockDevice;
        this._rootBlockDevice = config.rootBlockDevice;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get agentVersion() {
        return this.getStringAttribute('agent_version');
    }
    /**
     * @stability stable
     */
    set agentVersion(value) {
        this._agentVersion = value;
    }
    /**
     * @stability stable
     */
    resetAgentVersion() {
        this._agentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get agentVersionInput() {
        return this._agentVersion;
    }
    /**
     * @stability stable
     */
    get amiId() {
        return this.getStringAttribute('ami_id');
    }
    /**
     * @stability stable
     */
    set amiId(value) {
        this._amiId = value;
    }
    /**
     * @stability stable
     */
    resetAmiId() {
        this._amiId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get amiIdInput() {
        return this._amiId;
    }
    /**
     * @stability stable
     */
    get architecture() {
        return this.getStringAttribute('architecture');
    }
    /**
     * @stability stable
     */
    set architecture(value) {
        this._architecture = value;
    }
    /**
     * @stability stable
     */
    resetArchitecture() {
        this._architecture = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get architectureInput() {
        return this._architecture;
    }
    /**
     * @stability stable
     */
    get autoScalingType() {
        return this.getStringAttribute('auto_scaling_type');
    }
    /**
     * @stability stable
     */
    set autoScalingType(value) {
        this._autoScalingType = value;
    }
    /**
     * @stability stable
     */
    resetAutoScalingType() {
        this._autoScalingType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoScalingTypeInput() {
        return this._autoScalingType;
    }
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    /**
     * @stability stable
     */
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    /**
     * @stability stable
     */
    resetAvailabilityZone() {
        this._availabilityZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    /**
     * @stability stable
     */
    set createdAt(value) {
        this._createdAt = value;
    }
    /**
     * @stability stable
     */
    resetCreatedAt() {
        this._createdAt = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createdAtInput() {
        return this._createdAt;
    }
    /**
     * @stability stable
     */
    get deleteEbs() {
        return this.getBooleanAttribute('delete_ebs');
    }
    /**
     * @stability stable
     */
    set deleteEbs(value) {
        this._deleteEbs = value;
    }
    /**
     * @stability stable
     */
    resetDeleteEbs() {
        this._deleteEbs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteEbsInput() {
        return this._deleteEbs;
    }
    /**
     * @stability stable
     */
    get deleteEip() {
        return this.getBooleanAttribute('delete_eip');
    }
    /**
     * @stability stable
     */
    set deleteEip(value) {
        this._deleteEip = value;
    }
    /**
     * @stability stable
     */
    resetDeleteEip() {
        this._deleteEip = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteEipInput() {
        return this._deleteEip;
    }
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    /**
     * @stability stable
     */
    set ebsOptimized(value) {
        this._ebsOptimized = value;
    }
    /**
     * @stability stable
     */
    resetEbsOptimized() {
        this._ebsOptimized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsOptimizedInput() {
        return this._ebsOptimized;
    }
    // ec2_instance_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ec2InstanceId() {
        return this.getStringAttribute('ec2_instance_id');
    }
    /**
     * @stability stable
     */
    get ecsClusterArn() {
        return this.getStringAttribute('ecs_cluster_arn');
    }
    /**
     * @stability stable
     */
    set ecsClusterArn(value) {
        this._ecsClusterArn = value;
    }
    /**
     * @stability stable
     */
    resetEcsClusterArn() {
        this._ecsClusterArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ecsClusterArnInput() {
        return this._ecsClusterArn;
    }
    /**
     * @stability stable
     */
    get elasticIp() {
        return this.getStringAttribute('elastic_ip');
    }
    /**
     * @stability stable
     */
    set elasticIp(value) {
        this._elasticIp = value;
    }
    /**
     * @stability stable
     */
    resetElasticIp() {
        this._elasticIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticIpInput() {
        return this._elasticIp;
    }
    /**
     * @stability stable
     */
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    /**
     * @stability stable
     */
    set hostname(value) {
        this._hostname = value;
    }
    /**
     * @stability stable
     */
    resetHostname() {
        this._hostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostnameInput() {
        return this._hostname;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get infrastructureClass() {
        return this.getStringAttribute('infrastructure_class');
    }
    /**
     * @stability stable
     */
    set infrastructureClass(value) {
        this._infrastructureClass = value;
    }
    /**
     * @stability stable
     */
    resetInfrastructureClass() {
        this._infrastructureClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get infrastructureClassInput() {
        return this._infrastructureClass;
    }
    /**
     * @stability stable
     */
    get installUpdatesOnBoot() {
        return this.getBooleanAttribute('install_updates_on_boot');
    }
    /**
     * @stability stable
     */
    set installUpdatesOnBoot(value) {
        this._installUpdatesOnBoot = value;
    }
    /**
     * @stability stable
     */
    resetInstallUpdatesOnBoot() {
        this._installUpdatesOnBoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get installUpdatesOnBootInput() {
        return this._installUpdatesOnBoot;
    }
    /**
     * @stability stable
     */
    get instanceProfileArn() {
        return this.getStringAttribute('instance_profile_arn');
    }
    /**
     * @stability stable
     */
    set instanceProfileArn(value) {
        this._instanceProfileArn = value;
    }
    /**
     * @stability stable
     */
    resetInstanceProfileArn() {
        this._instanceProfileArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceProfileArnInput() {
        return this._instanceProfileArn;
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get lastServiceErrorId() {
        return this.getStringAttribute('last_service_error_id');
    }
    /**
     * @stability stable
     */
    set lastServiceErrorId(value) {
        this._lastServiceErrorId = value;
    }
    /**
     * @stability stable
     */
    resetLastServiceErrorId() {
        this._lastServiceErrorId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lastServiceErrorIdInput() {
        return this._lastServiceErrorId;
    }
    /**
     * @stability stable
     */
    get layerIds() {
        return this.getListAttribute('layer_ids');
    }
    /**
     * @stability stable
     */
    set layerIds(value) {
        this._layerIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get layerIdsInput() {
        return this._layerIds;
    }
    /**
     * @stability stable
     */
    get os() {
        return this.getStringAttribute('os');
    }
    /**
     * @stability stable
     */
    set os(value) {
        this._os = value;
    }
    /**
     * @stability stable
     */
    resetOs() {
        this._os = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get osInput() {
        return this._os;
    }
    /**
     * @stability stable
     */
    get platform() {
        return this.getStringAttribute('platform');
    }
    /**
     * @stability stable
     */
    set platform(value) {
        this._platform = value;
    }
    /**
     * @stability stable
     */
    resetPlatform() {
        this._platform = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get platformInput() {
        return this._platform;
    }
    /**
     * @stability stable
     */
    get privateDns() {
        return this.getStringAttribute('private_dns');
    }
    /**
     * @stability stable
     */
    set privateDns(value) {
        this._privateDns = value;
    }
    /**
     * @stability stable
     */
    resetPrivateDns() {
        this._privateDns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateDnsInput() {
        return this._privateDns;
    }
    /**
     * @stability stable
     */
    get privateIp() {
        return this.getStringAttribute('private_ip');
    }
    /**
     * @stability stable
     */
    set privateIp(value) {
        this._privateIp = value;
    }
    /**
     * @stability stable
     */
    resetPrivateIp() {
        this._privateIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateIpInput() {
        return this._privateIp;
    }
    /**
     * @stability stable
     */
    get publicDns() {
        return this.getStringAttribute('public_dns');
    }
    /**
     * @stability stable
     */
    set publicDns(value) {
        this._publicDns = value;
    }
    /**
     * @stability stable
     */
    resetPublicDns() {
        this._publicDns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicDnsInput() {
        return this._publicDns;
    }
    /**
     * @stability stable
     */
    get publicIp() {
        return this.getStringAttribute('public_ip');
    }
    /**
     * @stability stable
     */
    set publicIp(value) {
        this._publicIp = value;
    }
    /**
     * @stability stable
     */
    resetPublicIp() {
        this._publicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicIpInput() {
        return this._publicIp;
    }
    /**
     * @stability stable
     */
    get registeredBy() {
        return this.getStringAttribute('registered_by');
    }
    /**
     * @stability stable
     */
    set registeredBy(value) {
        this._registeredBy = value;
    }
    /**
     * @stability stable
     */
    resetRegisteredBy() {
        this._registeredBy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registeredByInput() {
        return this._registeredBy;
    }
    /**
     * @stability stable
     */
    get reportedAgentVersion() {
        return this.getStringAttribute('reported_agent_version');
    }
    /**
     * @stability stable
     */
    set reportedAgentVersion(value) {
        this._reportedAgentVersion = value;
    }
    /**
     * @stability stable
     */
    resetReportedAgentVersion() {
        this._reportedAgentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportedAgentVersionInput() {
        return this._reportedAgentVersion;
    }
    /**
     * @stability stable
     */
    get reportedOsFamily() {
        return this.getStringAttribute('reported_os_family');
    }
    /**
     * @stability stable
     */
    set reportedOsFamily(value) {
        this._reportedOsFamily = value;
    }
    /**
     * @stability stable
     */
    resetReportedOsFamily() {
        this._reportedOsFamily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportedOsFamilyInput() {
        return this._reportedOsFamily;
    }
    /**
     * @stability stable
     */
    get reportedOsName() {
        return this.getStringAttribute('reported_os_name');
    }
    /**
     * @stability stable
     */
    set reportedOsName(value) {
        this._reportedOsName = value;
    }
    /**
     * @stability stable
     */
    resetReportedOsName() {
        this._reportedOsName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportedOsNameInput() {
        return this._reportedOsName;
    }
    /**
     * @stability stable
     */
    get reportedOsVersion() {
        return this.getStringAttribute('reported_os_version');
    }
    /**
     * @stability stable
     */
    set reportedOsVersion(value) {
        this._reportedOsVersion = value;
    }
    /**
     * @stability stable
     */
    resetReportedOsVersion() {
        this._reportedOsVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportedOsVersionInput() {
        return this._reportedOsVersion;
    }
    /**
     * @stability stable
     */
    get rootDeviceType() {
        return this.getStringAttribute('root_device_type');
    }
    /**
     * @stability stable
     */
    set rootDeviceType(value) {
        this._rootDeviceType = value;
    }
    /**
     * @stability stable
     */
    resetRootDeviceType() {
        this._rootDeviceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootDeviceTypeInput() {
        return this._rootDeviceType;
    }
    /**
     * @stability stable
     */
    get rootDeviceVolumeId() {
        return this.getStringAttribute('root_device_volume_id');
    }
    /**
     * @stability stable
     */
    set rootDeviceVolumeId(value) {
        this._rootDeviceVolumeId = value;
    }
    /**
     * @stability stable
     */
    resetRootDeviceVolumeId() {
        this._rootDeviceVolumeId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootDeviceVolumeIdInput() {
        return this._rootDeviceVolumeId;
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get sshHostDsaKeyFingerprint() {
        return this.getStringAttribute('ssh_host_dsa_key_fingerprint');
    }
    /**
     * @stability stable
     */
    set sshHostDsaKeyFingerprint(value) {
        this._sshHostDsaKeyFingerprint = value;
    }
    /**
     * @stability stable
     */
    resetSshHostDsaKeyFingerprint() {
        this._sshHostDsaKeyFingerprint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sshHostDsaKeyFingerprintInput() {
        return this._sshHostDsaKeyFingerprint;
    }
    /**
     * @stability stable
     */
    get sshHostRsaKeyFingerprint() {
        return this.getStringAttribute('ssh_host_rsa_key_fingerprint');
    }
    /**
     * @stability stable
     */
    set sshHostRsaKeyFingerprint(value) {
        this._sshHostRsaKeyFingerprint = value;
    }
    /**
     * @stability stable
     */
    resetSshHostRsaKeyFingerprint() {
        this._sshHostRsaKeyFingerprint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sshHostRsaKeyFingerprintInput() {
        return this._sshHostRsaKeyFingerprint;
    }
    /**
     * @stability stable
     */
    get sshKeyName() {
        return this.getStringAttribute('ssh_key_name');
    }
    /**
     * @stability stable
     */
    set sshKeyName(value) {
        this._sshKeyName = value;
    }
    /**
     * @stability stable
     */
    resetSshKeyName() {
        this._sshKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sshKeyNameInput() {
        return this._sshKeyName;
    }
    /**
     * @stability stable
     */
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    /**
     * @stability stable
     */
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackIdInput() {
        return this._stackId;
    }
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    set state(value) {
        this._state = value;
    }
    /**
     * @stability stable
     */
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateInput() {
        return this._state;
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    /**
     * @stability stable
     */
    set subnetId(value) {
        this._subnetId = value;
    }
    /**
     * @stability stable
     */
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdInput() {
        return this._subnetId;
    }
    /**
     * @stability stable
     */
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
    /**
     * @stability stable
     */
    set tenancy(value) {
        this._tenancy = value;
    }
    /**
     * @stability stable
     */
    resetTenancy() {
        this._tenancy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tenancyInput() {
        return this._tenancy;
    }
    /**
     * @stability stable
     */
    get virtualizationType() {
        return this.getStringAttribute('virtualization_type');
    }
    /**
     * @stability stable
     */
    set virtualizationType(value) {
        this._virtualizationType = value;
    }
    /**
     * @stability stable
     */
    resetVirtualizationType() {
        this._virtualizationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get virtualizationTypeInput() {
        return this._virtualizationType;
    }
    /**
     * @stability stable
     */
    get ebsBlockDevice() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ebs_block_device');
    }
    /**
     * @stability stable
     */
    set ebsBlockDevice(value) {
        this._ebsBlockDevice = value;
    }
    /**
     * @stability stable
     */
    resetEbsBlockDevice() {
        this._ebsBlockDevice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsBlockDeviceInput() {
        return this._ebsBlockDevice;
    }
    /**
     * @stability stable
     */
    get ephemeralBlockDevice() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ephemeral_block_device');
    }
    /**
     * @stability stable
     */
    set ephemeralBlockDevice(value) {
        this._ephemeralBlockDevice = value;
    }
    /**
     * @stability stable
     */
    resetEphemeralBlockDevice() {
        this._ephemeralBlockDevice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ephemeralBlockDeviceInput() {
        return this._ephemeralBlockDevice;
    }
    /**
     * @stability stable
     */
    get rootBlockDevice() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('root_block_device');
    }
    /**
     * @stability stable
     */
    set rootBlockDevice(value) {
        this._rootBlockDevice = value;
    }
    /**
     * @stability stable
     */
    resetRootBlockDevice() {
        this._rootBlockDevice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootBlockDeviceInput() {
        return this._rootBlockDevice;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            agent_version: cdktf.stringToTerraform(this._agentVersion),
            ami_id: cdktf.stringToTerraform(this._amiId),
            architecture: cdktf.stringToTerraform(this._architecture),
            auto_scaling_type: cdktf.stringToTerraform(this._autoScalingType),
            availability_zone: cdktf.stringToTerraform(this._availabilityZone),
            created_at: cdktf.stringToTerraform(this._createdAt),
            delete_ebs: cdktf.booleanToTerraform(this._deleteEbs),
            delete_eip: cdktf.booleanToTerraform(this._deleteEip),
            ebs_optimized: cdktf.booleanToTerraform(this._ebsOptimized),
            ecs_cluster_arn: cdktf.stringToTerraform(this._ecsClusterArn),
            elastic_ip: cdktf.stringToTerraform(this._elasticIp),
            hostname: cdktf.stringToTerraform(this._hostname),
            infrastructure_class: cdktf.stringToTerraform(this._infrastructureClass),
            install_updates_on_boot: cdktf.booleanToTerraform(this._installUpdatesOnBoot),
            instance_profile_arn: cdktf.stringToTerraform(this._instanceProfileArn),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            last_service_error_id: cdktf.stringToTerraform(this._lastServiceErrorId),
            layer_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._layerIds),
            os: cdktf.stringToTerraform(this._os),
            platform: cdktf.stringToTerraform(this._platform),
            private_dns: cdktf.stringToTerraform(this._privateDns),
            private_ip: cdktf.stringToTerraform(this._privateIp),
            public_dns: cdktf.stringToTerraform(this._publicDns),
            public_ip: cdktf.stringToTerraform(this._publicIp),
            registered_by: cdktf.stringToTerraform(this._registeredBy),
            reported_agent_version: cdktf.stringToTerraform(this._reportedAgentVersion),
            reported_os_family: cdktf.stringToTerraform(this._reportedOsFamily),
            reported_os_name: cdktf.stringToTerraform(this._reportedOsName),
            reported_os_version: cdktf.stringToTerraform(this._reportedOsVersion),
            root_device_type: cdktf.stringToTerraform(this._rootDeviceType),
            root_device_volume_id: cdktf.stringToTerraform(this._rootDeviceVolumeId),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroupIds),
            ssh_host_dsa_key_fingerprint: cdktf.stringToTerraform(this._sshHostDsaKeyFingerprint),
            ssh_host_rsa_key_fingerprint: cdktf.stringToTerraform(this._sshHostRsaKeyFingerprint),
            ssh_key_name: cdktf.stringToTerraform(this._sshKeyName),
            stack_id: cdktf.stringToTerraform(this._stackId),
            state: cdktf.stringToTerraform(this._state),
            status: cdktf.stringToTerraform(this._status),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tenancy: cdktf.stringToTerraform(this._tenancy),
            virtualization_type: cdktf.stringToTerraform(this._virtualizationType),
            ebs_block_device: cdktf.listMapper(opsworksInstanceEbsBlockDeviceToTerraform)(this._ebsBlockDevice),
            ephemeral_block_device: cdktf.listMapper(opsworksInstanceEphemeralBlockDeviceToTerraform)(this._ephemeralBlockDevice),
            root_block_device: cdktf.listMapper(opsworksInstanceRootBlockDeviceToTerraform)(this._rootBlockDevice),
            timeouts: opsworksInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.OpsworksInstance = OpsworksInstance;
_b = JSII_RTTI_SYMBOL_1;
OpsworksInstance[_b] = { fqn: "@cdktf/provider-aws.opsworks.OpsworksInstance", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OpsworksInstance.tfResourceType = "aws_opsworks_instance";
