// @cdktf/provider-aws/lib/opsworks/opsworks-application.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksApplication = exports.opsworksApplicationSslConfigurationToTerraform = exports.opsworksApplicationEnvironmentToTerraform = exports.opsworksApplicationAppSourceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function opsworksApplicationAppSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        password: cdktf.stringToTerraform(struct.password),
        revision: cdktf.stringToTerraform(struct.revision),
        ssh_key: cdktf.stringToTerraform(struct.sshKey),
        type: cdktf.stringToTerraform(struct.type),
        url: cdktf.stringToTerraform(struct.url),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.opsworksApplicationAppSourceToTerraform = opsworksApplicationAppSourceToTerraform;
function opsworksApplicationEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        secure: cdktf.booleanToTerraform(struct.secure),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.opsworksApplicationEnvironmentToTerraform = opsworksApplicationEnvironmentToTerraform;
function opsworksApplicationSslConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate: cdktf.stringToTerraform(struct.certificate),
        chain: cdktf.stringToTerraform(struct.chain),
        private_key: cdktf.stringToTerraform(struct.privateKey),
    };
}
exports.opsworksApplicationSslConfigurationToTerraform = opsworksApplicationSslConfigurationToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/opsworks_application.html aws_opsworks_application}.
 *
 * @stability stable
 */
class OpsworksApplication extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/opsworks_application.html aws_opsworks_application} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_application',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._autoBundleOnDeploy = config.autoBundleOnDeploy;
        this._awsFlowRubySettings = config.awsFlowRubySettings;
        this._dataSourceArn = config.dataSourceArn;
        this._dataSourceDatabaseName = config.dataSourceDatabaseName;
        this._dataSourceType = config.dataSourceType;
        this._description = config.description;
        this._documentRoot = config.documentRoot;
        this._domains = config.domains;
        this._enableSsl = config.enableSsl;
        this._name = config.name;
        this._railsEnv = config.railsEnv;
        this._shortName = config.shortName;
        this._stackId = config.stackId;
        this._type = config.type;
        this._appSource = config.appSource;
        this._environment = config.environment;
        this._sslConfiguration = config.sslConfiguration;
    }
    /**
     * @stability stable
     */
    get autoBundleOnDeploy() {
        return this.getStringAttribute('auto_bundle_on_deploy');
    }
    /**
     * @stability stable
     */
    set autoBundleOnDeploy(value) {
        this._autoBundleOnDeploy = value;
    }
    /**
     * @stability stable
     */
    resetAutoBundleOnDeploy() {
        this._autoBundleOnDeploy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoBundleOnDeployInput() {
        return this._autoBundleOnDeploy;
    }
    /**
     * @stability stable
     */
    get awsFlowRubySettings() {
        return this.getStringAttribute('aws_flow_ruby_settings');
    }
    /**
     * @stability stable
     */
    set awsFlowRubySettings(value) {
        this._awsFlowRubySettings = value;
    }
    /**
     * @stability stable
     */
    resetAwsFlowRubySettings() {
        this._awsFlowRubySettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get awsFlowRubySettingsInput() {
        return this._awsFlowRubySettings;
    }
    /**
     * @stability stable
     */
    get dataSourceArn() {
        return this.getStringAttribute('data_source_arn');
    }
    /**
     * @stability stable
     */
    set dataSourceArn(value) {
        this._dataSourceArn = value;
    }
    /**
     * @stability stable
     */
    resetDataSourceArn() {
        this._dataSourceArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataSourceArnInput() {
        return this._dataSourceArn;
    }
    /**
     * @stability stable
     */
    get dataSourceDatabaseName() {
        return this.getStringAttribute('data_source_database_name');
    }
    /**
     * @stability stable
     */
    set dataSourceDatabaseName(value) {
        this._dataSourceDatabaseName = value;
    }
    /**
     * @stability stable
     */
    resetDataSourceDatabaseName() {
        this._dataSourceDatabaseName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataSourceDatabaseNameInput() {
        return this._dataSourceDatabaseName;
    }
    /**
     * @stability stable
     */
    get dataSourceType() {
        return this.getStringAttribute('data_source_type');
    }
    /**
     * @stability stable
     */
    set dataSourceType(value) {
        this._dataSourceType = value;
    }
    /**
     * @stability stable
     */
    resetDataSourceType() {
        this._dataSourceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataSourceTypeInput() {
        return this._dataSourceType;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get documentRoot() {
        return this.getStringAttribute('document_root');
    }
    /**
     * @stability stable
     */
    set documentRoot(value) {
        this._documentRoot = value;
    }
    /**
     * @stability stable
     */
    resetDocumentRoot() {
        this._documentRoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentRootInput() {
        return this._documentRoot;
    }
    /**
     * @stability stable
     */
    get domains() {
        return this.getListAttribute('domains');
    }
    /**
     * @stability stable
     */
    set domains(value) {
        this._domains = value;
    }
    /**
     * @stability stable
     */
    resetDomains() {
        this._domains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainsInput() {
        return this._domains;
    }
    /**
     * @stability stable
     */
    get enableSsl() {
        return this.getBooleanAttribute('enable_ssl');
    }
    /**
     * @stability stable
     */
    set enableSsl(value) {
        this._enableSsl = value;
    }
    /**
     * @stability stable
     */
    resetEnableSsl() {
        this._enableSsl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableSslInput() {
        return this._enableSsl;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get railsEnv() {
        return this.getStringAttribute('rails_env');
    }
    /**
     * @stability stable
     */
    set railsEnv(value) {
        this._railsEnv = value;
    }
    /**
     * @stability stable
     */
    resetRailsEnv() {
        this._railsEnv = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get railsEnvInput() {
        return this._railsEnv;
    }
    /**
     * @stability stable
     */
    get shortName() {
        return this.getStringAttribute('short_name');
    }
    /**
     * @stability stable
     */
    set shortName(value) {
        this._shortName = value;
    }
    /**
     * @stability stable
     */
    resetShortName() {
        this._shortName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get shortNameInput() {
        return this._shortName;
    }
    /**
     * @stability stable
     */
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    /**
     * @stability stable
     */
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackIdInput() {
        return this._stackId;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get appSource() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('app_source');
    }
    /**
     * @stability stable
     */
    set appSource(value) {
        this._appSource = value;
    }
    /**
     * @stability stable
     */
    resetAppSource() {
        this._appSource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get appSourceInput() {
        return this._appSource;
    }
    /**
     * @stability stable
     */
    get environment() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('environment');
    }
    /**
     * @stability stable
     */
    set environment(value) {
        this._environment = value;
    }
    /**
     * @stability stable
     */
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentInput() {
        return this._environment;
    }
    /**
     * @stability stable
     */
    get sslConfiguration() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ssl_configuration');
    }
    /**
     * @stability stable
     */
    set sslConfiguration(value) {
        this._sslConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetSslConfiguration() {
        this._sslConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sslConfigurationInput() {
        return this._sslConfiguration;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            auto_bundle_on_deploy: cdktf.stringToTerraform(this._autoBundleOnDeploy),
            aws_flow_ruby_settings: cdktf.stringToTerraform(this._awsFlowRubySettings),
            data_source_arn: cdktf.stringToTerraform(this._dataSourceArn),
            data_source_database_name: cdktf.stringToTerraform(this._dataSourceDatabaseName),
            data_source_type: cdktf.stringToTerraform(this._dataSourceType),
            description: cdktf.stringToTerraform(this._description),
            document_root: cdktf.stringToTerraform(this._documentRoot),
            domains: cdktf.listMapper(cdktf.stringToTerraform)(this._domains),
            enable_ssl: cdktf.booleanToTerraform(this._enableSsl),
            name: cdktf.stringToTerraform(this._name),
            rails_env: cdktf.stringToTerraform(this._railsEnv),
            short_name: cdktf.stringToTerraform(this._shortName),
            stack_id: cdktf.stringToTerraform(this._stackId),
            type: cdktf.stringToTerraform(this._type),
            app_source: cdktf.listMapper(opsworksApplicationAppSourceToTerraform)(this._appSource),
            environment: cdktf.listMapper(opsworksApplicationEnvironmentToTerraform)(this._environment),
            ssl_configuration: cdktf.listMapper(opsworksApplicationSslConfigurationToTerraform)(this._sslConfiguration),
        };
    }
}
exports.OpsworksApplication = OpsworksApplication;
_a = JSII_RTTI_SYMBOL_1;
OpsworksApplication[_a] = { fqn: "@cdktf/provider-aws.opsworks.OpsworksApplication", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OpsworksApplication.tfResourceType = "aws_opsworks_application";
