// @cdktf/provider-aws/lib/opsworks/opsworks-rds-db-instance.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksRdsDbInstance = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/opsworks_rds_db_instance.html aws_opsworks_rds_db_instance}.
 *
 * @stability stable
 */
class OpsworksRdsDbInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/opsworks_rds_db_instance.html aws_opsworks_rds_db_instance} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_rds_db_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dbPassword = config.dbPassword;
        this._dbUser = config.dbUser;
        this._rdsDbInstanceArn = config.rdsDbInstanceArn;
        this._stackId = config.stackId;
    }
    /**
     * @stability stable
     */
    get dbPassword() {
        return this.getStringAttribute('db_password');
    }
    /**
     * @stability stable
     */
    set dbPassword(value) {
        this._dbPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dbPasswordInput() {
        return this._dbPassword;
    }
    /**
     * @stability stable
     */
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    /**
     * @stability stable
     */
    set dbUser(value) {
        this._dbUser = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dbUserInput() {
        return this._dbUser;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get rdsDbInstanceArn() {
        return this.getStringAttribute('rds_db_instance_arn');
    }
    /**
     * @stability stable
     */
    set rdsDbInstanceArn(value) {
        this._rdsDbInstanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rdsDbInstanceArnInput() {
        return this._rdsDbInstanceArn;
    }
    /**
     * @stability stable
     */
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    /**
     * @stability stable
     */
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackIdInput() {
        return this._stackId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            db_password: cdktf.stringToTerraform(this._dbPassword),
            db_user: cdktf.stringToTerraform(this._dbUser),
            rds_db_instance_arn: cdktf.stringToTerraform(this._rdsDbInstanceArn),
            stack_id: cdktf.stringToTerraform(this._stackId),
        };
    }
}
exports.OpsworksRdsDbInstance = OpsworksRdsDbInstance;
_a = JSII_RTTI_SYMBOL_1;
OpsworksRdsDbInstance[_a] = { fqn: "@cdktf/provider-aws.opsworks.OpsworksRdsDbInstance", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OpsworksRdsDbInstance.tfResourceType = "aws_opsworks_rds_db_instance";
