// cdktf-cli/lib/server/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapWatchState = exports.readGitignore = void 0;
const parse_gitignore_1 = __importDefault(require("parse-gitignore"));
const path = __importStar(require("path"));
const fs_1 = require("fs");
async function readGitignore(directory) {
    const content = await fs_1.promises.readFile(path.join(directory, ".gitignore"));
    return parse_gitignore_1.default(content);
}
exports.readGitignore = readGitignore;
function mapWatchState(state) {
    return {
        status: state.status,
        stacks: state.stacks.map((s) => ({
            constructPath: s.constructPath,
            name: s.name,
            synthesizedStackPath: s.synthesizedStackPath,
            workingDirectory: s.workingDirectory,
        })),
        resources: Array.from(state.resources.values()).map((r) => ({
            id: r.id,
            action: plannedResourceActionReverseMap[r.action],
            deployState: deployingResourceApplyStateReverseMap[r.applyState],
            changedAt: r.changedAt,
        })),
        error: state.error
            ? {
                message: state.error.message,
                origin: state.error.origin,
                recoverable: state.error.recoverable,
                timestamp: state.error.timestamp,
            }
            : undefined,
    };
}
exports.mapWatchState = mapWatchState;
const plannedResourceActionReverseMap = {
    "no-op": "NO_OP",
    create: "CREATE",
    delete: "DELETE",
    read: "READ",
    update: "UPDATE",
};
const deployingResourceApplyStateReverseMap = {
    waiting: "WAITING",
    updating: "UPDATING",
    updated: "UPDATED",
    creating: "CREATING",
    created: "CREATED",
    destroying: "DESTROYING",
    destroyed: "DESTROYED",
    success: "SUCCESS",
    error: "ERROR",
};
