// @cdktf/provider-aws/lib/servicecatalog/servicecatalog-provisioned-product.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServicecatalogProvisionedProduct = exports.ServicecatalogProvisionedProductTimeoutsOutputReference = exports.servicecatalogProvisionedProductTimeoutsToTerraform = exports.ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference = exports.servicecatalogProvisionedProductStackSetProvisioningPreferencesToTerraform = exports.servicecatalogProvisionedProductProvisioningParametersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function servicecatalogProvisionedProductProvisioningParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        use_previous_value: cdktf.booleanToTerraform(struct.usePreviousValue),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.servicecatalogProvisionedProductProvisioningParametersToTerraform = servicecatalogProvisionedProductProvisioningParametersToTerraform;
function servicecatalogProvisionedProductStackSetProvisioningPreferencesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accounts: cdktf.listMapper(cdktf.stringToTerraform)(struct.accounts),
        failure_tolerance_count: cdktf.numberToTerraform(struct.failureToleranceCount),
        failure_tolerance_percentage: cdktf.numberToTerraform(struct.failureTolerancePercentage),
        max_concurrency_count: cdktf.numberToTerraform(struct.maxConcurrencyCount),
        max_concurrency_percentage: cdktf.numberToTerraform(struct.maxConcurrencyPercentage),
        regions: cdktf.listMapper(cdktf.stringToTerraform)(struct.regions),
    };
}
exports.servicecatalogProvisionedProductStackSetProvisioningPreferencesToTerraform = servicecatalogProvisionedProductStackSetProvisioningPreferencesToTerraform;
/**
 * @stability stable
 */
class ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accounts) {
            hasAnyValues = true;
            internalValueResult.accounts = this._accounts;
        }
        if (this._failureToleranceCount) {
            hasAnyValues = true;
            internalValueResult.failureToleranceCount = this._failureToleranceCount;
        }
        if (this._failureTolerancePercentage) {
            hasAnyValues = true;
            internalValueResult.failureTolerancePercentage = this._failureTolerancePercentage;
        }
        if (this._maxConcurrencyCount) {
            hasAnyValues = true;
            internalValueResult.maxConcurrencyCount = this._maxConcurrencyCount;
        }
        if (this._maxConcurrencyPercentage) {
            hasAnyValues = true;
            internalValueResult.maxConcurrencyPercentage = this._maxConcurrencyPercentage;
        }
        if (this._regions) {
            hasAnyValues = true;
            internalValueResult.regions = this._regions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accounts = undefined;
            this._failureToleranceCount = undefined;
            this._failureTolerancePercentage = undefined;
            this._maxConcurrencyCount = undefined;
            this._maxConcurrencyPercentage = undefined;
            this._regions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accounts = value.accounts;
            this._failureToleranceCount = value.failureToleranceCount;
            this._failureTolerancePercentage = value.failureTolerancePercentage;
            this._maxConcurrencyCount = value.maxConcurrencyCount;
            this._maxConcurrencyPercentage = value.maxConcurrencyPercentage;
            this._regions = value.regions;
        }
    }
    /**
     * @stability stable
     */
    get accounts() {
        return this.getListAttribute('accounts');
    }
    /**
     * @stability stable
     */
    set accounts(value) {
        this._accounts = value;
    }
    /**
     * @stability stable
     */
    resetAccounts() {
        this._accounts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountsInput() {
        return this._accounts;
    }
    /**
     * @stability stable
     */
    get failureToleranceCount() {
        return this.getNumberAttribute('failure_tolerance_count');
    }
    /**
     * @stability stable
     */
    set failureToleranceCount(value) {
        this._failureToleranceCount = value;
    }
    /**
     * @stability stable
     */
    resetFailureToleranceCount() {
        this._failureToleranceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get failureToleranceCountInput() {
        return this._failureToleranceCount;
    }
    /**
     * @stability stable
     */
    get failureTolerancePercentage() {
        return this.getNumberAttribute('failure_tolerance_percentage');
    }
    /**
     * @stability stable
     */
    set failureTolerancePercentage(value) {
        this._failureTolerancePercentage = value;
    }
    /**
     * @stability stable
     */
    resetFailureTolerancePercentage() {
        this._failureTolerancePercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get failureTolerancePercentageInput() {
        return this._failureTolerancePercentage;
    }
    /**
     * @stability stable
     */
    get maxConcurrencyCount() {
        return this.getNumberAttribute('max_concurrency_count');
    }
    /**
     * @stability stable
     */
    set maxConcurrencyCount(value) {
        this._maxConcurrencyCount = value;
    }
    /**
     * @stability stable
     */
    resetMaxConcurrencyCount() {
        this._maxConcurrencyCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrencyCountInput() {
        return this._maxConcurrencyCount;
    }
    /**
     * @stability stable
     */
    get maxConcurrencyPercentage() {
        return this.getNumberAttribute('max_concurrency_percentage');
    }
    /**
     * @stability stable
     */
    set maxConcurrencyPercentage(value) {
        this._maxConcurrencyPercentage = value;
    }
    /**
     * @stability stable
     */
    resetMaxConcurrencyPercentage() {
        this._maxConcurrencyPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrencyPercentageInput() {
        return this._maxConcurrencyPercentage;
    }
    /**
     * @stability stable
     */
    get regions() {
        return this.getListAttribute('regions');
    }
    /**
     * @stability stable
     */
    set regions(value) {
        this._regions = value;
    }
    /**
     * @stability stable
     */
    resetRegions() {
        this._regions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionsInput() {
        return this._regions;
    }
}
exports.ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference = ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference[_a] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference", version: "3.0.1" };
function servicecatalogProvisionedProductTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.servicecatalogProvisionedProductTimeoutsToTerraform = servicecatalogProvisionedProductTimeoutsToTerraform;
/**
 * @stability stable
 */
class ServicecatalogProvisionedProductTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.ServicecatalogProvisionedProductTimeoutsOutputReference = ServicecatalogProvisionedProductTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ServicecatalogProvisionedProductTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProvisionedProductTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_provisioned_product.html aws_servicecatalog_provisioned_product}.
 *
 * @stability stable
 */
class ServicecatalogProvisionedProduct extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/servicecatalog_provisioned_product.html aws_servicecatalog_provisioned_product} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_servicecatalog_provisioned_product',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // stack_set_provisioning_preferences - computed: false, optional: true, required: false
        this._stackSetProvisioningPreferences = new ServicecatalogProvisionedProductStackSetProvisioningPreferencesOutputReference(this, "stack_set_provisioning_preferences", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ServicecatalogProvisionedProductTimeoutsOutputReference(this, "timeouts", true);
        this._acceptLanguage = config.acceptLanguage;
        this._ignoreErrors = config.ignoreErrors;
        this._name = config.name;
        this._notificationArns = config.notificationArns;
        this._pathId = config.pathId;
        this._pathName = config.pathName;
        this._productId = config.productId;
        this._productName = config.productName;
        this._provisioningArtifactId = config.provisioningArtifactId;
        this._provisioningArtifactName = config.provisioningArtifactName;
        this._retainPhysicalResources = config.retainPhysicalResources;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._provisioningParameters = config.provisioningParameters;
        this._stackSetProvisioningPreferences.internalValue = config.stackSetProvisioningPreferences;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get acceptLanguage() {
        return this.getStringAttribute('accept_language');
    }
    /**
     * @stability stable
     */
    set acceptLanguage(value) {
        this._acceptLanguage = value;
    }
    /**
     * @stability stable
     */
    resetAcceptLanguage() {
        this._acceptLanguage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptLanguageInput() {
        return this._acceptLanguage;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // cloudwatch_dashboard_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudwatchDashboardNames() {
        return this.getListAttribute('cloudwatch_dashboard_names');
    }
    // created_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ignoreErrors() {
        return this.getBooleanAttribute('ignore_errors');
    }
    /**
     * @stability stable
     */
    set ignoreErrors(value) {
        this._ignoreErrors = value;
    }
    /**
     * @stability stable
     */
    resetIgnoreErrors() {
        this._ignoreErrors = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ignoreErrorsInput() {
        return this._ignoreErrors;
    }
    // last_provisioning_record_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastProvisioningRecordId() {
        return this.getStringAttribute('last_provisioning_record_id');
    }
    // last_record_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastRecordId() {
        return this.getStringAttribute('last_record_id');
    }
    // last_successful_provisioning_record_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastSuccessfulProvisioningRecordId() {
        return this.getStringAttribute('last_successful_provisioning_record_id');
    }
    // launch_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get launchRoleArn() {
        return this.getStringAttribute('launch_role_arn');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get notificationArns() {
        return this.getListAttribute('notification_arns');
    }
    /**
     * @stability stable
     */
    set notificationArns(value) {
        this._notificationArns = value;
    }
    /**
     * @stability stable
     */
    resetNotificationArns() {
        this._notificationArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationArnsInput() {
        return this._notificationArns;
    }
    /**
     * @stability stable
     */
    get pathId() {
        return this.getStringAttribute('path_id');
    }
    /**
     * @stability stable
     */
    set pathId(value) {
        this._pathId = value;
    }
    /**
     * @stability stable
     */
    resetPathId() {
        this._pathId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathIdInput() {
        return this._pathId;
    }
    /**
     * @stability stable
     */
    get pathName() {
        return this.getStringAttribute('path_name');
    }
    /**
     * @stability stable
     */
    set pathName(value) {
        this._pathName = value;
    }
    /**
     * @stability stable
     */
    resetPathName() {
        this._pathName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathNameInput() {
        return this._pathName;
    }
    /**
     * @stability stable
     */
    get productId() {
        return this.getStringAttribute('product_id');
    }
    /**
     * @stability stable
     */
    set productId(value) {
        this._productId = value;
    }
    /**
     * @stability stable
     */
    resetProductId() {
        this._productId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productIdInput() {
        return this._productId;
    }
    /**
     * @stability stable
     */
    get productName() {
        return this.getStringAttribute('product_name');
    }
    /**
     * @stability stable
     */
    set productName(value) {
        this._productName = value;
    }
    /**
     * @stability stable
     */
    resetProductName() {
        this._productName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get productNameInput() {
        return this._productName;
    }
    /**
     * @stability stable
     */
    get provisioningArtifactId() {
        return this.getStringAttribute('provisioning_artifact_id');
    }
    /**
     * @stability stable
     */
    set provisioningArtifactId(value) {
        this._provisioningArtifactId = value;
    }
    /**
     * @stability stable
     */
    resetProvisioningArtifactId() {
        this._provisioningArtifactId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get provisioningArtifactIdInput() {
        return this._provisioningArtifactId;
    }
    /**
     * @stability stable
     */
    get provisioningArtifactName() {
        return this.getStringAttribute('provisioning_artifact_name');
    }
    /**
     * @stability stable
     */
    set provisioningArtifactName(value) {
        this._provisioningArtifactName = value;
    }
    /**
     * @stability stable
     */
    resetProvisioningArtifactName() {
        this._provisioningArtifactName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get provisioningArtifactNameInput() {
        return this._provisioningArtifactName;
    }
    /**
     * @stability stable
     */
    get retainPhysicalResources() {
        return this.getBooleanAttribute('retain_physical_resources');
    }
    /**
     * @stability stable
     */
    set retainPhysicalResources(value) {
        this._retainPhysicalResources = value;
    }
    /**
     * @stability stable
     */
    resetRetainPhysicalResources() {
        this._retainPhysicalResources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get retainPhysicalResourcesInput() {
        return this._retainPhysicalResources;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // status_message - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    get provisioningParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('provisioning_parameters');
    }
    /**
     * @stability stable
     */
    set provisioningParameters(value) {
        this._provisioningParameters = value;
    }
    /**
     * @stability stable
     */
    resetProvisioningParameters() {
        this._provisioningParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get provisioningParametersInput() {
        return this._provisioningParameters;
    }
    /**
     * @stability stable
     */
    get stackSetProvisioningPreferences() {
        return this._stackSetProvisioningPreferences;
    }
    /**
     * @stability stable
     */
    putStackSetProvisioningPreferences(value) {
        this._stackSetProvisioningPreferences.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetStackSetProvisioningPreferences() {
        this._stackSetProvisioningPreferences.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackSetProvisioningPreferencesInput() {
        return this._stackSetProvisioningPreferences.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            accept_language: cdktf.stringToTerraform(this._acceptLanguage),
            ignore_errors: cdktf.booleanToTerraform(this._ignoreErrors),
            name: cdktf.stringToTerraform(this._name),
            notification_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._notificationArns),
            path_id: cdktf.stringToTerraform(this._pathId),
            path_name: cdktf.stringToTerraform(this._pathName),
            product_id: cdktf.stringToTerraform(this._productId),
            product_name: cdktf.stringToTerraform(this._productName),
            provisioning_artifact_id: cdktf.stringToTerraform(this._provisioningArtifactId),
            provisioning_artifact_name: cdktf.stringToTerraform(this._provisioningArtifactName),
            retain_physical_resources: cdktf.booleanToTerraform(this._retainPhysicalResources),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            provisioning_parameters: cdktf.listMapper(servicecatalogProvisionedProductProvisioningParametersToTerraform)(this._provisioningParameters),
            stack_set_provisioning_preferences: servicecatalogProvisionedProductStackSetProvisioningPreferencesToTerraform(this._stackSetProvisioningPreferences.internalValue),
            timeouts: servicecatalogProvisionedProductTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ServicecatalogProvisionedProduct = ServicecatalogProvisionedProduct;
_c = JSII_RTTI_SYMBOL_1;
ServicecatalogProvisionedProduct[_c] = { fqn: "@cdktf/provider-aws.servicecatalog.ServicecatalogProvisionedProduct", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ServicecatalogProvisionedProduct.tfResourceType = "aws_servicecatalog_provisioned_product";
