// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-route-response.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2RouteResponse = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route_response.html aws_apigatewayv2_route_response}.
 *
 * @stability stable
 */
class Apigatewayv2RouteResponse extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route_response.html aws_apigatewayv2_route_response} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_route_response',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._apiId = config.apiId;
        this._modelSelectionExpression = config.modelSelectionExpression;
        this._responseModels = config.responseModels;
        this._routeId = config.routeId;
        this._routeResponseKey = config.routeResponseKey;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get modelSelectionExpression() {
        return this.getStringAttribute('model_selection_expression');
    }
    /**
     * @stability stable
     */
    set modelSelectionExpression(value) {
        this._modelSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetModelSelectionExpression() {
        this._modelSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modelSelectionExpressionInput() {
        return this._modelSelectionExpression;
    }
    /**
     * @stability stable
     */
    get responseModels() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_models');
    }
    /**
     * @stability stable
     */
    set responseModels(value) {
        this._responseModels = value;
    }
    /**
     * @stability stable
     */
    resetResponseModels() {
        this._responseModels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseModelsInput() {
        return this._responseModels;
    }
    /**
     * @stability stable
     */
    get routeId() {
        return this.getStringAttribute('route_id');
    }
    /**
     * @stability stable
     */
    set routeId(value) {
        this._routeId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeIdInput() {
        return this._routeId;
    }
    /**
     * @stability stable
     */
    get routeResponseKey() {
        return this.getStringAttribute('route_response_key');
    }
    /**
     * @stability stable
     */
    set routeResponseKey(value) {
        this._routeResponseKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeResponseKeyInput() {
        return this._routeResponseKey;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            model_selection_expression: cdktf.stringToTerraform(this._modelSelectionExpression),
            response_models: cdktf.hashMapper(cdktf.anyToTerraform)(this._responseModels),
            route_id: cdktf.stringToTerraform(this._routeId),
            route_response_key: cdktf.stringToTerraform(this._routeResponseKey),
        };
    }
}
exports.Apigatewayv2RouteResponse = Apigatewayv2RouteResponse;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2RouteResponse[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2RouteResponse", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2RouteResponse.tfResourceType = "aws_apigatewayv2_route_response";
