// @cdktf/provider-aws/lib/chime/chime-voice-connector-logging.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChimeVoiceConnectorLogging = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/chime_voice_connector_logging.html aws_chime_voice_connector_logging}.
 *
 * @stability stable
 */
class ChimeVoiceConnectorLogging extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/chime_voice_connector_logging.html aws_chime_voice_connector_logging} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_chime_voice_connector_logging',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._enableSipLogs = config.enableSipLogs;
        this._voiceConnectorId = config.voiceConnectorId;
    }
    /**
     * @stability stable
     */
    get enableSipLogs() {
        return this.getBooleanAttribute('enable_sip_logs');
    }
    /**
     * @stability stable
     */
    set enableSipLogs(value) {
        this._enableSipLogs = value;
    }
    /**
     * @stability stable
     */
    resetEnableSipLogs() {
        this._enableSipLogs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableSipLogsInput() {
        return this._enableSipLogs;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get voiceConnectorId() {
        return this.getStringAttribute('voice_connector_id');
    }
    /**
     * @stability stable
     */
    set voiceConnectorId(value) {
        this._voiceConnectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get voiceConnectorIdInput() {
        return this._voiceConnectorId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enable_sip_logs: cdktf.booleanToTerraform(this._enableSipLogs),
            voice_connector_id: cdktf.stringToTerraform(this._voiceConnectorId),
        };
    }
}
exports.ChimeVoiceConnectorLogging = ChimeVoiceConnectorLogging;
_a = JSII_RTTI_SYMBOL_1;
ChimeVoiceConnectorLogging[_a] = { fqn: "@cdktf/provider-aws.chime.ChimeVoiceConnectorLogging", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ChimeVoiceConnectorLogging.tfResourceType = "aws_chime_voice_connector_logging";
