// @cdktf/provider-aws/lib/redshift/data-aws-redshift-orderable-cluster.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftOrderableCluster = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/redshift_orderable_cluster.html aws_redshift_orderable_cluster}.
 *
 * @stability stable
 */
class DataAwsRedshiftOrderableCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/redshift_orderable_cluster.html aws_redshift_orderable_cluster} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_orderable_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterType = config.clusterType;
        this._clusterVersion = config.clusterVersion;
        this._nodeType = config.nodeType;
        this._preferredNodeTypes = config.preferredNodeTypes;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // availability_zones - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZones() {
        return this.getListAttribute('availability_zones');
    }
    /**
     * @stability stable
     */
    get clusterType() {
        return this.getStringAttribute('cluster_type');
    }
    /**
     * @stability stable
     */
    set clusterType(value) {
        this._clusterType = value;
    }
    /**
     * @stability stable
     */
    resetClusterType() {
        this._clusterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterTypeInput() {
        return this._clusterType;
    }
    /**
     * @stability stable
     */
    get clusterVersion() {
        return this.getStringAttribute('cluster_version');
    }
    /**
     * @stability stable
     */
    set clusterVersion(value) {
        this._clusterVersion = value;
    }
    /**
     * @stability stable
     */
    resetClusterVersion() {
        this._clusterVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterVersionInput() {
        return this._clusterVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    /**
     * @stability stable
     */
    set nodeType(value) {
        this._nodeType = value;
    }
    /**
     * @stability stable
     */
    resetNodeType() {
        this._nodeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeTypeInput() {
        return this._nodeType;
    }
    /**
     * @stability stable
     */
    get preferredNodeTypes() {
        return this.getListAttribute('preferred_node_types');
    }
    /**
     * @stability stable
     */
    set preferredNodeTypes(value) {
        this._preferredNodeTypes = value;
    }
    /**
     * @stability stable
     */
    resetPreferredNodeTypes() {
        this._preferredNodeTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get preferredNodeTypesInput() {
        return this._preferredNodeTypes;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_type: cdktf.stringToTerraform(this._clusterType),
            cluster_version: cdktf.stringToTerraform(this._clusterVersion),
            node_type: cdktf.stringToTerraform(this._nodeType),
            preferred_node_types: cdktf.listMapper(cdktf.stringToTerraform)(this._preferredNodeTypes),
        };
    }
}
exports.DataAwsRedshiftOrderableCluster = DataAwsRedshiftOrderableCluster;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftOrderableCluster[_a] = { fqn: "@cdktf/provider-aws.redshift.DataAwsRedshiftOrderableCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsRedshiftOrderableCluster.tfResourceType = "aws_redshift_orderable_cluster";
