// @cdktf/provider-aws/lib/redshift/data-aws-redshift-cluster.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftCluster = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/redshift_cluster.html aws_redshift_cluster}.
 *
 * @stability stable
 */
class DataAwsRedshiftCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/redshift_cluster.html aws_redshift_cluster} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterIdentifier = config.clusterIdentifier;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // allow_version_upgrade - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get allowVersionUpgrade() {
        return this.getBooleanAttribute('allow_version_upgrade');
    }
    // automated_snapshot_retention_period - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get automatedSnapshotRetentionPeriod() {
        return this.getNumberAttribute('automated_snapshot_retention_period');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // bucket_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    /**
     * @stability stable
     */
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    /**
     * @stability stable
     */
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    // cluster_parameter_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterParameterGroupName() {
        return this.getStringAttribute('cluster_parameter_group_name');
    }
    // cluster_public_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterPublicKey() {
        return this.getStringAttribute('cluster_public_key');
    }
    // cluster_revision_number - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterRevisionNumber() {
        return this.getStringAttribute('cluster_revision_number');
    }
    // cluster_security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterSecurityGroups() {
        return this.getListAttribute('cluster_security_groups');
    }
    // cluster_subnet_group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterSubnetGroupName() {
        return this.getStringAttribute('cluster_subnet_group_name');
    }
    // cluster_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterType() {
        return this.getStringAttribute('cluster_type');
    }
    // cluster_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterVersion() {
        return this.getStringAttribute('cluster_version');
    }
    // database_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    // elastic_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get elasticIp() {
        return this.getStringAttribute('elastic_ip');
    }
    // enable_logging - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableLogging() {
        return this.getBooleanAttribute('enable_logging');
    }
    // encrypted - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // enhanced_vpc_routing - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enhancedVpcRouting() {
        return this.getBooleanAttribute('enhanced_vpc_routing');
    }
    // iam_roles - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iamRoles() {
        return this.getListAttribute('iam_roles');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // master_username - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get masterUsername() {
        return this.getStringAttribute('master_username');
    }
    // node_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    // number_of_nodes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get numberOfNodes() {
        return this.getNumberAttribute('number_of_nodes');
    }
    // port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    // preferred_maintenance_window - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get preferredMaintenanceWindow() {
        return this.getStringAttribute('preferred_maintenance_window');
    }
    // publicly_accessible - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    // s3_key_prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsRedshiftCluster = DataAwsRedshiftCluster;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftCluster[_a] = { fqn: "@cdktf/provider-aws.redshift.DataAwsRedshiftCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsRedshiftCluster.tfResourceType = "aws_redshift_cluster";
