// @cdktf/provider-aws/lib/redshift/redshift-cluster.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftCluster = exports.RedshiftClusterTimeoutsOutputReference = exports.redshiftClusterTimeoutsToTerraform = exports.RedshiftClusterSnapshotCopyOutputReference = exports.redshiftClusterSnapshotCopyToTerraform = exports.RedshiftClusterLoggingOutputReference = exports.redshiftClusterLoggingToTerraform = exports.RedshiftClusterClusterNodes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class RedshiftClusterClusterNodes extends cdktf.ComplexComputedList {
    // node_role - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nodeRole() {
        return this.getStringAttribute('node_role');
    }
    // private_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // public_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicIpAddress() {
        return this.getStringAttribute('public_ip_address');
    }
}
exports.RedshiftClusterClusterNodes = RedshiftClusterClusterNodes;
_a = JSII_RTTI_SYMBOL_1;
RedshiftClusterClusterNodes[_a] = { fqn: "@cdktf/provider-aws.redshift.RedshiftClusterClusterNodes", version: "3.0.1" };
function redshiftClusterLoggingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        enable: cdktf.booleanToTerraform(struct.enable),
        s3_key_prefix: cdktf.stringToTerraform(struct.s3KeyPrefix),
    };
}
exports.redshiftClusterLoggingToTerraform = redshiftClusterLoggingToTerraform;
/**
 * @stability stable
 */
class RedshiftClusterLoggingOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._enable) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._s3KeyPrefix) {
            hasAnyValues = true;
            internalValueResult.s3KeyPrefix = this._s3KeyPrefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._enable = undefined;
            this._s3KeyPrefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._enable = value.enable;
            this._s3KeyPrefix = value.s3KeyPrefix;
        }
    }
    /**
     * @stability stable
     */
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    /**
     * @stability stable
     */
    set bucketName(value) {
        this._bucketName = value;
    }
    /**
     * @stability stable
     */
    resetBucketName() {
        this._bucketName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketNameInput() {
        return this._bucketName;
    }
    /**
     * @stability stable
     */
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    /**
     * @stability stable
     */
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableInput() {
        return this._enable;
    }
    /**
     * @stability stable
     */
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    /**
     * @stability stable
     */
    set s3KeyPrefix(value) {
        this._s3KeyPrefix = value;
    }
    /**
     * @stability stable
     */
    resetS3KeyPrefix() {
        this._s3KeyPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KeyPrefixInput() {
        return this._s3KeyPrefix;
    }
}
exports.RedshiftClusterLoggingOutputReference = RedshiftClusterLoggingOutputReference;
_b = JSII_RTTI_SYMBOL_1;
RedshiftClusterLoggingOutputReference[_b] = { fqn: "@cdktf/provider-aws.redshift.RedshiftClusterLoggingOutputReference", version: "3.0.1" };
function redshiftClusterSnapshotCopyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_region: cdktf.stringToTerraform(struct.destinationRegion),
        grant_name: cdktf.stringToTerraform(struct.grantName),
        retention_period: cdktf.numberToTerraform(struct.retentionPeriod),
    };
}
exports.redshiftClusterSnapshotCopyToTerraform = redshiftClusterSnapshotCopyToTerraform;
/**
 * @stability stable
 */
class RedshiftClusterSnapshotCopyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destinationRegion) {
            hasAnyValues = true;
            internalValueResult.destinationRegion = this._destinationRegion;
        }
        if (this._grantName) {
            hasAnyValues = true;
            internalValueResult.grantName = this._grantName;
        }
        if (this._retentionPeriod) {
            hasAnyValues = true;
            internalValueResult.retentionPeriod = this._retentionPeriod;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destinationRegion = undefined;
            this._grantName = undefined;
            this._retentionPeriod = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destinationRegion = value.destinationRegion;
            this._grantName = value.grantName;
            this._retentionPeriod = value.retentionPeriod;
        }
    }
    /**
     * @stability stable
     */
    get destinationRegion() {
        return this.getStringAttribute('destination_region');
    }
    /**
     * @stability stable
     */
    set destinationRegion(value) {
        this._destinationRegion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationRegionInput() {
        return this._destinationRegion;
    }
    /**
     * @stability stable
     */
    get grantName() {
        return this.getStringAttribute('grant_name');
    }
    /**
     * @stability stable
     */
    set grantName(value) {
        this._grantName = value;
    }
    /**
     * @stability stable
     */
    resetGrantName() {
        this._grantName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get grantNameInput() {
        return this._grantName;
    }
    /**
     * @stability stable
     */
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    /**
     * @stability stable
     */
    set retentionPeriod(value) {
        this._retentionPeriod = value;
    }
    /**
     * @stability stable
     */
    resetRetentionPeriod() {
        this._retentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get retentionPeriodInput() {
        return this._retentionPeriod;
    }
}
exports.RedshiftClusterSnapshotCopyOutputReference = RedshiftClusterSnapshotCopyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RedshiftClusterSnapshotCopyOutputReference[_c] = { fqn: "@cdktf/provider-aws.redshift.RedshiftClusterSnapshotCopyOutputReference", version: "3.0.1" };
function redshiftClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.redshiftClusterTimeoutsToTerraform = redshiftClusterTimeoutsToTerraform;
/**
 * @stability stable
 */
class RedshiftClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.RedshiftClusterTimeoutsOutputReference = RedshiftClusterTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
RedshiftClusterTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.redshift.RedshiftClusterTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshift_cluster.html aws_redshift_cluster}.
 *
 * @stability stable
 */
class RedshiftCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshift_cluster.html aws_redshift_cluster} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // logging - computed: false, optional: true, required: false
        this._logging = new RedshiftClusterLoggingOutputReference(this, "logging", true);
        // snapshot_copy - computed: false, optional: true, required: false
        this._snapshotCopy = new RedshiftClusterSnapshotCopyOutputReference(this, "snapshot_copy", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RedshiftClusterTimeoutsOutputReference(this, "timeouts", true);
        this._allowVersionUpgrade = config.allowVersionUpgrade;
        this._automatedSnapshotRetentionPeriod = config.automatedSnapshotRetentionPeriod;
        this._availabilityZone = config.availabilityZone;
        this._clusterIdentifier = config.clusterIdentifier;
        this._clusterParameterGroupName = config.clusterParameterGroupName;
        this._clusterPublicKey = config.clusterPublicKey;
        this._clusterRevisionNumber = config.clusterRevisionNumber;
        this._clusterSecurityGroups = config.clusterSecurityGroups;
        this._clusterSubnetGroupName = config.clusterSubnetGroupName;
        this._clusterType = config.clusterType;
        this._clusterVersion = config.clusterVersion;
        this._databaseName = config.databaseName;
        this._elasticIp = config.elasticIp;
        this._encrypted = config.encrypted;
        this._endpoint = config.endpoint;
        this._enhancedVpcRouting = config.enhancedVpcRouting;
        this._finalSnapshotIdentifier = config.finalSnapshotIdentifier;
        this._iamRoles = config.iamRoles;
        this._kmsKeyId = config.kmsKeyId;
        this._masterPassword = config.masterPassword;
        this._masterUsername = config.masterUsername;
        this._nodeType = config.nodeType;
        this._numberOfNodes = config.numberOfNodes;
        this._ownerAccount = config.ownerAccount;
        this._port = config.port;
        this._preferredMaintenanceWindow = config.preferredMaintenanceWindow;
        this._publiclyAccessible = config.publiclyAccessible;
        this._skipFinalSnapshot = config.skipFinalSnapshot;
        this._snapshotClusterIdentifier = config.snapshotClusterIdentifier;
        this._snapshotIdentifier = config.snapshotIdentifier;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcSecurityGroupIds = config.vpcSecurityGroupIds;
        this._logging.internalValue = config.logging;
        this._snapshotCopy.internalValue = config.snapshotCopy;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get allowVersionUpgrade() {
        return this.getBooleanAttribute('allow_version_upgrade');
    }
    /**
     * @stability stable
     */
    set allowVersionUpgrade(value) {
        this._allowVersionUpgrade = value;
    }
    /**
     * @stability stable
     */
    resetAllowVersionUpgrade() {
        this._allowVersionUpgrade = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowVersionUpgradeInput() {
        return this._allowVersionUpgrade;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get automatedSnapshotRetentionPeriod() {
        return this.getNumberAttribute('automated_snapshot_retention_period');
    }
    /**
     * @stability stable
     */
    set automatedSnapshotRetentionPeriod(value) {
        this._automatedSnapshotRetentionPeriod = value;
    }
    /**
     * @stability stable
     */
    resetAutomatedSnapshotRetentionPeriod() {
        this._automatedSnapshotRetentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get automatedSnapshotRetentionPeriodInput() {
        return this._automatedSnapshotRetentionPeriod;
    }
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    /**
     * @stability stable
     */
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    /**
     * @stability stable
     */
    resetAvailabilityZone() {
        this._availabilityZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    /**
     * @stability stable
     */
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    /**
     * @stability stable
     */
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    // cluster_nodes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    clusterNodes(index) {
        return new RedshiftClusterClusterNodes(this, 'cluster_nodes', index);
    }
    /**
     * @stability stable
     */
    get clusterParameterGroupName() {
        return this.getStringAttribute('cluster_parameter_group_name');
    }
    /**
     * @stability stable
     */
    set clusterParameterGroupName(value) {
        this._clusterParameterGroupName = value;
    }
    /**
     * @stability stable
     */
    resetClusterParameterGroupName() {
        this._clusterParameterGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterParameterGroupNameInput() {
        return this._clusterParameterGroupName;
    }
    /**
     * @stability stable
     */
    get clusterPublicKey() {
        return this.getStringAttribute('cluster_public_key');
    }
    /**
     * @stability stable
     */
    set clusterPublicKey(value) {
        this._clusterPublicKey = value;
    }
    /**
     * @stability stable
     */
    resetClusterPublicKey() {
        this._clusterPublicKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterPublicKeyInput() {
        return this._clusterPublicKey;
    }
    /**
     * @stability stable
     */
    get clusterRevisionNumber() {
        return this.getStringAttribute('cluster_revision_number');
    }
    /**
     * @stability stable
     */
    set clusterRevisionNumber(value) {
        this._clusterRevisionNumber = value;
    }
    /**
     * @stability stable
     */
    resetClusterRevisionNumber() {
        this._clusterRevisionNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterRevisionNumberInput() {
        return this._clusterRevisionNumber;
    }
    /**
     * @stability stable
     */
    get clusterSecurityGroups() {
        return this.getListAttribute('cluster_security_groups');
    }
    /**
     * @stability stable
     */
    set clusterSecurityGroups(value) {
        this._clusterSecurityGroups = value;
    }
    /**
     * @stability stable
     */
    resetClusterSecurityGroups() {
        this._clusterSecurityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterSecurityGroupsInput() {
        return this._clusterSecurityGroups;
    }
    /**
     * @stability stable
     */
    get clusterSubnetGroupName() {
        return this.getStringAttribute('cluster_subnet_group_name');
    }
    /**
     * @stability stable
     */
    set clusterSubnetGroupName(value) {
        this._clusterSubnetGroupName = value;
    }
    /**
     * @stability stable
     */
    resetClusterSubnetGroupName() {
        this._clusterSubnetGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterSubnetGroupNameInput() {
        return this._clusterSubnetGroupName;
    }
    /**
     * @stability stable
     */
    get clusterType() {
        return this.getStringAttribute('cluster_type');
    }
    /**
     * @stability stable
     */
    set clusterType(value) {
        this._clusterType = value;
    }
    /**
     * @stability stable
     */
    resetClusterType() {
        this._clusterType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterTypeInput() {
        return this._clusterType;
    }
    /**
     * @stability stable
     */
    get clusterVersion() {
        return this.getStringAttribute('cluster_version');
    }
    /**
     * @stability stable
     */
    set clusterVersion(value) {
        this._clusterVersion = value;
    }
    /**
     * @stability stable
     */
    resetClusterVersion() {
        this._clusterVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterVersionInput() {
        return this._clusterVersion;
    }
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    /**
     * @stability stable
     */
    set databaseName(value) {
        this._databaseName = value;
    }
    /**
     * @stability stable
     */
    resetDatabaseName() {
        this._databaseName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseNameInput() {
        return this._databaseName;
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    /**
     * @stability stable
     */
    get elasticIp() {
        return this.getStringAttribute('elastic_ip');
    }
    /**
     * @stability stable
     */
    set elasticIp(value) {
        this._elasticIp = value;
    }
    /**
     * @stability stable
     */
    resetElasticIp() {
        this._elasticIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticIpInput() {
        return this._elasticIp;
    }
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    /**
     * @stability stable
     */
    set encrypted(value) {
        this._encrypted = value;
    }
    /**
     * @stability stable
     */
    resetEncrypted() {
        this._encrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptedInput() {
        return this._encrypted;
    }
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    /**
     * @stability stable
     */
    set endpoint(value) {
        this._endpoint = value;
    }
    /**
     * @stability stable
     */
    resetEndpoint() {
        this._endpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointInput() {
        return this._endpoint;
    }
    /**
     * @stability stable
     */
    get enhancedVpcRouting() {
        return this.getBooleanAttribute('enhanced_vpc_routing');
    }
    /**
     * @stability stable
     */
    set enhancedVpcRouting(value) {
        this._enhancedVpcRouting = value;
    }
    /**
     * @stability stable
     */
    resetEnhancedVpcRouting() {
        this._enhancedVpcRouting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enhancedVpcRoutingInput() {
        return this._enhancedVpcRouting;
    }
    /**
     * @stability stable
     */
    get finalSnapshotIdentifier() {
        return this.getStringAttribute('final_snapshot_identifier');
    }
    /**
     * @stability stable
     */
    set finalSnapshotIdentifier(value) {
        this._finalSnapshotIdentifier = value;
    }
    /**
     * @stability stable
     */
    resetFinalSnapshotIdentifier() {
        this._finalSnapshotIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get finalSnapshotIdentifierInput() {
        return this._finalSnapshotIdentifier;
    }
    /**
     * @stability stable
     */
    get iamRoles() {
        return this.getListAttribute('iam_roles');
    }
    /**
     * @stability stable
     */
    set iamRoles(value) {
        this._iamRoles = value;
    }
    /**
     * @stability stable
     */
    resetIamRoles() {
        this._iamRoles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamRolesInput() {
        return this._iamRoles;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    /**
     * @stability stable
     */
    get masterPassword() {
        return this.getStringAttribute('master_password');
    }
    /**
     * @stability stable
     */
    set masterPassword(value) {
        this._masterPassword = value;
    }
    /**
     * @stability stable
     */
    resetMasterPassword() {
        this._masterPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get masterPasswordInput() {
        return this._masterPassword;
    }
    /**
     * @stability stable
     */
    get masterUsername() {
        return this.getStringAttribute('master_username');
    }
    /**
     * @stability stable
     */
    set masterUsername(value) {
        this._masterUsername = value;
    }
    /**
     * @stability stable
     */
    resetMasterUsername() {
        this._masterUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get masterUsernameInput() {
        return this._masterUsername;
    }
    /**
     * @stability stable
     */
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    /**
     * @stability stable
     */
    set nodeType(value) {
        this._nodeType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeTypeInput() {
        return this._nodeType;
    }
    /**
     * @stability stable
     */
    get numberOfNodes() {
        return this.getNumberAttribute('number_of_nodes');
    }
    /**
     * @stability stable
     */
    set numberOfNodes(value) {
        this._numberOfNodes = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfNodes() {
        this._numberOfNodes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfNodesInput() {
        return this._numberOfNodes;
    }
    /**
     * @stability stable
     */
    get ownerAccount() {
        return this.getStringAttribute('owner_account');
    }
    /**
     * @stability stable
     */
    set ownerAccount(value) {
        this._ownerAccount = value;
    }
    /**
     * @stability stable
     */
    resetOwnerAccount() {
        this._ownerAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerAccountInput() {
        return this._ownerAccount;
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    /**
     * @stability stable
     */
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    /**
     * @stability stable
     */
    get preferredMaintenanceWindow() {
        return this.getStringAttribute('preferred_maintenance_window');
    }
    /**
     * @stability stable
     */
    set preferredMaintenanceWindow(value) {
        this._preferredMaintenanceWindow = value;
    }
    /**
     * @stability stable
     */
    resetPreferredMaintenanceWindow() {
        this._preferredMaintenanceWindow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get preferredMaintenanceWindowInput() {
        return this._preferredMaintenanceWindow;
    }
    /**
     * @stability stable
     */
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    /**
     * @stability stable
     */
    set publiclyAccessible(value) {
        this._publiclyAccessible = value;
    }
    /**
     * @stability stable
     */
    resetPubliclyAccessible() {
        this._publiclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publiclyAccessibleInput() {
        return this._publiclyAccessible;
    }
    /**
     * @stability stable
     */
    get skipFinalSnapshot() {
        return this.getBooleanAttribute('skip_final_snapshot');
    }
    /**
     * @stability stable
     */
    set skipFinalSnapshot(value) {
        this._skipFinalSnapshot = value;
    }
    /**
     * @stability stable
     */
    resetSkipFinalSnapshot() {
        this._skipFinalSnapshot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skipFinalSnapshotInput() {
        return this._skipFinalSnapshot;
    }
    /**
     * @stability stable
     */
    get snapshotClusterIdentifier() {
        return this.getStringAttribute('snapshot_cluster_identifier');
    }
    /**
     * @stability stable
     */
    set snapshotClusterIdentifier(value) {
        this._snapshotClusterIdentifier = value;
    }
    /**
     * @stability stable
     */
    resetSnapshotClusterIdentifier() {
        this._snapshotClusterIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snapshotClusterIdentifierInput() {
        return this._snapshotClusterIdentifier;
    }
    /**
     * @stability stable
     */
    get snapshotIdentifier() {
        return this.getStringAttribute('snapshot_identifier');
    }
    /**
     * @stability stable
     */
    set snapshotIdentifier(value) {
        this._snapshotIdentifier = value;
    }
    /**
     * @stability stable
     */
    resetSnapshotIdentifier() {
        this._snapshotIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snapshotIdentifierInput() {
        return this._snapshotIdentifier;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    /**
     * @stability stable
     */
    set vpcSecurityGroupIds(value) {
        this._vpcSecurityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetVpcSecurityGroupIds() {
        this._vpcSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcSecurityGroupIdsInput() {
        return this._vpcSecurityGroupIds;
    }
    /**
     * @stability stable
     */
    get logging() {
        return this._logging;
    }
    /**
     * @stability stable
     */
    putLogging(value) {
        this._logging.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLogging() {
        this._logging.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingInput() {
        return this._logging.internalValue;
    }
    /**
     * @stability stable
     */
    get snapshotCopy() {
        return this._snapshotCopy;
    }
    /**
     * @stability stable
     */
    putSnapshotCopy(value) {
        this._snapshotCopy.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSnapshotCopy() {
        this._snapshotCopy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snapshotCopyInput() {
        return this._snapshotCopy.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            allow_version_upgrade: cdktf.booleanToTerraform(this._allowVersionUpgrade),
            automated_snapshot_retention_period: cdktf.numberToTerraform(this._automatedSnapshotRetentionPeriod),
            availability_zone: cdktf.stringToTerraform(this._availabilityZone),
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            cluster_parameter_group_name: cdktf.stringToTerraform(this._clusterParameterGroupName),
            cluster_public_key: cdktf.stringToTerraform(this._clusterPublicKey),
            cluster_revision_number: cdktf.stringToTerraform(this._clusterRevisionNumber),
            cluster_security_groups: cdktf.listMapper(cdktf.stringToTerraform)(this._clusterSecurityGroups),
            cluster_subnet_group_name: cdktf.stringToTerraform(this._clusterSubnetGroupName),
            cluster_type: cdktf.stringToTerraform(this._clusterType),
            cluster_version: cdktf.stringToTerraform(this._clusterVersion),
            database_name: cdktf.stringToTerraform(this._databaseName),
            elastic_ip: cdktf.stringToTerraform(this._elasticIp),
            encrypted: cdktf.booleanToTerraform(this._encrypted),
            endpoint: cdktf.stringToTerraform(this._endpoint),
            enhanced_vpc_routing: cdktf.booleanToTerraform(this._enhancedVpcRouting),
            final_snapshot_identifier: cdktf.stringToTerraform(this._finalSnapshotIdentifier),
            iam_roles: cdktf.listMapper(cdktf.stringToTerraform)(this._iamRoles),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            master_password: cdktf.stringToTerraform(this._masterPassword),
            master_username: cdktf.stringToTerraform(this._masterUsername),
            node_type: cdktf.stringToTerraform(this._nodeType),
            number_of_nodes: cdktf.numberToTerraform(this._numberOfNodes),
            owner_account: cdktf.stringToTerraform(this._ownerAccount),
            port: cdktf.numberToTerraform(this._port),
            preferred_maintenance_window: cdktf.stringToTerraform(this._preferredMaintenanceWindow),
            publicly_accessible: cdktf.booleanToTerraform(this._publiclyAccessible),
            skip_final_snapshot: cdktf.booleanToTerraform(this._skipFinalSnapshot),
            snapshot_cluster_identifier: cdktf.stringToTerraform(this._snapshotClusterIdentifier),
            snapshot_identifier: cdktf.stringToTerraform(this._snapshotIdentifier),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            vpc_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._vpcSecurityGroupIds),
            logging: redshiftClusterLoggingToTerraform(this._logging.internalValue),
            snapshot_copy: redshiftClusterSnapshotCopyToTerraform(this._snapshotCopy.internalValue),
            timeouts: redshiftClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.RedshiftCluster = RedshiftCluster;
_e = JSII_RTTI_SYMBOL_1;
RedshiftCluster[_e] = { fqn: "@cdktf/provider-aws.redshift.RedshiftCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
RedshiftCluster.tfResourceType = "aws_redshift_cluster";
