// @cdktf/provider-aws/lib/eventbridge/cloudwatch-event-connection.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventConnection = exports.CloudwatchEventConnectionAuthParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform = exports.CloudwatchEventConnectionAuthParametersBasicOutputReference = exports.cloudwatchEventConnectionAuthParametersBasicToTerraform = exports.CloudwatchEventConnectionAuthParametersApiKeyOutputReference = exports.cloudwatchEventConnectionAuthParametersApiKeyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudwatchEventConnectionAuthParametersApiKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersApiKeyToTerraform = cloudwatchEventConnectionAuthParametersApiKeyToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersApiKeyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._value = value.value;
        }
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get value() {
        return this.getStringAttribute('value');
    }
    /**
     * @stability stable
     */
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersApiKeyOutputReference = CloudwatchEventConnectionAuthParametersApiKeyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersApiKeyOutputReference[_a] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersApiKeyOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersBasicToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        password: cdktf.stringToTerraform(struct.password),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.cloudwatchEventConnectionAuthParametersBasicToTerraform = cloudwatchEventConnectionAuthParametersBasicToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersBasicOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._password) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._username) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._password = undefined;
            this._username = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._password = value.password;
            this._username = value.username;
        }
    }
    /**
     * @stability stable
     */
    get password() {
        return this.getStringAttribute('password');
    }
    /**
     * @stability stable
     */
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passwordInput() {
        return this._password;
    }
    /**
     * @stability stable
     */
    get username() {
        return this.getStringAttribute('username');
    }
    /**
     * @stability stable
     */
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get usernameInput() {
        return this._username;
    }
}
exports.CloudwatchEventConnectionAuthParametersBasicOutputReference = CloudwatchEventConnectionAuthParametersBasicOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersBasicOutputReference[_b] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersBasicOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform;
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform;
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform;
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        body: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform)(struct.body),
        header: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform)(struct.header),
        query_string: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform)(struct.queryString),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._body) {
            hasAnyValues = true;
            internalValueResult.body = this._body;
        }
        if (this._header) {
            hasAnyValues = true;
            internalValueResult.header = this._header;
        }
        if (this._queryString) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._body = undefined;
            this._header = undefined;
            this._queryString = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._body = value.body;
            this._header = value.header;
            this._queryString = value.queryString;
        }
    }
    /**
     * @stability stable
     */
    get body() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('body');
    }
    /**
     * @stability stable
     */
    set body(value) {
        this._body = value;
    }
    /**
     * @stability stable
     */
    resetBody() {
        this._body = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bodyInput() {
        return this._body;
    }
    /**
     * @stability stable
     */
    get header() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('header');
    }
    /**
     * @stability stable
     */
    set header(value) {
        this._header = value;
    }
    /**
     * @stability stable
     */
    resetHeader() {
        this._header = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headerInput() {
        return this._header;
    }
    /**
     * @stability stable
     */
    get queryString() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('query_string');
    }
    /**
     * @stability stable
     */
    set queryString(value) {
        this._queryString = value;
    }
    /**
     * @stability stable
     */
    resetQueryString() {
        this._queryString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringInput() {
        return this._queryString;
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference = CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference[_c] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_id: cdktf.stringToTerraform(struct.clientId),
        client_secret: cdktf.stringToTerraform(struct.clientSecret),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform = cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientId) {
            hasAnyValues = true;
            internalValueResult.clientId = this._clientId;
        }
        if (this._clientSecret) {
            hasAnyValues = true;
            internalValueResult.clientSecret = this._clientSecret;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientId = undefined;
            this._clientSecret = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientId = value.clientId;
            this._clientSecret = value.clientSecret;
        }
    }
    /**
     * @stability stable
     */
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    /**
     * @stability stable
     */
    set clientId(value) {
        this._clientId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientIdInput() {
        return this._clientId;
    }
    /**
     * @stability stable
     */
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    /**
     * @stability stable
     */
    set clientSecret(value) {
        this._clientSecret = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientSecretInput() {
        return this._clientSecret;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference = CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference[_d] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform;
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform;
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform;
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        body: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform)(struct.body),
        header: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform)(struct.header),
        query_string: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform)(struct.queryString),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._body) {
            hasAnyValues = true;
            internalValueResult.body = this._body;
        }
        if (this._header) {
            hasAnyValues = true;
            internalValueResult.header = this._header;
        }
        if (this._queryString) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._body = undefined;
            this._header = undefined;
            this._queryString = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._body = value.body;
            this._header = value.header;
            this._queryString = value.queryString;
        }
    }
    /**
     * @stability stable
     */
    get body() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('body');
    }
    /**
     * @stability stable
     */
    set body(value) {
        this._body = value;
    }
    /**
     * @stability stable
     */
    resetBody() {
        this._body = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bodyInput() {
        return this._body;
    }
    /**
     * @stability stable
     */
    get header() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('header');
    }
    /**
     * @stability stable
     */
    set header(value) {
        this._header = value;
    }
    /**
     * @stability stable
     */
    resetHeader() {
        this._header = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get headerInput() {
        return this._header;
    }
    /**
     * @stability stable
     */
    get queryString() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('query_string');
    }
    /**
     * @stability stable
     */
    set queryString(value) {
        this._queryString = value;
    }
    /**
     * @stability stable
     */
    resetQueryString() {
        this._queryString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queryStringInput() {
        return this._queryString;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference[_e] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersOauthToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authorization_endpoint: cdktf.stringToTerraform(struct.authorizationEndpoint),
        http_method: cdktf.stringToTerraform(struct.httpMethod),
        client_parameters: cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform(struct.clientParameters),
        oauth_http_parameters: cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform(struct.oauthHttpParameters),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthToTerraform = cloudwatchEventConnectionAuthParametersOauthToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersOauthOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // client_parameters - computed: false, optional: true, required: false
        this._clientParameters = new CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference(this, "client_parameters", true);
        // oauth_http_parameters - computed: false, optional: false, required: true
        this._oauthHttpParameters = new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference(this, "oauth_http_parameters", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k, _l, _m;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authorizationEndpoint) {
            hasAnyValues = true;
            internalValueResult.authorizationEndpoint = this._authorizationEndpoint;
        }
        if (this._httpMethod) {
            hasAnyValues = true;
            internalValueResult.httpMethod = this._httpMethod;
        }
        if ((_j = this._clientParameters) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.clientParameters = (_k = this._clientParameters) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._oauthHttpParameters) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.oauthHttpParameters = (_m = this._oauthHttpParameters) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authorizationEndpoint = undefined;
            this._httpMethod = undefined;
            this._clientParameters.internalValue = undefined;
            this._oauthHttpParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authorizationEndpoint = value.authorizationEndpoint;
            this._httpMethod = value.httpMethod;
            this._clientParameters.internalValue = value.clientParameters;
            this._oauthHttpParameters.internalValue = value.oauthHttpParameters;
        }
    }
    /**
     * @stability stable
     */
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    /**
     * @stability stable
     */
    set authorizationEndpoint(value) {
        this._authorizationEndpoint = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizationEndpointInput() {
        return this._authorizationEndpoint;
    }
    /**
     * @stability stable
     */
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    /**
     * @stability stable
     */
    set httpMethod(value) {
        this._httpMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpMethodInput() {
        return this._httpMethod;
    }
    /**
     * @stability stable
     */
    get clientParameters() {
        return this._clientParameters;
    }
    /**
     * @stability stable
     */
    putClientParameters(value) {
        this._clientParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetClientParameters() {
        this._clientParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientParametersInput() {
        return this._clientParameters.internalValue;
    }
    /**
     * @stability stable
     */
    get oauthHttpParameters() {
        return this._oauthHttpParameters;
    }
    /**
     * @stability stable
     */
    putOauthHttpParameters(value) {
        this._oauthHttpParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get oauthHttpParametersInput() {
        return this._oauthHttpParameters.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOutputReference = CloudwatchEventConnectionAuthParametersOauthOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOutputReference[_f] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersOauthOutputReference", version: "3.0.1" };
function cloudwatchEventConnectionAuthParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        api_key: cloudwatchEventConnectionAuthParametersApiKeyToTerraform(struct.apiKey),
        basic: cloudwatchEventConnectionAuthParametersBasicToTerraform(struct.basic),
        invocation_http_parameters: cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform(struct.invocationHttpParameters),
        oauth: cloudwatchEventConnectionAuthParametersOauthToTerraform(struct.oauth),
    };
}
exports.cloudwatchEventConnectionAuthParametersToTerraform = cloudwatchEventConnectionAuthParametersToTerraform;
/**
 * @stability stable
 */
class CloudwatchEventConnectionAuthParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // api_key - computed: false, optional: true, required: false
        this._apiKey = new CloudwatchEventConnectionAuthParametersApiKeyOutputReference(this, "api_key", true);
        // basic - computed: false, optional: true, required: false
        this._basic = new CloudwatchEventConnectionAuthParametersBasicOutputReference(this, "basic", true);
        // invocation_http_parameters - computed: false, optional: true, required: false
        this._invocationHttpParameters = new CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference(this, "invocation_http_parameters", true);
        // oauth - computed: false, optional: true, required: false
        this._oauth = new CloudwatchEventConnectionAuthParametersOauthOutputReference(this, "oauth", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k, _l, _m, _o, _p, _q, _r;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_j = this._apiKey) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.apiKey = (_k = this._apiKey) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._basic) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.basic = (_m = this._basic) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        if ((_o = this._invocationHttpParameters) === null || _o === void 0 ? void 0 : _o.internalValue) {
            hasAnyValues = true;
            internalValueResult.invocationHttpParameters = (_p = this._invocationHttpParameters) === null || _p === void 0 ? void 0 : _p.internalValue;
        }
        if ((_q = this._oauth) === null || _q === void 0 ? void 0 : _q.internalValue) {
            hasAnyValues = true;
            internalValueResult.oauth = (_r = this._oauth) === null || _r === void 0 ? void 0 : _r.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._apiKey.internalValue = undefined;
            this._basic.internalValue = undefined;
            this._invocationHttpParameters.internalValue = undefined;
            this._oauth.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._apiKey.internalValue = value.apiKey;
            this._basic.internalValue = value.basic;
            this._invocationHttpParameters.internalValue = value.invocationHttpParameters;
            this._oauth.internalValue = value.oauth;
        }
    }
    /**
     * @stability stable
     */
    get apiKey() {
        return this._apiKey;
    }
    /**
     * @stability stable
     */
    putApiKey(value) {
        this._apiKey.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetApiKey() {
        this._apiKey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiKeyInput() {
        return this._apiKey.internalValue;
    }
    /**
     * @stability stable
     */
    get basic() {
        return this._basic;
    }
    /**
     * @stability stable
     */
    putBasic(value) {
        this._basic.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetBasic() {
        this._basic.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get basicInput() {
        return this._basic.internalValue;
    }
    /**
     * @stability stable
     */
    get invocationHttpParameters() {
        return this._invocationHttpParameters;
    }
    /**
     * @stability stable
     */
    putInvocationHttpParameters(value) {
        this._invocationHttpParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetInvocationHttpParameters() {
        this._invocationHttpParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get invocationHttpParametersInput() {
        return this._invocationHttpParameters.internalValue;
    }
    /**
     * @stability stable
     */
    get oauth() {
        return this._oauth;
    }
    /**
     * @stability stable
     */
    putOauth(value) {
        this._oauth.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetOauth() {
        this._oauth.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get oauthInput() {
        return this._oauth.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersOutputReference = CloudwatchEventConnectionAuthParametersOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOutputReference[_g] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnectionAuthParametersOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_connection.html aws_cloudwatch_event_connection}.
 *
 * @stability stable
 */
class CloudwatchEventConnection extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_connection.html aws_cloudwatch_event_connection} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // auth_parameters - computed: false, optional: false, required: true
        this._authParameters = new CloudwatchEventConnectionAuthParametersOutputReference(this, "auth_parameters", true);
        this._authorizationType = config.authorizationType;
        this._description = config.description;
        this._name = config.name;
        this._authParameters.internalValue = config.authParameters;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get authorizationType() {
        return this.getStringAttribute('authorization_type');
    }
    /**
     * @stability stable
     */
    set authorizationType(value) {
        this._authorizationType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizationTypeInput() {
        return this._authorizationType;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // secret_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    /**
     * @stability stable
     */
    get authParameters() {
        return this._authParameters;
    }
    /**
     * @stability stable
     */
    putAuthParameters(value) {
        this._authParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authParametersInput() {
        return this._authParameters.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            authorization_type: cdktf.stringToTerraform(this._authorizationType),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            auth_parameters: cloudwatchEventConnectionAuthParametersToTerraform(this._authParameters.internalValue),
        };
    }
}
exports.CloudwatchEventConnection = CloudwatchEventConnection;
_h = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnection[_h] = { fqn: "@cdktf/provider-aws.eventbridge.CloudwatchEventConnection", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudwatchEventConnection.tfResourceType = "aws_cloudwatch_event_connection";
