// @cdktf/provider-aws/lib/pinpoint/pinpoint-app.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinpointApp = exports.PinpointAppQuietTimeOutputReference = exports.pinpointAppQuietTimeToTerraform = exports.PinpointAppLimitsOutputReference = exports.pinpointAppLimitsToTerraform = exports.PinpointAppCampaignHookOutputReference = exports.pinpointAppCampaignHookToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function pinpointAppCampaignHookToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lambda_function_name: cdktf.stringToTerraform(struct.lambdaFunctionName),
        mode: cdktf.stringToTerraform(struct.mode),
        web_url: cdktf.stringToTerraform(struct.webUrl),
    };
}
exports.pinpointAppCampaignHookToTerraform = pinpointAppCampaignHookToTerraform;
/**
 * @stability stable
 */
class PinpointAppCampaignHookOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lambdaFunctionName) {
            hasAnyValues = true;
            internalValueResult.lambdaFunctionName = this._lambdaFunctionName;
        }
        if (this._mode) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._webUrl) {
            hasAnyValues = true;
            internalValueResult.webUrl = this._webUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lambdaFunctionName = undefined;
            this._mode = undefined;
            this._webUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lambdaFunctionName = value.lambdaFunctionName;
            this._mode = value.mode;
            this._webUrl = value.webUrl;
        }
    }
    /**
     * @stability stable
     */
    get lambdaFunctionName() {
        return this.getStringAttribute('lambda_function_name');
    }
    /**
     * @stability stable
     */
    set lambdaFunctionName(value) {
        this._lambdaFunctionName = value;
    }
    /**
     * @stability stable
     */
    resetLambdaFunctionName() {
        this._lambdaFunctionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaFunctionNameInput() {
        return this._lambdaFunctionName;
    }
    /**
     * @stability stable
     */
    get mode() {
        return this.getStringAttribute('mode');
    }
    /**
     * @stability stable
     */
    set mode(value) {
        this._mode = value;
    }
    /**
     * @stability stable
     */
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modeInput() {
        return this._mode;
    }
    /**
     * @stability stable
     */
    get webUrl() {
        return this.getStringAttribute('web_url');
    }
    /**
     * @stability stable
     */
    set webUrl(value) {
        this._webUrl = value;
    }
    /**
     * @stability stable
     */
    resetWebUrl() {
        this._webUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get webUrlInput() {
        return this._webUrl;
    }
}
exports.PinpointAppCampaignHookOutputReference = PinpointAppCampaignHookOutputReference;
_a = JSII_RTTI_SYMBOL_1;
PinpointAppCampaignHookOutputReference[_a] = { fqn: "@cdktf/provider-aws.pinpoint.PinpointAppCampaignHookOutputReference", version: "3.0.1" };
function pinpointAppLimitsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        daily: cdktf.numberToTerraform(struct.daily),
        maximum_duration: cdktf.numberToTerraform(struct.maximumDuration),
        messages_per_second: cdktf.numberToTerraform(struct.messagesPerSecond),
        total: cdktf.numberToTerraform(struct.total),
    };
}
exports.pinpointAppLimitsToTerraform = pinpointAppLimitsToTerraform;
/**
 * @stability stable
 */
class PinpointAppLimitsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daily) {
            hasAnyValues = true;
            internalValueResult.daily = this._daily;
        }
        if (this._maximumDuration) {
            hasAnyValues = true;
            internalValueResult.maximumDuration = this._maximumDuration;
        }
        if (this._messagesPerSecond) {
            hasAnyValues = true;
            internalValueResult.messagesPerSecond = this._messagesPerSecond;
        }
        if (this._total) {
            hasAnyValues = true;
            internalValueResult.total = this._total;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._daily = undefined;
            this._maximumDuration = undefined;
            this._messagesPerSecond = undefined;
            this._total = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._daily = value.daily;
            this._maximumDuration = value.maximumDuration;
            this._messagesPerSecond = value.messagesPerSecond;
            this._total = value.total;
        }
    }
    /**
     * @stability stable
     */
    get daily() {
        return this.getNumberAttribute('daily');
    }
    /**
     * @stability stable
     */
    set daily(value) {
        this._daily = value;
    }
    /**
     * @stability stable
     */
    resetDaily() {
        this._daily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dailyInput() {
        return this._daily;
    }
    /**
     * @stability stable
     */
    get maximumDuration() {
        return this.getNumberAttribute('maximum_duration');
    }
    /**
     * @stability stable
     */
    set maximumDuration(value) {
        this._maximumDuration = value;
    }
    /**
     * @stability stable
     */
    resetMaximumDuration() {
        this._maximumDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maximumDurationInput() {
        return this._maximumDuration;
    }
    /**
     * @stability stable
     */
    get messagesPerSecond() {
        return this.getNumberAttribute('messages_per_second');
    }
    /**
     * @stability stable
     */
    set messagesPerSecond(value) {
        this._messagesPerSecond = value;
    }
    /**
     * @stability stable
     */
    resetMessagesPerSecond() {
        this._messagesPerSecond = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messagesPerSecondInput() {
        return this._messagesPerSecond;
    }
    /**
     * @stability stable
     */
    get total() {
        return this.getNumberAttribute('total');
    }
    /**
     * @stability stable
     */
    set total(value) {
        this._total = value;
    }
    /**
     * @stability stable
     */
    resetTotal() {
        this._total = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get totalInput() {
        return this._total;
    }
}
exports.PinpointAppLimitsOutputReference = PinpointAppLimitsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
PinpointAppLimitsOutputReference[_b] = { fqn: "@cdktf/provider-aws.pinpoint.PinpointAppLimitsOutputReference", version: "3.0.1" };
function pinpointAppQuietTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end: cdktf.stringToTerraform(struct.end),
        start: cdktf.stringToTerraform(struct.start),
    };
}
exports.pinpointAppQuietTimeToTerraform = pinpointAppQuietTimeToTerraform;
/**
 * @stability stable
 */
class PinpointAppQuietTimeOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._end) {
            hasAnyValues = true;
            internalValueResult.end = this._end;
        }
        if (this._start) {
            hasAnyValues = true;
            internalValueResult.start = this._start;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._end = undefined;
            this._start = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._end = value.end;
            this._start = value.start;
        }
    }
    /**
     * @stability stable
     */
    get end() {
        return this.getStringAttribute('end');
    }
    /**
     * @stability stable
     */
    set end(value) {
        this._end = value;
    }
    /**
     * @stability stable
     */
    resetEnd() {
        this._end = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endInput() {
        return this._end;
    }
    /**
     * @stability stable
     */
    get start() {
        return this.getStringAttribute('start');
    }
    /**
     * @stability stable
     */
    set start(value) {
        this._start = value;
    }
    /**
     * @stability stable
     */
    resetStart() {
        this._start = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startInput() {
        return this._start;
    }
}
exports.PinpointAppQuietTimeOutputReference = PinpointAppQuietTimeOutputReference;
_c = JSII_RTTI_SYMBOL_1;
PinpointAppQuietTimeOutputReference[_c] = { fqn: "@cdktf/provider-aws.pinpoint.PinpointAppQuietTimeOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/pinpoint_app.html aws_pinpoint_app}.
 *
 * @stability stable
 */
class PinpointApp extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/pinpoint_app.html aws_pinpoint_app} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_pinpoint_app',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // campaign_hook - computed: false, optional: true, required: false
        this._campaignHook = new PinpointAppCampaignHookOutputReference(this, "campaign_hook", true);
        // limits - computed: false, optional: true, required: false
        this._limits = new PinpointAppLimitsOutputReference(this, "limits", true);
        // quiet_time - computed: false, optional: true, required: false
        this._quietTime = new PinpointAppQuietTimeOutputReference(this, "quiet_time", true);
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._campaignHook.internalValue = config.campaignHook;
        this._limits.internalValue = config.limits;
        this._quietTime.internalValue = config.quietTime;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // application_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get campaignHook() {
        return this._campaignHook;
    }
    /**
     * @stability stable
     */
    putCampaignHook(value) {
        this._campaignHook.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCampaignHook() {
        this._campaignHook.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get campaignHookInput() {
        return this._campaignHook.internalValue;
    }
    /**
     * @stability stable
     */
    get limits() {
        return this._limits;
    }
    /**
     * @stability stable
     */
    putLimits(value) {
        this._limits.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLimits() {
        this._limits.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get limitsInput() {
        return this._limits.internalValue;
    }
    /**
     * @stability stable
     */
    get quietTime() {
        return this._quietTime;
    }
    /**
     * @stability stable
     */
    putQuietTime(value) {
        this._quietTime.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetQuietTime() {
        this._quietTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get quietTimeInput() {
        return this._quietTime.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            campaign_hook: pinpointAppCampaignHookToTerraform(this._campaignHook.internalValue),
            limits: pinpointAppLimitsToTerraform(this._limits.internalValue),
            quiet_time: pinpointAppQuietTimeToTerraform(this._quietTime.internalValue),
        };
    }
}
exports.PinpointApp = PinpointApp;
_d = JSII_RTTI_SYMBOL_1;
PinpointApp[_d] = { fqn: "@cdktf/provider-aws.pinpoint.PinpointApp", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
PinpointApp.tfResourceType = "aws_pinpoint_app";
