// @cdktf/provider-aws/lib/lambdafunction/data-aws-lambda-function.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLambdaFunction = exports.DataAwsLambdaFunctionVpcConfig = exports.DataAwsLambdaFunctionTracingConfig = exports.DataAwsLambdaFunctionFileSystemConfig = exports.DataAwsLambdaFunctionEnvironment = exports.DataAwsLambdaFunctionDeadLetterConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsLambdaFunctionDeadLetterConfig extends cdktf.ComplexComputedList {
    // target_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
}
exports.DataAwsLambdaFunctionDeadLetterConfig = DataAwsLambdaFunctionDeadLetterConfig;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionDeadLetterConfig[_a] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunctionDeadLetterConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLambdaFunctionEnvironment extends cdktf.ComplexComputedList {
    // variables - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get variables() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('variables');
    }
}
exports.DataAwsLambdaFunctionEnvironment = DataAwsLambdaFunctionEnvironment;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionEnvironment[_b] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunctionEnvironment", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLambdaFunctionFileSystemConfig extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // local_mount_path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get localMountPath() {
        return this.getStringAttribute('local_mount_path');
    }
}
exports.DataAwsLambdaFunctionFileSystemConfig = DataAwsLambdaFunctionFileSystemConfig;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionFileSystemConfig[_c] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunctionFileSystemConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLambdaFunctionTracingConfig extends cdktf.ComplexComputedList {
    // mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataAwsLambdaFunctionTracingConfig = DataAwsLambdaFunctionTracingConfig;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionTracingConfig[_d] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunctionTracingConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLambdaFunctionVpcConfig extends cdktf.ComplexComputedList {
    // security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    // subnet_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsLambdaFunctionVpcConfig = DataAwsLambdaFunctionVpcConfig;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunctionVpcConfig[_e] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunctionVpcConfig", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/lambda_function.html aws_lambda_function}.
 *
 * @stability stable
 */
class DataAwsLambdaFunction extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/lambda_function.html aws_lambda_function} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_function',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._functionName = config.functionName;
        this._qualifier = config.qualifier;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // architectures - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get architectures() {
        return this.getListAttribute('architectures');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // code_signing_config_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get codeSigningConfigArn() {
        return this.getStringAttribute('code_signing_config_arn');
    }
    // dead_letter_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    deadLetterConfig(index) {
        return new DataAwsLambdaFunctionDeadLetterConfig(this, 'dead_letter_config', index);
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // environment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    environment(index) {
        return new DataAwsLambdaFunctionEnvironment(this, 'environment', index);
    }
    // file_system_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    fileSystemConfig(index) {
        return new DataAwsLambdaFunctionFileSystemConfig(this, 'file_system_config', index);
    }
    /**
     * @stability stable
     */
    get functionName() {
        return this.getStringAttribute('function_name');
    }
    /**
     * @stability stable
     */
    set functionName(value) {
        this._functionName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get functionNameInput() {
        return this._functionName;
    }
    // handler - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get handler() {
        return this.getStringAttribute('handler');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image_uri - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageUri() {
        return this.getStringAttribute('image_uri');
    }
    // invoke_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get invokeArn() {
        return this.getStringAttribute('invoke_arn');
    }
    // kms_key_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    // last_modified - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    // layers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get layers() {
        return this.getListAttribute('layers');
    }
    // memory_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get memorySize() {
        return this.getNumberAttribute('memory_size');
    }
    // qualified_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get qualifiedArn() {
        return this.getStringAttribute('qualified_arn');
    }
    /**
     * @stability stable
     */
    get qualifier() {
        return this.getStringAttribute('qualifier');
    }
    /**
     * @stability stable
     */
    set qualifier(value) {
        this._qualifier = value;
    }
    /**
     * @stability stable
     */
    resetQualifier() {
        this._qualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get qualifierInput() {
        return this._qualifier;
    }
    // reserved_concurrent_executions - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get reservedConcurrentExecutions() {
        return this.getNumberAttribute('reserved_concurrent_executions');
    }
    // role - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get role() {
        return this.getStringAttribute('role');
    }
    // runtime - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get runtime() {
        return this.getStringAttribute('runtime');
    }
    // signing_job_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get signingJobArn() {
        return this.getStringAttribute('signing_job_arn');
    }
    // signing_profile_version_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get signingProfileVersionArn() {
        return this.getStringAttribute('signing_profile_version_arn');
    }
    // source_code_hash - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceCodeHash() {
        return this.getStringAttribute('source_code_hash');
    }
    // source_code_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceCodeSize() {
        return this.getNumberAttribute('source_code_size');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    // tracing_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    tracingConfig(index) {
        return new DataAwsLambdaFunctionTracingConfig(this, 'tracing_config', index);
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    // vpc_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    vpcConfig(index) {
        return new DataAwsLambdaFunctionVpcConfig(this, 'vpc_config', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            function_name: cdktf.stringToTerraform(this._functionName),
            qualifier: cdktf.stringToTerraform(this._qualifier),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsLambdaFunction = DataAwsLambdaFunction;
_f = JSII_RTTI_SYMBOL_1;
DataAwsLambdaFunction[_f] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaFunction", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsLambdaFunction.tfResourceType = "aws_lambda_function";
