// @cdktf/provider-aws/lib/lambdafunction/data-aws-lambda-code-signing-config.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLambdaCodeSigningConfig = exports.DataAwsLambdaCodeSigningConfigPolicies = exports.DataAwsLambdaCodeSigningConfigAllowedPublishers = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsLambdaCodeSigningConfigAllowedPublishers extends cdktf.ComplexComputedList {
    // signing_profile_version_arns - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get signingProfileVersionArns() {
        return this.getListAttribute('signing_profile_version_arns');
    }
}
exports.DataAwsLambdaCodeSigningConfigAllowedPublishers = DataAwsLambdaCodeSigningConfigAllowedPublishers;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigAllowedPublishers[_a] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaCodeSigningConfigAllowedPublishers", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLambdaCodeSigningConfigPolicies extends cdktf.ComplexComputedList {
    // untrusted_artifact_on_deployment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get untrustedArtifactOnDeployment() {
        return this.getStringAttribute('untrusted_artifact_on_deployment');
    }
}
exports.DataAwsLambdaCodeSigningConfigPolicies = DataAwsLambdaCodeSigningConfigPolicies;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfigPolicies[_b] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaCodeSigningConfigPolicies", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/lambda_code_signing_config.html aws_lambda_code_signing_config}.
 *
 * @stability stable
 */
class DataAwsLambdaCodeSigningConfig extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/lambda_code_signing_config.html aws_lambda_code_signing_config} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_code_signing_config',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // allowed_publishers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    allowedPublishers(index) {
        return new DataAwsLambdaCodeSigningConfigAllowedPublishers(this, 'allowed_publishers', index);
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // config_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get configId() {
        return this.getStringAttribute('config_id');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_modified - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    // policies - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    policies(index) {
        return new DataAwsLambdaCodeSigningConfigPolicies(this, 'policies', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
        };
    }
}
exports.DataAwsLambdaCodeSigningConfig = DataAwsLambdaCodeSigningConfig;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLambdaCodeSigningConfig[_c] = { fqn: "@cdktf/provider-aws.lambdafunction.DataAwsLambdaCodeSigningConfig", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsLambdaCodeSigningConfig.tfResourceType = "aws_lambda_code_signing_config";
