// @cdktf/provider-aws/lib/lambdafunction/lambda-layer-version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaLayerVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lambda_layer_version.html aws_lambda_layer_version}.
 *
 * @stability stable
 */
class LambdaLayerVersion extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lambda_layer_version.html aws_lambda_layer_version} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_layer_version',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._compatibleArchitectures = config.compatibleArchitectures;
        this._compatibleRuntimes = config.compatibleRuntimes;
        this._description = config.description;
        this._filename = config.filename;
        this._layerName = config.layerName;
        this._licenseInfo = config.licenseInfo;
        this._s3Bucket = config.s3Bucket;
        this._s3Key = config.s3Key;
        this._s3ObjectVersion = config.s3ObjectVersion;
        this._skipDestroy = config.skipDestroy;
        this._sourceCodeHash = config.sourceCodeHash;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get compatibleArchitectures() {
        return this.getListAttribute('compatible_architectures');
    }
    /**
     * @stability stable
     */
    set compatibleArchitectures(value) {
        this._compatibleArchitectures = value;
    }
    /**
     * @stability stable
     */
    resetCompatibleArchitectures() {
        this._compatibleArchitectures = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get compatibleArchitecturesInput() {
        return this._compatibleArchitectures;
    }
    /**
     * @stability stable
     */
    get compatibleRuntimes() {
        return this.getListAttribute('compatible_runtimes');
    }
    /**
     * @stability stable
     */
    set compatibleRuntimes(value) {
        this._compatibleRuntimes = value;
    }
    /**
     * @stability stable
     */
    resetCompatibleRuntimes() {
        this._compatibleRuntimes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get compatibleRuntimesInput() {
        return this._compatibleRuntimes;
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get filename() {
        return this.getStringAttribute('filename');
    }
    /**
     * @stability stable
     */
    set filename(value) {
        this._filename = value;
    }
    /**
     * @stability stable
     */
    resetFilename() {
        this._filename = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filenameInput() {
        return this._filename;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // layer_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get layerArn() {
        return this.getStringAttribute('layer_arn');
    }
    /**
     * @stability stable
     */
    get layerName() {
        return this.getStringAttribute('layer_name');
    }
    /**
     * @stability stable
     */
    set layerName(value) {
        this._layerName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get layerNameInput() {
        return this._layerName;
    }
    /**
     * @stability stable
     */
    get licenseInfo() {
        return this.getStringAttribute('license_info');
    }
    /**
     * @stability stable
     */
    set licenseInfo(value) {
        this._licenseInfo = value;
    }
    /**
     * @stability stable
     */
    resetLicenseInfo() {
        this._licenseInfo = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get licenseInfoInput() {
        return this._licenseInfo;
    }
    /**
     * @stability stable
     */
    get s3Bucket() {
        return this.getStringAttribute('s3_bucket');
    }
    /**
     * @stability stable
     */
    set s3Bucket(value) {
        this._s3Bucket = value;
    }
    /**
     * @stability stable
     */
    resetS3Bucket() {
        this._s3Bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketInput() {
        return this._s3Bucket;
    }
    /**
     * @stability stable
     */
    get s3Key() {
        return this.getStringAttribute('s3_key');
    }
    /**
     * @stability stable
     */
    set s3Key(value) {
        this._s3Key = value;
    }
    /**
     * @stability stable
     */
    resetS3Key() {
        this._s3Key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KeyInput() {
        return this._s3Key;
    }
    /**
     * @stability stable
     */
    get s3ObjectVersion() {
        return this.getStringAttribute('s3_object_version');
    }
    /**
     * @stability stable
     */
    set s3ObjectVersion(value) {
        this._s3ObjectVersion = value;
    }
    /**
     * @stability stable
     */
    resetS3ObjectVersion() {
        this._s3ObjectVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3ObjectVersionInput() {
        return this._s3ObjectVersion;
    }
    // signing_job_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get signingJobArn() {
        return this.getStringAttribute('signing_job_arn');
    }
    // signing_profile_version_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get signingProfileVersionArn() {
        return this.getStringAttribute('signing_profile_version_arn');
    }
    /**
     * @stability stable
     */
    get skipDestroy() {
        return this.getBooleanAttribute('skip_destroy');
    }
    /**
     * @stability stable
     */
    set skipDestroy(value) {
        this._skipDestroy = value;
    }
    /**
     * @stability stable
     */
    resetSkipDestroy() {
        this._skipDestroy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skipDestroyInput() {
        return this._skipDestroy;
    }
    /**
     * @stability stable
     */
    get sourceCodeHash() {
        return this.getStringAttribute('source_code_hash');
    }
    /**
     * @stability stable
     */
    set sourceCodeHash(value) {
        this._sourceCodeHash = value;
    }
    /**
     * @stability stable
     */
    resetSourceCodeHash() {
        this._sourceCodeHash = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceCodeHashInput() {
        return this._sourceCodeHash;
    }
    // source_code_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceCodeSize() {
        return this.getNumberAttribute('source_code_size');
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            compatible_architectures: cdktf.listMapper(cdktf.stringToTerraform)(this._compatibleArchitectures),
            compatible_runtimes: cdktf.listMapper(cdktf.stringToTerraform)(this._compatibleRuntimes),
            description: cdktf.stringToTerraform(this._description),
            filename: cdktf.stringToTerraform(this._filename),
            layer_name: cdktf.stringToTerraform(this._layerName),
            license_info: cdktf.stringToTerraform(this._licenseInfo),
            s3_bucket: cdktf.stringToTerraform(this._s3Bucket),
            s3_key: cdktf.stringToTerraform(this._s3Key),
            s3_object_version: cdktf.stringToTerraform(this._s3ObjectVersion),
            skip_destroy: cdktf.booleanToTerraform(this._skipDestroy),
            source_code_hash: cdktf.stringToTerraform(this._sourceCodeHash),
        };
    }
}
exports.LambdaLayerVersion = LambdaLayerVersion;
_a = JSII_RTTI_SYMBOL_1;
LambdaLayerVersion[_a] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaLayerVersion", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LambdaLayerVersion.tfResourceType = "aws_lambda_layer_version";
