// @cdktf/provider-aws/lib/shield/shield-protection-group.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShieldProtectionGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/shield_protection_group.html aws_shield_protection_group}.
 *
 * @stability stable
 */
class ShieldProtectionGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/shield_protection_group.html aws_shield_protection_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_shield_protection_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._aggregation = config.aggregation;
        this._members = config.members;
        this._pattern = config.pattern;
        this._protectionGroupId = config.protectionGroupId;
        this._resourceType = config.resourceType;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    /**
     * @stability stable
     */
    get aggregation() {
        return this.getStringAttribute('aggregation');
    }
    /**
     * @stability stable
     */
    set aggregation(value) {
        this._aggregation = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get aggregationInput() {
        return this._aggregation;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get members() {
        return this.getListAttribute('members');
    }
    /**
     * @stability stable
     */
    set members(value) {
        this._members = value;
    }
    /**
     * @stability stable
     */
    resetMembers() {
        this._members = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get membersInput() {
        return this._members;
    }
    /**
     * @stability stable
     */
    get pattern() {
        return this.getStringAttribute('pattern');
    }
    /**
     * @stability stable
     */
    set pattern(value) {
        this._pattern = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get patternInput() {
        return this._pattern;
    }
    // protection_group_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get protectionGroupArn() {
        return this.getStringAttribute('protection_group_arn');
    }
    /**
     * @stability stable
     */
    get protectionGroupId() {
        return this.getStringAttribute('protection_group_id');
    }
    /**
     * @stability stable
     */
    set protectionGroupId(value) {
        this._protectionGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get protectionGroupIdInput() {
        return this._protectionGroupId;
    }
    /**
     * @stability stable
     */
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    /**
     * @stability stable
     */
    set resourceType(value) {
        this._resourceType = value;
    }
    /**
     * @stability stable
     */
    resetResourceType() {
        this._resourceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTypeInput() {
        return this._resourceType;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            aggregation: cdktf.stringToTerraform(this._aggregation),
            members: cdktf.listMapper(cdktf.stringToTerraform)(this._members),
            pattern: cdktf.stringToTerraform(this._pattern),
            protection_group_id: cdktf.stringToTerraform(this._protectionGroupId),
            resource_type: cdktf.stringToTerraform(this._resourceType),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.ShieldProtectionGroup = ShieldProtectionGroup;
_a = JSII_RTTI_SYMBOL_1;
ShieldProtectionGroup[_a] = { fqn: "@cdktf/provider-aws.shield.ShieldProtectionGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ShieldProtectionGroup.tfResourceType = "aws_shield_protection_group";
