// @cdktf/provider-aws/lib/eks/data-aws-eks-addon.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEksAddon = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/eks_addon.html aws_eks_addon}.
 *
 * @stability stable
 */
class DataAwsEksAddon extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/eks_addon.html aws_eks_addon} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_addon',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._addonName = config.addonName;
        this._clusterName = config.clusterName;
        this._tags = config.tags;
    }
    /**
     * @stability stable
     */
    get addonName() {
        return this.getStringAttribute('addon_name');
    }
    /**
     * @stability stable
     */
    set addonName(value) {
        this._addonName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get addonNameInput() {
        return this._addonName;
    }
    // addon_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get addonVersion() {
        return this.getStringAttribute('addon_version');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // modified_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get modifiedAt() {
        return this.getStringAttribute('modified_at');
    }
    // service_account_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceAccountRoleArn() {
        return this.getStringAttribute('service_account_role_arn');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            addon_name: cdktf.stringToTerraform(this._addonName),
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEksAddon = DataAwsEksAddon;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEksAddon[_a] = { fqn: "@cdktf/provider-aws.eks.DataAwsEksAddon", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEksAddon.tfResourceType = "aws_eks_addon";
