// @cdktf/provider-aws/lib/eks/eks-identity-provider-config.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksIdentityProviderConfig = exports.EksIdentityProviderConfigTimeoutsOutputReference = exports.eksIdentityProviderConfigTimeoutsToTerraform = exports.EksIdentityProviderConfigOidcOutputReference = exports.eksIdentityProviderConfigOidcToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function eksIdentityProviderConfigOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_id: cdktf.stringToTerraform(struct.clientId),
        groups_claim: cdktf.stringToTerraform(struct.groupsClaim),
        groups_prefix: cdktf.stringToTerraform(struct.groupsPrefix),
        identity_provider_config_name: cdktf.stringToTerraform(struct.identityProviderConfigName),
        issuer_url: cdktf.stringToTerraform(struct.issuerUrl),
        required_claims: cdktf.hashMapper(cdktf.anyToTerraform)(struct.requiredClaims),
        username_claim: cdktf.stringToTerraform(struct.usernameClaim),
        username_prefix: cdktf.stringToTerraform(struct.usernamePrefix),
    };
}
exports.eksIdentityProviderConfigOidcToTerraform = eksIdentityProviderConfigOidcToTerraform;
/**
 * @stability stable
 */
class EksIdentityProviderConfigOidcOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientId) {
            hasAnyValues = true;
            internalValueResult.clientId = this._clientId;
        }
        if (this._groupsClaim) {
            hasAnyValues = true;
            internalValueResult.groupsClaim = this._groupsClaim;
        }
        if (this._groupsPrefix) {
            hasAnyValues = true;
            internalValueResult.groupsPrefix = this._groupsPrefix;
        }
        if (this._identityProviderConfigName) {
            hasAnyValues = true;
            internalValueResult.identityProviderConfigName = this._identityProviderConfigName;
        }
        if (this._issuerUrl) {
            hasAnyValues = true;
            internalValueResult.issuerUrl = this._issuerUrl;
        }
        if (this._requiredClaims) {
            hasAnyValues = true;
            internalValueResult.requiredClaims = this._requiredClaims;
        }
        if (this._usernameClaim) {
            hasAnyValues = true;
            internalValueResult.usernameClaim = this._usernameClaim;
        }
        if (this._usernamePrefix) {
            hasAnyValues = true;
            internalValueResult.usernamePrefix = this._usernamePrefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientId = undefined;
            this._groupsClaim = undefined;
            this._groupsPrefix = undefined;
            this._identityProviderConfigName = undefined;
            this._issuerUrl = undefined;
            this._requiredClaims = undefined;
            this._usernameClaim = undefined;
            this._usernamePrefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientId = value.clientId;
            this._groupsClaim = value.groupsClaim;
            this._groupsPrefix = value.groupsPrefix;
            this._identityProviderConfigName = value.identityProviderConfigName;
            this._issuerUrl = value.issuerUrl;
            this._requiredClaims = value.requiredClaims;
            this._usernameClaim = value.usernameClaim;
            this._usernamePrefix = value.usernamePrefix;
        }
    }
    /**
     * @stability stable
     */
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    /**
     * @stability stable
     */
    set clientId(value) {
        this._clientId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientIdInput() {
        return this._clientId;
    }
    /**
     * @stability stable
     */
    get groupsClaim() {
        return this.getStringAttribute('groups_claim');
    }
    /**
     * @stability stable
     */
    set groupsClaim(value) {
        this._groupsClaim = value;
    }
    /**
     * @stability stable
     */
    resetGroupsClaim() {
        this._groupsClaim = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupsClaimInput() {
        return this._groupsClaim;
    }
    /**
     * @stability stable
     */
    get groupsPrefix() {
        return this.getStringAttribute('groups_prefix');
    }
    /**
     * @stability stable
     */
    set groupsPrefix(value) {
        this._groupsPrefix = value;
    }
    /**
     * @stability stable
     */
    resetGroupsPrefix() {
        this._groupsPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupsPrefixInput() {
        return this._groupsPrefix;
    }
    /**
     * @stability stable
     */
    get identityProviderConfigName() {
        return this.getStringAttribute('identity_provider_config_name');
    }
    /**
     * @stability stable
     */
    set identityProviderConfigName(value) {
        this._identityProviderConfigName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get identityProviderConfigNameInput() {
        return this._identityProviderConfigName;
    }
    /**
     * @stability stable
     */
    get issuerUrl() {
        return this.getStringAttribute('issuer_url');
    }
    /**
     * @stability stable
     */
    set issuerUrl(value) {
        this._issuerUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get issuerUrlInput() {
        return this._issuerUrl;
    }
    /**
     * @stability stable
     */
    get requiredClaims() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('required_claims');
    }
    /**
     * @stability stable
     */
    set requiredClaims(value) {
        this._requiredClaims = value;
    }
    /**
     * @stability stable
     */
    resetRequiredClaims() {
        this._requiredClaims = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requiredClaimsInput() {
        return this._requiredClaims;
    }
    /**
     * @stability stable
     */
    get usernameClaim() {
        return this.getStringAttribute('username_claim');
    }
    /**
     * @stability stable
     */
    set usernameClaim(value) {
        this._usernameClaim = value;
    }
    /**
     * @stability stable
     */
    resetUsernameClaim() {
        this._usernameClaim = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get usernameClaimInput() {
        return this._usernameClaim;
    }
    /**
     * @stability stable
     */
    get usernamePrefix() {
        return this.getStringAttribute('username_prefix');
    }
    /**
     * @stability stable
     */
    set usernamePrefix(value) {
        this._usernamePrefix = value;
    }
    /**
     * @stability stable
     */
    resetUsernamePrefix() {
        this._usernamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get usernamePrefixInput() {
        return this._usernamePrefix;
    }
}
exports.EksIdentityProviderConfigOidcOutputReference = EksIdentityProviderConfigOidcOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EksIdentityProviderConfigOidcOutputReference[_a] = { fqn: "@cdktf/provider-aws.eks.EksIdentityProviderConfigOidcOutputReference", version: "3.0.1" };
function eksIdentityProviderConfigTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.eksIdentityProviderConfigTimeoutsToTerraform = eksIdentityProviderConfigTimeoutsToTerraform;
/**
 * @stability stable
 */
class EksIdentityProviderConfigTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.EksIdentityProviderConfigTimeoutsOutputReference = EksIdentityProviderConfigTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EksIdentityProviderConfigTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.eks.EksIdentityProviderConfigTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/eks_identity_provider_config.html aws_eks_identity_provider_config}.
 *
 * @stability stable
 */
class EksIdentityProviderConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/eks_identity_provider_config.html aws_eks_identity_provider_config} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_identity_provider_config',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // oidc - computed: false, optional: false, required: true
        this._oidc = new EksIdentityProviderConfigOidcOutputReference(this, "oidc", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksIdentityProviderConfigTimeoutsOutputReference(this, "timeouts", true);
        this._clusterName = config.clusterName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._oidc.internalValue = config.oidc;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get oidc() {
        return this._oidc;
    }
    /**
     * @stability stable
     */
    putOidc(value) {
        this._oidc.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get oidcInput() {
        return this._oidc.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            oidc: eksIdentityProviderConfigOidcToTerraform(this._oidc.internalValue),
            timeouts: eksIdentityProviderConfigTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.EksIdentityProviderConfig = EksIdentityProviderConfig;
_c = JSII_RTTI_SYMBOL_1;
EksIdentityProviderConfig[_c] = { fqn: "@cdktf/provider-aws.eks.EksIdentityProviderConfig", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EksIdentityProviderConfig.tfResourceType = "aws_eks_identity_provider_config";
