// @cdktf/provider-aws/lib/eks/eks-node-group.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksNodeGroup = exports.EksNodeGroupUpdateConfigOutputReference = exports.eksNodeGroupUpdateConfigToTerraform = exports.EksNodeGroupTimeoutsOutputReference = exports.eksNodeGroupTimeoutsToTerraform = exports.eksNodeGroupTaintToTerraform = exports.EksNodeGroupScalingConfigOutputReference = exports.eksNodeGroupScalingConfigToTerraform = exports.EksNodeGroupRemoteAccessOutputReference = exports.eksNodeGroupRemoteAccessToTerraform = exports.EksNodeGroupLaunchTemplateOutputReference = exports.eksNodeGroupLaunchTemplateToTerraform = exports.EksNodeGroupResources = exports.EksNodeGroupResourcesAutoscalingGroups = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class EksNodeGroupResourcesAutoscalingGroups extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.EksNodeGroupResourcesAutoscalingGroups = EksNodeGroupResourcesAutoscalingGroups;
_a = JSII_RTTI_SYMBOL_1;
EksNodeGroupResourcesAutoscalingGroups[_a] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupResourcesAutoscalingGroups", version: "3.0.1" };
/**
 * @stability stable
 */
class EksNodeGroupResources extends cdktf.ComplexComputedList {
    // autoscaling_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get autoscalingGroups() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('autoscaling_groups');
    }
    // remote_access_security_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get remoteAccessSecurityGroupId() {
        return this.getStringAttribute('remote_access_security_group_id');
    }
}
exports.EksNodeGroupResources = EksNodeGroupResources;
_b = JSII_RTTI_SYMBOL_1;
EksNodeGroupResources[_b] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupResources", version: "3.0.1" };
function eksNodeGroupLaunchTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.eksNodeGroupLaunchTemplateToTerraform = eksNodeGroupLaunchTemplateToTerraform;
/**
 * @stability stable
 */
class EksNodeGroupLaunchTemplateOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._version) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._version = value.version;
        }
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
}
exports.EksNodeGroupLaunchTemplateOutputReference = EksNodeGroupLaunchTemplateOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EksNodeGroupLaunchTemplateOutputReference[_c] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupLaunchTemplateOutputReference", version: "3.0.1" };
function eksNodeGroupRemoteAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ec2_ssh_key: cdktf.stringToTerraform(struct.ec2SshKey),
        source_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.sourceSecurityGroupIds),
    };
}
exports.eksNodeGroupRemoteAccessToTerraform = eksNodeGroupRemoteAccessToTerraform;
/**
 * @stability stable
 */
class EksNodeGroupRemoteAccessOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ec2SshKey) {
            hasAnyValues = true;
            internalValueResult.ec2SshKey = this._ec2SshKey;
        }
        if (this._sourceSecurityGroupIds) {
            hasAnyValues = true;
            internalValueResult.sourceSecurityGroupIds = this._sourceSecurityGroupIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ec2SshKey = undefined;
            this._sourceSecurityGroupIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ec2SshKey = value.ec2SshKey;
            this._sourceSecurityGroupIds = value.sourceSecurityGroupIds;
        }
    }
    /**
     * @stability stable
     */
    get ec2SshKey() {
        return this.getStringAttribute('ec2_ssh_key');
    }
    /**
     * @stability stable
     */
    set ec2SshKey(value) {
        this._ec2SshKey = value;
    }
    /**
     * @stability stable
     */
    resetEc2SshKey() {
        this._ec2SshKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ec2SshKeyInput() {
        return this._ec2SshKey;
    }
    /**
     * @stability stable
     */
    get sourceSecurityGroupIds() {
        return this.getListAttribute('source_security_group_ids');
    }
    /**
     * @stability stable
     */
    set sourceSecurityGroupIds(value) {
        this._sourceSecurityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSourceSecurityGroupIds() {
        this._sourceSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceSecurityGroupIdsInput() {
        return this._sourceSecurityGroupIds;
    }
}
exports.EksNodeGroupRemoteAccessOutputReference = EksNodeGroupRemoteAccessOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EksNodeGroupRemoteAccessOutputReference[_d] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupRemoteAccessOutputReference", version: "3.0.1" };
function eksNodeGroupScalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        desired_size: cdktf.numberToTerraform(struct.desiredSize),
        max_size: cdktf.numberToTerraform(struct.maxSize),
        min_size: cdktf.numberToTerraform(struct.minSize),
    };
}
exports.eksNodeGroupScalingConfigToTerraform = eksNodeGroupScalingConfigToTerraform;
/**
 * @stability stable
 */
class EksNodeGroupScalingConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._desiredSize) {
            hasAnyValues = true;
            internalValueResult.desiredSize = this._desiredSize;
        }
        if (this._maxSize) {
            hasAnyValues = true;
            internalValueResult.maxSize = this._maxSize;
        }
        if (this._minSize) {
            hasAnyValues = true;
            internalValueResult.minSize = this._minSize;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._desiredSize = undefined;
            this._maxSize = undefined;
            this._minSize = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._desiredSize = value.desiredSize;
            this._maxSize = value.maxSize;
            this._minSize = value.minSize;
        }
    }
    /**
     * @stability stable
     */
    get desiredSize() {
        return this.getNumberAttribute('desired_size');
    }
    /**
     * @stability stable
     */
    set desiredSize(value) {
        this._desiredSize = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get desiredSizeInput() {
        return this._desiredSize;
    }
    /**
     * @stability stable
     */
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    /**
     * @stability stable
     */
    set maxSize(value) {
        this._maxSize = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxSizeInput() {
        return this._maxSize;
    }
    /**
     * @stability stable
     */
    get minSize() {
        return this.getNumberAttribute('min_size');
    }
    /**
     * @stability stable
     */
    set minSize(value) {
        this._minSize = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minSizeInput() {
        return this._minSize;
    }
}
exports.EksNodeGroupScalingConfigOutputReference = EksNodeGroupScalingConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EksNodeGroupScalingConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupScalingConfigOutputReference", version: "3.0.1" };
function eksNodeGroupTaintToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        effect: cdktf.stringToTerraform(struct.effect),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.eksNodeGroupTaintToTerraform = eksNodeGroupTaintToTerraform;
function eksNodeGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.eksNodeGroupTimeoutsToTerraform = eksNodeGroupTimeoutsToTerraform;
/**
 * @stability stable
 */
class EksNodeGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.EksNodeGroupTimeoutsOutputReference = EksNodeGroupTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EksNodeGroupTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupTimeoutsOutputReference", version: "3.0.1" };
function eksNodeGroupUpdateConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_unavailable: cdktf.numberToTerraform(struct.maxUnavailable),
        max_unavailable_percentage: cdktf.numberToTerraform(struct.maxUnavailablePercentage),
    };
}
exports.eksNodeGroupUpdateConfigToTerraform = eksNodeGroupUpdateConfigToTerraform;
/**
 * @stability stable
 */
class EksNodeGroupUpdateConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxUnavailable) {
            hasAnyValues = true;
            internalValueResult.maxUnavailable = this._maxUnavailable;
        }
        if (this._maxUnavailablePercentage) {
            hasAnyValues = true;
            internalValueResult.maxUnavailablePercentage = this._maxUnavailablePercentage;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxUnavailable = undefined;
            this._maxUnavailablePercentage = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxUnavailable = value.maxUnavailable;
            this._maxUnavailablePercentage = value.maxUnavailablePercentage;
        }
    }
    /**
     * @stability stable
     */
    get maxUnavailable() {
        return this.getNumberAttribute('max_unavailable');
    }
    /**
     * @stability stable
     */
    set maxUnavailable(value) {
        this._maxUnavailable = value;
    }
    /**
     * @stability stable
     */
    resetMaxUnavailable() {
        this._maxUnavailable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxUnavailableInput() {
        return this._maxUnavailable;
    }
    /**
     * @stability stable
     */
    get maxUnavailablePercentage() {
        return this.getNumberAttribute('max_unavailable_percentage');
    }
    /**
     * @stability stable
     */
    set maxUnavailablePercentage(value) {
        this._maxUnavailablePercentage = value;
    }
    /**
     * @stability stable
     */
    resetMaxUnavailablePercentage() {
        this._maxUnavailablePercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxUnavailablePercentageInput() {
        return this._maxUnavailablePercentage;
    }
}
exports.EksNodeGroupUpdateConfigOutputReference = EksNodeGroupUpdateConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EksNodeGroupUpdateConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroupUpdateConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/eks_node_group.html aws_eks_node_group}.
 *
 * @stability stable
 */
class EksNodeGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/eks_node_group.html aws_eks_node_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_node_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // launch_template - computed: false, optional: true, required: false
        this._launchTemplate = new EksNodeGroupLaunchTemplateOutputReference(this, "launch_template", true);
        // remote_access - computed: false, optional: true, required: false
        this._remoteAccess = new EksNodeGroupRemoteAccessOutputReference(this, "remote_access", true);
        // scaling_config - computed: false, optional: false, required: true
        this._scalingConfig = new EksNodeGroupScalingConfigOutputReference(this, "scaling_config", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksNodeGroupTimeoutsOutputReference(this, "timeouts", true);
        // update_config - computed: false, optional: true, required: false
        this._updateConfig = new EksNodeGroupUpdateConfigOutputReference(this, "update_config", true);
        this._amiType = config.amiType;
        this._capacityType = config.capacityType;
        this._clusterName = config.clusterName;
        this._diskSize = config.diskSize;
        this._forceUpdateVersion = config.forceUpdateVersion;
        this._instanceTypes = config.instanceTypes;
        this._labels = config.labels;
        this._nodeGroupName = config.nodeGroupName;
        this._nodeGroupNamePrefix = config.nodeGroupNamePrefix;
        this._nodeRoleArn = config.nodeRoleArn;
        this._releaseVersion = config.releaseVersion;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._version = config.version;
        this._launchTemplate.internalValue = config.launchTemplate;
        this._remoteAccess.internalValue = config.remoteAccess;
        this._scalingConfig.internalValue = config.scalingConfig;
        this._taint = config.taint;
        this._timeouts.internalValue = config.timeouts;
        this._updateConfig.internalValue = config.updateConfig;
    }
    /**
     * @stability stable
     */
    get amiType() {
        return this.getStringAttribute('ami_type');
    }
    /**
     * @stability stable
     */
    set amiType(value) {
        this._amiType = value;
    }
    /**
     * @stability stable
     */
    resetAmiType() {
        this._amiType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get amiTypeInput() {
        return this._amiType;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get capacityType() {
        return this.getStringAttribute('capacity_type');
    }
    /**
     * @stability stable
     */
    set capacityType(value) {
        this._capacityType = value;
    }
    /**
     * @stability stable
     */
    resetCapacityType() {
        this._capacityType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityTypeInput() {
        return this._capacityType;
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    /**
     * @stability stable
     */
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    /**
     * @stability stable
     */
    set diskSize(value) {
        this._diskSize = value;
    }
    /**
     * @stability stable
     */
    resetDiskSize() {
        this._diskSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get diskSizeInput() {
        return this._diskSize;
    }
    /**
     * @stability stable
     */
    get forceUpdateVersion() {
        return this.getBooleanAttribute('force_update_version');
    }
    /**
     * @stability stable
     */
    set forceUpdateVersion(value) {
        this._forceUpdateVersion = value;
    }
    /**
     * @stability stable
     */
    resetForceUpdateVersion() {
        this._forceUpdateVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forceUpdateVersionInput() {
        return this._forceUpdateVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceTypes() {
        return this.getListAttribute('instance_types');
    }
    /**
     * @stability stable
     */
    set instanceTypes(value) {
        this._instanceTypes = value;
    }
    /**
     * @stability stable
     */
    resetInstanceTypes() {
        this._instanceTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypesInput() {
        return this._instanceTypes;
    }
    /**
     * @stability stable
     */
    get labels() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('labels');
    }
    /**
     * @stability stable
     */
    set labels(value) {
        this._labels = value;
    }
    /**
     * @stability stable
     */
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get labelsInput() {
        return this._labels;
    }
    /**
     * @stability stable
     */
    get nodeGroupName() {
        return this.getStringAttribute('node_group_name');
    }
    /**
     * @stability stable
     */
    set nodeGroupName(value) {
        this._nodeGroupName = value;
    }
    /**
     * @stability stable
     */
    resetNodeGroupName() {
        this._nodeGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeGroupNameInput() {
        return this._nodeGroupName;
    }
    /**
     * @stability stable
     */
    get nodeGroupNamePrefix() {
        return this.getStringAttribute('node_group_name_prefix');
    }
    /**
     * @stability stable
     */
    set nodeGroupNamePrefix(value) {
        this._nodeGroupNamePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNodeGroupNamePrefix() {
        this._nodeGroupNamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeGroupNamePrefixInput() {
        return this._nodeGroupNamePrefix;
    }
    /**
     * @stability stable
     */
    get nodeRoleArn() {
        return this.getStringAttribute('node_role_arn');
    }
    /**
     * @stability stable
     */
    set nodeRoleArn(value) {
        this._nodeRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeRoleArnInput() {
        return this._nodeRoleArn;
    }
    /**
     * @stability stable
     */
    get releaseVersion() {
        return this.getStringAttribute('release_version');
    }
    /**
     * @stability stable
     */
    set releaseVersion(value) {
        this._releaseVersion = value;
    }
    /**
     * @stability stable
     */
    resetReleaseVersion() {
        this._releaseVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get releaseVersionInput() {
        return this._releaseVersion;
    }
    // resources - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    resources(index) {
        return new EksNodeGroupResources(this, 'resources', index);
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    /**
     * @stability stable
     */
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
    /**
     * @stability stable
     */
    get launchTemplate() {
        return this._launchTemplate;
    }
    /**
     * @stability stable
     */
    putLaunchTemplate(value) {
        this._launchTemplate.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLaunchTemplate() {
        this._launchTemplate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTemplateInput() {
        return this._launchTemplate.internalValue;
    }
    /**
     * @stability stable
     */
    get remoteAccess() {
        return this._remoteAccess;
    }
    /**
     * @stability stable
     */
    putRemoteAccess(value) {
        this._remoteAccess.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRemoteAccess() {
        this._remoteAccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get remoteAccessInput() {
        return this._remoteAccess.internalValue;
    }
    /**
     * @stability stable
     */
    get scalingConfig() {
        return this._scalingConfig;
    }
    /**
     * @stability stable
     */
    putScalingConfig(value) {
        this._scalingConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scalingConfigInput() {
        return this._scalingConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get taint() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('taint');
    }
    /**
     * @stability stable
     */
    set taint(value) {
        this._taint = value;
    }
    /**
     * @stability stable
     */
    resetTaint() {
        this._taint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taintInput() {
        return this._taint;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    /**
     * @stability stable
     */
    get updateConfig() {
        return this._updateConfig;
    }
    /**
     * @stability stable
     */
    putUpdateConfig(value) {
        this._updateConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetUpdateConfig() {
        this._updateConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateConfigInput() {
        return this._updateConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            ami_type: cdktf.stringToTerraform(this._amiType),
            capacity_type: cdktf.stringToTerraform(this._capacityType),
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            disk_size: cdktf.numberToTerraform(this._diskSize),
            force_update_version: cdktf.booleanToTerraform(this._forceUpdateVersion),
            instance_types: cdktf.listMapper(cdktf.stringToTerraform)(this._instanceTypes),
            labels: cdktf.hashMapper(cdktf.anyToTerraform)(this._labels),
            node_group_name: cdktf.stringToTerraform(this._nodeGroupName),
            node_group_name_prefix: cdktf.stringToTerraform(this._nodeGroupNamePrefix),
            node_role_arn: cdktf.stringToTerraform(this._nodeRoleArn),
            release_version: cdktf.stringToTerraform(this._releaseVersion),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            version: cdktf.stringToTerraform(this._version),
            launch_template: eksNodeGroupLaunchTemplateToTerraform(this._launchTemplate.internalValue),
            remote_access: eksNodeGroupRemoteAccessToTerraform(this._remoteAccess.internalValue),
            scaling_config: eksNodeGroupScalingConfigToTerraform(this._scalingConfig.internalValue),
            taint: cdktf.listMapper(eksNodeGroupTaintToTerraform)(this._taint),
            timeouts: eksNodeGroupTimeoutsToTerraform(this._timeouts.internalValue),
            update_config: eksNodeGroupUpdateConfigToTerraform(this._updateConfig.internalValue),
        };
    }
}
exports.EksNodeGroup = EksNodeGroup;
_h = JSII_RTTI_SYMBOL_1;
EksNodeGroup[_h] = { fqn: "@cdktf/provider-aws.eks.EksNodeGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EksNodeGroup.tfResourceType = "aws_eks_node_group";
