// @cdktf/provider-aws/lib/mwaa/mwaa-environment.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MwaaEnvironment = exports.MwaaEnvironmentNetworkConfigurationOutputReference = exports.mwaaEnvironmentNetworkConfigurationToTerraform = exports.MwaaEnvironmentLoggingConfigurationOutputReference = exports.mwaaEnvironmentLoggingConfigurationToTerraform = exports.MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference = exports.mwaaEnvironmentLoggingConfigurationWorkerLogsToTerraform = exports.MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference = exports.mwaaEnvironmentLoggingConfigurationWebserverLogsToTerraform = exports.MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference = exports.mwaaEnvironmentLoggingConfigurationTaskLogsToTerraform = exports.MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference = exports.mwaaEnvironmentLoggingConfigurationSchedulerLogsToTerraform = exports.MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference = exports.mwaaEnvironmentLoggingConfigurationDagProcessingLogsToTerraform = exports.MwaaEnvironmentLastUpdated = exports.MwaaEnvironmentLastUpdatedError = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class MwaaEnvironmentLastUpdatedError extends cdktf.ComplexComputedList {
    // error_code - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get errorCode() {
        return this.getStringAttribute('error_code');
    }
    // error_message - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get errorMessage() {
        return this.getStringAttribute('error_message');
    }
}
exports.MwaaEnvironmentLastUpdatedError = MwaaEnvironmentLastUpdatedError;
_a = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLastUpdatedError[_a] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLastUpdatedError", version: "3.0.1" };
/**
 * @stability stable
 */
class MwaaEnvironmentLastUpdated extends cdktf.ComplexComputedList {
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // error - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get error() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('error');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.MwaaEnvironmentLastUpdated = MwaaEnvironmentLastUpdated;
_b = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLastUpdated[_b] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLastUpdated", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationDagProcessingLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_level: cdktf.stringToTerraform(struct.logLevel),
    };
}
exports.mwaaEnvironmentLoggingConfigurationDagProcessingLogsToTerraform = mwaaEnvironmentLoggingConfigurationDagProcessingLogsToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logLevel) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logLevel = value.logLevel;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    /**
     * @stability stable
     */
    set logLevel(value) {
        this._logLevel = value;
    }
    /**
     * @stability stable
     */
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logLevelInput() {
        return this._logLevel;
    }
}
exports.MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference = MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference[_c] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationSchedulerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_level: cdktf.stringToTerraform(struct.logLevel),
    };
}
exports.mwaaEnvironmentLoggingConfigurationSchedulerLogsToTerraform = mwaaEnvironmentLoggingConfigurationSchedulerLogsToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logLevel) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logLevel = value.logLevel;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    /**
     * @stability stable
     */
    set logLevel(value) {
        this._logLevel = value;
    }
    /**
     * @stability stable
     */
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logLevelInput() {
        return this._logLevel;
    }
}
exports.MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference = MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference[_d] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationTaskLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_level: cdktf.stringToTerraform(struct.logLevel),
    };
}
exports.mwaaEnvironmentLoggingConfigurationTaskLogsToTerraform = mwaaEnvironmentLoggingConfigurationTaskLogsToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logLevel) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logLevel = value.logLevel;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    /**
     * @stability stable
     */
    set logLevel(value) {
        this._logLevel = value;
    }
    /**
     * @stability stable
     */
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logLevelInput() {
        return this._logLevel;
    }
}
exports.MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference = MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference[_e] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationWebserverLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_level: cdktf.stringToTerraform(struct.logLevel),
    };
}
exports.mwaaEnvironmentLoggingConfigurationWebserverLogsToTerraform = mwaaEnvironmentLoggingConfigurationWebserverLogsToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logLevel) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logLevel = value.logLevel;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    /**
     * @stability stable
     */
    set logLevel(value) {
        this._logLevel = value;
    }
    /**
     * @stability stable
     */
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logLevelInput() {
        return this._logLevel;
    }
}
exports.MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference = MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference[_f] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationWorkerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_level: cdktf.stringToTerraform(struct.logLevel),
    };
}
exports.mwaaEnvironmentLoggingConfigurationWorkerLogsToTerraform = mwaaEnvironmentLoggingConfigurationWorkerLogsToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logLevel) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logLevel = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logLevel = value.logLevel;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    /**
     * @stability stable
     */
    set logLevel(value) {
        this._logLevel = value;
    }
    /**
     * @stability stable
     */
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logLevelInput() {
        return this._logLevel;
    }
}
exports.MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference = MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference[_g] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference", version: "3.0.1" };
function mwaaEnvironmentLoggingConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dag_processing_logs: mwaaEnvironmentLoggingConfigurationDagProcessingLogsToTerraform(struct.dagProcessingLogs),
        scheduler_logs: mwaaEnvironmentLoggingConfigurationSchedulerLogsToTerraform(struct.schedulerLogs),
        task_logs: mwaaEnvironmentLoggingConfigurationTaskLogsToTerraform(struct.taskLogs),
        webserver_logs: mwaaEnvironmentLoggingConfigurationWebserverLogsToTerraform(struct.webserverLogs),
        worker_logs: mwaaEnvironmentLoggingConfigurationWorkerLogsToTerraform(struct.workerLogs),
    };
}
exports.mwaaEnvironmentLoggingConfigurationToTerraform = mwaaEnvironmentLoggingConfigurationToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentLoggingConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // dag_processing_logs - computed: false, optional: true, required: false
        this._dagProcessingLogs = new MwaaEnvironmentLoggingConfigurationDagProcessingLogsOutputReference(this, "dag_processing_logs", true);
        // scheduler_logs - computed: false, optional: true, required: false
        this._schedulerLogs = new MwaaEnvironmentLoggingConfigurationSchedulerLogsOutputReference(this, "scheduler_logs", true);
        // task_logs - computed: false, optional: true, required: false
        this._taskLogs = new MwaaEnvironmentLoggingConfigurationTaskLogsOutputReference(this, "task_logs", true);
        // webserver_logs - computed: false, optional: true, required: false
        this._webserverLogs = new MwaaEnvironmentLoggingConfigurationWebserverLogsOutputReference(this, "webserver_logs", true);
        // worker_logs - computed: false, optional: true, required: false
        this._workerLogs = new MwaaEnvironmentLoggingConfigurationWorkerLogsOutputReference(this, "worker_logs", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_l = this._dagProcessingLogs) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.dagProcessingLogs = (_m = this._dagProcessingLogs) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        if ((_o = this._schedulerLogs) === null || _o === void 0 ? void 0 : _o.internalValue) {
            hasAnyValues = true;
            internalValueResult.schedulerLogs = (_p = this._schedulerLogs) === null || _p === void 0 ? void 0 : _p.internalValue;
        }
        if ((_q = this._taskLogs) === null || _q === void 0 ? void 0 : _q.internalValue) {
            hasAnyValues = true;
            internalValueResult.taskLogs = (_r = this._taskLogs) === null || _r === void 0 ? void 0 : _r.internalValue;
        }
        if ((_s = this._webserverLogs) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.webserverLogs = (_t = this._webserverLogs) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        if ((_u = this._workerLogs) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.workerLogs = (_v = this._workerLogs) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dagProcessingLogs.internalValue = undefined;
            this._schedulerLogs.internalValue = undefined;
            this._taskLogs.internalValue = undefined;
            this._webserverLogs.internalValue = undefined;
            this._workerLogs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dagProcessingLogs.internalValue = value.dagProcessingLogs;
            this._schedulerLogs.internalValue = value.schedulerLogs;
            this._taskLogs.internalValue = value.taskLogs;
            this._webserverLogs.internalValue = value.webserverLogs;
            this._workerLogs.internalValue = value.workerLogs;
        }
    }
    /**
     * @stability stable
     */
    get dagProcessingLogs() {
        return this._dagProcessingLogs;
    }
    /**
     * @stability stable
     */
    putDagProcessingLogs(value) {
        this._dagProcessingLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDagProcessingLogs() {
        this._dagProcessingLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dagProcessingLogsInput() {
        return this._dagProcessingLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get schedulerLogs() {
        return this._schedulerLogs;
    }
    /**
     * @stability stable
     */
    putSchedulerLogs(value) {
        this._schedulerLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSchedulerLogs() {
        this._schedulerLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schedulerLogsInput() {
        return this._schedulerLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get taskLogs() {
        return this._taskLogs;
    }
    /**
     * @stability stable
     */
    putTaskLogs(value) {
        this._taskLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTaskLogs() {
        this._taskLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskLogsInput() {
        return this._taskLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get webserverLogs() {
        return this._webserverLogs;
    }
    /**
     * @stability stable
     */
    putWebserverLogs(value) {
        this._webserverLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetWebserverLogs() {
        this._webserverLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get webserverLogsInput() {
        return this._webserverLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get workerLogs() {
        return this._workerLogs;
    }
    /**
     * @stability stable
     */
    putWorkerLogs(value) {
        this._workerLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetWorkerLogs() {
        this._workerLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workerLogsInput() {
        return this._workerLogs.internalValue;
    }
}
exports.MwaaEnvironmentLoggingConfigurationOutputReference = MwaaEnvironmentLoggingConfigurationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentLoggingConfigurationOutputReference[_h] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentLoggingConfigurationOutputReference", version: "3.0.1" };
function mwaaEnvironmentNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnetIds),
    };
}
exports.mwaaEnvironmentNetworkConfigurationToTerraform = mwaaEnvironmentNetworkConfigurationToTerraform;
/**
 * @stability stable
 */
class MwaaEnvironmentNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.MwaaEnvironmentNetworkConfigurationOutputReference = MwaaEnvironmentNetworkConfigurationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MwaaEnvironmentNetworkConfigurationOutputReference[_j] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironmentNetworkConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/mwaa_environment.html aws_mwaa_environment}.
 *
 * @stability stable
 */
class MwaaEnvironment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/mwaa_environment.html aws_mwaa_environment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_mwaa_environment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // logging_configuration - computed: false, optional: true, required: false
        this._loggingConfiguration = new MwaaEnvironmentLoggingConfigurationOutputReference(this, "logging_configuration", true);
        // network_configuration - computed: false, optional: false, required: true
        this._networkConfiguration = new MwaaEnvironmentNetworkConfigurationOutputReference(this, "network_configuration", true);
        this._airflowConfigurationOptions = config.airflowConfigurationOptions;
        this._airflowVersion = config.airflowVersion;
        this._dagS3Path = config.dagS3Path;
        this._environmentClass = config.environmentClass;
        this._executionRoleArn = config.executionRoleArn;
        this._kmsKey = config.kmsKey;
        this._maxWorkers = config.maxWorkers;
        this._minWorkers = config.minWorkers;
        this._name = config.name;
        this._pluginsS3ObjectVersion = config.pluginsS3ObjectVersion;
        this._pluginsS3Path = config.pluginsS3Path;
        this._requirementsS3ObjectVersion = config.requirementsS3ObjectVersion;
        this._requirementsS3Path = config.requirementsS3Path;
        this._sourceBucketArn = config.sourceBucketArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._webserverAccessMode = config.webserverAccessMode;
        this._weeklyMaintenanceWindowStart = config.weeklyMaintenanceWindowStart;
        this._loggingConfiguration.internalValue = config.loggingConfiguration;
        this._networkConfiguration.internalValue = config.networkConfiguration;
    }
    /**
     * @stability stable
     */
    get airflowConfigurationOptions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('airflow_configuration_options');
    }
    /**
     * @stability stable
     */
    set airflowConfigurationOptions(value) {
        this._airflowConfigurationOptions = value;
    }
    /**
     * @stability stable
     */
    resetAirflowConfigurationOptions() {
        this._airflowConfigurationOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get airflowConfigurationOptionsInput() {
        return this._airflowConfigurationOptions;
    }
    /**
     * @stability stable
     */
    get airflowVersion() {
        return this.getStringAttribute('airflow_version');
    }
    /**
     * @stability stable
     */
    set airflowVersion(value) {
        this._airflowVersion = value;
    }
    /**
     * @stability stable
     */
    resetAirflowVersion() {
        this._airflowVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get airflowVersionInput() {
        return this._airflowVersion;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    /**
     * @stability stable
     */
    get dagS3Path() {
        return this.getStringAttribute('dag_s3_path');
    }
    /**
     * @stability stable
     */
    set dagS3Path(value) {
        this._dagS3Path = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dagS3PathInput() {
        return this._dagS3Path;
    }
    /**
     * @stability stable
     */
    get environmentClass() {
        return this.getStringAttribute('environment_class');
    }
    /**
     * @stability stable
     */
    set environmentClass(value) {
        this._environmentClass = value;
    }
    /**
     * @stability stable
     */
    resetEnvironmentClass() {
        this._environmentClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentClassInput() {
        return this._environmentClass;
    }
    /**
     * @stability stable
     */
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    /**
     * @stability stable
     */
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    /**
     * @stability stable
     */
    set kmsKey(value) {
        this._kmsKey = value;
    }
    /**
     * @stability stable
     */
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyInput() {
        return this._kmsKey;
    }
    // last_updated - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    lastUpdated(index) {
        return new MwaaEnvironmentLastUpdated(this, 'last_updated', index);
    }
    /**
     * @stability stable
     */
    get maxWorkers() {
        return this.getNumberAttribute('max_workers');
    }
    /**
     * @stability stable
     */
    set maxWorkers(value) {
        this._maxWorkers = value;
    }
    /**
     * @stability stable
     */
    resetMaxWorkers() {
        this._maxWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxWorkersInput() {
        return this._maxWorkers;
    }
    /**
     * @stability stable
     */
    get minWorkers() {
        return this.getNumberAttribute('min_workers');
    }
    /**
     * @stability stable
     */
    set minWorkers(value) {
        this._minWorkers = value;
    }
    /**
     * @stability stable
     */
    resetMinWorkers() {
        this._minWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minWorkersInput() {
        return this._minWorkers;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get pluginsS3ObjectVersion() {
        return this.getStringAttribute('plugins_s3_object_version');
    }
    /**
     * @stability stable
     */
    set pluginsS3ObjectVersion(value) {
        this._pluginsS3ObjectVersion = value;
    }
    /**
     * @stability stable
     */
    resetPluginsS3ObjectVersion() {
        this._pluginsS3ObjectVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pluginsS3ObjectVersionInput() {
        return this._pluginsS3ObjectVersion;
    }
    /**
     * @stability stable
     */
    get pluginsS3Path() {
        return this.getStringAttribute('plugins_s3_path');
    }
    /**
     * @stability stable
     */
    set pluginsS3Path(value) {
        this._pluginsS3Path = value;
    }
    /**
     * @stability stable
     */
    resetPluginsS3Path() {
        this._pluginsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pluginsS3PathInput() {
        return this._pluginsS3Path;
    }
    /**
     * @stability stable
     */
    get requirementsS3ObjectVersion() {
        return this.getStringAttribute('requirements_s3_object_version');
    }
    /**
     * @stability stable
     */
    set requirementsS3ObjectVersion(value) {
        this._requirementsS3ObjectVersion = value;
    }
    /**
     * @stability stable
     */
    resetRequirementsS3ObjectVersion() {
        this._requirementsS3ObjectVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requirementsS3ObjectVersionInput() {
        return this._requirementsS3ObjectVersion;
    }
    /**
     * @stability stable
     */
    get requirementsS3Path() {
        return this.getStringAttribute('requirements_s3_path');
    }
    /**
     * @stability stable
     */
    set requirementsS3Path(value) {
        this._requirementsS3Path = value;
    }
    /**
     * @stability stable
     */
    resetRequirementsS3Path() {
        this._requirementsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requirementsS3PathInput() {
        return this._requirementsS3Path;
    }
    // service_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceRoleArn() {
        return this.getStringAttribute('service_role_arn');
    }
    /**
     * @stability stable
     */
    get sourceBucketArn() {
        return this.getStringAttribute('source_bucket_arn');
    }
    /**
     * @stability stable
     */
    set sourceBucketArn(value) {
        this._sourceBucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceBucketArnInput() {
        return this._sourceBucketArn;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get webserverAccessMode() {
        return this.getStringAttribute('webserver_access_mode');
    }
    /**
     * @stability stable
     */
    set webserverAccessMode(value) {
        this._webserverAccessMode = value;
    }
    /**
     * @stability stable
     */
    resetWebserverAccessMode() {
        this._webserverAccessMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get webserverAccessModeInput() {
        return this._webserverAccessMode;
    }
    // webserver_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get webserverUrl() {
        return this.getStringAttribute('webserver_url');
    }
    /**
     * @stability stable
     */
    get weeklyMaintenanceWindowStart() {
        return this.getStringAttribute('weekly_maintenance_window_start');
    }
    /**
     * @stability stable
     */
    set weeklyMaintenanceWindowStart(value) {
        this._weeklyMaintenanceWindowStart = value;
    }
    /**
     * @stability stable
     */
    resetWeeklyMaintenanceWindowStart() {
        this._weeklyMaintenanceWindowStart = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get weeklyMaintenanceWindowStartInput() {
        return this._weeklyMaintenanceWindowStart;
    }
    /**
     * @stability stable
     */
    get loggingConfiguration() {
        return this._loggingConfiguration;
    }
    /**
     * @stability stable
     */
    putLoggingConfiguration(value) {
        this._loggingConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLoggingConfiguration() {
        this._loggingConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingConfigurationInput() {
        return this._loggingConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    /**
     * @stability stable
     */
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            airflow_configuration_options: cdktf.hashMapper(cdktf.anyToTerraform)(this._airflowConfigurationOptions),
            airflow_version: cdktf.stringToTerraform(this._airflowVersion),
            dag_s3_path: cdktf.stringToTerraform(this._dagS3Path),
            environment_class: cdktf.stringToTerraform(this._environmentClass),
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            kms_key: cdktf.stringToTerraform(this._kmsKey),
            max_workers: cdktf.numberToTerraform(this._maxWorkers),
            min_workers: cdktf.numberToTerraform(this._minWorkers),
            name: cdktf.stringToTerraform(this._name),
            plugins_s3_object_version: cdktf.stringToTerraform(this._pluginsS3ObjectVersion),
            plugins_s3_path: cdktf.stringToTerraform(this._pluginsS3Path),
            requirements_s3_object_version: cdktf.stringToTerraform(this._requirementsS3ObjectVersion),
            requirements_s3_path: cdktf.stringToTerraform(this._requirementsS3Path),
            source_bucket_arn: cdktf.stringToTerraform(this._sourceBucketArn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            webserver_access_mode: cdktf.stringToTerraform(this._webserverAccessMode),
            weekly_maintenance_window_start: cdktf.stringToTerraform(this._weeklyMaintenanceWindowStart),
            logging_configuration: mwaaEnvironmentLoggingConfigurationToTerraform(this._loggingConfiguration.internalValue),
            network_configuration: mwaaEnvironmentNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
        };
    }
}
exports.MwaaEnvironment = MwaaEnvironment;
_k = JSII_RTTI_SYMBOL_1;
MwaaEnvironment[_k] = { fqn: "@cdktf/provider-aws.mwaa.MwaaEnvironment", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MwaaEnvironment.tfResourceType = "aws_mwaa_environment";
