// @cdktf/provider-aws/lib/efs/efs-access-point.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EfsAccessPoint = exports.EfsAccessPointRootDirectoryOutputReference = exports.efsAccessPointRootDirectoryToTerraform = exports.EfsAccessPointRootDirectoryCreationInfoOutputReference = exports.efsAccessPointRootDirectoryCreationInfoToTerraform = exports.EfsAccessPointPosixUserOutputReference = exports.efsAccessPointPosixUserToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function efsAccessPointPosixUserToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gid: cdktf.numberToTerraform(struct.gid),
        secondary_gids: cdktf.listMapper(cdktf.numberToTerraform)(struct.secondaryGids),
        uid: cdktf.numberToTerraform(struct.uid),
    };
}
exports.efsAccessPointPosixUserToTerraform = efsAccessPointPosixUserToTerraform;
/**
 * @stability stable
 */
class EfsAccessPointPosixUserOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gid) {
            hasAnyValues = true;
            internalValueResult.gid = this._gid;
        }
        if (this._secondaryGids) {
            hasAnyValues = true;
            internalValueResult.secondaryGids = this._secondaryGids;
        }
        if (this._uid) {
            hasAnyValues = true;
            internalValueResult.uid = this._uid;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gid = undefined;
            this._secondaryGids = undefined;
            this._uid = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gid = value.gid;
            this._secondaryGids = value.secondaryGids;
            this._uid = value.uid;
        }
    }
    /**
     * @stability stable
     */
    get gid() {
        return this.getNumberAttribute('gid');
    }
    /**
     * @stability stable
     */
    set gid(value) {
        this._gid = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gidInput() {
        return this._gid;
    }
    /**
     * @stability stable
     */
    get secondaryGids() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('secondary_gids');
    }
    /**
     * @stability stable
     */
    set secondaryGids(value) {
        this._secondaryGids = value;
    }
    /**
     * @stability stable
     */
    resetSecondaryGids() {
        this._secondaryGids = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secondaryGidsInput() {
        return this._secondaryGids;
    }
    /**
     * @stability stable
     */
    get uid() {
        return this.getNumberAttribute('uid');
    }
    /**
     * @stability stable
     */
    set uid(value) {
        this._uid = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get uidInput() {
        return this._uid;
    }
}
exports.EfsAccessPointPosixUserOutputReference = EfsAccessPointPosixUserOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EfsAccessPointPosixUserOutputReference[_a] = { fqn: "@cdktf/provider-aws.efs.EfsAccessPointPosixUserOutputReference", version: "3.0.1" };
function efsAccessPointRootDirectoryCreationInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        owner_gid: cdktf.numberToTerraform(struct.ownerGid),
        owner_uid: cdktf.numberToTerraform(struct.ownerUid),
        permissions: cdktf.stringToTerraform(struct.permissions),
    };
}
exports.efsAccessPointRootDirectoryCreationInfoToTerraform = efsAccessPointRootDirectoryCreationInfoToTerraform;
/**
 * @stability stable
 */
class EfsAccessPointRootDirectoryCreationInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ownerGid) {
            hasAnyValues = true;
            internalValueResult.ownerGid = this._ownerGid;
        }
        if (this._ownerUid) {
            hasAnyValues = true;
            internalValueResult.ownerUid = this._ownerUid;
        }
        if (this._permissions) {
            hasAnyValues = true;
            internalValueResult.permissions = this._permissions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ownerGid = undefined;
            this._ownerUid = undefined;
            this._permissions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ownerGid = value.ownerGid;
            this._ownerUid = value.ownerUid;
            this._permissions = value.permissions;
        }
    }
    /**
     * @stability stable
     */
    get ownerGid() {
        return this.getNumberAttribute('owner_gid');
    }
    /**
     * @stability stable
     */
    set ownerGid(value) {
        this._ownerGid = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerGidInput() {
        return this._ownerGid;
    }
    /**
     * @stability stable
     */
    get ownerUid() {
        return this.getNumberAttribute('owner_uid');
    }
    /**
     * @stability stable
     */
    set ownerUid(value) {
        this._ownerUid = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerUidInput() {
        return this._ownerUid;
    }
    /**
     * @stability stable
     */
    get permissions() {
        return this.getStringAttribute('permissions');
    }
    /**
     * @stability stable
     */
    set permissions(value) {
        this._permissions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionsInput() {
        return this._permissions;
    }
}
exports.EfsAccessPointRootDirectoryCreationInfoOutputReference = EfsAccessPointRootDirectoryCreationInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EfsAccessPointRootDirectoryCreationInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.efs.EfsAccessPointRootDirectoryCreationInfoOutputReference", version: "3.0.1" };
function efsAccessPointRootDirectoryToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        creation_info: efsAccessPointRootDirectoryCreationInfoToTerraform(struct.creationInfo),
    };
}
exports.efsAccessPointRootDirectoryToTerraform = efsAccessPointRootDirectoryToTerraform;
/**
 * @stability stable
 */
class EfsAccessPointRootDirectoryOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // creation_info - computed: false, optional: true, required: false
        this._creationInfo = new EfsAccessPointRootDirectoryCreationInfoOutputReference(this, "creation_info", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if ((_e = this._creationInfo) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.creationInfo = (_f = this._creationInfo) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._creationInfo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._creationInfo.internalValue = value.creationInfo;
        }
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    /**
     * @stability stable
     */
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
    /**
     * @stability stable
     */
    get creationInfo() {
        return this._creationInfo;
    }
    /**
     * @stability stable
     */
    putCreationInfo(value) {
        this._creationInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCreationInfo() {
        this._creationInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get creationInfoInput() {
        return this._creationInfo.internalValue;
    }
}
exports.EfsAccessPointRootDirectoryOutputReference = EfsAccessPointRootDirectoryOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EfsAccessPointRootDirectoryOutputReference[_c] = { fqn: "@cdktf/provider-aws.efs.EfsAccessPointRootDirectoryOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/efs_access_point.html aws_efs_access_point}.
 *
 * @stability stable
 */
class EfsAccessPoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/efs_access_point.html aws_efs_access_point} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_access_point',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // posix_user - computed: false, optional: true, required: false
        this._posixUser = new EfsAccessPointPosixUserOutputReference(this, "posix_user", true);
        // root_directory - computed: false, optional: true, required: false
        this._rootDirectory = new EfsAccessPointRootDirectoryOutputReference(this, "root_directory", true);
        this._fileSystemId = config.fileSystemId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._posixUser.internalValue = config.posixUser;
        this._rootDirectory.internalValue = config.rootDirectory;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // file_system_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get fileSystemArn() {
        return this.getStringAttribute('file_system_arn');
    }
    /**
     * @stability stable
     */
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    /**
     * @stability stable
     */
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get posixUser() {
        return this._posixUser;
    }
    /**
     * @stability stable
     */
    putPosixUser(value) {
        this._posixUser.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPosixUser() {
        this._posixUser.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get posixUserInput() {
        return this._posixUser.internalValue;
    }
    /**
     * @stability stable
     */
    get rootDirectory() {
        return this._rootDirectory;
    }
    /**
     * @stability stable
     */
    putRootDirectory(value) {
        this._rootDirectory.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRootDirectory() {
        this._rootDirectory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootDirectoryInput() {
        return this._rootDirectory.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            file_system_id: cdktf.stringToTerraform(this._fileSystemId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            posix_user: efsAccessPointPosixUserToTerraform(this._posixUser.internalValue),
            root_directory: efsAccessPointRootDirectoryToTerraform(this._rootDirectory.internalValue),
        };
    }
}
exports.EfsAccessPoint = EfsAccessPoint;
_d = JSII_RTTI_SYMBOL_1;
EfsAccessPoint[_d] = { fqn: "@cdktf/provider-aws.efs.EfsAccessPoint", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EfsAccessPoint.tfResourceType = "aws_efs_access_point";
