// @cdktf/provider-aws/lib/elasticache/data-aws-elasticache-replication-group.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElasticacheReplicationGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/elasticache_replication_group.html aws_elasticache_replication_group}.
 *
 * @stability stable
 */
class DataAwsElasticacheReplicationGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/elasticache_replication_group.html aws_elasticache_replication_group} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_replication_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._replicationGroupId = config.replicationGroupId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // auth_token_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authTokenEnabled() {
        return this.getBooleanAttribute('auth_token_enabled');
    }
    // automatic_failover_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get automaticFailoverEnabled() {
        return this.getBooleanAttribute('automatic_failover_enabled');
    }
    // configuration_endpoint_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get configurationEndpointAddress() {
        return this.getStringAttribute('configuration_endpoint_address');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // member_clusters - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get memberClusters() {
        return this.getListAttribute('member_clusters');
    }
    // multi_az_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get multiAzEnabled() {
        return this.getBooleanAttribute('multi_az_enabled');
    }
    // node_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    // number_cache_clusters - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get numberCacheClusters() {
        return this.getNumberAttribute('number_cache_clusters');
    }
    // port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    // primary_endpoint_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get primaryEndpointAddress() {
        return this.getStringAttribute('primary_endpoint_address');
    }
    // reader_endpoint_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get readerEndpointAddress() {
        return this.getStringAttribute('reader_endpoint_address');
    }
    // replication_group_description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get replicationGroupDescription() {
        return this.getStringAttribute('replication_group_description');
    }
    /**
     * @stability stable
     */
    get replicationGroupId() {
        return this.getStringAttribute('replication_group_id');
    }
    /**
     * @stability stable
     */
    set replicationGroupId(value) {
        this._replicationGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replicationGroupIdInput() {
        return this._replicationGroupId;
    }
    // snapshot_retention_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotRetentionLimit() {
        return this.getNumberAttribute('snapshot_retention_limit');
    }
    // snapshot_window - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotWindow() {
        return this.getStringAttribute('snapshot_window');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            replication_group_id: cdktf.stringToTerraform(this._replicationGroupId),
        };
    }
}
exports.DataAwsElasticacheReplicationGroup = DataAwsElasticacheReplicationGroup;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElasticacheReplicationGroup[_a] = { fqn: "@cdktf/provider-aws.elasticache.DataAwsElasticacheReplicationGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsElasticacheReplicationGroup.tfResourceType = "aws_elasticache_replication_group";
