// @cdktf/provider-aws/lib/elasticache/elasticache-global-replication-group.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticacheGlobalReplicationGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/elasticache_global_replication_group.html aws_elasticache_global_replication_group}.
 *
 * @stability stable
 */
class ElasticacheGlobalReplicationGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elasticache_global_replication_group.html aws_elasticache_global_replication_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_global_replication_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._globalReplicationGroupDescription = config.globalReplicationGroupDescription;
        this._globalReplicationGroupIdSuffix = config.globalReplicationGroupIdSuffix;
        this._primaryReplicationGroupId = config.primaryReplicationGroupId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // actual_engine_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get actualEngineVersion() {
        return this.getStringAttribute('actual_engine_version');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // at_rest_encryption_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get atRestEncryptionEnabled() {
        return this.getBooleanAttribute('at_rest_encryption_enabled');
    }
    // auth_token_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authTokenEnabled() {
        return this.getBooleanAttribute('auth_token_enabled');
    }
    // cache_node_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cacheNodeType() {
        return this.getStringAttribute('cache_node_type');
    }
    // cluster_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clusterEnabled() {
        return this.getBooleanAttribute('cluster_enabled');
    }
    // engine - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engine() {
        return this.getStringAttribute('engine');
    }
    // engine_version_actual - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engineVersionActual() {
        return this.getStringAttribute('engine_version_actual');
    }
    /**
     * @stability stable
     */
    get globalReplicationGroupDescription() {
        return this.getStringAttribute('global_replication_group_description');
    }
    /**
     * @stability stable
     */
    set globalReplicationGroupDescription(value) {
        this._globalReplicationGroupDescription = value;
    }
    /**
     * @stability stable
     */
    resetGlobalReplicationGroupDescription() {
        this._globalReplicationGroupDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get globalReplicationGroupDescriptionInput() {
        return this._globalReplicationGroupDescription;
    }
    // global_replication_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get globalReplicationGroupId() {
        return this.getStringAttribute('global_replication_group_id');
    }
    /**
     * @stability stable
     */
    get globalReplicationGroupIdSuffix() {
        return this.getStringAttribute('global_replication_group_id_suffix');
    }
    /**
     * @stability stable
     */
    set globalReplicationGroupIdSuffix(value) {
        this._globalReplicationGroupIdSuffix = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get globalReplicationGroupIdSuffixInput() {
        return this._globalReplicationGroupIdSuffix;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get primaryReplicationGroupId() {
        return this.getStringAttribute('primary_replication_group_id');
    }
    /**
     * @stability stable
     */
    set primaryReplicationGroupId(value) {
        this._primaryReplicationGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get primaryReplicationGroupIdInput() {
        return this._primaryReplicationGroupId;
    }
    // transit_encryption_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get transitEncryptionEnabled() {
        return this.getBooleanAttribute('transit_encryption_enabled');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            global_replication_group_description: cdktf.stringToTerraform(this._globalReplicationGroupDescription),
            global_replication_group_id_suffix: cdktf.stringToTerraform(this._globalReplicationGroupIdSuffix),
            primary_replication_group_id: cdktf.stringToTerraform(this._primaryReplicationGroupId),
        };
    }
}
exports.ElasticacheGlobalReplicationGroup = ElasticacheGlobalReplicationGroup;
_a = JSII_RTTI_SYMBOL_1;
ElasticacheGlobalReplicationGroup[_a] = { fqn: "@cdktf/provider-aws.elasticache.ElasticacheGlobalReplicationGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ElasticacheGlobalReplicationGroup.tfResourceType = "aws_elasticache_global_replication_group";
