// @cdktf/provider-aws/lib/route53/route53recoverycontrolconfig-cluster.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53RecoverycontrolconfigCluster = exports.Route53RecoverycontrolconfigClusterClusterEndpoints = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class Route53RecoverycontrolconfigClusterClusterEndpoints extends cdktf.ComplexComputedList {
    // endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.Route53RecoverycontrolconfigClusterClusterEndpoints = Route53RecoverycontrolconfigClusterClusterEndpoints;
_a = JSII_RTTI_SYMBOL_1;
Route53RecoverycontrolconfigClusterClusterEndpoints[_a] = { fqn: "@cdktf/provider-aws.route53.Route53RecoverycontrolconfigClusterClusterEndpoints", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53recoverycontrolconfig_cluster.html aws_route53recoverycontrolconfig_cluster}.
 *
 * @stability stable
 */
class Route53RecoverycontrolconfigCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53recoverycontrolconfig_cluster.html aws_route53recoverycontrolconfig_cluster} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53recoverycontrolconfig_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // cluster_endpoints - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    clusterEndpoints(index) {
        return new Route53RecoverycontrolconfigClusterClusterEndpoints(this, 'cluster_endpoints', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.Route53RecoverycontrolconfigCluster = Route53RecoverycontrolconfigCluster;
_b = JSII_RTTI_SYMBOL_1;
Route53RecoverycontrolconfigCluster[_b] = { fqn: "@cdktf/provider-aws.route53.Route53RecoverycontrolconfigCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Route53RecoverycontrolconfigCluster.tfResourceType = "aws_route53recoverycontrolconfig_cluster";
