// @cdktf/provider-aws/lib/ram/ram-resource-share-accepter.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RamResourceShareAccepter = exports.RamResourceShareAccepterTimeoutsOutputReference = exports.ramResourceShareAccepterTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ramResourceShareAccepterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.ramResourceShareAccepterTimeoutsToTerraform = ramResourceShareAccepterTimeoutsToTerraform;
/**
 * @stability stable
 */
class RamResourceShareAccepterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.RamResourceShareAccepterTimeoutsOutputReference = RamResourceShareAccepterTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RamResourceShareAccepterTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.ram.RamResourceShareAccepterTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ram_resource_share_accepter.html aws_ram_resource_share_accepter}.
 *
 * @stability stable
 */
class RamResourceShareAccepter extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ram_resource_share_accepter.html aws_ram_resource_share_accepter} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ram_resource_share_accepter',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RamResourceShareAccepterTimeoutsOutputReference(this, "timeouts", true);
        this._shareArn = config.shareArn;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // invitation_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get invitationArn() {
        return this.getStringAttribute('invitation_arn');
    }
    // receiver_account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get receiverAccountId() {
        return this.getStringAttribute('receiver_account_id');
    }
    // resources - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get resources() {
        return this.getListAttribute('resources');
    }
    // sender_account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get senderAccountId() {
        return this.getStringAttribute('sender_account_id');
    }
    /**
     * @stability stable
     */
    get shareArn() {
        return this.getStringAttribute('share_arn');
    }
    /**
     * @stability stable
     */
    set shareArn(value) {
        this._shareArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get shareArnInput() {
        return this._shareArn;
    }
    // share_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get shareId() {
        return this.getStringAttribute('share_id');
    }
    // share_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get shareName() {
        return this.getStringAttribute('share_name');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            share_arn: cdktf.stringToTerraform(this._shareArn),
            timeouts: ramResourceShareAccepterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.RamResourceShareAccepter = RamResourceShareAccepter;
_b = JSII_RTTI_SYMBOL_1;
RamResourceShareAccepter[_b] = { fqn: "@cdktf/provider-aws.ram.RamResourceShareAccepter", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
RamResourceShareAccepter.tfResourceType = "aws_ram_resource_share_accepter";
