// @cdktf/provider-aws/lib/elb/data-aws-alb.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAlb = exports.DataAwsAlbSubnetMapping = exports.DataAwsAlbAccessLogs = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsAlbAccessLogs extends cdktf.ComplexComputedList {
    // bucket - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
}
exports.DataAwsAlbAccessLogs = DataAwsAlbAccessLogs;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAlbAccessLogs[_a] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbAccessLogs", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbSubnetMapping extends cdktf.ComplexComputedList {
    // allocation_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get allocationId() {
        return this.getStringAttribute('allocation_id');
    }
    // ipv6_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    // outpost_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get outpostId() {
        return this.getStringAttribute('outpost_id');
    }
    // private_ipv4_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIpv4Address() {
        return this.getStringAttribute('private_ipv4_address');
    }
    // subnet_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.DataAwsAlbSubnetMapping = DataAwsAlbSubnetMapping;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAlbSubnetMapping[_b] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbSubnetMapping", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/alb.html aws_alb}.
 *
 * @stability stable
 */
class DataAwsAlb extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/alb.html aws_alb} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_alb',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_logs - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    accessLogs(index) {
        return new DataAwsAlbAccessLogs(this, 'access_logs', index);
    }
    // arn - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // arn_suffix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arnSuffix() {
        return this.getStringAttribute('arn_suffix');
    }
    // customer_owned_ipv4_pool - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customerOwnedIpv4Pool() {
        return this.getStringAttribute('customer_owned_ipv4_pool');
    }
    // desync_mitigation_mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get desyncMitigationMode() {
        return this.getStringAttribute('desync_mitigation_mode');
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // drop_invalid_header_fields - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dropInvalidHeaderFields() {
        return this.getBooleanAttribute('drop_invalid_header_fields');
    }
    // enable_deletion_protection - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableDeletionProtection() {
        return this.getBooleanAttribute('enable_deletion_protection');
    }
    // enable_http2 - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableHttp2() {
        return this.getBooleanAttribute('enable_http2');
    }
    // enable_waf_fail_open - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableWafFailOpen() {
        return this.getBooleanAttribute('enable_waf_fail_open');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // idle_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get idleTimeout() {
        return this.getNumberAttribute('idle_timeout');
    }
    // internal - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get internal() {
        return this.getBooleanAttribute('internal');
    }
    // ip_address_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    // load_balancer_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get loadBalancerType() {
        return this.getStringAttribute('load_balancer_type');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    // subnet_mapping - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    subnetMapping(index) {
        return new DataAwsAlbSubnetMapping(this, 'subnet_mapping', index);
    }
    // subnets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnets() {
        return this.getListAttribute('subnets');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsAlb = DataAwsAlb;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAlb[_c] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlb", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAlb.tfResourceType = "aws_alb";
