// @cdktf/provider-aws/lib/rds/data-aws-rds-engine-version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRdsEngineVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/rds_engine_version.html aws_rds_engine_version}.
 *
 * @stability stable
 */
class DataAwsRdsEngineVersion extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/rds_engine_version.html aws_rds_engine_version} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_engine_version',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._engine = config.engine;
        this._parameterGroupFamily = config.parameterGroupFamily;
        this._preferredVersions = config.preferredVersions;
        this._version = config.version;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // default_character_set - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultCharacterSet() {
        return this.getStringAttribute('default_character_set');
    }
    /**
     * @stability stable
     */
    get engine() {
        return this.getStringAttribute('engine');
    }
    /**
     * @stability stable
     */
    set engine(value) {
        this._engine = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get engineInput() {
        return this._engine;
    }
    // engine_description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get engineDescription() {
        return this.getStringAttribute('engine_description');
    }
    // exportable_log_types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get exportableLogTypes() {
        return this.getListAttribute('exportable_log_types');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get parameterGroupFamily() {
        return this.getStringAttribute('parameter_group_family');
    }
    /**
     * @stability stable
     */
    set parameterGroupFamily(value) {
        this._parameterGroupFamily = value;
    }
    /**
     * @stability stable
     */
    resetParameterGroupFamily() {
        this._parameterGroupFamily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parameterGroupFamilyInput() {
        return this._parameterGroupFamily;
    }
    /**
     * @stability stable
     */
    get preferredVersions() {
        return this.getListAttribute('preferred_versions');
    }
    /**
     * @stability stable
     */
    set preferredVersions(value) {
        this._preferredVersions = value;
    }
    /**
     * @stability stable
     */
    resetPreferredVersions() {
        this._preferredVersions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get preferredVersionsInput() {
        return this._preferredVersions;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // supported_character_sets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportedCharacterSets() {
        return this.getListAttribute('supported_character_sets');
    }
    // supported_feature_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportedFeatureNames() {
        return this.getListAttribute('supported_feature_names');
    }
    // supported_modes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportedModes() {
        return this.getListAttribute('supported_modes');
    }
    // supported_timezones - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportedTimezones() {
        return this.getListAttribute('supported_timezones');
    }
    // supports_global_databases - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportsGlobalDatabases() {
        return this.getBooleanAttribute('supports_global_databases');
    }
    // supports_log_exports_to_cloudwatch - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportsLogExportsToCloudwatch() {
        return this.getBooleanAttribute('supports_log_exports_to_cloudwatch');
    }
    // supports_parallel_query - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportsParallelQuery() {
        return this.getBooleanAttribute('supports_parallel_query');
    }
    // supports_read_replica - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get supportsReadReplica() {
        return this.getBooleanAttribute('supports_read_replica');
    }
    // valid_upgrade_targets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get validUpgradeTargets() {
        return this.getListAttribute('valid_upgrade_targets');
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    /**
     * @stability stable
     */
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
    // version_description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get versionDescription() {
        return this.getStringAttribute('version_description');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            engine: cdktf.stringToTerraform(this._engine),
            parameter_group_family: cdktf.stringToTerraform(this._parameterGroupFamily),
            preferred_versions: cdktf.listMapper(cdktf.stringToTerraform)(this._preferredVersions),
            version: cdktf.stringToTerraform(this._version),
        };
    }
}
exports.DataAwsRdsEngineVersion = DataAwsRdsEngineVersion;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRdsEngineVersion[_a] = { fqn: "@cdktf/provider-aws.rds.DataAwsRdsEngineVersion", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsRdsEngineVersion.tfResourceType = "aws_rds_engine_version";
