// @cdktf/provider-aws/lib/rds/db-option-group.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbOptionGroup = exports.DbOptionGroupTimeoutsOutputReference = exports.dbOptionGroupTimeoutsToTerraform = exports.dbOptionGroupOptionToTerraform = exports.dbOptionGroupOptionOptionSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbOptionGroupOptionOptionSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dbOptionGroupOptionOptionSettingsToTerraform = dbOptionGroupOptionOptionSettingsToTerraform;
function dbOptionGroupOptionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        db_security_group_memberships: cdktf.listMapper(cdktf.stringToTerraform)(struct.dbSecurityGroupMemberships),
        option_name: cdktf.stringToTerraform(struct.optionName),
        port: cdktf.numberToTerraform(struct.port),
        version: cdktf.stringToTerraform(struct.version),
        vpc_security_group_memberships: cdktf.listMapper(cdktf.stringToTerraform)(struct.vpcSecurityGroupMemberships),
        option_settings: cdktf.listMapper(dbOptionGroupOptionOptionSettingsToTerraform)(struct.optionSettings),
    };
}
exports.dbOptionGroupOptionToTerraform = dbOptionGroupOptionToTerraform;
function dbOptionGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.dbOptionGroupTimeoutsToTerraform = dbOptionGroupTimeoutsToTerraform;
/**
 * @stability stable
 */
class DbOptionGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.DbOptionGroupTimeoutsOutputReference = DbOptionGroupTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbOptionGroupTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.rds.DbOptionGroupTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/db_option_group.html aws_db_option_group}.
 *
 * @stability stable
 */
class DbOptionGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/db_option_group.html aws_db_option_group} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_option_group',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbOptionGroupTimeoutsOutputReference(this, "timeouts", true);
        this._engineName = config.engineName;
        this._majorEngineVersion = config.majorEngineVersion;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._optionGroupDescription = config.optionGroupDescription;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._option = config.option;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get engineName() {
        return this.getStringAttribute('engine_name');
    }
    /**
     * @stability stable
     */
    set engineName(value) {
        this._engineName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get engineNameInput() {
        return this._engineName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get majorEngineVersion() {
        return this.getStringAttribute('major_engine_version');
    }
    /**
     * @stability stable
     */
    set majorEngineVersion(value) {
        this._majorEngineVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get majorEngineVersionInput() {
        return this._majorEngineVersion;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get optionGroupDescription() {
        return this.getStringAttribute('option_group_description');
    }
    /**
     * @stability stable
     */
    set optionGroupDescription(value) {
        this._optionGroupDescription = value;
    }
    /**
     * @stability stable
     */
    resetOptionGroupDescription() {
        this._optionGroupDescription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get optionGroupDescriptionInput() {
        return this._optionGroupDescription;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get option() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('option');
    }
    /**
     * @stability stable
     */
    set option(value) {
        this._option = value;
    }
    /**
     * @stability stable
     */
    resetOption() {
        this._option = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get optionInput() {
        return this._option;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            engine_name: cdktf.stringToTerraform(this._engineName),
            major_engine_version: cdktf.stringToTerraform(this._majorEngineVersion),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            option_group_description: cdktf.stringToTerraform(this._optionGroupDescription),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            option: cdktf.listMapper(dbOptionGroupOptionToTerraform)(this._option),
            timeouts: dbOptionGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DbOptionGroup = DbOptionGroup;
_b = JSII_RTTI_SYMBOL_1;
DbOptionGroup[_b] = { fqn: "@cdktf/provider-aws.rds.DbOptionGroup", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DbOptionGroup.tfResourceType = "aws_db_option_group";
