// @cdktf/provider-aws/lib/rds/rds-cluster-role-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RdsClusterRoleAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/rds_cluster_role_association.html aws_rds_cluster_role_association}.
 *
 * @stability stable
 */
class RdsClusterRoleAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/rds_cluster_role_association.html aws_rds_cluster_role_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_cluster_role_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dbClusterIdentifier = config.dbClusterIdentifier;
        this._featureName = config.featureName;
        this._roleArn = config.roleArn;
    }
    /**
     * @stability stable
     */
    get dbClusterIdentifier() {
        return this.getStringAttribute('db_cluster_identifier');
    }
    /**
     * @stability stable
     */
    set dbClusterIdentifier(value) {
        this._dbClusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dbClusterIdentifierInput() {
        return this._dbClusterIdentifier;
    }
    /**
     * @stability stable
     */
    get featureName() {
        return this.getStringAttribute('feature_name');
    }
    /**
     * @stability stable
     */
    set featureName(value) {
        this._featureName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get featureNameInput() {
        return this._featureName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            db_cluster_identifier: cdktf.stringToTerraform(this._dbClusterIdentifier),
            feature_name: cdktf.stringToTerraform(this._featureName),
            role_arn: cdktf.stringToTerraform(this._roleArn),
        };
    }
}
exports.RdsClusterRoleAssociation = RdsClusterRoleAssociation;
_a = JSII_RTTI_SYMBOL_1;
RdsClusterRoleAssociation[_a] = { fqn: "@cdktf/provider-aws.rds.RdsClusterRoleAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
RdsClusterRoleAssociation.tfResourceType = "aws_rds_cluster_role_association";
