// @cdktf/provider-aws/lib/codeartifact/data-aws-codeartifact-authorization-token.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCodeartifactAuthorizationToken = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/codeartifact_authorization_token.html aws_codeartifact_authorization_token}.
 *
 * @stability stable
 */
class DataAwsCodeartifactAuthorizationToken extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/codeartifact_authorization_token.html aws_codeartifact_authorization_token} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codeartifact_authorization_token',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._domain = config.domain;
        this._domainOwner = config.domainOwner;
        this._durationSeconds = config.durationSeconds;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // authorization_token - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authorizationToken() {
        return this.getStringAttribute('authorization_token');
    }
    /**
     * @stability stable
     */
    get domain() {
        return this.getStringAttribute('domain');
    }
    /**
     * @stability stable
     */
    set domain(value) {
        this._domain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainInput() {
        return this._domain;
    }
    /**
     * @stability stable
     */
    get domainOwner() {
        return this.getStringAttribute('domain_owner');
    }
    /**
     * @stability stable
     */
    set domainOwner(value) {
        this._domainOwner = value;
    }
    /**
     * @stability stable
     */
    resetDomainOwner() {
        this._domainOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainOwnerInput() {
        return this._domainOwner;
    }
    /**
     * @stability stable
     */
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    /**
     * @stability stable
     */
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    /**
     * @stability stable
     */
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    // expiration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get expiration() {
        return this.getStringAttribute('expiration');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            domain: cdktf.stringToTerraform(this._domain),
            domain_owner: cdktf.stringToTerraform(this._domainOwner),
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
        };
    }
}
exports.DataAwsCodeartifactAuthorizationToken = DataAwsCodeartifactAuthorizationToken;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCodeartifactAuthorizationToken[_a] = { fqn: "@cdktf/provider-aws.codeartifact.DataAwsCodeartifactAuthorizationToken", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCodeartifactAuthorizationToken.tfResourceType = "aws_codeartifact_authorization_token";
