// @cdktf/provider-aws/lib/workspaces/data-aws-workspaces-directory.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsWorkspacesDirectory = exports.DataAwsWorkspacesDirectoryWorkspaceCreationProperties = exports.DataAwsWorkspacesDirectoryWorkspaceAccessProperties = exports.DataAwsWorkspacesDirectorySelfServicePermissions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsWorkspacesDirectorySelfServicePermissions extends cdktf.ComplexComputedList {
    // change_compute_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get changeComputeType() {
        return this.getBooleanAttribute('change_compute_type');
    }
    // increase_volume_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get increaseVolumeSize() {
        return this.getBooleanAttribute('increase_volume_size');
    }
    // rebuild_workspace - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rebuildWorkspace() {
        return this.getBooleanAttribute('rebuild_workspace');
    }
    // restart_workspace - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get restartWorkspace() {
        return this.getBooleanAttribute('restart_workspace');
    }
    // switch_running_mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get switchRunningMode() {
        return this.getBooleanAttribute('switch_running_mode');
    }
}
exports.DataAwsWorkspacesDirectorySelfServicePermissions = DataAwsWorkspacesDirectorySelfServicePermissions;
_a = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectorySelfServicePermissions[_a] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesDirectorySelfServicePermissions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsWorkspacesDirectoryWorkspaceAccessProperties extends cdktf.ComplexComputedList {
    // device_type_android - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeAndroid() {
        return this.getStringAttribute('device_type_android');
    }
    // device_type_chromeos - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeChromeos() {
        return this.getStringAttribute('device_type_chromeos');
    }
    // device_type_ios - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeIos() {
        return this.getStringAttribute('device_type_ios');
    }
    // device_type_linux - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeLinux() {
        return this.getStringAttribute('device_type_linux');
    }
    // device_type_osx - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeOsx() {
        return this.getStringAttribute('device_type_osx');
    }
    // device_type_web - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeWeb() {
        return this.getStringAttribute('device_type_web');
    }
    // device_type_windows - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeWindows() {
        return this.getStringAttribute('device_type_windows');
    }
    // device_type_zeroclient - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceTypeZeroclient() {
        return this.getStringAttribute('device_type_zeroclient');
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceAccessProperties = DataAwsWorkspacesDirectoryWorkspaceAccessProperties;
_b = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceAccessProperties[_b] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesDirectoryWorkspaceAccessProperties", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsWorkspacesDirectoryWorkspaceCreationProperties extends cdktf.ComplexComputedList {
    // custom_security_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customSecurityGroupId() {
        return this.getStringAttribute('custom_security_group_id');
    }
    // default_ou - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultOu() {
        return this.getStringAttribute('default_ou');
    }
    // enable_internet_access - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableInternetAccess() {
        return this.getBooleanAttribute('enable_internet_access');
    }
    // enable_maintenance_mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enableMaintenanceMode() {
        return this.getBooleanAttribute('enable_maintenance_mode');
    }
    // user_enabled_as_local_administrator - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userEnabledAsLocalAdministrator() {
        return this.getBooleanAttribute('user_enabled_as_local_administrator');
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceCreationProperties = DataAwsWorkspacesDirectoryWorkspaceCreationProperties;
_c = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceCreationProperties[_c] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesDirectoryWorkspaceCreationProperties", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/workspaces_directory.html aws_workspaces_directory}.
 *
 * @stability stable
 */
class DataAwsWorkspacesDirectory extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/workspaces_directory.html aws_workspaces_directory} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_workspaces_directory',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._directoryId = config.directoryId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // alias - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get alias() {
        return this.getStringAttribute('alias');
    }
    // customer_user_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customerUserName() {
        return this.getStringAttribute('customer_user_name');
    }
    /**
     * @stability stable
     */
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    /**
     * @stability stable
     */
    set directoryId(value) {
        this._directoryId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get directoryIdInput() {
        return this._directoryId;
    }
    // directory_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get directoryName() {
        return this.getStringAttribute('directory_name');
    }
    // directory_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get directoryType() {
        return this.getStringAttribute('directory_type');
    }
    // dns_ip_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsIpAddresses() {
        return this.getListAttribute('dns_ip_addresses');
    }
    // iam_role_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iamRoleId() {
        return this.getStringAttribute('iam_role_id');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // ip_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipGroupIds() {
        return this.getListAttribute('ip_group_ids');
    }
    // registration_code - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get registrationCode() {
        return this.getStringAttribute('registration_code');
    }
    // self_service_permissions - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    selfServicePermissions(index) {
        return new DataAwsWorkspacesDirectorySelfServicePermissions(this, 'self_service_permissions', index);
    }
    // subnet_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // workspace_access_properties - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    workspaceAccessProperties(index) {
        return new DataAwsWorkspacesDirectoryWorkspaceAccessProperties(this, 'workspace_access_properties', index);
    }
    // workspace_creation_properties - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    workspaceCreationProperties(index) {
        return new DataAwsWorkspacesDirectoryWorkspaceCreationProperties(this, 'workspace_creation_properties', index);
    }
    // workspace_security_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get workspaceSecurityGroupId() {
        return this.getStringAttribute('workspace_security_group_id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            directory_id: cdktf.stringToTerraform(this._directoryId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsWorkspacesDirectory = DataAwsWorkspacesDirectory;
_d = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectory[_d] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesDirectory", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsWorkspacesDirectory.tfResourceType = "aws_workspaces_directory";
