// @cdktf/provider-aws/lib/workspaces/workspaces-workspace.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacesWorkspace = exports.WorkspacesWorkspaceWorkspacePropertiesOutputReference = exports.workspacesWorkspaceWorkspacePropertiesToTerraform = exports.WorkspacesWorkspaceTimeoutsOutputReference = exports.workspacesWorkspaceTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function workspacesWorkspaceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.workspacesWorkspaceTimeoutsToTerraform = workspacesWorkspaceTimeoutsToTerraform;
/**
 * @stability stable
 */
class WorkspacesWorkspaceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.WorkspacesWorkspaceTimeoutsOutputReference = WorkspacesWorkspaceTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WorkspacesWorkspaceTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.workspaces.WorkspacesWorkspaceTimeoutsOutputReference", version: "3.0.1" };
function workspacesWorkspaceWorkspacePropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        compute_type_name: cdktf.stringToTerraform(struct.computeTypeName),
        root_volume_size_gib: cdktf.numberToTerraform(struct.rootVolumeSizeGib),
        running_mode: cdktf.stringToTerraform(struct.runningMode),
        running_mode_auto_stop_timeout_in_minutes: cdktf.numberToTerraform(struct.runningModeAutoStopTimeoutInMinutes),
        user_volume_size_gib: cdktf.numberToTerraform(struct.userVolumeSizeGib),
    };
}
exports.workspacesWorkspaceWorkspacePropertiesToTerraform = workspacesWorkspaceWorkspacePropertiesToTerraform;
/**
 * @stability stable
 */
class WorkspacesWorkspaceWorkspacePropertiesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._computeTypeName) {
            hasAnyValues = true;
            internalValueResult.computeTypeName = this._computeTypeName;
        }
        if (this._rootVolumeSizeGib) {
            hasAnyValues = true;
            internalValueResult.rootVolumeSizeGib = this._rootVolumeSizeGib;
        }
        if (this._runningMode) {
            hasAnyValues = true;
            internalValueResult.runningMode = this._runningMode;
        }
        if (this._runningModeAutoStopTimeoutInMinutes) {
            hasAnyValues = true;
            internalValueResult.runningModeAutoStopTimeoutInMinutes = this._runningModeAutoStopTimeoutInMinutes;
        }
        if (this._userVolumeSizeGib) {
            hasAnyValues = true;
            internalValueResult.userVolumeSizeGib = this._userVolumeSizeGib;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._computeTypeName = undefined;
            this._rootVolumeSizeGib = undefined;
            this._runningMode = undefined;
            this._runningModeAutoStopTimeoutInMinutes = undefined;
            this._userVolumeSizeGib = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._computeTypeName = value.computeTypeName;
            this._rootVolumeSizeGib = value.rootVolumeSizeGib;
            this._runningMode = value.runningMode;
            this._runningModeAutoStopTimeoutInMinutes = value.runningModeAutoStopTimeoutInMinutes;
            this._userVolumeSizeGib = value.userVolumeSizeGib;
        }
    }
    /**
     * @stability stable
     */
    get computeTypeName() {
        return this.getStringAttribute('compute_type_name');
    }
    /**
     * @stability stable
     */
    set computeTypeName(value) {
        this._computeTypeName = value;
    }
    /**
     * @stability stable
     */
    resetComputeTypeName() {
        this._computeTypeName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get computeTypeNameInput() {
        return this._computeTypeName;
    }
    /**
     * @stability stable
     */
    get rootVolumeSizeGib() {
        return this.getNumberAttribute('root_volume_size_gib');
    }
    /**
     * @stability stable
     */
    set rootVolumeSizeGib(value) {
        this._rootVolumeSizeGib = value;
    }
    /**
     * @stability stable
     */
    resetRootVolumeSizeGib() {
        this._rootVolumeSizeGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootVolumeSizeGibInput() {
        return this._rootVolumeSizeGib;
    }
    /**
     * @stability stable
     */
    get runningMode() {
        return this.getStringAttribute('running_mode');
    }
    /**
     * @stability stable
     */
    set runningMode(value) {
        this._runningMode = value;
    }
    /**
     * @stability stable
     */
    resetRunningMode() {
        this._runningMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runningModeInput() {
        return this._runningMode;
    }
    /**
     * @stability stable
     */
    get runningModeAutoStopTimeoutInMinutes() {
        return this.getNumberAttribute('running_mode_auto_stop_timeout_in_minutes');
    }
    /**
     * @stability stable
     */
    set runningModeAutoStopTimeoutInMinutes(value) {
        this._runningModeAutoStopTimeoutInMinutes = value;
    }
    /**
     * @stability stable
     */
    resetRunningModeAutoStopTimeoutInMinutes() {
        this._runningModeAutoStopTimeoutInMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runningModeAutoStopTimeoutInMinutesInput() {
        return this._runningModeAutoStopTimeoutInMinutes;
    }
    /**
     * @stability stable
     */
    get userVolumeSizeGib() {
        return this.getNumberAttribute('user_volume_size_gib');
    }
    /**
     * @stability stable
     */
    set userVolumeSizeGib(value) {
        this._userVolumeSizeGib = value;
    }
    /**
     * @stability stable
     */
    resetUserVolumeSizeGib() {
        this._userVolumeSizeGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userVolumeSizeGibInput() {
        return this._userVolumeSizeGib;
    }
}
exports.WorkspacesWorkspaceWorkspacePropertiesOutputReference = WorkspacesWorkspaceWorkspacePropertiesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
WorkspacesWorkspaceWorkspacePropertiesOutputReference[_b] = { fqn: "@cdktf/provider-aws.workspaces.WorkspacesWorkspaceWorkspacePropertiesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/workspaces_workspace.html aws_workspaces_workspace}.
 *
 * @stability stable
 */
class WorkspacesWorkspace extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/workspaces_workspace.html aws_workspaces_workspace} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_workspaces_workspace',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new WorkspacesWorkspaceTimeoutsOutputReference(this, "timeouts", true);
        // workspace_properties - computed: false, optional: true, required: false
        this._workspaceProperties = new WorkspacesWorkspaceWorkspacePropertiesOutputReference(this, "workspace_properties", true);
        this._bundleId = config.bundleId;
        this._directoryId = config.directoryId;
        this._rootVolumeEncryptionEnabled = config.rootVolumeEncryptionEnabled;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._userName = config.userName;
        this._userVolumeEncryptionEnabled = config.userVolumeEncryptionEnabled;
        this._volumeEncryptionKey = config.volumeEncryptionKey;
        this._timeouts.internalValue = config.timeouts;
        this._workspaceProperties.internalValue = config.workspaceProperties;
    }
    /**
     * @stability stable
     */
    get bundleId() {
        return this.getStringAttribute('bundle_id');
    }
    /**
     * @stability stable
     */
    set bundleId(value) {
        this._bundleId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bundleIdInput() {
        return this._bundleId;
    }
    // computer_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get computerName() {
        return this.getStringAttribute('computer_name');
    }
    /**
     * @stability stable
     */
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    /**
     * @stability stable
     */
    set directoryId(value) {
        this._directoryId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get directoryIdInput() {
        return this._directoryId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    /**
     * @stability stable
     */
    get rootVolumeEncryptionEnabled() {
        return this.getBooleanAttribute('root_volume_encryption_enabled');
    }
    /**
     * @stability stable
     */
    set rootVolumeEncryptionEnabled(value) {
        this._rootVolumeEncryptionEnabled = value;
    }
    /**
     * @stability stable
     */
    resetRootVolumeEncryptionEnabled() {
        this._rootVolumeEncryptionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rootVolumeEncryptionEnabledInput() {
        return this._rootVolumeEncryptionEnabled;
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get userName() {
        return this.getStringAttribute('user_name');
    }
    /**
     * @stability stable
     */
    set userName(value) {
        this._userName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userNameInput() {
        return this._userName;
    }
    /**
     * @stability stable
     */
    get userVolumeEncryptionEnabled() {
        return this.getBooleanAttribute('user_volume_encryption_enabled');
    }
    /**
     * @stability stable
     */
    set userVolumeEncryptionEnabled(value) {
        this._userVolumeEncryptionEnabled = value;
    }
    /**
     * @stability stable
     */
    resetUserVolumeEncryptionEnabled() {
        this._userVolumeEncryptionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userVolumeEncryptionEnabledInput() {
        return this._userVolumeEncryptionEnabled;
    }
    /**
     * @stability stable
     */
    get volumeEncryptionKey() {
        return this.getStringAttribute('volume_encryption_key');
    }
    /**
     * @stability stable
     */
    set volumeEncryptionKey(value) {
        this._volumeEncryptionKey = value;
    }
    /**
     * @stability stable
     */
    resetVolumeEncryptionKey() {
        this._volumeEncryptionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get volumeEncryptionKeyInput() {
        return this._volumeEncryptionKey;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    /**
     * @stability stable
     */
    get workspaceProperties() {
        return this._workspaceProperties;
    }
    /**
     * @stability stable
     */
    putWorkspaceProperties(value) {
        this._workspaceProperties.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetWorkspaceProperties() {
        this._workspaceProperties.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get workspacePropertiesInput() {
        return this._workspaceProperties.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bundle_id: cdktf.stringToTerraform(this._bundleId),
            directory_id: cdktf.stringToTerraform(this._directoryId),
            root_volume_encryption_enabled: cdktf.booleanToTerraform(this._rootVolumeEncryptionEnabled),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            user_name: cdktf.stringToTerraform(this._userName),
            user_volume_encryption_enabled: cdktf.booleanToTerraform(this._userVolumeEncryptionEnabled),
            volume_encryption_key: cdktf.stringToTerraform(this._volumeEncryptionKey),
            timeouts: workspacesWorkspaceTimeoutsToTerraform(this._timeouts.internalValue),
            workspace_properties: workspacesWorkspaceWorkspacePropertiesToTerraform(this._workspaceProperties.internalValue),
        };
    }
}
exports.WorkspacesWorkspace = WorkspacesWorkspace;
_c = JSII_RTTI_SYMBOL_1;
WorkspacesWorkspace[_c] = { fqn: "@cdktf/provider-aws.workspaces.WorkspacesWorkspace", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WorkspacesWorkspace.tfResourceType = "aws_workspaces_workspace";
