// @cdktf/provider-aws/lib/backup/backup-plan.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupPlan = exports.backupPlanRuleToTerraform = exports.BackupPlanRuleLifecycleOutputReference = exports.backupPlanRuleLifecycleToTerraform = exports.backupPlanRuleCopyActionToTerraform = exports.BackupPlanRuleCopyActionLifecycleOutputReference = exports.backupPlanRuleCopyActionLifecycleToTerraform = exports.backupPlanAdvancedBackupSettingToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function backupPlanAdvancedBackupSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backup_options: cdktf.hashMapper(cdktf.anyToTerraform)(struct.backupOptions),
        resource_type: cdktf.stringToTerraform(struct.resourceType),
    };
}
exports.backupPlanAdvancedBackupSettingToTerraform = backupPlanAdvancedBackupSettingToTerraform;
function backupPlanRuleCopyActionLifecycleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cold_storage_after: cdktf.numberToTerraform(struct.coldStorageAfter),
        delete_after: cdktf.numberToTerraform(struct.deleteAfter),
    };
}
exports.backupPlanRuleCopyActionLifecycleToTerraform = backupPlanRuleCopyActionLifecycleToTerraform;
/**
 * @stability stable
 */
class BackupPlanRuleCopyActionLifecycleOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coldStorageAfter) {
            hasAnyValues = true;
            internalValueResult.coldStorageAfter = this._coldStorageAfter;
        }
        if (this._deleteAfter) {
            hasAnyValues = true;
            internalValueResult.deleteAfter = this._deleteAfter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coldStorageAfter = undefined;
            this._deleteAfter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coldStorageAfter = value.coldStorageAfter;
            this._deleteAfter = value.deleteAfter;
        }
    }
    /**
     * @stability stable
     */
    get coldStorageAfter() {
        return this.getNumberAttribute('cold_storage_after');
    }
    /**
     * @stability stable
     */
    set coldStorageAfter(value) {
        this._coldStorageAfter = value;
    }
    /**
     * @stability stable
     */
    resetColdStorageAfter() {
        this._coldStorageAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get coldStorageAfterInput() {
        return this._coldStorageAfter;
    }
    /**
     * @stability stable
     */
    get deleteAfter() {
        return this.getNumberAttribute('delete_after');
    }
    /**
     * @stability stable
     */
    set deleteAfter(value) {
        this._deleteAfter = value;
    }
    /**
     * @stability stable
     */
    resetDeleteAfter() {
        this._deleteAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteAfterInput() {
        return this._deleteAfter;
    }
}
exports.BackupPlanRuleCopyActionLifecycleOutputReference = BackupPlanRuleCopyActionLifecycleOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BackupPlanRuleCopyActionLifecycleOutputReference[_a] = { fqn: "@cdktf/provider-aws.backup.BackupPlanRuleCopyActionLifecycleOutputReference", version: "3.0.1" };
function backupPlanRuleCopyActionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        destination_vault_arn: cdktf.stringToTerraform(struct.destinationVaultArn),
        lifecycle: backupPlanRuleCopyActionLifecycleToTerraform(struct.lifecycle),
    };
}
exports.backupPlanRuleCopyActionToTerraform = backupPlanRuleCopyActionToTerraform;
function backupPlanRuleLifecycleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cold_storage_after: cdktf.numberToTerraform(struct.coldStorageAfter),
        delete_after: cdktf.numberToTerraform(struct.deleteAfter),
    };
}
exports.backupPlanRuleLifecycleToTerraform = backupPlanRuleLifecycleToTerraform;
/**
 * @stability stable
 */
class BackupPlanRuleLifecycleOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coldStorageAfter) {
            hasAnyValues = true;
            internalValueResult.coldStorageAfter = this._coldStorageAfter;
        }
        if (this._deleteAfter) {
            hasAnyValues = true;
            internalValueResult.deleteAfter = this._deleteAfter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coldStorageAfter = undefined;
            this._deleteAfter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coldStorageAfter = value.coldStorageAfter;
            this._deleteAfter = value.deleteAfter;
        }
    }
    /**
     * @stability stable
     */
    get coldStorageAfter() {
        return this.getNumberAttribute('cold_storage_after');
    }
    /**
     * @stability stable
     */
    set coldStorageAfter(value) {
        this._coldStorageAfter = value;
    }
    /**
     * @stability stable
     */
    resetColdStorageAfter() {
        this._coldStorageAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get coldStorageAfterInput() {
        return this._coldStorageAfter;
    }
    /**
     * @stability stable
     */
    get deleteAfter() {
        return this.getNumberAttribute('delete_after');
    }
    /**
     * @stability stable
     */
    set deleteAfter(value) {
        this._deleteAfter = value;
    }
    /**
     * @stability stable
     */
    resetDeleteAfter() {
        this._deleteAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteAfterInput() {
        return this._deleteAfter;
    }
}
exports.BackupPlanRuleLifecycleOutputReference = BackupPlanRuleLifecycleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BackupPlanRuleLifecycleOutputReference[_b] = { fqn: "@cdktf/provider-aws.backup.BackupPlanRuleLifecycleOutputReference", version: "3.0.1" };
function backupPlanRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        completion_window: cdktf.numberToTerraform(struct.completionWindow),
        enable_continuous_backup: cdktf.booleanToTerraform(struct.enableContinuousBackup),
        recovery_point_tags: cdktf.hashMapper(cdktf.anyToTerraform)(struct.recoveryPointTags),
        rule_name: cdktf.stringToTerraform(struct.ruleName),
        schedule: cdktf.stringToTerraform(struct.schedule),
        start_window: cdktf.numberToTerraform(struct.startWindow),
        target_vault_name: cdktf.stringToTerraform(struct.targetVaultName),
        copy_action: cdktf.listMapper(backupPlanRuleCopyActionToTerraform)(struct.copyAction),
        lifecycle: backupPlanRuleLifecycleToTerraform(struct.lifecycle),
    };
}
exports.backupPlanRuleToTerraform = backupPlanRuleToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/backup_plan.html aws_backup_plan}.
 *
 * @stability stable
 */
class BackupPlan extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/backup_plan.html aws_backup_plan} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_plan',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._advancedBackupSetting = config.advancedBackupSetting;
        this._rule = config.rule;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    get advancedBackupSetting() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('advanced_backup_setting');
    }
    /**
     * @stability stable
     */
    set advancedBackupSetting(value) {
        this._advancedBackupSetting = value;
    }
    /**
     * @stability stable
     */
    resetAdvancedBackupSetting() {
        this._advancedBackupSetting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get advancedBackupSettingInput() {
        return this._advancedBackupSetting;
    }
    /**
     * @stability stable
     */
    get rule() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('rule');
    }
    /**
     * @stability stable
     */
    set rule(value) {
        this._rule = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ruleInput() {
        return this._rule;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            advanced_backup_setting: cdktf.listMapper(backupPlanAdvancedBackupSettingToTerraform)(this._advancedBackupSetting),
            rule: cdktf.listMapper(backupPlanRuleToTerraform)(this._rule),
        };
    }
}
exports.BackupPlan = BackupPlan;
_c = JSII_RTTI_SYMBOL_1;
BackupPlan[_c] = { fqn: "@cdktf/provider-aws.backup.BackupPlan", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
BackupPlan.tfResourceType = "aws_backup_plan";
