// @cdktf/provider-aws/lib/backup/backup-selection.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupSelection = exports.backupSelectionSelectionTagToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function backupSelectionSelectionTagToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.backupSelectionSelectionTagToTerraform = backupSelectionSelectionTagToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/backup_selection.html aws_backup_selection}.
 *
 * @stability stable
 */
class BackupSelection extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/backup_selection.html aws_backup_selection} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_selection',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._iamRoleArn = config.iamRoleArn;
        this._name = config.name;
        this._planId = config.planId;
        this._resources = config.resources;
        this._selectionTag = config.selectionTag;
    }
    /**
     * @stability stable
     */
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    /**
     * @stability stable
     */
    set iamRoleArn(value) {
        this._iamRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamRoleArnInput() {
        return this._iamRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get planId() {
        return this.getStringAttribute('plan_id');
    }
    /**
     * @stability stable
     */
    set planId(value) {
        this._planId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get planIdInput() {
        return this._planId;
    }
    /**
     * @stability stable
     */
    get resources() {
        return this.getListAttribute('resources');
    }
    /**
     * @stability stable
     */
    set resources(value) {
        this._resources = value;
    }
    /**
     * @stability stable
     */
    resetResources() {
        this._resources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourcesInput() {
        return this._resources;
    }
    /**
     * @stability stable
     */
    get selectionTag() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('selection_tag');
    }
    /**
     * @stability stable
     */
    set selectionTag(value) {
        this._selectionTag = value;
    }
    /**
     * @stability stable
     */
    resetSelectionTag() {
        this._selectionTag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get selectionTagInput() {
        return this._selectionTag;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            iam_role_arn: cdktf.stringToTerraform(this._iamRoleArn),
            name: cdktf.stringToTerraform(this._name),
            plan_id: cdktf.stringToTerraform(this._planId),
            resources: cdktf.listMapper(cdktf.stringToTerraform)(this._resources),
            selection_tag: cdktf.listMapper(backupSelectionSelectionTagToTerraform)(this._selectionTag),
        };
    }
}
exports.BackupSelection = BackupSelection;
_a = JSII_RTTI_SYMBOL_1;
BackupSelection[_a] = { fqn: "@cdktf/provider-aws.backup.BackupSelection", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
BackupSelection.tfResourceType = "aws_backup_selection";
