// @cdktf/provider-aws/lib/backup/data-aws-backup-selection.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBackupSelection = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/backup_selection.html aws_backup_selection}.
 *
 * @stability stable
 */
class DataAwsBackupSelection extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/backup_selection.html aws_backup_selection} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_selection',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._planId = config.planId;
        this._selectionId = config.selectionId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // iam_role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    get planId() {
        return this.getStringAttribute('plan_id');
    }
    /**
     * @stability stable
     */
    set planId(value) {
        this._planId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get planIdInput() {
        return this._planId;
    }
    // resources - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get resources() {
        return this.getListAttribute('resources');
    }
    /**
     * @stability stable
     */
    get selectionId() {
        return this.getStringAttribute('selection_id');
    }
    /**
     * @stability stable
     */
    set selectionId(value) {
        this._selectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get selectionIdInput() {
        return this._selectionId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            plan_id: cdktf.stringToTerraform(this._planId),
            selection_id: cdktf.stringToTerraform(this._selectionId),
        };
    }
}
exports.DataAwsBackupSelection = DataAwsBackupSelection;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBackupSelection[_a] = { fqn: "@cdktf/provider-aws.backup.DataAwsBackupSelection", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsBackupSelection.tfResourceType = "aws_backup_selection";
