// @cdktf/provider-aws/lib/gamelift/gamelift-fleet.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameliftFleet = exports.GameliftFleetTimeoutsOutputReference = exports.gameliftFleetTimeoutsToTerraform = exports.GameliftFleetRuntimeConfigurationOutputReference = exports.gameliftFleetRuntimeConfigurationToTerraform = exports.gameliftFleetRuntimeConfigurationServerProcessToTerraform = exports.GameliftFleetResourceCreationLimitPolicyOutputReference = exports.gameliftFleetResourceCreationLimitPolicyToTerraform = exports.gameliftFleetEc2InboundPermissionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gameliftFleetEc2InboundPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        from_port: cdktf.numberToTerraform(struct.fromPort),
        ip_range: cdktf.stringToTerraform(struct.ipRange),
        protocol: cdktf.stringToTerraform(struct.protocol),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.gameliftFleetEc2InboundPermissionToTerraform = gameliftFleetEc2InboundPermissionToTerraform;
function gameliftFleetResourceCreationLimitPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        new_game_sessions_per_creator: cdktf.numberToTerraform(struct.newGameSessionsPerCreator),
        policy_period_in_minutes: cdktf.numberToTerraform(struct.policyPeriodInMinutes),
    };
}
exports.gameliftFleetResourceCreationLimitPolicyToTerraform = gameliftFleetResourceCreationLimitPolicyToTerraform;
/**
 * @stability stable
 */
class GameliftFleetResourceCreationLimitPolicyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._newGameSessionsPerCreator) {
            hasAnyValues = true;
            internalValueResult.newGameSessionsPerCreator = this._newGameSessionsPerCreator;
        }
        if (this._policyPeriodInMinutes) {
            hasAnyValues = true;
            internalValueResult.policyPeriodInMinutes = this._policyPeriodInMinutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._newGameSessionsPerCreator = undefined;
            this._policyPeriodInMinutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._newGameSessionsPerCreator = value.newGameSessionsPerCreator;
            this._policyPeriodInMinutes = value.policyPeriodInMinutes;
        }
    }
    /**
     * @stability stable
     */
    get newGameSessionsPerCreator() {
        return this.getNumberAttribute('new_game_sessions_per_creator');
    }
    /**
     * @stability stable
     */
    set newGameSessionsPerCreator(value) {
        this._newGameSessionsPerCreator = value;
    }
    /**
     * @stability stable
     */
    resetNewGameSessionsPerCreator() {
        this._newGameSessionsPerCreator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get newGameSessionsPerCreatorInput() {
        return this._newGameSessionsPerCreator;
    }
    /**
     * @stability stable
     */
    get policyPeriodInMinutes() {
        return this.getNumberAttribute('policy_period_in_minutes');
    }
    /**
     * @stability stable
     */
    set policyPeriodInMinutes(value) {
        this._policyPeriodInMinutes = value;
    }
    /**
     * @stability stable
     */
    resetPolicyPeriodInMinutes() {
        this._policyPeriodInMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyPeriodInMinutesInput() {
        return this._policyPeriodInMinutes;
    }
}
exports.GameliftFleetResourceCreationLimitPolicyOutputReference = GameliftFleetResourceCreationLimitPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GameliftFleetResourceCreationLimitPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.gamelift.GameliftFleetResourceCreationLimitPolicyOutputReference", version: "3.0.1" };
function gameliftFleetRuntimeConfigurationServerProcessToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        concurrent_executions: cdktf.numberToTerraform(struct.concurrentExecutions),
        launch_path: cdktf.stringToTerraform(struct.launchPath),
        parameters: cdktf.stringToTerraform(struct.parameters),
    };
}
exports.gameliftFleetRuntimeConfigurationServerProcessToTerraform = gameliftFleetRuntimeConfigurationServerProcessToTerraform;
function gameliftFleetRuntimeConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        game_session_activation_timeout_seconds: cdktf.numberToTerraform(struct.gameSessionActivationTimeoutSeconds),
        max_concurrent_game_session_activations: cdktf.numberToTerraform(struct.maxConcurrentGameSessionActivations),
        server_process: cdktf.listMapper(gameliftFleetRuntimeConfigurationServerProcessToTerraform)(struct.serverProcess),
    };
}
exports.gameliftFleetRuntimeConfigurationToTerraform = gameliftFleetRuntimeConfigurationToTerraform;
/**
 * @stability stable
 */
class GameliftFleetRuntimeConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gameSessionActivationTimeoutSeconds) {
            hasAnyValues = true;
            internalValueResult.gameSessionActivationTimeoutSeconds = this._gameSessionActivationTimeoutSeconds;
        }
        if (this._maxConcurrentGameSessionActivations) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentGameSessionActivations = this._maxConcurrentGameSessionActivations;
        }
        if (this._serverProcess) {
            hasAnyValues = true;
            internalValueResult.serverProcess = this._serverProcess;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gameSessionActivationTimeoutSeconds = undefined;
            this._maxConcurrentGameSessionActivations = undefined;
            this._serverProcess = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gameSessionActivationTimeoutSeconds = value.gameSessionActivationTimeoutSeconds;
            this._maxConcurrentGameSessionActivations = value.maxConcurrentGameSessionActivations;
            this._serverProcess = value.serverProcess;
        }
    }
    /**
     * @stability stable
     */
    get gameSessionActivationTimeoutSeconds() {
        return this.getNumberAttribute('game_session_activation_timeout_seconds');
    }
    /**
     * @stability stable
     */
    set gameSessionActivationTimeoutSeconds(value) {
        this._gameSessionActivationTimeoutSeconds = value;
    }
    /**
     * @stability stable
     */
    resetGameSessionActivationTimeoutSeconds() {
        this._gameSessionActivationTimeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gameSessionActivationTimeoutSecondsInput() {
        return this._gameSessionActivationTimeoutSeconds;
    }
    /**
     * @stability stable
     */
    get maxConcurrentGameSessionActivations() {
        return this.getNumberAttribute('max_concurrent_game_session_activations');
    }
    /**
     * @stability stable
     */
    set maxConcurrentGameSessionActivations(value) {
        this._maxConcurrentGameSessionActivations = value;
    }
    /**
     * @stability stable
     */
    resetMaxConcurrentGameSessionActivations() {
        this._maxConcurrentGameSessionActivations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrentGameSessionActivationsInput() {
        return this._maxConcurrentGameSessionActivations;
    }
    /**
     * @stability stable
     */
    get serverProcess() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('server_process');
    }
    /**
     * @stability stable
     */
    set serverProcess(value) {
        this._serverProcess = value;
    }
    /**
     * @stability stable
     */
    resetServerProcess() {
        this._serverProcess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serverProcessInput() {
        return this._serverProcess;
    }
}
exports.GameliftFleetRuntimeConfigurationOutputReference = GameliftFleetRuntimeConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GameliftFleetRuntimeConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.gamelift.GameliftFleetRuntimeConfigurationOutputReference", version: "3.0.1" };
function gameliftFleetTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.gameliftFleetTimeoutsToTerraform = gameliftFleetTimeoutsToTerraform;
/**
 * @stability stable
 */
class GameliftFleetTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.GameliftFleetTimeoutsOutputReference = GameliftFleetTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GameliftFleetTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.gamelift.GameliftFleetTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/gamelift_fleet.html aws_gamelift_fleet}.
 *
 * @stability stable
 */
class GameliftFleet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/gamelift_fleet.html aws_gamelift_fleet} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_gamelift_fleet',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // resource_creation_limit_policy - computed: false, optional: true, required: false
        this._resourceCreationLimitPolicy = new GameliftFleetResourceCreationLimitPolicyOutputReference(this, "resource_creation_limit_policy", true);
        // runtime_configuration - computed: false, optional: true, required: false
        this._runtimeConfiguration = new GameliftFleetRuntimeConfigurationOutputReference(this, "runtime_configuration", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GameliftFleetTimeoutsOutputReference(this, "timeouts", true);
        this._buildId = config.buildId;
        this._description = config.description;
        this._ec2InstanceType = config.ec2InstanceType;
        this._fleetType = config.fleetType;
        this._instanceRoleArn = config.instanceRoleArn;
        this._metricGroups = config.metricGroups;
        this._name = config.name;
        this._newGameSessionProtectionPolicy = config.newGameSessionProtectionPolicy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._ec2InboundPermission = config.ec2InboundPermission;
        this._resourceCreationLimitPolicy.internalValue = config.resourceCreationLimitPolicy;
        this._runtimeConfiguration.internalValue = config.runtimeConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get buildId() {
        return this.getStringAttribute('build_id');
    }
    /**
     * @stability stable
     */
    set buildId(value) {
        this._buildId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildIdInput() {
        return this._buildId;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get ec2InstanceType() {
        return this.getStringAttribute('ec2_instance_type');
    }
    /**
     * @stability stable
     */
    set ec2InstanceType(value) {
        this._ec2InstanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ec2InstanceTypeInput() {
        return this._ec2InstanceType;
    }
    /**
     * @stability stable
     */
    get fleetType() {
        return this.getStringAttribute('fleet_type');
    }
    /**
     * @stability stable
     */
    set fleetType(value) {
        this._fleetType = value;
    }
    /**
     * @stability stable
     */
    resetFleetType() {
        this._fleetType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fleetTypeInput() {
        return this._fleetType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceRoleArn() {
        return this.getStringAttribute('instance_role_arn');
    }
    /**
     * @stability stable
     */
    set instanceRoleArn(value) {
        this._instanceRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetInstanceRoleArn() {
        this._instanceRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceRoleArnInput() {
        return this._instanceRoleArn;
    }
    // log_paths - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get logPaths() {
        return this.getListAttribute('log_paths');
    }
    /**
     * @stability stable
     */
    get metricGroups() {
        return this.getListAttribute('metric_groups');
    }
    /**
     * @stability stable
     */
    set metricGroups(value) {
        this._metricGroups = value;
    }
    /**
     * @stability stable
     */
    resetMetricGroups() {
        this._metricGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricGroupsInput() {
        return this._metricGroups;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get newGameSessionProtectionPolicy() {
        return this.getStringAttribute('new_game_session_protection_policy');
    }
    /**
     * @stability stable
     */
    set newGameSessionProtectionPolicy(value) {
        this._newGameSessionProtectionPolicy = value;
    }
    /**
     * @stability stable
     */
    resetNewGameSessionProtectionPolicy() {
        this._newGameSessionProtectionPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get newGameSessionProtectionPolicyInput() {
        return this._newGameSessionProtectionPolicy;
    }
    // operating_system - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get operatingSystem() {
        return this.getStringAttribute('operating_system');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get ec2InboundPermission() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ec2_inbound_permission');
    }
    /**
     * @stability stable
     */
    set ec2InboundPermission(value) {
        this._ec2InboundPermission = value;
    }
    /**
     * @stability stable
     */
    resetEc2InboundPermission() {
        this._ec2InboundPermission = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ec2InboundPermissionInput() {
        return this._ec2InboundPermission;
    }
    /**
     * @stability stable
     */
    get resourceCreationLimitPolicy() {
        return this._resourceCreationLimitPolicy;
    }
    /**
     * @stability stable
     */
    putResourceCreationLimitPolicy(value) {
        this._resourceCreationLimitPolicy.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetResourceCreationLimitPolicy() {
        this._resourceCreationLimitPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceCreationLimitPolicyInput() {
        return this._resourceCreationLimitPolicy.internalValue;
    }
    /**
     * @stability stable
     */
    get runtimeConfiguration() {
        return this._runtimeConfiguration;
    }
    /**
     * @stability stable
     */
    putRuntimeConfiguration(value) {
        this._runtimeConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRuntimeConfiguration() {
        this._runtimeConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runtimeConfigurationInput() {
        return this._runtimeConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            build_id: cdktf.stringToTerraform(this._buildId),
            description: cdktf.stringToTerraform(this._description),
            ec2_instance_type: cdktf.stringToTerraform(this._ec2InstanceType),
            fleet_type: cdktf.stringToTerraform(this._fleetType),
            instance_role_arn: cdktf.stringToTerraform(this._instanceRoleArn),
            metric_groups: cdktf.listMapper(cdktf.stringToTerraform)(this._metricGroups),
            name: cdktf.stringToTerraform(this._name),
            new_game_session_protection_policy: cdktf.stringToTerraform(this._newGameSessionProtectionPolicy),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            ec2_inbound_permission: cdktf.listMapper(gameliftFleetEc2InboundPermissionToTerraform)(this._ec2InboundPermission),
            resource_creation_limit_policy: gameliftFleetResourceCreationLimitPolicyToTerraform(this._resourceCreationLimitPolicy.internalValue),
            runtime_configuration: gameliftFleetRuntimeConfigurationToTerraform(this._runtimeConfiguration.internalValue),
            timeouts: gameliftFleetTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GameliftFleet = GameliftFleet;
_d = JSII_RTTI_SYMBOL_1;
GameliftFleet[_d] = { fqn: "@cdktf/provider-aws.gamelift.GameliftFleet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GameliftFleet.tfResourceType = "aws_gamelift_fleet";
