// @cdktf/provider-aws/lib/cloudformation/cloudformation-type.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudformationType = exports.CloudformationTypeLoggingConfigOutputReference = exports.cloudformationTypeLoggingConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudformationTypeLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_group_name: cdktf.stringToTerraform(struct.logGroupName),
        log_role_arn: cdktf.stringToTerraform(struct.logRoleArn),
    };
}
exports.cloudformationTypeLoggingConfigToTerraform = cloudformationTypeLoggingConfigToTerraform;
/**
 * @stability stable
 */
class CloudformationTypeLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logGroupName) {
            hasAnyValues = true;
            internalValueResult.logGroupName = this._logGroupName;
        }
        if (this._logRoleArn) {
            hasAnyValues = true;
            internalValueResult.logRoleArn = this._logRoleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logGroupName = undefined;
            this._logRoleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logGroupName = value.logGroupName;
            this._logRoleArn = value.logRoleArn;
        }
    }
    /**
     * @stability stable
     */
    get logGroupName() {
        return this.getStringAttribute('log_group_name');
    }
    /**
     * @stability stable
     */
    set logGroupName(value) {
        this._logGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logGroupNameInput() {
        return this._logGroupName;
    }
    /**
     * @stability stable
     */
    get logRoleArn() {
        return this.getStringAttribute('log_role_arn');
    }
    /**
     * @stability stable
     */
    set logRoleArn(value) {
        this._logRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logRoleArnInput() {
        return this._logRoleArn;
    }
}
exports.CloudformationTypeLoggingConfigOutputReference = CloudformationTypeLoggingConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudformationTypeLoggingConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudformation.CloudformationTypeLoggingConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_type.html aws_cloudformation_type}.
 *
 * @stability stable
 */
class CloudformationType extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_type.html aws_cloudformation_type} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudformation_type',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // logging_config - computed: false, optional: true, required: false
        this._loggingConfig = new CloudformationTypeLoggingConfigOutputReference(this, "logging_config", true);
        this._executionRoleArn = config.executionRoleArn;
        this._schemaHandlerPackage = config.schemaHandlerPackage;
        this._type = config.type;
        this._typeName = config.typeName;
        this._loggingConfig.internalValue = config.loggingConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_version_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultVersionId() {
        return this.getStringAttribute('default_version_id');
    }
    // deprecated_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deprecatedStatus() {
        return this.getStringAttribute('deprecated_status');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // documentation_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get documentationUrl() {
        return this.getStringAttribute('documentation_url');
    }
    /**
     * @stability stable
     */
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    /**
     * @stability stable
     */
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetExecutionRoleArn() {
        this._executionRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // is_default_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get isDefaultVersion() {
        return this.getBooleanAttribute('is_default_version');
    }
    // provisioning_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get provisioningType() {
        return this.getStringAttribute('provisioning_type');
    }
    // schema - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get schema() {
        return this.getStringAttribute('schema');
    }
    /**
     * @stability stable
     */
    get schemaHandlerPackage() {
        return this.getStringAttribute('schema_handler_package');
    }
    /**
     * @stability stable
     */
    set schemaHandlerPackage(value) {
        this._schemaHandlerPackage = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaHandlerPackageInput() {
        return this._schemaHandlerPackage;
    }
    // source_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceUrl() {
        return this.getStringAttribute('source_url');
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    // type_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get typeArn() {
        return this.getStringAttribute('type_arn');
    }
    /**
     * @stability stable
     */
    get typeName() {
        return this.getStringAttribute('type_name');
    }
    /**
     * @stability stable
     */
    set typeName(value) {
        this._typeName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeNameInput() {
        return this._typeName;
    }
    // version_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    // visibility - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get visibility() {
        return this.getStringAttribute('visibility');
    }
    /**
     * @stability stable
     */
    get loggingConfig() {
        return this._loggingConfig;
    }
    /**
     * @stability stable
     */
    putLoggingConfig(value) {
        this._loggingConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLoggingConfig() {
        this._loggingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingConfigInput() {
        return this._loggingConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            schema_handler_package: cdktf.stringToTerraform(this._schemaHandlerPackage),
            type: cdktf.stringToTerraform(this._type),
            type_name: cdktf.stringToTerraform(this._typeName),
            logging_config: cloudformationTypeLoggingConfigToTerraform(this._loggingConfig.internalValue),
        };
    }
}
exports.CloudformationType = CloudformationType;
_b = JSII_RTTI_SYMBOL_1;
CloudformationType[_b] = { fqn: "@cdktf/provider-aws.cloudformation.CloudformationType", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudformationType.tfResourceType = "aws_cloudformation_type";
