// @cdktf/provider-aws/lib/macie2/macie2-account.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Macie2Account = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/macie2_account.html aws_macie2_account}.
 *
 * @stability stable
 */
class Macie2Account extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/macie2_account.html aws_macie2_account} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_macie2_account',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._findingPublishingFrequency = config.findingPublishingFrequency;
        this._status = config.status;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    /**
     * @stability stable
     */
    get findingPublishingFrequency() {
        return this.getStringAttribute('finding_publishing_frequency');
    }
    /**
     * @stability stable
     */
    set findingPublishingFrequency(value) {
        this._findingPublishingFrequency = value;
    }
    /**
     * @stability stable
     */
    resetFindingPublishingFrequency() {
        this._findingPublishingFrequency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get findingPublishingFrequencyInput() {
        return this._findingPublishingFrequency;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // service_role - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    // updated_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get updatedAt() {
        return this.getStringAttribute('updated_at');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            finding_publishing_frequency: cdktf.stringToTerraform(this._findingPublishingFrequency),
            status: cdktf.stringToTerraform(this._status),
        };
    }
}
exports.Macie2Account = Macie2Account;
_a = JSII_RTTI_SYMBOL_1;
Macie2Account[_a] = { fqn: "@cdktf/provider-aws.macie2.Macie2Account", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Macie2Account.tfResourceType = "aws_macie2_account";
