// @cdktf/provider-aws/lib/acm/acmpca-certificate-authority.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AcmpcaCertificateAuthority = exports.AcmpcaCertificateAuthorityTimeoutsOutputReference = exports.acmpcaCertificateAuthorityTimeoutsToTerraform = exports.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference = exports.acmpcaCertificateAuthorityRevocationConfigurationToTerraform = exports.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference = exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform = exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference = exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        common_name: cdktf.stringToTerraform(struct.commonName),
        country: cdktf.stringToTerraform(struct.country),
        distinguished_name_qualifier: cdktf.stringToTerraform(struct.distinguishedNameQualifier),
        generation_qualifier: cdktf.stringToTerraform(struct.generationQualifier),
        given_name: cdktf.stringToTerraform(struct.givenName),
        initials: cdktf.stringToTerraform(struct.initials),
        locality: cdktf.stringToTerraform(struct.locality),
        organization: cdktf.stringToTerraform(struct.organization),
        organizational_unit: cdktf.stringToTerraform(struct.organizationalUnit),
        pseudonym: cdktf.stringToTerraform(struct.pseudonym),
        state: cdktf.stringToTerraform(struct.state),
        surname: cdktf.stringToTerraform(struct.surname),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform;
/**
 * @stability stable
 */
class AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._commonName) {
            hasAnyValues = true;
            internalValueResult.commonName = this._commonName;
        }
        if (this._country) {
            hasAnyValues = true;
            internalValueResult.country = this._country;
        }
        if (this._distinguishedNameQualifier) {
            hasAnyValues = true;
            internalValueResult.distinguishedNameQualifier = this._distinguishedNameQualifier;
        }
        if (this._generationQualifier) {
            hasAnyValues = true;
            internalValueResult.generationQualifier = this._generationQualifier;
        }
        if (this._givenName) {
            hasAnyValues = true;
            internalValueResult.givenName = this._givenName;
        }
        if (this._initials) {
            hasAnyValues = true;
            internalValueResult.initials = this._initials;
        }
        if (this._locality) {
            hasAnyValues = true;
            internalValueResult.locality = this._locality;
        }
        if (this._organization) {
            hasAnyValues = true;
            internalValueResult.organization = this._organization;
        }
        if (this._organizationalUnit) {
            hasAnyValues = true;
            internalValueResult.organizationalUnit = this._organizationalUnit;
        }
        if (this._pseudonym) {
            hasAnyValues = true;
            internalValueResult.pseudonym = this._pseudonym;
        }
        if (this._state) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        if (this._surname) {
            hasAnyValues = true;
            internalValueResult.surname = this._surname;
        }
        if (this._title) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._commonName = undefined;
            this._country = undefined;
            this._distinguishedNameQualifier = undefined;
            this._generationQualifier = undefined;
            this._givenName = undefined;
            this._initials = undefined;
            this._locality = undefined;
            this._organization = undefined;
            this._organizationalUnit = undefined;
            this._pseudonym = undefined;
            this._state = undefined;
            this._surname = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._commonName = value.commonName;
            this._country = value.country;
            this._distinguishedNameQualifier = value.distinguishedNameQualifier;
            this._generationQualifier = value.generationQualifier;
            this._givenName = value.givenName;
            this._initials = value.initials;
            this._locality = value.locality;
            this._organization = value.organization;
            this._organizationalUnit = value.organizationalUnit;
            this._pseudonym = value.pseudonym;
            this._state = value.state;
            this._surname = value.surname;
            this._title = value.title;
        }
    }
    /**
     * @stability stable
     */
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    /**
     * @stability stable
     */
    set commonName(value) {
        this._commonName = value;
    }
    /**
     * @stability stable
     */
    resetCommonName() {
        this._commonName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commonNameInput() {
        return this._commonName;
    }
    /**
     * @stability stable
     */
    get country() {
        return this.getStringAttribute('country');
    }
    /**
     * @stability stable
     */
    set country(value) {
        this._country = value;
    }
    /**
     * @stability stable
     */
    resetCountry() {
        this._country = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get countryInput() {
        return this._country;
    }
    /**
     * @stability stable
     */
    get distinguishedNameQualifier() {
        return this.getStringAttribute('distinguished_name_qualifier');
    }
    /**
     * @stability stable
     */
    set distinguishedNameQualifier(value) {
        this._distinguishedNameQualifier = value;
    }
    /**
     * @stability stable
     */
    resetDistinguishedNameQualifier() {
        this._distinguishedNameQualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get distinguishedNameQualifierInput() {
        return this._distinguishedNameQualifier;
    }
    /**
     * @stability stable
     */
    get generationQualifier() {
        return this.getStringAttribute('generation_qualifier');
    }
    /**
     * @stability stable
     */
    set generationQualifier(value) {
        this._generationQualifier = value;
    }
    /**
     * @stability stable
     */
    resetGenerationQualifier() {
        this._generationQualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get generationQualifierInput() {
        return this._generationQualifier;
    }
    /**
     * @stability stable
     */
    get givenName() {
        return this.getStringAttribute('given_name');
    }
    /**
     * @stability stable
     */
    set givenName(value) {
        this._givenName = value;
    }
    /**
     * @stability stable
     */
    resetGivenName() {
        this._givenName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get givenNameInput() {
        return this._givenName;
    }
    /**
     * @stability stable
     */
    get initials() {
        return this.getStringAttribute('initials');
    }
    /**
     * @stability stable
     */
    set initials(value) {
        this._initials = value;
    }
    /**
     * @stability stable
     */
    resetInitials() {
        this._initials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get initialsInput() {
        return this._initials;
    }
    /**
     * @stability stable
     */
    get locality() {
        return this.getStringAttribute('locality');
    }
    /**
     * @stability stable
     */
    set locality(value) {
        this._locality = value;
    }
    /**
     * @stability stable
     */
    resetLocality() {
        this._locality = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get localityInput() {
        return this._locality;
    }
    /**
     * @stability stable
     */
    get organization() {
        return this.getStringAttribute('organization');
    }
    /**
     * @stability stable
     */
    set organization(value) {
        this._organization = value;
    }
    /**
     * @stability stable
     */
    resetOrganization() {
        this._organization = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get organizationInput() {
        return this._organization;
    }
    /**
     * @stability stable
     */
    get organizationalUnit() {
        return this.getStringAttribute('organizational_unit');
    }
    /**
     * @stability stable
     */
    set organizationalUnit(value) {
        this._organizationalUnit = value;
    }
    /**
     * @stability stable
     */
    resetOrganizationalUnit() {
        this._organizationalUnit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get organizationalUnitInput() {
        return this._organizationalUnit;
    }
    /**
     * @stability stable
     */
    get pseudonym() {
        return this.getStringAttribute('pseudonym');
    }
    /**
     * @stability stable
     */
    set pseudonym(value) {
        this._pseudonym = value;
    }
    /**
     * @stability stable
     */
    resetPseudonym() {
        this._pseudonym = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pseudonymInput() {
        return this._pseudonym;
    }
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    set state(value) {
        this._state = value;
    }
    /**
     * @stability stable
     */
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateInput() {
        return this._state;
    }
    /**
     * @stability stable
     */
    get surname() {
        return this.getStringAttribute('surname');
    }
    /**
     * @stability stable
     */
    set surname(value) {
        this._surname = value;
    }
    /**
     * @stability stable
     */
    resetSurname() {
        this._surname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get surnameInput() {
        return this._surname;
    }
    /**
     * @stability stable
     */
    get title() {
        return this.getStringAttribute('title');
    }
    /**
     * @stability stable
     */
    set title(value) {
        this._title = value;
    }
    /**
     * @stability stable
     */
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get titleInput() {
        return this._title;
    }
}
exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference = AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference[_a] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference", version: "3.0.1" };
function acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_algorithm: cdktf.stringToTerraform(struct.keyAlgorithm),
        signing_algorithm: cdktf.stringToTerraform(struct.signingAlgorithm),
        subject: acmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectToTerraform(struct.subject),
    };
}
exports.acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform = acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform;
/**
 * @stability stable
 */
class AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // subject - computed: false, optional: false, required: true
        this._subject = new AcmpcaCertificateAuthorityCertificateAuthorityConfigurationSubjectOutputReference(this, "subject", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _g, _h;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyAlgorithm) {
            hasAnyValues = true;
            internalValueResult.keyAlgorithm = this._keyAlgorithm;
        }
        if (this._signingAlgorithm) {
            hasAnyValues = true;
            internalValueResult.signingAlgorithm = this._signingAlgorithm;
        }
        if ((_g = this._subject) === null || _g === void 0 ? void 0 : _g.internalValue) {
            hasAnyValues = true;
            internalValueResult.subject = (_h = this._subject) === null || _h === void 0 ? void 0 : _h.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyAlgorithm = undefined;
            this._signingAlgorithm = undefined;
            this._subject.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyAlgorithm = value.keyAlgorithm;
            this._signingAlgorithm = value.signingAlgorithm;
            this._subject.internalValue = value.subject;
        }
    }
    /**
     * @stability stable
     */
    get keyAlgorithm() {
        return this.getStringAttribute('key_algorithm');
    }
    /**
     * @stability stable
     */
    set keyAlgorithm(value) {
        this._keyAlgorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyAlgorithmInput() {
        return this._keyAlgorithm;
    }
    /**
     * @stability stable
     */
    get signingAlgorithm() {
        return this.getStringAttribute('signing_algorithm');
    }
    /**
     * @stability stable
     */
    set signingAlgorithm(value) {
        this._signingAlgorithm = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signingAlgorithmInput() {
        return this._signingAlgorithm;
    }
    /**
     * @stability stable
     */
    get subject() {
        return this._subject;
    }
    /**
     * @stability stable
     */
    putSubject(value) {
        this._subject.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subjectInput() {
        return this._subject.internalValue;
    }
}
exports.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference = AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference", version: "3.0.1" };
function acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_cname: cdktf.stringToTerraform(struct.customCname),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        expiration_in_days: cdktf.numberToTerraform(struct.expirationInDays),
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_object_acl: cdktf.stringToTerraform(struct.s3ObjectAcl),
    };
}
exports.acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform;
/**
 * @stability stable
 */
class AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customCname) {
            hasAnyValues = true;
            internalValueResult.customCname = this._customCname;
        }
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._expirationInDays) {
            hasAnyValues = true;
            internalValueResult.expirationInDays = this._expirationInDays;
        }
        if (this._s3BucketName) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3ObjectAcl) {
            hasAnyValues = true;
            internalValueResult.s3ObjectAcl = this._s3ObjectAcl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customCname = undefined;
            this._enabled = undefined;
            this._expirationInDays = undefined;
            this._s3BucketName = undefined;
            this._s3ObjectAcl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customCname = value.customCname;
            this._enabled = value.enabled;
            this._expirationInDays = value.expirationInDays;
            this._s3BucketName = value.s3BucketName;
            this._s3ObjectAcl = value.s3ObjectAcl;
        }
    }
    /**
     * @stability stable
     */
    get customCname() {
        return this.getStringAttribute('custom_cname');
    }
    /**
     * @stability stable
     */
    set customCname(value) {
        this._customCname = value;
    }
    /**
     * @stability stable
     */
    resetCustomCname() {
        this._customCname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customCnameInput() {
        return this._customCname;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get expirationInDays() {
        return this.getNumberAttribute('expiration_in_days');
    }
    /**
     * @stability stable
     */
    set expirationInDays(value) {
        this._expirationInDays = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get expirationInDaysInput() {
        return this._expirationInDays;
    }
    /**
     * @stability stable
     */
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    /**
     * @stability stable
     */
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    /**
     * @stability stable
     */
    resetS3BucketName() {
        this._s3BucketName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    /**
     * @stability stable
     */
    get s3ObjectAcl() {
        return this.getStringAttribute('s3_object_acl');
    }
    /**
     * @stability stable
     */
    set s3ObjectAcl(value) {
        this._s3ObjectAcl = value;
    }
    /**
     * @stability stable
     */
    resetS3ObjectAcl() {
        this._s3ObjectAcl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3ObjectAclInput() {
        return this._s3ObjectAcl;
    }
}
exports.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference = AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference", version: "3.0.1" };
function acmpcaCertificateAuthorityRevocationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crl_configuration: acmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform(struct.crlConfiguration),
    };
}
exports.acmpcaCertificateAuthorityRevocationConfigurationToTerraform = acmpcaCertificateAuthorityRevocationConfigurationToTerraform;
/**
 * @stability stable
 */
class AcmpcaCertificateAuthorityRevocationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // crl_configuration - computed: false, optional: true, required: false
        this._crlConfiguration = new AcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationOutputReference(this, "crl_configuration", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _g, _h;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_g = this._crlConfiguration) === null || _g === void 0 ? void 0 : _g.internalValue) {
            hasAnyValues = true;
            internalValueResult.crlConfiguration = (_h = this._crlConfiguration) === null || _h === void 0 ? void 0 : _h.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._crlConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._crlConfiguration.internalValue = value.crlConfiguration;
        }
    }
    /**
     * @stability stable
     */
    get crlConfiguration() {
        return this._crlConfiguration;
    }
    /**
     * @stability stable
     */
    putCrlConfiguration(value) {
        this._crlConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCrlConfiguration() {
        this._crlConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get crlConfigurationInput() {
        return this._crlConfiguration.internalValue;
    }
}
exports.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference = AcmpcaCertificateAuthorityRevocationConfigurationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityRevocationConfigurationOutputReference[_d] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthorityRevocationConfigurationOutputReference", version: "3.0.1" };
function acmpcaCertificateAuthorityTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.acmpcaCertificateAuthorityTimeoutsToTerraform = acmpcaCertificateAuthorityTimeoutsToTerraform;
/**
 * @stability stable
 */
class AcmpcaCertificateAuthorityTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
}
exports.AcmpcaCertificateAuthorityTimeoutsOutputReference = AcmpcaCertificateAuthorityTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthorityTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthorityTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/acmpca_certificate_authority.html aws_acmpca_certificate_authority}.
 *
 * @stability stable
 */
class AcmpcaCertificateAuthority extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/acmpca_certificate_authority.html aws_acmpca_certificate_authority} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_acmpca_certificate_authority',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // certificate_authority_configuration - computed: false, optional: false, required: true
        this._certificateAuthorityConfiguration = new AcmpcaCertificateAuthorityCertificateAuthorityConfigurationOutputReference(this, "certificate_authority_configuration", true);
        // revocation_configuration - computed: false, optional: true, required: false
        this._revocationConfiguration = new AcmpcaCertificateAuthorityRevocationConfigurationOutputReference(this, "revocation_configuration", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new AcmpcaCertificateAuthorityTimeoutsOutputReference(this, "timeouts", true);
        this._enabled = config.enabled;
        this._permanentDeletionTimeInDays = config.permanentDeletionTimeInDays;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._certificateAuthorityConfiguration.internalValue = config.certificateAuthorityConfiguration;
        this._revocationConfiguration.internalValue = config.revocationConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    // certificate_chain - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    // certificate_signing_request - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateSigningRequest() {
        return this.getStringAttribute('certificate_signing_request');
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // not_after - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get notAfter() {
        return this.getStringAttribute('not_after');
    }
    // not_before - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get notBefore() {
        return this.getStringAttribute('not_before');
    }
    /**
     * @stability stable
     */
    get permanentDeletionTimeInDays() {
        return this.getNumberAttribute('permanent_deletion_time_in_days');
    }
    /**
     * @stability stable
     */
    set permanentDeletionTimeInDays(value) {
        this._permanentDeletionTimeInDays = value;
    }
    /**
     * @stability stable
     */
    resetPermanentDeletionTimeInDays() {
        this._permanentDeletionTimeInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permanentDeletionTimeInDaysInput() {
        return this._permanentDeletionTimeInDays;
    }
    // serial - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serial() {
        return this.getStringAttribute('serial');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get certificateAuthorityConfiguration() {
        return this._certificateAuthorityConfiguration;
    }
    /**
     * @stability stable
     */
    putCertificateAuthorityConfiguration(value) {
        this._certificateAuthorityConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateAuthorityConfigurationInput() {
        return this._certificateAuthorityConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get revocationConfiguration() {
        return this._revocationConfiguration;
    }
    /**
     * @stability stable
     */
    putRevocationConfiguration(value) {
        this._revocationConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRevocationConfiguration() {
        this._revocationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get revocationConfigurationInput() {
        return this._revocationConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enabled: cdktf.booleanToTerraform(this._enabled),
            permanent_deletion_time_in_days: cdktf.numberToTerraform(this._permanentDeletionTimeInDays),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            certificate_authority_configuration: acmpcaCertificateAuthorityCertificateAuthorityConfigurationToTerraform(this._certificateAuthorityConfiguration.internalValue),
            revocation_configuration: acmpcaCertificateAuthorityRevocationConfigurationToTerraform(this._revocationConfiguration.internalValue),
            timeouts: acmpcaCertificateAuthorityTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.AcmpcaCertificateAuthority = AcmpcaCertificateAuthority;
_f = JSII_RTTI_SYMBOL_1;
AcmpcaCertificateAuthority[_f] = { fqn: "@cdktf/provider-aws.acm.AcmpcaCertificateAuthority", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AcmpcaCertificateAuthority.tfResourceType = "aws_acmpca_certificate_authority";
