// @cdktf/provider-aws/lib/elasticbeanstalk/data-aws-elastic-beanstalk-application.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElasticBeanstalkApplication = exports.DataAwsElasticBeanstalkApplicationAppversionLifecycle = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsElasticBeanstalkApplicationAppversionLifecycle extends cdktf.ComplexComputedList {
    // delete_source_from_s3 - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleteSourceFromS3() {
        return this.getBooleanAttribute('delete_source_from_s3');
    }
    // max_age_in_days - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get maxAgeInDays() {
        return this.getNumberAttribute('max_age_in_days');
    }
    // max_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get maxCount() {
        return this.getNumberAttribute('max_count');
    }
    // service_role - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
}
exports.DataAwsElasticBeanstalkApplicationAppversionLifecycle = DataAwsElasticBeanstalkApplicationAppversionLifecycle;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElasticBeanstalkApplicationAppversionLifecycle[_a] = { fqn: "@cdktf/provider-aws.elasticbeanstalk.DataAwsElasticBeanstalkApplicationAppversionLifecycle", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/elastic_beanstalk_application.html aws_elastic_beanstalk_application}.
 *
 * @stability stable
 */
class DataAwsElasticBeanstalkApplication extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/elastic_beanstalk_application.html aws_elastic_beanstalk_application} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastic_beanstalk_application',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // appversion_lifecycle - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    appversionLifecycle(index) {
        return new DataAwsElasticBeanstalkApplicationAppversionLifecycle(this, 'appversion_lifecycle', index);
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsElasticBeanstalkApplication = DataAwsElasticBeanstalkApplication;
_b = JSII_RTTI_SYMBOL_1;
DataAwsElasticBeanstalkApplication[_b] = { fqn: "@cdktf/provider-aws.elasticbeanstalk.DataAwsElasticBeanstalkApplication", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsElasticBeanstalkApplication.tfResourceType = "aws_elastic_beanstalk_application";
