// @cdktf/provider-aws/lib/ses/ses-active-receipt-rule-set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesActiveReceiptRuleSet = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_active_receipt_rule_set.html aws_ses_active_receipt_rule_set}.
 *
 * @stability stable
 */
class SesActiveReceiptRuleSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_active_receipt_rule_set.html aws_ses_active_receipt_rule_set} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_active_receipt_rule_set',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._ruleSetName = config.ruleSetName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ruleSetName() {
        return this.getStringAttribute('rule_set_name');
    }
    /**
     * @stability stable
     */
    set ruleSetName(value) {
        this._ruleSetName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ruleSetNameInput() {
        return this._ruleSetName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            rule_set_name: cdktf.stringToTerraform(this._ruleSetName),
        };
    }
}
exports.SesActiveReceiptRuleSet = SesActiveReceiptRuleSet;
_a = JSII_RTTI_SYMBOL_1;
SesActiveReceiptRuleSet[_a] = { fqn: "@cdktf/provider-aws.ses.SesActiveReceiptRuleSet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SesActiveReceiptRuleSet.tfResourceType = "aws_ses_active_receipt_rule_set";
