// @cdktf/provider-aws/lib/wafregional/wafregional-regex-pattern-set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafregionalRegexPatternSet = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafregional_regex_pattern_set.html aws_wafregional_regex_pattern_set}.
 *
 * @stability stable
 */
class WafregionalRegexPatternSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafregional_regex_pattern_set.html aws_wafregional_regex_pattern_set} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafregional_regex_pattern_set',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._regexPatternStrings = config.regexPatternStrings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get regexPatternStrings() {
        return this.getListAttribute('regex_pattern_strings');
    }
    /**
     * @stability stable
     */
    set regexPatternStrings(value) {
        this._regexPatternStrings = value;
    }
    /**
     * @stability stable
     */
    resetRegexPatternStrings() {
        this._regexPatternStrings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regexPatternStringsInput() {
        return this._regexPatternStrings;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            regex_pattern_strings: cdktf.listMapper(cdktf.stringToTerraform)(this._regexPatternStrings),
        };
    }
}
exports.WafregionalRegexPatternSet = WafregionalRegexPatternSet;
_a = JSII_RTTI_SYMBOL_1;
WafregionalRegexPatternSet[_a] = { fqn: "@cdktf/provider-aws.wafregional.WafregionalRegexPatternSet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WafregionalRegexPatternSet.tfResourceType = "aws_wafregional_regex_pattern_set";
