// @cdktf/provider-aws/lib/cloudwatch/cloudwatch-log-destination-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchLogDestinationPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_log_destination_policy.html aws_cloudwatch_log_destination_policy}.
 *
 * @stability stable
 */
class CloudwatchLogDestinationPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_log_destination_policy.html aws_cloudwatch_log_destination_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_log_destination_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accessPolicy = config.accessPolicy;
        this._destinationName = config.destinationName;
    }
    /**
     * @stability stable
     */
    get accessPolicy() {
        return this.getStringAttribute('access_policy');
    }
    /**
     * @stability stable
     */
    set accessPolicy(value) {
        this._accessPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessPolicyInput() {
        return this._accessPolicy;
    }
    /**
     * @stability stable
     */
    get destinationName() {
        return this.getStringAttribute('destination_name');
    }
    /**
     * @stability stable
     */
    set destinationName(value) {
        this._destinationName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationNameInput() {
        return this._destinationName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            access_policy: cdktf.stringToTerraform(this._accessPolicy),
            destination_name: cdktf.stringToTerraform(this._destinationName),
        };
    }
}
exports.CloudwatchLogDestinationPolicy = CloudwatchLogDestinationPolicy;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchLogDestinationPolicy[_a] = { fqn: "@cdktf/provider-aws.cloudwatch.CloudwatchLogDestinationPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudwatchLogDestinationPolicy.tfResourceType = "aws_cloudwatch_log_destination_policy";
