// @cdktf/provider-aws/lib/cloudwatch/cloudwatch-dashboard.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchDashboard = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_dashboard.html aws_cloudwatch_dashboard}.
 *
 * @stability stable
 */
class CloudwatchDashboard extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_dashboard.html aws_cloudwatch_dashboard} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_dashboard',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dashboardBody = config.dashboardBody;
        this._dashboardName = config.dashboardName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // dashboard_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dashboardArn() {
        return this.getStringAttribute('dashboard_arn');
    }
    /**
     * @stability stable
     */
    get dashboardBody() {
        return this.getStringAttribute('dashboard_body');
    }
    /**
     * @stability stable
     */
    set dashboardBody(value) {
        this._dashboardBody = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dashboardBodyInput() {
        return this._dashboardBody;
    }
    /**
     * @stability stable
     */
    get dashboardName() {
        return this.getStringAttribute('dashboard_name');
    }
    /**
     * @stability stable
     */
    set dashboardName(value) {
        this._dashboardName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dashboardNameInput() {
        return this._dashboardName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            dashboard_body: cdktf.stringToTerraform(this._dashboardBody),
            dashboard_name: cdktf.stringToTerraform(this._dashboardName),
        };
    }
}
exports.CloudwatchDashboard = CloudwatchDashboard;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchDashboard[_a] = { fqn: "@cdktf/provider-aws.cloudwatch.CloudwatchDashboard", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CloudwatchDashboard.tfResourceType = "aws_cloudwatch_dashboard";
