// @cdktf/provider-aws/lib/cloudwatch/data-aws-cloudwatch-log-groups.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudwatchLogGroups = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudwatch_log_groups.html aws_cloudwatch_log_groups}.
 *
 * @stability stable
 */
class DataAwsCloudwatchLogGroups extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudwatch_log_groups.html aws_cloudwatch_log_groups} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_log_groups',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._logGroupNamePrefix = config.logGroupNamePrefix;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arns - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arns() {
        return this.getListAttribute('arns');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get logGroupNamePrefix() {
        return this.getStringAttribute('log_group_name_prefix');
    }
    /**
     * @stability stable
     */
    set logGroupNamePrefix(value) {
        this._logGroupNamePrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logGroupNamePrefixInput() {
        return this._logGroupNamePrefix;
    }
    // log_group_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get logGroupNames() {
        return this.getListAttribute('log_group_names');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            log_group_name_prefix: cdktf.stringToTerraform(this._logGroupNamePrefix),
        };
    }
}
exports.DataAwsCloudwatchLogGroups = DataAwsCloudwatchLogGroups;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudwatchLogGroups[_a] = { fqn: "@cdktf/provider-aws.cloudwatch.DataAwsCloudwatchLogGroups", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCloudwatchLogGroups.tfResourceType = "aws_cloudwatch_log_groups";
