// @cdktf/provider-aws/lib/ses/ses-domain-dkim.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesDomainDkim = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_domain_dkim.html aws_ses_domain_dkim}.
 *
 * @stability stable
 */
class SesDomainDkim extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_domain_dkim.html aws_ses_domain_dkim} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_domain_dkim',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._domain = config.domain;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // dkim_tokens - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dkimTokens() {
        return this.getListAttribute('dkim_tokens');
    }
    /**
     * @stability stable
     */
    get domain() {
        return this.getStringAttribute('domain');
    }
    /**
     * @stability stable
     */
    set domain(value) {
        this._domain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainInput() {
        return this._domain;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            domain: cdktf.stringToTerraform(this._domain),
        };
    }
}
exports.SesDomainDkim = SesDomainDkim;
_a = JSII_RTTI_SYMBOL_1;
SesDomainDkim[_a] = { fqn: "@cdktf/provider-aws.ses.SesDomainDkim", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SesDomainDkim.tfResourceType = "aws_ses_domain_dkim";
