// @cdktf/provider-aws/lib/ses/ses-template.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesTemplate = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_template.html aws_ses_template}.
 *
 * @stability stable
 */
class SesTemplate extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_template.html aws_ses_template} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_template',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._html = config.html;
        this._name = config.name;
        this._subject = config.subject;
        this._text = config.text;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get html() {
        return this.getStringAttribute('html');
    }
    /**
     * @stability stable
     */
    set html(value) {
        this._html = value;
    }
    /**
     * @stability stable
     */
    resetHtml() {
        this._html = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get htmlInput() {
        return this._html;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get subject() {
        return this.getStringAttribute('subject');
    }
    /**
     * @stability stable
     */
    set subject(value) {
        this._subject = value;
    }
    /**
     * @stability stable
     */
    resetSubject() {
        this._subject = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subjectInput() {
        return this._subject;
    }
    /**
     * @stability stable
     */
    get text() {
        return this.getStringAttribute('text');
    }
    /**
     * @stability stable
     */
    set text(value) {
        this._text = value;
    }
    /**
     * @stability stable
     */
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get textInput() {
        return this._text;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            html: cdktf.stringToTerraform(this._html),
            name: cdktf.stringToTerraform(this._name),
            subject: cdktf.stringToTerraform(this._subject),
            text: cdktf.stringToTerraform(this._text),
        };
    }
}
exports.SesTemplate = SesTemplate;
_a = JSII_RTTI_SYMBOL_1;
SesTemplate[_a] = { fqn: "@cdktf/provider-aws.ses.SesTemplate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SesTemplate.tfResourceType = "aws_ses_template";
