// @cdktf/provider-aws/lib/ses/ses-identity-notification-topic.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesIdentityNotificationTopic = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_identity_notification_topic.html aws_ses_identity_notification_topic}.
 *
 * @stability stable
 */
class SesIdentityNotificationTopic extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_identity_notification_topic.html aws_ses_identity_notification_topic} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_identity_notification_topic',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._identity = config.identity;
        this._includeOriginalHeaders = config.includeOriginalHeaders;
        this._notificationType = config.notificationType;
        this._topicArn = config.topicArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get identity() {
        return this.getStringAttribute('identity');
    }
    /**
     * @stability stable
     */
    set identity(value) {
        this._identity = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get identityInput() {
        return this._identity;
    }
    /**
     * @stability stable
     */
    get includeOriginalHeaders() {
        return this.getBooleanAttribute('include_original_headers');
    }
    /**
     * @stability stable
     */
    set includeOriginalHeaders(value) {
        this._includeOriginalHeaders = value;
    }
    /**
     * @stability stable
     */
    resetIncludeOriginalHeaders() {
        this._includeOriginalHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeOriginalHeadersInput() {
        return this._includeOriginalHeaders;
    }
    /**
     * @stability stable
     */
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    /**
     * @stability stable
     */
    set notificationType(value) {
        this._notificationType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationTypeInput() {
        return this._notificationType;
    }
    /**
     * @stability stable
     */
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    /**
     * @stability stable
     */
    set topicArn(value) {
        this._topicArn = value;
    }
    /**
     * @stability stable
     */
    resetTopicArn() {
        this._topicArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get topicArnInput() {
        return this._topicArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            identity: cdktf.stringToTerraform(this._identity),
            include_original_headers: cdktf.booleanToTerraform(this._includeOriginalHeaders),
            notification_type: cdktf.stringToTerraform(this._notificationType),
            topic_arn: cdktf.stringToTerraform(this._topicArn),
        };
    }
}
exports.SesIdentityNotificationTopic = SesIdentityNotificationTopic;
_a = JSII_RTTI_SYMBOL_1;
SesIdentityNotificationTopic[_a] = { fqn: "@cdktf/provider-aws.ses.SesIdentityNotificationTopic", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SesIdentityNotificationTopic.tfResourceType = "aws_ses_identity_notification_topic";
