// @cdktf/provider-aws/lib/secretsmanager/secretsmanager-secret-version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsmanagerSecretVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret_version.html aws_secretsmanager_secret_version}.
 *
 * @stability stable
 */
class SecretsmanagerSecretVersion extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret_version.html aws_secretsmanager_secret_version} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_version',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._secretBinary = config.secretBinary;
        this._secretId = config.secretId;
        this._secretString = config.secretString;
        this._versionStages = config.versionStages;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get secretBinary() {
        return this.getStringAttribute('secret_binary');
    }
    /**
     * @stability stable
     */
    set secretBinary(value) {
        this._secretBinary = value;
    }
    /**
     * @stability stable
     */
    resetSecretBinary() {
        this._secretBinary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretBinaryInput() {
        return this._secretBinary;
    }
    /**
     * @stability stable
     */
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    /**
     * @stability stable
     */
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretIdInput() {
        return this._secretId;
    }
    /**
     * @stability stable
     */
    get secretString() {
        return this.getStringAttribute('secret_string');
    }
    /**
     * @stability stable
     */
    set secretString(value) {
        this._secretString = value;
    }
    /**
     * @stability stable
     */
    resetSecretString() {
        this._secretString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretStringInput() {
        return this._secretString;
    }
    // version_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    /**
     * @stability stable
     */
    get versionStages() {
        return this.getListAttribute('version_stages');
    }
    /**
     * @stability stable
     */
    set versionStages(value) {
        this._versionStages = value;
    }
    /**
     * @stability stable
     */
    resetVersionStages() {
        this._versionStages = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionStagesInput() {
        return this._versionStages;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            secret_binary: cdktf.stringToTerraform(this._secretBinary),
            secret_id: cdktf.stringToTerraform(this._secretId),
            secret_string: cdktf.stringToTerraform(this._secretString),
            version_stages: cdktf.listMapper(cdktf.stringToTerraform)(this._versionStages),
        };
    }
}
exports.SecretsmanagerSecretVersion = SecretsmanagerSecretVersion;
_a = JSII_RTTI_SYMBOL_1;
SecretsmanagerSecretVersion[_a] = { fqn: "@cdktf/provider-aws.secretsmanager.SecretsmanagerSecretVersion", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecretsmanagerSecretVersion.tfResourceType = "aws_secretsmanager_secret_version";
