// @cdktf/provider-aws/lib/fsx/fsx-lustre-file-system.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxLustreFileSystem = exports.FsxLustreFileSystemTimeoutsOutputReference = exports.fsxLustreFileSystemTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxLustreFileSystemTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxLustreFileSystemTimeoutsToTerraform = fsxLustreFileSystemTimeoutsToTerraform;
/**
 * @stability stable
 */
class FsxLustreFileSystemTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.FsxLustreFileSystemTimeoutsOutputReference = FsxLustreFileSystemTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxLustreFileSystemTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsx.FsxLustreFileSystemTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/fsx_lustre_file_system.html aws_fsx_lustre_file_system}.
 *
 * @stability stable
 */
class FsxLustreFileSystem extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/fsx_lustre_file_system.html aws_fsx_lustre_file_system} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_lustre_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxLustreFileSystemTimeoutsOutputReference(this, "timeouts", true);
        this._autoImportPolicy = config.autoImportPolicy;
        this._automaticBackupRetentionDays = config.automaticBackupRetentionDays;
        this._backupId = config.backupId;
        this._copyTagsToBackups = config.copyTagsToBackups;
        this._dailyAutomaticBackupStartTime = config.dailyAutomaticBackupStartTime;
        this._dataCompressionType = config.dataCompressionType;
        this._deploymentType = config.deploymentType;
        this._driveCacheType = config.driveCacheType;
        this._exportPath = config.exportPath;
        this._importPath = config.importPath;
        this._importedFileChunkSize = config.importedFileChunkSize;
        this._kmsKeyId = config.kmsKeyId;
        this._perUnitStorageThroughput = config.perUnitStorageThroughput;
        this._securityGroupIds = config.securityGroupIds;
        this._storageCapacity = config.storageCapacity;
        this._storageType = config.storageType;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._weeklyMaintenanceStartTime = config.weeklyMaintenanceStartTime;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get autoImportPolicy() {
        return this.getStringAttribute('auto_import_policy');
    }
    /**
     * @stability stable
     */
    set autoImportPolicy(value) {
        this._autoImportPolicy = value;
    }
    /**
     * @stability stable
     */
    resetAutoImportPolicy() {
        this._autoImportPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoImportPolicyInput() {
        return this._autoImportPolicy;
    }
    /**
     * @stability stable
     */
    get automaticBackupRetentionDays() {
        return this.getNumberAttribute('automatic_backup_retention_days');
    }
    /**
     * @stability stable
     */
    set automaticBackupRetentionDays(value) {
        this._automaticBackupRetentionDays = value;
    }
    /**
     * @stability stable
     */
    resetAutomaticBackupRetentionDays() {
        this._automaticBackupRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get automaticBackupRetentionDaysInput() {
        return this._automaticBackupRetentionDays;
    }
    /**
     * @stability stable
     */
    get backupId() {
        return this.getStringAttribute('backup_id');
    }
    /**
     * @stability stable
     */
    set backupId(value) {
        this._backupId = value;
    }
    /**
     * @stability stable
     */
    resetBackupId() {
        this._backupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get backupIdInput() {
        return this._backupId;
    }
    /**
     * @stability stable
     */
    get copyTagsToBackups() {
        return this.getBooleanAttribute('copy_tags_to_backups');
    }
    /**
     * @stability stable
     */
    set copyTagsToBackups(value) {
        this._copyTagsToBackups = value;
    }
    /**
     * @stability stable
     */
    resetCopyTagsToBackups() {
        this._copyTagsToBackups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get copyTagsToBackupsInput() {
        return this._copyTagsToBackups;
    }
    /**
     * @stability stable
     */
    get dailyAutomaticBackupStartTime() {
        return this.getStringAttribute('daily_automatic_backup_start_time');
    }
    /**
     * @stability stable
     */
    set dailyAutomaticBackupStartTime(value) {
        this._dailyAutomaticBackupStartTime = value;
    }
    /**
     * @stability stable
     */
    resetDailyAutomaticBackupStartTime() {
        this._dailyAutomaticBackupStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dailyAutomaticBackupStartTimeInput() {
        return this._dailyAutomaticBackupStartTime;
    }
    /**
     * @stability stable
     */
    get dataCompressionType() {
        return this.getStringAttribute('data_compression_type');
    }
    /**
     * @stability stable
     */
    set dataCompressionType(value) {
        this._dataCompressionType = value;
    }
    /**
     * @stability stable
     */
    resetDataCompressionType() {
        this._dataCompressionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataCompressionTypeInput() {
        return this._dataCompressionType;
    }
    /**
     * @stability stable
     */
    get deploymentType() {
        return this.getStringAttribute('deployment_type');
    }
    /**
     * @stability stable
     */
    set deploymentType(value) {
        this._deploymentType = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentType() {
        this._deploymentType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentTypeInput() {
        return this._deploymentType;
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    /**
     * @stability stable
     */
    get driveCacheType() {
        return this.getStringAttribute('drive_cache_type');
    }
    /**
     * @stability stable
     */
    set driveCacheType(value) {
        this._driveCacheType = value;
    }
    /**
     * @stability stable
     */
    resetDriveCacheType() {
        this._driveCacheType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get driveCacheTypeInput() {
        return this._driveCacheType;
    }
    /**
     * @stability stable
     */
    get exportPath() {
        return this.getStringAttribute('export_path');
    }
    /**
     * @stability stable
     */
    set exportPath(value) {
        this._exportPath = value;
    }
    /**
     * @stability stable
     */
    resetExportPath() {
        this._exportPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get exportPathInput() {
        return this._exportPath;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get importPath() {
        return this.getStringAttribute('import_path');
    }
    /**
     * @stability stable
     */
    set importPath(value) {
        this._importPath = value;
    }
    /**
     * @stability stable
     */
    resetImportPath() {
        this._importPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get importPathInput() {
        return this._importPath;
    }
    /**
     * @stability stable
     */
    get importedFileChunkSize() {
        return this.getNumberAttribute('imported_file_chunk_size');
    }
    /**
     * @stability stable
     */
    set importedFileChunkSize(value) {
        this._importedFileChunkSize = value;
    }
    /**
     * @stability stable
     */
    resetImportedFileChunkSize() {
        this._importedFileChunkSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get importedFileChunkSizeInput() {
        return this._importedFileChunkSize;
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    // mount_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get mountName() {
        return this.getStringAttribute('mount_name');
    }
    // network_interface_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkInterfaceIds() {
        return this.getListAttribute('network_interface_ids');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get perUnitStorageThroughput() {
        return this.getNumberAttribute('per_unit_storage_throughput');
    }
    /**
     * @stability stable
     */
    set perUnitStorageThroughput(value) {
        this._perUnitStorageThroughput = value;
    }
    /**
     * @stability stable
     */
    resetPerUnitStorageThroughput() {
        this._perUnitStorageThroughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get perUnitStorageThroughputInput() {
        return this._perUnitStorageThroughput;
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get storageCapacity() {
        return this.getNumberAttribute('storage_capacity');
    }
    /**
     * @stability stable
     */
    set storageCapacity(value) {
        this._storageCapacity = value;
    }
    /**
     * @stability stable
     */
    resetStorageCapacity() {
        this._storageCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageCapacityInput() {
        return this._storageCapacity;
    }
    /**
     * @stability stable
     */
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    /**
     * @stability stable
     */
    set storageType(value) {
        this._storageType = value;
    }
    /**
     * @stability stable
     */
    resetStorageType() {
        this._storageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageTypeInput() {
        return this._storageType;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    get weeklyMaintenanceStartTime() {
        return this.getStringAttribute('weekly_maintenance_start_time');
    }
    /**
     * @stability stable
     */
    set weeklyMaintenanceStartTime(value) {
        this._weeklyMaintenanceStartTime = value;
    }
    /**
     * @stability stable
     */
    resetWeeklyMaintenanceStartTime() {
        this._weeklyMaintenanceStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get weeklyMaintenanceStartTimeInput() {
        return this._weeklyMaintenanceStartTime;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            auto_import_policy: cdktf.stringToTerraform(this._autoImportPolicy),
            automatic_backup_retention_days: cdktf.numberToTerraform(this._automaticBackupRetentionDays),
            backup_id: cdktf.stringToTerraform(this._backupId),
            copy_tags_to_backups: cdktf.booleanToTerraform(this._copyTagsToBackups),
            daily_automatic_backup_start_time: cdktf.stringToTerraform(this._dailyAutomaticBackupStartTime),
            data_compression_type: cdktf.stringToTerraform(this._dataCompressionType),
            deployment_type: cdktf.stringToTerraform(this._deploymentType),
            drive_cache_type: cdktf.stringToTerraform(this._driveCacheType),
            export_path: cdktf.stringToTerraform(this._exportPath),
            import_path: cdktf.stringToTerraform(this._importPath),
            imported_file_chunk_size: cdktf.numberToTerraform(this._importedFileChunkSize),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            per_unit_storage_throughput: cdktf.numberToTerraform(this._perUnitStorageThroughput),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroupIds),
            storage_capacity: cdktf.numberToTerraform(this._storageCapacity),
            storage_type: cdktf.stringToTerraform(this._storageType),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            weekly_maintenance_start_time: cdktf.stringToTerraform(this._weeklyMaintenanceStartTime),
            timeouts: fsxLustreFileSystemTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.FsxLustreFileSystem = FsxLustreFileSystem;
_b = JSII_RTTI_SYMBOL_1;
FsxLustreFileSystem[_b] = { fqn: "@cdktf/provider-aws.fsx.FsxLustreFileSystem", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
FsxLustreFileSystem.tfResourceType = "aws_fsx_lustre_file_system";
