// @cdktf/provider-aws/lib/secretsmanager/data-aws-secretsmanager-secret-version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSecretsmanagerSecretVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_version.html aws_secretsmanager_secret_version}.
 *
 * @stability stable
 */
class DataAwsSecretsmanagerSecretVersion extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_version.html aws_secretsmanager_secret_version} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_version',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._secretId = config.secretId;
        this._versionId = config.versionId;
        this._versionStage = config.versionStage;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // secret_binary - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secretBinary() {
        return this.getStringAttribute('secret_binary');
    }
    /**
     * @stability stable
     */
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    /**
     * @stability stable
     */
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretIdInput() {
        return this._secretId;
    }
    // secret_string - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secretString() {
        return this.getStringAttribute('secret_string');
    }
    /**
     * @stability stable
     */
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    /**
     * @stability stable
     */
    set versionId(value) {
        this._versionId = value;
    }
    /**
     * @stability stable
     */
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionIdInput() {
        return this._versionId;
    }
    /**
     * @stability stable
     */
    get versionStage() {
        return this.getStringAttribute('version_stage');
    }
    /**
     * @stability stable
     */
    set versionStage(value) {
        this._versionStage = value;
    }
    /**
     * @stability stable
     */
    resetVersionStage() {
        this._versionStage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionStageInput() {
        return this._versionStage;
    }
    // version_stages - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get versionStages() {
        return this.getListAttribute('version_stages');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            secret_id: cdktf.stringToTerraform(this._secretId),
            version_id: cdktf.stringToTerraform(this._versionId),
            version_stage: cdktf.stringToTerraform(this._versionStage),
        };
    }
}
exports.DataAwsSecretsmanagerSecretVersion = DataAwsSecretsmanagerSecretVersion;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecretVersion[_a] = { fqn: "@cdktf/provider-aws.secretsmanager.DataAwsSecretsmanagerSecretVersion", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSecretsmanagerSecretVersion.tfResourceType = "aws_secretsmanager_secret_version";
