// @cdktf/provider-aws/lib/secretsmanager/data-aws-secretsmanager-secret-rotation.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSecretsmanagerSecretRotation = exports.DataAwsSecretsmanagerSecretRotationRotationRules = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsSecretsmanagerSecretRotationRotationRules extends cdktf.ComplexComputedList {
    // automatically_after_days - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get automaticallyAfterDays() {
        return this.getNumberAttribute('automatically_after_days');
    }
}
exports.DataAwsSecretsmanagerSecretRotationRotationRules = DataAwsSecretsmanagerSecretRotationRotationRules;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecretRotationRotationRules[_a] = { fqn: "@cdktf/provider-aws.secretsmanager.DataAwsSecretsmanagerSecretRotationRotationRules", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_rotation.html aws_secretsmanager_secret_rotation}.
 *
 * @stability stable
 */
class DataAwsSecretsmanagerSecretRotation extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_rotation.html aws_secretsmanager_secret_rotation} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_rotation',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._secretId = config.secretId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // rotation_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rotationEnabled() {
        return this.getBooleanAttribute('rotation_enabled');
    }
    // rotation_lambda_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rotationLambdaArn() {
        return this.getStringAttribute('rotation_lambda_arn');
    }
    // rotation_rules - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    rotationRules(index) {
        return new DataAwsSecretsmanagerSecretRotationRotationRules(this, 'rotation_rules', index);
    }
    /**
     * @stability stable
     */
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    /**
     * @stability stable
     */
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretIdInput() {
        return this._secretId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            secret_id: cdktf.stringToTerraform(this._secretId),
        };
    }
}
exports.DataAwsSecretsmanagerSecretRotation = DataAwsSecretsmanagerSecretRotation;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecretRotation[_b] = { fqn: "@cdktf/provider-aws.secretsmanager.DataAwsSecretsmanagerSecretRotation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSecretsmanagerSecretRotation.tfResourceType = "aws_secretsmanager_secret_rotation";
