// @cdktf/provider-aws/lib/outposts/data-aws-outposts-outpost-instance-types.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOutpostsOutpostInstanceTypes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/outposts_outpost_instance_types.html aws_outposts_outpost_instance_types}.
 *
 * @stability stable
 */
class DataAwsOutpostsOutpostInstanceTypes extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/outposts_outpost_instance_types.html aws_outposts_outpost_instance_types} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_outposts_outpost_instance_types',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // instance_types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceTypes() {
        return this.getListAttribute('instance_types');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
        };
    }
}
exports.DataAwsOutpostsOutpostInstanceTypes = DataAwsOutpostsOutpostInstanceTypes;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOutpostsOutpostInstanceTypes[_a] = { fqn: "@cdktf/provider-aws.outposts.DataAwsOutpostsOutpostInstanceTypes", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsOutpostsOutpostInstanceTypes.tfResourceType = "aws_outposts_outpost_instance_types";
