// @cdktf/provider-aws/lib/outposts/data-aws-outposts-outpost-instance-type.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOutpostsOutpostInstanceType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/outposts_outpost_instance_type.html aws_outposts_outpost_instance_type}.
 *
 * @stability stable
 */
class DataAwsOutpostsOutpostInstanceType extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/outposts_outpost_instance_type.html aws_outposts_outpost_instance_type} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_outposts_outpost_instance_type',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
        this._instanceType = config.instanceType;
        this._preferredInstanceTypes = config.preferredInstanceTypes;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    /**
     * @stability stable
     */
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get preferredInstanceTypes() {
        return this.getListAttribute('preferred_instance_types');
    }
    /**
     * @stability stable
     */
    set preferredInstanceTypes(value) {
        this._preferredInstanceTypes = value;
    }
    /**
     * @stability stable
     */
    resetPreferredInstanceTypes() {
        this._preferredInstanceTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get preferredInstanceTypesInput() {
        return this._preferredInstanceTypes;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            preferred_instance_types: cdktf.listMapper(cdktf.stringToTerraform)(this._preferredInstanceTypes),
        };
    }
}
exports.DataAwsOutpostsOutpostInstanceType = DataAwsOutpostsOutpostInstanceType;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOutpostsOutpostInstanceType[_a] = { fqn: "@cdktf/provider-aws.outposts.DataAwsOutpostsOutpostInstanceType", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsOutpostsOutpostInstanceType.tfResourceType = "aws_outposts_outpost_instance_type";
