// @cdktf/provider-aws/lib/s3/s3-bucket-replication-configuration.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketReplicationConfigurationA = exports.s3BucketReplicationConfigurationRuleToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform = exports.S3BucketReplicationConfigurationRuleFilterOutputReference = exports.s3BucketReplicationConfigurationRuleFilterToTerraform = exports.S3BucketReplicationConfigurationRuleFilterTagOutputReference = exports.s3BucketReplicationConfigurationRuleFilterTagToTerraform = exports.S3BucketReplicationConfigurationRuleFilterAndOutputReference = exports.s3BucketReplicationConfigurationRuleFilterAndToTerraform = exports.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference = exports.s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform = exports.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference = exports.s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform = s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference = S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        owner: cdktf.stringToTerraform(struct.owner),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform = s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._owner) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._owner = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._owner = value.owner;
        }
    }
    /**
     * @stability stable
     */
    get owner() {
        return this.getStringAttribute('owner');
    }
    /**
     * @stability stable
     */
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerInput() {
        return this._owner;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference = S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        replica_kms_key_id: cdktf.stringToTerraform(struct.replicaKmsKeyId),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform = s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._replicaKmsKeyId) {
            hasAnyValues = true;
            internalValueResult.replicaKmsKeyId = this._replicaKmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._replicaKmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._replicaKmsKeyId = value.replicaKmsKeyId;
        }
    }
    /**
     * @stability stable
     */
    get replicaKmsKeyId() {
        return this.getStringAttribute('replica_kms_key_id');
    }
    /**
     * @stability stable
     */
    set replicaKmsKeyId(value) {
        this._replicaKmsKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replicaKmsKeyIdInput() {
        return this._replicaKmsKeyId;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference = S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform = s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minutes) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minutes = value.minutes;
        }
    }
    /**
     * @stability stable
     */
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    /**
     * @stability stable
     */
    set minutes(value) {
        this._minutes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minutesInput() {
        return this._minutes;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference = S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
        event_threshold: s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform(struct.eventThreshold),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform = s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // event_threshold - computed: false, optional: true, required: false
        this._eventThreshold = new S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference(this, "event_threshold", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _s, _t;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if ((_s = this._eventThreshold) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.eventThreshold = (_t = this._eventThreshold) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
            this._eventThreshold.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
            this._eventThreshold.internalValue = value.eventThreshold;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get eventThreshold() {
        return this._eventThreshold;
    }
    /**
     * @stability stable
     */
    putEventThreshold(value) {
        this._eventThreshold.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEventThreshold() {
        this._eventThreshold.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get eventThresholdInput() {
        return this._eventThreshold.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference = S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform = s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minutes) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minutes = value.minutes;
        }
    }
    /**
     * @stability stable
     */
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    /**
     * @stability stable
     */
    set minutes(value) {
        this._minutes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minutesInput() {
        return this._minutes;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference = S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference;
_f = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference[_f] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
        time: s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform(struct.time),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform = s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // time - computed: false, optional: false, required: true
        this._time = new S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference(this, "time", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _s, _t;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if ((_s = this._time) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.time = (_t = this._time) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
            this._time.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
            this._time.internalValue = value.time;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get time() {
        return this._time;
    }
    /**
     * @stability stable
     */
    putTime(value) {
        this._time.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeInput() {
        return this._time.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference = S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference[_g] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account: cdktf.stringToTerraform(struct.account),
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
        access_control_translation: s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform(struct.accessControlTranslation),
        encryption_configuration: s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform(struct.encryptionConfiguration),
        metrics: s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform(struct.metrics),
        replication_time: s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform(struct.replicationTime),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationToTerraform = s3BucketReplicationConfigurationRuleDestinationToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleDestinationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // access_control_translation - computed: false, optional: true, required: false
        this._accessControlTranslation = new S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference(this, "access_control_translation", true);
        // encryption_configuration - computed: false, optional: true, required: false
        this._encryptionConfiguration = new S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference(this, "encryption_configuration", true);
        // metrics - computed: false, optional: true, required: false
        this._metrics = new S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference(this, "metrics", true);
        // replication_time - computed: false, optional: true, required: false
        this._replicationTime = new S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference(this, "replication_time", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _s, _t, _u, _v, _w, _x, _y, _z;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._account) {
            hasAnyValues = true;
            internalValueResult.account = this._account;
        }
        if (this._bucket) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        if ((_s = this._accessControlTranslation) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.accessControlTranslation = (_t = this._accessControlTranslation) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        if ((_u = this._encryptionConfiguration) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.encryptionConfiguration = (_v = this._encryptionConfiguration) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._metrics) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.metrics = (_x = this._metrics) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        if ((_y = this._replicationTime) === null || _y === void 0 ? void 0 : _y.internalValue) {
            hasAnyValues = true;
            internalValueResult.replicationTime = (_z = this._replicationTime) === null || _z === void 0 ? void 0 : _z.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._account = undefined;
            this._bucket = undefined;
            this._storageClass = undefined;
            this._accessControlTranslation.internalValue = undefined;
            this._encryptionConfiguration.internalValue = undefined;
            this._metrics.internalValue = undefined;
            this._replicationTime.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._account = value.account;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
            this._accessControlTranslation.internalValue = value.accessControlTranslation;
            this._encryptionConfiguration.internalValue = value.encryptionConfiguration;
            this._metrics.internalValue = value.metrics;
            this._replicationTime.internalValue = value.replicationTime;
        }
    }
    /**
     * @stability stable
     */
    get account() {
        return this.getStringAttribute('account');
    }
    /**
     * @stability stable
     */
    set account(value) {
        this._account = value;
    }
    /**
     * @stability stable
     */
    resetAccount() {
        this._account = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountInput() {
        return this._account;
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    /**
     * @stability stable
     */
    set storageClass(value) {
        this._storageClass = value;
    }
    /**
     * @stability stable
     */
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageClassInput() {
        return this._storageClass;
    }
    /**
     * @stability stable
     */
    get accessControlTranslation() {
        return this._accessControlTranslation;
    }
    /**
     * @stability stable
     */
    putAccessControlTranslation(value) {
        this._accessControlTranslation.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAccessControlTranslation() {
        this._accessControlTranslation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessControlTranslationInput() {
        return this._accessControlTranslation.internalValue;
    }
    /**
     * @stability stable
     */
    get encryptionConfiguration() {
        return this._encryptionConfiguration;
    }
    /**
     * @stability stable
     */
    putEncryptionConfiguration(value) {
        this._encryptionConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionConfiguration() {
        this._encryptionConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionConfigurationInput() {
        return this._encryptionConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get metrics() {
        return this._metrics;
    }
    /**
     * @stability stable
     */
    putMetrics(value) {
        this._metrics.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMetrics() {
        this._metrics.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricsInput() {
        return this._metrics.internalValue;
    }
    /**
     * @stability stable
     */
    get replicationTime() {
        return this._replicationTime;
    }
    /**
     * @stability stable
     */
    putReplicationTime(value) {
        this._replicationTime.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetReplicationTime() {
        this._replicationTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replicationTimeInput() {
        return this._replicationTime.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationOutputReference = S3BucketReplicationConfigurationRuleDestinationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationOutputReference[_h] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleDestinationOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform = s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference = S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference[_j] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleFilterAndToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        tags: cdktf.hashMapper(cdktf.anyToTerraform)(struct.tags),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterAndToTerraform = s3BucketReplicationConfigurationRuleFilterAndToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleFilterAndOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._tags) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._tags = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._tags = value.tags;
        }
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterAndOutputReference = S3BucketReplicationConfigurationRuleFilterAndOutputReference;
_k = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterAndOutputReference[_k] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleFilterAndOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleFilterTagToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterTagToTerraform = s3BucketReplicationConfigurationRuleFilterTagToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleFilterTagOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._value = value.value;
        }
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get value() {
        return this.getStringAttribute('value');
    }
    /**
     * @stability stable
     */
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get valueInput() {
        return this._value;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterTagOutputReference = S3BucketReplicationConfigurationRuleFilterTagOutputReference;
_l = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterTagOutputReference[_l] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleFilterTagOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        and: s3BucketReplicationConfigurationRuleFilterAndToTerraform(struct.and),
        tag: s3BucketReplicationConfigurationRuleFilterTagToTerraform(struct.tag),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterToTerraform = s3BucketReplicationConfigurationRuleFilterToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleFilterOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // and - computed: false, optional: true, required: false
        this._and = new S3BucketReplicationConfigurationRuleFilterAndOutputReference(this, "and", true);
        // tag - computed: false, optional: true, required: false
        this._tag = new S3BucketReplicationConfigurationRuleFilterTagOutputReference(this, "tag", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _s, _t, _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if ((_s = this._and) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.and = (_t = this._and) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        if ((_u = this._tag) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.tag = (_v = this._tag) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._and.internalValue = undefined;
            this._tag.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._and.internalValue = value.and;
            this._tag.internalValue = value.tag;
        }
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
    /**
     * @stability stable
     */
    get and() {
        return this._and;
    }
    /**
     * @stability stable
     */
    putAnd(value) {
        this._and.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAnd() {
        this._and.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get andInput() {
        return this._and.internalValue;
    }
    /**
     * @stability stable
     */
    get tag() {
        return this._tag;
    }
    /**
     * @stability stable
     */
    putTag(value) {
        this._tag.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTag() {
        this._tag.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagInput() {
        return this._tag.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterOutputReference = S3BucketReplicationConfigurationRuleFilterOutputReference;
_m = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterOutputReference[_m] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleFilterOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference[_o] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference[_p] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        replica_modifications: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform(struct.replicaModifications),
        sse_kms_encrypted_objects: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform(struct.sseKmsEncryptedObjects),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform;
/**
 * @stability stable
 */
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // replica_modifications - computed: false, optional: true, required: false
        this._replicaModifications = new S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference(this, "replica_modifications", true);
        // sse_kms_encrypted_objects - computed: false, optional: true, required: false
        this._sseKmsEncryptedObjects = new S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference(this, "sse_kms_encrypted_objects", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _s, _t, _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_s = this._replicaModifications) === null || _s === void 0 ? void 0 : _s.internalValue) {
            hasAnyValues = true;
            internalValueResult.replicaModifications = (_t = this._replicaModifications) === null || _t === void 0 ? void 0 : _t.internalValue;
        }
        if ((_u = this._sseKmsEncryptedObjects) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.sseKmsEncryptedObjects = (_v = this._sseKmsEncryptedObjects) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._replicaModifications.internalValue = undefined;
            this._sseKmsEncryptedObjects.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._replicaModifications.internalValue = value.replicaModifications;
            this._sseKmsEncryptedObjects.internalValue = value.sseKmsEncryptedObjects;
        }
    }
    /**
     * @stability stable
     */
    get replicaModifications() {
        return this._replicaModifications;
    }
    /**
     * @stability stable
     */
    putReplicaModifications(value) {
        this._replicaModifications.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetReplicaModifications() {
        this._replicaModifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replicaModificationsInput() {
        return this._replicaModifications.internalValue;
    }
    /**
     * @stability stable
     */
    get sseKmsEncryptedObjects() {
        return this._sseKmsEncryptedObjects;
    }
    /**
     * @stability stable
     */
    putSseKmsEncryptedObjects(value) {
        this._sseKmsEncryptedObjects.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSseKmsEncryptedObjects() {
        this._sseKmsEncryptedObjects.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sseKmsEncryptedObjectsInput() {
        return this._sseKmsEncryptedObjects.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference;
_q = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference[_q] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference", version: "3.0.1" };
function s3BucketReplicationConfigurationRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        prefix: cdktf.stringToTerraform(struct.prefix),
        priority: cdktf.numberToTerraform(struct.priority),
        status: cdktf.stringToTerraform(struct.status),
        delete_marker_replication: s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform(struct.deleteMarkerReplication),
        destination: s3BucketReplicationConfigurationRuleDestinationToTerraform(struct.destination),
        existing_object_replication: s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform(struct.existingObjectReplication),
        filter: s3BucketReplicationConfigurationRuleFilterToTerraform(struct.filter),
        source_selection_criteria: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform(struct.sourceSelectionCriteria),
    };
}
exports.s3BucketReplicationConfigurationRuleToTerraform = s3BucketReplicationConfigurationRuleToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_replication_configuration.html aws_s3_bucket_replication_configuration}.
 *
 * @stability stable
 */
class S3BucketReplicationConfigurationA extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_replication_configuration.html aws_s3_bucket_replication_configuration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_replication_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._bucket = config.bucket;
        this._role = config.role;
        this._rule = config.rule;
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get role() {
        return this.getStringAttribute('role');
    }
    /**
     * @stability stable
     */
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleInput() {
        return this._role;
    }
    /**
     * @stability stable
     */
    get rule() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('rule');
    }
    /**
     * @stability stable
     */
    set rule(value) {
        this._rule = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ruleInput() {
        return this._rule;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            role: cdktf.stringToTerraform(this._role),
            rule: cdktf.listMapper(s3BucketReplicationConfigurationRuleToTerraform)(this._rule),
        };
    }
}
exports.S3BucketReplicationConfigurationA = S3BucketReplicationConfigurationA;
_r = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationA[_r] = { fqn: "@cdktf/provider-aws.s3.S3BucketReplicationConfigurationA", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
S3BucketReplicationConfigurationA.tfResourceType = "aws_s3_bucket_replication_configuration";
