// @cdktf/provider-aws/lib/lambdafunction/lambda-code-signing-config.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaCodeSigningConfig = exports.LambdaCodeSigningConfigPoliciesOutputReference = exports.lambdaCodeSigningConfigPoliciesToTerraform = exports.LambdaCodeSigningConfigAllowedPublishersOutputReference = exports.lambdaCodeSigningConfigAllowedPublishersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lambdaCodeSigningConfigAllowedPublishersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        signing_profile_version_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.signingProfileVersionArns),
    };
}
exports.lambdaCodeSigningConfigAllowedPublishersToTerraform = lambdaCodeSigningConfigAllowedPublishersToTerraform;
/**
 * @stability stable
 */
class LambdaCodeSigningConfigAllowedPublishersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._signingProfileVersionArns) {
            hasAnyValues = true;
            internalValueResult.signingProfileVersionArns = this._signingProfileVersionArns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._signingProfileVersionArns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._signingProfileVersionArns = value.signingProfileVersionArns;
        }
    }
    /**
     * @stability stable
     */
    get signingProfileVersionArns() {
        return this.getListAttribute('signing_profile_version_arns');
    }
    /**
     * @stability stable
     */
    set signingProfileVersionArns(value) {
        this._signingProfileVersionArns = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signingProfileVersionArnsInput() {
        return this._signingProfileVersionArns;
    }
}
exports.LambdaCodeSigningConfigAllowedPublishersOutputReference = LambdaCodeSigningConfigAllowedPublishersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LambdaCodeSigningConfigAllowedPublishersOutputReference[_a] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaCodeSigningConfigAllowedPublishersOutputReference", version: "3.0.1" };
function lambdaCodeSigningConfigPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        untrusted_artifact_on_deployment: cdktf.stringToTerraform(struct.untrustedArtifactOnDeployment),
    };
}
exports.lambdaCodeSigningConfigPoliciesToTerraform = lambdaCodeSigningConfigPoliciesToTerraform;
/**
 * @stability stable
 */
class LambdaCodeSigningConfigPoliciesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._untrustedArtifactOnDeployment) {
            hasAnyValues = true;
            internalValueResult.untrustedArtifactOnDeployment = this._untrustedArtifactOnDeployment;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._untrustedArtifactOnDeployment = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._untrustedArtifactOnDeployment = value.untrustedArtifactOnDeployment;
        }
    }
    /**
     * @stability stable
     */
    get untrustedArtifactOnDeployment() {
        return this.getStringAttribute('untrusted_artifact_on_deployment');
    }
    /**
     * @stability stable
     */
    set untrustedArtifactOnDeployment(value) {
        this._untrustedArtifactOnDeployment = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get untrustedArtifactOnDeploymentInput() {
        return this._untrustedArtifactOnDeployment;
    }
}
exports.LambdaCodeSigningConfigPoliciesOutputReference = LambdaCodeSigningConfigPoliciesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LambdaCodeSigningConfigPoliciesOutputReference[_b] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaCodeSigningConfigPoliciesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lambda_code_signing_config.html aws_lambda_code_signing_config}.
 *
 * @stability stable
 */
class LambdaCodeSigningConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lambda_code_signing_config.html aws_lambda_code_signing_config} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lambda_code_signing_config',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // allowed_publishers - computed: false, optional: false, required: true
        this._allowedPublishers = new LambdaCodeSigningConfigAllowedPublishersOutputReference(this, "allowed_publishers", true);
        // policies - computed: false, optional: true, required: false
        this._policies = new LambdaCodeSigningConfigPoliciesOutputReference(this, "policies", true);
        this._description = config.description;
        this._allowedPublishers.internalValue = config.allowedPublishers;
        this._policies.internalValue = config.policies;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // config_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get configId() {
        return this.getStringAttribute('config_id');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_modified - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModified() {
        return this.getStringAttribute('last_modified');
    }
    /**
     * @stability stable
     */
    get allowedPublishers() {
        return this._allowedPublishers;
    }
    /**
     * @stability stable
     */
    putAllowedPublishers(value) {
        this._allowedPublishers.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowedPublishersInput() {
        return this._allowedPublishers.internalValue;
    }
    /**
     * @stability stable
     */
    get policies() {
        return this._policies;
    }
    /**
     * @stability stable
     */
    putPolicies(value) {
        this._policies.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPolicies() {
        this._policies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policiesInput() {
        return this._policies.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            allowed_publishers: lambdaCodeSigningConfigAllowedPublishersToTerraform(this._allowedPublishers.internalValue),
            policies: lambdaCodeSigningConfigPoliciesToTerraform(this._policies.internalValue),
        };
    }
}
exports.LambdaCodeSigningConfig = LambdaCodeSigningConfig;
_c = JSII_RTTI_SYMBOL_1;
LambdaCodeSigningConfig[_c] = { fqn: "@cdktf/provider-aws.lambdafunction.LambdaCodeSigningConfig", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LambdaCodeSigningConfig.tfResourceType = "aws_lambda_code_signing_config";
