// @cdktf/provider-aws/lib/organizations/data-aws-organizations-delegated-services.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOrganizationsDelegatedServices = exports.DataAwsOrganizationsDelegatedServicesDelegatedServices = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsOrganizationsDelegatedServicesDelegatedServices extends cdktf.ComplexComputedList {
    // delegation_enabled_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get delegationEnabledDate() {
        return this.getStringAttribute('delegation_enabled_date');
    }
    // service_principal - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get servicePrincipal() {
        return this.getStringAttribute('service_principal');
    }
}
exports.DataAwsOrganizationsDelegatedServicesDelegatedServices = DataAwsOrganizationsDelegatedServicesDelegatedServices;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsDelegatedServicesDelegatedServices[_a] = { fqn: "@cdktf/provider-aws.organizations.DataAwsOrganizationsDelegatedServicesDelegatedServices", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/organizations_delegated_services.html aws_organizations_delegated_services}.
 *
 * @stability stable
 */
class DataAwsOrganizationsDelegatedServices extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/organizations_delegated_services.html aws_organizations_delegated_services} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_delegated_services',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accountId = config.accountId;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // delegated_services - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    delegatedServices(index) {
        return new DataAwsOrganizationsDelegatedServicesDelegatedServices(this, 'delegated_services', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
        };
    }
}
exports.DataAwsOrganizationsDelegatedServices = DataAwsOrganizationsDelegatedServices;
_b = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsDelegatedServices[_b] = { fqn: "@cdktf/provider-aws.organizations.DataAwsOrganizationsDelegatedServices", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsOrganizationsDelegatedServices.tfResourceType = "aws_organizations_delegated_services";
