// @cdktf/provider-aws/lib/organizations/data-aws-organizations-organizational-units.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOrganizationsOrganizationalUnits = exports.DataAwsOrganizationsOrganizationalUnitsChildren = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsOrganizationsOrganizationalUnitsChildren extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsOrganizationsOrganizationalUnitsChildren = DataAwsOrganizationsOrganizationalUnitsChildren;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationalUnitsChildren[_a] = { fqn: "@cdktf/provider-aws.organizations.DataAwsOrganizationsOrganizationalUnitsChildren", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/organizations_organizational_units.html aws_organizations_organizational_units}.
 *
 * @stability stable
 */
class DataAwsOrganizationsOrganizationalUnits extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/organizations_organizational_units.html aws_organizations_organizational_units} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_organizational_units',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._parentId = config.parentId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // children - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    children(index) {
        return new DataAwsOrganizationsOrganizationalUnitsChildren(this, 'children', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get parentId() {
        return this.getStringAttribute('parent_id');
    }
    /**
     * @stability stable
     */
    set parentId(value) {
        this._parentId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parentIdInput() {
        return this._parentId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            parent_id: cdktf.stringToTerraform(this._parentId),
        };
    }
}
exports.DataAwsOrganizationsOrganizationalUnits = DataAwsOrganizationsOrganizationalUnits;
_b = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationalUnits[_b] = { fqn: "@cdktf/provider-aws.organizations.DataAwsOrganizationsOrganizationalUnits", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsOrganizationsOrganizationalUnits.tfResourceType = "aws_organizations_organizational_units";
