// @cdktf/provider-aws/lib/organizations/organizations-delegated-administrator.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationsDelegatedAdministrator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/organizations_delegated_administrator.html aws_organizations_delegated_administrator}.
 *
 * @stability stable
 */
class OrganizationsDelegatedAdministrator extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/organizations_delegated_administrator.html aws_organizations_delegated_administrator} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_delegated_administrator',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accountId = config.accountId;
        this._servicePrincipal = config.servicePrincipal;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // delegation_enabled_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get delegationEnabledDate() {
        return this.getStringAttribute('delegation_enabled_date');
    }
    // email - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // joined_method - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get joinedMethod() {
        return this.getStringAttribute('joined_method');
    }
    // joined_timestamp - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get joinedTimestamp() {
        return this.getStringAttribute('joined_timestamp');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    get servicePrincipal() {
        return this.getStringAttribute('service_principal');
    }
    /**
     * @stability stable
     */
    set servicePrincipal(value) {
        this._servicePrincipal = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get servicePrincipalInput() {
        return this._servicePrincipal;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            service_principal: cdktf.stringToTerraform(this._servicePrincipal),
        };
    }
}
exports.OrganizationsDelegatedAdministrator = OrganizationsDelegatedAdministrator;
_a = JSII_RTTI_SYMBOL_1;
OrganizationsDelegatedAdministrator[_a] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsDelegatedAdministrator", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OrganizationsDelegatedAdministrator.tfResourceType = "aws_organizations_delegated_administrator";
