// @cdktf/provider-aws/lib/organizations/organizations-organization.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationsOrganization = exports.OrganizationsOrganizationRoots = exports.OrganizationsOrganizationRootsPolicyTypes = exports.OrganizationsOrganizationNonMasterAccounts = exports.OrganizationsOrganizationAccounts = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class OrganizationsOrganizationAccounts extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.OrganizationsOrganizationAccounts = OrganizationsOrganizationAccounts;
_a = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationAccounts[_a] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsOrganizationAccounts", version: "3.0.1" };
/**
 * @stability stable
 */
class OrganizationsOrganizationNonMasterAccounts extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.OrganizationsOrganizationNonMasterAccounts = OrganizationsOrganizationNonMasterAccounts;
_b = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationNonMasterAccounts[_b] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsOrganizationNonMasterAccounts", version: "3.0.1" };
/**
 * @stability stable
 */
class OrganizationsOrganizationRootsPolicyTypes extends cdktf.ComplexComputedList {
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.OrganizationsOrganizationRootsPolicyTypes = OrganizationsOrganizationRootsPolicyTypes;
_c = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRootsPolicyTypes[_c] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsOrganizationRootsPolicyTypes", version: "3.0.1" };
/**
 * @stability stable
 */
class OrganizationsOrganizationRoots extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // policy_types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get policyTypes() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('policy_types');
    }
}
exports.OrganizationsOrganizationRoots = OrganizationsOrganizationRoots;
_d = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRoots[_d] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsOrganizationRoots", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/organizations_organization.html aws_organizations_organization}.
 *
 * @stability stable
 */
class OrganizationsOrganization extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/organizations_organization.html aws_organizations_organization} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_organization',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._awsServiceAccessPrincipals = config.awsServiceAccessPrincipals;
        this._enabledPolicyTypes = config.enabledPolicyTypes;
        this._featureSet = config.featureSet;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // accounts - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    accounts(index) {
        return new OrganizationsOrganizationAccounts(this, 'accounts', index);
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get awsServiceAccessPrincipals() {
        return this.getListAttribute('aws_service_access_principals');
    }
    /**
     * @stability stable
     */
    set awsServiceAccessPrincipals(value) {
        this._awsServiceAccessPrincipals = value;
    }
    /**
     * @stability stable
     */
    resetAwsServiceAccessPrincipals() {
        this._awsServiceAccessPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get awsServiceAccessPrincipalsInput() {
        return this._awsServiceAccessPrincipals;
    }
    /**
     * @stability stable
     */
    get enabledPolicyTypes() {
        return this.getListAttribute('enabled_policy_types');
    }
    /**
     * @stability stable
     */
    set enabledPolicyTypes(value) {
        this._enabledPolicyTypes = value;
    }
    /**
     * @stability stable
     */
    resetEnabledPolicyTypes() {
        this._enabledPolicyTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledPolicyTypesInput() {
        return this._enabledPolicyTypes;
    }
    /**
     * @stability stable
     */
    get featureSet() {
        return this.getStringAttribute('feature_set');
    }
    /**
     * @stability stable
     */
    set featureSet(value) {
        this._featureSet = value;
    }
    /**
     * @stability stable
     */
    resetFeatureSet() {
        this._featureSet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get featureSetInput() {
        return this._featureSet;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // master_account_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get masterAccountArn() {
        return this.getStringAttribute('master_account_arn');
    }
    // master_account_email - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get masterAccountEmail() {
        return this.getStringAttribute('master_account_email');
    }
    // master_account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get masterAccountId() {
        return this.getStringAttribute('master_account_id');
    }
    // non_master_accounts - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    nonMasterAccounts(index) {
        return new OrganizationsOrganizationNonMasterAccounts(this, 'non_master_accounts', index);
    }
    // roots - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    roots(index) {
        return new OrganizationsOrganizationRoots(this, 'roots', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            aws_service_access_principals: cdktf.listMapper(cdktf.stringToTerraform)(this._awsServiceAccessPrincipals),
            enabled_policy_types: cdktf.listMapper(cdktf.stringToTerraform)(this._enabledPolicyTypes),
            feature_set: cdktf.stringToTerraform(this._featureSet),
        };
    }
}
exports.OrganizationsOrganization = OrganizationsOrganization;
_e = JSII_RTTI_SYMBOL_1;
OrganizationsOrganization[_e] = { fqn: "@cdktf/provider-aws.organizations.OrganizationsOrganization", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OrganizationsOrganization.tfResourceType = "aws_organizations_organization";
