// @cdktf/provider-aws/lib/macie2/macie2-classification-job.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Macie2ClassificationJob = exports.Macie2ClassificationJobScheduleFrequencyOutputReference = exports.macie2ClassificationJobScheduleFrequencyToTerraform = exports.Macie2ClassificationJobS3JobDefinitionOutputReference = exports.macie2ClassificationJobS3JobDefinitionToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingIncludesToTerraform = exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermToTerraform = exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermTagValuesToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingExcludesToTerraform = exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermToTerraform = exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermTagValuesToTerraform = exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference = exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermToTerraform = exports.macie2ClassificationJobS3JobDefinitionBucketDefinitionsToTerraform = exports.Macie2ClassificationJobUserPausedDetails = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class Macie2ClassificationJobUserPausedDetails extends cdktf.ComplexComputedList {
    // job_expires_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jobExpiresAt() {
        return this.getStringAttribute('job_expires_at');
    }
    // job_imminent_expiration_health_event_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jobImminentExpirationHealthEventArn() {
        return this.getStringAttribute('job_imminent_expiration_health_event_arn');
    }
    // job_paused_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jobPausedAt() {
        return this.getStringAttribute('job_paused_at');
    }
}
exports.Macie2ClassificationJobUserPausedDetails = Macie2ClassificationJobUserPausedDetails;
_a = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobUserPausedDetails[_a] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobUserPausedDetails", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionBucketDefinitionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account_id: cdktf.stringToTerraform(struct.accountId),
        buckets: cdktf.listMapper(cdktf.stringToTerraform)(struct.buckets),
    };
}
exports.macie2ClassificationJobS3JobDefinitionBucketDefinitionsToTerraform = macie2ClassificationJobS3JobDefinitionBucketDefinitionsToTerraform;
function macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comparator: cdktf.stringToTerraform(struct.comparator),
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermToTerraform = macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._comparator) {
            hasAnyValues = true;
            internalValueResult.comparator = this._comparator;
        }
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._values) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._comparator = undefined;
            this._key = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._comparator = value.comparator;
            this._key = value.key;
            this._values = value.values;
        }
    }
    /**
     * @stability stable
     */
    get comparator() {
        return this.getStringAttribute('comparator');
    }
    /**
     * @stability stable
     */
    set comparator(value) {
        this._comparator = value;
    }
    /**
     * @stability stable
     */
    resetComparator() {
        this._comparator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get comparatorInput() {
        return this._comparator;
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    /**
     * @stability stable
     */
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get values() {
        return this.getListAttribute('values');
    }
    /**
     * @stability stable
     */
    set values(value) {
        this._values = value;
    }
    /**
     * @stability stable
     */
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get valuesInput() {
        return this._values;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference = Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference[_b] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermTagValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermTagValuesToTerraform = macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermTagValuesToTerraform;
function macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comparator: cdktf.stringToTerraform(struct.comparator),
        key: cdktf.stringToTerraform(struct.key),
        target: cdktf.stringToTerraform(struct.target),
        tag_values: cdktf.listMapper(macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermTagValuesToTerraform)(struct.tagValues),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermToTerraform = macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._comparator) {
            hasAnyValues = true;
            internalValueResult.comparator = this._comparator;
        }
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._target) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._tagValues) {
            hasAnyValues = true;
            internalValueResult.tagValues = this._tagValues;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._comparator = undefined;
            this._key = undefined;
            this._target = undefined;
            this._tagValues = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._comparator = value.comparator;
            this._key = value.key;
            this._target = value.target;
            this._tagValues = value.tagValues;
        }
    }
    /**
     * @stability stable
     */
    get comparator() {
        return this.getStringAttribute('comparator');
    }
    /**
     * @stability stable
     */
    set comparator(value) {
        this._comparator = value;
    }
    /**
     * @stability stable
     */
    resetComparator() {
        this._comparator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get comparatorInput() {
        return this._comparator;
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    /**
     * @stability stable
     */
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get target() {
        return this.getStringAttribute('target');
    }
    /**
     * @stability stable
     */
    set target(value) {
        this._target = value;
    }
    /**
     * @stability stable
     */
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetInput() {
        return this._target;
    }
    /**
     * @stability stable
     */
    get tagValues() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tag_values');
    }
    /**
     * @stability stable
     */
    set tagValues(value) {
        this._tagValues = value;
    }
    /**
     * @stability stable
     */
    resetTagValues() {
        this._tagValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagValuesInput() {
        return this._tagValues;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference = Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference;
_c = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference[_c] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingExcludesAndToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        simple_scope_term: macie2ClassificationJobS3JobDefinitionScopingExcludesAndSimpleScopeTermToTerraform(struct.simpleScopeTerm),
        tag_scope_term: macie2ClassificationJobS3JobDefinitionScopingExcludesAndTagScopeTermToTerraform(struct.tagScopeTerm),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingExcludesAndToTerraform = macie2ClassificationJobS3JobDefinitionScopingExcludesAndToTerraform;
function macie2ClassificationJobS3JobDefinitionScopingExcludesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        and: cdktf.listMapper(macie2ClassificationJobS3JobDefinitionScopingExcludesAndToTerraform)(struct.and),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingExcludesToTerraform = macie2ClassificationJobS3JobDefinitionScopingExcludesToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._and) {
            hasAnyValues = true;
            internalValueResult.and = this._and;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._and = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._and = value.and;
        }
    }
    /**
     * @stability stable
     */
    get and() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('and');
    }
    /**
     * @stability stable
     */
    set and(value) {
        this._and = value;
    }
    /**
     * @stability stable
     */
    resetAnd() {
        this._and = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get andInput() {
        return this._and;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference = Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference[_d] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comparator: cdktf.stringToTerraform(struct.comparator),
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermToTerraform = macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._comparator) {
            hasAnyValues = true;
            internalValueResult.comparator = this._comparator;
        }
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._values) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._comparator = undefined;
            this._key = undefined;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._comparator = value.comparator;
            this._key = value.key;
            this._values = value.values;
        }
    }
    /**
     * @stability stable
     */
    get comparator() {
        return this.getStringAttribute('comparator');
    }
    /**
     * @stability stable
     */
    set comparator(value) {
        this._comparator = value;
    }
    /**
     * @stability stable
     */
    resetComparator() {
        this._comparator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get comparatorInput() {
        return this._comparator;
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    /**
     * @stability stable
     */
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get values() {
        return this.getListAttribute('values');
    }
    /**
     * @stability stable
     */
    set values(value) {
        this._values = value;
    }
    /**
     * @stability stable
     */
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get valuesInput() {
        return this._values;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference = Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference;
_e = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference[_e] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermTagValuesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermTagValuesToTerraform = macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermTagValuesToTerraform;
function macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comparator: cdktf.stringToTerraform(struct.comparator),
        key: cdktf.stringToTerraform(struct.key),
        target: cdktf.stringToTerraform(struct.target),
        tag_values: cdktf.listMapper(macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermTagValuesToTerraform)(struct.tagValues),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermToTerraform = macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._comparator) {
            hasAnyValues = true;
            internalValueResult.comparator = this._comparator;
        }
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._target) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._tagValues) {
            hasAnyValues = true;
            internalValueResult.tagValues = this._tagValues;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._comparator = undefined;
            this._key = undefined;
            this._target = undefined;
            this._tagValues = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._comparator = value.comparator;
            this._key = value.key;
            this._target = value.target;
            this._tagValues = value.tagValues;
        }
    }
    /**
     * @stability stable
     */
    get comparator() {
        return this.getStringAttribute('comparator');
    }
    /**
     * @stability stable
     */
    set comparator(value) {
        this._comparator = value;
    }
    /**
     * @stability stable
     */
    resetComparator() {
        this._comparator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get comparatorInput() {
        return this._comparator;
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    /**
     * @stability stable
     */
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get target() {
        return this.getStringAttribute('target');
    }
    /**
     * @stability stable
     */
    set target(value) {
        this._target = value;
    }
    /**
     * @stability stable
     */
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetInput() {
        return this._target;
    }
    /**
     * @stability stable
     */
    get tagValues() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tag_values');
    }
    /**
     * @stability stable
     */
    set tagValues(value) {
        this._tagValues = value;
    }
    /**
     * @stability stable
     */
    resetTagValues() {
        this._tagValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagValuesInput() {
        return this._tagValues;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference = Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference;
_f = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference[_f] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingIncludesAndToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        simple_scope_term: macie2ClassificationJobS3JobDefinitionScopingIncludesAndSimpleScopeTermToTerraform(struct.simpleScopeTerm),
        tag_scope_term: macie2ClassificationJobS3JobDefinitionScopingIncludesAndTagScopeTermToTerraform(struct.tagScopeTerm),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingIncludesAndToTerraform = macie2ClassificationJobS3JobDefinitionScopingIncludesAndToTerraform;
function macie2ClassificationJobS3JobDefinitionScopingIncludesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        and: cdktf.listMapper(macie2ClassificationJobS3JobDefinitionScopingIncludesAndToTerraform)(struct.and),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingIncludesToTerraform = macie2ClassificationJobS3JobDefinitionScopingIncludesToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._and) {
            hasAnyValues = true;
            internalValueResult.and = this._and;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._and = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._and = value.and;
        }
    }
    /**
     * @stability stable
     */
    get and() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('and');
    }
    /**
     * @stability stable
     */
    set and(value) {
        this._and = value;
    }
    /**
     * @stability stable
     */
    resetAnd() {
        this._and = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get andInput() {
        return this._and;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference = Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference[_g] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionScopingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        excludes: macie2ClassificationJobS3JobDefinitionScopingExcludesToTerraform(struct.excludes),
        includes: macie2ClassificationJobS3JobDefinitionScopingIncludesToTerraform(struct.includes),
    };
}
exports.macie2ClassificationJobS3JobDefinitionScopingToTerraform = macie2ClassificationJobS3JobDefinitionScopingToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionScopingOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // excludes - computed: false, optional: true, required: false
        this._excludes = new Macie2ClassificationJobS3JobDefinitionScopingExcludesOutputReference(this, "excludes", true);
        // includes - computed: false, optional: true, required: false
        this._includes = new Macie2ClassificationJobS3JobDefinitionScopingIncludesOutputReference(this, "includes", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _m, _o, _p, _q;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_m = this._excludes) === null || _m === void 0 ? void 0 : _m.internalValue) {
            hasAnyValues = true;
            internalValueResult.excludes = (_o = this._excludes) === null || _o === void 0 ? void 0 : _o.internalValue;
        }
        if ((_p = this._includes) === null || _p === void 0 ? void 0 : _p.internalValue) {
            hasAnyValues = true;
            internalValueResult.includes = (_q = this._includes) === null || _q === void 0 ? void 0 : _q.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludes.internalValue = undefined;
            this._includes.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludes.internalValue = value.excludes;
            this._includes.internalValue = value.includes;
        }
    }
    /**
     * @stability stable
     */
    get excludes() {
        return this._excludes;
    }
    /**
     * @stability stable
     */
    putExcludes(value) {
        this._excludes.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetExcludes() {
        this._excludes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludesInput() {
        return this._excludes.internalValue;
    }
    /**
     * @stability stable
     */
    get includes() {
        return this._includes;
    }
    /**
     * @stability stable
     */
    putIncludes(value) {
        this._includes.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIncludes() {
        this._includes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includesInput() {
        return this._includes.internalValue;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionScopingOutputReference = Macie2ClassificationJobS3JobDefinitionScopingOutputReference;
_h = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionScopingOutputReference[_h] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionScopingOutputReference", version: "3.0.1" };
function macie2ClassificationJobS3JobDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_definitions: cdktf.listMapper(macie2ClassificationJobS3JobDefinitionBucketDefinitionsToTerraform)(struct.bucketDefinitions),
        scoping: macie2ClassificationJobS3JobDefinitionScopingToTerraform(struct.scoping),
    };
}
exports.macie2ClassificationJobS3JobDefinitionToTerraform = macie2ClassificationJobS3JobDefinitionToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobS3JobDefinitionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // scoping - computed: false, optional: true, required: false
        this._scoping = new Macie2ClassificationJobS3JobDefinitionScopingOutputReference(this, "scoping", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _m, _o;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketDefinitions) {
            hasAnyValues = true;
            internalValueResult.bucketDefinitions = this._bucketDefinitions;
        }
        if ((_m = this._scoping) === null || _m === void 0 ? void 0 : _m.internalValue) {
            hasAnyValues = true;
            internalValueResult.scoping = (_o = this._scoping) === null || _o === void 0 ? void 0 : _o.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketDefinitions = undefined;
            this._scoping.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketDefinitions = value.bucketDefinitions;
            this._scoping.internalValue = value.scoping;
        }
    }
    /**
     * @stability stable
     */
    get bucketDefinitions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('bucket_definitions');
    }
    /**
     * @stability stable
     */
    set bucketDefinitions(value) {
        this._bucketDefinitions = value;
    }
    /**
     * @stability stable
     */
    resetBucketDefinitions() {
        this._bucketDefinitions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketDefinitionsInput() {
        return this._bucketDefinitions;
    }
    /**
     * @stability stable
     */
    get scoping() {
        return this._scoping;
    }
    /**
     * @stability stable
     */
    putScoping(value) {
        this._scoping.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetScoping() {
        this._scoping.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scopingInput() {
        return this._scoping.internalValue;
    }
}
exports.Macie2ClassificationJobS3JobDefinitionOutputReference = Macie2ClassificationJobS3JobDefinitionOutputReference;
_j = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobS3JobDefinitionOutputReference[_j] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobS3JobDefinitionOutputReference", version: "3.0.1" };
function macie2ClassificationJobScheduleFrequencyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        daily_schedule: cdktf.booleanToTerraform(struct.dailySchedule),
        monthly_schedule: cdktf.numberToTerraform(struct.monthlySchedule),
        weekly_schedule: cdktf.stringToTerraform(struct.weeklySchedule),
    };
}
exports.macie2ClassificationJobScheduleFrequencyToTerraform = macie2ClassificationJobScheduleFrequencyToTerraform;
/**
 * @stability stable
 */
class Macie2ClassificationJobScheduleFrequencyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dailySchedule) {
            hasAnyValues = true;
            internalValueResult.dailySchedule = this._dailySchedule;
        }
        if (this._monthlySchedule) {
            hasAnyValues = true;
            internalValueResult.monthlySchedule = this._monthlySchedule;
        }
        if (this._weeklySchedule) {
            hasAnyValues = true;
            internalValueResult.weeklySchedule = this._weeklySchedule;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dailySchedule = undefined;
            this._monthlySchedule = undefined;
            this._weeklySchedule = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dailySchedule = value.dailySchedule;
            this._monthlySchedule = value.monthlySchedule;
            this._weeklySchedule = value.weeklySchedule;
        }
    }
    /**
     * @stability stable
     */
    get dailySchedule() {
        return this.getBooleanAttribute('daily_schedule');
    }
    /**
     * @stability stable
     */
    set dailySchedule(value) {
        this._dailySchedule = value;
    }
    /**
     * @stability stable
     */
    resetDailySchedule() {
        this._dailySchedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dailyScheduleInput() {
        return this._dailySchedule;
    }
    /**
     * @stability stable
     */
    get monthlySchedule() {
        return this.getNumberAttribute('monthly_schedule');
    }
    /**
     * @stability stable
     */
    set monthlySchedule(value) {
        this._monthlySchedule = value;
    }
    /**
     * @stability stable
     */
    resetMonthlySchedule() {
        this._monthlySchedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get monthlyScheduleInput() {
        return this._monthlySchedule;
    }
    /**
     * @stability stable
     */
    get weeklySchedule() {
        return this.getStringAttribute('weekly_schedule');
    }
    /**
     * @stability stable
     */
    set weeklySchedule(value) {
        this._weeklySchedule = value;
    }
    /**
     * @stability stable
     */
    resetWeeklySchedule() {
        this._weeklySchedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get weeklyScheduleInput() {
        return this._weeklySchedule;
    }
}
exports.Macie2ClassificationJobScheduleFrequencyOutputReference = Macie2ClassificationJobScheduleFrequencyOutputReference;
_k = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJobScheduleFrequencyOutputReference[_k] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJobScheduleFrequencyOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/macie2_classification_job.html aws_macie2_classification_job}.
 *
 * @stability stable
 */
class Macie2ClassificationJob extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/macie2_classification_job.html aws_macie2_classification_job} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_macie2_classification_job',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // s3_job_definition - computed: false, optional: false, required: true
        this._s3JobDefinition = new Macie2ClassificationJobS3JobDefinitionOutputReference(this, "s3_job_definition", true);
        // schedule_frequency - computed: false, optional: true, required: false
        this._scheduleFrequency = new Macie2ClassificationJobScheduleFrequencyOutputReference(this, "schedule_frequency", true);
        this._customDataIdentifierIds = config.customDataIdentifierIds;
        this._description = config.description;
        this._initialRun = config.initialRun;
        this._jobStatus = config.jobStatus;
        this._jobType = config.jobType;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._samplingPercentage = config.samplingPercentage;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._s3JobDefinition.internalValue = config.s3JobDefinition;
        this._scheduleFrequency.internalValue = config.scheduleFrequency;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    /**
     * @stability stable
     */
    get customDataIdentifierIds() {
        return this.getListAttribute('custom_data_identifier_ids');
    }
    /**
     * @stability stable
     */
    set customDataIdentifierIds(value) {
        this._customDataIdentifierIds = value;
    }
    /**
     * @stability stable
     */
    resetCustomDataIdentifierIds() {
        this._customDataIdentifierIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customDataIdentifierIdsInput() {
        return this._customDataIdentifierIds;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get initialRun() {
        return this.getBooleanAttribute('initial_run');
    }
    /**
     * @stability stable
     */
    set initialRun(value) {
        this._initialRun = value;
    }
    /**
     * @stability stable
     */
    resetInitialRun() {
        this._initialRun = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get initialRunInput() {
        return this._initialRun;
    }
    // job_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jobArn() {
        return this.getStringAttribute('job_arn');
    }
    // job_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jobId() {
        return this.getStringAttribute('job_id');
    }
    /**
     * @stability stable
     */
    get jobStatus() {
        return this.getStringAttribute('job_status');
    }
    /**
     * @stability stable
     */
    set jobStatus(value) {
        this._jobStatus = value;
    }
    /**
     * @stability stable
     */
    resetJobStatus() {
        this._jobStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jobStatusInput() {
        return this._jobStatus;
    }
    /**
     * @stability stable
     */
    get jobType() {
        return this.getStringAttribute('job_type');
    }
    /**
     * @stability stable
     */
    set jobType(value) {
        this._jobType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jobTypeInput() {
        return this._jobType;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get samplingPercentage() {
        return this.getNumberAttribute('sampling_percentage');
    }
    /**
     * @stability stable
     */
    set samplingPercentage(value) {
        this._samplingPercentage = value;
    }
    /**
     * @stability stable
     */
    resetSamplingPercentage() {
        this._samplingPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get samplingPercentageInput() {
        return this._samplingPercentage;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // user_paused_details - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    userPausedDetails(index) {
        return new Macie2ClassificationJobUserPausedDetails(this, 'user_paused_details', index);
    }
    /**
     * @stability stable
     */
    get s3JobDefinition() {
        return this._s3JobDefinition;
    }
    /**
     * @stability stable
     */
    putS3JobDefinition(value) {
        this._s3JobDefinition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3JobDefinitionInput() {
        return this._s3JobDefinition.internalValue;
    }
    /**
     * @stability stable
     */
    get scheduleFrequency() {
        return this._scheduleFrequency;
    }
    /**
     * @stability stable
     */
    putScheduleFrequency(value) {
        this._scheduleFrequency.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetScheduleFrequency() {
        this._scheduleFrequency.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleFrequencyInput() {
        return this._scheduleFrequency.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            custom_data_identifier_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._customDataIdentifierIds),
            description: cdktf.stringToTerraform(this._description),
            initial_run: cdktf.booleanToTerraform(this._initialRun),
            job_status: cdktf.stringToTerraform(this._jobStatus),
            job_type: cdktf.stringToTerraform(this._jobType),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            sampling_percentage: cdktf.numberToTerraform(this._samplingPercentage),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            s3_job_definition: macie2ClassificationJobS3JobDefinitionToTerraform(this._s3JobDefinition.internalValue),
            schedule_frequency: macie2ClassificationJobScheduleFrequencyToTerraform(this._scheduleFrequency.internalValue),
        };
    }
}
exports.Macie2ClassificationJob = Macie2ClassificationJob;
_l = JSII_RTTI_SYMBOL_1;
Macie2ClassificationJob[_l] = { fqn: "@cdktf/provider-aws.macie2.Macie2ClassificationJob", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Macie2ClassificationJob.tfResourceType = "aws_macie2_classification_job";
