// @cdktf/provider-aws/lib/ssm/data-aws-ssm-document.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsmDocument = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ssm_document.html aws_ssm_document}.
 *
 * @stability stable
 */
class DataAwsSsmDocument extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ssm_document.html aws_ssm_document} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_document',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._documentFormat = config.documentFormat;
        this._documentVersion = config.documentVersion;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // content - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get content() {
        return this.getStringAttribute('content');
    }
    /**
     * @stability stable
     */
    get documentFormat() {
        return this.getStringAttribute('document_format');
    }
    /**
     * @stability stable
     */
    set documentFormat(value) {
        this._documentFormat = value;
    }
    /**
     * @stability stable
     */
    resetDocumentFormat() {
        this._documentFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentFormatInput() {
        return this._documentFormat;
    }
    // document_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get documentType() {
        return this.getStringAttribute('document_type');
    }
    /**
     * @stability stable
     */
    get documentVersion() {
        return this.getStringAttribute('document_version');
    }
    /**
     * @stability stable
     */
    set documentVersion(value) {
        this._documentVersion = value;
    }
    /**
     * @stability stable
     */
    resetDocumentVersion() {
        this._documentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentVersionInput() {
        return this._documentVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            document_format: cdktf.stringToTerraform(this._documentFormat),
            document_version: cdktf.stringToTerraform(this._documentVersion),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsSsmDocument = DataAwsSsmDocument;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsmDocument[_a] = { fqn: "@cdktf/provider-aws.ssm.DataAwsSsmDocument", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSsmDocument.tfResourceType = "aws_ssm_document";
