// @cdktf/provider-aws/lib/backup/backup-vault-lock-configuration.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupVaultLockConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/backup_vault_lock_configuration.html aws_backup_vault_lock_configuration}.
 *
 * @stability stable
 */
class BackupVaultLockConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/backup_vault_lock_configuration.html aws_backup_vault_lock_configuration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_vault_lock_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._backupVaultName = config.backupVaultName;
        this._changeableForDays = config.changeableForDays;
        this._maxRetentionDays = config.maxRetentionDays;
        this._minRetentionDays = config.minRetentionDays;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // backup_vault_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get backupVaultArn() {
        return this.getStringAttribute('backup_vault_arn');
    }
    /**
     * @stability stable
     */
    get backupVaultName() {
        return this.getStringAttribute('backup_vault_name');
    }
    /**
     * @stability stable
     */
    set backupVaultName(value) {
        this._backupVaultName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get backupVaultNameInput() {
        return this._backupVaultName;
    }
    /**
     * @stability stable
     */
    get changeableForDays() {
        return this.getNumberAttribute('changeable_for_days');
    }
    /**
     * @stability stable
     */
    set changeableForDays(value) {
        this._changeableForDays = value;
    }
    /**
     * @stability stable
     */
    resetChangeableForDays() {
        this._changeableForDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get changeableForDaysInput() {
        return this._changeableForDays;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get maxRetentionDays() {
        return this.getNumberAttribute('max_retention_days');
    }
    /**
     * @stability stable
     */
    set maxRetentionDays(value) {
        this._maxRetentionDays = value;
    }
    /**
     * @stability stable
     */
    resetMaxRetentionDays() {
        this._maxRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxRetentionDaysInput() {
        return this._maxRetentionDays;
    }
    /**
     * @stability stable
     */
    get minRetentionDays() {
        return this.getNumberAttribute('min_retention_days');
    }
    /**
     * @stability stable
     */
    set minRetentionDays(value) {
        this._minRetentionDays = value;
    }
    /**
     * @stability stable
     */
    resetMinRetentionDays() {
        this._minRetentionDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get minRetentionDaysInput() {
        return this._minRetentionDays;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            backup_vault_name: cdktf.stringToTerraform(this._backupVaultName),
            changeable_for_days: cdktf.numberToTerraform(this._changeableForDays),
            max_retention_days: cdktf.numberToTerraform(this._maxRetentionDays),
            min_retention_days: cdktf.numberToTerraform(this._minRetentionDays),
        };
    }
}
exports.BackupVaultLockConfiguration = BackupVaultLockConfiguration;
_a = JSII_RTTI_SYMBOL_1;
BackupVaultLockConfiguration[_a] = { fqn: "@cdktf/provider-aws.backup.BackupVaultLockConfiguration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
BackupVaultLockConfiguration.tfResourceType = "aws_backup_vault_lock_configuration";
