// @cdktf/provider-aws/lib/pricing/data-aws-pricing-product.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsPricingProduct = exports.dataAwsPricingProductFiltersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsPricingProductFiltersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataAwsPricingProductFiltersToTerraform = dataAwsPricingProductFiltersToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/pricing_product.html aws_pricing_product}.
 *
 * @stability stable
 */
class DataAwsPricingProduct extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/pricing_product.html aws_pricing_product} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_pricing_product',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._serviceCode = config.serviceCode;
        this._filters = config.filters;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // result - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get result() {
        return this.getStringAttribute('result');
    }
    /**
     * @stability stable
     */
    get serviceCode() {
        return this.getStringAttribute('service_code');
    }
    /**
     * @stability stable
     */
    set serviceCode(value) {
        this._serviceCode = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceCodeInput() {
        return this._serviceCode;
    }
    /**
     * @stability stable
     */
    get filters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filters');
    }
    /**
     * @stability stable
     */
    set filters(value) {
        this._filters = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filtersInput() {
        return this._filters;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            service_code: cdktf.stringToTerraform(this._serviceCode),
            filters: cdktf.listMapper(dataAwsPricingProductFiltersToTerraform)(this._filters),
        };
    }
}
exports.DataAwsPricingProduct = DataAwsPricingProduct;
_a = JSII_RTTI_SYMBOL_1;
DataAwsPricingProduct[_a] = { fqn: "@cdktf/provider-aws.pricing.DataAwsPricingProduct", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsPricingProduct.tfResourceType = "aws_pricing_product";
