// @cdktf/provider-aws/lib/datasync/datasync-agent.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncAgent = exports.DatasyncAgentTimeoutsOutputReference = exports.datasyncAgentTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncAgentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.datasyncAgentTimeoutsToTerraform = datasyncAgentTimeoutsToTerraform;
/**
 * @stability stable
 */
class DatasyncAgentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
}
exports.DatasyncAgentTimeoutsOutputReference = DatasyncAgentTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncAgentTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasync.DatasyncAgentTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/datasync_agent.html aws_datasync_agent}.
 *
 * @stability stable
 */
class DatasyncAgent extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/datasync_agent.html aws_datasync_agent} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_agent',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DatasyncAgentTimeoutsOutputReference(this, "timeouts", true);
        this._activationKey = config.activationKey;
        this._ipAddress = config.ipAddress;
        this._name = config.name;
        this._privateLinkEndpoint = config.privateLinkEndpoint;
        this._securityGroupArns = config.securityGroupArns;
        this._subnetArns = config.subnetArns;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcEndpointId = config.vpcEndpointId;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get activationKey() {
        return this.getStringAttribute('activation_key');
    }
    /**
     * @stability stable
     */
    set activationKey(value) {
        this._activationKey = value;
    }
    /**
     * @stability stable
     */
    resetActivationKey() {
        this._activationKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get activationKeyInput() {
        return this._activationKey;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    /**
     * @stability stable
     */
    set ipAddress(value) {
        this._ipAddress = value;
    }
    /**
     * @stability stable
     */
    resetIpAddress() {
        this._ipAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipAddressInput() {
        return this._ipAddress;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get privateLinkEndpoint() {
        return this.getStringAttribute('private_link_endpoint');
    }
    /**
     * @stability stable
     */
    set privateLinkEndpoint(value) {
        this._privateLinkEndpoint = value;
    }
    /**
     * @stability stable
     */
    resetPrivateLinkEndpoint() {
        this._privateLinkEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateLinkEndpointInput() {
        return this._privateLinkEndpoint;
    }
    /**
     * @stability stable
     */
    get securityGroupArns() {
        return this.getListAttribute('security_group_arns');
    }
    /**
     * @stability stable
     */
    set securityGroupArns(value) {
        this._securityGroupArns = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupArns() {
        this._securityGroupArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupArnsInput() {
        return this._securityGroupArns;
    }
    /**
     * @stability stable
     */
    get subnetArns() {
        return this.getListAttribute('subnet_arns');
    }
    /**
     * @stability stable
     */
    set subnetArns(value) {
        this._subnetArns = value;
    }
    /**
     * @stability stable
     */
    resetSubnetArns() {
        this._subnetArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetArnsInput() {
        return this._subnetArns;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get vpcEndpointId() {
        return this.getStringAttribute('vpc_endpoint_id');
    }
    /**
     * @stability stable
     */
    set vpcEndpointId(value) {
        this._vpcEndpointId = value;
    }
    /**
     * @stability stable
     */
    resetVpcEndpointId() {
        this._vpcEndpointId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcEndpointIdInput() {
        return this._vpcEndpointId;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            activation_key: cdktf.stringToTerraform(this._activationKey),
            ip_address: cdktf.stringToTerraform(this._ipAddress),
            name: cdktf.stringToTerraform(this._name),
            private_link_endpoint: cdktf.stringToTerraform(this._privateLinkEndpoint),
            security_group_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroupArns),
            subnet_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._subnetArns),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            vpc_endpoint_id: cdktf.stringToTerraform(this._vpcEndpointId),
            timeouts: datasyncAgentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DatasyncAgent = DatasyncAgent;
_b = JSII_RTTI_SYMBOL_1;
DatasyncAgent[_b] = { fqn: "@cdktf/provider-aws.datasync.DatasyncAgent", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DatasyncAgent.tfResourceType = "aws_datasync_agent";
