// @cdktf/provider-aws/lib/datasync/datasync-location-efs.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatasyncLocationEfs = exports.DatasyncLocationEfsEc2ConfigOutputReference = exports.datasyncLocationEfsEc2ConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datasyncLocationEfsEc2ConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupArns),
        subnet_arn: cdktf.stringToTerraform(struct.subnetArn),
    };
}
exports.datasyncLocationEfsEc2ConfigToTerraform = datasyncLocationEfsEc2ConfigToTerraform;
/**
 * @stability stable
 */
class DatasyncLocationEfsEc2ConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupArns) {
            hasAnyValues = true;
            internalValueResult.securityGroupArns = this._securityGroupArns;
        }
        if (this._subnetArn) {
            hasAnyValues = true;
            internalValueResult.subnetArn = this._subnetArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupArns = undefined;
            this._subnetArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupArns = value.securityGroupArns;
            this._subnetArn = value.subnetArn;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupArns() {
        return this.getListAttribute('security_group_arns');
    }
    /**
     * @stability stable
     */
    set securityGroupArns(value) {
        this._securityGroupArns = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupArnsInput() {
        return this._securityGroupArns;
    }
    /**
     * @stability stable
     */
    get subnetArn() {
        return this.getStringAttribute('subnet_arn');
    }
    /**
     * @stability stable
     */
    set subnetArn(value) {
        this._subnetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetArnInput() {
        return this._subnetArn;
    }
}
exports.DatasyncLocationEfsEc2ConfigOutputReference = DatasyncLocationEfsEc2ConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatasyncLocationEfsEc2ConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.datasync.DatasyncLocationEfsEc2ConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/datasync_location_efs.html aws_datasync_location_efs}.
 *
 * @stability stable
 */
class DatasyncLocationEfs extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/datasync_location_efs.html aws_datasync_location_efs} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datasync_location_efs',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // ec2_config - computed: false, optional: false, required: true
        this._ec2Config = new DatasyncLocationEfsEc2ConfigOutputReference(this, "ec2_config", true);
        this._efsFileSystemArn = config.efsFileSystemArn;
        this._subdirectory = config.subdirectory;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._ec2Config.internalValue = config.ec2Config;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get efsFileSystemArn() {
        return this.getStringAttribute('efs_file_system_arn');
    }
    /**
     * @stability stable
     */
    set efsFileSystemArn(value) {
        this._efsFileSystemArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get efsFileSystemArnInput() {
        return this._efsFileSystemArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get subdirectory() {
        return this.getStringAttribute('subdirectory');
    }
    /**
     * @stability stable
     */
    set subdirectory(value) {
        this._subdirectory = value;
    }
    /**
     * @stability stable
     */
    resetSubdirectory() {
        this._subdirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subdirectoryInput() {
        return this._subdirectory;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uri - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get uri() {
        return this.getStringAttribute('uri');
    }
    /**
     * @stability stable
     */
    get ec2Config() {
        return this._ec2Config;
    }
    /**
     * @stability stable
     */
    putEc2Config(value) {
        this._ec2Config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ec2ConfigInput() {
        return this._ec2Config.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            efs_file_system_arn: cdktf.stringToTerraform(this._efsFileSystemArn),
            subdirectory: cdktf.stringToTerraform(this._subdirectory),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            ec2_config: datasyncLocationEfsEc2ConfigToTerraform(this._ec2Config.internalValue),
        };
    }
}
exports.DatasyncLocationEfs = DatasyncLocationEfs;
_b = JSII_RTTI_SYMBOL_1;
DatasyncLocationEfs[_b] = { fqn: "@cdktf/provider-aws.datasync.DatasyncLocationEfs", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DatasyncLocationEfs.tfResourceType = "aws_datasync_location_efs";
