// @cdktf/provider-aws/lib/iam/iam-service-linked-role.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamServiceLinkedRole = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_service_linked_role.html aws_iam_service_linked_role}.
 *
 * @stability stable
 */
class IamServiceLinkedRole extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_service_linked_role.html aws_iam_service_linked_role} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_service_linked_role',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._awsServiceName = config.awsServiceName;
        this._customSuffix = config.customSuffix;
        this._description = config.description;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get awsServiceName() {
        return this.getStringAttribute('aws_service_name');
    }
    /**
     * @stability stable
     */
    set awsServiceName(value) {
        this._awsServiceName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get awsServiceNameInput() {
        return this._awsServiceName;
    }
    // create_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createDate() {
        return this.getStringAttribute('create_date');
    }
    /**
     * @stability stable
     */
    get customSuffix() {
        return this.getStringAttribute('custom_suffix');
    }
    /**
     * @stability stable
     */
    set customSuffix(value) {
        this._customSuffix = value;
    }
    /**
     * @stability stable
     */
    resetCustomSuffix() {
        this._customSuffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customSuffixInput() {
        return this._customSuffix;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    // unique_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get uniqueId() {
        return this.getStringAttribute('unique_id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            aws_service_name: cdktf.stringToTerraform(this._awsServiceName),
            custom_suffix: cdktf.stringToTerraform(this._customSuffix),
            description: cdktf.stringToTerraform(this._description),
        };
    }
}
exports.IamServiceLinkedRole = IamServiceLinkedRole;
_a = JSII_RTTI_SYMBOL_1;
IamServiceLinkedRole[_a] = { fqn: "@cdktf/provider-aws.iam.IamServiceLinkedRole", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IamServiceLinkedRole.tfResourceType = "aws_iam_service_linked_role";
