// @cdktf/provider-aws/lib/iam/iam-role.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamRole = exports.iamRoleInlinePolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamRoleInlinePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        policy: cdktf.stringToTerraform(struct.policy),
    };
}
exports.iamRoleInlinePolicyToTerraform = iamRoleInlinePolicyToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_role.html aws_iam_role}.
 *
 * @stability stable
 */
class IamRole extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_role.html aws_iam_role} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_role',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._assumeRolePolicy = config.assumeRolePolicy;
        this._description = config.description;
        this._forceDetachPolicies = config.forceDetachPolicies;
        this._managedPolicyArns = config.managedPolicyArns;
        this._maxSessionDuration = config.maxSessionDuration;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._path = config.path;
        this._permissionsBoundary = config.permissionsBoundary;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._inlinePolicy = config.inlinePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get assumeRolePolicy() {
        return this.getStringAttribute('assume_role_policy');
    }
    /**
     * @stability stable
     */
    set assumeRolePolicy(value) {
        this._assumeRolePolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get assumeRolePolicyInput() {
        return this._assumeRolePolicy;
    }
    // create_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createDate() {
        return this.getStringAttribute('create_date');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get forceDetachPolicies() {
        return this.getBooleanAttribute('force_detach_policies');
    }
    /**
     * @stability stable
     */
    set forceDetachPolicies(value) {
        this._forceDetachPolicies = value;
    }
    /**
     * @stability stable
     */
    resetForceDetachPolicies() {
        this._forceDetachPolicies = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forceDetachPoliciesInput() {
        return this._forceDetachPolicies;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get managedPolicyArns() {
        return this.getListAttribute('managed_policy_arns');
    }
    /**
     * @stability stable
     */
    set managedPolicyArns(value) {
        this._managedPolicyArns = value;
    }
    /**
     * @stability stable
     */
    resetManagedPolicyArns() {
        this._managedPolicyArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get managedPolicyArnsInput() {
        return this._managedPolicyArns;
    }
    /**
     * @stability stable
     */
    get maxSessionDuration() {
        return this.getNumberAttribute('max_session_duration');
    }
    /**
     * @stability stable
     */
    set maxSessionDuration(value) {
        this._maxSessionDuration = value;
    }
    /**
     * @stability stable
     */
    resetMaxSessionDuration() {
        this._maxSessionDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxSessionDurationInput() {
        return this._maxSessionDuration;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    /**
     * @stability stable
     */
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
    /**
     * @stability stable
     */
    get permissionsBoundary() {
        return this.getStringAttribute('permissions_boundary');
    }
    /**
     * @stability stable
     */
    set permissionsBoundary(value) {
        this._permissionsBoundary = value;
    }
    /**
     * @stability stable
     */
    resetPermissionsBoundary() {
        this._permissionsBoundary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionsBoundaryInput() {
        return this._permissionsBoundary;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // unique_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get uniqueId() {
        return this.getStringAttribute('unique_id');
    }
    /**
     * @stability stable
     */
    get inlinePolicy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('inline_policy');
    }
    /**
     * @stability stable
     */
    set inlinePolicy(value) {
        this._inlinePolicy = value;
    }
    /**
     * @stability stable
     */
    resetInlinePolicy() {
        this._inlinePolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inlinePolicyInput() {
        return this._inlinePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            assume_role_policy: cdktf.stringToTerraform(this._assumeRolePolicy),
            description: cdktf.stringToTerraform(this._description),
            force_detach_policies: cdktf.booleanToTerraform(this._forceDetachPolicies),
            managed_policy_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._managedPolicyArns),
            max_session_duration: cdktf.numberToTerraform(this._maxSessionDuration),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            path: cdktf.stringToTerraform(this._path),
            permissions_boundary: cdktf.stringToTerraform(this._permissionsBoundary),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            inline_policy: cdktf.listMapper(iamRoleInlinePolicyToTerraform)(this._inlinePolicy),
        };
    }
}
exports.IamRole = IamRole;
_a = JSII_RTTI_SYMBOL_1;
IamRole[_a] = { fqn: "@cdktf/provider-aws.iam.IamRole", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IamRole.tfResourceType = "aws_iam_role";
