// @cdktf/provider-aws/lib/iam/iam-user-login-profile.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamUserLoginProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_user_login_profile.html aws_iam_user_login_profile}.
 *
 * @stability stable
 */
class IamUserLoginProfile extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_user_login_profile.html aws_iam_user_login_profile} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_user_login_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._passwordLength = config.passwordLength;
        this._passwordResetRequired = config.passwordResetRequired;
        this._pgpKey = config.pgpKey;
        this._user = config.user;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // encrypted_password - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encryptedPassword() {
        return this.getStringAttribute('encrypted_password');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // key_fingerprint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyFingerprint() {
        return this.getStringAttribute('key_fingerprint');
    }
    /**
     * @stability stable
     */
    get passwordLength() {
        return this.getNumberAttribute('password_length');
    }
    /**
     * @stability stable
     */
    set passwordLength(value) {
        this._passwordLength = value;
    }
    /**
     * @stability stable
     */
    resetPasswordLength() {
        this._passwordLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passwordLengthInput() {
        return this._passwordLength;
    }
    /**
     * @stability stable
     */
    get passwordResetRequired() {
        return this.getBooleanAttribute('password_reset_required');
    }
    /**
     * @stability stable
     */
    set passwordResetRequired(value) {
        this._passwordResetRequired = value;
    }
    /**
     * @stability stable
     */
    resetPasswordResetRequired() {
        this._passwordResetRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passwordResetRequiredInput() {
        return this._passwordResetRequired;
    }
    /**
     * @stability stable
     */
    get pgpKey() {
        return this.getStringAttribute('pgp_key');
    }
    /**
     * @stability stable
     */
    set pgpKey(value) {
        this._pgpKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pgpKeyInput() {
        return this._pgpKey;
    }
    /**
     * @stability stable
     */
    get user() {
        return this.getStringAttribute('user');
    }
    /**
     * @stability stable
     */
    set user(value) {
        this._user = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userInput() {
        return this._user;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            password_length: cdktf.numberToTerraform(this._passwordLength),
            password_reset_required: cdktf.booleanToTerraform(this._passwordResetRequired),
            pgp_key: cdktf.stringToTerraform(this._pgpKey),
            user: cdktf.stringToTerraform(this._user),
        };
    }
}
exports.IamUserLoginProfile = IamUserLoginProfile;
_a = JSII_RTTI_SYMBOL_1;
IamUserLoginProfile[_a] = { fqn: "@cdktf/provider-aws.iam.IamUserLoginProfile", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IamUserLoginProfile.tfResourceType = "aws_iam_user_login_profile";
