// @cdktf/provider-aws/lib/appmesh/data-aws-appmesh-virtual-service.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAppmeshVirtualService = exports.DataAwsAppmeshVirtualServiceSpec = exports.DataAwsAppmeshVirtualServiceSpecProvider = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNode = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsAppmeshVirtualServiceSpecProviderVirtualNode extends cdktf.ComplexComputedList {
    // virtual_node_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualNodeName() {
        return this.getStringAttribute('virtual_node_name');
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNode = DataAwsAppmeshVirtualServiceSpecProviderVirtualNode;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualNode[_a] = { fqn: "@cdktf/provider-aws.appmesh.DataAwsAppmeshVirtualServiceSpecProviderVirtualNode", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter extends cdktf.ComplexComputedList {
    // virtual_router_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualRouterName() {
        return this.getStringAttribute('virtual_router_name');
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter = DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter[_b] = { fqn: "@cdktf/provider-aws.appmesh.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouter", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAppmeshVirtualServiceSpecProvider extends cdktf.ComplexComputedList {
    // virtual_node - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualNode() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('virtual_node');
    }
    // virtual_router - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualRouter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('virtual_router');
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProvider = DataAwsAppmeshVirtualServiceSpecProvider;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProvider[_c] = { fqn: "@cdktf/provider-aws.appmesh.DataAwsAppmeshVirtualServiceSpecProvider", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAppmeshVirtualServiceSpec extends cdktf.ComplexComputedList {
    // provider - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get provider() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('provider');
    }
}
exports.DataAwsAppmeshVirtualServiceSpec = DataAwsAppmeshVirtualServiceSpec;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpec[_d] = { fqn: "@cdktf/provider-aws.appmesh.DataAwsAppmeshVirtualServiceSpec", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/appmesh_virtual_service.html aws_appmesh_virtual_service}.
 *
 * @stability stable
 */
class DataAwsAppmeshVirtualService extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/appmesh_virtual_service.html aws_appmesh_virtual_service} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appmesh_virtual_service',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._meshName = config.meshName;
        this._meshOwner = config.meshOwner;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_updated_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    /**
     * @stability stable
     */
    get meshName() {
        return this.getStringAttribute('mesh_name');
    }
    /**
     * @stability stable
     */
    set meshName(value) {
        this._meshName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get meshNameInput() {
        return this._meshName;
    }
    /**
     * @stability stable
     */
    get meshOwner() {
        return this.getStringAttribute('mesh_owner');
    }
    /**
     * @stability stable
     */
    set meshOwner(value) {
        this._meshOwner = value;
    }
    /**
     * @stability stable
     */
    resetMeshOwner() {
        this._meshOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get meshOwnerInput() {
        return this._meshOwner;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // resource_owner - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get resourceOwner() {
        return this.getStringAttribute('resource_owner');
    }
    // spec - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    spec(index) {
        return new DataAwsAppmeshVirtualServiceSpec(this, 'spec', index);
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            mesh_name: cdktf.stringToTerraform(this._meshName),
            mesh_owner: cdktf.stringToTerraform(this._meshOwner),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsAppmeshVirtualService = DataAwsAppmeshVirtualService;
_e = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualService[_e] = { fqn: "@cdktf/provider-aws.appmesh.DataAwsAppmeshVirtualService", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAppmeshVirtualService.tfResourceType = "aws_appmesh_virtual_service";
