// @cdktf/provider-aws/lib/codecommit/data-aws-codecommit-approval-rule-template.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCodecommitApprovalRuleTemplate = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/codecommit_approval_rule_template.html aws_codecommit_approval_rule_template}.
 *
 * @stability stable
 */
class DataAwsCodecommitApprovalRuleTemplate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/codecommit_approval_rule_template.html aws_codecommit_approval_rule_template} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codecommit_approval_rule_template',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // approval_rule_template_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get approvalRuleTemplateId() {
        return this.getStringAttribute('approval_rule_template_id');
    }
    // content - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get content() {
        return this.getStringAttribute('content');
    }
    // creation_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_modified_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModifiedDate() {
        return this.getStringAttribute('last_modified_date');
    }
    // last_modified_user - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastModifiedUser() {
        return this.getStringAttribute('last_modified_user');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // rule_content_sha256 - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ruleContentSha256() {
        return this.getStringAttribute('rule_content_sha256');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsCodecommitApprovalRuleTemplate = DataAwsCodecommitApprovalRuleTemplate;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCodecommitApprovalRuleTemplate[_a] = { fqn: "@cdktf/provider-aws.codecommit.DataAwsCodecommitApprovalRuleTemplate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsCodecommitApprovalRuleTemplate.tfResourceType = "aws_codecommit_approval_rule_template";
