// @cdktf/provider-aws/lib/codebuild/codebuild-project.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodebuildProject = exports.CodebuildProjectVpcConfigOutputReference = exports.codebuildProjectVpcConfigToTerraform = exports.CodebuildProjectSourceOutputReference = exports.codebuildProjectSourceToTerraform = exports.CodebuildProjectSourceGitSubmodulesConfigOutputReference = exports.codebuildProjectSourceGitSubmodulesConfigToTerraform = exports.CodebuildProjectSourceBuildStatusConfigOutputReference = exports.codebuildProjectSourceBuildStatusConfigToTerraform = exports.CodebuildProjectSourceAuthOutputReference = exports.codebuildProjectSourceAuthToTerraform = exports.codebuildProjectSecondarySourcesToTerraform = exports.CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference = exports.codebuildProjectSecondarySourcesGitSubmodulesConfigToTerraform = exports.CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference = exports.codebuildProjectSecondarySourcesBuildStatusConfigToTerraform = exports.CodebuildProjectSecondarySourcesAuthOutputReference = exports.codebuildProjectSecondarySourcesAuthToTerraform = exports.codebuildProjectSecondaryArtifactsToTerraform = exports.CodebuildProjectLogsConfigOutputReference = exports.codebuildProjectLogsConfigToTerraform = exports.CodebuildProjectLogsConfigS3LogsOutputReference = exports.codebuildProjectLogsConfigS3LogsToTerraform = exports.CodebuildProjectLogsConfigCloudwatchLogsOutputReference = exports.codebuildProjectLogsConfigCloudwatchLogsToTerraform = exports.codebuildProjectFileSystemLocationsToTerraform = exports.CodebuildProjectEnvironmentOutputReference = exports.codebuildProjectEnvironmentToTerraform = exports.CodebuildProjectEnvironmentRegistryCredentialOutputReference = exports.codebuildProjectEnvironmentRegistryCredentialToTerraform = exports.codebuildProjectEnvironmentEnvironmentVariableToTerraform = exports.CodebuildProjectCacheOutputReference = exports.codebuildProjectCacheToTerraform = exports.CodebuildProjectBuildBatchConfigOutputReference = exports.codebuildProjectBuildBatchConfigToTerraform = exports.CodebuildProjectBuildBatchConfigRestrictionsOutputReference = exports.codebuildProjectBuildBatchConfigRestrictionsToTerraform = exports.CodebuildProjectArtifactsOutputReference = exports.codebuildProjectArtifactsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codebuildProjectArtifactsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        artifact_identifier: cdktf.stringToTerraform(struct.artifactIdentifier),
        encryption_disabled: cdktf.booleanToTerraform(struct.encryptionDisabled),
        location: cdktf.stringToTerraform(struct.location),
        name: cdktf.stringToTerraform(struct.name),
        namespace_type: cdktf.stringToTerraform(struct.namespaceType),
        override_artifact_name: cdktf.booleanToTerraform(struct.overrideArtifactName),
        packaging: cdktf.stringToTerraform(struct.packaging),
        path: cdktf.stringToTerraform(struct.path),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectArtifactsToTerraform = codebuildProjectArtifactsToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectArtifactsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._artifactIdentifier) {
            hasAnyValues = true;
            internalValueResult.artifactIdentifier = this._artifactIdentifier;
        }
        if (this._encryptionDisabled) {
            hasAnyValues = true;
            internalValueResult.encryptionDisabled = this._encryptionDisabled;
        }
        if (this._location) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._namespaceType) {
            hasAnyValues = true;
            internalValueResult.namespaceType = this._namespaceType;
        }
        if (this._overrideArtifactName) {
            hasAnyValues = true;
            internalValueResult.overrideArtifactName = this._overrideArtifactName;
        }
        if (this._packaging) {
            hasAnyValues = true;
            internalValueResult.packaging = this._packaging;
        }
        if (this._path) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._artifactIdentifier = undefined;
            this._encryptionDisabled = undefined;
            this._location = undefined;
            this._name = undefined;
            this._namespaceType = undefined;
            this._overrideArtifactName = undefined;
            this._packaging = undefined;
            this._path = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._artifactIdentifier = value.artifactIdentifier;
            this._encryptionDisabled = value.encryptionDisabled;
            this._location = value.location;
            this._name = value.name;
            this._namespaceType = value.namespaceType;
            this._overrideArtifactName = value.overrideArtifactName;
            this._packaging = value.packaging;
            this._path = value.path;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get artifactIdentifier() {
        return this.getStringAttribute('artifact_identifier');
    }
    /**
     * @stability stable
     */
    set artifactIdentifier(value) {
        this._artifactIdentifier = value;
    }
    /**
     * @stability stable
     */
    resetArtifactIdentifier() {
        this._artifactIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get artifactIdentifierInput() {
        return this._artifactIdentifier;
    }
    /**
     * @stability stable
     */
    get encryptionDisabled() {
        return this.getBooleanAttribute('encryption_disabled');
    }
    /**
     * @stability stable
     */
    set encryptionDisabled(value) {
        this._encryptionDisabled = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionDisabled() {
        this._encryptionDisabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionDisabledInput() {
        return this._encryptionDisabled;
    }
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    set location(value) {
        this._location = value;
    }
    /**
     * @stability stable
     */
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationInput() {
        return this._location;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namespaceType() {
        return this.getStringAttribute('namespace_type');
    }
    /**
     * @stability stable
     */
    set namespaceType(value) {
        this._namespaceType = value;
    }
    /**
     * @stability stable
     */
    resetNamespaceType() {
        this._namespaceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namespaceTypeInput() {
        return this._namespaceType;
    }
    /**
     * @stability stable
     */
    get overrideArtifactName() {
        return this.getBooleanAttribute('override_artifact_name');
    }
    /**
     * @stability stable
     */
    set overrideArtifactName(value) {
        this._overrideArtifactName = value;
    }
    /**
     * @stability stable
     */
    resetOverrideArtifactName() {
        this._overrideArtifactName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get overrideArtifactNameInput() {
        return this._overrideArtifactName;
    }
    /**
     * @stability stable
     */
    get packaging() {
        return this.getStringAttribute('packaging');
    }
    /**
     * @stability stable
     */
    set packaging(value) {
        this._packaging = value;
    }
    /**
     * @stability stable
     */
    resetPackaging() {
        this._packaging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get packagingInput() {
        return this._packaging;
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    /**
     * @stability stable
     */
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.CodebuildProjectArtifactsOutputReference = CodebuildProjectArtifactsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodebuildProjectArtifactsOutputReference[_a] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectArtifactsOutputReference", version: "3.0.1" };
function codebuildProjectBuildBatchConfigRestrictionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        compute_types_allowed: cdktf.listMapper(cdktf.stringToTerraform)(struct.computeTypesAllowed),
        maximum_builds_allowed: cdktf.numberToTerraform(struct.maximumBuildsAllowed),
    };
}
exports.codebuildProjectBuildBatchConfigRestrictionsToTerraform = codebuildProjectBuildBatchConfigRestrictionsToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectBuildBatchConfigRestrictionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._computeTypesAllowed) {
            hasAnyValues = true;
            internalValueResult.computeTypesAllowed = this._computeTypesAllowed;
        }
        if (this._maximumBuildsAllowed) {
            hasAnyValues = true;
            internalValueResult.maximumBuildsAllowed = this._maximumBuildsAllowed;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._computeTypesAllowed = undefined;
            this._maximumBuildsAllowed = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._computeTypesAllowed = value.computeTypesAllowed;
            this._maximumBuildsAllowed = value.maximumBuildsAllowed;
        }
    }
    /**
     * @stability stable
     */
    get computeTypesAllowed() {
        return this.getListAttribute('compute_types_allowed');
    }
    /**
     * @stability stable
     */
    set computeTypesAllowed(value) {
        this._computeTypesAllowed = value;
    }
    /**
     * @stability stable
     */
    resetComputeTypesAllowed() {
        this._computeTypesAllowed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get computeTypesAllowedInput() {
        return this._computeTypesAllowed;
    }
    /**
     * @stability stable
     */
    get maximumBuildsAllowed() {
        return this.getNumberAttribute('maximum_builds_allowed');
    }
    /**
     * @stability stable
     */
    set maximumBuildsAllowed(value) {
        this._maximumBuildsAllowed = value;
    }
    /**
     * @stability stable
     */
    resetMaximumBuildsAllowed() {
        this._maximumBuildsAllowed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maximumBuildsAllowedInput() {
        return this._maximumBuildsAllowed;
    }
}
exports.CodebuildProjectBuildBatchConfigRestrictionsOutputReference = CodebuildProjectBuildBatchConfigRestrictionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CodebuildProjectBuildBatchConfigRestrictionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectBuildBatchConfigRestrictionsOutputReference", version: "3.0.1" };
function codebuildProjectBuildBatchConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        combine_artifacts: cdktf.booleanToTerraform(struct.combineArtifacts),
        service_role: cdktf.stringToTerraform(struct.serviceRole),
        timeout_in_mins: cdktf.numberToTerraform(struct.timeoutInMins),
        restrictions: codebuildProjectBuildBatchConfigRestrictionsToTerraform(struct.restrictions),
    };
}
exports.codebuildProjectBuildBatchConfigToTerraform = codebuildProjectBuildBatchConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectBuildBatchConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // restrictions - computed: false, optional: true, required: false
        this._restrictions = new CodebuildProjectBuildBatchConfigRestrictionsOutputReference(this, "restrictions", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._combineArtifacts) {
            hasAnyValues = true;
            internalValueResult.combineArtifacts = this._combineArtifacts;
        }
        if (this._serviceRole) {
            hasAnyValues = true;
            internalValueResult.serviceRole = this._serviceRole;
        }
        if (this._timeoutInMins) {
            hasAnyValues = true;
            internalValueResult.timeoutInMins = this._timeoutInMins;
        }
        if ((_u = this._restrictions) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.restrictions = (_v = this._restrictions) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._combineArtifacts = undefined;
            this._serviceRole = undefined;
            this._timeoutInMins = undefined;
            this._restrictions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._combineArtifacts = value.combineArtifacts;
            this._serviceRole = value.serviceRole;
            this._timeoutInMins = value.timeoutInMins;
            this._restrictions.internalValue = value.restrictions;
        }
    }
    /**
     * @stability stable
     */
    get combineArtifacts() {
        return this.getBooleanAttribute('combine_artifacts');
    }
    /**
     * @stability stable
     */
    set combineArtifacts(value) {
        this._combineArtifacts = value;
    }
    /**
     * @stability stable
     */
    resetCombineArtifacts() {
        this._combineArtifacts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get combineArtifactsInput() {
        return this._combineArtifacts;
    }
    /**
     * @stability stable
     */
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    /**
     * @stability stable
     */
    set serviceRole(value) {
        this._serviceRole = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceRoleInput() {
        return this._serviceRole;
    }
    /**
     * @stability stable
     */
    get timeoutInMins() {
        return this.getNumberAttribute('timeout_in_mins');
    }
    /**
     * @stability stable
     */
    set timeoutInMins(value) {
        this._timeoutInMins = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutInMins() {
        this._timeoutInMins = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutInMinsInput() {
        return this._timeoutInMins;
    }
    /**
     * @stability stable
     */
    get restrictions() {
        return this._restrictions;
    }
    /**
     * @stability stable
     */
    putRestrictions(value) {
        this._restrictions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRestrictions() {
        this._restrictions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictionsInput() {
        return this._restrictions.internalValue;
    }
}
exports.CodebuildProjectBuildBatchConfigOutputReference = CodebuildProjectBuildBatchConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CodebuildProjectBuildBatchConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectBuildBatchConfigOutputReference", version: "3.0.1" };
function codebuildProjectCacheToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        location: cdktf.stringToTerraform(struct.location),
        modes: cdktf.listMapper(cdktf.stringToTerraform)(struct.modes),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectCacheToTerraform = codebuildProjectCacheToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectCacheOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._location) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._modes) {
            hasAnyValues = true;
            internalValueResult.modes = this._modes;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._location = undefined;
            this._modes = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._location = value.location;
            this._modes = value.modes;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    set location(value) {
        this._location = value;
    }
    /**
     * @stability stable
     */
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationInput() {
        return this._location;
    }
    /**
     * @stability stable
     */
    get modes() {
        return this.getListAttribute('modes');
    }
    /**
     * @stability stable
     */
    set modes(value) {
        this._modes = value;
    }
    /**
     * @stability stable
     */
    resetModes() {
        this._modes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modesInput() {
        return this._modes;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.CodebuildProjectCacheOutputReference = CodebuildProjectCacheOutputReference;
_d = JSII_RTTI_SYMBOL_1;
CodebuildProjectCacheOutputReference[_d] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectCacheOutputReference", version: "3.0.1" };
function codebuildProjectEnvironmentEnvironmentVariableToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.codebuildProjectEnvironmentEnvironmentVariableToTerraform = codebuildProjectEnvironmentEnvironmentVariableToTerraform;
function codebuildProjectEnvironmentRegistryCredentialToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        credential: cdktf.stringToTerraform(struct.credential),
        credential_provider: cdktf.stringToTerraform(struct.credentialProvider),
    };
}
exports.codebuildProjectEnvironmentRegistryCredentialToTerraform = codebuildProjectEnvironmentRegistryCredentialToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectEnvironmentRegistryCredentialOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._credential) {
            hasAnyValues = true;
            internalValueResult.credential = this._credential;
        }
        if (this._credentialProvider) {
            hasAnyValues = true;
            internalValueResult.credentialProvider = this._credentialProvider;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._credential = undefined;
            this._credentialProvider = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._credential = value.credential;
            this._credentialProvider = value.credentialProvider;
        }
    }
    /**
     * @stability stable
     */
    get credential() {
        return this.getStringAttribute('credential');
    }
    /**
     * @stability stable
     */
    set credential(value) {
        this._credential = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialInput() {
        return this._credential;
    }
    /**
     * @stability stable
     */
    get credentialProvider() {
        return this.getStringAttribute('credential_provider');
    }
    /**
     * @stability stable
     */
    set credentialProvider(value) {
        this._credentialProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialProviderInput() {
        return this._credentialProvider;
    }
}
exports.CodebuildProjectEnvironmentRegistryCredentialOutputReference = CodebuildProjectEnvironmentRegistryCredentialOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CodebuildProjectEnvironmentRegistryCredentialOutputReference[_e] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectEnvironmentRegistryCredentialOutputReference", version: "3.0.1" };
function codebuildProjectEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate: cdktf.stringToTerraform(struct.certificate),
        compute_type: cdktf.stringToTerraform(struct.computeType),
        image: cdktf.stringToTerraform(struct.image),
        image_pull_credentials_type: cdktf.stringToTerraform(struct.imagePullCredentialsType),
        privileged_mode: cdktf.booleanToTerraform(struct.privilegedMode),
        type: cdktf.stringToTerraform(struct.type),
        environment_variable: cdktf.listMapper(codebuildProjectEnvironmentEnvironmentVariableToTerraform)(struct.environmentVariable),
        registry_credential: codebuildProjectEnvironmentRegistryCredentialToTerraform(struct.registryCredential),
    };
}
exports.codebuildProjectEnvironmentToTerraform = codebuildProjectEnvironmentToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // registry_credential - computed: false, optional: true, required: false
        this._registryCredential = new CodebuildProjectEnvironmentRegistryCredentialOutputReference(this, "registry_credential", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificate) {
            hasAnyValues = true;
            internalValueResult.certificate = this._certificate;
        }
        if (this._computeType) {
            hasAnyValues = true;
            internalValueResult.computeType = this._computeType;
        }
        if (this._image) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._imagePullCredentialsType) {
            hasAnyValues = true;
            internalValueResult.imagePullCredentialsType = this._imagePullCredentialsType;
        }
        if (this._privilegedMode) {
            hasAnyValues = true;
            internalValueResult.privilegedMode = this._privilegedMode;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._environmentVariable) {
            hasAnyValues = true;
            internalValueResult.environmentVariable = this._environmentVariable;
        }
        if ((_u = this._registryCredential) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.registryCredential = (_v = this._registryCredential) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificate = undefined;
            this._computeType = undefined;
            this._image = undefined;
            this._imagePullCredentialsType = undefined;
            this._privilegedMode = undefined;
            this._type = undefined;
            this._environmentVariable = undefined;
            this._registryCredential.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificate = value.certificate;
            this._computeType = value.computeType;
            this._image = value.image;
            this._imagePullCredentialsType = value.imagePullCredentialsType;
            this._privilegedMode = value.privilegedMode;
            this._type = value.type;
            this._environmentVariable = value.environmentVariable;
            this._registryCredential.internalValue = value.registryCredential;
        }
    }
    /**
     * @stability stable
     */
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    /**
     * @stability stable
     */
    set certificate(value) {
        this._certificate = value;
    }
    /**
     * @stability stable
     */
    resetCertificate() {
        this._certificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateInput() {
        return this._certificate;
    }
    /**
     * @stability stable
     */
    get computeType() {
        return this.getStringAttribute('compute_type');
    }
    /**
     * @stability stable
     */
    set computeType(value) {
        this._computeType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get computeTypeInput() {
        return this._computeType;
    }
    /**
     * @stability stable
     */
    get image() {
        return this.getStringAttribute('image');
    }
    /**
     * @stability stable
     */
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get imageInput() {
        return this._image;
    }
    /**
     * @stability stable
     */
    get imagePullCredentialsType() {
        return this.getStringAttribute('image_pull_credentials_type');
    }
    /**
     * @stability stable
     */
    set imagePullCredentialsType(value) {
        this._imagePullCredentialsType = value;
    }
    /**
     * @stability stable
     */
    resetImagePullCredentialsType() {
        this._imagePullCredentialsType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get imagePullCredentialsTypeInput() {
        return this._imagePullCredentialsType;
    }
    /**
     * @stability stable
     */
    get privilegedMode() {
        return this.getBooleanAttribute('privileged_mode');
    }
    /**
     * @stability stable
     */
    set privilegedMode(value) {
        this._privilegedMode = value;
    }
    /**
     * @stability stable
     */
    resetPrivilegedMode() {
        this._privilegedMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privilegedModeInput() {
        return this._privilegedMode;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get environmentVariable() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('environment_variable');
    }
    /**
     * @stability stable
     */
    set environmentVariable(value) {
        this._environmentVariable = value;
    }
    /**
     * @stability stable
     */
    resetEnvironmentVariable() {
        this._environmentVariable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentVariableInput() {
        return this._environmentVariable;
    }
    /**
     * @stability stable
     */
    get registryCredential() {
        return this._registryCredential;
    }
    /**
     * @stability stable
     */
    putRegistryCredential(value) {
        this._registryCredential.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRegistryCredential() {
        this._registryCredential.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registryCredentialInput() {
        return this._registryCredential.internalValue;
    }
}
exports.CodebuildProjectEnvironmentOutputReference = CodebuildProjectEnvironmentOutputReference;
_f = JSII_RTTI_SYMBOL_1;
CodebuildProjectEnvironmentOutputReference[_f] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectEnvironmentOutputReference", version: "3.0.1" };
function codebuildProjectFileSystemLocationsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        identifier: cdktf.stringToTerraform(struct.identifier),
        location: cdktf.stringToTerraform(struct.location),
        mount_options: cdktf.stringToTerraform(struct.mountOptions),
        mount_point: cdktf.stringToTerraform(struct.mountPoint),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectFileSystemLocationsToTerraform = codebuildProjectFileSystemLocationsToTerraform;
function codebuildProjectLogsConfigCloudwatchLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        group_name: cdktf.stringToTerraform(struct.groupName),
        status: cdktf.stringToTerraform(struct.status),
        stream_name: cdktf.stringToTerraform(struct.streamName),
    };
}
exports.codebuildProjectLogsConfigCloudwatchLogsToTerraform = codebuildProjectLogsConfigCloudwatchLogsToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectLogsConfigCloudwatchLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._groupName) {
            hasAnyValues = true;
            internalValueResult.groupName = this._groupName;
        }
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if (this._streamName) {
            hasAnyValues = true;
            internalValueResult.streamName = this._streamName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._groupName = undefined;
            this._status = undefined;
            this._streamName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._groupName = value.groupName;
            this._status = value.status;
            this._streamName = value.streamName;
        }
    }
    /**
     * @stability stable
     */
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    /**
     * @stability stable
     */
    set groupName(value) {
        this._groupName = value;
    }
    /**
     * @stability stable
     */
    resetGroupName() {
        this._groupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupNameInput() {
        return this._groupName;
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get streamName() {
        return this.getStringAttribute('stream_name');
    }
    /**
     * @stability stable
     */
    set streamName(value) {
        this._streamName = value;
    }
    /**
     * @stability stable
     */
    resetStreamName() {
        this._streamName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get streamNameInput() {
        return this._streamName;
    }
}
exports.CodebuildProjectLogsConfigCloudwatchLogsOutputReference = CodebuildProjectLogsConfigCloudwatchLogsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CodebuildProjectLogsConfigCloudwatchLogsOutputReference[_g] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectLogsConfigCloudwatchLogsOutputReference", version: "3.0.1" };
function codebuildProjectLogsConfigS3LogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_disabled: cdktf.booleanToTerraform(struct.encryptionDisabled),
        location: cdktf.stringToTerraform(struct.location),
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.codebuildProjectLogsConfigS3LogsToTerraform = codebuildProjectLogsConfigS3LogsToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectLogsConfigS3LogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionDisabled) {
            hasAnyValues = true;
            internalValueResult.encryptionDisabled = this._encryptionDisabled;
        }
        if (this._location) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionDisabled = undefined;
            this._location = undefined;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionDisabled = value.encryptionDisabled;
            this._location = value.location;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get encryptionDisabled() {
        return this.getBooleanAttribute('encryption_disabled');
    }
    /**
     * @stability stable
     */
    set encryptionDisabled(value) {
        this._encryptionDisabled = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionDisabled() {
        this._encryptionDisabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionDisabledInput() {
        return this._encryptionDisabled;
    }
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    set location(value) {
        this._location = value;
    }
    /**
     * @stability stable
     */
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationInput() {
        return this._location;
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.CodebuildProjectLogsConfigS3LogsOutputReference = CodebuildProjectLogsConfigS3LogsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CodebuildProjectLogsConfigS3LogsOutputReference[_h] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectLogsConfigS3LogsOutputReference", version: "3.0.1" };
function codebuildProjectLogsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_logs: codebuildProjectLogsConfigCloudwatchLogsToTerraform(struct.cloudwatchLogs),
        s3_logs: codebuildProjectLogsConfigS3LogsToTerraform(struct.s3Logs),
    };
}
exports.codebuildProjectLogsConfigToTerraform = codebuildProjectLogsConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectLogsConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cloudwatch_logs - computed: false, optional: true, required: false
        this._cloudwatchLogs = new CodebuildProjectLogsConfigCloudwatchLogsOutputReference(this, "cloudwatch_logs", true);
        // s3_logs - computed: false, optional: true, required: false
        this._s3Logs = new CodebuildProjectLogsConfigS3LogsOutputReference(this, "s3_logs", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v, _w, _x;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._cloudwatchLogs) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogs = (_v = this._cloudwatchLogs) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._s3Logs) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3Logs = (_x = this._s3Logs) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchLogs.internalValue = undefined;
            this._s3Logs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchLogs.internalValue = value.cloudwatchLogs;
            this._s3Logs.internalValue = value.s3Logs;
        }
    }
    /**
     * @stability stable
     */
    get cloudwatchLogs() {
        return this._cloudwatchLogs;
    }
    /**
     * @stability stable
     */
    putCloudwatchLogs(value) {
        this._cloudwatchLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchLogs() {
        this._cloudwatchLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchLogsInput() {
        return this._cloudwatchLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get s3Logs() {
        return this._s3Logs;
    }
    /**
     * @stability stable
     */
    putS3Logs(value) {
        this._s3Logs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetS3Logs() {
        this._s3Logs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3LogsInput() {
        return this._s3Logs.internalValue;
    }
}
exports.CodebuildProjectLogsConfigOutputReference = CodebuildProjectLogsConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CodebuildProjectLogsConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectLogsConfigOutputReference", version: "3.0.1" };
function codebuildProjectSecondaryArtifactsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        artifact_identifier: cdktf.stringToTerraform(struct.artifactIdentifier),
        encryption_disabled: cdktf.booleanToTerraform(struct.encryptionDisabled),
        location: cdktf.stringToTerraform(struct.location),
        name: cdktf.stringToTerraform(struct.name),
        namespace_type: cdktf.stringToTerraform(struct.namespaceType),
        override_artifact_name: cdktf.booleanToTerraform(struct.overrideArtifactName),
        packaging: cdktf.stringToTerraform(struct.packaging),
        path: cdktf.stringToTerraform(struct.path),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectSecondaryArtifactsToTerraform = codebuildProjectSecondaryArtifactsToTerraform;
function codebuildProjectSecondarySourcesAuthToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource: cdktf.stringToTerraform(struct.resource),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectSecondarySourcesAuthToTerraform = codebuildProjectSecondarySourcesAuthToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSecondarySourcesAuthOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resource) {
            hasAnyValues = true;
            internalValueResult.resource = this._resource;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resource = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resource = value.resource;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get resource() {
        return this.getStringAttribute('resource');
    }
    /**
     * @stability stable
     */
    set resource(value) {
        this._resource = value;
    }
    /**
     * @stability stable
     */
    resetResource() {
        this._resource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceInput() {
        return this._resource;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.CodebuildProjectSecondarySourcesAuthOutputReference = CodebuildProjectSecondarySourcesAuthOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CodebuildProjectSecondarySourcesAuthOutputReference[_k] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSecondarySourcesAuthOutputReference", version: "3.0.1" };
function codebuildProjectSecondarySourcesBuildStatusConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        context: cdktf.stringToTerraform(struct.context),
        target_url: cdktf.stringToTerraform(struct.targetUrl),
    };
}
exports.codebuildProjectSecondarySourcesBuildStatusConfigToTerraform = codebuildProjectSecondarySourcesBuildStatusConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._context) {
            hasAnyValues = true;
            internalValueResult.context = this._context;
        }
        if (this._targetUrl) {
            hasAnyValues = true;
            internalValueResult.targetUrl = this._targetUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._context = undefined;
            this._targetUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._context = value.context;
            this._targetUrl = value.targetUrl;
        }
    }
    /**
     * @stability stable
     */
    get context() {
        return this.getStringAttribute('context');
    }
    /**
     * @stability stable
     */
    set context(value) {
        this._context = value;
    }
    /**
     * @stability stable
     */
    resetContext() {
        this._context = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contextInput() {
        return this._context;
    }
    /**
     * @stability stable
     */
    get targetUrl() {
        return this.getStringAttribute('target_url');
    }
    /**
     * @stability stable
     */
    set targetUrl(value) {
        this._targetUrl = value;
    }
    /**
     * @stability stable
     */
    resetTargetUrl() {
        this._targetUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetUrlInput() {
        return this._targetUrl;
    }
}
exports.CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference = CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference[_l] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSecondarySourcesBuildStatusConfigOutputReference", version: "3.0.1" };
function codebuildProjectSecondarySourcesGitSubmodulesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fetch_submodules: cdktf.booleanToTerraform(struct.fetchSubmodules),
    };
}
exports.codebuildProjectSecondarySourcesGitSubmodulesConfigToTerraform = codebuildProjectSecondarySourcesGitSubmodulesConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fetchSubmodules) {
            hasAnyValues = true;
            internalValueResult.fetchSubmodules = this._fetchSubmodules;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fetchSubmodules = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fetchSubmodules = value.fetchSubmodules;
        }
    }
    /**
     * @stability stable
     */
    get fetchSubmodules() {
        return this.getBooleanAttribute('fetch_submodules');
    }
    /**
     * @stability stable
     */
    set fetchSubmodules(value) {
        this._fetchSubmodules = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fetchSubmodulesInput() {
        return this._fetchSubmodules;
    }
}
exports.CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference = CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference[_m] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSecondarySourcesGitSubmodulesConfigOutputReference", version: "3.0.1" };
function codebuildProjectSecondarySourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        buildspec: cdktf.stringToTerraform(struct.buildspec),
        git_clone_depth: cdktf.numberToTerraform(struct.gitCloneDepth),
        insecure_ssl: cdktf.booleanToTerraform(struct.insecureSsl),
        location: cdktf.stringToTerraform(struct.location),
        report_build_status: cdktf.booleanToTerraform(struct.reportBuildStatus),
        source_identifier: cdktf.stringToTerraform(struct.sourceIdentifier),
        type: cdktf.stringToTerraform(struct.type),
        auth: codebuildProjectSecondarySourcesAuthToTerraform(struct.auth),
        build_status_config: codebuildProjectSecondarySourcesBuildStatusConfigToTerraform(struct.buildStatusConfig),
        git_submodules_config: codebuildProjectSecondarySourcesGitSubmodulesConfigToTerraform(struct.gitSubmodulesConfig),
    };
}
exports.codebuildProjectSecondarySourcesToTerraform = codebuildProjectSecondarySourcesToTerraform;
function codebuildProjectSourceAuthToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource: cdktf.stringToTerraform(struct.resource),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildProjectSourceAuthToTerraform = codebuildProjectSourceAuthToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSourceAuthOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resource) {
            hasAnyValues = true;
            internalValueResult.resource = this._resource;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resource = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resource = value.resource;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get resource() {
        return this.getStringAttribute('resource');
    }
    /**
     * @stability stable
     */
    set resource(value) {
        this._resource = value;
    }
    /**
     * @stability stable
     */
    resetResource() {
        this._resource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceInput() {
        return this._resource;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.CodebuildProjectSourceAuthOutputReference = CodebuildProjectSourceAuthOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CodebuildProjectSourceAuthOutputReference[_o] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSourceAuthOutputReference", version: "3.0.1" };
function codebuildProjectSourceBuildStatusConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        context: cdktf.stringToTerraform(struct.context),
        target_url: cdktf.stringToTerraform(struct.targetUrl),
    };
}
exports.codebuildProjectSourceBuildStatusConfigToTerraform = codebuildProjectSourceBuildStatusConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSourceBuildStatusConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._context) {
            hasAnyValues = true;
            internalValueResult.context = this._context;
        }
        if (this._targetUrl) {
            hasAnyValues = true;
            internalValueResult.targetUrl = this._targetUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._context = undefined;
            this._targetUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._context = value.context;
            this._targetUrl = value.targetUrl;
        }
    }
    /**
     * @stability stable
     */
    get context() {
        return this.getStringAttribute('context');
    }
    /**
     * @stability stable
     */
    set context(value) {
        this._context = value;
    }
    /**
     * @stability stable
     */
    resetContext() {
        this._context = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contextInput() {
        return this._context;
    }
    /**
     * @stability stable
     */
    get targetUrl() {
        return this.getStringAttribute('target_url');
    }
    /**
     * @stability stable
     */
    set targetUrl(value) {
        this._targetUrl = value;
    }
    /**
     * @stability stable
     */
    resetTargetUrl() {
        this._targetUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetUrlInput() {
        return this._targetUrl;
    }
}
exports.CodebuildProjectSourceBuildStatusConfigOutputReference = CodebuildProjectSourceBuildStatusConfigOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CodebuildProjectSourceBuildStatusConfigOutputReference[_p] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSourceBuildStatusConfigOutputReference", version: "3.0.1" };
function codebuildProjectSourceGitSubmodulesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fetch_submodules: cdktf.booleanToTerraform(struct.fetchSubmodules),
    };
}
exports.codebuildProjectSourceGitSubmodulesConfigToTerraform = codebuildProjectSourceGitSubmodulesConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSourceGitSubmodulesConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fetchSubmodules) {
            hasAnyValues = true;
            internalValueResult.fetchSubmodules = this._fetchSubmodules;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fetchSubmodules = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fetchSubmodules = value.fetchSubmodules;
        }
    }
    /**
     * @stability stable
     */
    get fetchSubmodules() {
        return this.getBooleanAttribute('fetch_submodules');
    }
    /**
     * @stability stable
     */
    set fetchSubmodules(value) {
        this._fetchSubmodules = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fetchSubmodulesInput() {
        return this._fetchSubmodules;
    }
}
exports.CodebuildProjectSourceGitSubmodulesConfigOutputReference = CodebuildProjectSourceGitSubmodulesConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CodebuildProjectSourceGitSubmodulesConfigOutputReference[_q] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSourceGitSubmodulesConfigOutputReference", version: "3.0.1" };
function codebuildProjectSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        buildspec: cdktf.stringToTerraform(struct.buildspec),
        git_clone_depth: cdktf.numberToTerraform(struct.gitCloneDepth),
        insecure_ssl: cdktf.booleanToTerraform(struct.insecureSsl),
        location: cdktf.stringToTerraform(struct.location),
        report_build_status: cdktf.booleanToTerraform(struct.reportBuildStatus),
        type: cdktf.stringToTerraform(struct.type),
        auth: codebuildProjectSourceAuthToTerraform(struct.auth),
        build_status_config: codebuildProjectSourceBuildStatusConfigToTerraform(struct.buildStatusConfig),
        git_submodules_config: codebuildProjectSourceGitSubmodulesConfigToTerraform(struct.gitSubmodulesConfig),
    };
}
exports.codebuildProjectSourceToTerraform = codebuildProjectSourceToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectSourceOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // auth - computed: false, optional: true, required: false
        this._auth = new CodebuildProjectSourceAuthOutputReference(this, "auth", true);
        // build_status_config - computed: false, optional: true, required: false
        this._buildStatusConfig = new CodebuildProjectSourceBuildStatusConfigOutputReference(this, "build_status_config", true);
        // git_submodules_config - computed: false, optional: true, required: false
        this._gitSubmodulesConfig = new CodebuildProjectSourceGitSubmodulesConfigOutputReference(this, "git_submodules_config", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v, _w, _x, _y, _z;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._buildspec) {
            hasAnyValues = true;
            internalValueResult.buildspec = this._buildspec;
        }
        if (this._gitCloneDepth) {
            hasAnyValues = true;
            internalValueResult.gitCloneDepth = this._gitCloneDepth;
        }
        if (this._insecureSsl) {
            hasAnyValues = true;
            internalValueResult.insecureSsl = this._insecureSsl;
        }
        if (this._location) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._reportBuildStatus) {
            hasAnyValues = true;
            internalValueResult.reportBuildStatus = this._reportBuildStatus;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if ((_u = this._auth) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.auth = (_v = this._auth) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._buildStatusConfig) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.buildStatusConfig = (_x = this._buildStatusConfig) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        if ((_y = this._gitSubmodulesConfig) === null || _y === void 0 ? void 0 : _y.internalValue) {
            hasAnyValues = true;
            internalValueResult.gitSubmodulesConfig = (_z = this._gitSubmodulesConfig) === null || _z === void 0 ? void 0 : _z.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._buildspec = undefined;
            this._gitCloneDepth = undefined;
            this._insecureSsl = undefined;
            this._location = undefined;
            this._reportBuildStatus = undefined;
            this._type = undefined;
            this._auth.internalValue = undefined;
            this._buildStatusConfig.internalValue = undefined;
            this._gitSubmodulesConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._buildspec = value.buildspec;
            this._gitCloneDepth = value.gitCloneDepth;
            this._insecureSsl = value.insecureSsl;
            this._location = value.location;
            this._reportBuildStatus = value.reportBuildStatus;
            this._type = value.type;
            this._auth.internalValue = value.auth;
            this._buildStatusConfig.internalValue = value.buildStatusConfig;
            this._gitSubmodulesConfig.internalValue = value.gitSubmodulesConfig;
        }
    }
    /**
     * @stability stable
     */
    get buildspec() {
        return this.getStringAttribute('buildspec');
    }
    /**
     * @stability stable
     */
    set buildspec(value) {
        this._buildspec = value;
    }
    /**
     * @stability stable
     */
    resetBuildspec() {
        this._buildspec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildspecInput() {
        return this._buildspec;
    }
    /**
     * @stability stable
     */
    get gitCloneDepth() {
        return this.getNumberAttribute('git_clone_depth');
    }
    /**
     * @stability stable
     */
    set gitCloneDepth(value) {
        this._gitCloneDepth = value;
    }
    /**
     * @stability stable
     */
    resetGitCloneDepth() {
        this._gitCloneDepth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gitCloneDepthInput() {
        return this._gitCloneDepth;
    }
    /**
     * @stability stable
     */
    get insecureSsl() {
        return this.getBooleanAttribute('insecure_ssl');
    }
    /**
     * @stability stable
     */
    set insecureSsl(value) {
        this._insecureSsl = value;
    }
    /**
     * @stability stable
     */
    resetInsecureSsl() {
        this._insecureSsl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get insecureSslInput() {
        return this._insecureSsl;
    }
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    set location(value) {
        this._location = value;
    }
    /**
     * @stability stable
     */
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationInput() {
        return this._location;
    }
    /**
     * @stability stable
     */
    get reportBuildStatus() {
        return this.getBooleanAttribute('report_build_status');
    }
    /**
     * @stability stable
     */
    set reportBuildStatus(value) {
        this._reportBuildStatus = value;
    }
    /**
     * @stability stable
     */
    resetReportBuildStatus() {
        this._reportBuildStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reportBuildStatusInput() {
        return this._reportBuildStatus;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get auth() {
        return this._auth;
    }
    /**
     * @stability stable
     */
    putAuth(value) {
        this._auth.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAuth() {
        this._auth.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authInput() {
        return this._auth.internalValue;
    }
    /**
     * @stability stable
     */
    get buildStatusConfig() {
        return this._buildStatusConfig;
    }
    /**
     * @stability stable
     */
    putBuildStatusConfig(value) {
        this._buildStatusConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetBuildStatusConfig() {
        this._buildStatusConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildStatusConfigInput() {
        return this._buildStatusConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get gitSubmodulesConfig() {
        return this._gitSubmodulesConfig;
    }
    /**
     * @stability stable
     */
    putGitSubmodulesConfig(value) {
        this._gitSubmodulesConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetGitSubmodulesConfig() {
        this._gitSubmodulesConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gitSubmodulesConfigInput() {
        return this._gitSubmodulesConfig.internalValue;
    }
}
exports.CodebuildProjectSourceOutputReference = CodebuildProjectSourceOutputReference;
_r = JSII_RTTI_SYMBOL_1;
CodebuildProjectSourceOutputReference[_r] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectSourceOutputReference", version: "3.0.1" };
function codebuildProjectVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnets: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnets),
        vpc_id: cdktf.stringToTerraform(struct.vpcId),
    };
}
exports.codebuildProjectVpcConfigToTerraform = codebuildProjectVpcConfigToTerraform;
/**
 * @stability stable
 */
class CodebuildProjectVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnets) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        if (this._vpcId) {
            hasAnyValues = true;
            internalValueResult.vpcId = this._vpcId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnets = undefined;
            this._vpcId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnets = value.subnets;
            this._vpcId = value.vpcId;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnets() {
        return this.getListAttribute('subnets');
    }
    /**
     * @stability stable
     */
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetsInput() {
        return this._subnets;
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
}
exports.CodebuildProjectVpcConfigOutputReference = CodebuildProjectVpcConfigOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CodebuildProjectVpcConfigOutputReference[_s] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProjectVpcConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/codebuild_project.html aws_codebuild_project}.
 *
 * @stability stable
 */
class CodebuildProject extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codebuild_project.html aws_codebuild_project} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codebuild_project',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // artifacts - computed: false, optional: false, required: true
        this._artifacts = new CodebuildProjectArtifactsOutputReference(this, "artifacts", true);
        // build_batch_config - computed: false, optional: true, required: false
        this._buildBatchConfig = new CodebuildProjectBuildBatchConfigOutputReference(this, "build_batch_config", true);
        // cache - computed: false, optional: true, required: false
        this._cache = new CodebuildProjectCacheOutputReference(this, "cache", true);
        // environment - computed: false, optional: false, required: true
        this._environment = new CodebuildProjectEnvironmentOutputReference(this, "environment", true);
        // logs_config - computed: false, optional: true, required: false
        this._logsConfig = new CodebuildProjectLogsConfigOutputReference(this, "logs_config", true);
        // source - computed: false, optional: false, required: true
        this._source = new CodebuildProjectSourceOutputReference(this, "source", true);
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new CodebuildProjectVpcConfigOutputReference(this, "vpc_config", true);
        this._badgeEnabled = config.badgeEnabled;
        this._buildTimeout = config.buildTimeout;
        this._concurrentBuildLimit = config.concurrentBuildLimit;
        this._description = config.description;
        this._encryptionKey = config.encryptionKey;
        this._name = config.name;
        this._queuedTimeout = config.queuedTimeout;
        this._serviceRole = config.serviceRole;
        this._sourceVersion = config.sourceVersion;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._artifacts.internalValue = config.artifacts;
        this._buildBatchConfig.internalValue = config.buildBatchConfig;
        this._cache.internalValue = config.cache;
        this._environment.internalValue = config.environment;
        this._fileSystemLocations = config.fileSystemLocations;
        this._logsConfig.internalValue = config.logsConfig;
        this._secondaryArtifacts = config.secondaryArtifacts;
        this._secondarySources = config.secondarySources;
        this._source.internalValue = config.source;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get badgeEnabled() {
        return this.getBooleanAttribute('badge_enabled');
    }
    /**
     * @stability stable
     */
    set badgeEnabled(value) {
        this._badgeEnabled = value;
    }
    /**
     * @stability stable
     */
    resetBadgeEnabled() {
        this._badgeEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get badgeEnabledInput() {
        return this._badgeEnabled;
    }
    // badge_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get badgeUrl() {
        return this.getStringAttribute('badge_url');
    }
    /**
     * @stability stable
     */
    get buildTimeout() {
        return this.getNumberAttribute('build_timeout');
    }
    /**
     * @stability stable
     */
    set buildTimeout(value) {
        this._buildTimeout = value;
    }
    /**
     * @stability stable
     */
    resetBuildTimeout() {
        this._buildTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildTimeoutInput() {
        return this._buildTimeout;
    }
    /**
     * @stability stable
     */
    get concurrentBuildLimit() {
        return this.getNumberAttribute('concurrent_build_limit');
    }
    /**
     * @stability stable
     */
    set concurrentBuildLimit(value) {
        this._concurrentBuildLimit = value;
    }
    /**
     * @stability stable
     */
    resetConcurrentBuildLimit() {
        this._concurrentBuildLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get concurrentBuildLimitInput() {
        return this._concurrentBuildLimit;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get encryptionKey() {
        return this.getStringAttribute('encryption_key');
    }
    /**
     * @stability stable
     */
    set encryptionKey(value) {
        this._encryptionKey = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionKey() {
        this._encryptionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionKeyInput() {
        return this._encryptionKey;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get queuedTimeout() {
        return this.getNumberAttribute('queued_timeout');
    }
    /**
     * @stability stable
     */
    set queuedTimeout(value) {
        this._queuedTimeout = value;
    }
    /**
     * @stability stable
     */
    resetQueuedTimeout() {
        this._queuedTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queuedTimeoutInput() {
        return this._queuedTimeout;
    }
    /**
     * @stability stable
     */
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    /**
     * @stability stable
     */
    set serviceRole(value) {
        this._serviceRole = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceRoleInput() {
        return this._serviceRole;
    }
    /**
     * @stability stable
     */
    get sourceVersion() {
        return this.getStringAttribute('source_version');
    }
    /**
     * @stability stable
     */
    set sourceVersion(value) {
        this._sourceVersion = value;
    }
    /**
     * @stability stable
     */
    resetSourceVersion() {
        this._sourceVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceVersionInput() {
        return this._sourceVersion;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get artifacts() {
        return this._artifacts;
    }
    /**
     * @stability stable
     */
    putArtifacts(value) {
        this._artifacts.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get artifactsInput() {
        return this._artifacts.internalValue;
    }
    /**
     * @stability stable
     */
    get buildBatchConfig() {
        return this._buildBatchConfig;
    }
    /**
     * @stability stable
     */
    putBuildBatchConfig(value) {
        this._buildBatchConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetBuildBatchConfig() {
        this._buildBatchConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildBatchConfigInput() {
        return this._buildBatchConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get cache() {
        return this._cache;
    }
    /**
     * @stability stable
     */
    putCache(value) {
        this._cache.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCache() {
        this._cache.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cacheInput() {
        return this._cache.internalValue;
    }
    /**
     * @stability stable
     */
    get environment() {
        return this._environment;
    }
    /**
     * @stability stable
     */
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentInput() {
        return this._environment.internalValue;
    }
    /**
     * @stability stable
     */
    get fileSystemLocations() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('file_system_locations');
    }
    /**
     * @stability stable
     */
    set fileSystemLocations(value) {
        this._fileSystemLocations = value;
    }
    /**
     * @stability stable
     */
    resetFileSystemLocations() {
        this._fileSystemLocations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileSystemLocationsInput() {
        return this._fileSystemLocations;
    }
    /**
     * @stability stable
     */
    get logsConfig() {
        return this._logsConfig;
    }
    /**
     * @stability stable
     */
    putLogsConfig(value) {
        this._logsConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLogsConfig() {
        this._logsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logsConfigInput() {
        return this._logsConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get secondaryArtifacts() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('secondary_artifacts');
    }
    /**
     * @stability stable
     */
    set secondaryArtifacts(value) {
        this._secondaryArtifacts = value;
    }
    /**
     * @stability stable
     */
    resetSecondaryArtifacts() {
        this._secondaryArtifacts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secondaryArtifactsInput() {
        return this._secondaryArtifacts;
    }
    /**
     * @stability stable
     */
    get secondarySources() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('secondary_sources');
    }
    /**
     * @stability stable
     */
    set secondarySources(value) {
        this._secondarySources = value;
    }
    /**
     * @stability stable
     */
    resetSecondarySources() {
        this._secondarySources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secondarySourcesInput() {
        return this._secondarySources;
    }
    /**
     * @stability stable
     */
    get source() {
        return this._source;
    }
    /**
     * @stability stable
     */
    putSource(value) {
        this._source.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceInput() {
        return this._source.internalValue;
    }
    /**
     * @stability stable
     */
    get vpcConfig() {
        return this._vpcConfig;
    }
    /**
     * @stability stable
     */
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            badge_enabled: cdktf.booleanToTerraform(this._badgeEnabled),
            build_timeout: cdktf.numberToTerraform(this._buildTimeout),
            concurrent_build_limit: cdktf.numberToTerraform(this._concurrentBuildLimit),
            description: cdktf.stringToTerraform(this._description),
            encryption_key: cdktf.stringToTerraform(this._encryptionKey),
            name: cdktf.stringToTerraform(this._name),
            queued_timeout: cdktf.numberToTerraform(this._queuedTimeout),
            service_role: cdktf.stringToTerraform(this._serviceRole),
            source_version: cdktf.stringToTerraform(this._sourceVersion),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            artifacts: codebuildProjectArtifactsToTerraform(this._artifacts.internalValue),
            build_batch_config: codebuildProjectBuildBatchConfigToTerraform(this._buildBatchConfig.internalValue),
            cache: codebuildProjectCacheToTerraform(this._cache.internalValue),
            environment: codebuildProjectEnvironmentToTerraform(this._environment.internalValue),
            file_system_locations: cdktf.listMapper(codebuildProjectFileSystemLocationsToTerraform)(this._fileSystemLocations),
            logs_config: codebuildProjectLogsConfigToTerraform(this._logsConfig.internalValue),
            secondary_artifacts: cdktf.listMapper(codebuildProjectSecondaryArtifactsToTerraform)(this._secondaryArtifacts),
            secondary_sources: cdktf.listMapper(codebuildProjectSecondarySourcesToTerraform)(this._secondarySources),
            source: codebuildProjectSourceToTerraform(this._source.internalValue),
            vpc_config: codebuildProjectVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
}
exports.CodebuildProject = CodebuildProject;
_t = JSII_RTTI_SYMBOL_1;
CodebuildProject[_t] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildProject", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CodebuildProject.tfResourceType = "aws_codebuild_project";
