// @cdktf/provider-aws/lib/codebuild/codebuild-source-credential.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodebuildSourceCredential = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/codebuild_source_credential.html aws_codebuild_source_credential}.
 *
 * @stability stable
 */
class CodebuildSourceCredential extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codebuild_source_credential.html aws_codebuild_source_credential} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codebuild_source_credential',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._authType = config.authType;
        this._serverType = config.serverType;
        this._token = config.token;
        this._userName = config.userName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get authType() {
        return this.getStringAttribute('auth_type');
    }
    /**
     * @stability stable
     */
    set authType(value) {
        this._authType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authTypeInput() {
        return this._authType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get serverType() {
        return this.getStringAttribute('server_type');
    }
    /**
     * @stability stable
     */
    set serverType(value) {
        this._serverType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serverTypeInput() {
        return this._serverType;
    }
    /**
     * @stability stable
     */
    get token() {
        return this.getStringAttribute('token');
    }
    /**
     * @stability stable
     */
    set token(value) {
        this._token = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tokenInput() {
        return this._token;
    }
    /**
     * @stability stable
     */
    get userName() {
        return this.getStringAttribute('user_name');
    }
    /**
     * @stability stable
     */
    set userName(value) {
        this._userName = value;
    }
    /**
     * @stability stable
     */
    resetUserName() {
        this._userName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userNameInput() {
        return this._userName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            auth_type: cdktf.stringToTerraform(this._authType),
            server_type: cdktf.stringToTerraform(this._serverType),
            token: cdktf.stringToTerraform(this._token),
            user_name: cdktf.stringToTerraform(this._userName),
        };
    }
}
exports.CodebuildSourceCredential = CodebuildSourceCredential;
_a = JSII_RTTI_SYMBOL_1;
CodebuildSourceCredential[_a] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildSourceCredential", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CodebuildSourceCredential.tfResourceType = "aws_codebuild_source_credential";
