// @cdktf/provider-aws/lib/msk/msk-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskCluster = exports.MskClusterTimeoutsOutputReference = exports.mskClusterTimeoutsToTerraform = exports.MskClusterOpenMonitoringOutputReference = exports.mskClusterOpenMonitoringToTerraform = exports.MskClusterOpenMonitoringPrometheusOutputReference = exports.mskClusterOpenMonitoringPrometheusToTerraform = exports.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference = exports.mskClusterOpenMonitoringPrometheusNodeExporterToTerraform = exports.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference = exports.mskClusterOpenMonitoringPrometheusJmxExporterToTerraform = exports.MskClusterLoggingInfoOutputReference = exports.mskClusterLoggingInfoToTerraform = exports.MskClusterLoggingInfoBrokerLogsOutputReference = exports.mskClusterLoggingInfoBrokerLogsToTerraform = exports.MskClusterLoggingInfoBrokerLogsS3OutputReference = exports.mskClusterLoggingInfoBrokerLogsS3ToTerraform = exports.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference = exports.mskClusterLoggingInfoBrokerLogsFirehoseToTerraform = exports.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference = exports.mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform = exports.MskClusterEncryptionInfoOutputReference = exports.mskClusterEncryptionInfoToTerraform = exports.MskClusterEncryptionInfoEncryptionInTransitOutputReference = exports.mskClusterEncryptionInfoEncryptionInTransitToTerraform = exports.MskClusterConfigurationInfoOutputReference = exports.mskClusterConfigurationInfoToTerraform = exports.MskClusterClientAuthenticationOutputReference = exports.mskClusterClientAuthenticationToTerraform = exports.MskClusterClientAuthenticationTlsOutputReference = exports.mskClusterClientAuthenticationTlsToTerraform = exports.MskClusterClientAuthenticationSaslOutputReference = exports.mskClusterClientAuthenticationSaslToTerraform = exports.MskClusterBrokerNodeGroupInfoOutputReference = exports.mskClusterBrokerNodeGroupInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mskClusterBrokerNodeGroupInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        az_distribution: cdktf.stringToTerraform(struct.azDistribution),
        client_subnets: cdktf.listMapper(cdktf.stringToTerraform)(struct.clientSubnets),
        ebs_volume_size: cdktf.numberToTerraform(struct.ebsVolumeSize),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroups),
    };
}
exports.mskClusterBrokerNodeGroupInfoToTerraform = mskClusterBrokerNodeGroupInfoToTerraform;
/**
 * @stability stable
 */
class MskClusterBrokerNodeGroupInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._azDistribution) {
            hasAnyValues = true;
            internalValueResult.azDistribution = this._azDistribution;
        }
        if (this._clientSubnets) {
            hasAnyValues = true;
            internalValueResult.clientSubnets = this._clientSubnets;
        }
        if (this._ebsVolumeSize) {
            hasAnyValues = true;
            internalValueResult.ebsVolumeSize = this._ebsVolumeSize;
        }
        if (this._instanceType) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._securityGroups) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._azDistribution = undefined;
            this._clientSubnets = undefined;
            this._ebsVolumeSize = undefined;
            this._instanceType = undefined;
            this._securityGroups = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._azDistribution = value.azDistribution;
            this._clientSubnets = value.clientSubnets;
            this._ebsVolumeSize = value.ebsVolumeSize;
            this._instanceType = value.instanceType;
            this._securityGroups = value.securityGroups;
        }
    }
    /**
     * @stability stable
     */
    get azDistribution() {
        return this.getStringAttribute('az_distribution');
    }
    /**
     * @stability stable
     */
    set azDistribution(value) {
        this._azDistribution = value;
    }
    /**
     * @stability stable
     */
    resetAzDistribution() {
        this._azDistribution = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get azDistributionInput() {
        return this._azDistribution;
    }
    /**
     * @stability stable
     */
    get clientSubnets() {
        return this.getListAttribute('client_subnets');
    }
    /**
     * @stability stable
     */
    set clientSubnets(value) {
        this._clientSubnets = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientSubnetsInput() {
        return this._clientSubnets;
    }
    /**
     * @stability stable
     */
    get ebsVolumeSize() {
        return this.getNumberAttribute('ebs_volume_size');
    }
    /**
     * @stability stable
     */
    set ebsVolumeSize(value) {
        this._ebsVolumeSize = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsVolumeSizeInput() {
        return this._ebsVolumeSize;
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    /**
     * @stability stable
     */
    set securityGroups(value) {
        this._securityGroups = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupsInput() {
        return this._securityGroups;
    }
}
exports.MskClusterBrokerNodeGroupInfoOutputReference = MskClusterBrokerNodeGroupInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoOutputReference[_a] = { fqn: "@cdktf/provider-aws.msk.MskClusterBrokerNodeGroupInfoOutputReference", version: "3.0.1" };
function mskClusterClientAuthenticationSaslToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam: cdktf.booleanToTerraform(struct.iam),
        scram: cdktf.booleanToTerraform(struct.scram),
    };
}
exports.mskClusterClientAuthenticationSaslToTerraform = mskClusterClientAuthenticationSaslToTerraform;
/**
 * @stability stable
 */
class MskClusterClientAuthenticationSaslOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iam) {
            hasAnyValues = true;
            internalValueResult.iam = this._iam;
        }
        if (this._scram) {
            hasAnyValues = true;
            internalValueResult.scram = this._scram;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iam = undefined;
            this._scram = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iam = value.iam;
            this._scram = value.scram;
        }
    }
    /**
     * @stability stable
     */
    get iam() {
        return this.getBooleanAttribute('iam');
    }
    /**
     * @stability stable
     */
    set iam(value) {
        this._iam = value;
    }
    /**
     * @stability stable
     */
    resetIam() {
        this._iam = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamInput() {
        return this._iam;
    }
    /**
     * @stability stable
     */
    get scram() {
        return this.getBooleanAttribute('scram');
    }
    /**
     * @stability stable
     */
    set scram(value) {
        this._scram = value;
    }
    /**
     * @stability stable
     */
    resetScram() {
        this._scram = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scramInput() {
        return this._scram;
    }
}
exports.MskClusterClientAuthenticationSaslOutputReference = MskClusterClientAuthenticationSaslOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationSaslOutputReference[_b] = { fqn: "@cdktf/provider-aws.msk.MskClusterClientAuthenticationSaslOutputReference", version: "3.0.1" };
function mskClusterClientAuthenticationTlsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_authority_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.certificateAuthorityArns),
    };
}
exports.mskClusterClientAuthenticationTlsToTerraform = mskClusterClientAuthenticationTlsToTerraform;
/**
 * @stability stable
 */
class MskClusterClientAuthenticationTlsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateAuthorityArns) {
            hasAnyValues = true;
            internalValueResult.certificateAuthorityArns = this._certificateAuthorityArns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateAuthorityArns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateAuthorityArns = value.certificateAuthorityArns;
        }
    }
    /**
     * @stability stable
     */
    get certificateAuthorityArns() {
        return this.getListAttribute('certificate_authority_arns');
    }
    /**
     * @stability stable
     */
    set certificateAuthorityArns(value) {
        this._certificateAuthorityArns = value;
    }
    /**
     * @stability stable
     */
    resetCertificateAuthorityArns() {
        this._certificateAuthorityArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateAuthorityArnsInput() {
        return this._certificateAuthorityArns;
    }
}
exports.MskClusterClientAuthenticationTlsOutputReference = MskClusterClientAuthenticationTlsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationTlsOutputReference[_c] = { fqn: "@cdktf/provider-aws.msk.MskClusterClientAuthenticationTlsOutputReference", version: "3.0.1" };
function mskClusterClientAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sasl: mskClusterClientAuthenticationSaslToTerraform(struct.sasl),
        tls: mskClusterClientAuthenticationTlsToTerraform(struct.tls),
    };
}
exports.mskClusterClientAuthenticationToTerraform = mskClusterClientAuthenticationToTerraform;
/**
 * @stability stable
 */
class MskClusterClientAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // sasl - computed: false, optional: true, required: false
        this._sasl = new MskClusterClientAuthenticationSaslOutputReference(this, "sasl", true);
        // tls - computed: false, optional: true, required: false
        this._tls = new MskClusterClientAuthenticationTlsOutputReference(this, "tls", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v, _w, _x;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._sasl) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.sasl = (_v = this._sasl) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._tls) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.tls = (_x = this._tls) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sasl.internalValue = undefined;
            this._tls.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sasl.internalValue = value.sasl;
            this._tls.internalValue = value.tls;
        }
    }
    /**
     * @stability stable
     */
    get sasl() {
        return this._sasl;
    }
    /**
     * @stability stable
     */
    putSasl(value) {
        this._sasl.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSasl() {
        this._sasl.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get saslInput() {
        return this._sasl.internalValue;
    }
    /**
     * @stability stable
     */
    get tls() {
        return this._tls;
    }
    /**
     * @stability stable
     */
    putTls(value) {
        this._tls.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTls() {
        this._tls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsInput() {
        return this._tls.internalValue;
    }
}
exports.MskClusterClientAuthenticationOutputReference = MskClusterClientAuthenticationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationOutputReference[_d] = { fqn: "@cdktf/provider-aws.msk.MskClusterClientAuthenticationOutputReference", version: "3.0.1" };
function mskClusterConfigurationInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mskClusterConfigurationInfoToTerraform = mskClusterConfigurationInfoToTerraform;
/**
 * @stability stable
 */
class MskClusterConfigurationInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._revision) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._revision = value.revision;
        }
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    /**
     * @stability stable
     */
    get revision() {
        return this.getNumberAttribute('revision');
    }
    /**
     * @stability stable
     */
    set revision(value) {
        this._revision = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get revisionInput() {
        return this._revision;
    }
}
exports.MskClusterConfigurationInfoOutputReference = MskClusterConfigurationInfoOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MskClusterConfigurationInfoOutputReference[_e] = { fqn: "@cdktf/provider-aws.msk.MskClusterConfigurationInfoOutputReference", version: "3.0.1" };
function mskClusterEncryptionInfoEncryptionInTransitToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_broker: cdktf.stringToTerraform(struct.clientBroker),
        in_cluster: cdktf.booleanToTerraform(struct.inCluster),
    };
}
exports.mskClusterEncryptionInfoEncryptionInTransitToTerraform = mskClusterEncryptionInfoEncryptionInTransitToTerraform;
/**
 * @stability stable
 */
class MskClusterEncryptionInfoEncryptionInTransitOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientBroker) {
            hasAnyValues = true;
            internalValueResult.clientBroker = this._clientBroker;
        }
        if (this._inCluster) {
            hasAnyValues = true;
            internalValueResult.inCluster = this._inCluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientBroker = undefined;
            this._inCluster = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientBroker = value.clientBroker;
            this._inCluster = value.inCluster;
        }
    }
    /**
     * @stability stable
     */
    get clientBroker() {
        return this.getStringAttribute('client_broker');
    }
    /**
     * @stability stable
     */
    set clientBroker(value) {
        this._clientBroker = value;
    }
    /**
     * @stability stable
     */
    resetClientBroker() {
        this._clientBroker = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientBrokerInput() {
        return this._clientBroker;
    }
    /**
     * @stability stable
     */
    get inCluster() {
        return this.getBooleanAttribute('in_cluster');
    }
    /**
     * @stability stable
     */
    set inCluster(value) {
        this._inCluster = value;
    }
    /**
     * @stability stable
     */
    resetInCluster() {
        this._inCluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inClusterInput() {
        return this._inCluster;
    }
}
exports.MskClusterEncryptionInfoEncryptionInTransitOutputReference = MskClusterEncryptionInfoEncryptionInTransitOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MskClusterEncryptionInfoEncryptionInTransitOutputReference[_f] = { fqn: "@cdktf/provider-aws.msk.MskClusterEncryptionInfoEncryptionInTransitOutputReference", version: "3.0.1" };
function mskClusterEncryptionInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_at_rest_kms_key_arn: cdktf.stringToTerraform(struct.encryptionAtRestKmsKeyArn),
        encryption_in_transit: mskClusterEncryptionInfoEncryptionInTransitToTerraform(struct.encryptionInTransit),
    };
}
exports.mskClusterEncryptionInfoToTerraform = mskClusterEncryptionInfoToTerraform;
/**
 * @stability stable
 */
class MskClusterEncryptionInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // encryption_in_transit - computed: false, optional: true, required: false
        this._encryptionInTransit = new MskClusterEncryptionInfoEncryptionInTransitOutputReference(this, "encryption_in_transit", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionAtRestKmsKeyArn) {
            hasAnyValues = true;
            internalValueResult.encryptionAtRestKmsKeyArn = this._encryptionAtRestKmsKeyArn;
        }
        if ((_u = this._encryptionInTransit) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.encryptionInTransit = (_v = this._encryptionInTransit) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionAtRestKmsKeyArn = undefined;
            this._encryptionInTransit.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionAtRestKmsKeyArn = value.encryptionAtRestKmsKeyArn;
            this._encryptionInTransit.internalValue = value.encryptionInTransit;
        }
    }
    /**
     * @stability stable
     */
    get encryptionAtRestKmsKeyArn() {
        return this.getStringAttribute('encryption_at_rest_kms_key_arn');
    }
    /**
     * @stability stable
     */
    set encryptionAtRestKmsKeyArn(value) {
        this._encryptionAtRestKmsKeyArn = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionAtRestKmsKeyArn() {
        this._encryptionAtRestKmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionAtRestKmsKeyArnInput() {
        return this._encryptionAtRestKmsKeyArn;
    }
    /**
     * @stability stable
     */
    get encryptionInTransit() {
        return this._encryptionInTransit;
    }
    /**
     * @stability stable
     */
    putEncryptionInTransit(value) {
        this._encryptionInTransit.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionInTransit() {
        this._encryptionInTransit.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionInTransitInput() {
        return this._encryptionInTransit.internalValue;
    }
}
exports.MskClusterEncryptionInfoOutputReference = MskClusterEncryptionInfoOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MskClusterEncryptionInfoOutputReference[_g] = { fqn: "@cdktf/provider-aws.msk.MskClusterEncryptionInfoOutputReference", version: "3.0.1" };
function mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_group: cdktf.stringToTerraform(struct.logGroup),
    };
}
exports.mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform = mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform;
/**
 * @stability stable
 */
class MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logGroup) {
            hasAnyValues = true;
            internalValueResult.logGroup = this._logGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logGroup = value.logGroup;
        }
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    /**
     * @stability stable
     */
    set logGroup(value) {
        this._logGroup = value;
    }
    /**
     * @stability stable
     */
    resetLogGroup() {
        this._logGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logGroupInput() {
        return this._logGroup;
    }
}
exports.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference = MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference[_h] = { fqn: "@cdktf/provider-aws.msk.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference", version: "3.0.1" };
function mskClusterLoggingInfoBrokerLogsFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream: cdktf.stringToTerraform(struct.deliveryStream),
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.mskClusterLoggingInfoBrokerLogsFirehoseToTerraform = mskClusterLoggingInfoBrokerLogsFirehoseToTerraform;
/**
 * @stability stable
 */
class MskClusterLoggingInfoBrokerLogsFirehoseOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deliveryStream) {
            hasAnyValues = true;
            internalValueResult.deliveryStream = this._deliveryStream;
        }
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deliveryStream = undefined;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deliveryStream = value.deliveryStream;
            this._enabled = value.enabled;
        }
    }
    /**
     * @stability stable
     */
    get deliveryStream() {
        return this.getStringAttribute('delivery_stream');
    }
    /**
     * @stability stable
     */
    set deliveryStream(value) {
        this._deliveryStream = value;
    }
    /**
     * @stability stable
     */
    resetDeliveryStream() {
        this._deliveryStream = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deliveryStreamInput() {
        return this._deliveryStream;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
}
exports.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference = MskClusterLoggingInfoBrokerLogsFirehoseOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsFirehoseOutputReference[_j] = { fqn: "@cdktf/provider-aws.msk.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference", version: "3.0.1" };
function mskClusterLoggingInfoBrokerLogsS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.mskClusterLoggingInfoBrokerLogsS3ToTerraform = mskClusterLoggingInfoBrokerLogsS3ToTerraform;
/**
 * @stability stable
 */
class MskClusterLoggingInfoBrokerLogsS3OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._enabled) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._enabled = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._enabled = value.enabled;
            this._prefix = value.prefix;
        }
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    /**
     * @stability stable
     */
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
}
exports.MskClusterLoggingInfoBrokerLogsS3OutputReference = MskClusterLoggingInfoBrokerLogsS3OutputReference;
_k = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsS3OutputReference[_k] = { fqn: "@cdktf/provider-aws.msk.MskClusterLoggingInfoBrokerLogsS3OutputReference", version: "3.0.1" };
function mskClusterLoggingInfoBrokerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_logs: mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform(struct.cloudwatchLogs),
        firehose: mskClusterLoggingInfoBrokerLogsFirehoseToTerraform(struct.firehose),
        s3: mskClusterLoggingInfoBrokerLogsS3ToTerraform(struct.s3),
    };
}
exports.mskClusterLoggingInfoBrokerLogsToTerraform = mskClusterLoggingInfoBrokerLogsToTerraform;
/**
 * @stability stable
 */
class MskClusterLoggingInfoBrokerLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cloudwatch_logs - computed: false, optional: true, required: false
        this._cloudwatchLogs = new MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference(this, "cloudwatch_logs", true);
        // firehose - computed: false, optional: true, required: false
        this._firehose = new MskClusterLoggingInfoBrokerLogsFirehoseOutputReference(this, "firehose", true);
        // s3 - computed: false, optional: true, required: false
        this._s3 = new MskClusterLoggingInfoBrokerLogsS3OutputReference(this, "s3", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v, _w, _x, _y, _z;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._cloudwatchLogs) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogs = (_v = this._cloudwatchLogs) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._firehose) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.firehose = (_x = this._firehose) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        if ((_y = this._s3) === null || _y === void 0 ? void 0 : _y.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3 = (_z = this._s3) === null || _z === void 0 ? void 0 : _z.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchLogs.internalValue = undefined;
            this._firehose.internalValue = undefined;
            this._s3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchLogs.internalValue = value.cloudwatchLogs;
            this._firehose.internalValue = value.firehose;
            this._s3.internalValue = value.s3;
        }
    }
    /**
     * @stability stable
     */
    get cloudwatchLogs() {
        return this._cloudwatchLogs;
    }
    /**
     * @stability stable
     */
    putCloudwatchLogs(value) {
        this._cloudwatchLogs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchLogs() {
        this._cloudwatchLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchLogsInput() {
        return this._cloudwatchLogs.internalValue;
    }
    /**
     * @stability stable
     */
    get firehose() {
        return this._firehose;
    }
    /**
     * @stability stable
     */
    putFirehose(value) {
        this._firehose.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFirehose() {
        this._firehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get firehoseInput() {
        return this._firehose.internalValue;
    }
    /**
     * @stability stable
     */
    get s3() {
        return this._s3;
    }
    /**
     * @stability stable
     */
    putS3(value) {
        this._s3.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3Input() {
        return this._s3.internalValue;
    }
}
exports.MskClusterLoggingInfoBrokerLogsOutputReference = MskClusterLoggingInfoBrokerLogsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsOutputReference[_l] = { fqn: "@cdktf/provider-aws.msk.MskClusterLoggingInfoBrokerLogsOutputReference", version: "3.0.1" };
function mskClusterLoggingInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        broker_logs: mskClusterLoggingInfoBrokerLogsToTerraform(struct.brokerLogs),
    };
}
exports.mskClusterLoggingInfoToTerraform = mskClusterLoggingInfoToTerraform;
/**
 * @stability stable
 */
class MskClusterLoggingInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // broker_logs - computed: false, optional: false, required: true
        this._brokerLogs = new MskClusterLoggingInfoBrokerLogsOutputReference(this, "broker_logs", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._brokerLogs) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.brokerLogs = (_v = this._brokerLogs) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._brokerLogs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._brokerLogs.internalValue = value.brokerLogs;
        }
    }
    /**
     * @stability stable
     */
    get brokerLogs() {
        return this._brokerLogs;
    }
    /**
     * @stability stable
     */
    putBrokerLogs(value) {
        this._brokerLogs.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get brokerLogsInput() {
        return this._brokerLogs.internalValue;
    }
}
exports.MskClusterLoggingInfoOutputReference = MskClusterLoggingInfoOutputReference;
_m = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoOutputReference[_m] = { fqn: "@cdktf/provider-aws.msk.MskClusterLoggingInfoOutputReference", version: "3.0.1" };
function mskClusterOpenMonitoringPrometheusJmxExporterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled_in_broker: cdktf.booleanToTerraform(struct.enabledInBroker),
    };
}
exports.mskClusterOpenMonitoringPrometheusJmxExporterToTerraform = mskClusterOpenMonitoringPrometheusJmxExporterToTerraform;
/**
 * @stability stable
 */
class MskClusterOpenMonitoringPrometheusJmxExporterOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabledInBroker) {
            hasAnyValues = true;
            internalValueResult.enabledInBroker = this._enabledInBroker;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabledInBroker = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabledInBroker = value.enabledInBroker;
        }
    }
    /**
     * @stability stable
     */
    get enabledInBroker() {
        return this.getBooleanAttribute('enabled_in_broker');
    }
    /**
     * @stability stable
     */
    set enabledInBroker(value) {
        this._enabledInBroker = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInBrokerInput() {
        return this._enabledInBroker;
    }
}
exports.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference = MskClusterOpenMonitoringPrometheusJmxExporterOutputReference;
_o = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusJmxExporterOutputReference[_o] = { fqn: "@cdktf/provider-aws.msk.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference", version: "3.0.1" };
function mskClusterOpenMonitoringPrometheusNodeExporterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled_in_broker: cdktf.booleanToTerraform(struct.enabledInBroker),
    };
}
exports.mskClusterOpenMonitoringPrometheusNodeExporterToTerraform = mskClusterOpenMonitoringPrometheusNodeExporterToTerraform;
/**
 * @stability stable
 */
class MskClusterOpenMonitoringPrometheusNodeExporterOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabledInBroker) {
            hasAnyValues = true;
            internalValueResult.enabledInBroker = this._enabledInBroker;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabledInBroker = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabledInBroker = value.enabledInBroker;
        }
    }
    /**
     * @stability stable
     */
    get enabledInBroker() {
        return this.getBooleanAttribute('enabled_in_broker');
    }
    /**
     * @stability stable
     */
    set enabledInBroker(value) {
        this._enabledInBroker = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInBrokerInput() {
        return this._enabledInBroker;
    }
}
exports.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference = MskClusterOpenMonitoringPrometheusNodeExporterOutputReference;
_p = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusNodeExporterOutputReference[_p] = { fqn: "@cdktf/provider-aws.msk.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference", version: "3.0.1" };
function mskClusterOpenMonitoringPrometheusToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        jmx_exporter: mskClusterOpenMonitoringPrometheusJmxExporterToTerraform(struct.jmxExporter),
        node_exporter: mskClusterOpenMonitoringPrometheusNodeExporterToTerraform(struct.nodeExporter),
    };
}
exports.mskClusterOpenMonitoringPrometheusToTerraform = mskClusterOpenMonitoringPrometheusToTerraform;
/**
 * @stability stable
 */
class MskClusterOpenMonitoringPrometheusOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // jmx_exporter - computed: false, optional: true, required: false
        this._jmxExporter = new MskClusterOpenMonitoringPrometheusJmxExporterOutputReference(this, "jmx_exporter", true);
        // node_exporter - computed: false, optional: true, required: false
        this._nodeExporter = new MskClusterOpenMonitoringPrometheusNodeExporterOutputReference(this, "node_exporter", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v, _w, _x;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._jmxExporter) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.jmxExporter = (_v = this._jmxExporter) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        if ((_w = this._nodeExporter) === null || _w === void 0 ? void 0 : _w.internalValue) {
            hasAnyValues = true;
            internalValueResult.nodeExporter = (_x = this._nodeExporter) === null || _x === void 0 ? void 0 : _x.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._jmxExporter.internalValue = undefined;
            this._nodeExporter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._jmxExporter.internalValue = value.jmxExporter;
            this._nodeExporter.internalValue = value.nodeExporter;
        }
    }
    /**
     * @stability stable
     */
    get jmxExporter() {
        return this._jmxExporter;
    }
    /**
     * @stability stable
     */
    putJmxExporter(value) {
        this._jmxExporter.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetJmxExporter() {
        this._jmxExporter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jmxExporterInput() {
        return this._jmxExporter.internalValue;
    }
    /**
     * @stability stable
     */
    get nodeExporter() {
        return this._nodeExporter;
    }
    /**
     * @stability stable
     */
    putNodeExporter(value) {
        this._nodeExporter.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNodeExporter() {
        this._nodeExporter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nodeExporterInput() {
        return this._nodeExporter.internalValue;
    }
}
exports.MskClusterOpenMonitoringPrometheusOutputReference = MskClusterOpenMonitoringPrometheusOutputReference;
_q = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusOutputReference[_q] = { fqn: "@cdktf/provider-aws.msk.MskClusterOpenMonitoringPrometheusOutputReference", version: "3.0.1" };
function mskClusterOpenMonitoringToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prometheus: mskClusterOpenMonitoringPrometheusToTerraform(struct.prometheus),
    };
}
exports.mskClusterOpenMonitoringToTerraform = mskClusterOpenMonitoringToTerraform;
/**
 * @stability stable
 */
class MskClusterOpenMonitoringOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // prometheus - computed: false, optional: false, required: true
        this._prometheus = new MskClusterOpenMonitoringPrometheusOutputReference(this, "prometheus", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _u, _v;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_u = this._prometheus) === null || _u === void 0 ? void 0 : _u.internalValue) {
            hasAnyValues = true;
            internalValueResult.prometheus = (_v = this._prometheus) === null || _v === void 0 ? void 0 : _v.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prometheus.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prometheus.internalValue = value.prometheus;
        }
    }
    /**
     * @stability stable
     */
    get prometheus() {
        return this._prometheus;
    }
    /**
     * @stability stable
     */
    putPrometheus(value) {
        this._prometheus.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prometheusInput() {
        return this._prometheus.internalValue;
    }
}
exports.MskClusterOpenMonitoringOutputReference = MskClusterOpenMonitoringOutputReference;
_r = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringOutputReference[_r] = { fqn: "@cdktf/provider-aws.msk.MskClusterOpenMonitoringOutputReference", version: "3.0.1" };
function mskClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mskClusterTimeoutsToTerraform = mskClusterTimeoutsToTerraform;
/**
 * @stability stable
 */
class MskClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.MskClusterTimeoutsOutputReference = MskClusterTimeoutsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
MskClusterTimeoutsOutputReference[_s] = { fqn: "@cdktf/provider-aws.msk.MskClusterTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_cluster.html aws_msk_cluster}.
 *
 * @stability stable
 */
class MskCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_cluster.html aws_msk_cluster} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // broker_node_group_info - computed: false, optional: false, required: true
        this._brokerNodeGroupInfo = new MskClusterBrokerNodeGroupInfoOutputReference(this, "broker_node_group_info", true);
        // client_authentication - computed: false, optional: true, required: false
        this._clientAuthentication = new MskClusterClientAuthenticationOutputReference(this, "client_authentication", true);
        // configuration_info - computed: false, optional: true, required: false
        this._configurationInfo = new MskClusterConfigurationInfoOutputReference(this, "configuration_info", true);
        // encryption_info - computed: false, optional: true, required: false
        this._encryptionInfo = new MskClusterEncryptionInfoOutputReference(this, "encryption_info", true);
        // logging_info - computed: false, optional: true, required: false
        this._loggingInfo = new MskClusterLoggingInfoOutputReference(this, "logging_info", true);
        // open_monitoring - computed: false, optional: true, required: false
        this._openMonitoring = new MskClusterOpenMonitoringOutputReference(this, "open_monitoring", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MskClusterTimeoutsOutputReference(this, "timeouts", true);
        this._clusterName = config.clusterName;
        this._enhancedMonitoring = config.enhancedMonitoring;
        this._kafkaVersion = config.kafkaVersion;
        this._numberOfBrokerNodes = config.numberOfBrokerNodes;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._brokerNodeGroupInfo.internalValue = config.brokerNodeGroupInfo;
        this._clientAuthentication.internalValue = config.clientAuthentication;
        this._configurationInfo.internalValue = config.configurationInfo;
        this._encryptionInfo.internalValue = config.encryptionInfo;
        this._loggingInfo.internalValue = config.loggingInfo;
        this._openMonitoring.internalValue = config.openMonitoring;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // bootstrap_brokers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bootstrapBrokers() {
        return this.getStringAttribute('bootstrap_brokers');
    }
    // bootstrap_brokers_sasl_iam - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bootstrapBrokersSaslIam() {
        return this.getStringAttribute('bootstrap_brokers_sasl_iam');
    }
    // bootstrap_brokers_sasl_scram - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bootstrapBrokersSaslScram() {
        return this.getStringAttribute('bootstrap_brokers_sasl_scram');
    }
    // bootstrap_brokers_tls - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bootstrapBrokersTls() {
        return this.getStringAttribute('bootstrap_brokers_tls');
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    // current_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get currentVersion() {
        return this.getStringAttribute('current_version');
    }
    /**
     * @stability stable
     */
    get enhancedMonitoring() {
        return this.getStringAttribute('enhanced_monitoring');
    }
    /**
     * @stability stable
     */
    set enhancedMonitoring(value) {
        this._enhancedMonitoring = value;
    }
    /**
     * @stability stable
     */
    resetEnhancedMonitoring() {
        this._enhancedMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enhancedMonitoringInput() {
        return this._enhancedMonitoring;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kafkaVersion() {
        return this.getStringAttribute('kafka_version');
    }
    /**
     * @stability stable
     */
    set kafkaVersion(value) {
        this._kafkaVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kafkaVersionInput() {
        return this._kafkaVersion;
    }
    /**
     * @stability stable
     */
    get numberOfBrokerNodes() {
        return this.getNumberAttribute('number_of_broker_nodes');
    }
    /**
     * @stability stable
     */
    set numberOfBrokerNodes(value) {
        this._numberOfBrokerNodes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfBrokerNodesInput() {
        return this._numberOfBrokerNodes;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // zookeeper_connect_string - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zookeeperConnectString() {
        return this.getStringAttribute('zookeeper_connect_string');
    }
    // zookeeper_connect_string_tls - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zookeeperConnectStringTls() {
        return this.getStringAttribute('zookeeper_connect_string_tls');
    }
    /**
     * @stability stable
     */
    get brokerNodeGroupInfo() {
        return this._brokerNodeGroupInfo;
    }
    /**
     * @stability stable
     */
    putBrokerNodeGroupInfo(value) {
        this._brokerNodeGroupInfo.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get brokerNodeGroupInfoInput() {
        return this._brokerNodeGroupInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get clientAuthentication() {
        return this._clientAuthentication;
    }
    /**
     * @stability stable
     */
    putClientAuthentication(value) {
        this._clientAuthentication.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetClientAuthentication() {
        this._clientAuthentication.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientAuthenticationInput() {
        return this._clientAuthentication.internalValue;
    }
    /**
     * @stability stable
     */
    get configurationInfo() {
        return this._configurationInfo;
    }
    /**
     * @stability stable
     */
    putConfigurationInfo(value) {
        this._configurationInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConfigurationInfo() {
        this._configurationInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get configurationInfoInput() {
        return this._configurationInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get encryptionInfo() {
        return this._encryptionInfo;
    }
    /**
     * @stability stable
     */
    putEncryptionInfo(value) {
        this._encryptionInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionInfo() {
        this._encryptionInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionInfoInput() {
        return this._encryptionInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get loggingInfo() {
        return this._loggingInfo;
    }
    /**
     * @stability stable
     */
    putLoggingInfo(value) {
        this._loggingInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLoggingInfo() {
        this._loggingInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingInfoInput() {
        return this._loggingInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get openMonitoring() {
        return this._openMonitoring;
    }
    /**
     * @stability stable
     */
    putOpenMonitoring(value) {
        this._openMonitoring.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetOpenMonitoring() {
        this._openMonitoring.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get openMonitoringInput() {
        return this._openMonitoring.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            enhanced_monitoring: cdktf.stringToTerraform(this._enhancedMonitoring),
            kafka_version: cdktf.stringToTerraform(this._kafkaVersion),
            number_of_broker_nodes: cdktf.numberToTerraform(this._numberOfBrokerNodes),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            broker_node_group_info: mskClusterBrokerNodeGroupInfoToTerraform(this._brokerNodeGroupInfo.internalValue),
            client_authentication: mskClusterClientAuthenticationToTerraform(this._clientAuthentication.internalValue),
            configuration_info: mskClusterConfigurationInfoToTerraform(this._configurationInfo.internalValue),
            encryption_info: mskClusterEncryptionInfoToTerraform(this._encryptionInfo.internalValue),
            logging_info: mskClusterLoggingInfoToTerraform(this._loggingInfo.internalValue),
            open_monitoring: mskClusterOpenMonitoringToTerraform(this._openMonitoring.internalValue),
            timeouts: mskClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.MskCluster = MskCluster;
_t = JSII_RTTI_SYMBOL_1;
MskCluster[_t] = { fqn: "@cdktf/provider-aws.msk.MskCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MskCluster.tfResourceType = "aws_msk_cluster";
