// @cdktf/provider-aws/lib/msk/msk-scram-secret-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskScramSecretAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_scram_secret_association.html aws_msk_scram_secret_association}.
 *
 * @stability stable
 */
class MskScramSecretAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_scram_secret_association.html aws_msk_scram_secret_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_scram_secret_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterArn = config.clusterArn;
        this._secretArnList = config.secretArnList;
    }
    /**
     * @stability stable
     */
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    /**
     * @stability stable
     */
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterArnInput() {
        return this._clusterArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get secretArnList() {
        return this.getListAttribute('secret_arn_list');
    }
    /**
     * @stability stable
     */
    set secretArnList(value) {
        this._secretArnList = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get secretArnListInput() {
        return this._secretArnList;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
            secret_arn_list: cdktf.listMapper(cdktf.stringToTerraform)(this._secretArnList),
        };
    }
}
exports.MskScramSecretAssociation = MskScramSecretAssociation;
_a = JSII_RTTI_SYMBOL_1;
MskScramSecretAssociation[_a] = { fqn: "@cdktf/provider-aws.msk.MskScramSecretAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MskScramSecretAssociation.tfResourceType = "aws_msk_scram_secret_association";
