// @cdktf/provider-aws/lib/appconfig/appconfig-deployment-strategy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppconfigDeploymentStrategy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/appconfig_deployment_strategy.html aws_appconfig_deployment_strategy}.
 *
 * @stability stable
 */
class AppconfigDeploymentStrategy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appconfig_deployment_strategy.html aws_appconfig_deployment_strategy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appconfig_deployment_strategy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._deploymentDurationInMinutes = config.deploymentDurationInMinutes;
        this._description = config.description;
        this._finalBakeTimeInMinutes = config.finalBakeTimeInMinutes;
        this._growthFactor = config.growthFactor;
        this._growthType = config.growthType;
        this._name = config.name;
        this._replicateTo = config.replicateTo;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get deploymentDurationInMinutes() {
        return this.getNumberAttribute('deployment_duration_in_minutes');
    }
    /**
     * @stability stable
     */
    set deploymentDurationInMinutes(value) {
        this._deploymentDurationInMinutes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentDurationInMinutesInput() {
        return this._deploymentDurationInMinutes;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get finalBakeTimeInMinutes() {
        return this.getNumberAttribute('final_bake_time_in_minutes');
    }
    /**
     * @stability stable
     */
    set finalBakeTimeInMinutes(value) {
        this._finalBakeTimeInMinutes = value;
    }
    /**
     * @stability stable
     */
    resetFinalBakeTimeInMinutes() {
        this._finalBakeTimeInMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get finalBakeTimeInMinutesInput() {
        return this._finalBakeTimeInMinutes;
    }
    /**
     * @stability stable
     */
    get growthFactor() {
        return this.getNumberAttribute('growth_factor');
    }
    /**
     * @stability stable
     */
    set growthFactor(value) {
        this._growthFactor = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get growthFactorInput() {
        return this._growthFactor;
    }
    /**
     * @stability stable
     */
    get growthType() {
        return this.getStringAttribute('growth_type');
    }
    /**
     * @stability stable
     */
    set growthType(value) {
        this._growthType = value;
    }
    /**
     * @stability stable
     */
    resetGrowthType() {
        this._growthType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get growthTypeInput() {
        return this._growthType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get replicateTo() {
        return this.getStringAttribute('replicate_to');
    }
    /**
     * @stability stable
     */
    set replicateTo(value) {
        this._replicateTo = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get replicateToInput() {
        return this._replicateTo;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            deployment_duration_in_minutes: cdktf.numberToTerraform(this._deploymentDurationInMinutes),
            description: cdktf.stringToTerraform(this._description),
            final_bake_time_in_minutes: cdktf.numberToTerraform(this._finalBakeTimeInMinutes),
            growth_factor: cdktf.numberToTerraform(this._growthFactor),
            growth_type: cdktf.stringToTerraform(this._growthType),
            name: cdktf.stringToTerraform(this._name),
            replicate_to: cdktf.stringToTerraform(this._replicateTo),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.AppconfigDeploymentStrategy = AppconfigDeploymentStrategy;
_a = JSII_RTTI_SYMBOL_1;
AppconfigDeploymentStrategy[_a] = { fqn: "@cdktf/provider-aws.appconfig.AppconfigDeploymentStrategy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AppconfigDeploymentStrategy.tfResourceType = "aws_appconfig_deployment_strategy";
