// @cdktf/provider-aws/lib/amplify/amplify-domain-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyDomainAssociation = exports.amplifyDomainAssociationSubDomainToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function amplifyDomainAssociationSubDomainToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        branch_name: cdktf.stringToTerraform(struct.branchName),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.amplifyDomainAssociationSubDomainToTerraform = amplifyDomainAssociationSubDomainToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/amplify_domain_association.html aws_amplify_domain_association}.
 *
 * @stability stable
 */
class AmplifyDomainAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/amplify_domain_association.html aws_amplify_domain_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_amplify_domain_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._appId = config.appId;
        this._domainName = config.domainName;
        this._waitForVerification = config.waitForVerification;
        this._subDomain = config.subDomain;
    }
    /**
     * @stability stable
     */
    get appId() {
        return this.getStringAttribute('app_id');
    }
    /**
     * @stability stable
     */
    set appId(value) {
        this._appId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get appIdInput() {
        return this._appId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_verification_dns_record - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateVerificationDnsRecord() {
        return this.getStringAttribute('certificate_verification_dns_record');
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get waitForVerification() {
        return this.getBooleanAttribute('wait_for_verification');
    }
    /**
     * @stability stable
     */
    set waitForVerification(value) {
        this._waitForVerification = value;
    }
    /**
     * @stability stable
     */
    resetWaitForVerification() {
        this._waitForVerification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get waitForVerificationInput() {
        return this._waitForVerification;
    }
    /**
     * @stability stable
     */
    get subDomain() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('sub_domain');
    }
    /**
     * @stability stable
     */
    set subDomain(value) {
        this._subDomain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subDomainInput() {
        return this._subDomain;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            app_id: cdktf.stringToTerraform(this._appId),
            domain_name: cdktf.stringToTerraform(this._domainName),
            wait_for_verification: cdktf.booleanToTerraform(this._waitForVerification),
            sub_domain: cdktf.listMapper(amplifyDomainAssociationSubDomainToTerraform)(this._subDomain),
        };
    }
}
exports.AmplifyDomainAssociation = AmplifyDomainAssociation;
_a = JSII_RTTI_SYMBOL_1;
AmplifyDomainAssociation[_a] = { fqn: "@cdktf/provider-aws.amplify.AmplifyDomainAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AmplifyDomainAssociation.tfResourceType = "aws_amplify_domain_association";
