// @cdktf/provider-aws/lib/amplify/amplify-branch.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyBranch = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/amplify_branch.html aws_amplify_branch}.
 *
 * @stability stable
 */
class AmplifyBranch extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/amplify_branch.html aws_amplify_branch} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_amplify_branch',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._appId = config.appId;
        this._backendEnvironmentArn = config.backendEnvironmentArn;
        this._basicAuthCredentials = config.basicAuthCredentials;
        this._branchName = config.branchName;
        this._description = config.description;
        this._displayName = config.displayName;
        this._enableAutoBuild = config.enableAutoBuild;
        this._enableBasicAuth = config.enableBasicAuth;
        this._enableNotification = config.enableNotification;
        this._enablePerformanceMode = config.enablePerformanceMode;
        this._enablePullRequestPreview = config.enablePullRequestPreview;
        this._environmentVariables = config.environmentVariables;
        this._framework = config.framework;
        this._pullRequestEnvironmentName = config.pullRequestEnvironmentName;
        this._stage = config.stage;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._ttl = config.ttl;
    }
    /**
     * @stability stable
     */
    get appId() {
        return this.getStringAttribute('app_id');
    }
    /**
     * @stability stable
     */
    set appId(value) {
        this._appId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get appIdInput() {
        return this._appId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // associated_resources - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associatedResources() {
        return this.getListAttribute('associated_resources');
    }
    /**
     * @stability stable
     */
    get backendEnvironmentArn() {
        return this.getStringAttribute('backend_environment_arn');
    }
    /**
     * @stability stable
     */
    set backendEnvironmentArn(value) {
        this._backendEnvironmentArn = value;
    }
    /**
     * @stability stable
     */
    resetBackendEnvironmentArn() {
        this._backendEnvironmentArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get backendEnvironmentArnInput() {
        return this._backendEnvironmentArn;
    }
    /**
     * @stability stable
     */
    get basicAuthCredentials() {
        return this.getStringAttribute('basic_auth_credentials');
    }
    /**
     * @stability stable
     */
    set basicAuthCredentials(value) {
        this._basicAuthCredentials = value;
    }
    /**
     * @stability stable
     */
    resetBasicAuthCredentials() {
        this._basicAuthCredentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get basicAuthCredentialsInput() {
        return this._basicAuthCredentials;
    }
    /**
     * @stability stable
     */
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    /**
     * @stability stable
     */
    set branchName(value) {
        this._branchName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get branchNameInput() {
        return this._branchName;
    }
    // custom_domains - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customDomains() {
        return this.getListAttribute('custom_domains');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // destination_branch - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get destinationBranch() {
        return this.getStringAttribute('destination_branch');
    }
    /**
     * @stability stable
     */
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    /**
     * @stability stable
     */
    set displayName(value) {
        this._displayName = value;
    }
    /**
     * @stability stable
     */
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get displayNameInput() {
        return this._displayName;
    }
    /**
     * @stability stable
     */
    get enableAutoBuild() {
        return this.getBooleanAttribute('enable_auto_build');
    }
    /**
     * @stability stable
     */
    set enableAutoBuild(value) {
        this._enableAutoBuild = value;
    }
    /**
     * @stability stable
     */
    resetEnableAutoBuild() {
        this._enableAutoBuild = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableAutoBuildInput() {
        return this._enableAutoBuild;
    }
    /**
     * @stability stable
     */
    get enableBasicAuth() {
        return this.getBooleanAttribute('enable_basic_auth');
    }
    /**
     * @stability stable
     */
    set enableBasicAuth(value) {
        this._enableBasicAuth = value;
    }
    /**
     * @stability stable
     */
    resetEnableBasicAuth() {
        this._enableBasicAuth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableBasicAuthInput() {
        return this._enableBasicAuth;
    }
    /**
     * @stability stable
     */
    get enableNotification() {
        return this.getBooleanAttribute('enable_notification');
    }
    /**
     * @stability stable
     */
    set enableNotification(value) {
        this._enableNotification = value;
    }
    /**
     * @stability stable
     */
    resetEnableNotification() {
        this._enableNotification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableNotificationInput() {
        return this._enableNotification;
    }
    /**
     * @stability stable
     */
    get enablePerformanceMode() {
        return this.getBooleanAttribute('enable_performance_mode');
    }
    /**
     * @stability stable
     */
    set enablePerformanceMode(value) {
        this._enablePerformanceMode = value;
    }
    /**
     * @stability stable
     */
    resetEnablePerformanceMode() {
        this._enablePerformanceMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enablePerformanceModeInput() {
        return this._enablePerformanceMode;
    }
    /**
     * @stability stable
     */
    get enablePullRequestPreview() {
        return this.getBooleanAttribute('enable_pull_request_preview');
    }
    /**
     * @stability stable
     */
    set enablePullRequestPreview(value) {
        this._enablePullRequestPreview = value;
    }
    /**
     * @stability stable
     */
    resetEnablePullRequestPreview() {
        this._enablePullRequestPreview = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enablePullRequestPreviewInput() {
        return this._enablePullRequestPreview;
    }
    /**
     * @stability stable
     */
    get environmentVariables() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('environment_variables');
    }
    /**
     * @stability stable
     */
    set environmentVariables(value) {
        this._environmentVariables = value;
    }
    /**
     * @stability stable
     */
    resetEnvironmentVariables() {
        this._environmentVariables = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentVariablesInput() {
        return this._environmentVariables;
    }
    /**
     * @stability stable
     */
    get framework() {
        return this.getStringAttribute('framework');
    }
    /**
     * @stability stable
     */
    set framework(value) {
        this._framework = value;
    }
    /**
     * @stability stable
     */
    resetFramework() {
        this._framework = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get frameworkInput() {
        return this._framework;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get pullRequestEnvironmentName() {
        return this.getStringAttribute('pull_request_environment_name');
    }
    /**
     * @stability stable
     */
    set pullRequestEnvironmentName(value) {
        this._pullRequestEnvironmentName = value;
    }
    /**
     * @stability stable
     */
    resetPullRequestEnvironmentName() {
        this._pullRequestEnvironmentName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pullRequestEnvironmentNameInput() {
        return this._pullRequestEnvironmentName;
    }
    // source_branch - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceBranch() {
        return this.getStringAttribute('source_branch');
    }
    /**
     * @stability stable
     */
    get stage() {
        return this.getStringAttribute('stage');
    }
    /**
     * @stability stable
     */
    set stage(value) {
        this._stage = value;
    }
    /**
     * @stability stable
     */
    resetStage() {
        this._stage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stageInput() {
        return this._stage;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get ttl() {
        return this.getStringAttribute('ttl');
    }
    /**
     * @stability stable
     */
    set ttl(value) {
        this._ttl = value;
    }
    /**
     * @stability stable
     */
    resetTtl() {
        this._ttl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ttlInput() {
        return this._ttl;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            app_id: cdktf.stringToTerraform(this._appId),
            backend_environment_arn: cdktf.stringToTerraform(this._backendEnvironmentArn),
            basic_auth_credentials: cdktf.stringToTerraform(this._basicAuthCredentials),
            branch_name: cdktf.stringToTerraform(this._branchName),
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            enable_auto_build: cdktf.booleanToTerraform(this._enableAutoBuild),
            enable_basic_auth: cdktf.booleanToTerraform(this._enableBasicAuth),
            enable_notification: cdktf.booleanToTerraform(this._enableNotification),
            enable_performance_mode: cdktf.booleanToTerraform(this._enablePerformanceMode),
            enable_pull_request_preview: cdktf.booleanToTerraform(this._enablePullRequestPreview),
            environment_variables: cdktf.hashMapper(cdktf.anyToTerraform)(this._environmentVariables),
            framework: cdktf.stringToTerraform(this._framework),
            pull_request_environment_name: cdktf.stringToTerraform(this._pullRequestEnvironmentName),
            stage: cdktf.stringToTerraform(this._stage),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            ttl: cdktf.stringToTerraform(this._ttl),
        };
    }
}
exports.AmplifyBranch = AmplifyBranch;
_a = JSII_RTTI_SYMBOL_1;
AmplifyBranch[_a] = { fqn: "@cdktf/provider-aws.amplify.AmplifyBranch", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AmplifyBranch.tfResourceType = "aws_amplify_branch";
