// @cdktf/provider-aws/lib/amplify/amplify-backend-environment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyBackendEnvironment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/amplify_backend_environment.html aws_amplify_backend_environment}.
 *
 * @stability stable
 */
class AmplifyBackendEnvironment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/amplify_backend_environment.html aws_amplify_backend_environment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_amplify_backend_environment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._appId = config.appId;
        this._deploymentArtifacts = config.deploymentArtifacts;
        this._environmentName = config.environmentName;
        this._stackName = config.stackName;
    }
    /**
     * @stability stable
     */
    get appId() {
        return this.getStringAttribute('app_id');
    }
    /**
     * @stability stable
     */
    set appId(value) {
        this._appId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get appIdInput() {
        return this._appId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get deploymentArtifacts() {
        return this.getStringAttribute('deployment_artifacts');
    }
    /**
     * @stability stable
     */
    set deploymentArtifacts(value) {
        this._deploymentArtifacts = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentArtifacts() {
        this._deploymentArtifacts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentArtifactsInput() {
        return this._deploymentArtifacts;
    }
    /**
     * @stability stable
     */
    get environmentName() {
        return this.getStringAttribute('environment_name');
    }
    /**
     * @stability stable
     */
    set environmentName(value) {
        this._environmentName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get environmentNameInput() {
        return this._environmentName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get stackName() {
        return this.getStringAttribute('stack_name');
    }
    /**
     * @stability stable
     */
    set stackName(value) {
        this._stackName = value;
    }
    /**
     * @stability stable
     */
    resetStackName() {
        this._stackName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackNameInput() {
        return this._stackName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            app_id: cdktf.stringToTerraform(this._appId),
            deployment_artifacts: cdktf.stringToTerraform(this._deploymentArtifacts),
            environment_name: cdktf.stringToTerraform(this._environmentName),
            stack_name: cdktf.stringToTerraform(this._stackName),
        };
    }
}
exports.AmplifyBackendEnvironment = AmplifyBackendEnvironment;
_a = JSII_RTTI_SYMBOL_1;
AmplifyBackendEnvironment[_a] = { fqn: "@cdktf/provider-aws.amplify.AmplifyBackendEnvironment", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AmplifyBackendEnvironment.tfResourceType = "aws_amplify_backend_environment";
