// @cdktf/provider-aws/lib/codebuild/codebuild-webhook.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodebuildWebhook = exports.codebuildWebhookFilterGroupToTerraform = exports.codebuildWebhookFilterGroupFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codebuildWebhookFilterGroupFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclude_matched_pattern: cdktf.booleanToTerraform(struct.excludeMatchedPattern),
        pattern: cdktf.stringToTerraform(struct.pattern),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildWebhookFilterGroupFilterToTerraform = codebuildWebhookFilterGroupFilterToTerraform;
function codebuildWebhookFilterGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.listMapper(codebuildWebhookFilterGroupFilterToTerraform)(struct.filter),
    };
}
exports.codebuildWebhookFilterGroupToTerraform = codebuildWebhookFilterGroupToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/codebuild_webhook.html aws_codebuild_webhook}.
 *
 * @stability stable
 */
class CodebuildWebhook extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codebuild_webhook.html aws_codebuild_webhook} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codebuild_webhook',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._branchFilter = config.branchFilter;
        this._buildType = config.buildType;
        this._projectName = config.projectName;
        this._filterGroup = config.filterGroup;
    }
    /**
     * @stability stable
     */
    get branchFilter() {
        return this.getStringAttribute('branch_filter');
    }
    /**
     * @stability stable
     */
    set branchFilter(value) {
        this._branchFilter = value;
    }
    /**
     * @stability stable
     */
    resetBranchFilter() {
        this._branchFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get branchFilterInput() {
        return this._branchFilter;
    }
    /**
     * @stability stable
     */
    get buildType() {
        return this.getStringAttribute('build_type');
    }
    /**
     * @stability stable
     */
    set buildType(value) {
        this._buildType = value;
    }
    /**
     * @stability stable
     */
    resetBuildType() {
        this._buildType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildTypeInput() {
        return this._buildType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // payload_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get payloadUrl() {
        return this.getStringAttribute('payload_url');
    }
    /**
     * @stability stable
     */
    get projectName() {
        return this.getStringAttribute('project_name');
    }
    /**
     * @stability stable
     */
    set projectName(value) {
        this._projectName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get projectNameInput() {
        return this._projectName;
    }
    // secret - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get secret() {
        return this.getStringAttribute('secret');
    }
    // url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get url() {
        return this.getStringAttribute('url');
    }
    /**
     * @stability stable
     */
    get filterGroup() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter_group');
    }
    /**
     * @stability stable
     */
    set filterGroup(value) {
        this._filterGroup = value;
    }
    /**
     * @stability stable
     */
    resetFilterGroup() {
        this._filterGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterGroupInput() {
        return this._filterGroup;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            branch_filter: cdktf.stringToTerraform(this._branchFilter),
            build_type: cdktf.stringToTerraform(this._buildType),
            project_name: cdktf.stringToTerraform(this._projectName),
            filter_group: cdktf.listMapper(codebuildWebhookFilterGroupToTerraform)(this._filterGroup),
        };
    }
}
exports.CodebuildWebhook = CodebuildWebhook;
_a = JSII_RTTI_SYMBOL_1;
CodebuildWebhook[_a] = { fqn: "@cdktf/provider-aws.codebuild.CodebuildWebhook", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CodebuildWebhook.tfResourceType = "aws_codebuild_webhook";
