// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-integration.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Integration = exports.Apigatewayv2IntegrationTlsConfigOutputReference = exports.apigatewayv2IntegrationTlsConfigToTerraform = exports.apigatewayv2IntegrationResponseParametersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2IntegrationResponseParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mappings: cdktf.hashMapper(cdktf.anyToTerraform)(struct.mappings),
        status_code: cdktf.stringToTerraform(struct.statusCode),
    };
}
exports.apigatewayv2IntegrationResponseParametersToTerraform = apigatewayv2IntegrationResponseParametersToTerraform;
function apigatewayv2IntegrationTlsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        server_name_to_verify: cdktf.stringToTerraform(struct.serverNameToVerify),
    };
}
exports.apigatewayv2IntegrationTlsConfigToTerraform = apigatewayv2IntegrationTlsConfigToTerraform;
/**
 * @stability stable
 */
class Apigatewayv2IntegrationTlsConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serverNameToVerify) {
            hasAnyValues = true;
            internalValueResult.serverNameToVerify = this._serverNameToVerify;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._serverNameToVerify = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._serverNameToVerify = value.serverNameToVerify;
        }
    }
    /**
     * @stability stable
     */
    get serverNameToVerify() {
        return this.getStringAttribute('server_name_to_verify');
    }
    /**
     * @stability stable
     */
    set serverNameToVerify(value) {
        this._serverNameToVerify = value;
    }
    /**
     * @stability stable
     */
    resetServerNameToVerify() {
        this._serverNameToVerify = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serverNameToVerifyInput() {
        return this._serverNameToVerify;
    }
}
exports.Apigatewayv2IntegrationTlsConfigOutputReference = Apigatewayv2IntegrationTlsConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2IntegrationTlsConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2IntegrationTlsConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_integration.html aws_apigatewayv2_integration}.
 *
 * @stability stable
 */
class Apigatewayv2Integration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_integration.html aws_apigatewayv2_integration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_integration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // tls_config - computed: false, optional: true, required: false
        this._tlsConfig = new Apigatewayv2IntegrationTlsConfigOutputReference(this, "tls_config", true);
        this._apiId = config.apiId;
        this._connectionId = config.connectionId;
        this._connectionType = config.connectionType;
        this._contentHandlingStrategy = config.contentHandlingStrategy;
        this._credentialsArn = config.credentialsArn;
        this._description = config.description;
        this._integrationMethod = config.integrationMethod;
        this._integrationSubtype = config.integrationSubtype;
        this._integrationType = config.integrationType;
        this._integrationUri = config.integrationUri;
        this._passthroughBehavior = config.passthroughBehavior;
        this._payloadFormatVersion = config.payloadFormatVersion;
        this._requestParameters = config.requestParameters;
        this._requestTemplates = config.requestTemplates;
        this._templateSelectionExpression = config.templateSelectionExpression;
        this._timeoutMilliseconds = config.timeoutMilliseconds;
        this._responseParameters = config.responseParameters;
        this._tlsConfig.internalValue = config.tlsConfig;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    /**
     * @stability stable
     */
    get connectionId() {
        return this.getStringAttribute('connection_id');
    }
    /**
     * @stability stable
     */
    set connectionId(value) {
        this._connectionId = value;
    }
    /**
     * @stability stable
     */
    resetConnectionId() {
        this._connectionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionIdInput() {
        return this._connectionId;
    }
    /**
     * @stability stable
     */
    get connectionType() {
        return this.getStringAttribute('connection_type');
    }
    /**
     * @stability stable
     */
    set connectionType(value) {
        this._connectionType = value;
    }
    /**
     * @stability stable
     */
    resetConnectionType() {
        this._connectionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionTypeInput() {
        return this._connectionType;
    }
    /**
     * @stability stable
     */
    get contentHandlingStrategy() {
        return this.getStringAttribute('content_handling_strategy');
    }
    /**
     * @stability stable
     */
    set contentHandlingStrategy(value) {
        this._contentHandlingStrategy = value;
    }
    /**
     * @stability stable
     */
    resetContentHandlingStrategy() {
        this._contentHandlingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentHandlingStrategyInput() {
        return this._contentHandlingStrategy;
    }
    /**
     * @stability stable
     */
    get credentialsArn() {
        return this.getStringAttribute('credentials_arn');
    }
    /**
     * @stability stable
     */
    set credentialsArn(value) {
        this._credentialsArn = value;
    }
    /**
     * @stability stable
     */
    resetCredentialsArn() {
        this._credentialsArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialsArnInput() {
        return this._credentialsArn;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get integrationMethod() {
        return this.getStringAttribute('integration_method');
    }
    /**
     * @stability stable
     */
    set integrationMethod(value) {
        this._integrationMethod = value;
    }
    /**
     * @stability stable
     */
    resetIntegrationMethod() {
        this._integrationMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationMethodInput() {
        return this._integrationMethod;
    }
    // integration_response_selection_expression - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get integrationResponseSelectionExpression() {
        return this.getStringAttribute('integration_response_selection_expression');
    }
    /**
     * @stability stable
     */
    get integrationSubtype() {
        return this.getStringAttribute('integration_subtype');
    }
    /**
     * @stability stable
     */
    set integrationSubtype(value) {
        this._integrationSubtype = value;
    }
    /**
     * @stability stable
     */
    resetIntegrationSubtype() {
        this._integrationSubtype = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationSubtypeInput() {
        return this._integrationSubtype;
    }
    /**
     * @stability stable
     */
    get integrationType() {
        return this.getStringAttribute('integration_type');
    }
    /**
     * @stability stable
     */
    set integrationType(value) {
        this._integrationType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationTypeInput() {
        return this._integrationType;
    }
    /**
     * @stability stable
     */
    get integrationUri() {
        return this.getStringAttribute('integration_uri');
    }
    /**
     * @stability stable
     */
    set integrationUri(value) {
        this._integrationUri = value;
    }
    /**
     * @stability stable
     */
    resetIntegrationUri() {
        this._integrationUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationUriInput() {
        return this._integrationUri;
    }
    /**
     * @stability stable
     */
    get passthroughBehavior() {
        return this.getStringAttribute('passthrough_behavior');
    }
    /**
     * @stability stable
     */
    set passthroughBehavior(value) {
        this._passthroughBehavior = value;
    }
    /**
     * @stability stable
     */
    resetPassthroughBehavior() {
        this._passthroughBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passthroughBehaviorInput() {
        return this._passthroughBehavior;
    }
    /**
     * @stability stable
     */
    get payloadFormatVersion() {
        return this.getStringAttribute('payload_format_version');
    }
    /**
     * @stability stable
     */
    set payloadFormatVersion(value) {
        this._payloadFormatVersion = value;
    }
    /**
     * @stability stable
     */
    resetPayloadFormatVersion() {
        this._payloadFormatVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get payloadFormatVersionInput() {
        return this._payloadFormatVersion;
    }
    /**
     * @stability stable
     */
    get requestParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_parameters');
    }
    /**
     * @stability stable
     */
    set requestParameters(value) {
        this._requestParameters = value;
    }
    /**
     * @stability stable
     */
    resetRequestParameters() {
        this._requestParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestParametersInput() {
        return this._requestParameters;
    }
    /**
     * @stability stable
     */
    get requestTemplates() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_templates');
    }
    /**
     * @stability stable
     */
    set requestTemplates(value) {
        this._requestTemplates = value;
    }
    /**
     * @stability stable
     */
    resetRequestTemplates() {
        this._requestTemplates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestTemplatesInput() {
        return this._requestTemplates;
    }
    /**
     * @stability stable
     */
    get templateSelectionExpression() {
        return this.getStringAttribute('template_selection_expression');
    }
    /**
     * @stability stable
     */
    set templateSelectionExpression(value) {
        this._templateSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetTemplateSelectionExpression() {
        this._templateSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get templateSelectionExpressionInput() {
        return this._templateSelectionExpression;
    }
    /**
     * @stability stable
     */
    get timeoutMilliseconds() {
        return this.getNumberAttribute('timeout_milliseconds');
    }
    /**
     * @stability stable
     */
    set timeoutMilliseconds(value) {
        this._timeoutMilliseconds = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutMilliseconds() {
        this._timeoutMilliseconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutMillisecondsInput() {
        return this._timeoutMilliseconds;
    }
    /**
     * @stability stable
     */
    get responseParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('response_parameters');
    }
    /**
     * @stability stable
     */
    set responseParameters(value) {
        this._responseParameters = value;
    }
    /**
     * @stability stable
     */
    resetResponseParameters() {
        this._responseParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get responseParametersInput() {
        return this._responseParameters;
    }
    /**
     * @stability stable
     */
    get tlsConfig() {
        return this._tlsConfig;
    }
    /**
     * @stability stable
     */
    putTlsConfig(value) {
        this._tlsConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTlsConfig() {
        this._tlsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsConfigInput() {
        return this._tlsConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            connection_id: cdktf.stringToTerraform(this._connectionId),
            connection_type: cdktf.stringToTerraform(this._connectionType),
            content_handling_strategy: cdktf.stringToTerraform(this._contentHandlingStrategy),
            credentials_arn: cdktf.stringToTerraform(this._credentialsArn),
            description: cdktf.stringToTerraform(this._description),
            integration_method: cdktf.stringToTerraform(this._integrationMethod),
            integration_subtype: cdktf.stringToTerraform(this._integrationSubtype),
            integration_type: cdktf.stringToTerraform(this._integrationType),
            integration_uri: cdktf.stringToTerraform(this._integrationUri),
            passthrough_behavior: cdktf.stringToTerraform(this._passthroughBehavior),
            payload_format_version: cdktf.stringToTerraform(this._payloadFormatVersion),
            request_parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._requestParameters),
            request_templates: cdktf.hashMapper(cdktf.anyToTerraform)(this._requestTemplates),
            template_selection_expression: cdktf.stringToTerraform(this._templateSelectionExpression),
            timeout_milliseconds: cdktf.numberToTerraform(this._timeoutMilliseconds),
            response_parameters: cdktf.listMapper(apigatewayv2IntegrationResponseParametersToTerraform)(this._responseParameters),
            tls_config: apigatewayv2IntegrationTlsConfigToTerraform(this._tlsConfig.internalValue),
        };
    }
}
exports.Apigatewayv2Integration = Apigatewayv2Integration;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2Integration[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2Integration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2Integration.tfResourceType = "aws_apigatewayv2_integration";
