// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-api.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Api = exports.Apigatewayv2ApiCorsConfigurationOutputReference = exports.apigatewayv2ApiCorsConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2ApiCorsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allow_credentials: cdktf.booleanToTerraform(struct.allowCredentials),
        allow_headers: cdktf.listMapper(cdktf.stringToTerraform)(struct.allowHeaders),
        allow_methods: cdktf.listMapper(cdktf.stringToTerraform)(struct.allowMethods),
        allow_origins: cdktf.listMapper(cdktf.stringToTerraform)(struct.allowOrigins),
        expose_headers: cdktf.listMapper(cdktf.stringToTerraform)(struct.exposeHeaders),
        max_age: cdktf.numberToTerraform(struct.maxAge),
    };
}
exports.apigatewayv2ApiCorsConfigurationToTerraform = apigatewayv2ApiCorsConfigurationToTerraform;
/**
 * @stability stable
 */
class Apigatewayv2ApiCorsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowCredentials) {
            hasAnyValues = true;
            internalValueResult.allowCredentials = this._allowCredentials;
        }
        if (this._allowHeaders) {
            hasAnyValues = true;
            internalValueResult.allowHeaders = this._allowHeaders;
        }
        if (this._allowMethods) {
            hasAnyValues = true;
            internalValueResult.allowMethods = this._allowMethods;
        }
        if (this._allowOrigins) {
            hasAnyValues = true;
            internalValueResult.allowOrigins = this._allowOrigins;
        }
        if (this._exposeHeaders) {
            hasAnyValues = true;
            internalValueResult.exposeHeaders = this._exposeHeaders;
        }
        if (this._maxAge) {
            hasAnyValues = true;
            internalValueResult.maxAge = this._maxAge;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowCredentials = undefined;
            this._allowHeaders = undefined;
            this._allowMethods = undefined;
            this._allowOrigins = undefined;
            this._exposeHeaders = undefined;
            this._maxAge = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowCredentials = value.allowCredentials;
            this._allowHeaders = value.allowHeaders;
            this._allowMethods = value.allowMethods;
            this._allowOrigins = value.allowOrigins;
            this._exposeHeaders = value.exposeHeaders;
            this._maxAge = value.maxAge;
        }
    }
    /**
     * @stability stable
     */
    get allowCredentials() {
        return this.getBooleanAttribute('allow_credentials');
    }
    /**
     * @stability stable
     */
    set allowCredentials(value) {
        this._allowCredentials = value;
    }
    /**
     * @stability stable
     */
    resetAllowCredentials() {
        this._allowCredentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowCredentialsInput() {
        return this._allowCredentials;
    }
    /**
     * @stability stable
     */
    get allowHeaders() {
        return this.getListAttribute('allow_headers');
    }
    /**
     * @stability stable
     */
    set allowHeaders(value) {
        this._allowHeaders = value;
    }
    /**
     * @stability stable
     */
    resetAllowHeaders() {
        this._allowHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowHeadersInput() {
        return this._allowHeaders;
    }
    /**
     * @stability stable
     */
    get allowMethods() {
        return this.getListAttribute('allow_methods');
    }
    /**
     * @stability stable
     */
    set allowMethods(value) {
        this._allowMethods = value;
    }
    /**
     * @stability stable
     */
    resetAllowMethods() {
        this._allowMethods = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowMethodsInput() {
        return this._allowMethods;
    }
    /**
     * @stability stable
     */
    get allowOrigins() {
        return this.getListAttribute('allow_origins');
    }
    /**
     * @stability stable
     */
    set allowOrigins(value) {
        this._allowOrigins = value;
    }
    /**
     * @stability stable
     */
    resetAllowOrigins() {
        this._allowOrigins = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowOriginsInput() {
        return this._allowOrigins;
    }
    /**
     * @stability stable
     */
    get exposeHeaders() {
        return this.getListAttribute('expose_headers');
    }
    /**
     * @stability stable
     */
    set exposeHeaders(value) {
        this._exposeHeaders = value;
    }
    /**
     * @stability stable
     */
    resetExposeHeaders() {
        this._exposeHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get exposeHeadersInput() {
        return this._exposeHeaders;
    }
    /**
     * @stability stable
     */
    get maxAge() {
        return this.getNumberAttribute('max_age');
    }
    /**
     * @stability stable
     */
    set maxAge(value) {
        this._maxAge = value;
    }
    /**
     * @stability stable
     */
    resetMaxAge() {
        this._maxAge = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxAgeInput() {
        return this._maxAge;
    }
}
exports.Apigatewayv2ApiCorsConfigurationOutputReference = Apigatewayv2ApiCorsConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2ApiCorsConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2ApiCorsConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_api.html aws_apigatewayv2_api}.
 *
 * @stability stable
 */
class Apigatewayv2Api extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_api.html aws_apigatewayv2_api} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_api',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // cors_configuration - computed: false, optional: true, required: false
        this._corsConfiguration = new Apigatewayv2ApiCorsConfigurationOutputReference(this, "cors_configuration", true);
        this._apiKeySelectionExpression = config.apiKeySelectionExpression;
        this._body = config.body;
        this._credentialsArn = config.credentialsArn;
        this._description = config.description;
        this._disableExecuteApiEndpoint = config.disableExecuteApiEndpoint;
        this._failOnWarnings = config.failOnWarnings;
        this._name = config.name;
        this._protocolType = config.protocolType;
        this._routeKey = config.routeKey;
        this._routeSelectionExpression = config.routeSelectionExpression;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._target = config.target;
        this._version = config.version;
        this._corsConfiguration.internalValue = config.corsConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // api_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get apiEndpoint() {
        return this.getStringAttribute('api_endpoint');
    }
    /**
     * @stability stable
     */
    get apiKeySelectionExpression() {
        return this.getStringAttribute('api_key_selection_expression');
    }
    /**
     * @stability stable
     */
    set apiKeySelectionExpression(value) {
        this._apiKeySelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetApiKeySelectionExpression() {
        this._apiKeySelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiKeySelectionExpressionInput() {
        return this._apiKeySelectionExpression;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get body() {
        return this.getStringAttribute('body');
    }
    /**
     * @stability stable
     */
    set body(value) {
        this._body = value;
    }
    /**
     * @stability stable
     */
    resetBody() {
        this._body = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bodyInput() {
        return this._body;
    }
    /**
     * @stability stable
     */
    get credentialsArn() {
        return this.getStringAttribute('credentials_arn');
    }
    /**
     * @stability stable
     */
    set credentialsArn(value) {
        this._credentialsArn = value;
    }
    /**
     * @stability stable
     */
    resetCredentialsArn() {
        this._credentialsArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialsArnInput() {
        return this._credentialsArn;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get disableExecuteApiEndpoint() {
        return this.getBooleanAttribute('disable_execute_api_endpoint');
    }
    /**
     * @stability stable
     */
    set disableExecuteApiEndpoint(value) {
        this._disableExecuteApiEndpoint = value;
    }
    /**
     * @stability stable
     */
    resetDisableExecuteApiEndpoint() {
        this._disableExecuteApiEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableExecuteApiEndpointInput() {
        return this._disableExecuteApiEndpoint;
    }
    // execution_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get executionArn() {
        return this.getStringAttribute('execution_arn');
    }
    /**
     * @stability stable
     */
    get failOnWarnings() {
        return this.getBooleanAttribute('fail_on_warnings');
    }
    /**
     * @stability stable
     */
    set failOnWarnings(value) {
        this._failOnWarnings = value;
    }
    /**
     * @stability stable
     */
    resetFailOnWarnings() {
        this._failOnWarnings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get failOnWarningsInput() {
        return this._failOnWarnings;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get protocolType() {
        return this.getStringAttribute('protocol_type');
    }
    /**
     * @stability stable
     */
    set protocolType(value) {
        this._protocolType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get protocolTypeInput() {
        return this._protocolType;
    }
    /**
     * @stability stable
     */
    get routeKey() {
        return this.getStringAttribute('route_key');
    }
    /**
     * @stability stable
     */
    set routeKey(value) {
        this._routeKey = value;
    }
    /**
     * @stability stable
     */
    resetRouteKey() {
        this._routeKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeKeyInput() {
        return this._routeKey;
    }
    /**
     * @stability stable
     */
    get routeSelectionExpression() {
        return this.getStringAttribute('route_selection_expression');
    }
    /**
     * @stability stable
     */
    set routeSelectionExpression(value) {
        this._routeSelectionExpression = value;
    }
    /**
     * @stability stable
     */
    resetRouteSelectionExpression() {
        this._routeSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routeSelectionExpressionInput() {
        return this._routeSelectionExpression;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get target() {
        return this.getStringAttribute('target');
    }
    /**
     * @stability stable
     */
    set target(value) {
        this._target = value;
    }
    /**
     * @stability stable
     */
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetInput() {
        return this._target;
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    /**
     * @stability stable
     */
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
    /**
     * @stability stable
     */
    get corsConfiguration() {
        return this._corsConfiguration;
    }
    /**
     * @stability stable
     */
    putCorsConfiguration(value) {
        this._corsConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCorsConfiguration() {
        this._corsConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get corsConfigurationInput() {
        return this._corsConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_key_selection_expression: cdktf.stringToTerraform(this._apiKeySelectionExpression),
            body: cdktf.stringToTerraform(this._body),
            credentials_arn: cdktf.stringToTerraform(this._credentialsArn),
            description: cdktf.stringToTerraform(this._description),
            disable_execute_api_endpoint: cdktf.booleanToTerraform(this._disableExecuteApiEndpoint),
            fail_on_warnings: cdktf.booleanToTerraform(this._failOnWarnings),
            name: cdktf.stringToTerraform(this._name),
            protocol_type: cdktf.stringToTerraform(this._protocolType),
            route_key: cdktf.stringToTerraform(this._routeKey),
            route_selection_expression: cdktf.stringToTerraform(this._routeSelectionExpression),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            target: cdktf.stringToTerraform(this._target),
            version: cdktf.stringToTerraform(this._version),
            cors_configuration: apigatewayv2ApiCorsConfigurationToTerraform(this._corsConfiguration.internalValue),
        };
    }
}
exports.Apigatewayv2Api = Apigatewayv2Api;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2Api[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2Api", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2Api.tfResourceType = "aws_apigatewayv2_api";
