// @cdktf/provider-aws/lib/apigatewayv2/apigatewayv2-authorizer.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Authorizer = exports.Apigatewayv2AuthorizerJwtConfigurationOutputReference = exports.apigatewayv2AuthorizerJwtConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2AuthorizerJwtConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audience: cdktf.listMapper(cdktf.stringToTerraform)(struct.audience),
        issuer: cdktf.stringToTerraform(struct.issuer),
    };
}
exports.apigatewayv2AuthorizerJwtConfigurationToTerraform = apigatewayv2AuthorizerJwtConfigurationToTerraform;
/**
 * @stability stable
 */
class Apigatewayv2AuthorizerJwtConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audience) {
            hasAnyValues = true;
            internalValueResult.audience = this._audience;
        }
        if (this._issuer) {
            hasAnyValues = true;
            internalValueResult.issuer = this._issuer;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audience = undefined;
            this._issuer = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audience = value.audience;
            this._issuer = value.issuer;
        }
    }
    /**
     * @stability stable
     */
    get audience() {
        return this.getListAttribute('audience');
    }
    /**
     * @stability stable
     */
    set audience(value) {
        this._audience = value;
    }
    /**
     * @stability stable
     */
    resetAudience() {
        this._audience = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get audienceInput() {
        return this._audience;
    }
    /**
     * @stability stable
     */
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    /**
     * @stability stable
     */
    set issuer(value) {
        this._issuer = value;
    }
    /**
     * @stability stable
     */
    resetIssuer() {
        this._issuer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get issuerInput() {
        return this._issuer;
    }
}
exports.Apigatewayv2AuthorizerJwtConfigurationOutputReference = Apigatewayv2AuthorizerJwtConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2AuthorizerJwtConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2AuthorizerJwtConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_authorizer.html aws_apigatewayv2_authorizer}.
 *
 * @stability stable
 */
class Apigatewayv2Authorizer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_authorizer.html aws_apigatewayv2_authorizer} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_authorizer',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // jwt_configuration - computed: false, optional: true, required: false
        this._jwtConfiguration = new Apigatewayv2AuthorizerJwtConfigurationOutputReference(this, "jwt_configuration", true);
        this._apiId = config.apiId;
        this._authorizerCredentialsArn = config.authorizerCredentialsArn;
        this._authorizerPayloadFormatVersion = config.authorizerPayloadFormatVersion;
        this._authorizerResultTtlInSeconds = config.authorizerResultTtlInSeconds;
        this._authorizerType = config.authorizerType;
        this._authorizerUri = config.authorizerUri;
        this._enableSimpleResponses = config.enableSimpleResponses;
        this._identitySources = config.identitySources;
        this._name = config.name;
        this._jwtConfiguration.internalValue = config.jwtConfiguration;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    /**
     * @stability stable
     */
    get authorizerCredentialsArn() {
        return this.getStringAttribute('authorizer_credentials_arn');
    }
    /**
     * @stability stable
     */
    set authorizerCredentialsArn(value) {
        this._authorizerCredentialsArn = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizerCredentialsArn() {
        this._authorizerCredentialsArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerCredentialsArnInput() {
        return this._authorizerCredentialsArn;
    }
    /**
     * @stability stable
     */
    get authorizerPayloadFormatVersion() {
        return this.getStringAttribute('authorizer_payload_format_version');
    }
    /**
     * @stability stable
     */
    set authorizerPayloadFormatVersion(value) {
        this._authorizerPayloadFormatVersion = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizerPayloadFormatVersion() {
        this._authorizerPayloadFormatVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerPayloadFormatVersionInput() {
        return this._authorizerPayloadFormatVersion;
    }
    /**
     * @stability stable
     */
    get authorizerResultTtlInSeconds() {
        return this.getNumberAttribute('authorizer_result_ttl_in_seconds');
    }
    /**
     * @stability stable
     */
    set authorizerResultTtlInSeconds(value) {
        this._authorizerResultTtlInSeconds = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizerResultTtlInSeconds() {
        this._authorizerResultTtlInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerResultTtlInSecondsInput() {
        return this._authorizerResultTtlInSeconds;
    }
    /**
     * @stability stable
     */
    get authorizerType() {
        return this.getStringAttribute('authorizer_type');
    }
    /**
     * @stability stable
     */
    set authorizerType(value) {
        this._authorizerType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerTypeInput() {
        return this._authorizerType;
    }
    /**
     * @stability stable
     */
    get authorizerUri() {
        return this.getStringAttribute('authorizer_uri');
    }
    /**
     * @stability stable
     */
    set authorizerUri(value) {
        this._authorizerUri = value;
    }
    /**
     * @stability stable
     */
    resetAuthorizerUri() {
        this._authorizerUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerUriInput() {
        return this._authorizerUri;
    }
    /**
     * @stability stable
     */
    get enableSimpleResponses() {
        return this.getBooleanAttribute('enable_simple_responses');
    }
    /**
     * @stability stable
     */
    set enableSimpleResponses(value) {
        this._enableSimpleResponses = value;
    }
    /**
     * @stability stable
     */
    resetEnableSimpleResponses() {
        this._enableSimpleResponses = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableSimpleResponsesInput() {
        return this._enableSimpleResponses;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get identitySources() {
        return this.getListAttribute('identity_sources');
    }
    /**
     * @stability stable
     */
    set identitySources(value) {
        this._identitySources = value;
    }
    /**
     * @stability stable
     */
    resetIdentitySources() {
        this._identitySources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get identitySourcesInput() {
        return this._identitySources;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get jwtConfiguration() {
        return this._jwtConfiguration;
    }
    /**
     * @stability stable
     */
    putJwtConfiguration(value) {
        this._jwtConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetJwtConfiguration() {
        this._jwtConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jwtConfigurationInput() {
        return this._jwtConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            authorizer_credentials_arn: cdktf.stringToTerraform(this._authorizerCredentialsArn),
            authorizer_payload_format_version: cdktf.stringToTerraform(this._authorizerPayloadFormatVersion),
            authorizer_result_ttl_in_seconds: cdktf.numberToTerraform(this._authorizerResultTtlInSeconds),
            authorizer_type: cdktf.stringToTerraform(this._authorizerType),
            authorizer_uri: cdktf.stringToTerraform(this._authorizerUri),
            enable_simple_responses: cdktf.booleanToTerraform(this._enableSimpleResponses),
            identity_sources: cdktf.listMapper(cdktf.stringToTerraform)(this._identitySources),
            name: cdktf.stringToTerraform(this._name),
            jwt_configuration: apigatewayv2AuthorizerJwtConfigurationToTerraform(this._jwtConfiguration.internalValue),
        };
    }
}
exports.Apigatewayv2Authorizer = Apigatewayv2Authorizer;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2Authorizer[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2.Apigatewayv2Authorizer", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Apigatewayv2Authorizer.tfResourceType = "aws_apigatewayv2_authorizer";
