// @cdktf/provider-aws/lib/efs/efs-file-system.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EfsFileSystem = exports.efsFileSystemLifecyclePolicyToTerraform = exports.EfsFileSystemSizeInBytes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class EfsFileSystemSizeInBytes extends cdktf.ComplexComputedList {
    // value - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get value() {
        return this.getNumberAttribute('value');
    }
    // value_in_ia - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get valueInIa() {
        return this.getNumberAttribute('value_in_ia');
    }
    // value_in_standard - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get valueInStandard() {
        return this.getNumberAttribute('value_in_standard');
    }
}
exports.EfsFileSystemSizeInBytes = EfsFileSystemSizeInBytes;
_a = JSII_RTTI_SYMBOL_1;
EfsFileSystemSizeInBytes[_a] = { fqn: "@cdktf/provider-aws.efs.EfsFileSystemSizeInBytes", version: "3.0.1" };
function efsFileSystemLifecyclePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        transition_to_ia: cdktf.stringToTerraform(struct.transitionToIa),
        transition_to_primary_storage_class: cdktf.stringToTerraform(struct.transitionToPrimaryStorageClass),
    };
}
exports.efsFileSystemLifecyclePolicyToTerraform = efsFileSystemLifecyclePolicyToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/efs_file_system.html aws_efs_file_system}.
 *
 * @stability stable
 */
class EfsFileSystem extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/efs_file_system.html aws_efs_file_system} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._availabilityZoneName = config.availabilityZoneName;
        this._creationToken = config.creationToken;
        this._encrypted = config.encrypted;
        this._kmsKeyId = config.kmsKeyId;
        this._performanceMode = config.performanceMode;
        this._provisionedThroughputInMibps = config.provisionedThroughputInMibps;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._throughputMode = config.throughputMode;
        this._lifecyclePolicy = config.lifecyclePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZoneId() {
        return this.getStringAttribute('availability_zone_id');
    }
    /**
     * @stability stable
     */
    get availabilityZoneName() {
        return this.getStringAttribute('availability_zone_name');
    }
    /**
     * @stability stable
     */
    set availabilityZoneName(value) {
        this._availabilityZoneName = value;
    }
    /**
     * @stability stable
     */
    resetAvailabilityZoneName() {
        this._availabilityZoneName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get availabilityZoneNameInput() {
        return this._availabilityZoneName;
    }
    /**
     * @stability stable
     */
    get creationToken() {
        return this.getStringAttribute('creation_token');
    }
    /**
     * @stability stable
     */
    set creationToken(value) {
        this._creationToken = value;
    }
    /**
     * @stability stable
     */
    resetCreationToken() {
        this._creationToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get creationTokenInput() {
        return this._creationToken;
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    /**
     * @stability stable
     */
    set encrypted(value) {
        this._encrypted = value;
    }
    /**
     * @stability stable
     */
    resetEncrypted() {
        this._encrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptedInput() {
        return this._encrypted;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    // number_of_mount_targets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get numberOfMountTargets() {
        return this.getNumberAttribute('number_of_mount_targets');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get performanceMode() {
        return this.getStringAttribute('performance_mode');
    }
    /**
     * @stability stable
     */
    set performanceMode(value) {
        this._performanceMode = value;
    }
    /**
     * @stability stable
     */
    resetPerformanceMode() {
        this._performanceMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get performanceModeInput() {
        return this._performanceMode;
    }
    /**
     * @stability stable
     */
    get provisionedThroughputInMibps() {
        return this.getNumberAttribute('provisioned_throughput_in_mibps');
    }
    /**
     * @stability stable
     */
    set provisionedThroughputInMibps(value) {
        this._provisionedThroughputInMibps = value;
    }
    /**
     * @stability stable
     */
    resetProvisionedThroughputInMibps() {
        this._provisionedThroughputInMibps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get provisionedThroughputInMibpsInput() {
        return this._provisionedThroughputInMibps;
    }
    // size_in_bytes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    sizeInBytes(index) {
        return new EfsFileSystemSizeInBytes(this, 'size_in_bytes', index);
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get throughputMode() {
        return this.getStringAttribute('throughput_mode');
    }
    /**
     * @stability stable
     */
    set throughputMode(value) {
        this._throughputMode = value;
    }
    /**
     * @stability stable
     */
    resetThroughputMode() {
        this._throughputMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get throughputModeInput() {
        return this._throughputMode;
    }
    /**
     * @stability stable
     */
    get lifecyclePolicy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('lifecycle_policy');
    }
    /**
     * @stability stable
     */
    set lifecyclePolicy(value) {
        this._lifecyclePolicy = value;
    }
    /**
     * @stability stable
     */
    resetLifecyclePolicy() {
        this._lifecyclePolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lifecyclePolicyInput() {
        return this._lifecyclePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            availability_zone_name: cdktf.stringToTerraform(this._availabilityZoneName),
            creation_token: cdktf.stringToTerraform(this._creationToken),
            encrypted: cdktf.booleanToTerraform(this._encrypted),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            performance_mode: cdktf.stringToTerraform(this._performanceMode),
            provisioned_throughput_in_mibps: cdktf.numberToTerraform(this._provisionedThroughputInMibps),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            throughput_mode: cdktf.stringToTerraform(this._throughputMode),
            lifecycle_policy: cdktf.listMapper(efsFileSystemLifecyclePolicyToTerraform)(this._lifecyclePolicy),
        };
    }
}
exports.EfsFileSystem = EfsFileSystem;
_b = JSII_RTTI_SYMBOL_1;
EfsFileSystem[_b] = { fqn: "@cdktf/provider-aws.efs.EfsFileSystem", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EfsFileSystem.tfResourceType = "aws_efs_file_system";
