// @cdktf/provider-aws/lib/efs/efs-backup-policy.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EfsBackupPolicy = exports.EfsBackupPolicyBackupPolicyOutputReference = exports.efsBackupPolicyBackupPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function efsBackupPolicyBackupPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.efsBackupPolicyBackupPolicyToTerraform = efsBackupPolicyBackupPolicyToTerraform;
/**
 * @stability stable
 */
class EfsBackupPolicyBackupPolicyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
}
exports.EfsBackupPolicyBackupPolicyOutputReference = EfsBackupPolicyBackupPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EfsBackupPolicyBackupPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.efs.EfsBackupPolicyBackupPolicyOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/efs_backup_policy.html aws_efs_backup_policy}.
 *
 * @stability stable
 */
class EfsBackupPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/efs_backup_policy.html aws_efs_backup_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_backup_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // backup_policy - computed: false, optional: false, required: true
        this._backupPolicy = new EfsBackupPolicyBackupPolicyOutputReference(this, "backup_policy", true);
        this._fileSystemId = config.fileSystemId;
        this._backupPolicy.internalValue = config.backupPolicy;
    }
    /**
     * @stability stable
     */
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    /**
     * @stability stable
     */
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get backupPolicy() {
        return this._backupPolicy;
    }
    /**
     * @stability stable
     */
    putBackupPolicy(value) {
        this._backupPolicy.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get backupPolicyInput() {
        return this._backupPolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            file_system_id: cdktf.stringToTerraform(this._fileSystemId),
            backup_policy: efsBackupPolicyBackupPolicyToTerraform(this._backupPolicy.internalValue),
        };
    }
}
exports.EfsBackupPolicy = EfsBackupPolicy;
_b = JSII_RTTI_SYMBOL_1;
EfsBackupPolicy[_b] = { fqn: "@cdktf/provider-aws.efs.EfsBackupPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EfsBackupPolicy.tfResourceType = "aws_efs_backup_policy";
