// @cdktf/provider-aws/lib/acm/data-aws-acmpca-certificate-authority.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAcmpcaCertificateAuthority = exports.dataAwsAcmpcaCertificateAuthorityRevocationConfigurationToTerraform = exports.dataAwsAcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform = dataAwsAcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform;
function dataAwsAcmpcaCertificateAuthorityRevocationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crl_configuration: cdktf.listMapper(dataAwsAcmpcaCertificateAuthorityRevocationConfigurationCrlConfigurationToTerraform)(struct.crlConfiguration),
    };
}
exports.dataAwsAcmpcaCertificateAuthorityRevocationConfigurationToTerraform = dataAwsAcmpcaCertificateAuthorityRevocationConfigurationToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/acmpca_certificate_authority.html aws_acmpca_certificate_authority}.
 *
 * @stability stable
 */
class DataAwsAcmpcaCertificateAuthority extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/acmpca_certificate_authority.html aws_acmpca_certificate_authority} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_acmpca_certificate_authority',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
        this._tags = config.tags;
        this._revocationConfiguration = config.revocationConfiguration;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // certificate - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    // certificate_chain - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateChain() {
        return this.getStringAttribute('certificate_chain');
    }
    // certificate_signing_request - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateSigningRequest() {
        return this.getStringAttribute('certificate_signing_request');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // not_after - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get notAfter() {
        return this.getStringAttribute('not_after');
    }
    // not_before - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get notBefore() {
        return this.getStringAttribute('not_before');
    }
    // serial - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serial() {
        return this.getStringAttribute('serial');
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    get revocationConfiguration() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('revocation_configuration');
    }
    /**
     * @stability stable
     */
    set revocationConfiguration(value) {
        this._revocationConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetRevocationConfiguration() {
        this._revocationConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get revocationConfigurationInput() {
        return this._revocationConfiguration;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            revocation_configuration: cdktf.listMapper(dataAwsAcmpcaCertificateAuthorityRevocationConfigurationToTerraform)(this._revocationConfiguration),
        };
    }
}
exports.DataAwsAcmpcaCertificateAuthority = DataAwsAcmpcaCertificateAuthority;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAcmpcaCertificateAuthority[_a] = { fqn: "@cdktf/provider-aws.acm.DataAwsAcmpcaCertificateAuthority", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAcmpcaCertificateAuthority.tfResourceType = "aws_acmpca_certificate_authority";
