// @cdktf/provider-aws/lib/kinesis/kinesis-analytics-application.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisAnalyticsApplication = exports.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform = exports.kinesisAnalyticsApplicationOutputsToTerraform = exports.KinesisAnalyticsApplicationOutputsSchemaOutputReference = exports.kinesisAnalyticsApplicationOutputsSchemaToTerraform = exports.KinesisAnalyticsApplicationOutputsLambdaOutputReference = exports.kinesisAnalyticsApplicationOutputsLambdaToTerraform = exports.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference = exports.kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform = exports.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference = exports.kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform = exports.KinesisAnalyticsApplicationInputsOutputReference = exports.kinesisAnalyticsApplicationInputsToTerraform = exports.kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform = exports.kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform = exports.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference = exports.kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform = exports.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference = exports.kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform = exports.KinesisAnalyticsApplicationInputsParallelismOutputReference = exports.kinesisAnalyticsApplicationInputsParallelismToTerraform = exports.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference = exports.kinesisAnalyticsApplicationInputsKinesisStreamToTerraform = exports.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference = exports.kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform = exports.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference = exports.kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_stream_arn: cdktf.stringToTerraform(struct.logStreamArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform = kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logStreamArn) {
            hasAnyValues = true;
            internalValueResult.logStreamArn = this._logStreamArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logStreamArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logStreamArn = value.logStreamArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get logStreamArn() {
        return this.getStringAttribute('log_stream_arn');
    }
    /**
     * @stability stable
     */
    set logStreamArn(value) {
        this._logStreamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logStreamArnInput() {
        return this._logStreamArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference = KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform = kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference = KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference;
_b = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference[_b] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsKinesisStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsKinesisStreamToTerraform = kinesisAnalyticsApplicationInputsKinesisStreamToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsKinesisStreamOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference = KinesisAnalyticsApplicationInputsKinesisStreamOutputReference;
_c = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsKinesisStreamOutputReference[_c] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsParallelismToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
    };
}
exports.kinesisAnalyticsApplicationInputsParallelismToTerraform = kinesisAnalyticsApplicationInputsParallelismToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsParallelismOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
        }
    }
    /**
     * @stability stable
     */
    get count() {
        return this.getNumberAttribute('count');
    }
    /**
     * @stability stable
     */
    set count(value) {
        this._count = value;
    }
    /**
     * @stability stable
     */
    resetCount() {
        this._count = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get countInput() {
        return this._count;
    }
}
exports.KinesisAnalyticsApplicationInputsParallelismOutputReference = KinesisAnalyticsApplicationInputsParallelismOutputReference;
_d = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsParallelismOutputReference[_d] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsParallelismOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform = kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference = KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference;
_e = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference[_e] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lambda: kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform(struct.lambda),
    };
}
exports.kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform = kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // lambda - computed: false, optional: false, required: true
        this._lambda = new KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference(this, "lambda", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_0 = this._lambda) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.lambda = (_1 = this._lambda) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lambda.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lambda.internalValue = value.lambda;
        }
    }
    /**
     * @stability stable
     */
    get lambda() {
        return this._lambda;
    }
    /**
     * @stability stable
     */
    putLambda(value) {
        this._lambda.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaInput() {
        return this._lambda.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference = KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference[_f] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping: cdktf.stringToTerraform(struct.mapping),
        name: cdktf.stringToTerraform(struct.name),
        sql_type: cdktf.stringToTerraform(struct.sqlType),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform;
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_column_delimiter: cdktf.stringToTerraform(struct.recordColumnDelimiter),
        record_row_delimiter: cdktf.stringToTerraform(struct.recordRowDelimiter),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordColumnDelimiter) {
            hasAnyValues = true;
            internalValueResult.recordColumnDelimiter = this._recordColumnDelimiter;
        }
        if (this._recordRowDelimiter) {
            hasAnyValues = true;
            internalValueResult.recordRowDelimiter = this._recordRowDelimiter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordColumnDelimiter = undefined;
            this._recordRowDelimiter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordColumnDelimiter = value.recordColumnDelimiter;
            this._recordRowDelimiter = value.recordRowDelimiter;
        }
    }
    /**
     * @stability stable
     */
    get recordColumnDelimiter() {
        return this.getStringAttribute('record_column_delimiter');
    }
    /**
     * @stability stable
     */
    set recordColumnDelimiter(value) {
        this._recordColumnDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordColumnDelimiterInput() {
        return this._recordColumnDelimiter;
    }
    /**
     * @stability stable
     */
    get recordRowDelimiter() {
        return this.getStringAttribute('record_row_delimiter');
    }
    /**
     * @stability stable
     */
    set recordRowDelimiter(value) {
        this._recordRowDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordRowDelimiterInput() {
        return this._recordRowDelimiter;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference;
_g = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference[_g] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_row_path: cdktf.stringToTerraform(struct.recordRowPath),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordRowPath) {
            hasAnyValues = true;
            internalValueResult.recordRowPath = this._recordRowPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordRowPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordRowPath = value.recordRowPath;
        }
    }
    /**
     * @stability stable
     */
    get recordRowPath() {
        return this.getStringAttribute('record_row_path');
    }
    /**
     * @stability stable
     */
    set recordRowPath(value) {
        this._recordRowPath = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordRowPathInput() {
        return this._recordRowPath;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference;
_h = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference[_h] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        csv: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform(struct.csv),
        json: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform(struct.json),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // csv - computed: false, optional: true, required: false
        this._csv = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference(this, "csv", true);
        // json - computed: false, optional: true, required: false
        this._json = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference(this, "json", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1, _2, _3;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_0 = this._csv) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.csv = (_1 = this._csv) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        if ((_2 = this._json) === null || _2 === void 0 ? void 0 : _2.internalValue) {
            hasAnyValues = true;
            internalValueResult.json = (_3 = this._json) === null || _3 === void 0 ? void 0 : _3.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._csv.internalValue = undefined;
            this._json.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._csv.internalValue = value.csv;
            this._json.internalValue = value.json;
        }
    }
    /**
     * @stability stable
     */
    get csv() {
        return this._csv;
    }
    /**
     * @stability stable
     */
    putCsv(value) {
        this._csv.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCsv() {
        this._csv.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get csvInput() {
        return this._csv.internalValue;
    }
    /**
     * @stability stable
     */
    get json() {
        return this._json;
    }
    /**
     * @stability stable
     */
    putJson(value) {
        this._json.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetJson() {
        this._json.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jsonInput() {
        return this._json.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference;
_j = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference[_j] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping_parameters: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform(struct.mappingParameters),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // mapping_parameters - computed: false, optional: true, required: false
        this._mappingParameters = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference(this, "mapping_parameters", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_0 = this._mappingParameters) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.mappingParameters = (_1 = this._mappingParameters) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mappingParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mappingParameters.internalValue = value.mappingParameters;
        }
    }
    /**
     * @stability stable
     */
    get mappingParameters() {
        return this._mappingParameters;
    }
    /**
     * @stability stable
     */
    putMappingParameters(value) {
        this._mappingParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMappingParameters() {
        this._mappingParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get mappingParametersInput() {
        return this._mappingParameters.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference;
_k = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference[_k] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_encoding: cdktf.stringToTerraform(struct.recordEncoding),
        record_columns: cdktf.listMapper(kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform)(struct.recordColumns),
        record_format: kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform(struct.recordFormat),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaToTerraform = kinesisAnalyticsApplicationInputsSchemaToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsSchemaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // record_format - computed: false, optional: false, required: true
        this._recordFormat = new KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference(this, "record_format", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordEncoding) {
            hasAnyValues = true;
            internalValueResult.recordEncoding = this._recordEncoding;
        }
        if (this._recordColumns) {
            hasAnyValues = true;
            internalValueResult.recordColumns = this._recordColumns;
        }
        if ((_0 = this._recordFormat) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.recordFormat = (_1 = this._recordFormat) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordEncoding = undefined;
            this._recordColumns = undefined;
            this._recordFormat.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordEncoding = value.recordEncoding;
            this._recordColumns = value.recordColumns;
            this._recordFormat.internalValue = value.recordFormat;
        }
    }
    /**
     * @stability stable
     */
    get recordEncoding() {
        return this.getStringAttribute('record_encoding');
    }
    /**
     * @stability stable
     */
    set recordEncoding(value) {
        this._recordEncoding = value;
    }
    /**
     * @stability stable
     */
    resetRecordEncoding() {
        this._recordEncoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordEncodingInput() {
        return this._recordEncoding;
    }
    /**
     * @stability stable
     */
    get recordColumns() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('record_columns');
    }
    /**
     * @stability stable
     */
    set recordColumns(value) {
        this._recordColumns = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordColumnsInput() {
        return this._recordColumns;
    }
    /**
     * @stability stable
     */
    get recordFormat() {
        return this._recordFormat;
    }
    /**
     * @stability stable
     */
    putRecordFormat(value) {
        this._recordFormat.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordFormatInput() {
        return this._recordFormat.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaOutputReference = KinesisAnalyticsApplicationInputsSchemaOutputReference;
_l = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaOutputReference[_l] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsSchemaOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        starting_position: cdktf.stringToTerraform(struct.startingPosition),
    };
}
exports.kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform = kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform;
function kinesisAnalyticsApplicationInputsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name_prefix: cdktf.stringToTerraform(struct.namePrefix),
        kinesis_firehose: kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform(struct.kinesisFirehose),
        kinesis_stream: kinesisAnalyticsApplicationInputsKinesisStreamToTerraform(struct.kinesisStream),
        parallelism: kinesisAnalyticsApplicationInputsParallelismToTerraform(struct.parallelism),
        processing_configuration: kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform(struct.processingConfiguration),
        schema: kinesisAnalyticsApplicationInputsSchemaToTerraform(struct.schema),
        starting_position_configuration: cdktf.listMapper(kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform)(struct.startingPositionConfiguration),
    };
}
exports.kinesisAnalyticsApplicationInputsToTerraform = kinesisAnalyticsApplicationInputsToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationInputsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // kinesis_firehose - computed: false, optional: true, required: false
        this._kinesisFirehose = new KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference(this, "kinesis_firehose", true);
        // kinesis_stream - computed: false, optional: true, required: false
        this._kinesisStream = new KinesisAnalyticsApplicationInputsKinesisStreamOutputReference(this, "kinesis_stream", true);
        // parallelism - computed: false, optional: true, required: false
        this._parallelism = new KinesisAnalyticsApplicationInputsParallelismOutputReference(this, "parallelism", true);
        // processing_configuration - computed: false, optional: true, required: false
        this._processingConfiguration = new KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference(this, "processing_configuration", true);
        // schema - computed: false, optional: false, required: true
        this._schema = new KinesisAnalyticsApplicationInputsSchemaOutputReference(this, "schema", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1, _2, _3, _4, _5, _6, _7, _8, _9;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namePrefix) {
            hasAnyValues = true;
            internalValueResult.namePrefix = this._namePrefix;
        }
        if ((_0 = this._kinesisFirehose) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.kinesisFirehose = (_1 = this._kinesisFirehose) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        if ((_2 = this._kinesisStream) === null || _2 === void 0 ? void 0 : _2.internalValue) {
            hasAnyValues = true;
            internalValueResult.kinesisStream = (_3 = this._kinesisStream) === null || _3 === void 0 ? void 0 : _3.internalValue;
        }
        if ((_4 = this._parallelism) === null || _4 === void 0 ? void 0 : _4.internalValue) {
            hasAnyValues = true;
            internalValueResult.parallelism = (_5 = this._parallelism) === null || _5 === void 0 ? void 0 : _5.internalValue;
        }
        if ((_6 = this._processingConfiguration) === null || _6 === void 0 ? void 0 : _6.internalValue) {
            hasAnyValues = true;
            internalValueResult.processingConfiguration = (_7 = this._processingConfiguration) === null || _7 === void 0 ? void 0 : _7.internalValue;
        }
        if ((_8 = this._schema) === null || _8 === void 0 ? void 0 : _8.internalValue) {
            hasAnyValues = true;
            internalValueResult.schema = (_9 = this._schema) === null || _9 === void 0 ? void 0 : _9.internalValue;
        }
        if (this._startingPositionConfiguration) {
            hasAnyValues = true;
            internalValueResult.startingPositionConfiguration = this._startingPositionConfiguration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namePrefix = undefined;
            this._kinesisFirehose.internalValue = undefined;
            this._kinesisStream.internalValue = undefined;
            this._parallelism.internalValue = undefined;
            this._processingConfiguration.internalValue = undefined;
            this._schema.internalValue = undefined;
            this._startingPositionConfiguration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namePrefix = value.namePrefix;
            this._kinesisFirehose.internalValue = value.kinesisFirehose;
            this._kinesisStream.internalValue = value.kinesisStream;
            this._parallelism.internalValue = value.parallelism;
            this._processingConfiguration.internalValue = value.processingConfiguration;
            this._schema.internalValue = value.schema;
            this._startingPositionConfiguration = value.startingPositionConfiguration;
        }
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get kinesisFirehose() {
        return this._kinesisFirehose;
    }
    /**
     * @stability stable
     */
    putKinesisFirehose(value) {
        this._kinesisFirehose.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetKinesisFirehose() {
        this._kinesisFirehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kinesisFirehoseInput() {
        return this._kinesisFirehose.internalValue;
    }
    /**
     * @stability stable
     */
    get kinesisStream() {
        return this._kinesisStream;
    }
    /**
     * @stability stable
     */
    putKinesisStream(value) {
        this._kinesisStream.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetKinesisStream() {
        this._kinesisStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kinesisStreamInput() {
        return this._kinesisStream.internalValue;
    }
    /**
     * @stability stable
     */
    get parallelism() {
        return this._parallelism;
    }
    /**
     * @stability stable
     */
    putParallelism(value) {
        this._parallelism.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetParallelism() {
        this._parallelism.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parallelismInput() {
        return this._parallelism.internalValue;
    }
    /**
     * @stability stable
     */
    get processingConfiguration() {
        return this._processingConfiguration;
    }
    /**
     * @stability stable
     */
    putProcessingConfiguration(value) {
        this._processingConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetProcessingConfiguration() {
        this._processingConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get processingConfigurationInput() {
        return this._processingConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get schema() {
        return this._schema;
    }
    /**
     * @stability stable
     */
    putSchema(value) {
        this._schema.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaInput() {
        return this._schema.internalValue;
    }
    /**
     * @stability stable
     */
    get startingPositionConfiguration() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('starting_position_configuration');
    }
    /**
     * @stability stable
     */
    set startingPositionConfiguration(value) {
        this._startingPositionConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetStartingPositionConfiguration() {
        this._startingPositionConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startingPositionConfigurationInput() {
        return this._startingPositionConfiguration;
    }
}
exports.KinesisAnalyticsApplicationInputsOutputReference = KinesisAnalyticsApplicationInputsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsOutputReference[_m] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationInputsOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform = kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference = KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference;
_o = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference[_o] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform = kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference = KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference;
_p = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference[_p] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationOutputsLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsLambdaToTerraform = kinesisAnalyticsApplicationOutputsLambdaToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationOutputsLambdaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsLambdaOutputReference = KinesisAnalyticsApplicationOutputsLambdaOutputReference;
_q = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsLambdaOutputReference[_q] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationOutputsLambdaOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationOutputsSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_format_type: cdktf.stringToTerraform(struct.recordFormatType),
    };
}
exports.kinesisAnalyticsApplicationOutputsSchemaToTerraform = kinesisAnalyticsApplicationOutputsSchemaToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationOutputsSchemaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordFormatType) {
            hasAnyValues = true;
            internalValueResult.recordFormatType = this._recordFormatType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordFormatType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordFormatType = value.recordFormatType;
        }
    }
    /**
     * @stability stable
     */
    get recordFormatType() {
        return this.getStringAttribute('record_format_type');
    }
    /**
     * @stability stable
     */
    set recordFormatType(value) {
        this._recordFormatType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordFormatTypeInput() {
        return this._recordFormatType;
    }
}
exports.KinesisAnalyticsApplicationOutputsSchemaOutputReference = KinesisAnalyticsApplicationOutputsSchemaOutputReference;
_r = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsSchemaOutputReference[_r] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationOutputsSchemaOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationOutputsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        kinesis_firehose: kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform(struct.kinesisFirehose),
        kinesis_stream: kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform(struct.kinesisStream),
        lambda: kinesisAnalyticsApplicationOutputsLambdaToTerraform(struct.lambda),
        schema: kinesisAnalyticsApplicationOutputsSchemaToTerraform(struct.schema),
    };
}
exports.kinesisAnalyticsApplicationOutputsToTerraform = kinesisAnalyticsApplicationOutputsToTerraform;
function kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_arn: cdktf.stringToTerraform(struct.bucketArn),
        file_key: cdktf.stringToTerraform(struct.fileKey),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketArn) {
            hasAnyValues = true;
            internalValueResult.bucketArn = this._bucketArn;
        }
        if (this._fileKey) {
            hasAnyValues = true;
            internalValueResult.fileKey = this._fileKey;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketArn = undefined;
            this._fileKey = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketArn = value.bucketArn;
            this._fileKey = value.fileKey;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get bucketArn() {
        return this.getStringAttribute('bucket_arn');
    }
    /**
     * @stability stable
     */
    set bucketArn(value) {
        this._bucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketArnInput() {
        return this._bucketArn;
    }
    /**
     * @stability stable
     */
    get fileKey() {
        return this.getStringAttribute('file_key');
    }
    /**
     * @stability stable
     */
    set fileKey(value) {
        this._fileKey = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fileKeyInput() {
        return this._fileKey;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference = KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference;
_s = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference[_s] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping: cdktf.stringToTerraform(struct.mapping),
        name: cdktf.stringToTerraform(struct.name),
        sql_type: cdktf.stringToTerraform(struct.sqlType),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform;
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_column_delimiter: cdktf.stringToTerraform(struct.recordColumnDelimiter),
        record_row_delimiter: cdktf.stringToTerraform(struct.recordRowDelimiter),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordColumnDelimiter) {
            hasAnyValues = true;
            internalValueResult.recordColumnDelimiter = this._recordColumnDelimiter;
        }
        if (this._recordRowDelimiter) {
            hasAnyValues = true;
            internalValueResult.recordRowDelimiter = this._recordRowDelimiter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordColumnDelimiter = undefined;
            this._recordRowDelimiter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordColumnDelimiter = value.recordColumnDelimiter;
            this._recordRowDelimiter = value.recordRowDelimiter;
        }
    }
    /**
     * @stability stable
     */
    get recordColumnDelimiter() {
        return this.getStringAttribute('record_column_delimiter');
    }
    /**
     * @stability stable
     */
    set recordColumnDelimiter(value) {
        this._recordColumnDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordColumnDelimiterInput() {
        return this._recordColumnDelimiter;
    }
    /**
     * @stability stable
     */
    get recordRowDelimiter() {
        return this.getStringAttribute('record_row_delimiter');
    }
    /**
     * @stability stable
     */
    set recordRowDelimiter(value) {
        this._recordRowDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordRowDelimiterInput() {
        return this._recordRowDelimiter;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference;
_t = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference[_t] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_row_path: cdktf.stringToTerraform(struct.recordRowPath),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordRowPath) {
            hasAnyValues = true;
            internalValueResult.recordRowPath = this._recordRowPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordRowPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordRowPath = value.recordRowPath;
        }
    }
    /**
     * @stability stable
     */
    get recordRowPath() {
        return this.getStringAttribute('record_row_path');
    }
    /**
     * @stability stable
     */
    set recordRowPath(value) {
        this._recordRowPath = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordRowPathInput() {
        return this._recordRowPath;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference;
_u = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference[_u] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        csv: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform(struct.csv),
        json: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform(struct.json),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // csv - computed: false, optional: true, required: false
        this._csv = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference(this, "csv", true);
        // json - computed: false, optional: true, required: false
        this._json = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference(this, "json", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1, _2, _3;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_0 = this._csv) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.csv = (_1 = this._csv) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        if ((_2 = this._json) === null || _2 === void 0 ? void 0 : _2.internalValue) {
            hasAnyValues = true;
            internalValueResult.json = (_3 = this._json) === null || _3 === void 0 ? void 0 : _3.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._csv.internalValue = undefined;
            this._json.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._csv.internalValue = value.csv;
            this._json.internalValue = value.json;
        }
    }
    /**
     * @stability stable
     */
    get csv() {
        return this._csv;
    }
    /**
     * @stability stable
     */
    putCsv(value) {
        this._csv.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCsv() {
        this._csv.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get csvInput() {
        return this._csv.internalValue;
    }
    /**
     * @stability stable
     */
    get json() {
        return this._json;
    }
    /**
     * @stability stable
     */
    putJson(value) {
        this._json.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetJson() {
        this._json.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jsonInput() {
        return this._json.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference;
_v = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference[_v] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping_parameters: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform(struct.mappingParameters),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // mapping_parameters - computed: false, optional: true, required: false
        this._mappingParameters = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference(this, "mapping_parameters", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_0 = this._mappingParameters) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.mappingParameters = (_1 = this._mappingParameters) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mappingParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mappingParameters.internalValue = value.mappingParameters;
        }
    }
    /**
     * @stability stable
     */
    get mappingParameters() {
        return this._mappingParameters;
    }
    /**
     * @stability stable
     */
    putMappingParameters(value) {
        this._mappingParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMappingParameters() {
        this._mappingParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get mappingParametersInput() {
        return this._mappingParameters.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference;
_w = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference[_w] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_encoding: cdktf.stringToTerraform(struct.recordEncoding),
        record_columns: cdktf.listMapper(kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform)(struct.recordColumns),
        record_format: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform(struct.recordFormat),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // record_format - computed: false, optional: false, required: true
        this._recordFormat = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference(this, "record_format", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordEncoding) {
            hasAnyValues = true;
            internalValueResult.recordEncoding = this._recordEncoding;
        }
        if (this._recordColumns) {
            hasAnyValues = true;
            internalValueResult.recordColumns = this._recordColumns;
        }
        if ((_0 = this._recordFormat) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.recordFormat = (_1 = this._recordFormat) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordEncoding = undefined;
            this._recordColumns = undefined;
            this._recordFormat.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordEncoding = value.recordEncoding;
            this._recordColumns = value.recordColumns;
            this._recordFormat.internalValue = value.recordFormat;
        }
    }
    /**
     * @stability stable
     */
    get recordEncoding() {
        return this.getStringAttribute('record_encoding');
    }
    /**
     * @stability stable
     */
    set recordEncoding(value) {
        this._recordEncoding = value;
    }
    /**
     * @stability stable
     */
    resetRecordEncoding() {
        this._recordEncoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordEncodingInput() {
        return this._recordEncoding;
    }
    /**
     * @stability stable
     */
    get recordColumns() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('record_columns');
    }
    /**
     * @stability stable
     */
    set recordColumns(value) {
        this._recordColumns = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordColumnsInput() {
        return this._recordColumns;
    }
    /**
     * @stability stable
     */
    get recordFormat() {
        return this._recordFormat;
    }
    /**
     * @stability stable
     */
    putRecordFormat(value) {
        this._recordFormat.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recordFormatInput() {
        return this._recordFormat.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference;
_x = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference[_x] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference", version: "3.0.1" };
function kinesisAnalyticsApplicationReferenceDataSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        table_name: cdktf.stringToTerraform(struct.tableName),
        s3: kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform(struct.s3),
        schema: kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform(struct.schema),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesToTerraform;
/**
 * @stability stable
 */
class KinesisAnalyticsApplicationReferenceDataSourcesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // s3 - computed: false, optional: false, required: true
        this._s3 = new KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference(this, "s3", true);
        // schema - computed: false, optional: false, required: true
        this._schema = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference(this, "schema", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _0, _1, _2, _3;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tableName) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        if ((_0 = this._s3) === null || _0 === void 0 ? void 0 : _0.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3 = (_1 = this._s3) === null || _1 === void 0 ? void 0 : _1.internalValue;
        }
        if ((_2 = this._schema) === null || _2 === void 0 ? void 0 : _2.internalValue) {
            hasAnyValues = true;
            internalValueResult.schema = (_3 = this._schema) === null || _3 === void 0 ? void 0 : _3.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tableName = undefined;
            this._s3.internalValue = undefined;
            this._schema.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tableName = value.tableName;
            this._s3.internalValue = value.s3;
            this._schema.internalValue = value.schema;
        }
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
    /**
     * @stability stable
     */
    get s3() {
        return this._s3;
    }
    /**
     * @stability stable
     */
    putS3(value) {
        this._s3.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3Input() {
        return this._s3.internalValue;
    }
    /**
     * @stability stable
     */
    get schema() {
        return this._schema;
    }
    /**
     * @stability stable
     */
    putSchema(value) {
        this._schema.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaInput() {
        return this._schema.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesOutputReference[_y] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/kinesis_analytics_application.html aws_kinesis_analytics_application}.
 *
 * @stability stable
 */
class KinesisAnalyticsApplication extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kinesis_analytics_application.html aws_kinesis_analytics_application} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kinesis_analytics_application',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // cloudwatch_logging_options - computed: false, optional: true, required: false
        this._cloudwatchLoggingOptions = new KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference(this, "cloudwatch_logging_options", true);
        // inputs - computed: false, optional: true, required: false
        this._inputs = new KinesisAnalyticsApplicationInputsOutputReference(this, "inputs", true);
        // reference_data_sources - computed: false, optional: true, required: false
        this._referenceDataSources = new KinesisAnalyticsApplicationReferenceDataSourcesOutputReference(this, "reference_data_sources", true);
        this._code = config.code;
        this._description = config.description;
        this._name = config.name;
        this._startApplication = config.startApplication;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._cloudwatchLoggingOptions.internalValue = config.cloudwatchLoggingOptions;
        this._inputs.internalValue = config.inputs;
        this._outputs = config.outputs;
        this._referenceDataSources.internalValue = config.referenceDataSources;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get code() {
        return this.getStringAttribute('code');
    }
    /**
     * @stability stable
     */
    set code(value) {
        this._code = value;
    }
    /**
     * @stability stable
     */
    resetCode() {
        this._code = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get codeInput() {
        return this._code;
    }
    // create_timestamp - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createTimestamp() {
        return this.getStringAttribute('create_timestamp');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_update_timestamp - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastUpdateTimestamp() {
        return this.getStringAttribute('last_update_timestamp');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get startApplication() {
        return this.getBooleanAttribute('start_application');
    }
    /**
     * @stability stable
     */
    set startApplication(value) {
        this._startApplication = value;
    }
    /**
     * @stability stable
     */
    resetStartApplication() {
        this._startApplication = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startApplicationInput() {
        return this._startApplication;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getNumberAttribute('version');
    }
    /**
     * @stability stable
     */
    get cloudwatchLoggingOptions() {
        return this._cloudwatchLoggingOptions;
    }
    /**
     * @stability stable
     */
    putCloudwatchLoggingOptions(value) {
        this._cloudwatchLoggingOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchLoggingOptions() {
        this._cloudwatchLoggingOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchLoggingOptionsInput() {
        return this._cloudwatchLoggingOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get inputs() {
        return this._inputs;
    }
    /**
     * @stability stable
     */
    putInputs(value) {
        this._inputs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetInputs() {
        this._inputs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputsInput() {
        return this._inputs.internalValue;
    }
    /**
     * @stability stable
     */
    get outputs() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('outputs');
    }
    /**
     * @stability stable
     */
    set outputs(value) {
        this._outputs = value;
    }
    /**
     * @stability stable
     */
    resetOutputs() {
        this._outputs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputsInput() {
        return this._outputs;
    }
    /**
     * @stability stable
     */
    get referenceDataSources() {
        return this._referenceDataSources;
    }
    /**
     * @stability stable
     */
    putReferenceDataSources(value) {
        this._referenceDataSources.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetReferenceDataSources() {
        this._referenceDataSources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get referenceDataSourcesInput() {
        return this._referenceDataSources.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            code: cdktf.stringToTerraform(this._code),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            start_application: cdktf.booleanToTerraform(this._startApplication),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            cloudwatch_logging_options: kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform(this._cloudwatchLoggingOptions.internalValue),
            inputs: kinesisAnalyticsApplicationInputsToTerraform(this._inputs.internalValue),
            outputs: cdktf.listMapper(kinesisAnalyticsApplicationOutputsToTerraform)(this._outputs),
            reference_data_sources: kinesisAnalyticsApplicationReferenceDataSourcesToTerraform(this._referenceDataSources.internalValue),
        };
    }
}
exports.KinesisAnalyticsApplication = KinesisAnalyticsApplication;
_z = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplication[_z] = { fqn: "@cdktf/provider-aws.kinesis.KinesisAnalyticsApplication", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
KinesisAnalyticsApplication.tfResourceType = "aws_kinesis_analytics_application";
