// @cdktf/provider-aws/lib/mq/mq-broker.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MqBroker = exports.mqBrokerUserToTerraform = exports.MqBrokerMaintenanceWindowStartTimeOutputReference = exports.mqBrokerMaintenanceWindowStartTimeToTerraform = exports.MqBrokerLogsOutputReference = exports.mqBrokerLogsToTerraform = exports.MqBrokerLdapServerMetadataOutputReference = exports.mqBrokerLdapServerMetadataToTerraform = exports.MqBrokerEncryptionOptionsOutputReference = exports.mqBrokerEncryptionOptionsToTerraform = exports.MqBrokerConfigurationOutputReference = exports.mqBrokerConfigurationToTerraform = exports.MqBrokerInstances = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class MqBrokerInstances extends cdktf.ComplexComputedList {
    // console_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get consoleUrl() {
        return this.getStringAttribute('console_url');
    }
    // endpoints - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoints() {
        return this.getListAttribute('endpoints');
    }
    // ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
}
exports.MqBrokerInstances = MqBrokerInstances;
_a = JSII_RTTI_SYMBOL_1;
MqBrokerInstances[_a] = { fqn: "@cdktf/provider-aws.mq.MqBrokerInstances", version: "3.0.1" };
function mqBrokerConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mqBrokerConfigurationToTerraform = mqBrokerConfigurationToTerraform;
/**
 * @stability stable
 */
class MqBrokerConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._revision) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._revision = value.revision;
        }
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get revision() {
        return this.getNumberAttribute('revision');
    }
    /**
     * @stability stable
     */
    set revision(value) {
        this._revision = value;
    }
    /**
     * @stability stable
     */
    resetRevision() {
        this._revision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get revisionInput() {
        return this._revision;
    }
}
exports.MqBrokerConfigurationOutputReference = MqBrokerConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MqBrokerConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.mq.MqBrokerConfigurationOutputReference", version: "3.0.1" };
function mqBrokerEncryptionOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        use_aws_owned_key: cdktf.booleanToTerraform(struct.useAwsOwnedKey),
    };
}
exports.mqBrokerEncryptionOptionsToTerraform = mqBrokerEncryptionOptionsToTerraform;
/**
 * @stability stable
 */
class MqBrokerEncryptionOptionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._useAwsOwnedKey) {
            hasAnyValues = true;
            internalValueResult.useAwsOwnedKey = this._useAwsOwnedKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._useAwsOwnedKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._useAwsOwnedKey = value.useAwsOwnedKey;
        }
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    /**
     * @stability stable
     */
    get useAwsOwnedKey() {
        return this.getBooleanAttribute('use_aws_owned_key');
    }
    /**
     * @stability stable
     */
    set useAwsOwnedKey(value) {
        this._useAwsOwnedKey = value;
    }
    /**
     * @stability stable
     */
    resetUseAwsOwnedKey() {
        this._useAwsOwnedKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get useAwsOwnedKeyInput() {
        return this._useAwsOwnedKey;
    }
}
exports.MqBrokerEncryptionOptionsOutputReference = MqBrokerEncryptionOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MqBrokerEncryptionOptionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.mq.MqBrokerEncryptionOptionsOutputReference", version: "3.0.1" };
function mqBrokerLdapServerMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hosts: cdktf.listMapper(cdktf.stringToTerraform)(struct.hosts),
        role_base: cdktf.stringToTerraform(struct.roleBase),
        role_name: cdktf.stringToTerraform(struct.roleName),
        role_search_matching: cdktf.stringToTerraform(struct.roleSearchMatching),
        role_search_subtree: cdktf.booleanToTerraform(struct.roleSearchSubtree),
        service_account_password: cdktf.stringToTerraform(struct.serviceAccountPassword),
        service_account_username: cdktf.stringToTerraform(struct.serviceAccountUsername),
        user_base: cdktf.stringToTerraform(struct.userBase),
        user_role_name: cdktf.stringToTerraform(struct.userRoleName),
        user_search_matching: cdktf.stringToTerraform(struct.userSearchMatching),
        user_search_subtree: cdktf.booleanToTerraform(struct.userSearchSubtree),
    };
}
exports.mqBrokerLdapServerMetadataToTerraform = mqBrokerLdapServerMetadataToTerraform;
/**
 * @stability stable
 */
class MqBrokerLdapServerMetadataOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hosts) {
            hasAnyValues = true;
            internalValueResult.hosts = this._hosts;
        }
        if (this._roleBase) {
            hasAnyValues = true;
            internalValueResult.roleBase = this._roleBase;
        }
        if (this._roleName) {
            hasAnyValues = true;
            internalValueResult.roleName = this._roleName;
        }
        if (this._roleSearchMatching) {
            hasAnyValues = true;
            internalValueResult.roleSearchMatching = this._roleSearchMatching;
        }
        if (this._roleSearchSubtree) {
            hasAnyValues = true;
            internalValueResult.roleSearchSubtree = this._roleSearchSubtree;
        }
        if (this._serviceAccountPassword) {
            hasAnyValues = true;
            internalValueResult.serviceAccountPassword = this._serviceAccountPassword;
        }
        if (this._serviceAccountUsername) {
            hasAnyValues = true;
            internalValueResult.serviceAccountUsername = this._serviceAccountUsername;
        }
        if (this._userBase) {
            hasAnyValues = true;
            internalValueResult.userBase = this._userBase;
        }
        if (this._userRoleName) {
            hasAnyValues = true;
            internalValueResult.userRoleName = this._userRoleName;
        }
        if (this._userSearchMatching) {
            hasAnyValues = true;
            internalValueResult.userSearchMatching = this._userSearchMatching;
        }
        if (this._userSearchSubtree) {
            hasAnyValues = true;
            internalValueResult.userSearchSubtree = this._userSearchSubtree;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hosts = undefined;
            this._roleBase = undefined;
            this._roleName = undefined;
            this._roleSearchMatching = undefined;
            this._roleSearchSubtree = undefined;
            this._serviceAccountPassword = undefined;
            this._serviceAccountUsername = undefined;
            this._userBase = undefined;
            this._userRoleName = undefined;
            this._userSearchMatching = undefined;
            this._userSearchSubtree = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hosts = value.hosts;
            this._roleBase = value.roleBase;
            this._roleName = value.roleName;
            this._roleSearchMatching = value.roleSearchMatching;
            this._roleSearchSubtree = value.roleSearchSubtree;
            this._serviceAccountPassword = value.serviceAccountPassword;
            this._serviceAccountUsername = value.serviceAccountUsername;
            this._userBase = value.userBase;
            this._userRoleName = value.userRoleName;
            this._userSearchMatching = value.userSearchMatching;
            this._userSearchSubtree = value.userSearchSubtree;
        }
    }
    /**
     * @stability stable
     */
    get hosts() {
        return this.getListAttribute('hosts');
    }
    /**
     * @stability stable
     */
    set hosts(value) {
        this._hosts = value;
    }
    /**
     * @stability stable
     */
    resetHosts() {
        this._hosts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostsInput() {
        return this._hosts;
    }
    /**
     * @stability stable
     */
    get roleBase() {
        return this.getStringAttribute('role_base');
    }
    /**
     * @stability stable
     */
    set roleBase(value) {
        this._roleBase = value;
    }
    /**
     * @stability stable
     */
    resetRoleBase() {
        this._roleBase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleBaseInput() {
        return this._roleBase;
    }
    /**
     * @stability stable
     */
    get roleName() {
        return this.getStringAttribute('role_name');
    }
    /**
     * @stability stable
     */
    set roleName(value) {
        this._roleName = value;
    }
    /**
     * @stability stable
     */
    resetRoleName() {
        this._roleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleNameInput() {
        return this._roleName;
    }
    /**
     * @stability stable
     */
    get roleSearchMatching() {
        return this.getStringAttribute('role_search_matching');
    }
    /**
     * @stability stable
     */
    set roleSearchMatching(value) {
        this._roleSearchMatching = value;
    }
    /**
     * @stability stable
     */
    resetRoleSearchMatching() {
        this._roleSearchMatching = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleSearchMatchingInput() {
        return this._roleSearchMatching;
    }
    /**
     * @stability stable
     */
    get roleSearchSubtree() {
        return this.getBooleanAttribute('role_search_subtree');
    }
    /**
     * @stability stable
     */
    set roleSearchSubtree(value) {
        this._roleSearchSubtree = value;
    }
    /**
     * @stability stable
     */
    resetRoleSearchSubtree() {
        this._roleSearchSubtree = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleSearchSubtreeInput() {
        return this._roleSearchSubtree;
    }
    /**
     * @stability stable
     */
    get serviceAccountPassword() {
        return this.getStringAttribute('service_account_password');
    }
    /**
     * @stability stable
     */
    set serviceAccountPassword(value) {
        this._serviceAccountPassword = value;
    }
    /**
     * @stability stable
     */
    resetServiceAccountPassword() {
        this._serviceAccountPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceAccountPasswordInput() {
        return this._serviceAccountPassword;
    }
    /**
     * @stability stable
     */
    get serviceAccountUsername() {
        return this.getStringAttribute('service_account_username');
    }
    /**
     * @stability stable
     */
    set serviceAccountUsername(value) {
        this._serviceAccountUsername = value;
    }
    /**
     * @stability stable
     */
    resetServiceAccountUsername() {
        this._serviceAccountUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceAccountUsernameInput() {
        return this._serviceAccountUsername;
    }
    /**
     * @stability stable
     */
    get userBase() {
        return this.getStringAttribute('user_base');
    }
    /**
     * @stability stable
     */
    set userBase(value) {
        this._userBase = value;
    }
    /**
     * @stability stable
     */
    resetUserBase() {
        this._userBase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userBaseInput() {
        return this._userBase;
    }
    /**
     * @stability stable
     */
    get userRoleName() {
        return this.getStringAttribute('user_role_name');
    }
    /**
     * @stability stable
     */
    set userRoleName(value) {
        this._userRoleName = value;
    }
    /**
     * @stability stable
     */
    resetUserRoleName() {
        this._userRoleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userRoleNameInput() {
        return this._userRoleName;
    }
    /**
     * @stability stable
     */
    get userSearchMatching() {
        return this.getStringAttribute('user_search_matching');
    }
    /**
     * @stability stable
     */
    set userSearchMatching(value) {
        this._userSearchMatching = value;
    }
    /**
     * @stability stable
     */
    resetUserSearchMatching() {
        this._userSearchMatching = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userSearchMatchingInput() {
        return this._userSearchMatching;
    }
    /**
     * @stability stable
     */
    get userSearchSubtree() {
        return this.getBooleanAttribute('user_search_subtree');
    }
    /**
     * @stability stable
     */
    set userSearchSubtree(value) {
        this._userSearchSubtree = value;
    }
    /**
     * @stability stable
     */
    resetUserSearchSubtree() {
        this._userSearchSubtree = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userSearchSubtreeInput() {
        return this._userSearchSubtree;
    }
}
exports.MqBrokerLdapServerMetadataOutputReference = MqBrokerLdapServerMetadataOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MqBrokerLdapServerMetadataOutputReference[_d] = { fqn: "@cdktf/provider-aws.mq.MqBrokerLdapServerMetadataOutputReference", version: "3.0.1" };
function mqBrokerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit: cdktf.stringToTerraform(struct.audit),
        general: cdktf.booleanToTerraform(struct.general),
    };
}
exports.mqBrokerLogsToTerraform = mqBrokerLogsToTerraform;
/**
 * @stability stable
 */
class MqBrokerLogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audit) {
            hasAnyValues = true;
            internalValueResult.audit = this._audit;
        }
        if (this._general) {
            hasAnyValues = true;
            internalValueResult.general = this._general;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audit = undefined;
            this._general = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audit = value.audit;
            this._general = value.general;
        }
    }
    /**
     * @stability stable
     */
    get audit() {
        return this.getStringAttribute('audit');
    }
    /**
     * @stability stable
     */
    set audit(value) {
        this._audit = value;
    }
    /**
     * @stability stable
     */
    resetAudit() {
        this._audit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get auditInput() {
        return this._audit;
    }
    /**
     * @stability stable
     */
    get general() {
        return this.getBooleanAttribute('general');
    }
    /**
     * @stability stable
     */
    set general(value) {
        this._general = value;
    }
    /**
     * @stability stable
     */
    resetGeneral() {
        this._general = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get generalInput() {
        return this._general;
    }
}
exports.MqBrokerLogsOutputReference = MqBrokerLogsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MqBrokerLogsOutputReference[_e] = { fqn: "@cdktf/provider-aws.mq.MqBrokerLogsOutputReference", version: "3.0.1" };
function mqBrokerMaintenanceWindowStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day_of_week: cdktf.stringToTerraform(struct.dayOfWeek),
        time_of_day: cdktf.stringToTerraform(struct.timeOfDay),
        time_zone: cdktf.stringToTerraform(struct.timeZone),
    };
}
exports.mqBrokerMaintenanceWindowStartTimeToTerraform = mqBrokerMaintenanceWindowStartTimeToTerraform;
/**
 * @stability stable
 */
class MqBrokerMaintenanceWindowStartTimeOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dayOfWeek) {
            hasAnyValues = true;
            internalValueResult.dayOfWeek = this._dayOfWeek;
        }
        if (this._timeOfDay) {
            hasAnyValues = true;
            internalValueResult.timeOfDay = this._timeOfDay;
        }
        if (this._timeZone) {
            hasAnyValues = true;
            internalValueResult.timeZone = this._timeZone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dayOfWeek = undefined;
            this._timeOfDay = undefined;
            this._timeZone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dayOfWeek = value.dayOfWeek;
            this._timeOfDay = value.timeOfDay;
            this._timeZone = value.timeZone;
        }
    }
    /**
     * @stability stable
     */
    get dayOfWeek() {
        return this.getStringAttribute('day_of_week');
    }
    /**
     * @stability stable
     */
    set dayOfWeek(value) {
        this._dayOfWeek = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dayOfWeekInput() {
        return this._dayOfWeek;
    }
    /**
     * @stability stable
     */
    get timeOfDay() {
        return this.getStringAttribute('time_of_day');
    }
    /**
     * @stability stable
     */
    set timeOfDay(value) {
        this._timeOfDay = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeOfDayInput() {
        return this._timeOfDay;
    }
    /**
     * @stability stable
     */
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    /**
     * @stability stable
     */
    set timeZone(value) {
        this._timeZone = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeZoneInput() {
        return this._timeZone;
    }
}
exports.MqBrokerMaintenanceWindowStartTimeOutputReference = MqBrokerMaintenanceWindowStartTimeOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MqBrokerMaintenanceWindowStartTimeOutputReference[_f] = { fqn: "@cdktf/provider-aws.mq.MqBrokerMaintenanceWindowStartTimeOutputReference", version: "3.0.1" };
function mqBrokerUserToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        console_access: cdktf.booleanToTerraform(struct.consoleAccess),
        groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.groups),
        password: cdktf.stringToTerraform(struct.password),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.mqBrokerUserToTerraform = mqBrokerUserToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/mq_broker.html aws_mq_broker}.
 *
 * @stability stable
 */
class MqBroker extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/mq_broker.html aws_mq_broker} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_mq_broker',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // configuration - computed: false, optional: true, required: false
        this._configuration = new MqBrokerConfigurationOutputReference(this, "configuration", true);
        // encryption_options - computed: false, optional: true, required: false
        this._encryptionOptions = new MqBrokerEncryptionOptionsOutputReference(this, "encryption_options", true);
        // ldap_server_metadata - computed: false, optional: true, required: false
        this._ldapServerMetadata = new MqBrokerLdapServerMetadataOutputReference(this, "ldap_server_metadata", true);
        // logs - computed: false, optional: true, required: false
        this._logs = new MqBrokerLogsOutputReference(this, "logs", true);
        // maintenance_window_start_time - computed: false, optional: true, required: false
        this._maintenanceWindowStartTime = new MqBrokerMaintenanceWindowStartTimeOutputReference(this, "maintenance_window_start_time", true);
        this._applyImmediately = config.applyImmediately;
        this._authenticationStrategy = config.authenticationStrategy;
        this._autoMinorVersionUpgrade = config.autoMinorVersionUpgrade;
        this._brokerName = config.brokerName;
        this._deploymentMode = config.deploymentMode;
        this._engineType = config.engineType;
        this._engineVersion = config.engineVersion;
        this._hostInstanceType = config.hostInstanceType;
        this._publiclyAccessible = config.publiclyAccessible;
        this._securityGroups = config.securityGroups;
        this._storageType = config.storageType;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._configuration.internalValue = config.configuration;
        this._encryptionOptions.internalValue = config.encryptionOptions;
        this._ldapServerMetadata.internalValue = config.ldapServerMetadata;
        this._logs.internalValue = config.logs;
        this._maintenanceWindowStartTime.internalValue = config.maintenanceWindowStartTime;
        this._user = config.user;
    }
    /**
     * @stability stable
     */
    get applyImmediately() {
        return this.getBooleanAttribute('apply_immediately');
    }
    /**
     * @stability stable
     */
    set applyImmediately(value) {
        this._applyImmediately = value;
    }
    /**
     * @stability stable
     */
    resetApplyImmediately() {
        this._applyImmediately = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applyImmediatelyInput() {
        return this._applyImmediately;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get authenticationStrategy() {
        return this.getStringAttribute('authentication_strategy');
    }
    /**
     * @stability stable
     */
    set authenticationStrategy(value) {
        this._authenticationStrategy = value;
    }
    /**
     * @stability stable
     */
    resetAuthenticationStrategy() {
        this._authenticationStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authenticationStrategyInput() {
        return this._authenticationStrategy;
    }
    /**
     * @stability stable
     */
    get autoMinorVersionUpgrade() {
        return this.getBooleanAttribute('auto_minor_version_upgrade');
    }
    /**
     * @stability stable
     */
    set autoMinorVersionUpgrade(value) {
        this._autoMinorVersionUpgrade = value;
    }
    /**
     * @stability stable
     */
    resetAutoMinorVersionUpgrade() {
        this._autoMinorVersionUpgrade = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoMinorVersionUpgradeInput() {
        return this._autoMinorVersionUpgrade;
    }
    /**
     * @stability stable
     */
    get brokerName() {
        return this.getStringAttribute('broker_name');
    }
    /**
     * @stability stable
     */
    set brokerName(value) {
        this._brokerName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get brokerNameInput() {
        return this._brokerName;
    }
    /**
     * @stability stable
     */
    get deploymentMode() {
        return this.getStringAttribute('deployment_mode');
    }
    /**
     * @stability stable
     */
    set deploymentMode(value) {
        this._deploymentMode = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentMode() {
        this._deploymentMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentModeInput() {
        return this._deploymentMode;
    }
    /**
     * @stability stable
     */
    get engineType() {
        return this.getStringAttribute('engine_type');
    }
    /**
     * @stability stable
     */
    set engineType(value) {
        this._engineType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get engineTypeInput() {
        return this._engineType;
    }
    /**
     * @stability stable
     */
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    /**
     * @stability stable
     */
    set engineVersion(value) {
        this._engineVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get engineVersionInput() {
        return this._engineVersion;
    }
    /**
     * @stability stable
     */
    get hostInstanceType() {
        return this.getStringAttribute('host_instance_type');
    }
    /**
     * @stability stable
     */
    set hostInstanceType(value) {
        this._hostInstanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostInstanceTypeInput() {
        return this._hostInstanceType;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // instances - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    instances(index) {
        return new MqBrokerInstances(this, 'instances', index);
    }
    /**
     * @stability stable
     */
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    /**
     * @stability stable
     */
    set publiclyAccessible(value) {
        this._publiclyAccessible = value;
    }
    /**
     * @stability stable
     */
    resetPubliclyAccessible() {
        this._publiclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publiclyAccessibleInput() {
        return this._publiclyAccessible;
    }
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    /**
     * @stability stable
     */
    set securityGroups(value) {
        this._securityGroups = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupsInput() {
        return this._securityGroups;
    }
    /**
     * @stability stable
     */
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    /**
     * @stability stable
     */
    set storageType(value) {
        this._storageType = value;
    }
    /**
     * @stability stable
     */
    resetStorageType() {
        this._storageType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageTypeInput() {
        return this._storageType;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    /**
     * @stability stable
     */
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get configuration() {
        return this._configuration;
    }
    /**
     * @stability stable
     */
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get configurationInput() {
        return this._configuration.internalValue;
    }
    /**
     * @stability stable
     */
    get encryptionOptions() {
        return this._encryptionOptions;
    }
    /**
     * @stability stable
     */
    putEncryptionOptions(value) {
        this._encryptionOptions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionOptions() {
        this._encryptionOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionOptionsInput() {
        return this._encryptionOptions.internalValue;
    }
    /**
     * @stability stable
     */
    get ldapServerMetadata() {
        return this._ldapServerMetadata;
    }
    /**
     * @stability stable
     */
    putLdapServerMetadata(value) {
        this._ldapServerMetadata.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLdapServerMetadata() {
        this._ldapServerMetadata.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ldapServerMetadataInput() {
        return this._ldapServerMetadata.internalValue;
    }
    /**
     * @stability stable
     */
    get logs() {
        return this._logs;
    }
    /**
     * @stability stable
     */
    putLogs(value) {
        this._logs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLogs() {
        this._logs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logsInput() {
        return this._logs.internalValue;
    }
    /**
     * @stability stable
     */
    get maintenanceWindowStartTime() {
        return this._maintenanceWindowStartTime;
    }
    /**
     * @stability stable
     */
    putMaintenanceWindowStartTime(value) {
        this._maintenanceWindowStartTime.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetMaintenanceWindowStartTime() {
        this._maintenanceWindowStartTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maintenanceWindowStartTimeInput() {
        return this._maintenanceWindowStartTime.internalValue;
    }
    /**
     * @stability stable
     */
    get user() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('user');
    }
    /**
     * @stability stable
     */
    set user(value) {
        this._user = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get userInput() {
        return this._user;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            apply_immediately: cdktf.booleanToTerraform(this._applyImmediately),
            authentication_strategy: cdktf.stringToTerraform(this._authenticationStrategy),
            auto_minor_version_upgrade: cdktf.booleanToTerraform(this._autoMinorVersionUpgrade),
            broker_name: cdktf.stringToTerraform(this._brokerName),
            deployment_mode: cdktf.stringToTerraform(this._deploymentMode),
            engine_type: cdktf.stringToTerraform(this._engineType),
            engine_version: cdktf.stringToTerraform(this._engineVersion),
            host_instance_type: cdktf.stringToTerraform(this._hostInstanceType),
            publicly_accessible: cdktf.booleanToTerraform(this._publiclyAccessible),
            security_groups: cdktf.listMapper(cdktf.stringToTerraform)(this._securityGroups),
            storage_type: cdktf.stringToTerraform(this._storageType),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            configuration: mqBrokerConfigurationToTerraform(this._configuration.internalValue),
            encryption_options: mqBrokerEncryptionOptionsToTerraform(this._encryptionOptions.internalValue),
            ldap_server_metadata: mqBrokerLdapServerMetadataToTerraform(this._ldapServerMetadata.internalValue),
            logs: mqBrokerLogsToTerraform(this._logs.internalValue),
            maintenance_window_start_time: mqBrokerMaintenanceWindowStartTimeToTerraform(this._maintenanceWindowStartTime.internalValue),
            user: cdktf.listMapper(mqBrokerUserToTerraform)(this._user),
        };
    }
}
exports.MqBroker = MqBroker;
_g = JSII_RTTI_SYMBOL_1;
MqBroker[_g] = { fqn: "@cdktf/provider-aws.mq.MqBroker", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MqBroker.tfResourceType = "aws_mq_broker";
