// @cdktf/provider-aws/lib/lakeformation/lakeformation-permissions.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LakeformationPermissions = exports.LakeformationPermissionsTableWithColumnsOutputReference = exports.lakeformationPermissionsTableWithColumnsToTerraform = exports.LakeformationPermissionsTableOutputReference = exports.lakeformationPermissionsTableToTerraform = exports.LakeformationPermissionsDatabaseOutputReference = exports.lakeformationPermissionsDatabaseToTerraform = exports.LakeformationPermissionsDataLocationOutputReference = exports.lakeformationPermissionsDataLocationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lakeformationPermissionsDataLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
    };
}
exports.lakeformationPermissionsDataLocationToTerraform = lakeformationPermissionsDataLocationToTerraform;
/**
 * @stability stable
 */
class LakeformationPermissionsDataLocationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._catalogId) {
            hasAnyValues = true;
            internalValueResult.catalogId = this._catalogId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._catalogId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._catalogId = value.catalogId;
        }
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
}
exports.LakeformationPermissionsDataLocationOutputReference = LakeformationPermissionsDataLocationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LakeformationPermissionsDataLocationOutputReference[_a] = { fqn: "@cdktf/provider-aws.lakeformation.LakeformationPermissionsDataLocationOutputReference", version: "3.0.1" };
function lakeformationPermissionsDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.lakeformationPermissionsDatabaseToTerraform = lakeformationPermissionsDatabaseToTerraform;
/**
 * @stability stable
 */
class LakeformationPermissionsDatabaseOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._catalogId) {
            hasAnyValues = true;
            internalValueResult.catalogId = this._catalogId;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._catalogId = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._catalogId = value.catalogId;
            this._name = value.name;
        }
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
}
exports.LakeformationPermissionsDatabaseOutputReference = LakeformationPermissionsDatabaseOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LakeformationPermissionsDatabaseOutputReference[_b] = { fqn: "@cdktf/provider-aws.lakeformation.LakeformationPermissionsDatabaseOutputReference", version: "3.0.1" };
function lakeformationPermissionsTableToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        name: cdktf.stringToTerraform(struct.name),
        wildcard: cdktf.booleanToTerraform(struct.wildcard),
    };
}
exports.lakeformationPermissionsTableToTerraform = lakeformationPermissionsTableToTerraform;
/**
 * @stability stable
 */
class LakeformationPermissionsTableOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._catalogId) {
            hasAnyValues = true;
            internalValueResult.catalogId = this._catalogId;
        }
        if (this._databaseName) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._wildcard) {
            hasAnyValues = true;
            internalValueResult.wildcard = this._wildcard;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._catalogId = undefined;
            this._databaseName = undefined;
            this._name = undefined;
            this._wildcard = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._catalogId = value.catalogId;
            this._databaseName = value.databaseName;
            this._name = value.name;
            this._wildcard = value.wildcard;
        }
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    /**
     * @stability stable
     */
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseNameInput() {
        return this._databaseName;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get wildcard() {
        return this.getBooleanAttribute('wildcard');
    }
    /**
     * @stability stable
     */
    set wildcard(value) {
        this._wildcard = value;
    }
    /**
     * @stability stable
     */
    resetWildcard() {
        this._wildcard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get wildcardInput() {
        return this._wildcard;
    }
}
exports.LakeformationPermissionsTableOutputReference = LakeformationPermissionsTableOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LakeformationPermissionsTableOutputReference[_c] = { fqn: "@cdktf/provider-aws.lakeformation.LakeformationPermissionsTableOutputReference", version: "3.0.1" };
function lakeformationPermissionsTableWithColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog_id: cdktf.stringToTerraform(struct.catalogId),
        column_names: cdktf.listMapper(cdktf.stringToTerraform)(struct.columnNames),
        database_name: cdktf.stringToTerraform(struct.databaseName),
        excluded_column_names: cdktf.listMapper(cdktf.stringToTerraform)(struct.excludedColumnNames),
        name: cdktf.stringToTerraform(struct.name),
        wildcard: cdktf.booleanToTerraform(struct.wildcard),
    };
}
exports.lakeformationPermissionsTableWithColumnsToTerraform = lakeformationPermissionsTableWithColumnsToTerraform;
/**
 * @stability stable
 */
class LakeformationPermissionsTableWithColumnsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._catalogId) {
            hasAnyValues = true;
            internalValueResult.catalogId = this._catalogId;
        }
        if (this._columnNames) {
            hasAnyValues = true;
            internalValueResult.columnNames = this._columnNames;
        }
        if (this._databaseName) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._excludedColumnNames) {
            hasAnyValues = true;
            internalValueResult.excludedColumnNames = this._excludedColumnNames;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._wildcard) {
            hasAnyValues = true;
            internalValueResult.wildcard = this._wildcard;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._catalogId = undefined;
            this._columnNames = undefined;
            this._databaseName = undefined;
            this._excludedColumnNames = undefined;
            this._name = undefined;
            this._wildcard = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._catalogId = value.catalogId;
            this._columnNames = value.columnNames;
            this._databaseName = value.databaseName;
            this._excludedColumnNames = value.excludedColumnNames;
            this._name = value.name;
            this._wildcard = value.wildcard;
        }
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
    /**
     * @stability stable
     */
    get columnNames() {
        return this.getListAttribute('column_names');
    }
    /**
     * @stability stable
     */
    set columnNames(value) {
        this._columnNames = value;
    }
    /**
     * @stability stable
     */
    resetColumnNames() {
        this._columnNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get columnNamesInput() {
        return this._columnNames;
    }
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    /**
     * @stability stable
     */
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseNameInput() {
        return this._databaseName;
    }
    /**
     * @stability stable
     */
    get excludedColumnNames() {
        return this.getListAttribute('excluded_column_names');
    }
    /**
     * @stability stable
     */
    set excludedColumnNames(value) {
        this._excludedColumnNames = value;
    }
    /**
     * @stability stable
     */
    resetExcludedColumnNames() {
        this._excludedColumnNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludedColumnNamesInput() {
        return this._excludedColumnNames;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get wildcard() {
        return this.getBooleanAttribute('wildcard');
    }
    /**
     * @stability stable
     */
    set wildcard(value) {
        this._wildcard = value;
    }
    /**
     * @stability stable
     */
    resetWildcard() {
        this._wildcard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get wildcardInput() {
        return this._wildcard;
    }
}
exports.LakeformationPermissionsTableWithColumnsOutputReference = LakeformationPermissionsTableWithColumnsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
LakeformationPermissionsTableWithColumnsOutputReference[_d] = { fqn: "@cdktf/provider-aws.lakeformation.LakeformationPermissionsTableWithColumnsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lakeformation_permissions.html aws_lakeformation_permissions}.
 *
 * @stability stable
 */
class LakeformationPermissions extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lakeformation_permissions.html aws_lakeformation_permissions} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lakeformation_permissions',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // data_location - computed: false, optional: true, required: false
        this._dataLocation = new LakeformationPermissionsDataLocationOutputReference(this, "data_location", true);
        // database - computed: false, optional: true, required: false
        this._database = new LakeformationPermissionsDatabaseOutputReference(this, "database", true);
        // table - computed: false, optional: true, required: false
        this._table = new LakeformationPermissionsTableOutputReference(this, "table", true);
        // table_with_columns - computed: false, optional: true, required: false
        this._tableWithColumns = new LakeformationPermissionsTableWithColumnsOutputReference(this, "table_with_columns", true);
        this._catalogId = config.catalogId;
        this._catalogResource = config.catalogResource;
        this._permissions = config.permissions;
        this._permissionsWithGrantOption = config.permissionsWithGrantOption;
        this._principal = config.principal;
        this._dataLocation.internalValue = config.dataLocation;
        this._database.internalValue = config.database;
        this._table.internalValue = config.table;
        this._tableWithColumns.internalValue = config.tableWithColumns;
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
    /**
     * @stability stable
     */
    get catalogResource() {
        return this.getBooleanAttribute('catalog_resource');
    }
    /**
     * @stability stable
     */
    set catalogResource(value) {
        this._catalogResource = value;
    }
    /**
     * @stability stable
     */
    resetCatalogResource() {
        this._catalogResource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogResourceInput() {
        return this._catalogResource;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get permissions() {
        return this.getListAttribute('permissions');
    }
    /**
     * @stability stable
     */
    set permissions(value) {
        this._permissions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionsInput() {
        return this._permissions;
    }
    /**
     * @stability stable
     */
    get permissionsWithGrantOption() {
        return this.getListAttribute('permissions_with_grant_option');
    }
    /**
     * @stability stable
     */
    set permissionsWithGrantOption(value) {
        this._permissionsWithGrantOption = value;
    }
    /**
     * @stability stable
     */
    resetPermissionsWithGrantOption() {
        this._permissionsWithGrantOption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionsWithGrantOptionInput() {
        return this._permissionsWithGrantOption;
    }
    /**
     * @stability stable
     */
    get principal() {
        return this.getStringAttribute('principal');
    }
    /**
     * @stability stable
     */
    set principal(value) {
        this._principal = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get principalInput() {
        return this._principal;
    }
    /**
     * @stability stable
     */
    get dataLocation() {
        return this._dataLocation;
    }
    /**
     * @stability stable
     */
    putDataLocation(value) {
        this._dataLocation.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDataLocation() {
        this._dataLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dataLocationInput() {
        return this._dataLocation.internalValue;
    }
    /**
     * @stability stable
     */
    get database() {
        return this._database;
    }
    /**
     * @stability stable
     */
    putDatabase(value) {
        this._database.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDatabase() {
        this._database.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseInput() {
        return this._database.internalValue;
    }
    /**
     * @stability stable
     */
    get table() {
        return this._table;
    }
    /**
     * @stability stable
     */
    putTable(value) {
        this._table.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTable() {
        this._table.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableInput() {
        return this._table.internalValue;
    }
    /**
     * @stability stable
     */
    get tableWithColumns() {
        return this._tableWithColumns;
    }
    /**
     * @stability stable
     */
    putTableWithColumns(value) {
        this._tableWithColumns.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTableWithColumns() {
        this._tableWithColumns.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableWithColumnsInput() {
        return this._tableWithColumns.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            catalog_resource: cdktf.booleanToTerraform(this._catalogResource),
            permissions: cdktf.listMapper(cdktf.stringToTerraform)(this._permissions),
            permissions_with_grant_option: cdktf.listMapper(cdktf.stringToTerraform)(this._permissionsWithGrantOption),
            principal: cdktf.stringToTerraform(this._principal),
            data_location: lakeformationPermissionsDataLocationToTerraform(this._dataLocation.internalValue),
            database: lakeformationPermissionsDatabaseToTerraform(this._database.internalValue),
            table: lakeformationPermissionsTableToTerraform(this._table.internalValue),
            table_with_columns: lakeformationPermissionsTableWithColumnsToTerraform(this._tableWithColumns.internalValue),
        };
    }
}
exports.LakeformationPermissions = LakeformationPermissions;
_e = JSII_RTTI_SYMBOL_1;
LakeformationPermissions[_e] = { fqn: "@cdktf/provider-aws.lakeformation.LakeformationPermissions", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LakeformationPermissions.tfResourceType = "aws_lakeformation_permissions";
