// @cdktf/provider-aws/lib/securityhub/securityhub-standards-control.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubStandardsControl = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_control.html aws_securityhub_standards_control}.
 *
 * @stability stable
 */
class SecurityhubStandardsControl extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_control.html aws_securityhub_standards_control} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_standards_control',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._controlStatus = config.controlStatus;
        this._disabledReason = config.disabledReason;
        this._standardsControlArn = config.standardsControlArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // control_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get controlId() {
        return this.getStringAttribute('control_id');
    }
    /**
     * @stability stable
     */
    get controlStatus() {
        return this.getStringAttribute('control_status');
    }
    /**
     * @stability stable
     */
    set controlStatus(value) {
        this._controlStatus = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get controlStatusInput() {
        return this._controlStatus;
    }
    // control_status_updated_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get controlStatusUpdatedAt() {
        return this.getStringAttribute('control_status_updated_at');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    get disabledReason() {
        return this.getStringAttribute('disabled_reason');
    }
    /**
     * @stability stable
     */
    set disabledReason(value) {
        this._disabledReason = value;
    }
    /**
     * @stability stable
     */
    resetDisabledReason() {
        this._disabledReason = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disabledReasonInput() {
        return this._disabledReason;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // related_requirements - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get relatedRequirements() {
        return this.getListAttribute('related_requirements');
    }
    // remediation_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get remediationUrl() {
        return this.getStringAttribute('remediation_url');
    }
    // severity_rating - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get severityRating() {
        return this.getStringAttribute('severity_rating');
    }
    /**
     * @stability stable
     */
    get standardsControlArn() {
        return this.getStringAttribute('standards_control_arn');
    }
    /**
     * @stability stable
     */
    set standardsControlArn(value) {
        this._standardsControlArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get standardsControlArnInput() {
        return this._standardsControlArn;
    }
    // title - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get title() {
        return this.getStringAttribute('title');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            control_status: cdktf.stringToTerraform(this._controlStatus),
            disabled_reason: cdktf.stringToTerraform(this._disabledReason),
            standards_control_arn: cdktf.stringToTerraform(this._standardsControlArn),
        };
    }
}
exports.SecurityhubStandardsControl = SecurityhubStandardsControl;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubStandardsControl[_a] = { fqn: "@cdktf/provider-aws.securityhub.SecurityhubStandardsControl", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecurityhubStandardsControl.tfResourceType = "aws_securityhub_standards_control";
