// @cdktf/provider-aws/lib/kms/data-aws-kms-key.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKmsKey = exports.DataAwsKmsKeyMultiRegionConfiguration = exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeys = exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsKmsKeyMultiRegionConfigurationPrimaryKey extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKey = DataAwsKmsKeyMultiRegionConfigurationPrimaryKey;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationPrimaryKey[_a] = { fqn: "@cdktf/provider-aws.kms.DataAwsKmsKeyMultiRegionConfigurationPrimaryKey", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsKmsKeyMultiRegionConfigurationReplicaKeys extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeys = DataAwsKmsKeyMultiRegionConfigurationReplicaKeys;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationReplicaKeys[_b] = { fqn: "@cdktf/provider-aws.kms.DataAwsKmsKeyMultiRegionConfigurationReplicaKeys", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsKmsKeyMultiRegionConfiguration extends cdktf.ComplexComputedList {
    // multi_region_key_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get multiRegionKeyType() {
        return this.getStringAttribute('multi_region_key_type');
    }
    // primary_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get primaryKey() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('primary_key');
    }
    // replica_keys - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get replicaKeys() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('replica_keys');
    }
}
exports.DataAwsKmsKeyMultiRegionConfiguration = DataAwsKmsKeyMultiRegionConfiguration;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfiguration[_c] = { fqn: "@cdktf/provider-aws.kms.DataAwsKmsKeyMultiRegionConfiguration", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/kms_key.html aws_kms_key}.
 *
 * @stability stable
 */
class DataAwsKmsKey extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/kms_key.html aws_kms_key} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_key',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._grantTokens = config.grantTokens;
        this._keyId = config.keyId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // aws_account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    // creation_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    // customer_master_key_spec - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customerMasterKeySpec() {
        return this.getStringAttribute('customer_master_key_spec');
    }
    // deletion_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deletionDate() {
        return this.getStringAttribute('deletion_date');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // expiration_model - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get expirationModel() {
        return this.getStringAttribute('expiration_model');
    }
    /**
     * @stability stable
     */
    get grantTokens() {
        return this.getListAttribute('grant_tokens');
    }
    /**
     * @stability stable
     */
    set grantTokens(value) {
        this._grantTokens = value;
    }
    /**
     * @stability stable
     */
    resetGrantTokens() {
        this._grantTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get grantTokensInput() {
        return this._grantTokens;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    /**
     * @stability stable
     */
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyIdInput() {
        return this._keyId;
    }
    // key_manager - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyManager() {
        return this.getStringAttribute('key_manager');
    }
    // key_state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyState() {
        return this.getStringAttribute('key_state');
    }
    // key_usage - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyUsage() {
        return this.getStringAttribute('key_usage');
    }
    // multi_region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get multiRegion() {
        return this.getBooleanAttribute('multi_region');
    }
    // multi_region_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    multiRegionConfiguration(index) {
        return new DataAwsKmsKeyMultiRegionConfiguration(this, 'multi_region_configuration', index);
    }
    // origin - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get origin() {
        return this.getStringAttribute('origin');
    }
    // valid_to - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get validTo() {
        return this.getStringAttribute('valid_to');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            grant_tokens: cdktf.listMapper(cdktf.stringToTerraform)(this._grantTokens),
            key_id: cdktf.stringToTerraform(this._keyId),
        };
    }
}
exports.DataAwsKmsKey = DataAwsKmsKey;
_d = JSII_RTTI_SYMBOL_1;
DataAwsKmsKey[_d] = { fqn: "@cdktf/provider-aws.kms.DataAwsKmsKey", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsKmsKey.tfResourceType = "aws_kms_key";
