// @cdktf/provider-aws/lib/worklink/worklink-website-certificate-authority-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorklinkWebsiteCertificateAuthorityAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/worklink_website_certificate_authority_association.html aws_worklink_website_certificate_authority_association}.
 *
 * @stability stable
 */
class WorklinkWebsiteCertificateAuthorityAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/worklink_website_certificate_authority_association.html aws_worklink_website_certificate_authority_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_worklink_website_certificate_authority_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._certificate = config.certificate;
        this._displayName = config.displayName;
        this._fleetArn = config.fleetArn;
    }
    /**
     * @stability stable
     */
    get certificate() {
        return this.getStringAttribute('certificate');
    }
    /**
     * @stability stable
     */
    set certificate(value) {
        this._certificate = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get certificateInput() {
        return this._certificate;
    }
    /**
     * @stability stable
     */
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    /**
     * @stability stable
     */
    set displayName(value) {
        this._displayName = value;
    }
    /**
     * @stability stable
     */
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get displayNameInput() {
        return this._displayName;
    }
    /**
     * @stability stable
     */
    get fleetArn() {
        return this.getStringAttribute('fleet_arn');
    }
    /**
     * @stability stable
     */
    set fleetArn(value) {
        this._fleetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fleetArnInput() {
        return this._fleetArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // website_ca_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get websiteCaId() {
        return this.getStringAttribute('website_ca_id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            certificate: cdktf.stringToTerraform(this._certificate),
            display_name: cdktf.stringToTerraform(this._displayName),
            fleet_arn: cdktf.stringToTerraform(this._fleetArn),
        };
    }
}
exports.WorklinkWebsiteCertificateAuthorityAssociation = WorklinkWebsiteCertificateAuthorityAssociation;
_a = JSII_RTTI_SYMBOL_1;
WorklinkWebsiteCertificateAuthorityAssociation[_a] = { fqn: "@cdktf/provider-aws.worklink.WorklinkWebsiteCertificateAuthorityAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WorklinkWebsiteCertificateAuthorityAssociation.tfResourceType = "aws_worklink_website_certificate_authority_association";
