// @cdktf/provider-aws/lib/config/config-organization-custom-rule.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigOrganizationCustomRule = exports.ConfigOrganizationCustomRuleTimeoutsOutputReference = exports.configOrganizationCustomRuleTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function configOrganizationCustomRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.configOrganizationCustomRuleTimeoutsToTerraform = configOrganizationCustomRuleTimeoutsToTerraform;
/**
 * @stability stable
 */
class ConfigOrganizationCustomRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.ConfigOrganizationCustomRuleTimeoutsOutputReference = ConfigOrganizationCustomRuleTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ConfigOrganizationCustomRuleTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.config.ConfigOrganizationCustomRuleTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/config_organization_custom_rule.html aws_config_organization_custom_rule}.
 *
 * @stability stable
 */
class ConfigOrganizationCustomRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/config_organization_custom_rule.html aws_config_organization_custom_rule} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_config_organization_custom_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ConfigOrganizationCustomRuleTimeoutsOutputReference(this, "timeouts", true);
        this._description = config.description;
        this._excludedAccounts = config.excludedAccounts;
        this._inputParameters = config.inputParameters;
        this._lambdaFunctionArn = config.lambdaFunctionArn;
        this._maximumExecutionFrequency = config.maximumExecutionFrequency;
        this._name = config.name;
        this._resourceIdScope = config.resourceIdScope;
        this._resourceTypesScope = config.resourceTypesScope;
        this._tagKeyScope = config.tagKeyScope;
        this._tagValueScope = config.tagValueScope;
        this._triggerTypes = config.triggerTypes;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get excludedAccounts() {
        return this.getListAttribute('excluded_accounts');
    }
    /**
     * @stability stable
     */
    set excludedAccounts(value) {
        this._excludedAccounts = value;
    }
    /**
     * @stability stable
     */
    resetExcludedAccounts() {
        this._excludedAccounts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludedAccountsInput() {
        return this._excludedAccounts;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get inputParameters() {
        return this.getStringAttribute('input_parameters');
    }
    /**
     * @stability stable
     */
    set inputParameters(value) {
        this._inputParameters = value;
    }
    /**
     * @stability stable
     */
    resetInputParameters() {
        this._inputParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputParametersInput() {
        return this._inputParameters;
    }
    /**
     * @stability stable
     */
    get lambdaFunctionArn() {
        return this.getStringAttribute('lambda_function_arn');
    }
    /**
     * @stability stable
     */
    set lambdaFunctionArn(value) {
        this._lambdaFunctionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaFunctionArnInput() {
        return this._lambdaFunctionArn;
    }
    /**
     * @stability stable
     */
    get maximumExecutionFrequency() {
        return this.getStringAttribute('maximum_execution_frequency');
    }
    /**
     * @stability stable
     */
    set maximumExecutionFrequency(value) {
        this._maximumExecutionFrequency = value;
    }
    /**
     * @stability stable
     */
    resetMaximumExecutionFrequency() {
        this._maximumExecutionFrequency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maximumExecutionFrequencyInput() {
        return this._maximumExecutionFrequency;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get resourceIdScope() {
        return this.getStringAttribute('resource_id_scope');
    }
    /**
     * @stability stable
     */
    set resourceIdScope(value) {
        this._resourceIdScope = value;
    }
    /**
     * @stability stable
     */
    resetResourceIdScope() {
        this._resourceIdScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceIdScopeInput() {
        return this._resourceIdScope;
    }
    /**
     * @stability stable
     */
    get resourceTypesScope() {
        return this.getListAttribute('resource_types_scope');
    }
    /**
     * @stability stable
     */
    set resourceTypesScope(value) {
        this._resourceTypesScope = value;
    }
    /**
     * @stability stable
     */
    resetResourceTypesScope() {
        this._resourceTypesScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTypesScopeInput() {
        return this._resourceTypesScope;
    }
    /**
     * @stability stable
     */
    get tagKeyScope() {
        return this.getStringAttribute('tag_key_scope');
    }
    /**
     * @stability stable
     */
    set tagKeyScope(value) {
        this._tagKeyScope = value;
    }
    /**
     * @stability stable
     */
    resetTagKeyScope() {
        this._tagKeyScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagKeyScopeInput() {
        return this._tagKeyScope;
    }
    /**
     * @stability stable
     */
    get tagValueScope() {
        return this.getStringAttribute('tag_value_scope');
    }
    /**
     * @stability stable
     */
    set tagValueScope(value) {
        this._tagValueScope = value;
    }
    /**
     * @stability stable
     */
    resetTagValueScope() {
        this._tagValueScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagValueScopeInput() {
        return this._tagValueScope;
    }
    /**
     * @stability stable
     */
    get triggerTypes() {
        return this.getListAttribute('trigger_types');
    }
    /**
     * @stability stable
     */
    set triggerTypes(value) {
        this._triggerTypes = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get triggerTypesInput() {
        return this._triggerTypes;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            excluded_accounts: cdktf.listMapper(cdktf.stringToTerraform)(this._excludedAccounts),
            input_parameters: cdktf.stringToTerraform(this._inputParameters),
            lambda_function_arn: cdktf.stringToTerraform(this._lambdaFunctionArn),
            maximum_execution_frequency: cdktf.stringToTerraform(this._maximumExecutionFrequency),
            name: cdktf.stringToTerraform(this._name),
            resource_id_scope: cdktf.stringToTerraform(this._resourceIdScope),
            resource_types_scope: cdktf.listMapper(cdktf.stringToTerraform)(this._resourceTypesScope),
            tag_key_scope: cdktf.stringToTerraform(this._tagKeyScope),
            tag_value_scope: cdktf.stringToTerraform(this._tagValueScope),
            trigger_types: cdktf.listMapper(cdktf.stringToTerraform)(this._triggerTypes),
            timeouts: configOrganizationCustomRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ConfigOrganizationCustomRule = ConfigOrganizationCustomRule;
_b = JSII_RTTI_SYMBOL_1;
ConfigOrganizationCustomRule[_b] = { fqn: "@cdktf/provider-aws.config.ConfigOrganizationCustomRule", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ConfigOrganizationCustomRule.tfResourceType = "aws_config_organization_custom_rule";
