// @cdktf/provider-aws/lib/guardduty/guardduty-member.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuarddutyMember = exports.GuarddutyMemberTimeoutsOutputReference = exports.guarddutyMemberTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function guarddutyMemberTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.guarddutyMemberTimeoutsToTerraform = guarddutyMemberTimeoutsToTerraform;
/**
 * @stability stable
 */
class GuarddutyMemberTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.GuarddutyMemberTimeoutsOutputReference = GuarddutyMemberTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GuarddutyMemberTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.guardduty.GuarddutyMemberTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/guardduty_member.html aws_guardduty_member}.
 *
 * @stability stable
 */
class GuarddutyMember extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/guardduty_member.html aws_guardduty_member} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_member',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GuarddutyMemberTimeoutsOutputReference(this, "timeouts", true);
        this._accountId = config.accountId;
        this._detectorId = config.detectorId;
        this._disableEmailNotification = config.disableEmailNotification;
        this._email = config.email;
        this._invitationMessage = config.invitationMessage;
        this._invite = config.invite;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    /**
     * @stability stable
     */
    get detectorId() {
        return this.getStringAttribute('detector_id');
    }
    /**
     * @stability stable
     */
    set detectorId(value) {
        this._detectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get detectorIdInput() {
        return this._detectorId;
    }
    /**
     * @stability stable
     */
    get disableEmailNotification() {
        return this.getBooleanAttribute('disable_email_notification');
    }
    /**
     * @stability stable
     */
    set disableEmailNotification(value) {
        this._disableEmailNotification = value;
    }
    /**
     * @stability stable
     */
    resetDisableEmailNotification() {
        this._disableEmailNotification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disableEmailNotificationInput() {
        return this._disableEmailNotification;
    }
    /**
     * @stability stable
     */
    get email() {
        return this.getStringAttribute('email');
    }
    /**
     * @stability stable
     */
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get emailInput() {
        return this._email;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get invitationMessage() {
        return this.getStringAttribute('invitation_message');
    }
    /**
     * @stability stable
     */
    set invitationMessage(value) {
        this._invitationMessage = value;
    }
    /**
     * @stability stable
     */
    resetInvitationMessage() {
        this._invitationMessage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get invitationMessageInput() {
        return this._invitationMessage;
    }
    /**
     * @stability stable
     */
    get invite() {
        return this.getBooleanAttribute('invite');
    }
    /**
     * @stability stable
     */
    set invite(value) {
        this._invite = value;
    }
    /**
     * @stability stable
     */
    resetInvite() {
        this._invite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inviteInput() {
        return this._invite;
    }
    // relationship_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get relationshipStatus() {
        return this.getStringAttribute('relationship_status');
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            detector_id: cdktf.stringToTerraform(this._detectorId),
            disable_email_notification: cdktf.booleanToTerraform(this._disableEmailNotification),
            email: cdktf.stringToTerraform(this._email),
            invitation_message: cdktf.stringToTerraform(this._invitationMessage),
            invite: cdktf.booleanToTerraform(this._invite),
            timeouts: guarddutyMemberTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GuarddutyMember = GuarddutyMember;
_b = JSII_RTTI_SYMBOL_1;
GuarddutyMember[_b] = { fqn: "@cdktf/provider-aws.guardduty.GuarddutyMember", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GuarddutyMember.tfResourceType = "aws_guardduty_member";
