// @cdktf/provider-aws/lib/guardduty/guardduty-detector.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuarddutyDetector = exports.GuarddutyDetectorDatasourcesOutputReference = exports.guarddutyDetectorDatasourcesToTerraform = exports.GuarddutyDetectorDatasourcesS3LogsOutputReference = exports.guarddutyDetectorDatasourcesS3LogsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function guarddutyDetectorDatasourcesS3LogsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable: cdktf.booleanToTerraform(struct.enable),
    };
}
exports.guarddutyDetectorDatasourcesS3LogsToTerraform = guarddutyDetectorDatasourcesS3LogsToTerraform;
/**
 * @stability stable
 */
class GuarddutyDetectorDatasourcesS3LogsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enable) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enable = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enable = value.enable;
        }
    }
    /**
     * @stability stable
     */
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    /**
     * @stability stable
     */
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableInput() {
        return this._enable;
    }
}
exports.GuarddutyDetectorDatasourcesS3LogsOutputReference = GuarddutyDetectorDatasourcesS3LogsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GuarddutyDetectorDatasourcesS3LogsOutputReference[_a] = { fqn: "@cdktf/provider-aws.guardduty.GuarddutyDetectorDatasourcesS3LogsOutputReference", version: "3.0.1" };
function guarddutyDetectorDatasourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_logs: guarddutyDetectorDatasourcesS3LogsToTerraform(struct.s3Logs),
    };
}
exports.guarddutyDetectorDatasourcesToTerraform = guarddutyDetectorDatasourcesToTerraform;
/**
 * @stability stable
 */
class GuarddutyDetectorDatasourcesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // s3_logs - computed: false, optional: true, required: false
        this._s3Logs = new GuarddutyDetectorDatasourcesS3LogsOutputReference(this, "s3_logs", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _d, _e;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_d = this._s3Logs) === null || _d === void 0 ? void 0 : _d.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3Logs = (_e = this._s3Logs) === null || _e === void 0 ? void 0 : _e.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3Logs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3Logs.internalValue = value.s3Logs;
        }
    }
    /**
     * @stability stable
     */
    get s3Logs() {
        return this._s3Logs;
    }
    /**
     * @stability stable
     */
    putS3Logs(value) {
        this._s3Logs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetS3Logs() {
        this._s3Logs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3LogsInput() {
        return this._s3Logs.internalValue;
    }
}
exports.GuarddutyDetectorDatasourcesOutputReference = GuarddutyDetectorDatasourcesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GuarddutyDetectorDatasourcesOutputReference[_b] = { fqn: "@cdktf/provider-aws.guardduty.GuarddutyDetectorDatasourcesOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/guardduty_detector.html aws_guardduty_detector}.
 *
 * @stability stable
 */
class GuarddutyDetector extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/guardduty_detector.html aws_guardduty_detector} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_detector',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // datasources - computed: false, optional: true, required: false
        this._datasources = new GuarddutyDetectorDatasourcesOutputReference(this, "datasources", true);
        this._enable = config.enable;
        this._findingPublishingFrequency = config.findingPublishingFrequency;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._datasources.internalValue = config.datasources;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    /**
     * @stability stable
     */
    set enable(value) {
        this._enable = value;
    }
    /**
     * @stability stable
     */
    resetEnable() {
        this._enable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableInput() {
        return this._enable;
    }
    /**
     * @stability stable
     */
    get findingPublishingFrequency() {
        return this.getStringAttribute('finding_publishing_frequency');
    }
    /**
     * @stability stable
     */
    set findingPublishingFrequency(value) {
        this._findingPublishingFrequency = value;
    }
    /**
     * @stability stable
     */
    resetFindingPublishingFrequency() {
        this._findingPublishingFrequency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get findingPublishingFrequencyInput() {
        return this._findingPublishingFrequency;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get datasources() {
        return this._datasources;
    }
    /**
     * @stability stable
     */
    putDatasources(value) {
        this._datasources.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDatasources() {
        this._datasources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get datasourcesInput() {
        return this._datasources.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enable: cdktf.booleanToTerraform(this._enable),
            finding_publishing_frequency: cdktf.stringToTerraform(this._findingPublishingFrequency),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            datasources: guarddutyDetectorDatasourcesToTerraform(this._datasources.internalValue),
        };
    }
}
exports.GuarddutyDetector = GuarddutyDetector;
_c = JSII_RTTI_SYMBOL_1;
GuarddutyDetector[_c] = { fqn: "@cdktf/provider-aws.guardduty.GuarddutyDetector", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GuarddutyDetector.tfResourceType = "aws_guardduty_detector";
