// @cdktf/provider-aws/lib/licensemanager/licensemanager-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicensemanagerAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/licensemanager_association.html aws_licensemanager_association}.
 *
 * @stability stable
 */
class LicensemanagerAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/licensemanager_association.html aws_licensemanager_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_licensemanager_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._licenseConfigurationArn = config.licenseConfigurationArn;
        this._resourceArn = config.resourceArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get licenseConfigurationArn() {
        return this.getStringAttribute('license_configuration_arn');
    }
    /**
     * @stability stable
     */
    set licenseConfigurationArn(value) {
        this._licenseConfigurationArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get licenseConfigurationArnInput() {
        return this._licenseConfigurationArn;
    }
    /**
     * @stability stable
     */
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    /**
     * @stability stable
     */
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceArnInput() {
        return this._resourceArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            license_configuration_arn: cdktf.stringToTerraform(this._licenseConfigurationArn),
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
        };
    }
}
exports.LicensemanagerAssociation = LicensemanagerAssociation;
_a = JSII_RTTI_SYMBOL_1;
LicensemanagerAssociation[_a] = { fqn: "@cdktf/provider-aws.licensemanager.LicensemanagerAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LicensemanagerAssociation.tfResourceType = "aws_licensemanager_association";
