// @cdktf/provider-aws/lib/opsworks/opsworks-ganglia-layer.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksGangliaLayer = exports.opsworksGangliaLayerEbsVolumeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function opsworksGangliaLayerEbsVolumeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encrypted: cdktf.booleanToTerraform(struct.encrypted),
        iops: cdktf.numberToTerraform(struct.iops),
        mount_point: cdktf.stringToTerraform(struct.mountPoint),
        number_of_disks: cdktf.numberToTerraform(struct.numberOfDisks),
        raid_level: cdktf.stringToTerraform(struct.raidLevel),
        size: cdktf.numberToTerraform(struct.size),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.opsworksGangliaLayerEbsVolumeToTerraform = opsworksGangliaLayerEbsVolumeToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/opsworks_ganglia_layer.html aws_opsworks_ganglia_layer}.
 *
 * @stability stable
 */
class OpsworksGangliaLayer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/opsworks_ganglia_layer.html aws_opsworks_ganglia_layer} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_ganglia_layer',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._autoAssignElasticIps = config.autoAssignElasticIps;
        this._autoAssignPublicIps = config.autoAssignPublicIps;
        this._autoHealing = config.autoHealing;
        this._customConfigureRecipes = config.customConfigureRecipes;
        this._customDeployRecipes = config.customDeployRecipes;
        this._customInstanceProfileArn = config.customInstanceProfileArn;
        this._customJson = config.customJson;
        this._customSecurityGroupIds = config.customSecurityGroupIds;
        this._customSetupRecipes = config.customSetupRecipes;
        this._customShutdownRecipes = config.customShutdownRecipes;
        this._customUndeployRecipes = config.customUndeployRecipes;
        this._drainElbOnShutdown = config.drainElbOnShutdown;
        this._elasticLoadBalancer = config.elasticLoadBalancer;
        this._installUpdatesOnBoot = config.installUpdatesOnBoot;
        this._instanceShutdownTimeout = config.instanceShutdownTimeout;
        this._name = config.name;
        this._password = config.password;
        this._stackId = config.stackId;
        this._systemPackages = config.systemPackages;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._url = config.url;
        this._useEbsOptimizedInstances = config.useEbsOptimizedInstances;
        this._username = config.username;
        this._ebsVolume = config.ebsVolume;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get autoAssignElasticIps() {
        return this.getBooleanAttribute('auto_assign_elastic_ips');
    }
    /**
     * @stability stable
     */
    set autoAssignElasticIps(value) {
        this._autoAssignElasticIps = value;
    }
    /**
     * @stability stable
     */
    resetAutoAssignElasticIps() {
        this._autoAssignElasticIps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoAssignElasticIpsInput() {
        return this._autoAssignElasticIps;
    }
    /**
     * @stability stable
     */
    get autoAssignPublicIps() {
        return this.getBooleanAttribute('auto_assign_public_ips');
    }
    /**
     * @stability stable
     */
    set autoAssignPublicIps(value) {
        this._autoAssignPublicIps = value;
    }
    /**
     * @stability stable
     */
    resetAutoAssignPublicIps() {
        this._autoAssignPublicIps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoAssignPublicIpsInput() {
        return this._autoAssignPublicIps;
    }
    /**
     * @stability stable
     */
    get autoHealing() {
        return this.getBooleanAttribute('auto_healing');
    }
    /**
     * @stability stable
     */
    set autoHealing(value) {
        this._autoHealing = value;
    }
    /**
     * @stability stable
     */
    resetAutoHealing() {
        this._autoHealing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get autoHealingInput() {
        return this._autoHealing;
    }
    /**
     * @stability stable
     */
    get customConfigureRecipes() {
        return this.getListAttribute('custom_configure_recipes');
    }
    /**
     * @stability stable
     */
    set customConfigureRecipes(value) {
        this._customConfigureRecipes = value;
    }
    /**
     * @stability stable
     */
    resetCustomConfigureRecipes() {
        this._customConfigureRecipes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customConfigureRecipesInput() {
        return this._customConfigureRecipes;
    }
    /**
     * @stability stable
     */
    get customDeployRecipes() {
        return this.getListAttribute('custom_deploy_recipes');
    }
    /**
     * @stability stable
     */
    set customDeployRecipes(value) {
        this._customDeployRecipes = value;
    }
    /**
     * @stability stable
     */
    resetCustomDeployRecipes() {
        this._customDeployRecipes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customDeployRecipesInput() {
        return this._customDeployRecipes;
    }
    /**
     * @stability stable
     */
    get customInstanceProfileArn() {
        return this.getStringAttribute('custom_instance_profile_arn');
    }
    /**
     * @stability stable
     */
    set customInstanceProfileArn(value) {
        this._customInstanceProfileArn = value;
    }
    /**
     * @stability stable
     */
    resetCustomInstanceProfileArn() {
        this._customInstanceProfileArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customInstanceProfileArnInput() {
        return this._customInstanceProfileArn;
    }
    /**
     * @stability stable
     */
    get customJson() {
        return this.getStringAttribute('custom_json');
    }
    /**
     * @stability stable
     */
    set customJson(value) {
        this._customJson = value;
    }
    /**
     * @stability stable
     */
    resetCustomJson() {
        this._customJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customJsonInput() {
        return this._customJson;
    }
    /**
     * @stability stable
     */
    get customSecurityGroupIds() {
        return this.getListAttribute('custom_security_group_ids');
    }
    /**
     * @stability stable
     */
    set customSecurityGroupIds(value) {
        this._customSecurityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetCustomSecurityGroupIds() {
        this._customSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customSecurityGroupIdsInput() {
        return this._customSecurityGroupIds;
    }
    /**
     * @stability stable
     */
    get customSetupRecipes() {
        return this.getListAttribute('custom_setup_recipes');
    }
    /**
     * @stability stable
     */
    set customSetupRecipes(value) {
        this._customSetupRecipes = value;
    }
    /**
     * @stability stable
     */
    resetCustomSetupRecipes() {
        this._customSetupRecipes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customSetupRecipesInput() {
        return this._customSetupRecipes;
    }
    /**
     * @stability stable
     */
    get customShutdownRecipes() {
        return this.getListAttribute('custom_shutdown_recipes');
    }
    /**
     * @stability stable
     */
    set customShutdownRecipes(value) {
        this._customShutdownRecipes = value;
    }
    /**
     * @stability stable
     */
    resetCustomShutdownRecipes() {
        this._customShutdownRecipes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customShutdownRecipesInput() {
        return this._customShutdownRecipes;
    }
    /**
     * @stability stable
     */
    get customUndeployRecipes() {
        return this.getListAttribute('custom_undeploy_recipes');
    }
    /**
     * @stability stable
     */
    set customUndeployRecipes(value) {
        this._customUndeployRecipes = value;
    }
    /**
     * @stability stable
     */
    resetCustomUndeployRecipes() {
        this._customUndeployRecipes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get customUndeployRecipesInput() {
        return this._customUndeployRecipes;
    }
    /**
     * @stability stable
     */
    get drainElbOnShutdown() {
        return this.getBooleanAttribute('drain_elb_on_shutdown');
    }
    /**
     * @stability stable
     */
    set drainElbOnShutdown(value) {
        this._drainElbOnShutdown = value;
    }
    /**
     * @stability stable
     */
    resetDrainElbOnShutdown() {
        this._drainElbOnShutdown = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get drainElbOnShutdownInput() {
        return this._drainElbOnShutdown;
    }
    /**
     * @stability stable
     */
    get elasticLoadBalancer() {
        return this.getStringAttribute('elastic_load_balancer');
    }
    /**
     * @stability stable
     */
    set elasticLoadBalancer(value) {
        this._elasticLoadBalancer = value;
    }
    /**
     * @stability stable
     */
    resetElasticLoadBalancer() {
        this._elasticLoadBalancer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticLoadBalancerInput() {
        return this._elasticLoadBalancer;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get installUpdatesOnBoot() {
        return this.getBooleanAttribute('install_updates_on_boot');
    }
    /**
     * @stability stable
     */
    set installUpdatesOnBoot(value) {
        this._installUpdatesOnBoot = value;
    }
    /**
     * @stability stable
     */
    resetInstallUpdatesOnBoot() {
        this._installUpdatesOnBoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get installUpdatesOnBootInput() {
        return this._installUpdatesOnBoot;
    }
    /**
     * @stability stable
     */
    get instanceShutdownTimeout() {
        return this.getNumberAttribute('instance_shutdown_timeout');
    }
    /**
     * @stability stable
     */
    set instanceShutdownTimeout(value) {
        this._instanceShutdownTimeout = value;
    }
    /**
     * @stability stable
     */
    resetInstanceShutdownTimeout() {
        this._instanceShutdownTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceShutdownTimeoutInput() {
        return this._instanceShutdownTimeout;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get password() {
        return this.getStringAttribute('password');
    }
    /**
     * @stability stable
     */
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passwordInput() {
        return this._password;
    }
    /**
     * @stability stable
     */
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    /**
     * @stability stable
     */
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stackIdInput() {
        return this._stackId;
    }
    /**
     * @stability stable
     */
    get systemPackages() {
        return this.getListAttribute('system_packages');
    }
    /**
     * @stability stable
     */
    set systemPackages(value) {
        this._systemPackages = value;
    }
    /**
     * @stability stable
     */
    resetSystemPackages() {
        this._systemPackages = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get systemPackagesInput() {
        return this._systemPackages;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get url() {
        return this.getStringAttribute('url');
    }
    /**
     * @stability stable
     */
    set url(value) {
        this._url = value;
    }
    /**
     * @stability stable
     */
    resetUrl() {
        this._url = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get urlInput() {
        return this._url;
    }
    /**
     * @stability stable
     */
    get useEbsOptimizedInstances() {
        return this.getBooleanAttribute('use_ebs_optimized_instances');
    }
    /**
     * @stability stable
     */
    set useEbsOptimizedInstances(value) {
        this._useEbsOptimizedInstances = value;
    }
    /**
     * @stability stable
     */
    resetUseEbsOptimizedInstances() {
        this._useEbsOptimizedInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get useEbsOptimizedInstancesInput() {
        return this._useEbsOptimizedInstances;
    }
    /**
     * @stability stable
     */
    get username() {
        return this.getStringAttribute('username');
    }
    /**
     * @stability stable
     */
    set username(value) {
        this._username = value;
    }
    /**
     * @stability stable
     */
    resetUsername() {
        this._username = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get usernameInput() {
        return this._username;
    }
    /**
     * @stability stable
     */
    get ebsVolume() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ebs_volume');
    }
    /**
     * @stability stable
     */
    set ebsVolume(value) {
        this._ebsVolume = value;
    }
    /**
     * @stability stable
     */
    resetEbsVolume() {
        this._ebsVolume = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsVolumeInput() {
        return this._ebsVolume;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            auto_assign_elastic_ips: cdktf.booleanToTerraform(this._autoAssignElasticIps),
            auto_assign_public_ips: cdktf.booleanToTerraform(this._autoAssignPublicIps),
            auto_healing: cdktf.booleanToTerraform(this._autoHealing),
            custom_configure_recipes: cdktf.listMapper(cdktf.stringToTerraform)(this._customConfigureRecipes),
            custom_deploy_recipes: cdktf.listMapper(cdktf.stringToTerraform)(this._customDeployRecipes),
            custom_instance_profile_arn: cdktf.stringToTerraform(this._customInstanceProfileArn),
            custom_json: cdktf.stringToTerraform(this._customJson),
            custom_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._customSecurityGroupIds),
            custom_setup_recipes: cdktf.listMapper(cdktf.stringToTerraform)(this._customSetupRecipes),
            custom_shutdown_recipes: cdktf.listMapper(cdktf.stringToTerraform)(this._customShutdownRecipes),
            custom_undeploy_recipes: cdktf.listMapper(cdktf.stringToTerraform)(this._customUndeployRecipes),
            drain_elb_on_shutdown: cdktf.booleanToTerraform(this._drainElbOnShutdown),
            elastic_load_balancer: cdktf.stringToTerraform(this._elasticLoadBalancer),
            install_updates_on_boot: cdktf.booleanToTerraform(this._installUpdatesOnBoot),
            instance_shutdown_timeout: cdktf.numberToTerraform(this._instanceShutdownTimeout),
            name: cdktf.stringToTerraform(this._name),
            password: cdktf.stringToTerraform(this._password),
            stack_id: cdktf.stringToTerraform(this._stackId),
            system_packages: cdktf.listMapper(cdktf.stringToTerraform)(this._systemPackages),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            url: cdktf.stringToTerraform(this._url),
            use_ebs_optimized_instances: cdktf.booleanToTerraform(this._useEbsOptimizedInstances),
            username: cdktf.stringToTerraform(this._username),
            ebs_volume: cdktf.listMapper(opsworksGangliaLayerEbsVolumeToTerraform)(this._ebsVolume),
        };
    }
}
exports.OpsworksGangliaLayer = OpsworksGangliaLayer;
_a = JSII_RTTI_SYMBOL_1;
OpsworksGangliaLayer[_a] = { fqn: "@cdktf/provider-aws.opsworks.OpsworksGangliaLayer", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
OpsworksGangliaLayer.tfResourceType = "aws_opsworks_ganglia_layer";
