// @cdktf/provider-aws/lib/budgets/budgets-budget.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BudgetsBudget = exports.budgetsBudgetNotificationToTerraform = exports.BudgetsBudgetCostTypesOutputReference = exports.budgetsBudgetCostTypesToTerraform = exports.budgetsBudgetCostFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function budgetsBudgetCostFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.budgetsBudgetCostFilterToTerraform = budgetsBudgetCostFilterToTerraform;
function budgetsBudgetCostTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        include_credit: cdktf.booleanToTerraform(struct.includeCredit),
        include_discount: cdktf.booleanToTerraform(struct.includeDiscount),
        include_other_subscription: cdktf.booleanToTerraform(struct.includeOtherSubscription),
        include_recurring: cdktf.booleanToTerraform(struct.includeRecurring),
        include_refund: cdktf.booleanToTerraform(struct.includeRefund),
        include_subscription: cdktf.booleanToTerraform(struct.includeSubscription),
        include_support: cdktf.booleanToTerraform(struct.includeSupport),
        include_tax: cdktf.booleanToTerraform(struct.includeTax),
        include_upfront: cdktf.booleanToTerraform(struct.includeUpfront),
        use_amortized: cdktf.booleanToTerraform(struct.useAmortized),
        use_blended: cdktf.booleanToTerraform(struct.useBlended),
    };
}
exports.budgetsBudgetCostTypesToTerraform = budgetsBudgetCostTypesToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetCostTypesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._includeCredit) {
            hasAnyValues = true;
            internalValueResult.includeCredit = this._includeCredit;
        }
        if (this._includeDiscount) {
            hasAnyValues = true;
            internalValueResult.includeDiscount = this._includeDiscount;
        }
        if (this._includeOtherSubscription) {
            hasAnyValues = true;
            internalValueResult.includeOtherSubscription = this._includeOtherSubscription;
        }
        if (this._includeRecurring) {
            hasAnyValues = true;
            internalValueResult.includeRecurring = this._includeRecurring;
        }
        if (this._includeRefund) {
            hasAnyValues = true;
            internalValueResult.includeRefund = this._includeRefund;
        }
        if (this._includeSubscription) {
            hasAnyValues = true;
            internalValueResult.includeSubscription = this._includeSubscription;
        }
        if (this._includeSupport) {
            hasAnyValues = true;
            internalValueResult.includeSupport = this._includeSupport;
        }
        if (this._includeTax) {
            hasAnyValues = true;
            internalValueResult.includeTax = this._includeTax;
        }
        if (this._includeUpfront) {
            hasAnyValues = true;
            internalValueResult.includeUpfront = this._includeUpfront;
        }
        if (this._useAmortized) {
            hasAnyValues = true;
            internalValueResult.useAmortized = this._useAmortized;
        }
        if (this._useBlended) {
            hasAnyValues = true;
            internalValueResult.useBlended = this._useBlended;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._includeCredit = undefined;
            this._includeDiscount = undefined;
            this._includeOtherSubscription = undefined;
            this._includeRecurring = undefined;
            this._includeRefund = undefined;
            this._includeSubscription = undefined;
            this._includeSupport = undefined;
            this._includeTax = undefined;
            this._includeUpfront = undefined;
            this._useAmortized = undefined;
            this._useBlended = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._includeCredit = value.includeCredit;
            this._includeDiscount = value.includeDiscount;
            this._includeOtherSubscription = value.includeOtherSubscription;
            this._includeRecurring = value.includeRecurring;
            this._includeRefund = value.includeRefund;
            this._includeSubscription = value.includeSubscription;
            this._includeSupport = value.includeSupport;
            this._includeTax = value.includeTax;
            this._includeUpfront = value.includeUpfront;
            this._useAmortized = value.useAmortized;
            this._useBlended = value.useBlended;
        }
    }
    /**
     * @stability stable
     */
    get includeCredit() {
        return this.getBooleanAttribute('include_credit');
    }
    /**
     * @stability stable
     */
    set includeCredit(value) {
        this._includeCredit = value;
    }
    /**
     * @stability stable
     */
    resetIncludeCredit() {
        this._includeCredit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeCreditInput() {
        return this._includeCredit;
    }
    /**
     * @stability stable
     */
    get includeDiscount() {
        return this.getBooleanAttribute('include_discount');
    }
    /**
     * @stability stable
     */
    set includeDiscount(value) {
        this._includeDiscount = value;
    }
    /**
     * @stability stable
     */
    resetIncludeDiscount() {
        this._includeDiscount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeDiscountInput() {
        return this._includeDiscount;
    }
    /**
     * @stability stable
     */
    get includeOtherSubscription() {
        return this.getBooleanAttribute('include_other_subscription');
    }
    /**
     * @stability stable
     */
    set includeOtherSubscription(value) {
        this._includeOtherSubscription = value;
    }
    /**
     * @stability stable
     */
    resetIncludeOtherSubscription() {
        this._includeOtherSubscription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeOtherSubscriptionInput() {
        return this._includeOtherSubscription;
    }
    /**
     * @stability stable
     */
    get includeRecurring() {
        return this.getBooleanAttribute('include_recurring');
    }
    /**
     * @stability stable
     */
    set includeRecurring(value) {
        this._includeRecurring = value;
    }
    /**
     * @stability stable
     */
    resetIncludeRecurring() {
        this._includeRecurring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeRecurringInput() {
        return this._includeRecurring;
    }
    /**
     * @stability stable
     */
    get includeRefund() {
        return this.getBooleanAttribute('include_refund');
    }
    /**
     * @stability stable
     */
    set includeRefund(value) {
        this._includeRefund = value;
    }
    /**
     * @stability stable
     */
    resetIncludeRefund() {
        this._includeRefund = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeRefundInput() {
        return this._includeRefund;
    }
    /**
     * @stability stable
     */
    get includeSubscription() {
        return this.getBooleanAttribute('include_subscription');
    }
    /**
     * @stability stable
     */
    set includeSubscription(value) {
        this._includeSubscription = value;
    }
    /**
     * @stability stable
     */
    resetIncludeSubscription() {
        this._includeSubscription = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeSubscriptionInput() {
        return this._includeSubscription;
    }
    /**
     * @stability stable
     */
    get includeSupport() {
        return this.getBooleanAttribute('include_support');
    }
    /**
     * @stability stable
     */
    set includeSupport(value) {
        this._includeSupport = value;
    }
    /**
     * @stability stable
     */
    resetIncludeSupport() {
        this._includeSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeSupportInput() {
        return this._includeSupport;
    }
    /**
     * @stability stable
     */
    get includeTax() {
        return this.getBooleanAttribute('include_tax');
    }
    /**
     * @stability stable
     */
    set includeTax(value) {
        this._includeTax = value;
    }
    /**
     * @stability stable
     */
    resetIncludeTax() {
        this._includeTax = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeTaxInput() {
        return this._includeTax;
    }
    /**
     * @stability stable
     */
    get includeUpfront() {
        return this.getBooleanAttribute('include_upfront');
    }
    /**
     * @stability stable
     */
    set includeUpfront(value) {
        this._includeUpfront = value;
    }
    /**
     * @stability stable
     */
    resetIncludeUpfront() {
        this._includeUpfront = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeUpfrontInput() {
        return this._includeUpfront;
    }
    /**
     * @stability stable
     */
    get useAmortized() {
        return this.getBooleanAttribute('use_amortized');
    }
    /**
     * @stability stable
     */
    set useAmortized(value) {
        this._useAmortized = value;
    }
    /**
     * @stability stable
     */
    resetUseAmortized() {
        this._useAmortized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get useAmortizedInput() {
        return this._useAmortized;
    }
    /**
     * @stability stable
     */
    get useBlended() {
        return this.getBooleanAttribute('use_blended');
    }
    /**
     * @stability stable
     */
    set useBlended(value) {
        this._useBlended = value;
    }
    /**
     * @stability stable
     */
    resetUseBlended() {
        this._useBlended = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get useBlendedInput() {
        return this._useBlended;
    }
}
exports.BudgetsBudgetCostTypesOutputReference = BudgetsBudgetCostTypesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BudgetsBudgetCostTypesOutputReference[_a] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetCostTypesOutputReference", version: "3.0.1" };
function budgetsBudgetNotificationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comparison_operator: cdktf.stringToTerraform(struct.comparisonOperator),
        notification_type: cdktf.stringToTerraform(struct.notificationType),
        subscriber_email_addresses: cdktf.listMapper(cdktf.stringToTerraform)(struct.subscriberEmailAddresses),
        subscriber_sns_topic_arns: cdktf.listMapper(cdktf.stringToTerraform)(struct.subscriberSnsTopicArns),
        threshold: cdktf.numberToTerraform(struct.threshold),
        threshold_type: cdktf.stringToTerraform(struct.thresholdType),
    };
}
exports.budgetsBudgetNotificationToTerraform = budgetsBudgetNotificationToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/budgets_budget.html aws_budgets_budget}.
 *
 * @stability stable
 */
class BudgetsBudget extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/budgets_budget.html aws_budgets_budget} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_budgets_budget',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // cost_types - computed: false, optional: true, required: false
        this._costTypes = new BudgetsBudgetCostTypesOutputReference(this, "cost_types", true);
        this._accountId = config.accountId;
        this._budgetType = config.budgetType;
        this._costFilters = config.costFilters;
        this._limitAmount = config.limitAmount;
        this._limitUnit = config.limitUnit;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._timePeriodEnd = config.timePeriodEnd;
        this._timePeriodStart = config.timePeriodStart;
        this._timeUnit = config.timeUnit;
        this._costFilter = config.costFilter;
        this._costTypes.internalValue = config.costTypes;
        this._notification = config.notification;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get budgetType() {
        return this.getStringAttribute('budget_type');
    }
    /**
     * @stability stable
     */
    set budgetType(value) {
        this._budgetType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get budgetTypeInput() {
        return this._budgetType;
    }
    /**
     * @stability stable
     */
    get costFilters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('cost_filters');
    }
    /**
     * @stability stable
     */
    set costFilters(value) {
        this._costFilters = value;
    }
    /**
     * @stability stable
     */
    resetCostFilters() {
        this._costFilters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get costFiltersInput() {
        return this._costFilters;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get limitAmount() {
        return this.getStringAttribute('limit_amount');
    }
    /**
     * @stability stable
     */
    set limitAmount(value) {
        this._limitAmount = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get limitAmountInput() {
        return this._limitAmount;
    }
    /**
     * @stability stable
     */
    get limitUnit() {
        return this.getStringAttribute('limit_unit');
    }
    /**
     * @stability stable
     */
    set limitUnit(value) {
        this._limitUnit = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get limitUnitInput() {
        return this._limitUnit;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    /**
     * @stability stable
     */
    get timePeriodEnd() {
        return this.getStringAttribute('time_period_end');
    }
    /**
     * @stability stable
     */
    set timePeriodEnd(value) {
        this._timePeriodEnd = value;
    }
    /**
     * @stability stable
     */
    resetTimePeriodEnd() {
        this._timePeriodEnd = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timePeriodEndInput() {
        return this._timePeriodEnd;
    }
    /**
     * @stability stable
     */
    get timePeriodStart() {
        return this.getStringAttribute('time_period_start');
    }
    /**
     * @stability stable
     */
    set timePeriodStart(value) {
        this._timePeriodStart = value;
    }
    /**
     * @stability stable
     */
    resetTimePeriodStart() {
        this._timePeriodStart = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timePeriodStartInput() {
        return this._timePeriodStart;
    }
    /**
     * @stability stable
     */
    get timeUnit() {
        return this.getStringAttribute('time_unit');
    }
    /**
     * @stability stable
     */
    set timeUnit(value) {
        this._timeUnit = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeUnitInput() {
        return this._timeUnit;
    }
    /**
     * @stability stable
     */
    get costFilter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('cost_filter');
    }
    /**
     * @stability stable
     */
    set costFilter(value) {
        this._costFilter = value;
    }
    /**
     * @stability stable
     */
    resetCostFilter() {
        this._costFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get costFilterInput() {
        return this._costFilter;
    }
    /**
     * @stability stable
     */
    get costTypes() {
        return this._costTypes;
    }
    /**
     * @stability stable
     */
    putCostTypes(value) {
        this._costTypes.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCostTypes() {
        this._costTypes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get costTypesInput() {
        return this._costTypes.internalValue;
    }
    /**
     * @stability stable
     */
    get notification() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('notification');
    }
    /**
     * @stability stable
     */
    set notification(value) {
        this._notification = value;
    }
    /**
     * @stability stable
     */
    resetNotification() {
        this._notification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationInput() {
        return this._notification;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            budget_type: cdktf.stringToTerraform(this._budgetType),
            cost_filters: cdktf.hashMapper(cdktf.anyToTerraform)(this._costFilters),
            limit_amount: cdktf.stringToTerraform(this._limitAmount),
            limit_unit: cdktf.stringToTerraform(this._limitUnit),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            time_period_end: cdktf.stringToTerraform(this._timePeriodEnd),
            time_period_start: cdktf.stringToTerraform(this._timePeriodStart),
            time_unit: cdktf.stringToTerraform(this._timeUnit),
            cost_filter: cdktf.listMapper(budgetsBudgetCostFilterToTerraform)(this._costFilter),
            cost_types: budgetsBudgetCostTypesToTerraform(this._costTypes.internalValue),
            notification: cdktf.listMapper(budgetsBudgetNotificationToTerraform)(this._notification),
        };
    }
}
exports.BudgetsBudget = BudgetsBudget;
_b = JSII_RTTI_SYMBOL_1;
BudgetsBudget[_b] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudget", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
BudgetsBudget.tfResourceType = "aws_budgets_budget";
