// @cdktf/provider-aws/lib/msk/data-aws-msk-broker-nodes.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsMskBrokerNodes = exports.DataAwsMskBrokerNodesNodeInfoList = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsMskBrokerNodesNodeInfoList extends cdktf.ComplexComputedList {
    // attached_eni_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get attachedEniId() {
        return this.getStringAttribute('attached_eni_id');
    }
    // broker_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get brokerId() {
        return this.getNumberAttribute('broker_id');
    }
    // client_subnet - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clientSubnet() {
        return this.getStringAttribute('client_subnet');
    }
    // client_vpc_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clientVpcIpAddress() {
        return this.getStringAttribute('client_vpc_ip_address');
    }
    // endpoints - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoints() {
        return this.getListAttribute('endpoints');
    }
    // node_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nodeArn() {
        return this.getStringAttribute('node_arn');
    }
}
exports.DataAwsMskBrokerNodesNodeInfoList = DataAwsMskBrokerNodesNodeInfoList;
_a = JSII_RTTI_SYMBOL_1;
DataAwsMskBrokerNodesNodeInfoList[_a] = { fqn: "@cdktf/provider-aws.msk.DataAwsMskBrokerNodesNodeInfoList", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/msk_broker_nodes.html aws_msk_broker_nodes}.
 *
 * @stability stable
 */
class DataAwsMskBrokerNodes extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/msk_broker_nodes.html aws_msk_broker_nodes} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_broker_nodes',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterArn = config.clusterArn;
    }
    /**
     * @stability stable
     */
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    /**
     * @stability stable
     */
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterArnInput() {
        return this._clusterArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // node_info_list - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    nodeInfoList(index) {
        return new DataAwsMskBrokerNodesNodeInfoList(this, 'node_info_list', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
        };
    }
}
exports.DataAwsMskBrokerNodes = DataAwsMskBrokerNodes;
_b = JSII_RTTI_SYMBOL_1;
DataAwsMskBrokerNodes[_b] = { fqn: "@cdktf/provider-aws.msk.DataAwsMskBrokerNodes", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsMskBrokerNodes.tfResourceType = "aws_msk_broker_nodes";
