// @cdktf/provider-aws/lib/msk/msk-configuration.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_configuration.html aws_msk_configuration}.
 *
 * @stability stable
 */
class MskConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_configuration.html aws_msk_configuration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._description = config.description;
        this._kafkaVersions = config.kafkaVersions;
        this._name = config.name;
        this._serverProperties = config.serverProperties;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kafkaVersions() {
        return this.getListAttribute('kafka_versions');
    }
    /**
     * @stability stable
     */
    set kafkaVersions(value) {
        this._kafkaVersions = value;
    }
    /**
     * @stability stable
     */
    resetKafkaVersions() {
        this._kafkaVersions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kafkaVersionsInput() {
        return this._kafkaVersions;
    }
    // latest_revision - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get latestRevision() {
        return this.getNumberAttribute('latest_revision');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get serverProperties() {
        return this.getStringAttribute('server_properties');
    }
    /**
     * @stability stable
     */
    set serverProperties(value) {
        this._serverProperties = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serverPropertiesInput() {
        return this._serverProperties;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            kafka_versions: cdktf.listMapper(cdktf.stringToTerraform)(this._kafkaVersions),
            name: cdktf.stringToTerraform(this._name),
            server_properties: cdktf.stringToTerraform(this._serverProperties),
        };
    }
}
exports.MskConfiguration = MskConfiguration;
_a = JSII_RTTI_SYMBOL_1;
MskConfiguration[_a] = { fqn: "@cdktf/provider-aws.msk.MskConfiguration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MskConfiguration.tfResourceType = "aws_msk_configuration";
