// @cdktf/provider-aws/lib/macie/macie-s3-bucket-association.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MacieS3BucketAssociation = exports.MacieS3BucketAssociationClassificationTypeOutputReference = exports.macieS3BucketAssociationClassificationTypeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function macieS3BucketAssociationClassificationTypeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        continuous: cdktf.stringToTerraform(struct.continuous),
        one_time: cdktf.stringToTerraform(struct.oneTime),
    };
}
exports.macieS3BucketAssociationClassificationTypeToTerraform = macieS3BucketAssociationClassificationTypeToTerraform;
/**
 * @stability stable
 */
class MacieS3BucketAssociationClassificationTypeOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._continuous) {
            hasAnyValues = true;
            internalValueResult.continuous = this._continuous;
        }
        if (this._oneTime) {
            hasAnyValues = true;
            internalValueResult.oneTime = this._oneTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._continuous = undefined;
            this._oneTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._continuous = value.continuous;
            this._oneTime = value.oneTime;
        }
    }
    /**
     * @stability stable
     */
    get continuous() {
        return this.getStringAttribute('continuous');
    }
    /**
     * @stability stable
     */
    set continuous(value) {
        this._continuous = value;
    }
    /**
     * @stability stable
     */
    resetContinuous() {
        this._continuous = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get continuousInput() {
        return this._continuous;
    }
    /**
     * @stability stable
     */
    get oneTime() {
        return this.getStringAttribute('one_time');
    }
    /**
     * @stability stable
     */
    set oneTime(value) {
        this._oneTime = value;
    }
    /**
     * @stability stable
     */
    resetOneTime() {
        this._oneTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get oneTimeInput() {
        return this._oneTime;
    }
}
exports.MacieS3BucketAssociationClassificationTypeOutputReference = MacieS3BucketAssociationClassificationTypeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MacieS3BucketAssociationClassificationTypeOutputReference[_a] = { fqn: "@cdktf/provider-aws.macie.MacieS3BucketAssociationClassificationTypeOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/macie_s3_bucket_association.html aws_macie_s3_bucket_association}.
 *
 * @stability stable
 */
class MacieS3BucketAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/macie_s3_bucket_association.html aws_macie_s3_bucket_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_macie_s3_bucket_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // classification_type - computed: false, optional: true, required: false
        this._classificationType = new MacieS3BucketAssociationClassificationTypeOutputReference(this, "classification_type", true);
        this._bucketName = config.bucketName;
        this._memberAccountId = config.memberAccountId;
        this._prefix = config.prefix;
        this._classificationType.internalValue = config.classificationType;
    }
    /**
     * @stability stable
     */
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    /**
     * @stability stable
     */
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketNameInput() {
        return this._bucketName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get memberAccountId() {
        return this.getStringAttribute('member_account_id');
    }
    /**
     * @stability stable
     */
    set memberAccountId(value) {
        this._memberAccountId = value;
    }
    /**
     * @stability stable
     */
    resetMemberAccountId() {
        this._memberAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get memberAccountIdInput() {
        return this._memberAccountId;
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
    /**
     * @stability stable
     */
    get classificationType() {
        return this._classificationType;
    }
    /**
     * @stability stable
     */
    putClassificationType(value) {
        this._classificationType.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetClassificationType() {
        this._classificationType.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get classificationTypeInput() {
        return this._classificationType.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bucket_name: cdktf.stringToTerraform(this._bucketName),
            member_account_id: cdktf.stringToTerraform(this._memberAccountId),
            prefix: cdktf.stringToTerraform(this._prefix),
            classification_type: macieS3BucketAssociationClassificationTypeToTerraform(this._classificationType.internalValue),
        };
    }
}
exports.MacieS3BucketAssociation = MacieS3BucketAssociation;
_b = JSII_RTTI_SYMBOL_1;
MacieS3BucketAssociation[_b] = { fqn: "@cdktf/provider-aws.macie.MacieS3BucketAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MacieS3BucketAssociation.tfResourceType = "aws_macie_s3_bucket_association";
