// cdktf-cli/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Errors = void 0;
const checkpoint_1 = require("./checkpoint");
const version_check_1 = require("../bin/cmds/helper/version-check");
// Errors that will emit telemetry events
async function report(command, payload) {
    const reportParams = {
        command,
        product: "cdktf",
        version: version_check_1.versionNumber(),
        dateTime: new Date(),
        payload,
    };
    await checkpoint_1.ReportRequest(reportParams);
}
function reportPrefixedError(type) {
    return (command, message, context) => {
        report(command, { ...context, message, type });
        const err = new Error(`${type} Error: ${message}`);
        Object.entries(context || {}).forEach(([key, value]) => {
            err[key] = value;
        });
        return err;
    };
}
exports.Errors = {
    // Error within our control
    Internal: reportPrefixedError("Internal"),
    // Error in the usage
    Usage: reportPrefixedError("Usage"),
};
