// @cdktf/provider-aws/lib/wafregional/wafregional-geo-match-set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafregionalGeoMatchSet = exports.wafregionalGeoMatchSetGeoMatchConstraintToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafregionalGeoMatchSetGeoMatchConstraintToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.wafregionalGeoMatchSetGeoMatchConstraintToTerraform = wafregionalGeoMatchSetGeoMatchConstraintToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafregional_geo_match_set.html aws_wafregional_geo_match_set}.
 *
 * @stability stable
 */
class WafregionalGeoMatchSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafregional_geo_match_set.html aws_wafregional_geo_match_set} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafregional_geo_match_set',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._geoMatchConstraint = config.geoMatchConstraint;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get geoMatchConstraint() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('geo_match_constraint');
    }
    /**
     * @stability stable
     */
    set geoMatchConstraint(value) {
        this._geoMatchConstraint = value;
    }
    /**
     * @stability stable
     */
    resetGeoMatchConstraint() {
        this._geoMatchConstraint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get geoMatchConstraintInput() {
        return this._geoMatchConstraint;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            geo_match_constraint: cdktf.listMapper(wafregionalGeoMatchSetGeoMatchConstraintToTerraform)(this._geoMatchConstraint),
        };
    }
}
exports.WafregionalGeoMatchSet = WafregionalGeoMatchSet;
_a = JSII_RTTI_SYMBOL_1;
WafregionalGeoMatchSet[_a] = { fqn: "@cdktf/provider-aws.wafregional.WafregionalGeoMatchSet", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WafregionalGeoMatchSet.tfResourceType = "aws_wafregional_geo_match_set";
