// @cdktf/provider-aws/lib/wafregional/wafregional-ipset.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafregionalIpset = exports.wafregionalIpsetIpSetDescriptorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafregionalIpsetIpSetDescriptorToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.wafregionalIpsetIpSetDescriptorToTerraform = wafregionalIpsetIpSetDescriptorToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafregional_ipset.html aws_wafregional_ipset}.
 *
 * @stability stable
 */
class WafregionalIpset extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafregional_ipset.html aws_wafregional_ipset} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafregional_ipset',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._ipSetDescriptor = config.ipSetDescriptor;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get ipSetDescriptor() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ip_set_descriptor');
    }
    /**
     * @stability stable
     */
    set ipSetDescriptor(value) {
        this._ipSetDescriptor = value;
    }
    /**
     * @stability stable
     */
    resetIpSetDescriptor() {
        this._ipSetDescriptor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipSetDescriptorInput() {
        return this._ipSetDescriptor;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            ip_set_descriptor: cdktf.listMapper(wafregionalIpsetIpSetDescriptorToTerraform)(this._ipSetDescriptor),
        };
    }
}
exports.WafregionalIpset = WafregionalIpset;
_a = JSII_RTTI_SYMBOL_1;
WafregionalIpset[_a] = { fqn: "@cdktf/provider-aws.wafregional.WafregionalIpset", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WafregionalIpset.tfResourceType = "aws_wafregional_ipset";
