// @cdktf/provider-aws/lib/vpc/data-aws-internet-gateway.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsInternetGateway = exports.dataAwsInternetGatewayFilterToTerraform = exports.DataAwsInternetGatewayAttachments = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsInternetGatewayAttachments extends cdktf.ComplexComputedList {
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsInternetGatewayAttachments = DataAwsInternetGatewayAttachments;
_a = JSII_RTTI_SYMBOL_1;
DataAwsInternetGatewayAttachments[_a] = { fqn: "@cdktf/provider-aws.vpc.DataAwsInternetGatewayAttachments", version: "3.0.1" };
function dataAwsInternetGatewayFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsInternetGatewayFilterToTerraform = dataAwsInternetGatewayFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/internet_gateway.html aws_internet_gateway}.
 *
 * @stability stable
 */
class DataAwsInternetGateway extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/internet_gateway.html aws_internet_gateway} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_internet_gateway',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._internetGatewayId = config.internetGatewayId;
        this._tags = config.tags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attachments - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    attachments(index) {
        return new DataAwsInternetGatewayAttachments(this, 'attachments', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get internetGatewayId() {
        return this.getStringAttribute('internet_gateway_id');
    }
    /**
     * @stability stable
     */
    set internetGatewayId(value) {
        this._internetGatewayId = value;
    }
    /**
     * @stability stable
     */
    resetInternetGatewayId() {
        this._internetGatewayId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get internetGatewayIdInput() {
        return this._internetGatewayId;
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            internet_gateway_id: cdktf.stringToTerraform(this._internetGatewayId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsInternetGatewayFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsInternetGateway = DataAwsInternetGateway;
_b = JSII_RTTI_SYMBOL_1;
DataAwsInternetGateway[_b] = { fqn: "@cdktf/provider-aws.vpc.DataAwsInternetGateway", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsInternetGateway.tfResourceType = "aws_internet_gateway";
