// @cdktf/provider-aws/lib/vpc/data-aws-vpc-endpoint.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpcEndpoint = exports.dataAwsVpcEndpointFilterToTerraform = exports.DataAwsVpcEndpointDnsEntry = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsVpcEndpointDnsEntry extends cdktf.ComplexComputedList {
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
}
exports.DataAwsVpcEndpointDnsEntry = DataAwsVpcEndpointDnsEntry;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpcEndpointDnsEntry[_a] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcEndpointDnsEntry", version: "3.0.1" };
function dataAwsVpcEndpointFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsVpcEndpointFilterToTerraform = dataAwsVpcEndpointFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/vpc_endpoint.html aws_vpc_endpoint}.
 *
 * @stability stable
 */
class DataAwsVpcEndpoint extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/vpc_endpoint.html aws_vpc_endpoint} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._serviceName = config.serviceName;
        this._state = config.state;
        this._tags = config.tags;
        this._vpcId = config.vpcId;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // cidr_blocks - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cidrBlocks() {
        return this.getListAttribute('cidr_blocks');
    }
    // dns_entry - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    dnsEntry(index) {
        return new DataAwsVpcEndpointDnsEntry(this, 'dns_entry', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // network_interface_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkInterfaceIds() {
        return this.getListAttribute('network_interface_ids');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    // prefix_list_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get prefixListId() {
        return this.getStringAttribute('prefix_list_id');
    }
    // private_dns_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateDnsEnabled() {
        return this.getBooleanAttribute('private_dns_enabled');
    }
    // requester_managed - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get requesterManaged() {
        return this.getBooleanAttribute('requester_managed');
    }
    // route_table_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get routeTableIds() {
        return this.getListAttribute('route_table_ids');
    }
    // security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    /**
     * @stability stable
     */
    set serviceName(value) {
        this._serviceName = value;
    }
    /**
     * @stability stable
     */
    resetServiceName() {
        this._serviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceNameInput() {
        return this._serviceName;
    }
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    set state(value) {
        this._state = value;
    }
    /**
     * @stability stable
     */
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateInput() {
        return this._state;
    }
    // subnet_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // vpc_endpoint_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcEndpointType() {
        return this.getStringAttribute('vpc_endpoint_type');
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    /**
     * @stability stable
     */
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            service_name: cdktf.stringToTerraform(this._serviceName),
            state: cdktf.stringToTerraform(this._state),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            filter: cdktf.listMapper(dataAwsVpcEndpointFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsVpcEndpoint = DataAwsVpcEndpoint;
_b = JSII_RTTI_SYMBOL_1;
DataAwsVpcEndpoint[_b] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcEndpoint", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsVpcEndpoint.tfResourceType = "aws_vpc_endpoint";
