// @cdktf/provider-aws/lib/vpc/data-aws-network-interface.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsNetworkInterface = exports.dataAwsNetworkInterfaceFilterToTerraform = exports.DataAwsNetworkInterfaceAttachment = exports.DataAwsNetworkInterfaceAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsNetworkInterfaceAssociation extends cdktf.ComplexComputedList {
    // allocation_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get allocationId() {
        return this.getStringAttribute('allocation_id');
    }
    // association_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associationId() {
        return this.getStringAttribute('association_id');
    }
    // carrier_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get carrierIp() {
        return this.getStringAttribute('carrier_ip');
    }
    // customer_owned_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get customerOwnedIp() {
        return this.getStringAttribute('customer_owned_ip');
    }
    // ip_owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipOwnerId() {
        return this.getStringAttribute('ip_owner_id');
    }
    // public_dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicDnsName() {
        return this.getStringAttribute('public_dns_name');
    }
    // public_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicIp() {
        return this.getStringAttribute('public_ip');
    }
}
exports.DataAwsNetworkInterfaceAssociation = DataAwsNetworkInterfaceAssociation;
_a = JSII_RTTI_SYMBOL_1;
DataAwsNetworkInterfaceAssociation[_a] = { fqn: "@cdktf/provider-aws.vpc.DataAwsNetworkInterfaceAssociation", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsNetworkInterfaceAttachment extends cdktf.ComplexComputedList {
    // attachment_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get attachmentId() {
        return this.getStringAttribute('attachment_id');
    }
    // device_index - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceIndex() {
        return this.getNumberAttribute('device_index');
    }
    // instance_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // instance_owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceOwnerId() {
        return this.getStringAttribute('instance_owner_id');
    }
}
exports.DataAwsNetworkInterfaceAttachment = DataAwsNetworkInterfaceAttachment;
_b = JSII_RTTI_SYMBOL_1;
DataAwsNetworkInterfaceAttachment[_b] = { fqn: "@cdktf/provider-aws.vpc.DataAwsNetworkInterfaceAttachment", version: "3.0.1" };
function dataAwsNetworkInterfaceFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsNetworkInterfaceFilterToTerraform = dataAwsNetworkInterfaceFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/network_interface.html aws_network_interface}.
 *
 * @stability stable
 */
class DataAwsNetworkInterface extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/network_interface.html aws_network_interface} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_network_interface',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._tags = config.tags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // association - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    association(index) {
        return new DataAwsNetworkInterfaceAssociation(this, 'association', index);
    }
    // attachment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    attachment(index) {
        return new DataAwsNetworkInterfaceAttachment(this, 'attachment', index);
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // interface_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get interfaceType() {
        return this.getStringAttribute('interface_type');
    }
    // ipv6_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6Addresses() {
        return this.getListAttribute('ipv6_addresses');
    }
    // mac_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get macAddress() {
        return this.getStringAttribute('mac_address');
    }
    // outpost_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get outpostArn() {
        return this.getStringAttribute('outpost_arn');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // private_dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateDnsName() {
        return this.getStringAttribute('private_dns_name');
    }
    // private_ip - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIp() {
        return this.getStringAttribute('private_ip');
    }
    // private_ips - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIps() {
        return this.getListAttribute('private_ips');
    }
    // requester_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get requesterId() {
        return this.getStringAttribute('requester_id');
    }
    // security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    // subnet_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsNetworkInterfaceFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsNetworkInterface = DataAwsNetworkInterface;
_c = JSII_RTTI_SYMBOL_1;
DataAwsNetworkInterface[_c] = { fqn: "@cdktf/provider-aws.vpc.DataAwsNetworkInterface", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsNetworkInterface.tfResourceType = "aws_network_interface";
