// @cdktf/provider-aws/lib/vpc/data-aws-vpc-dhcp-options.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpcDhcpOptions = exports.dataAwsVpcDhcpOptionsFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsVpcDhcpOptionsFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsVpcDhcpOptionsFilterToTerraform = dataAwsVpcDhcpOptionsFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/vpc_dhcp_options.html aws_vpc_dhcp_options}.
 *
 * @stability stable
 */
class DataAwsVpcDhcpOptions extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/vpc_dhcp_options.html aws_vpc_dhcp_options} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_dhcp_options',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dhcpOptionsId = config.dhcpOptionsId;
        this._tags = config.tags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get dhcpOptionsId() {
        return this.getStringAttribute('dhcp_options_id');
    }
    /**
     * @stability stable
     */
    set dhcpOptionsId(value) {
        this._dhcpOptionsId = value;
    }
    /**
     * @stability stable
     */
    resetDhcpOptionsId() {
        this._dhcpOptionsId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dhcpOptionsIdInput() {
        return this._dhcpOptionsId;
    }
    // domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // domain_name_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainNameServers() {
        return this.getListAttribute('domain_name_servers');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // netbios_name_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get netbiosNameServers() {
        return this.getListAttribute('netbios_name_servers');
    }
    // netbios_node_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get netbiosNodeType() {
        return this.getStringAttribute('netbios_node_type');
    }
    // ntp_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ntpServers() {
        return this.getListAttribute('ntp_servers');
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            dhcp_options_id: cdktf.stringToTerraform(this._dhcpOptionsId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsVpcDhcpOptionsFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsVpcDhcpOptions = DataAwsVpcDhcpOptions;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpcDhcpOptions[_a] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcDhcpOptions", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsVpcDhcpOptions.tfResourceType = "aws_vpc_dhcp_options";
