// @cdktf/provider-aws/lib/vpc/default-vpc-dhcp-options.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultVpcDhcpOptions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/default_vpc_dhcp_options.html aws_default_vpc_dhcp_options}.
 *
 * @stability stable
 */
class DefaultVpcDhcpOptions extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/default_vpc_dhcp_options.html aws_default_vpc_dhcp_options} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_default_vpc_dhcp_options',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._netbiosNameServers = config.netbiosNameServers;
        this._netbiosNodeType = config.netbiosNodeType;
        this._ownerId = config.ownerId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // domain_name_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainNameServers() {
        return this.getStringAttribute('domain_name_servers');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get netbiosNameServers() {
        return this.getListAttribute('netbios_name_servers');
    }
    /**
     * @stability stable
     */
    set netbiosNameServers(value) {
        this._netbiosNameServers = value;
    }
    /**
     * @stability stable
     */
    resetNetbiosNameServers() {
        this._netbiosNameServers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get netbiosNameServersInput() {
        return this._netbiosNameServers;
    }
    /**
     * @stability stable
     */
    get netbiosNodeType() {
        return this.getStringAttribute('netbios_node_type');
    }
    /**
     * @stability stable
     */
    set netbiosNodeType(value) {
        this._netbiosNodeType = value;
    }
    /**
     * @stability stable
     */
    resetNetbiosNodeType() {
        this._netbiosNodeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get netbiosNodeTypeInput() {
        return this._netbiosNodeType;
    }
    // ntp_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ntpServers() {
        return this.getStringAttribute('ntp_servers');
    }
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    set ownerId(value) {
        this._ownerId = value;
    }
    /**
     * @stability stable
     */
    resetOwnerId() {
        this._ownerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerIdInput() {
        return this._ownerId;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            netbios_name_servers: cdktf.listMapper(cdktf.stringToTerraform)(this._netbiosNameServers),
            netbios_node_type: cdktf.stringToTerraform(this._netbiosNodeType),
            owner_id: cdktf.stringToTerraform(this._ownerId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.DefaultVpcDhcpOptions = DefaultVpcDhcpOptions;
_a = JSII_RTTI_SYMBOL_1;
DefaultVpcDhcpOptions[_a] = { fqn: "@cdktf/provider-aws.vpc.DefaultVpcDhcpOptions", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DefaultVpcDhcpOptions.tfResourceType = "aws_default_vpc_dhcp_options";
