// @cdktf/provider-aws/lib/vpc/network-interface-attachment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkInterfaceAttachmentA = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/network_interface_attachment.html aws_network_interface_attachment}.
 *
 * @stability stable
 */
class NetworkInterfaceAttachmentA extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/network_interface_attachment.html aws_network_interface_attachment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_network_interface_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._deviceIndex = config.deviceIndex;
        this._instanceId = config.instanceId;
        this._networkInterfaceId = config.networkInterfaceId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // attachment_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get attachmentId() {
        return this.getStringAttribute('attachment_id');
    }
    /**
     * @stability stable
     */
    get deviceIndex() {
        return this.getNumberAttribute('device_index');
    }
    /**
     * @stability stable
     */
    set deviceIndex(value) {
        this._deviceIndex = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deviceIndexInput() {
        return this._deviceIndex;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    /**
     * @stability stable
     */
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceIdInput() {
        return this._instanceId;
    }
    /**
     * @stability stable
     */
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    /**
     * @stability stable
     */
    set networkInterfaceId(value) {
        this._networkInterfaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkInterfaceIdInput() {
        return this._networkInterfaceId;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            device_index: cdktf.numberToTerraform(this._deviceIndex),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            network_interface_id: cdktf.stringToTerraform(this._networkInterfaceId),
        };
    }
}
exports.NetworkInterfaceAttachmentA = NetworkInterfaceAttachmentA;
_a = JSII_RTTI_SYMBOL_1;
NetworkInterfaceAttachmentA[_a] = { fqn: "@cdktf/provider-aws.vpc.NetworkInterfaceAttachmentA", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
NetworkInterfaceAttachmentA.tfResourceType = "aws_network_interface_attachment";
