// @cdktf/provider-aws/lib/vpc/security-group-rule.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityGroupRule = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/security_group_rule.html aws_security_group_rule}.
 *
 * @stability stable
 */
class SecurityGroupRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/security_group_rule.html aws_security_group_rule} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_security_group_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._cidrBlocks = config.cidrBlocks;
        this._description = config.description;
        this._fromPort = config.fromPort;
        this._ipv6CidrBlocks = config.ipv6CidrBlocks;
        this._prefixListIds = config.prefixListIds;
        this._protocol = config.protocol;
        this._securityGroupId = config.securityGroupId;
        this._self = config.selfAttribute;
        this._sourceSecurityGroupId = config.sourceSecurityGroupId;
        this._toPort = config.toPort;
        this._type = config.type;
    }
    /**
     * @stability stable
     */
    get cidrBlocks() {
        return this.getListAttribute('cidr_blocks');
    }
    /**
     * @stability stable
     */
    set cidrBlocks(value) {
        this._cidrBlocks = value;
    }
    /**
     * @stability stable
     */
    resetCidrBlocks() {
        this._cidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cidrBlocksInput() {
        return this._cidrBlocks;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    /**
     * @stability stable
     */
    set fromPort(value) {
        this._fromPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fromPortInput() {
        return this._fromPort;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ipv6CidrBlocks() {
        return this.getListAttribute('ipv6_cidr_blocks');
    }
    /**
     * @stability stable
     */
    set ipv6CidrBlocks(value) {
        this._ipv6CidrBlocks = value;
    }
    /**
     * @stability stable
     */
    resetIpv6CidrBlocks() {
        this._ipv6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipv6CidrBlocksInput() {
        return this._ipv6CidrBlocks;
    }
    /**
     * @stability stable
     */
    get prefixListIds() {
        return this.getListAttribute('prefix_list_ids');
    }
    /**
     * @stability stable
     */
    set prefixListIds(value) {
        this._prefixListIds = value;
    }
    /**
     * @stability stable
     */
    resetPrefixListIds() {
        this._prefixListIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixListIdsInput() {
        return this._prefixListIds;
    }
    /**
     * @stability stable
     */
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    /**
     * @stability stable
     */
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get protocolInput() {
        return this._protocol;
    }
    /**
     * @stability stable
     */
    get securityGroupId() {
        return this.getStringAttribute('security_group_id');
    }
    /**
     * @stability stable
     */
    set securityGroupId(value) {
        this._securityGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdInput() {
        return this._securityGroupId;
    }
    /**
     * @stability stable
     */
    get selfAttribute() {
        return this.getBooleanAttribute('self');
    }
    /**
     * @stability stable
     */
    set selfAttribute(value) {
        this._self = value;
    }
    /**
     * @stability stable
     */
    resetSelfAttribute() {
        this._self = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get selfAttributeInput() {
        return this._self;
    }
    /**
     * @stability stable
     */
    get sourceSecurityGroupId() {
        return this.getStringAttribute('source_security_group_id');
    }
    /**
     * @stability stable
     */
    set sourceSecurityGroupId(value) {
        this._sourceSecurityGroupId = value;
    }
    /**
     * @stability stable
     */
    resetSourceSecurityGroupId() {
        this._sourceSecurityGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sourceSecurityGroupIdInput() {
        return this._sourceSecurityGroupId;
    }
    /**
     * @stability stable
     */
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    /**
     * @stability stable
     */
    set toPort(value) {
        this._toPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get toPortInput() {
        return this._toPort;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform)(this._cidrBlocks),
            description: cdktf.stringToTerraform(this._description),
            from_port: cdktf.numberToTerraform(this._fromPort),
            ipv6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform)(this._ipv6CidrBlocks),
            prefix_list_ids: cdktf.listMapper(cdktf.stringToTerraform)(this._prefixListIds),
            protocol: cdktf.stringToTerraform(this._protocol),
            security_group_id: cdktf.stringToTerraform(this._securityGroupId),
            self: cdktf.booleanToTerraform(this._self),
            source_security_group_id: cdktf.stringToTerraform(this._sourceSecurityGroupId),
            to_port: cdktf.numberToTerraform(this._toPort),
            type: cdktf.stringToTerraform(this._type),
        };
    }
}
exports.SecurityGroupRule = SecurityGroupRule;
_a = JSII_RTTI_SYMBOL_1;
SecurityGroupRule[_a] = { fqn: "@cdktf/provider-aws.vpc.SecurityGroupRule", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecurityGroupRule.tfResourceType = "aws_security_group_rule";
