// @cdktf/provider-aws/lib/vpc/vpc-endpoint-service.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcEndpointService = exports.VpcEndpointServicePrivateDnsNameConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class VpcEndpointServicePrivateDnsNameConfiguration extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    // value - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.VpcEndpointServicePrivateDnsNameConfiguration = VpcEndpointServicePrivateDnsNameConfiguration;
_a = JSII_RTTI_SYMBOL_1;
VpcEndpointServicePrivateDnsNameConfiguration[_a] = { fqn: "@cdktf/provider-aws.vpc.VpcEndpointServicePrivateDnsNameConfiguration", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service.html aws_vpc_endpoint_service}.
 *
 * @stability stable
 */
class VpcEndpointService extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service.html aws_vpc_endpoint_service} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint_service',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._acceptanceRequired = config.acceptanceRequired;
        this._allowedPrincipals = config.allowedPrincipals;
        this._gatewayLoadBalancerArns = config.gatewayLoadBalancerArns;
        this._networkLoadBalancerArns = config.networkLoadBalancerArns;
        this._privateDnsName = config.privateDnsName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    /**
     * @stability stable
     */
    get acceptanceRequired() {
        return this.getBooleanAttribute('acceptance_required');
    }
    /**
     * @stability stable
     */
    set acceptanceRequired(value) {
        this._acceptanceRequired = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get acceptanceRequiredInput() {
        return this._acceptanceRequired;
    }
    /**
     * @stability stable
     */
    get allowedPrincipals() {
        return this.getListAttribute('allowed_principals');
    }
    /**
     * @stability stable
     */
    set allowedPrincipals(value) {
        this._allowedPrincipals = value;
    }
    /**
     * @stability stable
     */
    resetAllowedPrincipals() {
        this._allowedPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowedPrincipalsInput() {
        return this._allowedPrincipals;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zones - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZones() {
        return this.getListAttribute('availability_zones');
    }
    // base_endpoint_dns_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get baseEndpointDnsNames() {
        return this.getListAttribute('base_endpoint_dns_names');
    }
    /**
     * @stability stable
     */
    get gatewayLoadBalancerArns() {
        return this.getListAttribute('gateway_load_balancer_arns');
    }
    /**
     * @stability stable
     */
    set gatewayLoadBalancerArns(value) {
        this._gatewayLoadBalancerArns = value;
    }
    /**
     * @stability stable
     */
    resetGatewayLoadBalancerArns() {
        this._gatewayLoadBalancerArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gatewayLoadBalancerArnsInput() {
        return this._gatewayLoadBalancerArns;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // manages_vpc_endpoints - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get managesVpcEndpoints() {
        return this.getBooleanAttribute('manages_vpc_endpoints');
    }
    /**
     * @stability stable
     */
    get networkLoadBalancerArns() {
        return this.getListAttribute('network_load_balancer_arns');
    }
    /**
     * @stability stable
     */
    set networkLoadBalancerArns(value) {
        this._networkLoadBalancerArns = value;
    }
    /**
     * @stability stable
     */
    resetNetworkLoadBalancerArns() {
        this._networkLoadBalancerArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkLoadBalancerArnsInput() {
        return this._networkLoadBalancerArns;
    }
    /**
     * @stability stable
     */
    get privateDnsName() {
        return this.getStringAttribute('private_dns_name');
    }
    /**
     * @stability stable
     */
    set privateDnsName(value) {
        this._privateDnsName = value;
    }
    /**
     * @stability stable
     */
    resetPrivateDnsName() {
        this._privateDnsName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateDnsNameInput() {
        return this._privateDnsName;
    }
    // private_dns_name_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    privateDnsNameConfiguration(index) {
        return new VpcEndpointServicePrivateDnsNameConfiguration(this, 'private_dns_name_configuration', index);
    }
    // service_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    // service_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get serviceType() {
        return this.getStringAttribute('service_type');
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            acceptance_required: cdktf.booleanToTerraform(this._acceptanceRequired),
            allowed_principals: cdktf.listMapper(cdktf.stringToTerraform)(this._allowedPrincipals),
            gateway_load_balancer_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._gatewayLoadBalancerArns),
            network_load_balancer_arns: cdktf.listMapper(cdktf.stringToTerraform)(this._networkLoadBalancerArns),
            private_dns_name: cdktf.stringToTerraform(this._privateDnsName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.VpcEndpointService = VpcEndpointService;
_b = JSII_RTTI_SYMBOL_1;
VpcEndpointService[_b] = { fqn: "@cdktf/provider-aws.vpc.VpcEndpointService", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
VpcEndpointService.tfResourceType = "aws_vpc_endpoint_service";
