// @cdktf/provider-aws/lib/vpc/vpc-dhcp-options-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcDhcpOptionsAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_dhcp_options_association.html aws_vpc_dhcp_options_association}.
 *
 * @stability stable
 */
class VpcDhcpOptionsAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_dhcp_options_association.html aws_vpc_dhcp_options_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_dhcp_options_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._dhcpOptionsId = config.dhcpOptionsId;
        this._vpcId = config.vpcId;
    }
    /**
     * @stability stable
     */
    get dhcpOptionsId() {
        return this.getStringAttribute('dhcp_options_id');
    }
    /**
     * @stability stable
     */
    set dhcpOptionsId(value) {
        this._dhcpOptionsId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dhcpOptionsIdInput() {
        return this._dhcpOptionsId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            dhcp_options_id: cdktf.stringToTerraform(this._dhcpOptionsId),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
        };
    }
}
exports.VpcDhcpOptionsAssociation = VpcDhcpOptionsAssociation;
_a = JSII_RTTI_SYMBOL_1;
VpcDhcpOptionsAssociation[_a] = { fqn: "@cdktf/provider-aws.vpc.VpcDhcpOptionsAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
VpcDhcpOptionsAssociation.tfResourceType = "aws_vpc_dhcp_options_association";
