// @cdktf/provider-aws/lib/iot/iot-role-alias.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotRoleAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_role_alias.html aws_iot_role_alias}.
 *
 * @stability stable
 */
class IotRoleAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_role_alias.html aws_iot_role_alias} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_role_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._alias = config.alias;
        this._credentialDuration = config.credentialDuration;
        this._roleArn = config.roleArn;
    }
    /**
     * @stability stable
     */
    get alias() {
        return this.getStringAttribute('alias');
    }
    /**
     * @stability stable
     */
    set alias(value) {
        this._alias = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get aliasInput() {
        return this._alias;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get credentialDuration() {
        return this.getNumberAttribute('credential_duration');
    }
    /**
     * @stability stable
     */
    set credentialDuration(value) {
        this._credentialDuration = value;
    }
    /**
     * @stability stable
     */
    resetCredentialDuration() {
        this._credentialDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialDurationInput() {
        return this._credentialDuration;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            credential_duration: cdktf.numberToTerraform(this._credentialDuration),
            role_arn: cdktf.stringToTerraform(this._roleArn),
        };
    }
}
exports.IotRoleAlias = IotRoleAlias;
_a = JSII_RTTI_SYMBOL_1;
IotRoleAlias[_a] = { fqn: "@cdktf/provider-aws.iot.IotRoleAlias", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotRoleAlias.tfResourceType = "aws_iot_role_alias";
