// @cdktf/provider-aws/lib/elasticsearch/elasticsearch-domain-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticsearchDomainPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain_policy.html aws_elasticsearch_domain_policy}.
 *
 * @stability stable
 */
class ElasticsearchDomainPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain_policy.html aws_elasticsearch_domain_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticsearch_domain_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._accessPolicies = config.accessPolicies;
        this._domainName = config.domainName;
    }
    /**
     * @stability stable
     */
    get accessPolicies() {
        return this.getStringAttribute('access_policies');
    }
    /**
     * @stability stable
     */
    set accessPolicies(value) {
        this._accessPolicies = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accessPoliciesInput() {
        return this._accessPolicies;
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            access_policies: cdktf.stringToTerraform(this._accessPolicies),
            domain_name: cdktf.stringToTerraform(this._domainName),
        };
    }
}
exports.ElasticsearchDomainPolicy = ElasticsearchDomainPolicy;
_a = JSII_RTTI_SYMBOL_1;
ElasticsearchDomainPolicy[_a] = { fqn: "@cdktf/provider-aws.elasticsearch.ElasticsearchDomainPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ElasticsearchDomainPolicy.tfResourceType = "aws_elasticsearch_domain_policy";
