// @cdktf/provider-aws/lib/signer/signer-signing-profile.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignerSigningProfile = exports.SignerSigningProfileSignatureValidityPeriodOutputReference = exports.signerSigningProfileSignatureValidityPeriodToTerraform = exports.SignerSigningProfileRevocationRecord = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class SignerSigningProfileRevocationRecord extends cdktf.ComplexComputedList {
    // revocation_effective_from - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get revocationEffectiveFrom() {
        return this.getStringAttribute('revocation_effective_from');
    }
    // revoked_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get revokedAt() {
        return this.getStringAttribute('revoked_at');
    }
    // revoked_by - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get revokedBy() {
        return this.getStringAttribute('revoked_by');
    }
}
exports.SignerSigningProfileRevocationRecord = SignerSigningProfileRevocationRecord;
_a = JSII_RTTI_SYMBOL_1;
SignerSigningProfileRevocationRecord[_a] = { fqn: "@cdktf/provider-aws.signer.SignerSigningProfileRevocationRecord", version: "3.0.1" };
function signerSigningProfileSignatureValidityPeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.signerSigningProfileSignatureValidityPeriodToTerraform = signerSigningProfileSignatureValidityPeriodToTerraform;
/**
 * @stability stable
 */
class SignerSigningProfileSignatureValidityPeriodOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._value = value.value;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get value() {
        return this.getNumberAttribute('value');
    }
    /**
     * @stability stable
     */
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get valueInput() {
        return this._value;
    }
}
exports.SignerSigningProfileSignatureValidityPeriodOutputReference = SignerSigningProfileSignatureValidityPeriodOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SignerSigningProfileSignatureValidityPeriodOutputReference[_b] = { fqn: "@cdktf/provider-aws.signer.SignerSigningProfileSignatureValidityPeriodOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/signer_signing_profile.html aws_signer_signing_profile}.
 *
 * @stability stable
 */
class SignerSigningProfile extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/signer_signing_profile.html aws_signer_signing_profile} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_signer_signing_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // signature_validity_period - computed: false, optional: true, required: false
        this._signatureValidityPeriod = new SignerSigningProfileSignatureValidityPeriodOutputReference(this, "signature_validity_period", true);
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._platformId = config.platformId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._signatureValidityPeriod.internalValue = config.signatureValidityPeriod;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    /**
     * @stability stable
     */
    set namePrefix(value) {
        this._namePrefix = value;
    }
    /**
     * @stability stable
     */
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get namePrefixInput() {
        return this._namePrefix;
    }
    // platform_display_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get platformDisplayName() {
        return this.getStringAttribute('platform_display_name');
    }
    /**
     * @stability stable
     */
    get platformId() {
        return this.getStringAttribute('platform_id');
    }
    /**
     * @stability stable
     */
    set platformId(value) {
        this._platformId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get platformIdInput() {
        return this._platformId;
    }
    // revocation_record - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    revocationRecord(index) {
        return new SignerSigningProfileRevocationRecord(this, 'revocation_record', index);
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    // version_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get versionArn() {
        return this.getStringAttribute('version_arn');
    }
    /**
     * @stability stable
     */
    get signatureValidityPeriod() {
        return this._signatureValidityPeriod;
    }
    /**
     * @stability stable
     */
    putSignatureValidityPeriod(value) {
        this._signatureValidityPeriod.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSignatureValidityPeriod() {
        this._signatureValidityPeriod.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signatureValidityPeriodInput() {
        return this._signatureValidityPeriod.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            platform_id: cdktf.stringToTerraform(this._platformId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            signature_validity_period: signerSigningProfileSignatureValidityPeriodToTerraform(this._signatureValidityPeriod.internalValue),
        };
    }
}
exports.SignerSigningProfile = SignerSigningProfile;
_c = JSII_RTTI_SYMBOL_1;
SignerSigningProfile[_c] = { fqn: "@cdktf/provider-aws.signer.SignerSigningProfile", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SignerSigningProfile.tfResourceType = "aws_signer_signing_profile";
