// @cdktf/provider-aws/lib/elasticsearch/data-aws-elasticsearch-domain.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElasticsearchDomain = exports.DataAwsElasticsearchDomainVpcOptions = exports.DataAwsElasticsearchDomainSnapshotOptions = exports.DataAwsElasticsearchDomainNodeToNodeEncryption = exports.DataAwsElasticsearchDomainLogPublishingOptions = exports.DataAwsElasticsearchDomainEncryptionAtRest = exports.DataAwsElasticsearchDomainEbsOptions = exports.DataAwsElasticsearchDomainCognitoOptions = exports.DataAwsElasticsearchDomainClusterConfig = exports.DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig = exports.DataAwsElasticsearchDomainAdvancedSecurityOptions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainAdvancedSecurityOptions extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // internal_user_database_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get internalUserDatabaseEnabled() {
        return this.getBooleanAttribute('internal_user_database_enabled');
    }
}
exports.DataAwsElasticsearchDomainAdvancedSecurityOptions = DataAwsElasticsearchDomainAdvancedSecurityOptions;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainAdvancedSecurityOptions[_a] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainAdvancedSecurityOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig extends cdktf.ComplexComputedList {
    // availability_zone_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZoneCount() {
        return this.getNumberAttribute('availability_zone_count');
    }
}
exports.DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig = DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig;
_b = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig[_b] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainClusterConfigZoneAwarenessConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainClusterConfig extends cdktf.ComplexComputedList {
    // dedicated_master_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dedicatedMasterCount() {
        return this.getNumberAttribute('dedicated_master_count');
    }
    // dedicated_master_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dedicatedMasterEnabled() {
        return this.getBooleanAttribute('dedicated_master_enabled');
    }
    // dedicated_master_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dedicatedMasterType() {
        return this.getStringAttribute('dedicated_master_type');
    }
    // instance_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    // instance_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // warm_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get warmCount() {
        return this.getNumberAttribute('warm_count');
    }
    // warm_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get warmEnabled() {
        return this.getBooleanAttribute('warm_enabled');
    }
    // warm_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get warmType() {
        return this.getStringAttribute('warm_type');
    }
    // zone_awareness_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneAwarenessConfig() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('zone_awareness_config');
    }
    // zone_awareness_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneAwarenessEnabled() {
        return this.getBooleanAttribute('zone_awareness_enabled');
    }
}
exports.DataAwsElasticsearchDomainClusterConfig = DataAwsElasticsearchDomainClusterConfig;
_c = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainClusterConfig[_c] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainClusterConfig", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainCognitoOptions extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // identity_pool_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get identityPoolId() {
        return this.getStringAttribute('identity_pool_id');
    }
    // role_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // user_pool_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
}
exports.DataAwsElasticsearchDomainCognitoOptions = DataAwsElasticsearchDomainCognitoOptions;
_d = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainCognitoOptions[_d] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainCognitoOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainEbsOptions extends cdktf.ComplexComputedList {
    // ebs_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ebsEnabled() {
        return this.getBooleanAttribute('ebs_enabled');
    }
    // iops - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // volume_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsElasticsearchDomainEbsOptions = DataAwsElasticsearchDomainEbsOptions;
_e = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainEbsOptions[_e] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainEbsOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainEncryptionAtRest extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
}
exports.DataAwsElasticsearchDomainEncryptionAtRest = DataAwsElasticsearchDomainEncryptionAtRest;
_f = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainEncryptionAtRest[_f] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainEncryptionAtRest", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainLogPublishingOptions extends cdktf.ComplexComputedList {
    // cloudwatch_log_group_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudwatchLogGroupArn() {
        return this.getStringAttribute('cloudwatch_log_group_arn');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // log_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get logType() {
        return this.getStringAttribute('log_type');
    }
}
exports.DataAwsElasticsearchDomainLogPublishingOptions = DataAwsElasticsearchDomainLogPublishingOptions;
_g = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainLogPublishingOptions[_g] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainLogPublishingOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainNodeToNodeEncryption extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsElasticsearchDomainNodeToNodeEncryption = DataAwsElasticsearchDomainNodeToNodeEncryption;
_h = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainNodeToNodeEncryption[_h] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainNodeToNodeEncryption", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainSnapshotOptions extends cdktf.ComplexComputedList {
    // automated_snapshot_start_hour - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get automatedSnapshotStartHour() {
        return this.getNumberAttribute('automated_snapshot_start_hour');
    }
}
exports.DataAwsElasticsearchDomainSnapshotOptions = DataAwsElasticsearchDomainSnapshotOptions;
_j = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainSnapshotOptions[_j] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainSnapshotOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElasticsearchDomainVpcOptions extends cdktf.ComplexComputedList {
    // availability_zones - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZones() {
        return this.getListAttribute('availability_zones');
    }
    // security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    // subnet_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    // vpc_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsElasticsearchDomainVpcOptions = DataAwsElasticsearchDomainVpcOptions;
_k = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomainVpcOptions[_k] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomainVpcOptions", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/elasticsearch_domain.html aws_elasticsearch_domain}.
 *
 * @stability stable
 */
class DataAwsElasticsearchDomain extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/elasticsearch_domain.html aws_elasticsearch_domain} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticsearch_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._domainName = config.domainName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_policies - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get accessPolicies() {
        return this.getStringAttribute('access_policies');
    }
    // advanced_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    advancedOptions(key) {
        return new cdktf.StringMap(this, 'advanced_options').lookup(key);
    }
    // advanced_security_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    advancedSecurityOptions(index) {
        return new DataAwsElasticsearchDomainAdvancedSecurityOptions(this, 'advanced_security_options', index);
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // cluster_config - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    clusterConfig(index) {
        return new DataAwsElasticsearchDomainClusterConfig(this, 'cluster_config', index);
    }
    // cognito_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    cognitoOptions(index) {
        return new DataAwsElasticsearchDomainCognitoOptions(this, 'cognito_options', index);
    }
    // created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get created() {
        return this.getBooleanAttribute('created');
    }
    // deleted - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleted() {
        return this.getBooleanAttribute('deleted');
    }
    // domain_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // ebs_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ebsOptions(index) {
        return new DataAwsElasticsearchDomainEbsOptions(this, 'ebs_options', index);
    }
    // elasticsearch_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get elasticsearchVersion() {
        return this.getStringAttribute('elasticsearch_version');
    }
    // encryption_at_rest - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    encryptionAtRest(index) {
        return new DataAwsElasticsearchDomainEncryptionAtRest(this, 'encryption_at_rest', index);
    }
    // endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // kibana_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kibanaEndpoint() {
        return this.getStringAttribute('kibana_endpoint');
    }
    // log_publishing_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    logPublishingOptions(index) {
        return new DataAwsElasticsearchDomainLogPublishingOptions(this, 'log_publishing_options', index);
    }
    // node_to_node_encryption - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    nodeToNodeEncryption(index) {
        return new DataAwsElasticsearchDomainNodeToNodeEncryption(this, 'node_to_node_encryption', index);
    }
    // processing - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get processing() {
        return this.getBooleanAttribute('processing');
    }
    // snapshot_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    snapshotOptions(index) {
        return new DataAwsElasticsearchDomainSnapshotOptions(this, 'snapshot_options', index);
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // vpc_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    vpcOptions(index) {
        return new DataAwsElasticsearchDomainVpcOptions(this, 'vpc_options', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsElasticsearchDomain = DataAwsElasticsearchDomain;
_l = JSII_RTTI_SYMBOL_1;
DataAwsElasticsearchDomain[_l] = { fqn: "@cdktf/provider-aws.elasticsearch.DataAwsElasticsearchDomain", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsElasticsearchDomain.tfResourceType = "aws_elasticsearch_domain";
