// @cdktf/provider-aws/lib/dynamodb/data-aws-dynamodb-table.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDynamodbTable = exports.DataAwsDynamodbTableServerSideEncryptionOutputReference = exports.dataAwsDynamodbTableServerSideEncryptionToTerraform = exports.DataAwsDynamodbTableTtl = exports.DataAwsDynamodbTableReplica = exports.DataAwsDynamodbTablePointInTimeRecovery = exports.DataAwsDynamodbTableLocalSecondaryIndex = exports.DataAwsDynamodbTableGlobalSecondaryIndex = exports.DataAwsDynamodbTableAttribute = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsDynamodbTableAttribute extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsDynamodbTableAttribute = DataAwsDynamodbTableAttribute;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableAttribute[_a] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableAttribute", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsDynamodbTableGlobalSecondaryIndex extends cdktf.ComplexComputedList {
    // hash_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // non_key_attributes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nonKeyAttributes() {
        return this.getListAttribute('non_key_attributes');
    }
    // projection_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get projectionType() {
        return this.getStringAttribute('projection_type');
    }
    // range_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    // read_capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get readCapacity() {
        return this.getNumberAttribute('read_capacity');
    }
    // write_capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get writeCapacity() {
        return this.getNumberAttribute('write_capacity');
    }
}
exports.DataAwsDynamodbTableGlobalSecondaryIndex = DataAwsDynamodbTableGlobalSecondaryIndex;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableGlobalSecondaryIndex[_b] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableGlobalSecondaryIndex", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsDynamodbTableLocalSecondaryIndex extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // non_key_attributes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nonKeyAttributes() {
        return this.getListAttribute('non_key_attributes');
    }
    // projection_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get projectionType() {
        return this.getStringAttribute('projection_type');
    }
    // range_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
}
exports.DataAwsDynamodbTableLocalSecondaryIndex = DataAwsDynamodbTableLocalSecondaryIndex;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableLocalSecondaryIndex[_c] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableLocalSecondaryIndex", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsDynamodbTablePointInTimeRecovery extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsDynamodbTablePointInTimeRecovery = DataAwsDynamodbTablePointInTimeRecovery;
_d = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTablePointInTimeRecovery[_d] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTablePointInTimeRecovery", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsDynamodbTableReplica extends cdktf.ComplexComputedList {
    // kms_key_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    // region_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionName() {
        return this.getStringAttribute('region_name');
    }
}
exports.DataAwsDynamodbTableReplica = DataAwsDynamodbTableReplica;
_e = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableReplica[_e] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableReplica", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsDynamodbTableTtl extends cdktf.ComplexComputedList {
    // attribute_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get attributeName() {
        return this.getStringAttribute('attribute_name');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsDynamodbTableTtl = DataAwsDynamodbTableTtl;
_f = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableTtl[_f] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableTtl", version: "3.0.1" };
function dataAwsDynamodbTableServerSideEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableServerSideEncryptionToTerraform = dataAwsDynamodbTableServerSideEncryptionToTerraform;
/**
 * @stability stable
 */
class DataAwsDynamodbTableServerSideEncryptionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataAwsDynamodbTableServerSideEncryptionOutputReference = DataAwsDynamodbTableServerSideEncryptionOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableServerSideEncryptionOutputReference[_g] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTableServerSideEncryptionOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table.html aws_dynamodb_table}.
 *
 * @stability stable
 */
class DataAwsDynamodbTable extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table.html aws_dynamodb_table} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // server_side_encryption - computed: false, optional: true, required: false
        this._serverSideEncryption = new DataAwsDynamodbTableServerSideEncryptionOutputReference(this, "server_side_encryption", true);
        this._name = config.name;
        this._tags = config.tags;
        this._serverSideEncryption.internalValue = config.serverSideEncryption;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // attribute - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    attribute(index) {
        return new DataAwsDynamodbTableAttribute(this, 'attribute', index);
    }
    // billing_mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get billingMode() {
        return this.getStringAttribute('billing_mode');
    }
    // global_secondary_index - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    globalSecondaryIndex(index) {
        return new DataAwsDynamodbTableGlobalSecondaryIndex(this, 'global_secondary_index', index);
    }
    // hash_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // local_secondary_index - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    localSecondaryIndex(index) {
        return new DataAwsDynamodbTableLocalSecondaryIndex(this, 'local_secondary_index', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // point_in_time_recovery - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    pointInTimeRecovery(index) {
        return new DataAwsDynamodbTablePointInTimeRecovery(this, 'point_in_time_recovery', index);
    }
    // range_key - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    // read_capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get readCapacity() {
        return this.getNumberAttribute('read_capacity');
    }
    // replica - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    replica(index) {
        return new DataAwsDynamodbTableReplica(this, 'replica', index);
    }
    // stream_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    // stream_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get streamEnabled() {
        return this.getBooleanAttribute('stream_enabled');
    }
    // stream_label - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get streamLabel() {
        return this.getStringAttribute('stream_label');
    }
    // stream_view_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get streamViewType() {
        return this.getStringAttribute('stream_view_type');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // ttl - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    ttl(index) {
        return new DataAwsDynamodbTableTtl(this, 'ttl', index);
    }
    // write_capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get writeCapacity() {
        return this.getNumberAttribute('write_capacity');
    }
    /**
     * @stability stable
     */
    get serverSideEncryption() {
        return this._serverSideEncryption;
    }
    /**
     * @stability stable
     */
    putServerSideEncryption(value) {
        this._serverSideEncryption.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetServerSideEncryption() {
        this._serverSideEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serverSideEncryptionInput() {
        return this._serverSideEncryption.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            server_side_encryption: dataAwsDynamodbTableServerSideEncryptionToTerraform(this._serverSideEncryption.internalValue),
        };
    }
}
exports.DataAwsDynamodbTable = DataAwsDynamodbTable;
_h = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTable[_h] = { fqn: "@cdktf/provider-aws.dynamodb.DataAwsDynamodbTable", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsDynamodbTable.tfResourceType = "aws_dynamodb_table";
