// @cdktf/provider-aws/lib/waf/waf-web-acl.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafWebAcl = exports.wafWebAclRulesToTerraform = exports.WafWebAclRulesOverrideActionOutputReference = exports.wafWebAclRulesOverrideActionToTerraform = exports.WafWebAclRulesActionOutputReference = exports.wafWebAclRulesActionToTerraform = exports.WafWebAclLoggingConfigurationOutputReference = exports.wafWebAclLoggingConfigurationToTerraform = exports.WafWebAclLoggingConfigurationRedactedFieldsOutputReference = exports.wafWebAclLoggingConfigurationRedactedFieldsToTerraform = exports.wafWebAclLoggingConfigurationRedactedFieldsFieldToMatchToTerraform = exports.WafWebAclDefaultActionOutputReference = exports.wafWebAclDefaultActionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafWebAclDefaultActionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafWebAclDefaultActionToTerraform = wafWebAclDefaultActionToTerraform;
/**
 * @stability stable
 */
class WafWebAclDefaultActionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.WafWebAclDefaultActionOutputReference = WafWebAclDefaultActionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafWebAclDefaultActionOutputReference[_a] = { fqn: "@cdktf/provider-aws.waf.WafWebAclDefaultActionOutputReference", version: "3.0.1" };
function wafWebAclLoggingConfigurationRedactedFieldsFieldToMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data: cdktf.stringToTerraform(struct.data),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafWebAclLoggingConfigurationRedactedFieldsFieldToMatchToTerraform = wafWebAclLoggingConfigurationRedactedFieldsFieldToMatchToTerraform;
function wafWebAclLoggingConfigurationRedactedFieldsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field_to_match: cdktf.listMapper(wafWebAclLoggingConfigurationRedactedFieldsFieldToMatchToTerraform)(struct.fieldToMatch),
    };
}
exports.wafWebAclLoggingConfigurationRedactedFieldsToTerraform = wafWebAclLoggingConfigurationRedactedFieldsToTerraform;
/**
 * @stability stable
 */
class WafWebAclLoggingConfigurationRedactedFieldsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fieldToMatch) {
            hasAnyValues = true;
            internalValueResult.fieldToMatch = this._fieldToMatch;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fieldToMatch = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fieldToMatch = value.fieldToMatch;
        }
    }
    /**
     * @stability stable
     */
    get fieldToMatch() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('field_to_match');
    }
    /**
     * @stability stable
     */
    set fieldToMatch(value) {
        this._fieldToMatch = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fieldToMatchInput() {
        return this._fieldToMatch;
    }
}
exports.WafWebAclLoggingConfigurationRedactedFieldsOutputReference = WafWebAclLoggingConfigurationRedactedFieldsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
WafWebAclLoggingConfigurationRedactedFieldsOutputReference[_b] = { fqn: "@cdktf/provider-aws.waf.WafWebAclLoggingConfigurationRedactedFieldsOutputReference", version: "3.0.1" };
function wafWebAclLoggingConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_destination: cdktf.stringToTerraform(struct.logDestination),
        redacted_fields: wafWebAclLoggingConfigurationRedactedFieldsToTerraform(struct.redactedFields),
    };
}
exports.wafWebAclLoggingConfigurationToTerraform = wafWebAclLoggingConfigurationToTerraform;
/**
 * @stability stable
 */
class WafWebAclLoggingConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // redacted_fields - computed: false, optional: true, required: false
        this._redactedFields = new WafWebAclLoggingConfigurationRedactedFieldsOutputReference(this, "redacted_fields", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _g, _h;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logDestination) {
            hasAnyValues = true;
            internalValueResult.logDestination = this._logDestination;
        }
        if ((_g = this._redactedFields) === null || _g === void 0 ? void 0 : _g.internalValue) {
            hasAnyValues = true;
            internalValueResult.redactedFields = (_h = this._redactedFields) === null || _h === void 0 ? void 0 : _h.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logDestination = undefined;
            this._redactedFields.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logDestination = value.logDestination;
            this._redactedFields.internalValue = value.redactedFields;
        }
    }
    /**
     * @stability stable
     */
    get logDestination() {
        return this.getStringAttribute('log_destination');
    }
    /**
     * @stability stable
     */
    set logDestination(value) {
        this._logDestination = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logDestinationInput() {
        return this._logDestination;
    }
    /**
     * @stability stable
     */
    get redactedFields() {
        return this._redactedFields;
    }
    /**
     * @stability stable
     */
    putRedactedFields(value) {
        this._redactedFields.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRedactedFields() {
        this._redactedFields.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get redactedFieldsInput() {
        return this._redactedFields.internalValue;
    }
}
exports.WafWebAclLoggingConfigurationOutputReference = WafWebAclLoggingConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
WafWebAclLoggingConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.waf.WafWebAclLoggingConfigurationOutputReference", version: "3.0.1" };
function wafWebAclRulesActionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafWebAclRulesActionToTerraform = wafWebAclRulesActionToTerraform;
/**
 * @stability stable
 */
class WafWebAclRulesActionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.WafWebAclRulesActionOutputReference = WafWebAclRulesActionOutputReference;
_d = JSII_RTTI_SYMBOL_1;
WafWebAclRulesActionOutputReference[_d] = { fqn: "@cdktf/provider-aws.waf.WafWebAclRulesActionOutputReference", version: "3.0.1" };
function wafWebAclRulesOverrideActionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafWebAclRulesOverrideActionToTerraform = wafWebAclRulesOverrideActionToTerraform;
/**
 * @stability stable
 */
class WafWebAclRulesOverrideActionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.WafWebAclRulesOverrideActionOutputReference = WafWebAclRulesOverrideActionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
WafWebAclRulesOverrideActionOutputReference[_e] = { fqn: "@cdktf/provider-aws.waf.WafWebAclRulesOverrideActionOutputReference", version: "3.0.1" };
function wafWebAclRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        priority: cdktf.numberToTerraform(struct.priority),
        rule_id: cdktf.stringToTerraform(struct.ruleId),
        type: cdktf.stringToTerraform(struct.type),
        action: wafWebAclRulesActionToTerraform(struct.action),
        override_action: wafWebAclRulesOverrideActionToTerraform(struct.overrideAction),
    };
}
exports.wafWebAclRulesToTerraform = wafWebAclRulesToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/waf_web_acl.html aws_waf_web_acl}.
 *
 * @stability stable
 */
class WafWebAcl extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/waf_web_acl.html aws_waf_web_acl} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_waf_web_acl',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // default_action - computed: false, optional: false, required: true
        this._defaultAction = new WafWebAclDefaultActionOutputReference(this, "default_action", true);
        // logging_configuration - computed: false, optional: true, required: false
        this._loggingConfiguration = new WafWebAclLoggingConfigurationOutputReference(this, "logging_configuration", true);
        this._metricName = config.metricName;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._defaultAction.internalValue = config.defaultAction;
        this._loggingConfiguration.internalValue = config.loggingConfiguration;
        this._rules = config.rules;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    /**
     * @stability stable
     */
    set metricName(value) {
        this._metricName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricNameInput() {
        return this._metricName;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get defaultAction() {
        return this._defaultAction;
    }
    /**
     * @stability stable
     */
    putDefaultAction(value) {
        this._defaultAction.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultActionInput() {
        return this._defaultAction.internalValue;
    }
    /**
     * @stability stable
     */
    get loggingConfiguration() {
        return this._loggingConfiguration;
    }
    /**
     * @stability stable
     */
    putLoggingConfiguration(value) {
        this._loggingConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLoggingConfiguration() {
        this._loggingConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingConfigurationInput() {
        return this._loggingConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get rules() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('rules');
    }
    /**
     * @stability stable
     */
    set rules(value) {
        this._rules = value;
    }
    /**
     * @stability stable
     */
    resetRules() {
        this._rules = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rulesInput() {
        return this._rules;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            metric_name: cdktf.stringToTerraform(this._metricName),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            default_action: wafWebAclDefaultActionToTerraform(this._defaultAction.internalValue),
            logging_configuration: wafWebAclLoggingConfigurationToTerraform(this._loggingConfiguration.internalValue),
            rules: cdktf.listMapper(wafWebAclRulesToTerraform)(this._rules),
        };
    }
}
exports.WafWebAcl = WafWebAcl;
_f = JSII_RTTI_SYMBOL_1;
WafWebAcl[_f] = { fqn: "@cdktf/provider-aws.waf.WafWebAcl", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
WafWebAcl.tfResourceType = "aws_waf_web_acl";
