// @cdktf/provider-aws/lib/eks/eks-addon.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksAddon = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/eks_addon.html aws_eks_addon}.
 *
 * @stability stable
 */
class EksAddon extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/eks_addon.html aws_eks_addon} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_addon',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._addonName = config.addonName;
        this._addonVersion = config.addonVersion;
        this._clusterName = config.clusterName;
        this._resolveConflicts = config.resolveConflicts;
        this._serviceAccountRoleArn = config.serviceAccountRoleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    /**
     * @stability stable
     */
    get addonName() {
        return this.getStringAttribute('addon_name');
    }
    /**
     * @stability stable
     */
    set addonName(value) {
        this._addonName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get addonNameInput() {
        return this._addonName;
    }
    /**
     * @stability stable
     */
    get addonVersion() {
        return this.getStringAttribute('addon_version');
    }
    /**
     * @stability stable
     */
    set addonVersion(value) {
        this._addonVersion = value;
    }
    /**
     * @stability stable
     */
    resetAddonVersion() {
        this._addonVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get addonVersionInput() {
        return this._addonVersion;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    /**
     * @stability stable
     */
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterNameInput() {
        return this._clusterName;
    }
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // modified_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get modifiedAt() {
        return this.getStringAttribute('modified_at');
    }
    /**
     * @stability stable
     */
    get resolveConflicts() {
        return this.getStringAttribute('resolve_conflicts');
    }
    /**
     * @stability stable
     */
    set resolveConflicts(value) {
        this._resolveConflicts = value;
    }
    /**
     * @stability stable
     */
    resetResolveConflicts() {
        this._resolveConflicts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resolveConflictsInput() {
        return this._resolveConflicts;
    }
    /**
     * @stability stable
     */
    get serviceAccountRoleArn() {
        return this.getStringAttribute('service_account_role_arn');
    }
    /**
     * @stability stable
     */
    set serviceAccountRoleArn(value) {
        this._serviceAccountRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetServiceAccountRoleArn() {
        this._serviceAccountRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceAccountRoleArnInput() {
        return this._serviceAccountRoleArn;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            addon_name: cdktf.stringToTerraform(this._addonName),
            addon_version: cdktf.stringToTerraform(this._addonVersion),
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            resolve_conflicts: cdktf.stringToTerraform(this._resolveConflicts),
            service_account_role_arn: cdktf.stringToTerraform(this._serviceAccountRoleArn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.EksAddon = EksAddon;
_a = JSII_RTTI_SYMBOL_1;
EksAddon[_a] = { fqn: "@cdktf/provider-aws.eks.EksAddon", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EksAddon.tfResourceType = "aws_eks_addon";
