// @cdktf/provider-aws/lib/s3/data-aws-s3-bucket-objects.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsS3BucketObjects = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/s3_bucket_objects.html aws_s3_bucket_objects}.
 *
 * @stability stable
 */
class DataAwsS3BucketObjects extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/s3_bucket_objects.html aws_s3_bucket_objects} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_objects',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._bucket = config.bucket;
        this._delimiter = config.delimiter;
        this._encodingType = config.encodingType;
        this._fetchOwner = config.fetchOwner;
        this._maxKeys = config.maxKeys;
        this._prefix = config.prefix;
        this._startAfter = config.startAfter;
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    // common_prefixes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get commonPrefixes() {
        return this.getListAttribute('common_prefixes');
    }
    /**
     * @stability stable
     */
    get delimiter() {
        return this.getStringAttribute('delimiter');
    }
    /**
     * @stability stable
     */
    set delimiter(value) {
        this._delimiter = value;
    }
    /**
     * @stability stable
     */
    resetDelimiter() {
        this._delimiter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get delimiterInput() {
        return this._delimiter;
    }
    /**
     * @stability stable
     */
    get encodingType() {
        return this.getStringAttribute('encoding_type');
    }
    /**
     * @stability stable
     */
    set encodingType(value) {
        this._encodingType = value;
    }
    /**
     * @stability stable
     */
    resetEncodingType() {
        this._encodingType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encodingTypeInput() {
        return this._encodingType;
    }
    /**
     * @stability stable
     */
    get fetchOwner() {
        return this.getBooleanAttribute('fetch_owner');
    }
    /**
     * @stability stable
     */
    set fetchOwner(value) {
        this._fetchOwner = value;
    }
    /**
     * @stability stable
     */
    resetFetchOwner() {
        this._fetchOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fetchOwnerInput() {
        return this._fetchOwner;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // keys - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keys() {
        return this.getListAttribute('keys');
    }
    /**
     * @stability stable
     */
    get maxKeys() {
        return this.getNumberAttribute('max_keys');
    }
    /**
     * @stability stable
     */
    set maxKeys(value) {
        this._maxKeys = value;
    }
    /**
     * @stability stable
     */
    resetMaxKeys() {
        this._maxKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxKeysInput() {
        return this._maxKeys;
    }
    // owners - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get owners() {
        return this.getListAttribute('owners');
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
    /**
     * @stability stable
     */
    get startAfter() {
        return this.getStringAttribute('start_after');
    }
    /**
     * @stability stable
     */
    set startAfter(value) {
        this._startAfter = value;
    }
    /**
     * @stability stable
     */
    resetStartAfter() {
        this._startAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startAfterInput() {
        return this._startAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            delimiter: cdktf.stringToTerraform(this._delimiter),
            encoding_type: cdktf.stringToTerraform(this._encodingType),
            fetch_owner: cdktf.booleanToTerraform(this._fetchOwner),
            max_keys: cdktf.numberToTerraform(this._maxKeys),
            prefix: cdktf.stringToTerraform(this._prefix),
            start_after: cdktf.stringToTerraform(this._startAfter),
        };
    }
}
exports.DataAwsS3BucketObjects = DataAwsS3BucketObjects;
_a = JSII_RTTI_SYMBOL_1;
DataAwsS3BucketObjects[_a] = { fqn: "@cdktf/provider-aws.s3.DataAwsS3BucketObjects", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsS3BucketObjects.tfResourceType = "aws_s3_bucket_objects";
