// @cdktf/provider-aws/lib/s3/s3-access-point.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3AccessPoint = exports.S3AccessPointVpcConfigurationOutputReference = exports.s3AccessPointVpcConfigurationToTerraform = exports.S3AccessPointPublicAccessBlockConfigurationOutputReference = exports.s3AccessPointPublicAccessBlockConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3AccessPointPublicAccessBlockConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        block_public_acls: cdktf.booleanToTerraform(struct.blockPublicAcls),
        block_public_policy: cdktf.booleanToTerraform(struct.blockPublicPolicy),
        ignore_public_acls: cdktf.booleanToTerraform(struct.ignorePublicAcls),
        restrict_public_buckets: cdktf.booleanToTerraform(struct.restrictPublicBuckets),
    };
}
exports.s3AccessPointPublicAccessBlockConfigurationToTerraform = s3AccessPointPublicAccessBlockConfigurationToTerraform;
/**
 * @stability stable
 */
class S3AccessPointPublicAccessBlockConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._blockPublicAcls) {
            hasAnyValues = true;
            internalValueResult.blockPublicAcls = this._blockPublicAcls;
        }
        if (this._blockPublicPolicy) {
            hasAnyValues = true;
            internalValueResult.blockPublicPolicy = this._blockPublicPolicy;
        }
        if (this._ignorePublicAcls) {
            hasAnyValues = true;
            internalValueResult.ignorePublicAcls = this._ignorePublicAcls;
        }
        if (this._restrictPublicBuckets) {
            hasAnyValues = true;
            internalValueResult.restrictPublicBuckets = this._restrictPublicBuckets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._blockPublicAcls = undefined;
            this._blockPublicPolicy = undefined;
            this._ignorePublicAcls = undefined;
            this._restrictPublicBuckets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._blockPublicAcls = value.blockPublicAcls;
            this._blockPublicPolicy = value.blockPublicPolicy;
            this._ignorePublicAcls = value.ignorePublicAcls;
            this._restrictPublicBuckets = value.restrictPublicBuckets;
        }
    }
    /**
     * @stability stable
     */
    get blockPublicAcls() {
        return this.getBooleanAttribute('block_public_acls');
    }
    /**
     * @stability stable
     */
    set blockPublicAcls(value) {
        this._blockPublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicAcls() {
        this._blockPublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicAclsInput() {
        return this._blockPublicAcls;
    }
    /**
     * @stability stable
     */
    get blockPublicPolicy() {
        return this.getBooleanAttribute('block_public_policy');
    }
    /**
     * @stability stable
     */
    set blockPublicPolicy(value) {
        this._blockPublicPolicy = value;
    }
    /**
     * @stability stable
     */
    resetBlockPublicPolicy() {
        this._blockPublicPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get blockPublicPolicyInput() {
        return this._blockPublicPolicy;
    }
    /**
     * @stability stable
     */
    get ignorePublicAcls() {
        return this.getBooleanAttribute('ignore_public_acls');
    }
    /**
     * @stability stable
     */
    set ignorePublicAcls(value) {
        this._ignorePublicAcls = value;
    }
    /**
     * @stability stable
     */
    resetIgnorePublicAcls() {
        this._ignorePublicAcls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ignorePublicAclsInput() {
        return this._ignorePublicAcls;
    }
    /**
     * @stability stable
     */
    get restrictPublicBuckets() {
        return this.getBooleanAttribute('restrict_public_buckets');
    }
    /**
     * @stability stable
     */
    set restrictPublicBuckets(value) {
        this._restrictPublicBuckets = value;
    }
    /**
     * @stability stable
     */
    resetRestrictPublicBuckets() {
        this._restrictPublicBuckets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restrictPublicBucketsInput() {
        return this._restrictPublicBuckets;
    }
}
exports.S3AccessPointPublicAccessBlockConfigurationOutputReference = S3AccessPointPublicAccessBlockConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3AccessPointPublicAccessBlockConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3.S3AccessPointPublicAccessBlockConfigurationOutputReference", version: "3.0.1" };
function s3AccessPointVpcConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        vpc_id: cdktf.stringToTerraform(struct.vpcId),
    };
}
exports.s3AccessPointVpcConfigurationToTerraform = s3AccessPointVpcConfigurationToTerraform;
/**
 * @stability stable
 */
class S3AccessPointVpcConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._vpcId) {
            hasAnyValues = true;
            internalValueResult.vpcId = this._vpcId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._vpcId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._vpcId = value.vpcId;
        }
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
}
exports.S3AccessPointVpcConfigurationOutputReference = S3AccessPointVpcConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3AccessPointVpcConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3.S3AccessPointVpcConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_access_point.html aws_s3_access_point}.
 *
 * @stability stable
 */
class S3AccessPoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_access_point.html aws_s3_access_point} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_access_point',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // public_access_block_configuration - computed: false, optional: true, required: false
        this._publicAccessBlockConfiguration = new S3AccessPointPublicAccessBlockConfigurationOutputReference(this, "public_access_block_configuration", true);
        // vpc_configuration - computed: false, optional: true, required: false
        this._vpcConfiguration = new S3AccessPointVpcConfigurationOutputReference(this, "vpc_configuration", true);
        this._accountId = config.accountId;
        this._bucket = config.bucket;
        this._name = config.name;
        this._policy = config.policy;
        this._publicAccessBlockConfiguration.internalValue = config.publicAccessBlockConfiguration;
        this._vpcConfiguration.internalValue = config.vpcConfiguration;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // alias - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get alias() {
        return this.getStringAttribute('alias');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    // domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // endpoints - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    endpoints(key) {
        return new cdktf.StringMap(this, 'endpoints').lookup(key);
    }
    // has_public_access_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hasPublicAccessPolicy() {
        return this.getBooleanAttribute('has_public_access_policy');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // network_origin - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkOrigin() {
        return this.getStringAttribute('network_origin');
    }
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    /**
     * @stability stable
     */
    set policy(value) {
        this._policy = value;
    }
    /**
     * @stability stable
     */
    resetPolicy() {
        this._policy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyInput() {
        return this._policy;
    }
    /**
     * @stability stable
     */
    get publicAccessBlockConfiguration() {
        return this._publicAccessBlockConfiguration;
    }
    /**
     * @stability stable
     */
    putPublicAccessBlockConfiguration(value) {
        this._publicAccessBlockConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPublicAccessBlockConfiguration() {
        this._publicAccessBlockConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicAccessBlockConfigurationInput() {
        return this._publicAccessBlockConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get vpcConfiguration() {
        return this._vpcConfiguration;
    }
    /**
     * @stability stable
     */
    putVpcConfiguration(value) {
        this._vpcConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetVpcConfiguration() {
        this._vpcConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcConfigurationInput() {
        return this._vpcConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            bucket: cdktf.stringToTerraform(this._bucket),
            name: cdktf.stringToTerraform(this._name),
            policy: cdktf.stringToTerraform(this._policy),
            public_access_block_configuration: s3AccessPointPublicAccessBlockConfigurationToTerraform(this._publicAccessBlockConfiguration.internalValue),
            vpc_configuration: s3AccessPointVpcConfigurationToTerraform(this._vpcConfiguration.internalValue),
        };
    }
}
exports.S3AccessPoint = S3AccessPoint;
_c = JSII_RTTI_SYMBOL_1;
S3AccessPoint[_c] = { fqn: "@cdktf/provider-aws.s3.S3AccessPoint", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
S3AccessPoint.tfResourceType = "aws_s3_access_point";
