// @cdktf/provider-aws/lib/s3/s3-bucket-inventory.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketInventory = exports.S3BucketInventoryScheduleOutputReference = exports.s3BucketInventoryScheduleToTerraform = exports.S3BucketInventoryFilterOutputReference = exports.s3BucketInventoryFilterToTerraform = exports.S3BucketInventoryDestinationOutputReference = exports.s3BucketInventoryDestinationToTerraform = exports.S3BucketInventoryDestinationBucketOutputReference = exports.s3BucketInventoryDestinationBucketToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionOutputReference = exports.s3BucketInventoryDestinationBucketEncryptionToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference = exports.s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference = exports.s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_id: cdktf.stringToTerraform(struct.keyId),
    };
}
exports.s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform = s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyId) {
            hasAnyValues = true;
            internalValueResult.keyId = this._keyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyId = value.keyId;
        }
    }
    /**
     * @stability stable
     */
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    /**
     * @stability stable
     */
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyIdInput() {
        return this._keyId;
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference = S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference", version: "3.0.1" };
function s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform = s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference = S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference[_b] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference", version: "3.0.1" };
function s3BucketInventoryDestinationBucketEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sse_kms: s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform(struct.sseKms),
        sse_s3: s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform(struct.sseS3),
    };
}
exports.s3BucketInventoryDestinationBucketEncryptionToTerraform = s3BucketInventoryDestinationBucketEncryptionToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryDestinationBucketEncryptionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // sse_kms - computed: false, optional: true, required: false
        this._sseKms = new S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference(this, "sse_kms", true);
        // sse_s3 - computed: false, optional: true, required: false
        this._sseS3 = new S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference(this, "sse_s3", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k, _l, _m;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_j = this._sseKms) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.sseKms = (_k = this._sseKms) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._sseS3) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.sseS3 = (_m = this._sseS3) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sseKms.internalValue = undefined;
            this._sseS3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sseKms.internalValue = value.sseKms;
            this._sseS3.internalValue = value.sseS3;
        }
    }
    /**
     * @stability stable
     */
    get sseKms() {
        return this._sseKms;
    }
    /**
     * @stability stable
     */
    putSseKms(value) {
        this._sseKms.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSseKms() {
        this._sseKms.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sseKmsInput() {
        return this._sseKms.internalValue;
    }
    /**
     * @stability stable
     */
    get sseS3() {
        return this._sseS3;
    }
    /**
     * @stability stable
     */
    putSseS3(value) {
        this._sseS3.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSseS3() {
        this._sseS3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sseS3Input() {
        return this._sseS3.internalValue;
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionOutputReference = S3BucketInventoryDestinationBucketEncryptionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryDestinationBucketEncryptionOutputReference", version: "3.0.1" };
function s3BucketInventoryDestinationBucketToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account_id: cdktf.stringToTerraform(struct.accountId),
        bucket_arn: cdktf.stringToTerraform(struct.bucketArn),
        format: cdktf.stringToTerraform(struct.format),
        prefix: cdktf.stringToTerraform(struct.prefix),
        encryption: s3BucketInventoryDestinationBucketEncryptionToTerraform(struct.encryption),
    };
}
exports.s3BucketInventoryDestinationBucketToTerraform = s3BucketInventoryDestinationBucketToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryDestinationBucketOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // encryption - computed: false, optional: true, required: false
        this._encryption = new S3BucketInventoryDestinationBucketEncryptionOutputReference(this, "encryption", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accountId) {
            hasAnyValues = true;
            internalValueResult.accountId = this._accountId;
        }
        if (this._bucketArn) {
            hasAnyValues = true;
            internalValueResult.bucketArn = this._bucketArn;
        }
        if (this._format) {
            hasAnyValues = true;
            internalValueResult.format = this._format;
        }
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if ((_j = this._encryption) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.encryption = (_k = this._encryption) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accountId = undefined;
            this._bucketArn = undefined;
            this._format = undefined;
            this._prefix = undefined;
            this._encryption.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accountId = value.accountId;
            this._bucketArn = value.bucketArn;
            this._format = value.format;
            this._prefix = value.prefix;
            this._encryption.internalValue = value.encryption;
        }
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    /**
     * @stability stable
     */
    get bucketArn() {
        return this.getStringAttribute('bucket_arn');
    }
    /**
     * @stability stable
     */
    set bucketArn(value) {
        this._bucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketArnInput() {
        return this._bucketArn;
    }
    /**
     * @stability stable
     */
    get format() {
        return this.getStringAttribute('format');
    }
    /**
     * @stability stable
     */
    set format(value) {
        this._format = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get formatInput() {
        return this._format;
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
    /**
     * @stability stable
     */
    get encryption() {
        return this._encryption;
    }
    /**
     * @stability stable
     */
    putEncryption(value) {
        this._encryption.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryption() {
        this._encryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionInput() {
        return this._encryption.internalValue;
    }
}
exports.S3BucketInventoryDestinationBucketOutputReference = S3BucketInventoryDestinationBucketOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryDestinationBucketOutputReference", version: "3.0.1" };
function s3BucketInventoryDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: s3BucketInventoryDestinationBucketToTerraform(struct.bucket),
    };
}
exports.s3BucketInventoryDestinationToTerraform = s3BucketInventoryDestinationToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryDestinationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // bucket - computed: false, optional: false, required: true
        this._bucket = new S3BucketInventoryDestinationBucketOutputReference(this, "bucket", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_j = this._bucket) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.bucket = (_k = this._bucket) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket.internalValue = value.bucket;
        }
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    putBucket(value) {
        this._bucket.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket.internalValue;
    }
}
exports.S3BucketInventoryDestinationOutputReference = S3BucketInventoryDestinationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryDestinationOutputReference", version: "3.0.1" };
function s3BucketInventoryFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.s3BucketInventoryFilterToTerraform = s3BucketInventoryFilterToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryFilterOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
        }
    }
    /**
     * @stability stable
     */
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    /**
     * @stability stable
     */
    set prefix(value) {
        this._prefix = value;
    }
    /**
     * @stability stable
     */
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get prefixInput() {
        return this._prefix;
    }
}
exports.S3BucketInventoryFilterOutputReference = S3BucketInventoryFilterOutputReference;
_f = JSII_RTTI_SYMBOL_1;
S3BucketInventoryFilterOutputReference[_f] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryFilterOutputReference", version: "3.0.1" };
function s3BucketInventoryScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        frequency: cdktf.stringToTerraform(struct.frequency),
    };
}
exports.s3BucketInventoryScheduleToTerraform = s3BucketInventoryScheduleToTerraform;
/**
 * @stability stable
 */
class S3BucketInventoryScheduleOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._frequency) {
            hasAnyValues = true;
            internalValueResult.frequency = this._frequency;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._frequency = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._frequency = value.frequency;
        }
    }
    /**
     * @stability stable
     */
    get frequency() {
        return this.getStringAttribute('frequency');
    }
    /**
     * @stability stable
     */
    set frequency(value) {
        this._frequency = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get frequencyInput() {
        return this._frequency;
    }
}
exports.S3BucketInventoryScheduleOutputReference = S3BucketInventoryScheduleOutputReference;
_g = JSII_RTTI_SYMBOL_1;
S3BucketInventoryScheduleOutputReference[_g] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventoryScheduleOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_inventory.html aws_s3_bucket_inventory}.
 *
 * @stability stable
 */
class S3BucketInventory extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_inventory.html aws_s3_bucket_inventory} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_inventory',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // destination - computed: false, optional: false, required: true
        this._destination = new S3BucketInventoryDestinationOutputReference(this, "destination", true);
        // filter - computed: false, optional: true, required: false
        this._filter = new S3BucketInventoryFilterOutputReference(this, "filter", true);
        // schedule - computed: false, optional: false, required: true
        this._schedule = new S3BucketInventoryScheduleOutputReference(this, "schedule", true);
        this._bucket = config.bucket;
        this._enabled = config.enabled;
        this._includedObjectVersions = config.includedObjectVersions;
        this._name = config.name;
        this._optionalFields = config.optionalFields;
        this._destination.internalValue = config.destination;
        this._filter.internalValue = config.filter;
        this._schedule.internalValue = config.schedule;
    }
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    /**
     * @stability stable
     */
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketInput() {
        return this._bucket;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get includedObjectVersions() {
        return this.getStringAttribute('included_object_versions');
    }
    /**
     * @stability stable
     */
    set includedObjectVersions(value) {
        this._includedObjectVersions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includedObjectVersionsInput() {
        return this._includedObjectVersions;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get optionalFields() {
        return this.getListAttribute('optional_fields');
    }
    /**
     * @stability stable
     */
    set optionalFields(value) {
        this._optionalFields = value;
    }
    /**
     * @stability stable
     */
    resetOptionalFields() {
        this._optionalFields = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get optionalFieldsInput() {
        return this._optionalFields;
    }
    /**
     * @stability stable
     */
    get destination() {
        return this._destination;
    }
    /**
     * @stability stable
     */
    putDestination(value) {
        this._destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationInput() {
        return this._destination.internalValue;
    }
    /**
     * @stability stable
     */
    get filter() {
        return this._filter;
    }
    /**
     * @stability stable
     */
    putFilter(value) {
        this._filter.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter.internalValue;
    }
    /**
     * @stability stable
     */
    get schedule() {
        return this._schedule;
    }
    /**
     * @stability stable
     */
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            enabled: cdktf.booleanToTerraform(this._enabled),
            included_object_versions: cdktf.stringToTerraform(this._includedObjectVersions),
            name: cdktf.stringToTerraform(this._name),
            optional_fields: cdktf.listMapper(cdktf.stringToTerraform)(this._optionalFields),
            destination: s3BucketInventoryDestinationToTerraform(this._destination.internalValue),
            filter: s3BucketInventoryFilterToTerraform(this._filter.internalValue),
            schedule: s3BucketInventoryScheduleToTerraform(this._schedule.internalValue),
        };
    }
}
exports.S3BucketInventory = S3BucketInventory;
_h = JSII_RTTI_SYMBOL_1;
S3BucketInventory[_h] = { fqn: "@cdktf/provider-aws.s3.S3BucketInventory", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
S3BucketInventory.tfResourceType = "aws_s3_bucket_inventory";
