// @cdktf/provider-aws/lib/datasources/data-aws-availability-zone.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAvailabilityZone = exports.dataAwsAvailabilityZoneFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAvailabilityZoneFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsAvailabilityZoneFilterToTerraform = dataAwsAvailabilityZoneFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/availability_zone.html aws_availability_zone}.
 *
 * @stability stable
 */
class DataAwsAvailabilityZone extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/availability_zone.html aws_availability_zone} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_availability_zone',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._allAvailabilityZones = config.allAvailabilityZones;
        this._name = config.name;
        this._state = config.state;
        this._zoneId = config.zoneId;
        this._filter = config.filter;
    }
    /**
     * @stability stable
     */
    get allAvailabilityZones() {
        return this.getBooleanAttribute('all_availability_zones');
    }
    /**
     * @stability stable
     */
    set allAvailabilityZones(value) {
        this._allAvailabilityZones = value;
    }
    /**
     * @stability stable
     */
    resetAllAvailabilityZones() {
        this._allAvailabilityZones = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allAvailabilityZonesInput() {
        return this._allAvailabilityZones;
    }
    // group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // name_suffix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nameSuffix() {
        return this.getStringAttribute('name_suffix');
    }
    // network_border_group - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkBorderGroup() {
        return this.getStringAttribute('network_border_group');
    }
    // opt_in_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get optInStatus() {
        return this.getStringAttribute('opt_in_status');
    }
    // parent_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get parentZoneId() {
        return this.getStringAttribute('parent_zone_id');
    }
    // parent_zone_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get parentZoneName() {
        return this.getStringAttribute('parent_zone_name');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    set state(value) {
        this._state = value;
    }
    /**
     * @stability stable
     */
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateInput() {
        return this._state;
    }
    /**
     * @stability stable
     */
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    /**
     * @stability stable
     */
    set zoneId(value) {
        this._zoneId = value;
    }
    /**
     * @stability stable
     */
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get zoneIdInput() {
        return this._zoneId;
    }
    // zone_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneType() {
        return this.getStringAttribute('zone_type');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            all_availability_zones: cdktf.booleanToTerraform(this._allAvailabilityZones),
            name: cdktf.stringToTerraform(this._name),
            state: cdktf.stringToTerraform(this._state),
            zone_id: cdktf.stringToTerraform(this._zoneId),
            filter: cdktf.listMapper(dataAwsAvailabilityZoneFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsAvailabilityZone = DataAwsAvailabilityZone;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAvailabilityZone[_a] = { fqn: "@cdktf/provider-aws.datasources.DataAwsAvailabilityZone", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAvailabilityZone.tfResourceType = "aws_availability_zone";
