// @cdktf/provider-aws/lib/datasources/data-aws-elb.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsElb = exports.DataAwsElbListener = exports.DataAwsElbHealthCheck = exports.DataAwsElbAccessLogs = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsElbAccessLogs extends cdktf.ComplexComputedList {
    // bucket - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // bucket_prefix - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get bucketPrefix() {
        return this.getStringAttribute('bucket_prefix');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // interval - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get interval() {
        return this.getNumberAttribute('interval');
    }
}
exports.DataAwsElbAccessLogs = DataAwsElbAccessLogs;
_a = JSII_RTTI_SYMBOL_1;
DataAwsElbAccessLogs[_a] = { fqn: "@cdktf/provider-aws.datasources.DataAwsElbAccessLogs", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElbHealthCheck extends cdktf.ComplexComputedList {
    // healthy_threshold - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    // interval - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get interval() {
        return this.getNumberAttribute('interval');
    }
    // target - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get target() {
        return this.getStringAttribute('target');
    }
    // timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    // unhealthy_threshold - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
}
exports.DataAwsElbHealthCheck = DataAwsElbHealthCheck;
_b = JSII_RTTI_SYMBOL_1;
DataAwsElbHealthCheck[_b] = { fqn: "@cdktf/provider-aws.datasources.DataAwsElbHealthCheck", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsElbListener extends cdktf.ComplexComputedList {
    // instance_port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instancePort() {
        return this.getNumberAttribute('instance_port');
    }
    // instance_protocol - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceProtocol() {
        return this.getStringAttribute('instance_protocol');
    }
    // lb_port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lbPort() {
        return this.getNumberAttribute('lb_port');
    }
    // lb_protocol - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lbProtocol() {
        return this.getStringAttribute('lb_protocol');
    }
    // ssl_certificate_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sslCertificateId() {
        return this.getStringAttribute('ssl_certificate_id');
    }
}
exports.DataAwsElbListener = DataAwsElbListener;
_c = JSII_RTTI_SYMBOL_1;
DataAwsElbListener[_c] = { fqn: "@cdktf/provider-aws.datasources.DataAwsElbListener", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/elb.html aws_elb}.
 *
 * @stability stable
 */
class DataAwsElb extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/elb.html aws_elb} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elb',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_logs - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    accessLogs(index) {
        return new DataAwsElbAccessLogs(this, 'access_logs', index);
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zones - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZones() {
        return this.getListAttribute('availability_zones');
    }
    // connection_draining - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get connectionDraining() {
        return this.getBooleanAttribute('connection_draining');
    }
    // connection_draining_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get connectionDrainingTimeout() {
        return this.getNumberAttribute('connection_draining_timeout');
    }
    // cross_zone_load_balancing - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get crossZoneLoadBalancing() {
        return this.getBooleanAttribute('cross_zone_load_balancing');
    }
    // desync_mitigation_mode - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get desyncMitigationMode() {
        return this.getStringAttribute('desync_mitigation_mode');
    }
    // dns_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // health_check - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    healthCheck(index) {
        return new DataAwsElbHealthCheck(this, 'health_check', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // idle_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get idleTimeout() {
        return this.getNumberAttribute('idle_timeout');
    }
    // instances - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instances() {
        return this.getListAttribute('instances');
    }
    // internal - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get internal() {
        return this.getBooleanAttribute('internal');
    }
    // listener - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    listener(index) {
        return new DataAwsElbListener(this, 'listener', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    // source_security_group - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceSecurityGroup() {
        return this.getStringAttribute('source_security_group');
    }
    // source_security_group_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceSecurityGroupId() {
        return this.getStringAttribute('source_security_group_id');
    }
    // subnets - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnets() {
        return this.getListAttribute('subnets');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsElb = DataAwsElb;
_d = JSII_RTTI_SYMBOL_1;
DataAwsElb[_d] = { fqn: "@cdktf/provider-aws.datasources.DataAwsElb", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsElb.tfResourceType = "aws_elb";
