// @cdktf/provider-aws/lib/datasources/data-aws-ip-ranges.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIpRanges = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ip_ranges.html aws_ip_ranges}.
 *
 * @stability stable
 */
class DataAwsIpRanges extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ip_ranges.html aws_ip_ranges} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ip_ranges',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._regions = config.regions;
        this._services = config.services;
        this._url = config.url;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // cidr_blocks - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cidrBlocks() {
        return this.getListAttribute('cidr_blocks');
    }
    // create_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createDate() {
        return this.getStringAttribute('create_date');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // ipv6_cidr_blocks - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6CidrBlocks() {
        return this.getListAttribute('ipv6_cidr_blocks');
    }
    /**
     * @stability stable
     */
    get regions() {
        return this.getListAttribute('regions');
    }
    /**
     * @stability stable
     */
    set regions(value) {
        this._regions = value;
    }
    /**
     * @stability stable
     */
    resetRegions() {
        this._regions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionsInput() {
        return this._regions;
    }
    /**
     * @stability stable
     */
    get services() {
        return this.getListAttribute('services');
    }
    /**
     * @stability stable
     */
    set services(value) {
        this._services = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get servicesInput() {
        return this._services;
    }
    // sync_token - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get syncToken() {
        return this.getNumberAttribute('sync_token');
    }
    /**
     * @stability stable
     */
    get url() {
        return this.getStringAttribute('url');
    }
    /**
     * @stability stable
     */
    set url(value) {
        this._url = value;
    }
    /**
     * @stability stable
     */
    resetUrl() {
        this._url = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get urlInput() {
        return this._url;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            regions: cdktf.listMapper(cdktf.stringToTerraform)(this._regions),
            services: cdktf.listMapper(cdktf.stringToTerraform)(this._services),
            url: cdktf.stringToTerraform(this._url),
        };
    }
}
exports.DataAwsIpRanges = DataAwsIpRanges;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIpRanges[_a] = { fqn: "@cdktf/provider-aws.datasources.DataAwsIpRanges", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsIpRanges.tfResourceType = "aws_ip_ranges";
