// @cdktf/provider-aws/lib/datasources/elb-attachment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElbAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/elb_attachment.html aws_elb_attachment}.
 *
 * @stability stable
 */
class ElbAttachment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elb_attachment.html aws_elb_attachment} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elb_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._elb = config.elb;
        this._instance = config.instance;
    }
    /**
     * @stability stable
     */
    get elb() {
        return this.getStringAttribute('elb');
    }
    /**
     * @stability stable
     */
    set elb(value) {
        this._elb = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elbInput() {
        return this._elb;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instance() {
        return this.getStringAttribute('instance');
    }
    /**
     * @stability stable
     */
    set instance(value) {
        this._instance = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceInput() {
        return this._instance;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            elb: cdktf.stringToTerraform(this._elb),
            instance: cdktf.stringToTerraform(this._instance),
        };
    }
}
exports.ElbAttachment = ElbAttachment;
_a = JSII_RTTI_SYMBOL_1;
ElbAttachment[_a] = { fqn: "@cdktf/provider-aws.datasources.ElbAttachment", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ElbAttachment.tfResourceType = "aws_elb_attachment";
