// cdktf-cli/bin/cmds/destroy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const destroy_1 = require("./ui/destroy");
const provider_generator_1 = require("@cdktf/provider-generator");
const render_ink_1 = require("./helper/render-ink");
const version_check_1 = require("./helper/version-check");
const check_directory_1 = require("./helper/check-directory");
const check_environment_1 = require("./helper/check-environment");
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "destroy [stack] [OPTIONS]";
        this.describe = "Destroy the given stack";
        this.builder = (args) => args
            .positional("stack", {
            desc: "Destroy stack which matches the given id only. Required when more than one stack is present in the app",
            type: "string",
        })
            .option("app", {
            default: config.app,
            required: true,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            required: true,
            desc: "Output directory",
            alias: "o",
        })
            .option("auto-approve", {
            type: "boolean",
            default: false,
            required: false,
            desc: "Auto approve",
        })
            .showHelpOnFail(true);
    }
    async handler(argv) {
        check_directory_1.throwIfNotProjectDirectory("destroy");
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("destroy");
        const command = argv.app;
        const outdir = argv.output;
        const autoApprove = argv.autoApprove;
        const stack = argv.stack;
        await render_ink_1.renderInk(react_1.default.createElement(destroy_1.Destroy, {
            targetDir: outdir,
            targetStack: stack,
            synthCommand: command,
            autoApprove,
        }));
    }
}
module.exports = new Command();
