// cdktf-cli/bin/cmds/diff.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const diff_1 = require("./ui/diff");
const provider_generator_1 = require("@cdktf/provider-generator");
const render_ink_1 = require("./helper/render-ink");
const version_check_1 = require("./helper/version-check");
const check_directory_1 = require("./helper/check-directory");
const check_environment_1 = require("./helper/check-environment");
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "diff [stack] [OPTIONS]";
        this.describe = "Perform a diff (terraform plan) for the given stack";
        this.aliases = ["plan"];
        this.builder = (args) => args
            .positional("stack", {
            desc: "Diff stack which matches the given id only. Required when more than one stack is present in the app",
            type: "string",
        })
            .option("app", {
            default: config.app,
            required: true,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            required: true,
            desc: "Output directory",
            alias: "o",
        })
            .showHelpOnFail(true);
    }
    async handler(argv) {
        check_directory_1.throwIfNotProjectDirectory("diff");
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("diff");
        const command = argv.app;
        const outdir = argv.output;
        const stack = argv.stack;
        await render_ink_1.renderInk(react_1.default.createElement(diff_1.Diff, {
            targetDir: outdir,
            targetStack: stack,
            synthCommand: command,
        }));
    }
}
module.exports = new Command();
