// cdktf-cli/bin/cmds/synth.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const synth_1 = require("./ui/synth");
const provider_generator_1 = require("@cdktf/provider-generator");
const render_ink_1 = require("./helper/render-ink");
const fs = __importStar(require("fs-extra"));
const version_check_1 = require("./helper/version-check");
const check_directory_1 = require("./helper/check-directory");
const check_environment_1 = require("./helper/check-environment");
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "synth [stack] [OPTIONS]";
        this.describe = "Synthesizes Terraform code for the given app in a directory.";
        this.aliases = ["synthesize"];
        this.builder = (args) => args
            .positional("stack", {
            desc: "Stack to output when using --json flag",
            type: "string",
        })
            .option("app", {
            default: config.app,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            desc: "Output directory",
            alias: "o",
        })
            .option("json", {
            type: "boolean",
            desc: "Provide JSON output for the generated Terraform configuration.",
            default: false,
        })
            .showHelpOnFail(true);
    }
    async handler(argv) {
        check_directory_1.throwIfNotProjectDirectory("synth");
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("synth");
        const command = argv.app;
        const outdir = argv.output;
        const jsonOutput = argv.json;
        const stack = argv.stack;
        if (config.checkCodeMakerOutput &&
            !(await fs.pathExists(config.codeMakerOutput))) {
            console.error(`ERROR: synthesis failed, run "cdktf get" to generate providers in ${config.codeMakerOutput}`);
            process.exit(1);
        }
        await render_ink_1.renderInk(react_1.default.createElement(synth_1.Synth, {
            targetDir: outdir,
            targetStack: stack,
            synthCommand: command,
            jsonOutput: jsonOutput,
        }));
    }
}
module.exports = new Command();
