// cdktf-cli/bin/cmds/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const terraform_check_1 = require("./helper/terraform-check");
const version_check_1 = require("./helper/version-check");
const init_1 = require("./helper/init");
const check_environment_1 = require("./helper/check-environment");
const utilities_1 = require("./helper/utilities");
const pkg = utilities_1.readPackageJson();
class Command {
    constructor() {
        this.command = "init [OPTIONS]";
        this.describe = "Create a new cdktf project from a template.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .option("template", {
            type: "string",
            desc: `The template to be used to create a new project. Either URL to zip file or one of the built-in templates: [${init_1.templates
                .map((t) => `"${t}"`)
                .join(", ")}]`,
        })
            .option("project-name", {
            type: "string",
            desc: "The name of the project.",
        })
            .option("project-description", {
            type: "string",
            desc: "The description of the project.",
        })
            .option("dist", {
            type: "string",
            desc: 'Install dependencies from a "dist" directory (for development)',
        })
            .option("local", {
            type: "boolean",
            desc: "Use local state storage for generated Terraform.",
            default: false,
        })
            .option("cdktf-version", {
            type: "string",
            desc: "The cdktf version to use while creating a new project.",
            default: pkg.version,
        })
            .option("from-terraform-project", {
            type: "string",
            desc: "Use a terraform project as the basis, CDK constructs will be generated based on the .tf files in the path",
        })
            .strict();
    }
    async handler(argv) {
        await terraform_check_1.terraformCheck();
        await version_check_1.displayVersionMessage();
        await check_environment_1.checkEnvironment("init");
        init_1.checkForEmptyDirectory(".");
        await init_1.runInit(argv);
    }
}
module.exports = new Command();
