// cdktf-cli/bin/cmds/watch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const watch_1 = require("./ui/watch");
const provider_generator_1 = require("@cdktf/provider-generator");
const render_ink_1 = require("./helper/render-ink");
const chalk = __importStar(require("chalk"));
const version_check_1 = require("./helper/version-check");
const react_2 = require("../../lib/client/react");
const check_directory_1 = require("./helper/check-directory");
const chalkColour = new chalk.Instance();
const config = provider_generator_1.config.readConfigSync();
class Command {
    constructor() {
        this.command = "watch [stack] [OPTIONS]";
        this.describe = "[experimental] Watch for file changes and automatically trigger a deploy";
        this.builder = (args) => args
            .positional("stack", {
            desc: "Deploy stack which matches the given id only. Required when more than one stack is present in the app",
            type: "string",
        })
            .option("app", {
            default: config.app,
            required: true,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            required: true,
            desc: "Output directory",
            alias: "o",
        })
            .option("auto-approve", {
            type: "boolean",
            default: false,
            required: false,
            desc: "Auto approve",
        })
            .showHelpOnFail(true);
    }
    async handler(argv) {
        check_directory_1.throwIfNotProjectDirectory("watch");
        await version_check_1.displayVersionMessage();
        const command = argv.app;
        const outdir = argv.output;
        const autoApprove = argv.autoApprove;
        const stack = argv.stack;
        if (!autoApprove) {
            console.error(chalkColour `{redBright ERROR: The watch command always automatically deploys and approves changes. To make this behaviour explicit the --auto-approve flag must be set}`);
            process.exit(1);
        }
        await render_ink_1.renderInk(react_1.default.createElement(react_2.GraphQLServerProvider, undefined, react_1.default.createElement(watch_1.Watch, {
            targetDir: outdir,
            targetStack: stack,
            synthCommand: command,
            autoApprove,
        })));
    }
}
module.exports = new Command();
