// @cdktf/provider-aws/lib/ssm/ssm-maintenance-window-target.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmMaintenanceWindowTarget = exports.ssmMaintenanceWindowTargetTargetsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssmMaintenanceWindowTargetTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.ssmMaintenanceWindowTargetTargetsToTerraform = ssmMaintenanceWindowTargetTargetsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window_target.html aws_ssm_maintenance_window_target}.
 *
 * @stability stable
 */
class SsmMaintenanceWindowTarget extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window_target.html aws_ssm_maintenance_window_target} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_maintenance_window_target',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._description = config.description;
        this._name = config.name;
        this._ownerInformation = config.ownerInformation;
        this._resourceType = config.resourceType;
        this._windowId = config.windowId;
        this._targets = config.targets;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get ownerInformation() {
        return this.getStringAttribute('owner_information');
    }
    /**
     * @stability stable
     */
    set ownerInformation(value) {
        this._ownerInformation = value;
    }
    /**
     * @stability stable
     */
    resetOwnerInformation() {
        this._ownerInformation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerInformationInput() {
        return this._ownerInformation;
    }
    /**
     * @stability stable
     */
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    /**
     * @stability stable
     */
    set resourceType(value) {
        this._resourceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTypeInput() {
        return this._resourceType;
    }
    /**
     * @stability stable
     */
    get windowId() {
        return this.getStringAttribute('window_id');
    }
    /**
     * @stability stable
     */
    set windowId(value) {
        this._windowId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get windowIdInput() {
        return this._windowId;
    }
    /**
     * @stability stable
     */
    get targets() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('targets');
    }
    /**
     * @stability stable
     */
    set targets(value) {
        this._targets = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetsInput() {
        return this._targets;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            owner_information: cdktf.stringToTerraform(this._ownerInformation),
            resource_type: cdktf.stringToTerraform(this._resourceType),
            window_id: cdktf.stringToTerraform(this._windowId),
            targets: cdktf.listMapper(ssmMaintenanceWindowTargetTargetsToTerraform)(this._targets),
        };
    }
}
exports.SsmMaintenanceWindowTarget = SsmMaintenanceWindowTarget;
_a = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTarget[_a] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTarget", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsmMaintenanceWindowTarget.tfResourceType = "aws_ssm_maintenance_window_target";
