// @cdktf/provider-aws/lib/mediaconvert/media-convert-queue.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaConvertQueue = exports.MediaConvertQueueReservationPlanSettingsOutputReference = exports.mediaConvertQueueReservationPlanSettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mediaConvertQueueReservationPlanSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        commitment: cdktf.stringToTerraform(struct.commitment),
        renewal_type: cdktf.stringToTerraform(struct.renewalType),
        reserved_slots: cdktf.numberToTerraform(struct.reservedSlots),
    };
}
exports.mediaConvertQueueReservationPlanSettingsToTerraform = mediaConvertQueueReservationPlanSettingsToTerraform;
/**
 * @stability stable
 */
class MediaConvertQueueReservationPlanSettingsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._commitment) {
            hasAnyValues = true;
            internalValueResult.commitment = this._commitment;
        }
        if (this._renewalType) {
            hasAnyValues = true;
            internalValueResult.renewalType = this._renewalType;
        }
        if (this._reservedSlots) {
            hasAnyValues = true;
            internalValueResult.reservedSlots = this._reservedSlots;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._commitment = undefined;
            this._renewalType = undefined;
            this._reservedSlots = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._commitment = value.commitment;
            this._renewalType = value.renewalType;
            this._reservedSlots = value.reservedSlots;
        }
    }
    /**
     * @stability stable
     */
    get commitment() {
        return this.getStringAttribute('commitment');
    }
    /**
     * @stability stable
     */
    set commitment(value) {
        this._commitment = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commitmentInput() {
        return this._commitment;
    }
    /**
     * @stability stable
     */
    get renewalType() {
        return this.getStringAttribute('renewal_type');
    }
    /**
     * @stability stable
     */
    set renewalType(value) {
        this._renewalType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get renewalTypeInput() {
        return this._renewalType;
    }
    /**
     * @stability stable
     */
    get reservedSlots() {
        return this.getNumberAttribute('reserved_slots');
    }
    /**
     * @stability stable
     */
    set reservedSlots(value) {
        this._reservedSlots = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reservedSlotsInput() {
        return this._reservedSlots;
    }
}
exports.MediaConvertQueueReservationPlanSettingsOutputReference = MediaConvertQueueReservationPlanSettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MediaConvertQueueReservationPlanSettingsOutputReference[_a] = { fqn: "@cdktf/provider-aws.mediaconvert.MediaConvertQueueReservationPlanSettingsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/media_convert_queue.html aws_media_convert_queue}.
 *
 * @stability stable
 */
class MediaConvertQueue extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/media_convert_queue.html aws_media_convert_queue} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_media_convert_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // reservation_plan_settings - computed: false, optional: true, required: false
        this._reservationPlanSettings = new MediaConvertQueueReservationPlanSettingsOutputReference(this, "reservation_plan_settings", true);
        this._description = config.description;
        this._name = config.name;
        this._pricingPlan = config.pricingPlan;
        this._status = config.status;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._reservationPlanSettings.internalValue = config.reservationPlanSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get pricingPlan() {
        return this.getStringAttribute('pricing_plan');
    }
    /**
     * @stability stable
     */
    set pricingPlan(value) {
        this._pricingPlan = value;
    }
    /**
     * @stability stable
     */
    resetPricingPlan() {
        this._pricingPlan = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pricingPlanInput() {
        return this._pricingPlan;
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get reservationPlanSettings() {
        return this._reservationPlanSettings;
    }
    /**
     * @stability stable
     */
    putReservationPlanSettings(value) {
        this._reservationPlanSettings.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetReservationPlanSettings() {
        this._reservationPlanSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get reservationPlanSettingsInput() {
        return this._reservationPlanSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            pricing_plan: cdktf.stringToTerraform(this._pricingPlan),
            status: cdktf.stringToTerraform(this._status),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            reservation_plan_settings: mediaConvertQueueReservationPlanSettingsToTerraform(this._reservationPlanSettings.internalValue),
        };
    }
}
exports.MediaConvertQueue = MediaConvertQueue;
_b = JSII_RTTI_SYMBOL_1;
MediaConvertQueue[_b] = { fqn: "@cdktf/provider-aws.mediaconvert.MediaConvertQueue", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
MediaConvertQueue.tfResourceType = "aws_media_convert_queue";
