// @cdktf/provider-aws/lib/eks/eks-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksCluster = exports.EksClusterVpcConfigOutputReference = exports.eksClusterVpcConfigToTerraform = exports.EksClusterTimeoutsOutputReference = exports.eksClusterTimeoutsToTerraform = exports.EksClusterKubernetesNetworkConfigOutputReference = exports.eksClusterKubernetesNetworkConfigToTerraform = exports.EksClusterEncryptionConfigOutputReference = exports.eksClusterEncryptionConfigToTerraform = exports.EksClusterEncryptionConfigProviderOutputReference = exports.eksClusterEncryptionConfigProviderToTerraform = exports.EksClusterIdentity = exports.EksClusterIdentityOidc = exports.EksClusterCertificateAuthority = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class EksClusterCertificateAuthority extends cdktf.ComplexComputedList {
    // data - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get data() {
        return this.getStringAttribute('data');
    }
}
exports.EksClusterCertificateAuthority = EksClusterCertificateAuthority;
_a = JSII_RTTI_SYMBOL_1;
EksClusterCertificateAuthority[_a] = { fqn: "@cdktf/provider-aws.eks.EksClusterCertificateAuthority", version: "3.0.1" };
/**
 * @stability stable
 */
class EksClusterIdentityOidc extends cdktf.ComplexComputedList {
    // issuer - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get issuer() {
        return this.getStringAttribute('issuer');
    }
}
exports.EksClusterIdentityOidc = EksClusterIdentityOidc;
_b = JSII_RTTI_SYMBOL_1;
EksClusterIdentityOidc[_b] = { fqn: "@cdktf/provider-aws.eks.EksClusterIdentityOidc", version: "3.0.1" };
/**
 * @stability stable
 */
class EksClusterIdentity extends cdktf.ComplexComputedList {
    // oidc - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get oidc() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('oidc');
    }
}
exports.EksClusterIdentity = EksClusterIdentity;
_c = JSII_RTTI_SYMBOL_1;
EksClusterIdentity[_c] = { fqn: "@cdktf/provider-aws.eks.EksClusterIdentity", version: "3.0.1" };
function eksClusterEncryptionConfigProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_arn: cdktf.stringToTerraform(struct.keyArn),
    };
}
exports.eksClusterEncryptionConfigProviderToTerraform = eksClusterEncryptionConfigProviderToTerraform;
/**
 * @stability stable
 */
class EksClusterEncryptionConfigProviderOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyArn) {
            hasAnyValues = true;
            internalValueResult.keyArn = this._keyArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyArn = value.keyArn;
        }
    }
    /**
     * @stability stable
     */
    get keyArn() {
        return this.getStringAttribute('key_arn');
    }
    /**
     * @stability stable
     */
    set keyArn(value) {
        this._keyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyArnInput() {
        return this._keyArn;
    }
}
exports.EksClusterEncryptionConfigProviderOutputReference = EksClusterEncryptionConfigProviderOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EksClusterEncryptionConfigProviderOutputReference[_d] = { fqn: "@cdktf/provider-aws.eks.EksClusterEncryptionConfigProviderOutputReference", version: "3.0.1" };
function eksClusterEncryptionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: cdktf.listMapper(cdktf.stringToTerraform)(struct.resources),
        provider: eksClusterEncryptionConfigProviderToTerraform(struct.provider),
    };
}
exports.eksClusterEncryptionConfigToTerraform = eksClusterEncryptionConfigToTerraform;
/**
 * @stability stable
 */
class EksClusterEncryptionConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // provider - computed: false, optional: false, required: true
        this._provider = new EksClusterEncryptionConfigProviderOutputReference(this, "provider", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _k, _l;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        if ((_k = this._provider) === null || _k === void 0 ? void 0 : _k.internalValue) {
            hasAnyValues = true;
            internalValueResult.provider = (_l = this._provider) === null || _l === void 0 ? void 0 : _l.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources = undefined;
            this._provider.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources = value.resources;
            this._provider.internalValue = value.provider;
        }
    }
    /**
     * @stability stable
     */
    get resources() {
        return this.getListAttribute('resources');
    }
    /**
     * @stability stable
     */
    set resources(value) {
        this._resources = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourcesInput() {
        return this._resources;
    }
    /**
     * @stability stable
     */
    get provider() {
        return this._provider;
    }
    /**
     * @stability stable
     */
    putProvider(value) {
        this._provider.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get providerInput() {
        return this._provider.internalValue;
    }
}
exports.EksClusterEncryptionConfigOutputReference = EksClusterEncryptionConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EksClusterEncryptionConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.eks.EksClusterEncryptionConfigOutputReference", version: "3.0.1" };
function eksClusterKubernetesNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        service_ipv4_cidr: cdktf.stringToTerraform(struct.serviceIpv4Cidr),
    };
}
exports.eksClusterKubernetesNetworkConfigToTerraform = eksClusterKubernetesNetworkConfigToTerraform;
/**
 * @stability stable
 */
class EksClusterKubernetesNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serviceIpv4Cidr) {
            hasAnyValues = true;
            internalValueResult.serviceIpv4Cidr = this._serviceIpv4Cidr;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._serviceIpv4Cidr = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._serviceIpv4Cidr = value.serviceIpv4Cidr;
        }
    }
    /**
     * @stability stable
     */
    get serviceIpv4Cidr() {
        return this.getStringAttribute('service_ipv4_cidr');
    }
    /**
     * @stability stable
     */
    set serviceIpv4Cidr(value) {
        this._serviceIpv4Cidr = value;
    }
    /**
     * @stability stable
     */
    resetServiceIpv4Cidr() {
        this._serviceIpv4Cidr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceIpv4CidrInput() {
        return this._serviceIpv4Cidr;
    }
}
exports.EksClusterKubernetesNetworkConfigOutputReference = EksClusterKubernetesNetworkConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EksClusterKubernetesNetworkConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.eks.EksClusterKubernetesNetworkConfigOutputReference", version: "3.0.1" };
function eksClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.eksClusterTimeoutsToTerraform = eksClusterTimeoutsToTerraform;
/**
 * @stability stable
 */
class EksClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    /**
     * @stability stable
     */
    get create() {
        return this.getStringAttribute('create');
    }
    /**
     * @stability stable
     */
    set create(value) {
        this._create = value;
    }
    /**
     * @stability stable
     */
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get createInput() {
        return this._create;
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
    /**
     * @stability stable
     */
    get update() {
        return this.getStringAttribute('update');
    }
    /**
     * @stability stable
     */
    set update(value) {
        this._update = value;
    }
    /**
     * @stability stable
     */
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateInput() {
        return this._update;
    }
}
exports.EksClusterTimeoutsOutputReference = EksClusterTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EksClusterTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-aws.eks.EksClusterTimeoutsOutputReference", version: "3.0.1" };
function eksClusterVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_private_access: cdktf.booleanToTerraform(struct.endpointPrivateAccess),
        endpoint_public_access: cdktf.booleanToTerraform(struct.endpointPublicAccess),
        public_access_cidrs: cdktf.listMapper(cdktf.stringToTerraform)(struct.publicAccessCidrs),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnetIds),
    };
}
exports.eksClusterVpcConfigToTerraform = eksClusterVpcConfigToTerraform;
/**
 * @stability stable
 */
class EksClusterVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointPrivateAccess) {
            hasAnyValues = true;
            internalValueResult.endpointPrivateAccess = this._endpointPrivateAccess;
        }
        if (this._endpointPublicAccess) {
            hasAnyValues = true;
            internalValueResult.endpointPublicAccess = this._endpointPublicAccess;
        }
        if (this._publicAccessCidrs) {
            hasAnyValues = true;
            internalValueResult.publicAccessCidrs = this._publicAccessCidrs;
        }
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpointPrivateAccess = undefined;
            this._endpointPublicAccess = undefined;
            this._publicAccessCidrs = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpointPrivateAccess = value.endpointPrivateAccess;
            this._endpointPublicAccess = value.endpointPublicAccess;
            this._publicAccessCidrs = value.publicAccessCidrs;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    /**
     * @stability stable
     */
    get endpointPrivateAccess() {
        return this.getBooleanAttribute('endpoint_private_access');
    }
    /**
     * @stability stable
     */
    set endpointPrivateAccess(value) {
        this._endpointPrivateAccess = value;
    }
    /**
     * @stability stable
     */
    resetEndpointPrivateAccess() {
        this._endpointPrivateAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointPrivateAccessInput() {
        return this._endpointPrivateAccess;
    }
    /**
     * @stability stable
     */
    get endpointPublicAccess() {
        return this.getBooleanAttribute('endpoint_public_access');
    }
    /**
     * @stability stable
     */
    set endpointPublicAccess(value) {
        this._endpointPublicAccess = value;
    }
    /**
     * @stability stable
     */
    resetEndpointPublicAccess() {
        this._endpointPublicAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointPublicAccessInput() {
        return this._endpointPublicAccess;
    }
    /**
     * @stability stable
     */
    get publicAccessCidrs() {
        return this.getListAttribute('public_access_cidrs');
    }
    /**
     * @stability stable
     */
    set publicAccessCidrs(value) {
        this._publicAccessCidrs = value;
    }
    /**
     * @stability stable
     */
    resetPublicAccessCidrs() {
        this._publicAccessCidrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get publicAccessCidrsInput() {
        return this._publicAccessCidrs;
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.EksClusterVpcConfigOutputReference = EksClusterVpcConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EksClusterVpcConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.eks.EksClusterVpcConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/eks_cluster.html aws_eks_cluster}.
 *
 * @stability stable
 */
class EksCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/eks_cluster.html aws_eks_cluster} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // encryption_config - computed: false, optional: true, required: false
        this._encryptionConfig = new EksClusterEncryptionConfigOutputReference(this, "encryption_config", true);
        // kubernetes_network_config - computed: false, optional: true, required: false
        this._kubernetesNetworkConfig = new EksClusterKubernetesNetworkConfigOutputReference(this, "kubernetes_network_config", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksClusterTimeoutsOutputReference(this, "timeouts", true);
        // vpc_config - computed: false, optional: false, required: true
        this._vpcConfig = new EksClusterVpcConfigOutputReference(this, "vpc_config", true);
        this._enabledClusterLogTypes = config.enabledClusterLogTypes;
        this._name = config.name;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._version = config.version;
        this._encryptionConfig.internalValue = config.encryptionConfig;
        this._kubernetesNetworkConfig.internalValue = config.kubernetesNetworkConfig;
        this._timeouts.internalValue = config.timeouts;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_authority - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    certificateAuthority(index) {
        return new EksClusterCertificateAuthority(this, 'certificate_authority', index);
    }
    // created_at - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    /**
     * @stability stable
     */
    get enabledClusterLogTypes() {
        return this.getListAttribute('enabled_cluster_log_types');
    }
    /**
     * @stability stable
     */
    set enabledClusterLogTypes(value) {
        this._enabledClusterLogTypes = value;
    }
    /**
     * @stability stable
     */
    resetEnabledClusterLogTypes() {
        this._enabledClusterLogTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledClusterLogTypesInput() {
        return this._enabledClusterLogTypes;
    }
    // endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // identity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    identity(index) {
        return new EksClusterIdentity(this, 'identity', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // platform_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get version() {
        return this.getStringAttribute('version');
    }
    /**
     * @stability stable
     */
    set version(value) {
        this._version = value;
    }
    /**
     * @stability stable
     */
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get versionInput() {
        return this._version;
    }
    /**
     * @stability stable
     */
    get encryptionConfig() {
        return this._encryptionConfig;
    }
    /**
     * @stability stable
     */
    putEncryptionConfig(value) {
        this._encryptionConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetEncryptionConfig() {
        this._encryptionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get encryptionConfigInput() {
        return this._encryptionConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get kubernetesNetworkConfig() {
        return this._kubernetesNetworkConfig;
    }
    /**
     * @stability stable
     */
    putKubernetesNetworkConfig(value) {
        this._kubernetesNetworkConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetKubernetesNetworkConfig() {
        this._kubernetesNetworkConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kubernetesNetworkConfigInput() {
        return this._kubernetesNetworkConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    /**
     * @stability stable
     */
    get vpcConfig() {
        return this._vpcConfig;
    }
    /**
     * @stability stable
     */
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            enabled_cluster_log_types: cdktf.listMapper(cdktf.stringToTerraform)(this._enabledClusterLogTypes),
            name: cdktf.stringToTerraform(this._name),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            version: cdktf.stringToTerraform(this._version),
            encryption_config: eksClusterEncryptionConfigToTerraform(this._encryptionConfig.internalValue),
            kubernetes_network_config: eksClusterKubernetesNetworkConfigToTerraform(this._kubernetesNetworkConfig.internalValue),
            timeouts: eksClusterTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_config: eksClusterVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
}
exports.EksCluster = EksCluster;
_j = JSII_RTTI_SYMBOL_1;
EksCluster[_j] = { fqn: "@cdktf/provider-aws.eks.EksCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EksCluster.tfResourceType = "aws_eks_cluster";
