// @cdktf/provider-aws/lib/elb/data-aws-alb-listener.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAlbListener = exports.DataAwsAlbListenerDefaultAction = exports.DataAwsAlbListenerDefaultActionRedirect = exports.DataAwsAlbListenerDefaultActionForward = exports.DataAwsAlbListenerDefaultActionForwardTargetGroup = exports.DataAwsAlbListenerDefaultActionForwardStickiness = exports.DataAwsAlbListenerDefaultActionFixedResponse = exports.DataAwsAlbListenerDefaultActionAuthenticateOidc = exports.DataAwsAlbListenerDefaultActionAuthenticateCognito = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionAuthenticateCognito extends cdktf.ComplexComputedList {
    // authentication_request_extra_params - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authenticationRequestExtraParams() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('authentication_request_extra_params');
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // user_pool_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userPoolArn() {
        return this.getStringAttribute('user_pool_arn');
    }
    // user_pool_client_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userPoolClientId() {
        return this.getStringAttribute('user_pool_client_id');
    }
    // user_pool_domain - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userPoolDomain() {
        return this.getStringAttribute('user_pool_domain');
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateCognito = DataAwsAlbListenerDefaultActionAuthenticateCognito;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateCognito[_a] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionAuthenticateCognito", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionAuthenticateOidc extends cdktf.ComplexComputedList {
    // authentication_request_extra_params - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authenticationRequestExtraParams() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('authentication_request_extra_params');
    }
    // authorization_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    // client_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    // client_secret - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    // issuer - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // on_unauthenticated_request - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get onUnauthenticatedRequest() {
        return this.getStringAttribute('on_unauthenticated_request');
    }
    // scope - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get scope() {
        return this.getStringAttribute('scope');
    }
    // session_cookie_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sessionCookieName() {
        return this.getStringAttribute('session_cookie_name');
    }
    // session_timeout - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sessionTimeout() {
        return this.getNumberAttribute('session_timeout');
    }
    // token_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tokenEndpoint() {
        return this.getStringAttribute('token_endpoint');
    }
    // user_info_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userInfoEndpoint() {
        return this.getStringAttribute('user_info_endpoint');
    }
}
exports.DataAwsAlbListenerDefaultActionAuthenticateOidc = DataAwsAlbListenerDefaultActionAuthenticateOidc;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionAuthenticateOidc[_b] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionAuthenticateOidc", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionFixedResponse extends cdktf.ComplexComputedList {
    // content_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get contentType() {
        return this.getStringAttribute('content_type');
    }
    // message_body - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get messageBody() {
        return this.getStringAttribute('message_body');
    }
    // status_code - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsAlbListenerDefaultActionFixedResponse = DataAwsAlbListenerDefaultActionFixedResponse;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionFixedResponse[_c] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionFixedResponse", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionForwardStickiness extends cdktf.ComplexComputedList {
    // duration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get duration() {
        return this.getNumberAttribute('duration');
    }
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsAlbListenerDefaultActionForwardStickiness = DataAwsAlbListenerDefaultActionForwardStickiness;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardStickiness[_d] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionForwardStickiness", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionForwardTargetGroup extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // weight - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get weight() {
        return this.getNumberAttribute('weight');
    }
}
exports.DataAwsAlbListenerDefaultActionForwardTargetGroup = DataAwsAlbListenerDefaultActionForwardTargetGroup;
_e = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForwardTargetGroup[_e] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionForwardTargetGroup", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionForward extends cdktf.ComplexComputedList {
    // stickiness - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get stickiness() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('stickiness');
    }
    // target_group - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get targetGroup() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('target_group');
    }
}
exports.DataAwsAlbListenerDefaultActionForward = DataAwsAlbListenerDefaultActionForward;
_f = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionForward[_f] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionForward", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultActionRedirect extends cdktf.ComplexComputedList {
    // host - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get host() {
        return this.getStringAttribute('host');
    }
    // path - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get port() {
        return this.getStringAttribute('port');
    }
    // protocol - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // query - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get query() {
        return this.getStringAttribute('query');
    }
    // status_code - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get statusCode() {
        return this.getStringAttribute('status_code');
    }
}
exports.DataAwsAlbListenerDefaultActionRedirect = DataAwsAlbListenerDefaultActionRedirect;
_g = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultActionRedirect[_g] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultActionRedirect", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsAlbListenerDefaultAction extends cdktf.ComplexComputedList {
    // authenticate_cognito - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authenticateCognito() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('authenticate_cognito');
    }
    // authenticate_oidc - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get authenticateOidc() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('authenticate_oidc');
    }
    // fixed_response - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get fixedResponse() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('fixed_response');
    }
    // forward - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get forward() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('forward');
    }
    // order - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get order() {
        return this.getNumberAttribute('order');
    }
    // redirect - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get redirect() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('redirect');
    }
    // target_group_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get targetGroupArn() {
        return this.getStringAttribute('target_group_arn');
    }
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsAlbListenerDefaultAction = DataAwsAlbListenerDefaultAction;
_h = JSII_RTTI_SYMBOL_1;
DataAwsAlbListenerDefaultAction[_h] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListenerDefaultAction", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/alb_listener.html aws_alb_listener}.
 *
 * @stability stable
 */
class DataAwsAlbListener extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/alb_listener.html aws_alb_listener} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_alb_listener',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._loadBalancerArn = config.loadBalancerArn;
        this._port = config.port;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // alpn_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get alpnPolicy() {
        return this.getStringAttribute('alpn_policy');
    }
    // arn - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    // default_action - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    defaultAction(index) {
        return new DataAwsAlbListenerDefaultAction(this, 'default_action', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get loadBalancerArn() {
        return this.getStringAttribute('load_balancer_arn');
    }
    /**
     * @stability stable
     */
    set loadBalancerArn(value) {
        this._loadBalancerArn = value;
    }
    /**
     * @stability stable
     */
    resetLoadBalancerArn() {
        this._loadBalancerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerArnInput() {
        return this._loadBalancerArn;
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    /**
     * @stability stable
     */
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    // protocol - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // ssl_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sslPolicy() {
        return this.getStringAttribute('ssl_policy');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            load_balancer_arn: cdktf.stringToTerraform(this._loadBalancerArn),
            port: cdktf.numberToTerraform(this._port),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsAlbListener = DataAwsAlbListener;
_j = JSII_RTTI_SYMBOL_1;
DataAwsAlbListener[_j] = { fqn: "@cdktf/provider-aws.elb.DataAwsAlbListener", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsAlbListener.tfResourceType = "aws_alb_listener";
