// @cdktf/provider-aws/lib/elb/lb-ssl-negotiation-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LbSslNegotiationPolicy = exports.lbSslNegotiationPolicyAttributeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lbSslNegotiationPolicyAttributeToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.lbSslNegotiationPolicyAttributeToTerraform = lbSslNegotiationPolicyAttributeToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lb_ssl_negotiation_policy.html aws_lb_ssl_negotiation_policy}.
 *
 * @stability stable
 */
class LbSslNegotiationPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lb_ssl_negotiation_policy.html aws_lb_ssl_negotiation_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lb_ssl_negotiation_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._lbPort = config.lbPort;
        this._loadBalancer = config.loadBalancer;
        this._name = config.name;
        this._attribute = config.attribute;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get lbPort() {
        return this.getNumberAttribute('lb_port');
    }
    /**
     * @stability stable
     */
    set lbPort(value) {
        this._lbPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lbPortInput() {
        return this._lbPort;
    }
    /**
     * @stability stable
     */
    get loadBalancer() {
        return this.getStringAttribute('load_balancer');
    }
    /**
     * @stability stable
     */
    set loadBalancer(value) {
        this._loadBalancer = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerInput() {
        return this._loadBalancer;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get attribute() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('attribute');
    }
    /**
     * @stability stable
     */
    set attribute(value) {
        this._attribute = value;
    }
    /**
     * @stability stable
     */
    resetAttribute() {
        this._attribute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get attributeInput() {
        return this._attribute;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            lb_port: cdktf.numberToTerraform(this._lbPort),
            load_balancer: cdktf.stringToTerraform(this._loadBalancer),
            name: cdktf.stringToTerraform(this._name),
            attribute: cdktf.listMapper(lbSslNegotiationPolicyAttributeToTerraform)(this._attribute),
        };
    }
}
exports.LbSslNegotiationPolicy = LbSslNegotiationPolicy;
_a = JSII_RTTI_SYMBOL_1;
LbSslNegotiationPolicy[_a] = { fqn: "@cdktf/provider-aws.elb.LbSslNegotiationPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LbSslNegotiationPolicy.tfResourceType = "aws_lb_ssl_negotiation_policy";
