// @cdktf/provider-aws/lib/directconnect/data-aws-dx-location.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDxLocation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/dx_location.html aws_dx_location}.
 *
 * @stability stable
 */
class DataAwsDxLocation extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/dx_location.html aws_dx_location} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_location',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._locationCode = config.locationCode;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // available_port_speeds - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availablePortSpeeds() {
        return this.getListAttribute('available_port_speeds');
    }
    // available_providers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availableProviders() {
        return this.getListAttribute('available_providers');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get locationCode() {
        return this.getStringAttribute('location_code');
    }
    /**
     * @stability stable
     */
    set locationCode(value) {
        this._locationCode = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationCodeInput() {
        return this._locationCode;
    }
    // location_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get locationName() {
        return this.getStringAttribute('location_name');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            location_code: cdktf.stringToTerraform(this._locationCode),
        };
    }
}
exports.DataAwsDxLocation = DataAwsDxLocation;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDxLocation[_a] = { fqn: "@cdktf/provider-aws.directconnect.DataAwsDxLocation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsDxLocation.tfResourceType = "aws_dx_location";
