// @cdktf/provider-aws/lib/directconnect/dx-connection-association.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DxConnectionAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/dx_connection_association.html aws_dx_connection_association}.
 *
 * @stability stable
 */
class DxConnectionAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dx_connection_association.html aws_dx_connection_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_connection_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._connectionId = config.connectionId;
        this._lagId = config.lagId;
    }
    /**
     * @stability stable
     */
    get connectionId() {
        return this.getStringAttribute('connection_id');
    }
    /**
     * @stability stable
     */
    set connectionId(value) {
        this._connectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionIdInput() {
        return this._connectionId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get lagId() {
        return this.getStringAttribute('lag_id');
    }
    /**
     * @stability stable
     */
    set lagId(value) {
        this._lagId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lagIdInput() {
        return this._lagId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            connection_id: cdktf.stringToTerraform(this._connectionId),
            lag_id: cdktf.stringToTerraform(this._lagId),
        };
    }
}
exports.DxConnectionAssociation = DxConnectionAssociation;
_a = JSII_RTTI_SYMBOL_1;
DxConnectionAssociation[_a] = { fqn: "@cdktf/provider-aws.directconnect.DxConnectionAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DxConnectionAssociation.tfResourceType = "aws_dx_connection_association";
