// @cdktf/provider-aws/lib/directconnect/dx-gateway-association-proposal.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DxGatewayAssociationProposal = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/dx_gateway_association_proposal.html aws_dx_gateway_association_proposal}.
 *
 * @stability stable
 */
class DxGatewayAssociationProposal extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dx_gateway_association_proposal.html aws_dx_gateway_association_proposal} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_gateway_association_proposal',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._allowedPrefixes = config.allowedPrefixes;
        this._associatedGatewayId = config.associatedGatewayId;
        this._dxGatewayId = config.dxGatewayId;
        this._dxGatewayOwnerAccountId = config.dxGatewayOwnerAccountId;
    }
    /**
     * @stability stable
     */
    get allowedPrefixes() {
        return this.getListAttribute('allowed_prefixes');
    }
    /**
     * @stability stable
     */
    set allowedPrefixes(value) {
        this._allowedPrefixes = value;
    }
    /**
     * @stability stable
     */
    resetAllowedPrefixes() {
        this._allowedPrefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowedPrefixesInput() {
        return this._allowedPrefixes;
    }
    /**
     * @stability stable
     */
    get associatedGatewayId() {
        return this.getStringAttribute('associated_gateway_id');
    }
    /**
     * @stability stable
     */
    set associatedGatewayId(value) {
        this._associatedGatewayId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get associatedGatewayIdInput() {
        return this._associatedGatewayId;
    }
    // associated_gateway_owner_account_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associatedGatewayOwnerAccountId() {
        return this.getStringAttribute('associated_gateway_owner_account_id');
    }
    // associated_gateway_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associatedGatewayType() {
        return this.getStringAttribute('associated_gateway_type');
    }
    /**
     * @stability stable
     */
    get dxGatewayId() {
        return this.getStringAttribute('dx_gateway_id');
    }
    /**
     * @stability stable
     */
    set dxGatewayId(value) {
        this._dxGatewayId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dxGatewayIdInput() {
        return this._dxGatewayId;
    }
    /**
     * @stability stable
     */
    get dxGatewayOwnerAccountId() {
        return this.getStringAttribute('dx_gateway_owner_account_id');
    }
    /**
     * @stability stable
     */
    set dxGatewayOwnerAccountId(value) {
        this._dxGatewayOwnerAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dxGatewayOwnerAccountIdInput() {
        return this._dxGatewayOwnerAccountId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            allowed_prefixes: cdktf.listMapper(cdktf.stringToTerraform)(this._allowedPrefixes),
            associated_gateway_id: cdktf.stringToTerraform(this._associatedGatewayId),
            dx_gateway_id: cdktf.stringToTerraform(this._dxGatewayId),
            dx_gateway_owner_account_id: cdktf.stringToTerraform(this._dxGatewayOwnerAccountId),
        };
    }
}
exports.DxGatewayAssociationProposal = DxGatewayAssociationProposal;
_a = JSII_RTTI_SYMBOL_1;
DxGatewayAssociationProposal[_a] = { fqn: "@cdktf/provider-aws.directconnect.DxGatewayAssociationProposal", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DxGatewayAssociationProposal.tfResourceType = "aws_dx_gateway_association_proposal";
