// @cdktf/provider-aws/lib/storagegateway/storagegateway-stored-iscsi-volume.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoragegatewayStoredIscsiVolume = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/storagegateway_stored_iscsi_volume.html aws_storagegateway_stored_iscsi_volume}.
 *
 * @stability stable
 */
class StoragegatewayStoredIscsiVolume extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/storagegateway_stored_iscsi_volume.html aws_storagegateway_stored_iscsi_volume} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_storagegateway_stored_iscsi_volume',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._diskId = config.diskId;
        this._gatewayArn = config.gatewayArn;
        this._kmsEncrypted = config.kmsEncrypted;
        this._kmsKey = config.kmsKey;
        this._networkInterfaceId = config.networkInterfaceId;
        this._preserveExistingData = config.preserveExistingData;
        this._snapshotId = config.snapshotId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._targetName = config.targetName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // chap_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get chapEnabled() {
        return this.getBooleanAttribute('chap_enabled');
    }
    /**
     * @stability stable
     */
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    /**
     * @stability stable
     */
    set diskId(value) {
        this._diskId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get diskIdInput() {
        return this._diskId;
    }
    /**
     * @stability stable
     */
    get gatewayArn() {
        return this.getStringAttribute('gateway_arn');
    }
    /**
     * @stability stable
     */
    set gatewayArn(value) {
        this._gatewayArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gatewayArnInput() {
        return this._gatewayArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get kmsEncrypted() {
        return this.getBooleanAttribute('kms_encrypted');
    }
    /**
     * @stability stable
     */
    set kmsEncrypted(value) {
        this._kmsEncrypted = value;
    }
    /**
     * @stability stable
     */
    resetKmsEncrypted() {
        this._kmsEncrypted = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsEncryptedInput() {
        return this._kmsEncrypted;
    }
    /**
     * @stability stable
     */
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    /**
     * @stability stable
     */
    set kmsKey(value) {
        this._kmsKey = value;
    }
    /**
     * @stability stable
     */
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyInput() {
        return this._kmsKey;
    }
    // lun_number - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lunNumber() {
        return this.getNumberAttribute('lun_number');
    }
    /**
     * @stability stable
     */
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    /**
     * @stability stable
     */
    set networkInterfaceId(value) {
        this._networkInterfaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkInterfaceIdInput() {
        return this._networkInterfaceId;
    }
    // network_interface_port - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkInterfacePort() {
        return this.getNumberAttribute('network_interface_port');
    }
    /**
     * @stability stable
     */
    get preserveExistingData() {
        return this.getBooleanAttribute('preserve_existing_data');
    }
    /**
     * @stability stable
     */
    set preserveExistingData(value) {
        this._preserveExistingData = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get preserveExistingDataInput() {
        return this._preserveExistingData;
    }
    /**
     * @stability stable
     */
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    /**
     * @stability stable
     */
    set snapshotId(value) {
        this._snapshotId = value;
    }
    /**
     * @stability stable
     */
    resetSnapshotId() {
        this._snapshotId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snapshotIdInput() {
        return this._snapshotId;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // target_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    /**
     * @stability stable
     */
    get targetName() {
        return this.getStringAttribute('target_name');
    }
    /**
     * @stability stable
     */
    set targetName(value) {
        this._targetName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetNameInput() {
        return this._targetName;
    }
    // volume_attachment_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeAttachmentStatus() {
        return this.getStringAttribute('volume_attachment_status');
    }
    // volume_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeId() {
        return this.getStringAttribute('volume_id');
    }
    // volume_size_in_bytes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeSizeInBytes() {
        return this.getNumberAttribute('volume_size_in_bytes');
    }
    // volume_status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeStatus() {
        return this.getStringAttribute('volume_status');
    }
    // volume_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            disk_id: cdktf.stringToTerraform(this._diskId),
            gateway_arn: cdktf.stringToTerraform(this._gatewayArn),
            kms_encrypted: cdktf.booleanToTerraform(this._kmsEncrypted),
            kms_key: cdktf.stringToTerraform(this._kmsKey),
            network_interface_id: cdktf.stringToTerraform(this._networkInterfaceId),
            preserve_existing_data: cdktf.booleanToTerraform(this._preserveExistingData),
            snapshot_id: cdktf.stringToTerraform(this._snapshotId),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            target_name: cdktf.stringToTerraform(this._targetName),
        };
    }
}
exports.StoragegatewayStoredIscsiVolume = StoragegatewayStoredIscsiVolume;
_a = JSII_RTTI_SYMBOL_1;
StoragegatewayStoredIscsiVolume[_a] = { fqn: "@cdktf/provider-aws.storagegateway.StoragegatewayStoredIscsiVolume", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
StoragegatewayStoredIscsiVolume.tfResourceType = "aws_storagegateway_stored_iscsi_volume";
