// @cdktf/provider-aws/lib/storagegateway/storagegateway-working-storage.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StoragegatewayWorkingStorage = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/storagegateway_working_storage.html aws_storagegateway_working_storage}.
 *
 * @stability stable
 */
class StoragegatewayWorkingStorage extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/storagegateway_working_storage.html aws_storagegateway_working_storage} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_storagegateway_working_storage',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._diskId = config.diskId;
        this._gatewayArn = config.gatewayArn;
    }
    /**
     * @stability stable
     */
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    /**
     * @stability stable
     */
    set diskId(value) {
        this._diskId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get diskIdInput() {
        return this._diskId;
    }
    /**
     * @stability stable
     */
    get gatewayArn() {
        return this.getStringAttribute('gateway_arn');
    }
    /**
     * @stability stable
     */
    set gatewayArn(value) {
        this._gatewayArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get gatewayArnInput() {
        return this._gatewayArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            disk_id: cdktf.stringToTerraform(this._diskId),
            gateway_arn: cdktf.stringToTerraform(this._gatewayArn),
        };
    }
}
exports.StoragegatewayWorkingStorage = StoragegatewayWorkingStorage;
_a = JSII_RTTI_SYMBOL_1;
StoragegatewayWorkingStorage[_a] = { fqn: "@cdktf/provider-aws.storagegateway.StoragegatewayWorkingStorage", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
StoragegatewayWorkingStorage.tfResourceType = "aws_storagegateway_working_storage";
