// @cdktf/provider-aws/lib/securityhub/securityhub-standards-subscription.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubStandardsSubscription = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_subscription.html aws_securityhub_standards_subscription}.
 *
 * @stability stable
 */
class SecurityhubStandardsSubscription extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_subscription.html aws_securityhub_standards_subscription} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_standards_subscription',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._standardsArn = config.standardsArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get standardsArn() {
        return this.getStringAttribute('standards_arn');
    }
    /**
     * @stability stable
     */
    set standardsArn(value) {
        this._standardsArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get standardsArnInput() {
        return this._standardsArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            standards_arn: cdktf.stringToTerraform(this._standardsArn),
        };
    }
}
exports.SecurityhubStandardsSubscription = SecurityhubStandardsSubscription;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubStandardsSubscription[_a] = { fqn: "@cdktf/provider-aws.securityhub.SecurityhubStandardsSubscription", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecurityhubStandardsSubscription.tfResourceType = "aws_securityhub_standards_subscription";
