// @cdktf/provider-aws/lib/securityhub/securityhub-organization-admin-account.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubOrganizationAdminAccount = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_organization_admin_account.html aws_securityhub_organization_admin_account}.
 *
 * @stability stable
 */
class SecurityhubOrganizationAdminAccount extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_organization_admin_account.html aws_securityhub_organization_admin_account} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_organization_admin_account',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._adminAccountId = config.adminAccountId;
    }
    /**
     * @stability stable
     */
    get adminAccountId() {
        return this.getStringAttribute('admin_account_id');
    }
    /**
     * @stability stable
     */
    set adminAccountId(value) {
        this._adminAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get adminAccountIdInput() {
        return this._adminAccountId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            admin_account_id: cdktf.stringToTerraform(this._adminAccountId),
        };
    }
}
exports.SecurityhubOrganizationAdminAccount = SecurityhubOrganizationAdminAccount;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubOrganizationAdminAccount[_a] = { fqn: "@cdktf/provider-aws.securityhub.SecurityhubOrganizationAdminAccount", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecurityhubOrganizationAdminAccount.tfResourceType = "aws_securityhub_organization_admin_account";
