// @cdktf/provider-aws/lib/securityhub/securityhub-finding-aggregator.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubFindingAggregator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_finding_aggregator.html aws_securityhub_finding_aggregator}.
 *
 * @stability stable
 */
class SecurityhubFindingAggregator extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_finding_aggregator.html aws_securityhub_finding_aggregator} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_finding_aggregator',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._linkingMode = config.linkingMode;
        this._specifiedRegions = config.specifiedRegions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get linkingMode() {
        return this.getStringAttribute('linking_mode');
    }
    /**
     * @stability stable
     */
    set linkingMode(value) {
        this._linkingMode = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get linkingModeInput() {
        return this._linkingMode;
    }
    /**
     * @stability stable
     */
    get specifiedRegions() {
        return this.getListAttribute('specified_regions');
    }
    /**
     * @stability stable
     */
    set specifiedRegions(value) {
        this._specifiedRegions = value;
    }
    /**
     * @stability stable
     */
    resetSpecifiedRegions() {
        this._specifiedRegions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get specifiedRegionsInput() {
        return this._specifiedRegions;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            linking_mode: cdktf.stringToTerraform(this._linkingMode),
            specified_regions: cdktf.listMapper(cdktf.stringToTerraform)(this._specifiedRegions),
        };
    }
}
exports.SecurityhubFindingAggregator = SecurityhubFindingAggregator;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubFindingAggregator[_a] = { fqn: "@cdktf/provider-aws.securityhub.SecurityhubFindingAggregator", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SecurityhubFindingAggregator.tfResourceType = "aws_securityhub_finding_aggregator";
