// @cdktf/provider-aws/lib/ssoadmin/ssoadmin-permission-set-inline-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminPermissionSetInlinePolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_permission_set_inline_policy.html aws_ssoadmin_permission_set_inline_policy}.
 *
 * @stability stable
 */
class SsoadminPermissionSetInlinePolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_permission_set_inline_policy.html aws_ssoadmin_permission_set_inline_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_permission_set_inline_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._inlinePolicy = config.inlinePolicy;
        this._instanceArn = config.instanceArn;
        this._permissionSetArn = config.permissionSetArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get inlinePolicy() {
        return this.getStringAttribute('inline_policy');
    }
    /**
     * @stability stable
     */
    set inlinePolicy(value) {
        this._inlinePolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inlinePolicyInput() {
        return this._inlinePolicy;
    }
    /**
     * @stability stable
     */
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    /**
     * @stability stable
     */
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceArnInput() {
        return this._instanceArn;
    }
    /**
     * @stability stable
     */
    get permissionSetArn() {
        return this.getStringAttribute('permission_set_arn');
    }
    /**
     * @stability stable
     */
    set permissionSetArn(value) {
        this._permissionSetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get permissionSetArnInput() {
        return this._permissionSetArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            inline_policy: cdktf.stringToTerraform(this._inlinePolicy),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            permission_set_arn: cdktf.stringToTerraform(this._permissionSetArn),
        };
    }
}
exports.SsoadminPermissionSetInlinePolicy = SsoadminPermissionSetInlinePolicy;
_a = JSII_RTTI_SYMBOL_1;
SsoadminPermissionSetInlinePolicy[_a] = { fqn: "@cdktf/provider-aws.ssoadmin.SsoadminPermissionSetInlinePolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsoadminPermissionSetInlinePolicy.tfResourceType = "aws_ssoadmin_permission_set_inline_policy";
