// @cdktf/provider-aws/lib/serverlessapplicationrepository/data-aws-serverlessapplicationrepository-application.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServerlessapplicationrepositoryApplication = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/serverlessapplicationrepository_application.html aws_serverlessapplicationrepository_application}.
 *
 * @stability stable
 */
class DataAwsServerlessapplicationrepositoryApplication extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/serverlessapplicationrepository_application.html aws_serverlessapplicationrepository_application} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_serverlessapplicationrepository_application',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._applicationId = config.applicationId;
        this._semanticVersion = config.semanticVersion;
    }
    /**
     * @stability stable
     */
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    /**
     * @stability stable
     */
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applicationIdInput() {
        return this._applicationId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // required_capabilities - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get requiredCapabilities() {
        return this.getListAttribute('required_capabilities');
    }
    /**
     * @stability stable
     */
    get semanticVersion() {
        return this.getStringAttribute('semantic_version');
    }
    /**
     * @stability stable
     */
    set semanticVersion(value) {
        this._semanticVersion = value;
    }
    /**
     * @stability stable
     */
    resetSemanticVersion() {
        this._semanticVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get semanticVersionInput() {
        return this._semanticVersion;
    }
    // source_code_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sourceCodeUrl() {
        return this.getStringAttribute('source_code_url');
    }
    // template_url - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            semantic_version: cdktf.stringToTerraform(this._semanticVersion),
        };
    }
}
exports.DataAwsServerlessapplicationrepositoryApplication = DataAwsServerlessapplicationrepositoryApplication;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServerlessapplicationrepositoryApplication[_a] = { fqn: "@cdktf/provider-aws.serverlessapplicationrepository.DataAwsServerlessapplicationrepositoryApplication", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsServerlessapplicationrepositoryApplication.tfResourceType = "aws_serverlessapplicationrepository_application";
