// @cdktf/provider-aws/lib/ecs/ecs-cluster.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcsCluster = exports.ecsClusterSettingToTerraform = exports.ecsClusterDefaultCapacityProviderStrategyToTerraform = exports.EcsClusterConfigurationOutputReference = exports.ecsClusterConfigurationToTerraform = exports.EcsClusterConfigurationExecuteCommandConfigurationOutputReference = exports.ecsClusterConfigurationExecuteCommandConfigurationToTerraform = exports.EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference = exports.ecsClusterConfigurationExecuteCommandConfigurationLogConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecsClusterConfigurationExecuteCommandConfigurationLogConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloud_watch_encryption_enabled: cdktf.booleanToTerraform(struct.cloudWatchEncryptionEnabled),
        cloud_watch_log_group_name: cdktf.stringToTerraform(struct.cloudWatchLogGroupName),
        s3_bucket_encryption_enabled: cdktf.booleanToTerraform(struct.s3BucketEncryptionEnabled),
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_key_prefix: cdktf.stringToTerraform(struct.s3KeyPrefix),
    };
}
exports.ecsClusterConfigurationExecuteCommandConfigurationLogConfigurationToTerraform = ecsClusterConfigurationExecuteCommandConfigurationLogConfigurationToTerraform;
/**
 * @stability stable
 */
class EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudWatchEncryptionEnabled) {
            hasAnyValues = true;
            internalValueResult.cloudWatchEncryptionEnabled = this._cloudWatchEncryptionEnabled;
        }
        if (this._cloudWatchLogGroupName) {
            hasAnyValues = true;
            internalValueResult.cloudWatchLogGroupName = this._cloudWatchLogGroupName;
        }
        if (this._s3BucketEncryptionEnabled) {
            hasAnyValues = true;
            internalValueResult.s3BucketEncryptionEnabled = this._s3BucketEncryptionEnabled;
        }
        if (this._s3BucketName) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3KeyPrefix) {
            hasAnyValues = true;
            internalValueResult.s3KeyPrefix = this._s3KeyPrefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudWatchEncryptionEnabled = undefined;
            this._cloudWatchLogGroupName = undefined;
            this._s3BucketEncryptionEnabled = undefined;
            this._s3BucketName = undefined;
            this._s3KeyPrefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudWatchEncryptionEnabled = value.cloudWatchEncryptionEnabled;
            this._cloudWatchLogGroupName = value.cloudWatchLogGroupName;
            this._s3BucketEncryptionEnabled = value.s3BucketEncryptionEnabled;
            this._s3BucketName = value.s3BucketName;
            this._s3KeyPrefix = value.s3KeyPrefix;
        }
    }
    /**
     * @stability stable
     */
    get cloudWatchEncryptionEnabled() {
        return this.getBooleanAttribute('cloud_watch_encryption_enabled');
    }
    /**
     * @stability stable
     */
    set cloudWatchEncryptionEnabled(value) {
        this._cloudWatchEncryptionEnabled = value;
    }
    /**
     * @stability stable
     */
    resetCloudWatchEncryptionEnabled() {
        this._cloudWatchEncryptionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudWatchEncryptionEnabledInput() {
        return this._cloudWatchEncryptionEnabled;
    }
    /**
     * @stability stable
     */
    get cloudWatchLogGroupName() {
        return this.getStringAttribute('cloud_watch_log_group_name');
    }
    /**
     * @stability stable
     */
    set cloudWatchLogGroupName(value) {
        this._cloudWatchLogGroupName = value;
    }
    /**
     * @stability stable
     */
    resetCloudWatchLogGroupName() {
        this._cloudWatchLogGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudWatchLogGroupNameInput() {
        return this._cloudWatchLogGroupName;
    }
    /**
     * @stability stable
     */
    get s3BucketEncryptionEnabled() {
        return this.getBooleanAttribute('s3_bucket_encryption_enabled');
    }
    /**
     * @stability stable
     */
    set s3BucketEncryptionEnabled(value) {
        this._s3BucketEncryptionEnabled = value;
    }
    /**
     * @stability stable
     */
    resetS3BucketEncryptionEnabled() {
        this._s3BucketEncryptionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketEncryptionEnabledInput() {
        return this._s3BucketEncryptionEnabled;
    }
    /**
     * @stability stable
     */
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    /**
     * @stability stable
     */
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    /**
     * @stability stable
     */
    resetS3BucketName() {
        this._s3BucketName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    /**
     * @stability stable
     */
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    /**
     * @stability stable
     */
    set s3KeyPrefix(value) {
        this._s3KeyPrefix = value;
    }
    /**
     * @stability stable
     */
    resetS3KeyPrefix() {
        this._s3KeyPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KeyPrefixInput() {
        return this._s3KeyPrefix;
    }
}
exports.EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference = EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecs.EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference", version: "3.0.1" };
function ecsClusterConfigurationExecuteCommandConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        logging: cdktf.stringToTerraform(struct.logging),
        log_configuration: ecsClusterConfigurationExecuteCommandConfigurationLogConfigurationToTerraform(struct.logConfiguration),
    };
}
exports.ecsClusterConfigurationExecuteCommandConfigurationToTerraform = ecsClusterConfigurationExecuteCommandConfigurationToTerraform;
/**
 * @stability stable
 */
class EcsClusterConfigurationExecuteCommandConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // log_configuration - computed: false, optional: true, required: false
        this._logConfiguration = new EcsClusterConfigurationExecuteCommandConfigurationLogConfigurationOutputReference(this, "log_configuration", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._logging) {
            hasAnyValues = true;
            internalValueResult.logging = this._logging;
        }
        if ((_e = this._logConfiguration) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.logConfiguration = (_f = this._logConfiguration) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._logging = undefined;
            this._logConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._logging = value.logging;
            this._logConfiguration.internalValue = value.logConfiguration;
        }
    }
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    /**
     * @stability stable
     */
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    /**
     * @stability stable
     */
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    /**
     * @stability stable
     */
    get logging() {
        return this.getStringAttribute('logging');
    }
    /**
     * @stability stable
     */
    set logging(value) {
        this._logging = value;
    }
    /**
     * @stability stable
     */
    resetLogging() {
        this._logging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loggingInput() {
        return this._logging;
    }
    /**
     * @stability stable
     */
    get logConfiguration() {
        return this._logConfiguration;
    }
    /**
     * @stability stable
     */
    putLogConfiguration(value) {
        this._logConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLogConfiguration() {
        this._logConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get logConfigurationInput() {
        return this._logConfiguration.internalValue;
    }
}
exports.EcsClusterConfigurationExecuteCommandConfigurationOutputReference = EcsClusterConfigurationExecuteCommandConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EcsClusterConfigurationExecuteCommandConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.ecs.EcsClusterConfigurationExecuteCommandConfigurationOutputReference", version: "3.0.1" };
function ecsClusterConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execute_command_configuration: ecsClusterConfigurationExecuteCommandConfigurationToTerraform(struct.executeCommandConfiguration),
    };
}
exports.ecsClusterConfigurationToTerraform = ecsClusterConfigurationToTerraform;
/**
 * @stability stable
 */
class EcsClusterConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // execute_command_configuration - computed: false, optional: true, required: false
        this._executeCommandConfiguration = new EcsClusterConfigurationExecuteCommandConfigurationOutputReference(this, "execute_command_configuration", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_e = this._executeCommandConfiguration) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.executeCommandConfiguration = (_f = this._executeCommandConfiguration) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._executeCommandConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._executeCommandConfiguration.internalValue = value.executeCommandConfiguration;
        }
    }
    /**
     * @stability stable
     */
    get executeCommandConfiguration() {
        return this._executeCommandConfiguration;
    }
    /**
     * @stability stable
     */
    putExecuteCommandConfiguration(value) {
        this._executeCommandConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetExecuteCommandConfiguration() {
        this._executeCommandConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executeCommandConfigurationInput() {
        return this._executeCommandConfiguration.internalValue;
    }
}
exports.EcsClusterConfigurationOutputReference = EcsClusterConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EcsClusterConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.ecs.EcsClusterConfigurationOutputReference", version: "3.0.1" };
function ecsClusterDefaultCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.ecsClusterDefaultCapacityProviderStrategyToTerraform = ecsClusterDefaultCapacityProviderStrategyToTerraform;
function ecsClusterSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.ecsClusterSettingToTerraform = ecsClusterSettingToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ecs_cluster.html aws_ecs_cluster}.
 *
 * @stability stable
 */
class EcsCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ecs_cluster.html aws_ecs_cluster} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // configuration - computed: false, optional: true, required: false
        this._configuration = new EcsClusterConfigurationOutputReference(this, "configuration", true);
        this._capacityProviders = config.capacityProviders;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._configuration.internalValue = config.configuration;
        this._defaultCapacityProviderStrategy = config.defaultCapacityProviderStrategy;
        this._setting = config.setting;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get capacityProviders() {
        return this.getListAttribute('capacity_providers');
    }
    /**
     * @stability stable
     */
    set capacityProviders(value) {
        this._capacityProviders = value;
    }
    /**
     * @stability stable
     */
    resetCapacityProviders() {
        this._capacityProviders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityProvidersInput() {
        return this._capacityProviders;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get configuration() {
        return this._configuration;
    }
    /**
     * @stability stable
     */
    putConfiguration(value) {
        this._configuration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetConfiguration() {
        this._configuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get configurationInput() {
        return this._configuration.internalValue;
    }
    /**
     * @stability stable
     */
    get defaultCapacityProviderStrategy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('default_capacity_provider_strategy');
    }
    /**
     * @stability stable
     */
    set defaultCapacityProviderStrategy(value) {
        this._defaultCapacityProviderStrategy = value;
    }
    /**
     * @stability stable
     */
    resetDefaultCapacityProviderStrategy() {
        this._defaultCapacityProviderStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultCapacityProviderStrategyInput() {
        return this._defaultCapacityProviderStrategy;
    }
    /**
     * @stability stable
     */
    get setting() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('setting');
    }
    /**
     * @stability stable
     */
    set setting(value) {
        this._setting = value;
    }
    /**
     * @stability stable
     */
    resetSetting() {
        this._setting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get settingInput() {
        return this._setting;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            capacity_providers: cdktf.listMapper(cdktf.stringToTerraform)(this._capacityProviders),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            configuration: ecsClusterConfigurationToTerraform(this._configuration.internalValue),
            default_capacity_provider_strategy: cdktf.listMapper(ecsClusterDefaultCapacityProviderStrategyToTerraform)(this._defaultCapacityProviderStrategy),
            setting: cdktf.listMapper(ecsClusterSettingToTerraform)(this._setting),
        };
    }
}
exports.EcsCluster = EcsCluster;
_d = JSII_RTTI_SYMBOL_1;
EcsCluster[_d] = { fqn: "@cdktf/provider-aws.ecs.EcsCluster", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EcsCluster.tfResourceType = "aws_ecs_cluster";
