// @cdktf/provider-aws/lib/directconnect/dx-hosted-connection.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DxHostedConnection = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/dx_hosted_connection.html aws_dx_hosted_connection}.
 *
 * @stability stable
 */
class DxHostedConnection extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dx_hosted_connection.html aws_dx_hosted_connection} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_hosted_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._bandwidth = config.bandwidth;
        this._connectionId = config.connectionId;
        this._name = config.name;
        this._ownerAccountId = config.ownerAccountId;
        this._vlan = config.vlan;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // aws_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get awsDevice() {
        return this.getStringAttribute('aws_device');
    }
    /**
     * @stability stable
     */
    get bandwidth() {
        return this.getStringAttribute('bandwidth');
    }
    /**
     * @stability stable
     */
    set bandwidth(value) {
        this._bandwidth = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bandwidthInput() {
        return this._bandwidth;
    }
    /**
     * @stability stable
     */
    get connectionId() {
        return this.getStringAttribute('connection_id');
    }
    /**
     * @stability stable
     */
    set connectionId(value) {
        this._connectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionIdInput() {
        return this._connectionId;
    }
    // has_logical_redundancy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hasLogicalRedundancy() {
        return this.getStringAttribute('has_logical_redundancy');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // jumbo_frame_capable - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get jumboFrameCapable() {
        return this.getBooleanAttribute('jumbo_frame_capable');
    }
    // lag_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lagId() {
        return this.getStringAttribute('lag_id');
    }
    // loa_issue_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get loaIssueTime() {
        return this.getStringAttribute('loa_issue_time');
    }
    // location - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    /**
     * @stability stable
     */
    set ownerAccountId(value) {
        this._ownerAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerAccountIdInput() {
        return this._ownerAccountId;
    }
    // partner_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get partnerName() {
        return this.getStringAttribute('partner_name');
    }
    // provider_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get providerName() {
        return this.getStringAttribute('provider_name');
    }
    // region - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
    // state - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get state() {
        return this.getStringAttribute('state');
    }
    /**
     * @stability stable
     */
    get vlan() {
        return this.getNumberAttribute('vlan');
    }
    /**
     * @stability stable
     */
    set vlan(value) {
        this._vlan = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vlanInput() {
        return this._vlan;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bandwidth: cdktf.stringToTerraform(this._bandwidth),
            connection_id: cdktf.stringToTerraform(this._connectionId),
            name: cdktf.stringToTerraform(this._name),
            owner_account_id: cdktf.stringToTerraform(this._ownerAccountId),
            vlan: cdktf.numberToTerraform(this._vlan),
        };
    }
}
exports.DxHostedConnection = DxHostedConnection;
_a = JSII_RTTI_SYMBOL_1;
DxHostedConnection[_a] = { fqn: "@cdktf/provider-aws.directconnect.DxHostedConnection", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DxHostedConnection.tfResourceType = "aws_dx_hosted_connection";
