// @cdktf/provider-aws/lib/apigateway/api-gateway-integration.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayIntegration = exports.ApiGatewayIntegrationTlsConfigOutputReference = exports.apiGatewayIntegrationTlsConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apiGatewayIntegrationTlsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        insecure_skip_verification: cdktf.booleanToTerraform(struct.insecureSkipVerification),
    };
}
exports.apiGatewayIntegrationTlsConfigToTerraform = apiGatewayIntegrationTlsConfigToTerraform;
/**
 * @stability stable
 */
class ApiGatewayIntegrationTlsConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._insecureSkipVerification) {
            hasAnyValues = true;
            internalValueResult.insecureSkipVerification = this._insecureSkipVerification;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._insecureSkipVerification = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._insecureSkipVerification = value.insecureSkipVerification;
        }
    }
    /**
     * @stability stable
     */
    get insecureSkipVerification() {
        return this.getBooleanAttribute('insecure_skip_verification');
    }
    /**
     * @stability stable
     */
    set insecureSkipVerification(value) {
        this._insecureSkipVerification = value;
    }
    /**
     * @stability stable
     */
    resetInsecureSkipVerification() {
        this._insecureSkipVerification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get insecureSkipVerificationInput() {
        return this._insecureSkipVerification;
    }
}
exports.ApiGatewayIntegrationTlsConfigOutputReference = ApiGatewayIntegrationTlsConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayIntegrationTlsConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayIntegrationTlsConfigOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_integration.html aws_api_gateway_integration}.
 *
 * @stability stable
 */
class ApiGatewayIntegration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_integration.html aws_api_gateway_integration} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_integration',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // tls_config - computed: false, optional: true, required: false
        this._tlsConfig = new ApiGatewayIntegrationTlsConfigOutputReference(this, "tls_config", true);
        this._cacheKeyParameters = config.cacheKeyParameters;
        this._cacheNamespace = config.cacheNamespace;
        this._connectionId = config.connectionId;
        this._connectionType = config.connectionType;
        this._contentHandling = config.contentHandling;
        this._credentials = config.credentials;
        this._httpMethod = config.httpMethod;
        this._integrationHttpMethod = config.integrationHttpMethod;
        this._passthroughBehavior = config.passthroughBehavior;
        this._requestParameters = config.requestParameters;
        this._requestTemplates = config.requestTemplates;
        this._resourceId = config.resourceId;
        this._restApiId = config.restApiId;
        this._timeoutMilliseconds = config.timeoutMilliseconds;
        this._type = config.type;
        this._uri = config.uri;
        this._tlsConfig.internalValue = config.tlsConfig;
    }
    /**
     * @stability stable
     */
    get cacheKeyParameters() {
        return this.getListAttribute('cache_key_parameters');
    }
    /**
     * @stability stable
     */
    set cacheKeyParameters(value) {
        this._cacheKeyParameters = value;
    }
    /**
     * @stability stable
     */
    resetCacheKeyParameters() {
        this._cacheKeyParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cacheKeyParametersInput() {
        return this._cacheKeyParameters;
    }
    /**
     * @stability stable
     */
    get cacheNamespace() {
        return this.getStringAttribute('cache_namespace');
    }
    /**
     * @stability stable
     */
    set cacheNamespace(value) {
        this._cacheNamespace = value;
    }
    /**
     * @stability stable
     */
    resetCacheNamespace() {
        this._cacheNamespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cacheNamespaceInput() {
        return this._cacheNamespace;
    }
    /**
     * @stability stable
     */
    get connectionId() {
        return this.getStringAttribute('connection_id');
    }
    /**
     * @stability stable
     */
    set connectionId(value) {
        this._connectionId = value;
    }
    /**
     * @stability stable
     */
    resetConnectionId() {
        this._connectionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionIdInput() {
        return this._connectionId;
    }
    /**
     * @stability stable
     */
    get connectionType() {
        return this.getStringAttribute('connection_type');
    }
    /**
     * @stability stable
     */
    set connectionType(value) {
        this._connectionType = value;
    }
    /**
     * @stability stable
     */
    resetConnectionType() {
        this._connectionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get connectionTypeInput() {
        return this._connectionType;
    }
    /**
     * @stability stable
     */
    get contentHandling() {
        return this.getStringAttribute('content_handling');
    }
    /**
     * @stability stable
     */
    set contentHandling(value) {
        this._contentHandling = value;
    }
    /**
     * @stability stable
     */
    resetContentHandling() {
        this._contentHandling = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get contentHandlingInput() {
        return this._contentHandling;
    }
    /**
     * @stability stable
     */
    get credentials() {
        return this.getStringAttribute('credentials');
    }
    /**
     * @stability stable
     */
    set credentials(value) {
        this._credentials = value;
    }
    /**
     * @stability stable
     */
    resetCredentials() {
        this._credentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get credentialsInput() {
        return this._credentials;
    }
    /**
     * @stability stable
     */
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    /**
     * @stability stable
     */
    set httpMethod(value) {
        this._httpMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get httpMethodInput() {
        return this._httpMethod;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get integrationHttpMethod() {
        return this.getStringAttribute('integration_http_method');
    }
    /**
     * @stability stable
     */
    set integrationHttpMethod(value) {
        this._integrationHttpMethod = value;
    }
    /**
     * @stability stable
     */
    resetIntegrationHttpMethod() {
        this._integrationHttpMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get integrationHttpMethodInput() {
        return this._integrationHttpMethod;
    }
    /**
     * @stability stable
     */
    get passthroughBehavior() {
        return this.getStringAttribute('passthrough_behavior');
    }
    /**
     * @stability stable
     */
    set passthroughBehavior(value) {
        this._passthroughBehavior = value;
    }
    /**
     * @stability stable
     */
    resetPassthroughBehavior() {
        this._passthroughBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get passthroughBehaviorInput() {
        return this._passthroughBehavior;
    }
    /**
     * @stability stable
     */
    get requestParameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_parameters');
    }
    /**
     * @stability stable
     */
    set requestParameters(value) {
        this._requestParameters = value;
    }
    /**
     * @stability stable
     */
    resetRequestParameters() {
        this._requestParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestParametersInput() {
        return this._requestParameters;
    }
    /**
     * @stability stable
     */
    get requestTemplates() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('request_templates');
    }
    /**
     * @stability stable
     */
    set requestTemplates(value) {
        this._requestTemplates = value;
    }
    /**
     * @stability stable
     */
    resetRequestTemplates() {
        this._requestTemplates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestTemplatesInput() {
        return this._requestTemplates;
    }
    /**
     * @stability stable
     */
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    /**
     * @stability stable
     */
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceIdInput() {
        return this._resourceId;
    }
    /**
     * @stability stable
     */
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    /**
     * @stability stable
     */
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get restApiIdInput() {
        return this._restApiId;
    }
    /**
     * @stability stable
     */
    get timeoutMilliseconds() {
        return this.getNumberAttribute('timeout_milliseconds');
    }
    /**
     * @stability stable
     */
    set timeoutMilliseconds(value) {
        this._timeoutMilliseconds = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutMilliseconds() {
        this._timeoutMilliseconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutMillisecondsInput() {
        return this._timeoutMilliseconds;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    /**
     * @stability stable
     */
    get uri() {
        return this.getStringAttribute('uri');
    }
    /**
     * @stability stable
     */
    set uri(value) {
        this._uri = value;
    }
    /**
     * @stability stable
     */
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get uriInput() {
        return this._uri;
    }
    /**
     * @stability stable
     */
    get tlsConfig() {
        return this._tlsConfig;
    }
    /**
     * @stability stable
     */
    putTlsConfig(value) {
        this._tlsConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTlsConfig() {
        this._tlsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsConfigInput() {
        return this._tlsConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cache_key_parameters: cdktf.listMapper(cdktf.stringToTerraform)(this._cacheKeyParameters),
            cache_namespace: cdktf.stringToTerraform(this._cacheNamespace),
            connection_id: cdktf.stringToTerraform(this._connectionId),
            connection_type: cdktf.stringToTerraform(this._connectionType),
            content_handling: cdktf.stringToTerraform(this._contentHandling),
            credentials: cdktf.stringToTerraform(this._credentials),
            http_method: cdktf.stringToTerraform(this._httpMethod),
            integration_http_method: cdktf.stringToTerraform(this._integrationHttpMethod),
            passthrough_behavior: cdktf.stringToTerraform(this._passthroughBehavior),
            request_parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._requestParameters),
            request_templates: cdktf.hashMapper(cdktf.anyToTerraform)(this._requestTemplates),
            resource_id: cdktf.stringToTerraform(this._resourceId),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            timeout_milliseconds: cdktf.numberToTerraform(this._timeoutMilliseconds),
            type: cdktf.stringToTerraform(this._type),
            uri: cdktf.stringToTerraform(this._uri),
            tls_config: apiGatewayIntegrationTlsConfigToTerraform(this._tlsConfig.internalValue),
        };
    }
}
exports.ApiGatewayIntegration = ApiGatewayIntegration;
_b = JSII_RTTI_SYMBOL_1;
ApiGatewayIntegration[_b] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayIntegration", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayIntegration.tfResourceType = "aws_api_gateway_integration";
