// @cdktf/provider-aws/lib/codestar/codestarconnections-host.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodestarconnectionsHost = exports.CodestarconnectionsHostVpcConfigurationOutputReference = exports.codestarconnectionsHostVpcConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codestarconnectionsHostVpcConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnetIds),
        tls_certificate: cdktf.stringToTerraform(struct.tlsCertificate),
        vpc_id: cdktf.stringToTerraform(struct.vpcId),
    };
}
exports.codestarconnectionsHostVpcConfigurationToTerraform = codestarconnectionsHostVpcConfigurationToTerraform;
/**
 * @stability stable
 */
class CodestarconnectionsHostVpcConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        if (this._tlsCertificate) {
            hasAnyValues = true;
            internalValueResult.tlsCertificate = this._tlsCertificate;
        }
        if (this._vpcId) {
            hasAnyValues = true;
            internalValueResult.vpcId = this._vpcId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
            this._tlsCertificate = undefined;
            this._vpcId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
            this._tlsCertificate = value.tlsCertificate;
            this._vpcId = value.vpcId;
        }
    }
    /**
     * @stability stable
     */
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    /**
     * @stability stable
     */
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    /**
     * @stability stable
     */
    get subnetIds() {
        return this.getListAttribute('subnet_ids');
    }
    /**
     * @stability stable
     */
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetIdsInput() {
        return this._subnetIds;
    }
    /**
     * @stability stable
     */
    get tlsCertificate() {
        return this.getStringAttribute('tls_certificate');
    }
    /**
     * @stability stable
     */
    set tlsCertificate(value) {
        this._tlsCertificate = value;
    }
    /**
     * @stability stable
     */
    resetTlsCertificate() {
        this._tlsCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tlsCertificateInput() {
        return this._tlsCertificate;
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
}
exports.CodestarconnectionsHostVpcConfigurationOutputReference = CodestarconnectionsHostVpcConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodestarconnectionsHostVpcConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.codestar.CodestarconnectionsHostVpcConfigurationOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/codestarconnections_host.html aws_codestarconnections_host}.
 *
 * @stability stable
 */
class CodestarconnectionsHost extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codestarconnections_host.html aws_codestarconnections_host} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codestarconnections_host',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // vpc_configuration - computed: false, optional: true, required: false
        this._vpcConfiguration = new CodestarconnectionsHostVpcConfigurationOutputReference(this, "vpc_configuration", true);
        this._name = config.name;
        this._providerEndpoint = config.providerEndpoint;
        this._providerType = config.providerType;
        this._vpcConfiguration.internalValue = config.vpcConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get providerEndpoint() {
        return this.getStringAttribute('provider_endpoint');
    }
    /**
     * @stability stable
     */
    set providerEndpoint(value) {
        this._providerEndpoint = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get providerEndpointInput() {
        return this._providerEndpoint;
    }
    /**
     * @stability stable
     */
    get providerType() {
        return this.getStringAttribute('provider_type');
    }
    /**
     * @stability stable
     */
    set providerType(value) {
        this._providerType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get providerTypeInput() {
        return this._providerType;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get vpcConfiguration() {
        return this._vpcConfiguration;
    }
    /**
     * @stability stable
     */
    putVpcConfiguration(value) {
        this._vpcConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetVpcConfiguration() {
        this._vpcConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcConfigurationInput() {
        return this._vpcConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            provider_endpoint: cdktf.stringToTerraform(this._providerEndpoint),
            provider_type: cdktf.stringToTerraform(this._providerType),
            vpc_configuration: codestarconnectionsHostVpcConfigurationToTerraform(this._vpcConfiguration.internalValue),
        };
    }
}
exports.CodestarconnectionsHost = CodestarconnectionsHost;
_b = JSII_RTTI_SYMBOL_1;
CodestarconnectionsHost[_b] = { fqn: "@cdktf/provider-aws.codestar.CodestarconnectionsHost", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CodestarconnectionsHost.tfResourceType = "aws_codestarconnections_host";
