// @cdktf/provider-aws/lib/budgets/budgets-budget-action.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BudgetsBudgetAction = exports.budgetsBudgetActionSubscriberToTerraform = exports.BudgetsBudgetActionDefinitionOutputReference = exports.budgetsBudgetActionDefinitionToTerraform = exports.BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference = exports.budgetsBudgetActionDefinitionSsmActionDefinitionToTerraform = exports.BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference = exports.budgetsBudgetActionDefinitionScpActionDefinitionToTerraform = exports.BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference = exports.budgetsBudgetActionDefinitionIamActionDefinitionToTerraform = exports.BudgetsBudgetActionActionThresholdOutputReference = exports.budgetsBudgetActionActionThresholdToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function budgetsBudgetActionActionThresholdToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_threshold_type: cdktf.stringToTerraform(struct.actionThresholdType),
        action_threshold_value: cdktf.numberToTerraform(struct.actionThresholdValue),
    };
}
exports.budgetsBudgetActionActionThresholdToTerraform = budgetsBudgetActionActionThresholdToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetActionActionThresholdOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionThresholdType) {
            hasAnyValues = true;
            internalValueResult.actionThresholdType = this._actionThresholdType;
        }
        if (this._actionThresholdValue) {
            hasAnyValues = true;
            internalValueResult.actionThresholdValue = this._actionThresholdValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._actionThresholdType = undefined;
            this._actionThresholdValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._actionThresholdType = value.actionThresholdType;
            this._actionThresholdValue = value.actionThresholdValue;
        }
    }
    /**
     * @stability stable
     */
    get actionThresholdType() {
        return this.getStringAttribute('action_threshold_type');
    }
    /**
     * @stability stable
     */
    set actionThresholdType(value) {
        this._actionThresholdType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionThresholdTypeInput() {
        return this._actionThresholdType;
    }
    /**
     * @stability stable
     */
    get actionThresholdValue() {
        return this.getNumberAttribute('action_threshold_value');
    }
    /**
     * @stability stable
     */
    set actionThresholdValue(value) {
        this._actionThresholdValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionThresholdValueInput() {
        return this._actionThresholdValue;
    }
}
exports.BudgetsBudgetActionActionThresholdOutputReference = BudgetsBudgetActionActionThresholdOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BudgetsBudgetActionActionThresholdOutputReference[_a] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetActionActionThresholdOutputReference", version: "3.0.1" };
function budgetsBudgetActionDefinitionIamActionDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.groups),
        policy_arn: cdktf.stringToTerraform(struct.policyArn),
        roles: cdktf.listMapper(cdktf.stringToTerraform)(struct.roles),
        users: cdktf.listMapper(cdktf.stringToTerraform)(struct.users),
    };
}
exports.budgetsBudgetActionDefinitionIamActionDefinitionToTerraform = budgetsBudgetActionDefinitionIamActionDefinitionToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._groups) {
            hasAnyValues = true;
            internalValueResult.groups = this._groups;
        }
        if (this._policyArn) {
            hasAnyValues = true;
            internalValueResult.policyArn = this._policyArn;
        }
        if (this._roles) {
            hasAnyValues = true;
            internalValueResult.roles = this._roles;
        }
        if (this._users) {
            hasAnyValues = true;
            internalValueResult.users = this._users;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._groups = undefined;
            this._policyArn = undefined;
            this._roles = undefined;
            this._users = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._groups = value.groups;
            this._policyArn = value.policyArn;
            this._roles = value.roles;
            this._users = value.users;
        }
    }
    /**
     * @stability stable
     */
    get groups() {
        return this.getListAttribute('groups');
    }
    /**
     * @stability stable
     */
    set groups(value) {
        this._groups = value;
    }
    /**
     * @stability stable
     */
    resetGroups() {
        this._groups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get groupsInput() {
        return this._groups;
    }
    /**
     * @stability stable
     */
    get policyArn() {
        return this.getStringAttribute('policy_arn');
    }
    /**
     * @stability stable
     */
    set policyArn(value) {
        this._policyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyArnInput() {
        return this._policyArn;
    }
    /**
     * @stability stable
     */
    get roles() {
        return this.getListAttribute('roles');
    }
    /**
     * @stability stable
     */
    set roles(value) {
        this._roles = value;
    }
    /**
     * @stability stable
     */
    resetRoles() {
        this._roles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rolesInput() {
        return this._roles;
    }
    /**
     * @stability stable
     */
    get users() {
        return this.getListAttribute('users');
    }
    /**
     * @stability stable
     */
    set users(value) {
        this._users = value;
    }
    /**
     * @stability stable
     */
    resetUsers() {
        this._users = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get usersInput() {
        return this._users;
    }
}
exports.BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference = BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference[_b] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference", version: "3.0.1" };
function budgetsBudgetActionDefinitionScpActionDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        policy_id: cdktf.stringToTerraform(struct.policyId),
        target_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.targetIds),
    };
}
exports.budgetsBudgetActionDefinitionScpActionDefinitionToTerraform = budgetsBudgetActionDefinitionScpActionDefinitionToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._policyId) {
            hasAnyValues = true;
            internalValueResult.policyId = this._policyId;
        }
        if (this._targetIds) {
            hasAnyValues = true;
            internalValueResult.targetIds = this._targetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._policyId = undefined;
            this._targetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._policyId = value.policyId;
            this._targetIds = value.targetIds;
        }
    }
    /**
     * @stability stable
     */
    get policyId() {
        return this.getStringAttribute('policy_id');
    }
    /**
     * @stability stable
     */
    set policyId(value) {
        this._policyId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyIdInput() {
        return this._policyId;
    }
    /**
     * @stability stable
     */
    get targetIds() {
        return this.getListAttribute('target_ids');
    }
    /**
     * @stability stable
     */
    set targetIds(value) {
        this._targetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetIdsInput() {
        return this._targetIds;
    }
}
exports.BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference = BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference[_c] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference", version: "3.0.1" };
function budgetsBudgetActionDefinitionSsmActionDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_sub_type: cdktf.stringToTerraform(struct.actionSubType),
        instance_ids: cdktf.listMapper(cdktf.stringToTerraform)(struct.instanceIds),
        region: cdktf.stringToTerraform(struct.region),
    };
}
exports.budgetsBudgetActionDefinitionSsmActionDefinitionToTerraform = budgetsBudgetActionDefinitionSsmActionDefinitionToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionSubType) {
            hasAnyValues = true;
            internalValueResult.actionSubType = this._actionSubType;
        }
        if (this._instanceIds) {
            hasAnyValues = true;
            internalValueResult.instanceIds = this._instanceIds;
        }
        if (this._region) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._actionSubType = undefined;
            this._instanceIds = undefined;
            this._region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._actionSubType = value.actionSubType;
            this._instanceIds = value.instanceIds;
            this._region = value.region;
        }
    }
    /**
     * @stability stable
     */
    get actionSubType() {
        return this.getStringAttribute('action_sub_type');
    }
    /**
     * @stability stable
     */
    set actionSubType(value) {
        this._actionSubType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionSubTypeInput() {
        return this._actionSubType;
    }
    /**
     * @stability stable
     */
    get instanceIds() {
        return this.getListAttribute('instance_ids');
    }
    /**
     * @stability stable
     */
    set instanceIds(value) {
        this._instanceIds = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceIdsInput() {
        return this._instanceIds;
    }
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
    /**
     * @stability stable
     */
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionInput() {
        return this._region;
    }
}
exports.BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference = BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference;
_d = JSII_RTTI_SYMBOL_1;
BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference[_d] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference", version: "3.0.1" };
function budgetsBudgetActionDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_action_definition: budgetsBudgetActionDefinitionIamActionDefinitionToTerraform(struct.iamActionDefinition),
        scp_action_definition: budgetsBudgetActionDefinitionScpActionDefinitionToTerraform(struct.scpActionDefinition),
        ssm_action_definition: budgetsBudgetActionDefinitionSsmActionDefinitionToTerraform(struct.ssmActionDefinition),
    };
}
exports.budgetsBudgetActionDefinitionToTerraform = budgetsBudgetActionDefinitionToTerraform;
/**
 * @stability stable
 */
class BudgetsBudgetActionDefinitionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // iam_action_definition - computed: false, optional: true, required: false
        this._iamActionDefinition = new BudgetsBudgetActionDefinitionIamActionDefinitionOutputReference(this, "iam_action_definition", true);
        // scp_action_definition - computed: false, optional: true, required: false
        this._scpActionDefinition = new BudgetsBudgetActionDefinitionScpActionDefinitionOutputReference(this, "scp_action_definition", true);
        // ssm_action_definition - computed: false, optional: true, required: false
        this._ssmActionDefinition = new BudgetsBudgetActionDefinitionSsmActionDefinitionOutputReference(this, "ssm_action_definition", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _g, _h, _j, _k, _l, _m;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_g = this._iamActionDefinition) === null || _g === void 0 ? void 0 : _g.internalValue) {
            hasAnyValues = true;
            internalValueResult.iamActionDefinition = (_h = this._iamActionDefinition) === null || _h === void 0 ? void 0 : _h.internalValue;
        }
        if ((_j = this._scpActionDefinition) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.scpActionDefinition = (_k = this._scpActionDefinition) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._ssmActionDefinition) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.ssmActionDefinition = (_m = this._ssmActionDefinition) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iamActionDefinition.internalValue = undefined;
            this._scpActionDefinition.internalValue = undefined;
            this._ssmActionDefinition.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iamActionDefinition.internalValue = value.iamActionDefinition;
            this._scpActionDefinition.internalValue = value.scpActionDefinition;
            this._ssmActionDefinition.internalValue = value.ssmActionDefinition;
        }
    }
    /**
     * @stability stable
     */
    get iamActionDefinition() {
        return this._iamActionDefinition;
    }
    /**
     * @stability stable
     */
    putIamActionDefinition(value) {
        this._iamActionDefinition.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIamActionDefinition() {
        this._iamActionDefinition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamActionDefinitionInput() {
        return this._iamActionDefinition.internalValue;
    }
    /**
     * @stability stable
     */
    get scpActionDefinition() {
        return this._scpActionDefinition;
    }
    /**
     * @stability stable
     */
    putScpActionDefinition(value) {
        this._scpActionDefinition.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetScpActionDefinition() {
        this._scpActionDefinition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scpActionDefinitionInput() {
        return this._scpActionDefinition.internalValue;
    }
    /**
     * @stability stable
     */
    get ssmActionDefinition() {
        return this._ssmActionDefinition;
    }
    /**
     * @stability stable
     */
    putSsmActionDefinition(value) {
        this._ssmActionDefinition.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSsmActionDefinition() {
        this._ssmActionDefinition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ssmActionDefinitionInput() {
        return this._ssmActionDefinition.internalValue;
    }
}
exports.BudgetsBudgetActionDefinitionOutputReference = BudgetsBudgetActionDefinitionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
BudgetsBudgetActionDefinitionOutputReference[_e] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetActionDefinitionOutputReference", version: "3.0.1" };
function budgetsBudgetActionSubscriberToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address: cdktf.stringToTerraform(struct.address),
        subscription_type: cdktf.stringToTerraform(struct.subscriptionType),
    };
}
exports.budgetsBudgetActionSubscriberToTerraform = budgetsBudgetActionSubscriberToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/budgets_budget_action.html aws_budgets_budget_action}.
 *
 * @stability stable
 */
class BudgetsBudgetAction extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/budgets_budget_action.html aws_budgets_budget_action} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_budgets_budget_action',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // action_threshold - computed: false, optional: false, required: true
        this._actionThreshold = new BudgetsBudgetActionActionThresholdOutputReference(this, "action_threshold", true);
        // definition - computed: false, optional: false, required: true
        this._definition = new BudgetsBudgetActionDefinitionOutputReference(this, "definition", true);
        this._accountId = config.accountId;
        this._actionType = config.actionType;
        this._approvalModel = config.approvalModel;
        this._budgetName = config.budgetName;
        this._executionRoleArn = config.executionRoleArn;
        this._notificationType = config.notificationType;
        this._actionThreshold.internalValue = config.actionThreshold;
        this._definition.internalValue = config.definition;
        this._subscriber = config.subscriber;
    }
    /**
     * @stability stable
     */
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    /**
     * @stability stable
     */
    set accountId(value) {
        this._accountId = value;
    }
    /**
     * @stability stable
     */
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountIdInput() {
        return this._accountId;
    }
    // action_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get actionId() {
        return this.getStringAttribute('action_id');
    }
    /**
     * @stability stable
     */
    get actionType() {
        return this.getStringAttribute('action_type');
    }
    /**
     * @stability stable
     */
    set actionType(value) {
        this._actionType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionTypeInput() {
        return this._actionType;
    }
    /**
     * @stability stable
     */
    get approvalModel() {
        return this.getStringAttribute('approval_model');
    }
    /**
     * @stability stable
     */
    set approvalModel(value) {
        this._approvalModel = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get approvalModelInput() {
        return this._approvalModel;
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get budgetName() {
        return this.getStringAttribute('budget_name');
    }
    /**
     * @stability stable
     */
    set budgetName(value) {
        this._budgetName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get budgetNameInput() {
        return this._budgetName;
    }
    /**
     * @stability stable
     */
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    /**
     * @stability stable
     */
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    /**
     * @stability stable
     */
    set notificationType(value) {
        this._notificationType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationTypeInput() {
        return this._notificationType;
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get actionThreshold() {
        return this._actionThreshold;
    }
    /**
     * @stability stable
     */
    putActionThreshold(value) {
        this._actionThreshold.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get actionThresholdInput() {
        return this._actionThreshold.internalValue;
    }
    /**
     * @stability stable
     */
    get definition() {
        return this._definition;
    }
    /**
     * @stability stable
     */
    putDefinition(value) {
        this._definition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get definitionInput() {
        return this._definition.internalValue;
    }
    /**
     * @stability stable
     */
    get subscriber() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('subscriber');
    }
    /**
     * @stability stable
     */
    set subscriber(value) {
        this._subscriber = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subscriberInput() {
        return this._subscriber;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            action_type: cdktf.stringToTerraform(this._actionType),
            approval_model: cdktf.stringToTerraform(this._approvalModel),
            budget_name: cdktf.stringToTerraform(this._budgetName),
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            notification_type: cdktf.stringToTerraform(this._notificationType),
            action_threshold: budgetsBudgetActionActionThresholdToTerraform(this._actionThreshold.internalValue),
            definition: budgetsBudgetActionDefinitionToTerraform(this._definition.internalValue),
            subscriber: cdktf.listMapper(budgetsBudgetActionSubscriberToTerraform)(this._subscriber),
        };
    }
}
exports.BudgetsBudgetAction = BudgetsBudgetAction;
_f = JSII_RTTI_SYMBOL_1;
BudgetsBudgetAction[_f] = { fqn: "@cdktf/provider-aws.budgets.BudgetsBudgetAction", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
BudgetsBudgetAction.tfResourceType = "aws_budgets_budget_action";
