// @cdktf/provider-aws/lib/iot/iot-authorizer.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotAuthorizer = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_authorizer.html aws_iot_authorizer}.
 *
 * @stability stable
 */
class IotAuthorizer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_authorizer.html aws_iot_authorizer} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_authorizer',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._authorizerFunctionArn = config.authorizerFunctionArn;
        this._name = config.name;
        this._signingDisabled = config.signingDisabled;
        this._status = config.status;
        this._tokenKeyName = config.tokenKeyName;
        this._tokenSigningPublicKeys = config.tokenSigningPublicKeys;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get authorizerFunctionArn() {
        return this.getStringAttribute('authorizer_function_arn');
    }
    /**
     * @stability stable
     */
    set authorizerFunctionArn(value) {
        this._authorizerFunctionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get authorizerFunctionArnInput() {
        return this._authorizerFunctionArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get signingDisabled() {
        return this.getBooleanAttribute('signing_disabled');
    }
    /**
     * @stability stable
     */
    set signingDisabled(value) {
        this._signingDisabled = value;
    }
    /**
     * @stability stable
     */
    resetSigningDisabled() {
        this._signingDisabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signingDisabledInput() {
        return this._signingDisabled;
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get tokenKeyName() {
        return this.getStringAttribute('token_key_name');
    }
    /**
     * @stability stable
     */
    set tokenKeyName(value) {
        this._tokenKeyName = value;
    }
    /**
     * @stability stable
     */
    resetTokenKeyName() {
        this._tokenKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tokenKeyNameInput() {
        return this._tokenKeyName;
    }
    /**
     * @stability stable
     */
    get tokenSigningPublicKeys() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('token_signing_public_keys');
    }
    /**
     * @stability stable
     */
    set tokenSigningPublicKeys(value) {
        this._tokenSigningPublicKeys = value;
    }
    /**
     * @stability stable
     */
    resetTokenSigningPublicKeys() {
        this._tokenSigningPublicKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tokenSigningPublicKeysInput() {
        return this._tokenSigningPublicKeys;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            authorizer_function_arn: cdktf.stringToTerraform(this._authorizerFunctionArn),
            name: cdktf.stringToTerraform(this._name),
            signing_disabled: cdktf.booleanToTerraform(this._signingDisabled),
            status: cdktf.stringToTerraform(this._status),
            token_key_name: cdktf.stringToTerraform(this._tokenKeyName),
            token_signing_public_keys: cdktf.hashMapper(cdktf.anyToTerraform)(this._tokenSigningPublicKeys),
        };
    }
}
exports.IotAuthorizer = IotAuthorizer;
_a = JSII_RTTI_SYMBOL_1;
IotAuthorizer[_a] = { fqn: "@cdktf/provider-aws.iot.IotAuthorizer", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotAuthorizer.tfResourceType = "aws_iot_authorizer";
