// @cdktf/provider-aws/lib/iot/iot-thing-group-membership.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThingGroupMembership = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group_membership.html aws_iot_thing_group_membership}.
 *
 * @stability stable
 */
class IotThingGroupMembership extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group_membership.html aws_iot_thing_group_membership} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing_group_membership',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._overrideDynamicGroup = config.overrideDynamicGroup;
        this._thingGroupName = config.thingGroupName;
        this._thingName = config.thingName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get overrideDynamicGroup() {
        return this.getBooleanAttribute('override_dynamic_group');
    }
    /**
     * @stability stable
     */
    set overrideDynamicGroup(value) {
        this._overrideDynamicGroup = value;
    }
    /**
     * @stability stable
     */
    resetOverrideDynamicGroup() {
        this._overrideDynamicGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get overrideDynamicGroupInput() {
        return this._overrideDynamicGroup;
    }
    /**
     * @stability stable
     */
    get thingGroupName() {
        return this.getStringAttribute('thing_group_name');
    }
    /**
     * @stability stable
     */
    set thingGroupName(value) {
        this._thingGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thingGroupNameInput() {
        return this._thingGroupName;
    }
    /**
     * @stability stable
     */
    get thingName() {
        return this.getStringAttribute('thing_name');
    }
    /**
     * @stability stable
     */
    set thingName(value) {
        this._thingName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thingNameInput() {
        return this._thingName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            override_dynamic_group: cdktf.booleanToTerraform(this._overrideDynamicGroup),
            thing_group_name: cdktf.stringToTerraform(this._thingGroupName),
            thing_name: cdktf.stringToTerraform(this._thingName),
        };
    }
}
exports.IotThingGroupMembership = IotThingGroupMembership;
_a = JSII_RTTI_SYMBOL_1;
IotThingGroupMembership[_a] = { fqn: "@cdktf/provider-aws.iot.IotThingGroupMembership", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotThingGroupMembership.tfResourceType = "aws_iot_thing_group_membership";
