// @cdktf/provider-aws/lib/iot/iot-topic-rule.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotTopicRule = exports.iotTopicRuleStepFunctionsToTerraform = exports.iotTopicRuleSqsToTerraform = exports.iotTopicRuleSnsToTerraform = exports.iotTopicRuleS3ToTerraform = exports.iotTopicRuleRepublishToTerraform = exports.iotTopicRuleLambdaToTerraform = exports.iotTopicRuleKinesisToTerraform = exports.iotTopicRuleIotEventsToTerraform = exports.iotTopicRuleIotAnalyticsToTerraform = exports.iotTopicRuleFirehoseToTerraform = exports.IotTopicRuleErrorActionOutputReference = exports.iotTopicRuleErrorActionToTerraform = exports.IotTopicRuleErrorActionStepFunctionsOutputReference = exports.iotTopicRuleErrorActionStepFunctionsToTerraform = exports.IotTopicRuleErrorActionSqsOutputReference = exports.iotTopicRuleErrorActionSqsToTerraform = exports.IotTopicRuleErrorActionSnsOutputReference = exports.iotTopicRuleErrorActionSnsToTerraform = exports.IotTopicRuleErrorActionS3OutputReference = exports.iotTopicRuleErrorActionS3ToTerraform = exports.IotTopicRuleErrorActionRepublishOutputReference = exports.iotTopicRuleErrorActionRepublishToTerraform = exports.IotTopicRuleErrorActionLambdaOutputReference = exports.iotTopicRuleErrorActionLambdaToTerraform = exports.IotTopicRuleErrorActionKinesisOutputReference = exports.iotTopicRuleErrorActionKinesisToTerraform = exports.IotTopicRuleErrorActionIotEventsOutputReference = exports.iotTopicRuleErrorActionIotEventsToTerraform = exports.IotTopicRuleErrorActionIotAnalyticsOutputReference = exports.iotTopicRuleErrorActionIotAnalyticsToTerraform = exports.IotTopicRuleErrorActionFirehoseOutputReference = exports.iotTopicRuleErrorActionFirehoseToTerraform = exports.IotTopicRuleErrorActionElasticsearchOutputReference = exports.iotTopicRuleErrorActionElasticsearchToTerraform = exports.IotTopicRuleErrorActionDynamodbv2OutputReference = exports.iotTopicRuleErrorActionDynamodbv2ToTerraform = exports.IotTopicRuleErrorActionDynamodbv2PutItemOutputReference = exports.iotTopicRuleErrorActionDynamodbv2PutItemToTerraform = exports.IotTopicRuleErrorActionDynamodbOutputReference = exports.iotTopicRuleErrorActionDynamodbToTerraform = exports.IotTopicRuleErrorActionCloudwatchMetricOutputReference = exports.iotTopicRuleErrorActionCloudwatchMetricToTerraform = exports.IotTopicRuleErrorActionCloudwatchAlarmOutputReference = exports.iotTopicRuleErrorActionCloudwatchAlarmToTerraform = exports.iotTopicRuleElasticsearchToTerraform = exports.iotTopicRuleDynamodbv2ToTerraform = exports.IotTopicRuleDynamodbv2PutItemOutputReference = exports.iotTopicRuleDynamodbv2PutItemToTerraform = exports.iotTopicRuleDynamodbToTerraform = exports.iotTopicRuleCloudwatchMetricToTerraform = exports.iotTopicRuleCloudwatchAlarmToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iotTopicRuleCloudwatchAlarmToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        alarm_name: cdktf.stringToTerraform(struct.alarmName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        state_reason: cdktf.stringToTerraform(struct.stateReason),
        state_value: cdktf.stringToTerraform(struct.stateValue),
    };
}
exports.iotTopicRuleCloudwatchAlarmToTerraform = iotTopicRuleCloudwatchAlarmToTerraform;
function iotTopicRuleCloudwatchMetricToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metric_name: cdktf.stringToTerraform(struct.metricName),
        metric_namespace: cdktf.stringToTerraform(struct.metricNamespace),
        metric_timestamp: cdktf.stringToTerraform(struct.metricTimestamp),
        metric_unit: cdktf.stringToTerraform(struct.metricUnit),
        metric_value: cdktf.stringToTerraform(struct.metricValue),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleCloudwatchMetricToTerraform = iotTopicRuleCloudwatchMetricToTerraform;
function iotTopicRuleDynamodbToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hash_key_field: cdktf.stringToTerraform(struct.hashKeyField),
        hash_key_type: cdktf.stringToTerraform(struct.hashKeyType),
        hash_key_value: cdktf.stringToTerraform(struct.hashKeyValue),
        operation: cdktf.stringToTerraform(struct.operation),
        payload_field: cdktf.stringToTerraform(struct.payloadField),
        range_key_field: cdktf.stringToTerraform(struct.rangeKeyField),
        range_key_type: cdktf.stringToTerraform(struct.rangeKeyType),
        range_key_value: cdktf.stringToTerraform(struct.rangeKeyValue),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.iotTopicRuleDynamodbToTerraform = iotTopicRuleDynamodbToTerraform;
function iotTopicRuleDynamodbv2PutItemToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.iotTopicRuleDynamodbv2PutItemToTerraform = iotTopicRuleDynamodbv2PutItemToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleDynamodbv2PutItemOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tableName) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tableName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tableName = value.tableName;
        }
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
}
exports.IotTopicRuleDynamodbv2PutItemOutputReference = IotTopicRuleDynamodbv2PutItemOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IotTopicRuleDynamodbv2PutItemOutputReference[_a] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleDynamodbv2PutItemOutputReference", version: "3.0.1" };
function iotTopicRuleDynamodbv2ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        put_item: iotTopicRuleDynamodbv2PutItemToTerraform(struct.putItem),
    };
}
exports.iotTopicRuleDynamodbv2ToTerraform = iotTopicRuleDynamodbv2ToTerraform;
function iotTopicRuleElasticsearchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint: cdktf.stringToTerraform(struct.endpoint),
        id: cdktf.stringToTerraform(struct.id),
        index: cdktf.stringToTerraform(struct.index),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.iotTopicRuleElasticsearchToTerraform = iotTopicRuleElasticsearchToTerraform;
function iotTopicRuleErrorActionCloudwatchAlarmToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        alarm_name: cdktf.stringToTerraform(struct.alarmName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        state_reason: cdktf.stringToTerraform(struct.stateReason),
        state_value: cdktf.stringToTerraform(struct.stateValue),
    };
}
exports.iotTopicRuleErrorActionCloudwatchAlarmToTerraform = iotTopicRuleErrorActionCloudwatchAlarmToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionCloudwatchAlarmOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._alarmName) {
            hasAnyValues = true;
            internalValueResult.alarmName = this._alarmName;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._stateReason) {
            hasAnyValues = true;
            internalValueResult.stateReason = this._stateReason;
        }
        if (this._stateValue) {
            hasAnyValues = true;
            internalValueResult.stateValue = this._stateValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._alarmName = undefined;
            this._roleArn = undefined;
            this._stateReason = undefined;
            this._stateValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._alarmName = value.alarmName;
            this._roleArn = value.roleArn;
            this._stateReason = value.stateReason;
            this._stateValue = value.stateValue;
        }
    }
    /**
     * @stability stable
     */
    get alarmName() {
        return this.getStringAttribute('alarm_name');
    }
    /**
     * @stability stable
     */
    set alarmName(value) {
        this._alarmName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get alarmNameInput() {
        return this._alarmName;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get stateReason() {
        return this.getStringAttribute('state_reason');
    }
    /**
     * @stability stable
     */
    set stateReason(value) {
        this._stateReason = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateReasonInput() {
        return this._stateReason;
    }
    /**
     * @stability stable
     */
    get stateValue() {
        return this.getStringAttribute('state_value');
    }
    /**
     * @stability stable
     */
    set stateValue(value) {
        this._stateValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateValueInput() {
        return this._stateValue;
    }
}
exports.IotTopicRuleErrorActionCloudwatchAlarmOutputReference = IotTopicRuleErrorActionCloudwatchAlarmOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionCloudwatchAlarmOutputReference[_b] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionCloudwatchAlarmOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionCloudwatchMetricToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metric_name: cdktf.stringToTerraform(struct.metricName),
        metric_namespace: cdktf.stringToTerraform(struct.metricNamespace),
        metric_timestamp: cdktf.stringToTerraform(struct.metricTimestamp),
        metric_unit: cdktf.stringToTerraform(struct.metricUnit),
        metric_value: cdktf.stringToTerraform(struct.metricValue),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleErrorActionCloudwatchMetricToTerraform = iotTopicRuleErrorActionCloudwatchMetricToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionCloudwatchMetricOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metricName) {
            hasAnyValues = true;
            internalValueResult.metricName = this._metricName;
        }
        if (this._metricNamespace) {
            hasAnyValues = true;
            internalValueResult.metricNamespace = this._metricNamespace;
        }
        if (this._metricTimestamp) {
            hasAnyValues = true;
            internalValueResult.metricTimestamp = this._metricTimestamp;
        }
        if (this._metricUnit) {
            hasAnyValues = true;
            internalValueResult.metricUnit = this._metricUnit;
        }
        if (this._metricValue) {
            hasAnyValues = true;
            internalValueResult.metricValue = this._metricValue;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metricName = undefined;
            this._metricNamespace = undefined;
            this._metricTimestamp = undefined;
            this._metricUnit = undefined;
            this._metricValue = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metricName = value.metricName;
            this._metricNamespace = value.metricNamespace;
            this._metricTimestamp = value.metricTimestamp;
            this._metricUnit = value.metricUnit;
            this._metricValue = value.metricValue;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    /**
     * @stability stable
     */
    set metricName(value) {
        this._metricName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricNameInput() {
        return this._metricName;
    }
    /**
     * @stability stable
     */
    get metricNamespace() {
        return this.getStringAttribute('metric_namespace');
    }
    /**
     * @stability stable
     */
    set metricNamespace(value) {
        this._metricNamespace = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricNamespaceInput() {
        return this._metricNamespace;
    }
    /**
     * @stability stable
     */
    get metricTimestamp() {
        return this.getStringAttribute('metric_timestamp');
    }
    /**
     * @stability stable
     */
    set metricTimestamp(value) {
        this._metricTimestamp = value;
    }
    /**
     * @stability stable
     */
    resetMetricTimestamp() {
        this._metricTimestamp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricTimestampInput() {
        return this._metricTimestamp;
    }
    /**
     * @stability stable
     */
    get metricUnit() {
        return this.getStringAttribute('metric_unit');
    }
    /**
     * @stability stable
     */
    set metricUnit(value) {
        this._metricUnit = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricUnitInput() {
        return this._metricUnit;
    }
    /**
     * @stability stable
     */
    get metricValue() {
        return this.getStringAttribute('metric_value');
    }
    /**
     * @stability stable
     */
    set metricValue(value) {
        this._metricValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get metricValueInput() {
        return this._metricValue;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.IotTopicRuleErrorActionCloudwatchMetricOutputReference = IotTopicRuleErrorActionCloudwatchMetricOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionCloudwatchMetricOutputReference[_c] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionCloudwatchMetricOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionDynamodbToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hash_key_field: cdktf.stringToTerraform(struct.hashKeyField),
        hash_key_type: cdktf.stringToTerraform(struct.hashKeyType),
        hash_key_value: cdktf.stringToTerraform(struct.hashKeyValue),
        operation: cdktf.stringToTerraform(struct.operation),
        payload_field: cdktf.stringToTerraform(struct.payloadField),
        range_key_field: cdktf.stringToTerraform(struct.rangeKeyField),
        range_key_type: cdktf.stringToTerraform(struct.rangeKeyType),
        range_key_value: cdktf.stringToTerraform(struct.rangeKeyValue),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.iotTopicRuleErrorActionDynamodbToTerraform = iotTopicRuleErrorActionDynamodbToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionDynamodbOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hashKeyField) {
            hasAnyValues = true;
            internalValueResult.hashKeyField = this._hashKeyField;
        }
        if (this._hashKeyType) {
            hasAnyValues = true;
            internalValueResult.hashKeyType = this._hashKeyType;
        }
        if (this._hashKeyValue) {
            hasAnyValues = true;
            internalValueResult.hashKeyValue = this._hashKeyValue;
        }
        if (this._operation) {
            hasAnyValues = true;
            internalValueResult.operation = this._operation;
        }
        if (this._payloadField) {
            hasAnyValues = true;
            internalValueResult.payloadField = this._payloadField;
        }
        if (this._rangeKeyField) {
            hasAnyValues = true;
            internalValueResult.rangeKeyField = this._rangeKeyField;
        }
        if (this._rangeKeyType) {
            hasAnyValues = true;
            internalValueResult.rangeKeyType = this._rangeKeyType;
        }
        if (this._rangeKeyValue) {
            hasAnyValues = true;
            internalValueResult.rangeKeyValue = this._rangeKeyValue;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._tableName) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hashKeyField = undefined;
            this._hashKeyType = undefined;
            this._hashKeyValue = undefined;
            this._operation = undefined;
            this._payloadField = undefined;
            this._rangeKeyField = undefined;
            this._rangeKeyType = undefined;
            this._rangeKeyValue = undefined;
            this._roleArn = undefined;
            this._tableName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hashKeyField = value.hashKeyField;
            this._hashKeyType = value.hashKeyType;
            this._hashKeyValue = value.hashKeyValue;
            this._operation = value.operation;
            this._payloadField = value.payloadField;
            this._rangeKeyField = value.rangeKeyField;
            this._rangeKeyType = value.rangeKeyType;
            this._rangeKeyValue = value.rangeKeyValue;
            this._roleArn = value.roleArn;
            this._tableName = value.tableName;
        }
    }
    /**
     * @stability stable
     */
    get hashKeyField() {
        return this.getStringAttribute('hash_key_field');
    }
    /**
     * @stability stable
     */
    set hashKeyField(value) {
        this._hashKeyField = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hashKeyFieldInput() {
        return this._hashKeyField;
    }
    /**
     * @stability stable
     */
    get hashKeyType() {
        return this.getStringAttribute('hash_key_type');
    }
    /**
     * @stability stable
     */
    set hashKeyType(value) {
        this._hashKeyType = value;
    }
    /**
     * @stability stable
     */
    resetHashKeyType() {
        this._hashKeyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hashKeyTypeInput() {
        return this._hashKeyType;
    }
    /**
     * @stability stable
     */
    get hashKeyValue() {
        return this.getStringAttribute('hash_key_value');
    }
    /**
     * @stability stable
     */
    set hashKeyValue(value) {
        this._hashKeyValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hashKeyValueInput() {
        return this._hashKeyValue;
    }
    /**
     * @stability stable
     */
    get operation() {
        return this.getStringAttribute('operation');
    }
    /**
     * @stability stable
     */
    set operation(value) {
        this._operation = value;
    }
    /**
     * @stability stable
     */
    resetOperation() {
        this._operation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get operationInput() {
        return this._operation;
    }
    /**
     * @stability stable
     */
    get payloadField() {
        return this.getStringAttribute('payload_field');
    }
    /**
     * @stability stable
     */
    set payloadField(value) {
        this._payloadField = value;
    }
    /**
     * @stability stable
     */
    resetPayloadField() {
        this._payloadField = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get payloadFieldInput() {
        return this._payloadField;
    }
    /**
     * @stability stable
     */
    get rangeKeyField() {
        return this.getStringAttribute('range_key_field');
    }
    /**
     * @stability stable
     */
    set rangeKeyField(value) {
        this._rangeKeyField = value;
    }
    /**
     * @stability stable
     */
    resetRangeKeyField() {
        this._rangeKeyField = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rangeKeyFieldInput() {
        return this._rangeKeyField;
    }
    /**
     * @stability stable
     */
    get rangeKeyType() {
        return this.getStringAttribute('range_key_type');
    }
    /**
     * @stability stable
     */
    set rangeKeyType(value) {
        this._rangeKeyType = value;
    }
    /**
     * @stability stable
     */
    resetRangeKeyType() {
        this._rangeKeyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rangeKeyTypeInput() {
        return this._rangeKeyType;
    }
    /**
     * @stability stable
     */
    get rangeKeyValue() {
        return this.getStringAttribute('range_key_value');
    }
    /**
     * @stability stable
     */
    set rangeKeyValue(value) {
        this._rangeKeyValue = value;
    }
    /**
     * @stability stable
     */
    resetRangeKeyValue() {
        this._rangeKeyValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rangeKeyValueInput() {
        return this._rangeKeyValue;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
}
exports.IotTopicRuleErrorActionDynamodbOutputReference = IotTopicRuleErrorActionDynamodbOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionDynamodbOutputReference[_d] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionDynamodbOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionDynamodbv2PutItemToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.iotTopicRuleErrorActionDynamodbv2PutItemToTerraform = iotTopicRuleErrorActionDynamodbv2PutItemToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionDynamodbv2PutItemOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tableName) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tableName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tableName = value.tableName;
        }
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
}
exports.IotTopicRuleErrorActionDynamodbv2PutItemOutputReference = IotTopicRuleErrorActionDynamodbv2PutItemOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionDynamodbv2PutItemOutputReference[_e] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionDynamodbv2PutItemOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionDynamodbv2ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        put_item: iotTopicRuleErrorActionDynamodbv2PutItemToTerraform(struct.putItem),
    };
}
exports.iotTopicRuleErrorActionDynamodbv2ToTerraform = iotTopicRuleErrorActionDynamodbv2ToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionDynamodbv2OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // put_item - computed: false, optional: true, required: false
        this._putItem = new IotTopicRuleErrorActionDynamodbv2PutItemOutputReference(this, "put_item", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _v, _w;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if ((_v = this._putItem) === null || _v === void 0 ? void 0 : _v.internalValue) {
            hasAnyValues = true;
            internalValueResult.putItem = (_w = this._putItem) === null || _w === void 0 ? void 0 : _w.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._roleArn = undefined;
            this._putItem.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._roleArn = value.roleArn;
            this._putItem.internalValue = value.putItem;
        }
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get putItem() {
        return this._putItem;
    }
    /**
     * @stability stable
     */
    putPutItem(value) {
        this._putItem.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetPutItem() {
        this._putItem.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get putItemInput() {
        return this._putItem.internalValue;
    }
}
exports.IotTopicRuleErrorActionDynamodbv2OutputReference = IotTopicRuleErrorActionDynamodbv2OutputReference;
_f = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionDynamodbv2OutputReference[_f] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionDynamodbv2OutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionElasticsearchToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint: cdktf.stringToTerraform(struct.endpoint),
        id: cdktf.stringToTerraform(struct.id),
        index: cdktf.stringToTerraform(struct.index),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.iotTopicRuleErrorActionElasticsearchToTerraform = iotTopicRuleErrorActionElasticsearchToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionElasticsearchOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpoint) {
            hasAnyValues = true;
            internalValueResult.endpoint = this._endpoint;
        }
        if (this._id) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._index) {
            hasAnyValues = true;
            internalValueResult.index = this._index;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpoint = undefined;
            this._id = undefined;
            this._index = undefined;
            this._roleArn = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpoint = value.endpoint;
            this._id = value.id;
            this._index = value.index;
            this._roleArn = value.roleArn;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    /**
     * @stability stable
     */
    set endpoint(value) {
        this._endpoint = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endpointInput() {
        return this._endpoint;
    }
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get idInput() {
        return this._id;
    }
    /**
     * @stability stable
     */
    get index() {
        return this.getStringAttribute('index');
    }
    /**
     * @stability stable
     */
    set index(value) {
        this._index = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get indexInput() {
        return this._index;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.IotTopicRuleErrorActionElasticsearchOutputReference = IotTopicRuleErrorActionElasticsearchOutputReference;
_g = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionElasticsearchOutputReference[_g] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionElasticsearchOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream_name: cdktf.stringToTerraform(struct.deliveryStreamName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        separator: cdktf.stringToTerraform(struct.separator),
    };
}
exports.iotTopicRuleErrorActionFirehoseToTerraform = iotTopicRuleErrorActionFirehoseToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionFirehoseOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deliveryStreamName) {
            hasAnyValues = true;
            internalValueResult.deliveryStreamName = this._deliveryStreamName;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._separator) {
            hasAnyValues = true;
            internalValueResult.separator = this._separator;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deliveryStreamName = undefined;
            this._roleArn = undefined;
            this._separator = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deliveryStreamName = value.deliveryStreamName;
            this._roleArn = value.roleArn;
            this._separator = value.separator;
        }
    }
    /**
     * @stability stable
     */
    get deliveryStreamName() {
        return this.getStringAttribute('delivery_stream_name');
    }
    /**
     * @stability stable
     */
    set deliveryStreamName(value) {
        this._deliveryStreamName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deliveryStreamNameInput() {
        return this._deliveryStreamName;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get separator() {
        return this.getStringAttribute('separator');
    }
    /**
     * @stability stable
     */
    set separator(value) {
        this._separator = value;
    }
    /**
     * @stability stable
     */
    resetSeparator() {
        this._separator = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get separatorInput() {
        return this._separator;
    }
}
exports.IotTopicRuleErrorActionFirehoseOutputReference = IotTopicRuleErrorActionFirehoseOutputReference;
_h = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionFirehoseOutputReference[_h] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionFirehoseOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionIotAnalyticsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        channel_name: cdktf.stringToTerraform(struct.channelName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleErrorActionIotAnalyticsToTerraform = iotTopicRuleErrorActionIotAnalyticsToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionIotAnalyticsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._channelName) {
            hasAnyValues = true;
            internalValueResult.channelName = this._channelName;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._channelName = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._channelName = value.channelName;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get channelName() {
        return this.getStringAttribute('channel_name');
    }
    /**
     * @stability stable
     */
    set channelName(value) {
        this._channelName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get channelNameInput() {
        return this._channelName;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.IotTopicRuleErrorActionIotAnalyticsOutputReference = IotTopicRuleErrorActionIotAnalyticsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionIotAnalyticsOutputReference[_j] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionIotAnalyticsOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionIotEventsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_name: cdktf.stringToTerraform(struct.inputName),
        message_id: cdktf.stringToTerraform(struct.messageId),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleErrorActionIotEventsToTerraform = iotTopicRuleErrorActionIotEventsToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionIotEventsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputName) {
            hasAnyValues = true;
            internalValueResult.inputName = this._inputName;
        }
        if (this._messageId) {
            hasAnyValues = true;
            internalValueResult.messageId = this._messageId;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._inputName = undefined;
            this._messageId = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._inputName = value.inputName;
            this._messageId = value.messageId;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get inputName() {
        return this.getStringAttribute('input_name');
    }
    /**
     * @stability stable
     */
    set inputName(value) {
        this._inputName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputNameInput() {
        return this._inputName;
    }
    /**
     * @stability stable
     */
    get messageId() {
        return this.getStringAttribute('message_id');
    }
    /**
     * @stability stable
     */
    set messageId(value) {
        this._messageId = value;
    }
    /**
     * @stability stable
     */
    resetMessageId() {
        this._messageId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageIdInput() {
        return this._messageId;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.IotTopicRuleErrorActionIotEventsOutputReference = IotTopicRuleErrorActionIotEventsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionIotEventsOutputReference[_k] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionIotEventsOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionKinesisToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partition_key: cdktf.stringToTerraform(struct.partitionKey),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        stream_name: cdktf.stringToTerraform(struct.streamName),
    };
}
exports.iotTopicRuleErrorActionKinesisToTerraform = iotTopicRuleErrorActionKinesisToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionKinesisOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._partitionKey) {
            hasAnyValues = true;
            internalValueResult.partitionKey = this._partitionKey;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._streamName) {
            hasAnyValues = true;
            internalValueResult.streamName = this._streamName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._partitionKey = undefined;
            this._roleArn = undefined;
            this._streamName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._partitionKey = value.partitionKey;
            this._roleArn = value.roleArn;
            this._streamName = value.streamName;
        }
    }
    /**
     * @stability stable
     */
    get partitionKey() {
        return this.getStringAttribute('partition_key');
    }
    /**
     * @stability stable
     */
    set partitionKey(value) {
        this._partitionKey = value;
    }
    /**
     * @stability stable
     */
    resetPartitionKey() {
        this._partitionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get partitionKeyInput() {
        return this._partitionKey;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get streamName() {
        return this.getStringAttribute('stream_name');
    }
    /**
     * @stability stable
     */
    set streamName(value) {
        this._streamName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get streamNameInput() {
        return this._streamName;
    }
}
exports.IotTopicRuleErrorActionKinesisOutputReference = IotTopicRuleErrorActionKinesisOutputReference;
_l = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionKinesisOutputReference[_l] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionKinesisOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        function_arn: cdktf.stringToTerraform(struct.functionArn),
    };
}
exports.iotTopicRuleErrorActionLambdaToTerraform = iotTopicRuleErrorActionLambdaToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionLambdaOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._functionArn) {
            hasAnyValues = true;
            internalValueResult.functionArn = this._functionArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._functionArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._functionArn = value.functionArn;
        }
    }
    /**
     * @stability stable
     */
    get functionArn() {
        return this.getStringAttribute('function_arn');
    }
    /**
     * @stability stable
     */
    set functionArn(value) {
        this._functionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get functionArnInput() {
        return this._functionArn;
    }
}
exports.IotTopicRuleErrorActionLambdaOutputReference = IotTopicRuleErrorActionLambdaOutputReference;
_m = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionLambdaOutputReference[_m] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionLambdaOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionRepublishToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        qos: cdktf.numberToTerraform(struct.qos),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        topic: cdktf.stringToTerraform(struct.topic),
    };
}
exports.iotTopicRuleErrorActionRepublishToTerraform = iotTopicRuleErrorActionRepublishToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionRepublishOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._qos) {
            hasAnyValues = true;
            internalValueResult.qos = this._qos;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._topic) {
            hasAnyValues = true;
            internalValueResult.topic = this._topic;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._qos = undefined;
            this._roleArn = undefined;
            this._topic = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._qos = value.qos;
            this._roleArn = value.roleArn;
            this._topic = value.topic;
        }
    }
    /**
     * @stability stable
     */
    get qos() {
        return this.getNumberAttribute('qos');
    }
    /**
     * @stability stable
     */
    set qos(value) {
        this._qos = value;
    }
    /**
     * @stability stable
     */
    resetQos() {
        this._qos = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get qosInput() {
        return this._qos;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get topic() {
        return this.getStringAttribute('topic');
    }
    /**
     * @stability stable
     */
    set topic(value) {
        this._topic = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get topicInput() {
        return this._topic;
    }
}
exports.IotTopicRuleErrorActionRepublishOutputReference = IotTopicRuleErrorActionRepublishOutputReference;
_o = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionRepublishOutputReference[_o] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionRepublishOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        key: cdktf.stringToTerraform(struct.key),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleErrorActionS3ToTerraform = iotTopicRuleErrorActionS3ToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionS3OutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._key) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._key = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._key = value.key;
            this._roleArn = value.roleArn;
        }
    }
    /**
     * @stability stable
     */
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    /**
     * @stability stable
     */
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketNameInput() {
        return this._bucketName;
    }
    /**
     * @stability stable
     */
    get key() {
        return this.getStringAttribute('key');
    }
    /**
     * @stability stable
     */
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keyInput() {
        return this._key;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.IotTopicRuleErrorActionS3OutputReference = IotTopicRuleErrorActionS3OutputReference;
_p = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionS3OutputReference[_p] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionS3OutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionSnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_format: cdktf.stringToTerraform(struct.messageFormat),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        target_arn: cdktf.stringToTerraform(struct.targetArn),
    };
}
exports.iotTopicRuleErrorActionSnsToTerraform = iotTopicRuleErrorActionSnsToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionSnsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._messageFormat) {
            hasAnyValues = true;
            internalValueResult.messageFormat = this._messageFormat;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._targetArn) {
            hasAnyValues = true;
            internalValueResult.targetArn = this._targetArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._messageFormat = undefined;
            this._roleArn = undefined;
            this._targetArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._messageFormat = value.messageFormat;
            this._roleArn = value.roleArn;
            this._targetArn = value.targetArn;
        }
    }
    /**
     * @stability stable
     */
    get messageFormat() {
        return this.getStringAttribute('message_format');
    }
    /**
     * @stability stable
     */
    set messageFormat(value) {
        this._messageFormat = value;
    }
    /**
     * @stability stable
     */
    resetMessageFormat() {
        this._messageFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageFormatInput() {
        return this._messageFormat;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    /**
     * @stability stable
     */
    set targetArn(value) {
        this._targetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetArnInput() {
        return this._targetArn;
    }
}
exports.IotTopicRuleErrorActionSnsOutputReference = IotTopicRuleErrorActionSnsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionSnsOutputReference[_q] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionSnsOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionSqsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        queue_url: cdktf.stringToTerraform(struct.queueUrl),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        use_base64: cdktf.booleanToTerraform(struct.useBase64),
    };
}
exports.iotTopicRuleErrorActionSqsToTerraform = iotTopicRuleErrorActionSqsToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionSqsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._queueUrl) {
            hasAnyValues = true;
            internalValueResult.queueUrl = this._queueUrl;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._useBase64) {
            hasAnyValues = true;
            internalValueResult.useBase64 = this._useBase64;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._queueUrl = undefined;
            this._roleArn = undefined;
            this._useBase64 = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._queueUrl = value.queueUrl;
            this._roleArn = value.roleArn;
            this._useBase64 = value.useBase64;
        }
    }
    /**
     * @stability stable
     */
    get queueUrl() {
        return this.getStringAttribute('queue_url');
    }
    /**
     * @stability stable
     */
    set queueUrl(value) {
        this._queueUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get queueUrlInput() {
        return this._queueUrl;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get useBase64() {
        return this.getBooleanAttribute('use_base64');
    }
    /**
     * @stability stable
     */
    set useBase64(value) {
        this._useBase64 = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get useBase64Input() {
        return this._useBase64;
    }
}
exports.IotTopicRuleErrorActionSqsOutputReference = IotTopicRuleErrorActionSqsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionSqsOutputReference[_r] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionSqsOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionStepFunctionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execution_name_prefix: cdktf.stringToTerraform(struct.executionNamePrefix),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        state_machine_name: cdktf.stringToTerraform(struct.stateMachineName),
    };
}
exports.iotTopicRuleErrorActionStepFunctionsToTerraform = iotTopicRuleErrorActionStepFunctionsToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionStepFunctionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._executionNamePrefix) {
            hasAnyValues = true;
            internalValueResult.executionNamePrefix = this._executionNamePrefix;
        }
        if (this._roleArn) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._stateMachineName) {
            hasAnyValues = true;
            internalValueResult.stateMachineName = this._stateMachineName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._executionNamePrefix = undefined;
            this._roleArn = undefined;
            this._stateMachineName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._executionNamePrefix = value.executionNamePrefix;
            this._roleArn = value.roleArn;
            this._stateMachineName = value.stateMachineName;
        }
    }
    /**
     * @stability stable
     */
    get executionNamePrefix() {
        return this.getStringAttribute('execution_name_prefix');
    }
    /**
     * @stability stable
     */
    set executionNamePrefix(value) {
        this._executionNamePrefix = value;
    }
    /**
     * @stability stable
     */
    resetExecutionNamePrefix() {
        this._executionNamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get executionNamePrefixInput() {
        return this._executionNamePrefix;
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    /**
     * @stability stable
     */
    get stateMachineName() {
        return this.getStringAttribute('state_machine_name');
    }
    /**
     * @stability stable
     */
    set stateMachineName(value) {
        this._stateMachineName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stateMachineNameInput() {
        return this._stateMachineName;
    }
}
exports.IotTopicRuleErrorActionStepFunctionsOutputReference = IotTopicRuleErrorActionStepFunctionsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionStepFunctionsOutputReference[_s] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionStepFunctionsOutputReference", version: "3.0.1" };
function iotTopicRuleErrorActionToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_alarm: iotTopicRuleErrorActionCloudwatchAlarmToTerraform(struct.cloudwatchAlarm),
        cloudwatch_metric: iotTopicRuleErrorActionCloudwatchMetricToTerraform(struct.cloudwatchMetric),
        dynamodb: iotTopicRuleErrorActionDynamodbToTerraform(struct.dynamodb),
        dynamodbv2: iotTopicRuleErrorActionDynamodbv2ToTerraform(struct.dynamodbv2),
        elasticsearch: iotTopicRuleErrorActionElasticsearchToTerraform(struct.elasticsearch),
        firehose: iotTopicRuleErrorActionFirehoseToTerraform(struct.firehose),
        iot_analytics: iotTopicRuleErrorActionIotAnalyticsToTerraform(struct.iotAnalytics),
        iot_events: iotTopicRuleErrorActionIotEventsToTerraform(struct.iotEvents),
        kinesis: iotTopicRuleErrorActionKinesisToTerraform(struct.kinesis),
        lambda: iotTopicRuleErrorActionLambdaToTerraform(struct.lambda),
        republish: iotTopicRuleErrorActionRepublishToTerraform(struct.republish),
        s3: iotTopicRuleErrorActionS3ToTerraform(struct.s3),
        sns: iotTopicRuleErrorActionSnsToTerraform(struct.sns),
        sqs: iotTopicRuleErrorActionSqsToTerraform(struct.sqs),
        step_functions: iotTopicRuleErrorActionStepFunctionsToTerraform(struct.stepFunctions),
    };
}
exports.iotTopicRuleErrorActionToTerraform = iotTopicRuleErrorActionToTerraform;
/**
 * @stability stable
 */
class IotTopicRuleErrorActionOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cloudwatch_alarm - computed: false, optional: true, required: false
        this._cloudwatchAlarm = new IotTopicRuleErrorActionCloudwatchAlarmOutputReference(this, "cloudwatch_alarm", true);
        // cloudwatch_metric - computed: false, optional: true, required: false
        this._cloudwatchMetric = new IotTopicRuleErrorActionCloudwatchMetricOutputReference(this, "cloudwatch_metric", true);
        // dynamodb - computed: false, optional: true, required: false
        this._dynamodb = new IotTopicRuleErrorActionDynamodbOutputReference(this, "dynamodb", true);
        // dynamodbv2 - computed: false, optional: true, required: false
        this._dynamodbv2 = new IotTopicRuleErrorActionDynamodbv2OutputReference(this, "dynamodbv2", true);
        // elasticsearch - computed: false, optional: true, required: false
        this._elasticsearch = new IotTopicRuleErrorActionElasticsearchOutputReference(this, "elasticsearch", true);
        // firehose - computed: false, optional: true, required: false
        this._firehose = new IotTopicRuleErrorActionFirehoseOutputReference(this, "firehose", true);
        // iot_analytics - computed: false, optional: true, required: false
        this._iotAnalytics = new IotTopicRuleErrorActionIotAnalyticsOutputReference(this, "iot_analytics", true);
        // iot_events - computed: false, optional: true, required: false
        this._iotEvents = new IotTopicRuleErrorActionIotEventsOutputReference(this, "iot_events", true);
        // kinesis - computed: false, optional: true, required: false
        this._kinesis = new IotTopicRuleErrorActionKinesisOutputReference(this, "kinesis", true);
        // lambda - computed: false, optional: true, required: false
        this._lambda = new IotTopicRuleErrorActionLambdaOutputReference(this, "lambda", true);
        // republish - computed: false, optional: true, required: false
        this._republish = new IotTopicRuleErrorActionRepublishOutputReference(this, "republish", true);
        // s3 - computed: false, optional: true, required: false
        this._s3 = new IotTopicRuleErrorActionS3OutputReference(this, "s3", true);
        // sns - computed: false, optional: true, required: false
        this._sns = new IotTopicRuleErrorActionSnsOutputReference(this, "sns", true);
        // sqs - computed: false, optional: true, required: false
        this._sqs = new IotTopicRuleErrorActionSqsOutputReference(this, "sqs", true);
        // step_functions - computed: false, optional: true, required: false
        this._stepFunctions = new IotTopicRuleErrorActionStepFunctionsOutputReference(this, "step_functions", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_v = this._cloudwatchAlarm) === null || _v === void 0 ? void 0 : _v.internalValue) {
            hasAnyValues = true;
            internalValueResult.cloudwatchAlarm = (_w = this._cloudwatchAlarm) === null || _w === void 0 ? void 0 : _w.internalValue;
        }
        if ((_x = this._cloudwatchMetric) === null || _x === void 0 ? void 0 : _x.internalValue) {
            hasAnyValues = true;
            internalValueResult.cloudwatchMetric = (_y = this._cloudwatchMetric) === null || _y === void 0 ? void 0 : _y.internalValue;
        }
        if ((_z = this._dynamodb) === null || _z === void 0 ? void 0 : _z.internalValue) {
            hasAnyValues = true;
            internalValueResult.dynamodb = (_0 = this._dynamodb) === null || _0 === void 0 ? void 0 : _0.internalValue;
        }
        if ((_1 = this._dynamodbv2) === null || _1 === void 0 ? void 0 : _1.internalValue) {
            hasAnyValues = true;
            internalValueResult.dynamodbv2 = (_2 = this._dynamodbv2) === null || _2 === void 0 ? void 0 : _2.internalValue;
        }
        if ((_3 = this._elasticsearch) === null || _3 === void 0 ? void 0 : _3.internalValue) {
            hasAnyValues = true;
            internalValueResult.elasticsearch = (_4 = this._elasticsearch) === null || _4 === void 0 ? void 0 : _4.internalValue;
        }
        if ((_5 = this._firehose) === null || _5 === void 0 ? void 0 : _5.internalValue) {
            hasAnyValues = true;
            internalValueResult.firehose = (_6 = this._firehose) === null || _6 === void 0 ? void 0 : _6.internalValue;
        }
        if ((_7 = this._iotAnalytics) === null || _7 === void 0 ? void 0 : _7.internalValue) {
            hasAnyValues = true;
            internalValueResult.iotAnalytics = (_8 = this._iotAnalytics) === null || _8 === void 0 ? void 0 : _8.internalValue;
        }
        if ((_9 = this._iotEvents) === null || _9 === void 0 ? void 0 : _9.internalValue) {
            hasAnyValues = true;
            internalValueResult.iotEvents = (_10 = this._iotEvents) === null || _10 === void 0 ? void 0 : _10.internalValue;
        }
        if ((_11 = this._kinesis) === null || _11 === void 0 ? void 0 : _11.internalValue) {
            hasAnyValues = true;
            internalValueResult.kinesis = (_12 = this._kinesis) === null || _12 === void 0 ? void 0 : _12.internalValue;
        }
        if ((_13 = this._lambda) === null || _13 === void 0 ? void 0 : _13.internalValue) {
            hasAnyValues = true;
            internalValueResult.lambda = (_14 = this._lambda) === null || _14 === void 0 ? void 0 : _14.internalValue;
        }
        if ((_15 = this._republish) === null || _15 === void 0 ? void 0 : _15.internalValue) {
            hasAnyValues = true;
            internalValueResult.republish = (_16 = this._republish) === null || _16 === void 0 ? void 0 : _16.internalValue;
        }
        if ((_17 = this._s3) === null || _17 === void 0 ? void 0 : _17.internalValue) {
            hasAnyValues = true;
            internalValueResult.s3 = (_18 = this._s3) === null || _18 === void 0 ? void 0 : _18.internalValue;
        }
        if ((_19 = this._sns) === null || _19 === void 0 ? void 0 : _19.internalValue) {
            hasAnyValues = true;
            internalValueResult.sns = (_20 = this._sns) === null || _20 === void 0 ? void 0 : _20.internalValue;
        }
        if ((_21 = this._sqs) === null || _21 === void 0 ? void 0 : _21.internalValue) {
            hasAnyValues = true;
            internalValueResult.sqs = (_22 = this._sqs) === null || _22 === void 0 ? void 0 : _22.internalValue;
        }
        if ((_23 = this._stepFunctions) === null || _23 === void 0 ? void 0 : _23.internalValue) {
            hasAnyValues = true;
            internalValueResult.stepFunctions = (_24 = this._stepFunctions) === null || _24 === void 0 ? void 0 : _24.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchAlarm.internalValue = undefined;
            this._cloudwatchMetric.internalValue = undefined;
            this._dynamodb.internalValue = undefined;
            this._dynamodbv2.internalValue = undefined;
            this._elasticsearch.internalValue = undefined;
            this._firehose.internalValue = undefined;
            this._iotAnalytics.internalValue = undefined;
            this._iotEvents.internalValue = undefined;
            this._kinesis.internalValue = undefined;
            this._lambda.internalValue = undefined;
            this._republish.internalValue = undefined;
            this._s3.internalValue = undefined;
            this._sns.internalValue = undefined;
            this._sqs.internalValue = undefined;
            this._stepFunctions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchAlarm.internalValue = value.cloudwatchAlarm;
            this._cloudwatchMetric.internalValue = value.cloudwatchMetric;
            this._dynamodb.internalValue = value.dynamodb;
            this._dynamodbv2.internalValue = value.dynamodbv2;
            this._elasticsearch.internalValue = value.elasticsearch;
            this._firehose.internalValue = value.firehose;
            this._iotAnalytics.internalValue = value.iotAnalytics;
            this._iotEvents.internalValue = value.iotEvents;
            this._kinesis.internalValue = value.kinesis;
            this._lambda.internalValue = value.lambda;
            this._republish.internalValue = value.republish;
            this._s3.internalValue = value.s3;
            this._sns.internalValue = value.sns;
            this._sqs.internalValue = value.sqs;
            this._stepFunctions.internalValue = value.stepFunctions;
        }
    }
    /**
     * @stability stable
     */
    get cloudwatchAlarm() {
        return this._cloudwatchAlarm;
    }
    /**
     * @stability stable
     */
    putCloudwatchAlarm(value) {
        this._cloudwatchAlarm.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchAlarm() {
        this._cloudwatchAlarm.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchAlarmInput() {
        return this._cloudwatchAlarm.internalValue;
    }
    /**
     * @stability stable
     */
    get cloudwatchMetric() {
        return this._cloudwatchMetric;
    }
    /**
     * @stability stable
     */
    putCloudwatchMetric(value) {
        this._cloudwatchMetric.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchMetric() {
        this._cloudwatchMetric.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchMetricInput() {
        return this._cloudwatchMetric.internalValue;
    }
    /**
     * @stability stable
     */
    get dynamodb() {
        return this._dynamodb;
    }
    /**
     * @stability stable
     */
    putDynamodb(value) {
        this._dynamodb.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDynamodb() {
        this._dynamodb.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dynamodbInput() {
        return this._dynamodb.internalValue;
    }
    /**
     * @stability stable
     */
    get dynamodbv2() {
        return this._dynamodbv2;
    }
    /**
     * @stability stable
     */
    putDynamodbv2(value) {
        this._dynamodbv2.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDynamodbv2() {
        this._dynamodbv2.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dynamodbv2Input() {
        return this._dynamodbv2.internalValue;
    }
    /**
     * @stability stable
     */
    get elasticsearch() {
        return this._elasticsearch;
    }
    /**
     * @stability stable
     */
    putElasticsearch(value) {
        this._elasticsearch.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetElasticsearch() {
        this._elasticsearch.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticsearchInput() {
        return this._elasticsearch.internalValue;
    }
    /**
     * @stability stable
     */
    get firehose() {
        return this._firehose;
    }
    /**
     * @stability stable
     */
    putFirehose(value) {
        this._firehose.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetFirehose() {
        this._firehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get firehoseInput() {
        return this._firehose.internalValue;
    }
    /**
     * @stability stable
     */
    get iotAnalytics() {
        return this._iotAnalytics;
    }
    /**
     * @stability stable
     */
    putIotAnalytics(value) {
        this._iotAnalytics.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIotAnalytics() {
        this._iotAnalytics.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iotAnalyticsInput() {
        return this._iotAnalytics.internalValue;
    }
    /**
     * @stability stable
     */
    get iotEvents() {
        return this._iotEvents;
    }
    /**
     * @stability stable
     */
    putIotEvents(value) {
        this._iotEvents.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIotEvents() {
        this._iotEvents.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iotEventsInput() {
        return this._iotEvents.internalValue;
    }
    /**
     * @stability stable
     */
    get kinesis() {
        return this._kinesis;
    }
    /**
     * @stability stable
     */
    putKinesis(value) {
        this._kinesis.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetKinesis() {
        this._kinesis.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kinesisInput() {
        return this._kinesis.internalValue;
    }
    /**
     * @stability stable
     */
    get lambda() {
        return this._lambda;
    }
    /**
     * @stability stable
     */
    putLambda(value) {
        this._lambda.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLambda() {
        this._lambda.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaInput() {
        return this._lambda.internalValue;
    }
    /**
     * @stability stable
     */
    get republish() {
        return this._republish;
    }
    /**
     * @stability stable
     */
    putRepublish(value) {
        this._republish.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRepublish() {
        this._republish.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get republishInput() {
        return this._republish.internalValue;
    }
    /**
     * @stability stable
     */
    get s3() {
        return this._s3;
    }
    /**
     * @stability stable
     */
    putS3(value) {
        this._s3.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3Input() {
        return this._s3.internalValue;
    }
    /**
     * @stability stable
     */
    get sns() {
        return this._sns;
    }
    /**
     * @stability stable
     */
    putSns(value) {
        this._sns.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSns() {
        this._sns.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snsInput() {
        return this._sns.internalValue;
    }
    /**
     * @stability stable
     */
    get sqs() {
        return this._sqs;
    }
    /**
     * @stability stable
     */
    putSqs(value) {
        this._sqs.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSqs() {
        this._sqs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sqsInput() {
        return this._sqs.internalValue;
    }
    /**
     * @stability stable
     */
    get stepFunctions() {
        return this._stepFunctions;
    }
    /**
     * @stability stable
     */
    putStepFunctions(value) {
        this._stepFunctions.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetStepFunctions() {
        this._stepFunctions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stepFunctionsInput() {
        return this._stepFunctions.internalValue;
    }
}
exports.IotTopicRuleErrorActionOutputReference = IotTopicRuleErrorActionOutputReference;
_t = JSII_RTTI_SYMBOL_1;
IotTopicRuleErrorActionOutputReference[_t] = { fqn: "@cdktf/provider-aws.iot.IotTopicRuleErrorActionOutputReference", version: "3.0.1" };
function iotTopicRuleFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream_name: cdktf.stringToTerraform(struct.deliveryStreamName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        separator: cdktf.stringToTerraform(struct.separator),
    };
}
exports.iotTopicRuleFirehoseToTerraform = iotTopicRuleFirehoseToTerraform;
function iotTopicRuleIotAnalyticsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        channel_name: cdktf.stringToTerraform(struct.channelName),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleIotAnalyticsToTerraform = iotTopicRuleIotAnalyticsToTerraform;
function iotTopicRuleIotEventsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_name: cdktf.stringToTerraform(struct.inputName),
        message_id: cdktf.stringToTerraform(struct.messageId),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleIotEventsToTerraform = iotTopicRuleIotEventsToTerraform;
function iotTopicRuleKinesisToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        partition_key: cdktf.stringToTerraform(struct.partitionKey),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        stream_name: cdktf.stringToTerraform(struct.streamName),
    };
}
exports.iotTopicRuleKinesisToTerraform = iotTopicRuleKinesisToTerraform;
function iotTopicRuleLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        function_arn: cdktf.stringToTerraform(struct.functionArn),
    };
}
exports.iotTopicRuleLambdaToTerraform = iotTopicRuleLambdaToTerraform;
function iotTopicRuleRepublishToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        qos: cdktf.numberToTerraform(struct.qos),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        topic: cdktf.stringToTerraform(struct.topic),
    };
}
exports.iotTopicRuleRepublishToTerraform = iotTopicRuleRepublishToTerraform;
function iotTopicRuleS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        key: cdktf.stringToTerraform(struct.key),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.iotTopicRuleS3ToTerraform = iotTopicRuleS3ToTerraform;
function iotTopicRuleSnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message_format: cdktf.stringToTerraform(struct.messageFormat),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        target_arn: cdktf.stringToTerraform(struct.targetArn),
    };
}
exports.iotTopicRuleSnsToTerraform = iotTopicRuleSnsToTerraform;
function iotTopicRuleSqsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        queue_url: cdktf.stringToTerraform(struct.queueUrl),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        use_base64: cdktf.booleanToTerraform(struct.useBase64),
    };
}
exports.iotTopicRuleSqsToTerraform = iotTopicRuleSqsToTerraform;
function iotTopicRuleStepFunctionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        execution_name_prefix: cdktf.stringToTerraform(struct.executionNamePrefix),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        state_machine_name: cdktf.stringToTerraform(struct.stateMachineName),
    };
}
exports.iotTopicRuleStepFunctionsToTerraform = iotTopicRuleStepFunctionsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_topic_rule.html aws_iot_topic_rule}.
 *
 * @stability stable
 */
class IotTopicRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_topic_rule.html aws_iot_topic_rule} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_topic_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // error_action - computed: false, optional: true, required: false
        this._errorAction = new IotTopicRuleErrorActionOutputReference(this, "error_action", true);
        this._description = config.description;
        this._enabled = config.enabled;
        this._name = config.name;
        this._sql = config.sql;
        this._sqlVersion = config.sqlVersion;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._cloudwatchAlarm = config.cloudwatchAlarm;
        this._cloudwatchMetric = config.cloudwatchMetric;
        this._dynamodb = config.dynamodb;
        this._dynamodbv2 = config.dynamodbv2;
        this._elasticsearch = config.elasticsearch;
        this._errorAction.internalValue = config.errorAction;
        this._firehose = config.firehose;
        this._iotAnalytics = config.iotAnalytics;
        this._iotEvents = config.iotEvents;
        this._kinesis = config.kinesis;
        this._lambda = config.lambda;
        this._republish = config.republish;
        this._s3 = config.s3;
        this._sns = config.sns;
        this._sqs = config.sqs;
        this._stepFunctions = config.stepFunctions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get sql() {
        return this.getStringAttribute('sql');
    }
    /**
     * @stability stable
     */
    set sql(value) {
        this._sql = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sqlInput() {
        return this._sql;
    }
    /**
     * @stability stable
     */
    get sqlVersion() {
        return this.getStringAttribute('sql_version');
    }
    /**
     * @stability stable
     */
    set sqlVersion(value) {
        this._sqlVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sqlVersionInput() {
        return this._sqlVersion;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get cloudwatchAlarm() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('cloudwatch_alarm');
    }
    /**
     * @stability stable
     */
    set cloudwatchAlarm(value) {
        this._cloudwatchAlarm = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchAlarm() {
        this._cloudwatchAlarm = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchAlarmInput() {
        return this._cloudwatchAlarm;
    }
    /**
     * @stability stable
     */
    get cloudwatchMetric() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('cloudwatch_metric');
    }
    /**
     * @stability stable
     */
    set cloudwatchMetric(value) {
        this._cloudwatchMetric = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchMetric() {
        this._cloudwatchMetric = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchMetricInput() {
        return this._cloudwatchMetric;
    }
    /**
     * @stability stable
     */
    get dynamodb() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('dynamodb');
    }
    /**
     * @stability stable
     */
    set dynamodb(value) {
        this._dynamodb = value;
    }
    /**
     * @stability stable
     */
    resetDynamodb() {
        this._dynamodb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dynamodbInput() {
        return this._dynamodb;
    }
    /**
     * @stability stable
     */
    get dynamodbv2() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('dynamodbv2');
    }
    /**
     * @stability stable
     */
    set dynamodbv2(value) {
        this._dynamodbv2 = value;
    }
    /**
     * @stability stable
     */
    resetDynamodbv2() {
        this._dynamodbv2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dynamodbv2Input() {
        return this._dynamodbv2;
    }
    /**
     * @stability stable
     */
    get elasticsearch() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('elasticsearch');
    }
    /**
     * @stability stable
     */
    set elasticsearch(value) {
        this._elasticsearch = value;
    }
    /**
     * @stability stable
     */
    resetElasticsearch() {
        this._elasticsearch = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get elasticsearchInput() {
        return this._elasticsearch;
    }
    /**
     * @stability stable
     */
    get errorAction() {
        return this._errorAction;
    }
    /**
     * @stability stable
     */
    putErrorAction(value) {
        this._errorAction.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetErrorAction() {
        this._errorAction.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get errorActionInput() {
        return this._errorAction.internalValue;
    }
    /**
     * @stability stable
     */
    get firehose() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('firehose');
    }
    /**
     * @stability stable
     */
    set firehose(value) {
        this._firehose = value;
    }
    /**
     * @stability stable
     */
    resetFirehose() {
        this._firehose = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get firehoseInput() {
        return this._firehose;
    }
    /**
     * @stability stable
     */
    get iotAnalytics() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('iot_analytics');
    }
    /**
     * @stability stable
     */
    set iotAnalytics(value) {
        this._iotAnalytics = value;
    }
    /**
     * @stability stable
     */
    resetIotAnalytics() {
        this._iotAnalytics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iotAnalyticsInput() {
        return this._iotAnalytics;
    }
    /**
     * @stability stable
     */
    get iotEvents() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('iot_events');
    }
    /**
     * @stability stable
     */
    set iotEvents(value) {
        this._iotEvents = value;
    }
    /**
     * @stability stable
     */
    resetIotEvents() {
        this._iotEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iotEventsInput() {
        return this._iotEvents;
    }
    /**
     * @stability stable
     */
    get kinesis() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('kinesis');
    }
    /**
     * @stability stable
     */
    set kinesis(value) {
        this._kinesis = value;
    }
    /**
     * @stability stable
     */
    resetKinesis() {
        this._kinesis = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get kinesisInput() {
        return this._kinesis;
    }
    /**
     * @stability stable
     */
    get lambda() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('lambda');
    }
    /**
     * @stability stable
     */
    set lambda(value) {
        this._lambda = value;
    }
    /**
     * @stability stable
     */
    resetLambda() {
        this._lambda = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaInput() {
        return this._lambda;
    }
    /**
     * @stability stable
     */
    get republish() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('republish');
    }
    /**
     * @stability stable
     */
    set republish(value) {
        this._republish = value;
    }
    /**
     * @stability stable
     */
    resetRepublish() {
        this._republish = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get republishInput() {
        return this._republish;
    }
    /**
     * @stability stable
     */
    get s3() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('s3');
    }
    /**
     * @stability stable
     */
    set s3(value) {
        this._s3 = value;
    }
    /**
     * @stability stable
     */
    resetS3() {
        this._s3 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3Input() {
        return this._s3;
    }
    /**
     * @stability stable
     */
    get sns() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('sns');
    }
    /**
     * @stability stable
     */
    set sns(value) {
        this._sns = value;
    }
    /**
     * @stability stable
     */
    resetSns() {
        this._sns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get snsInput() {
        return this._sns;
    }
    /**
     * @stability stable
     */
    get sqs() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('sqs');
    }
    /**
     * @stability stable
     */
    set sqs(value) {
        this._sqs = value;
    }
    /**
     * @stability stable
     */
    resetSqs() {
        this._sqs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sqsInput() {
        return this._sqs;
    }
    /**
     * @stability stable
     */
    get stepFunctions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('step_functions');
    }
    /**
     * @stability stable
     */
    set stepFunctions(value) {
        this._stepFunctions = value;
    }
    /**
     * @stability stable
     */
    resetStepFunctions() {
        this._stepFunctions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stepFunctionsInput() {
        return this._stepFunctions;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            enabled: cdktf.booleanToTerraform(this._enabled),
            name: cdktf.stringToTerraform(this._name),
            sql: cdktf.stringToTerraform(this._sql),
            sql_version: cdktf.stringToTerraform(this._sqlVersion),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            cloudwatch_alarm: cdktf.listMapper(iotTopicRuleCloudwatchAlarmToTerraform)(this._cloudwatchAlarm),
            cloudwatch_metric: cdktf.listMapper(iotTopicRuleCloudwatchMetricToTerraform)(this._cloudwatchMetric),
            dynamodb: cdktf.listMapper(iotTopicRuleDynamodbToTerraform)(this._dynamodb),
            dynamodbv2: cdktf.listMapper(iotTopicRuleDynamodbv2ToTerraform)(this._dynamodbv2),
            elasticsearch: cdktf.listMapper(iotTopicRuleElasticsearchToTerraform)(this._elasticsearch),
            error_action: iotTopicRuleErrorActionToTerraform(this._errorAction.internalValue),
            firehose: cdktf.listMapper(iotTopicRuleFirehoseToTerraform)(this._firehose),
            iot_analytics: cdktf.listMapper(iotTopicRuleIotAnalyticsToTerraform)(this._iotAnalytics),
            iot_events: cdktf.listMapper(iotTopicRuleIotEventsToTerraform)(this._iotEvents),
            kinesis: cdktf.listMapper(iotTopicRuleKinesisToTerraform)(this._kinesis),
            lambda: cdktf.listMapper(iotTopicRuleLambdaToTerraform)(this._lambda),
            republish: cdktf.listMapper(iotTopicRuleRepublishToTerraform)(this._republish),
            s3: cdktf.listMapper(iotTopicRuleS3ToTerraform)(this._s3),
            sns: cdktf.listMapper(iotTopicRuleSnsToTerraform)(this._sns),
            sqs: cdktf.listMapper(iotTopicRuleSqsToTerraform)(this._sqs),
            step_functions: cdktf.listMapper(iotTopicRuleStepFunctionsToTerraform)(this._stepFunctions),
        };
    }
}
exports.IotTopicRule = IotTopicRule;
_u = JSII_RTTI_SYMBOL_1;
IotTopicRule[_u] = { fqn: "@cdktf/provider-aws.iot.IotTopicRule", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotTopicRule.tfResourceType = "aws_iot_topic_rule";
