// @cdktf/provider-aws/lib/iot/iot-thing.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThing = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing.html aws_iot_thing}.
 *
 * @stability stable
 */
class IotThing extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing.html aws_iot_thing} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._attributes = config.attributes;
        this._name = config.name;
        this._thingTypeName = config.thingTypeName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get attributes() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('attributes');
    }
    /**
     * @stability stable
     */
    set attributes(value) {
        this._attributes = value;
    }
    /**
     * @stability stable
     */
    resetAttributes() {
        this._attributes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get attributesInput() {
        return this._attributes;
    }
    // default_client_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultClientId() {
        return this.getStringAttribute('default_client_id');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get thingTypeName() {
        return this.getStringAttribute('thing_type_name');
    }
    /**
     * @stability stable
     */
    set thingTypeName(value) {
        this._thingTypeName = value;
    }
    /**
     * @stability stable
     */
    resetThingTypeName() {
        this._thingTypeName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get thingTypeNameInput() {
        return this._thingTypeName;
    }
    // version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get version() {
        return this.getNumberAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            attributes: cdktf.hashMapper(cdktf.anyToTerraform)(this._attributes),
            name: cdktf.stringToTerraform(this._name),
            thing_type_name: cdktf.stringToTerraform(this._thingTypeName),
        };
    }
}
exports.IotThing = IotThing;
_a = JSII_RTTI_SYMBOL_1;
IotThing[_a] = { fqn: "@cdktf/provider-aws.iot.IotThing", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
IotThing.tfResourceType = "aws_iot_thing";
