// @cdktf/provider-aws/lib/imagebuilder/data-aws-imagebuilder-image-pipeline.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsImagebuilderImagePipeline = exports.DataAwsImagebuilderImagePipelineSchedule = exports.DataAwsImagebuilderImagePipelineImageTestsConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsImagebuilderImagePipelineImageTestsConfiguration extends cdktf.ComplexComputedList {
    // image_tests_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageTestsEnabled() {
        return this.getBooleanAttribute('image_tests_enabled');
    }
    // timeout_minutes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get timeoutMinutes() {
        return this.getNumberAttribute('timeout_minutes');
    }
}
exports.DataAwsImagebuilderImagePipelineImageTestsConfiguration = DataAwsImagebuilderImagePipelineImageTestsConfiguration;
_a = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImagePipelineImageTestsConfiguration[_a] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImagePipelineImageTestsConfiguration", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsImagebuilderImagePipelineSchedule extends cdktf.ComplexComputedList {
    // pipeline_execution_start_condition - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get pipelineExecutionStartCondition() {
        return this.getStringAttribute('pipeline_execution_start_condition');
    }
    // schedule_expression - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get scheduleExpression() {
        return this.getStringAttribute('schedule_expression');
    }
}
exports.DataAwsImagebuilderImagePipelineSchedule = DataAwsImagebuilderImagePipelineSchedule;
_b = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImagePipelineSchedule[_b] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImagePipelineSchedule", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image_pipeline.html aws_imagebuilder_image_pipeline}.
 *
 * @stability stable
 */
class DataAwsImagebuilderImagePipeline extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image_pipeline.html aws_imagebuilder_image_pipeline} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_image_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._arn = config.arn;
        this._tags = config.tags;
    }
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get arnInput() {
        return this._arn;
    }
    // date_created - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // date_last_run - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateLastRun() {
        return this.getStringAttribute('date_last_run');
    }
    // date_next_run - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateNextRun() {
        return this.getStringAttribute('date_next_run');
    }
    // date_updated - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get dateUpdated() {
        return this.getStringAttribute('date_updated');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // distribution_configuration_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get distributionConfigurationArn() {
        return this.getStringAttribute('distribution_configuration_arn');
    }
    // enhanced_image_metadata_enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enhancedImageMetadataEnabled() {
        return this.getBooleanAttribute('enhanced_image_metadata_enabled');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image_recipe_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageRecipeArn() {
        return this.getStringAttribute('image_recipe_arn');
    }
    // image_tests_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    imageTestsConfiguration(index) {
        return new DataAwsImagebuilderImagePipelineImageTestsConfiguration(this, 'image_tests_configuration', index);
    }
    // infrastructure_configuration_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get infrastructureConfigurationArn() {
        return this.getStringAttribute('infrastructure_configuration_arn');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    // platform - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get platform() {
        return this.getStringAttribute('platform');
    }
    // schedule - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    schedule(index) {
        return new DataAwsImagebuilderImagePipelineSchedule(this, 'schedule', index);
    }
    // status - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsImagebuilderImagePipeline = DataAwsImagebuilderImagePipeline;
_c = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImagePipeline[_c] = { fqn: "@cdktf/provider-aws.imagebuilder.DataAwsImagebuilderImagePipeline", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsImagebuilderImagePipeline.tfResourceType = "aws_imagebuilder_image_pipeline";
