// @cdktf/provider-aws/lib/route53/data-aws-route53-zone.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRoute53Zone = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/route53_zone.html aws_route53_zone}.
 *
 * @stability stable
 */
class DataAwsRoute53Zone extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/route53_zone.html aws_route53_zone} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_zone',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._privateZone = config.privateZone;
        this._resourceRecordSetCount = config.resourceRecordSetCount;
        this._tags = config.tags;
        this._vpcId = config.vpcId;
        this._zoneId = config.zoneId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // caller_reference - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get callerReference() {
        return this.getStringAttribute('caller_reference');
    }
    // comment - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get comment() {
        return this.getStringAttribute('comment');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // linked_service_description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get linkedServiceDescription() {
        return this.getStringAttribute('linked_service_description');
    }
    // linked_service_principal - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get linkedServicePrincipal() {
        return this.getStringAttribute('linked_service_principal');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // name_servers - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get nameServers() {
        return this.getListAttribute('name_servers');
    }
    /**
     * @stability stable
     */
    get privateZone() {
        return this.getBooleanAttribute('private_zone');
    }
    /**
     * @stability stable
     */
    set privateZone(value) {
        this._privateZone = value;
    }
    /**
     * @stability stable
     */
    resetPrivateZone() {
        this._privateZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get privateZoneInput() {
        return this._privateZone;
    }
    /**
     * @stability stable
     */
    get resourceRecordSetCount() {
        return this.getNumberAttribute('resource_record_set_count');
    }
    /**
     * @stability stable
     */
    set resourceRecordSetCount(value) {
        this._resourceRecordSetCount = value;
    }
    /**
     * @stability stable
     */
    resetResourceRecordSetCount() {
        this._resourceRecordSetCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceRecordSetCountInput() {
        return this._resourceRecordSetCount;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    /**
     * @stability stable
     */
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
    /**
     * @stability stable
     */
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    /**
     * @stability stable
     */
    set zoneId(value) {
        this._zoneId = value;
    }
    /**
     * @stability stable
     */
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get zoneIdInput() {
        return this._zoneId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            private_zone: cdktf.booleanToTerraform(this._privateZone),
            resource_record_set_count: cdktf.numberToTerraform(this._resourceRecordSetCount),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            zone_id: cdktf.stringToTerraform(this._zoneId),
        };
    }
}
exports.DataAwsRoute53Zone = DataAwsRoute53Zone;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRoute53Zone[_a] = { fqn: "@cdktf/provider-aws.route53.DataAwsRoute53Zone", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsRoute53Zone.tfResourceType = "aws_route53_zone";
