// @cdktf/provider-aws/lib/route53/route53-hosted-zone-dnssec.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53HostedZoneDnssec = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_hosted_zone_dnssec.html aws_route53_hosted_zone_dnssec}.
 *
 * @stability stable
 */
class Route53HostedZoneDnssec extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_hosted_zone_dnssec.html aws_route53_hosted_zone_dnssec} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_hosted_zone_dnssec',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._hostedZoneId = config.hostedZoneId;
        this._signingStatus = config.signingStatus;
    }
    /**
     * @stability stable
     */
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    /**
     * @stability stable
     */
    set hostedZoneId(value) {
        this._hostedZoneId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get hostedZoneIdInput() {
        return this._hostedZoneId;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get signingStatus() {
        return this.getStringAttribute('signing_status');
    }
    /**
     * @stability stable
     */
    set signingStatus(value) {
        this._signingStatus = value;
    }
    /**
     * @stability stable
     */
    resetSigningStatus() {
        this._signingStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signingStatusInput() {
        return this._signingStatus;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            hosted_zone_id: cdktf.stringToTerraform(this._hostedZoneId),
            signing_status: cdktf.stringToTerraform(this._signingStatus),
        };
    }
}
exports.Route53HostedZoneDnssec = Route53HostedZoneDnssec;
_a = JSII_RTTI_SYMBOL_1;
Route53HostedZoneDnssec[_a] = { fqn: "@cdktf/provider-aws.route53.Route53HostedZoneDnssec", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Route53HostedZoneDnssec.tfResourceType = "aws_route53_hosted_zone_dnssec";
