// @cdktf/provider-aws/lib/route53/route53-health-check.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53HealthCheck = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_health_check.html aws_route53_health_check}.
 *
 * @stability stable
 */
class Route53HealthCheck extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_health_check.html aws_route53_health_check} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_health_check',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._childHealthThreshold = config.childHealthThreshold;
        this._childHealthchecks = config.childHealthchecks;
        this._cloudwatchAlarmName = config.cloudwatchAlarmName;
        this._cloudwatchAlarmRegion = config.cloudwatchAlarmRegion;
        this._disabled = config.disabled;
        this._enableSni = config.enableSni;
        this._failureThreshold = config.failureThreshold;
        this._fqdn = config.fqdn;
        this._insufficientDataHealthStatus = config.insufficientDataHealthStatus;
        this._invertHealthcheck = config.invertHealthcheck;
        this._ipAddress = config.ipAddress;
        this._measureLatency = config.measureLatency;
        this._port = config.port;
        this._referenceName = config.referenceName;
        this._regions = config.regions;
        this._requestInterval = config.requestInterval;
        this._resourcePath = config.resourcePath;
        this._routingControlArn = config.routingControlArn;
        this._searchString = config.searchString;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get childHealthThreshold() {
        return this.getNumberAttribute('child_health_threshold');
    }
    /**
     * @stability stable
     */
    set childHealthThreshold(value) {
        this._childHealthThreshold = value;
    }
    /**
     * @stability stable
     */
    resetChildHealthThreshold() {
        this._childHealthThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get childHealthThresholdInput() {
        return this._childHealthThreshold;
    }
    /**
     * @stability stable
     */
    get childHealthchecks() {
        return this.getListAttribute('child_healthchecks');
    }
    /**
     * @stability stable
     */
    set childHealthchecks(value) {
        this._childHealthchecks = value;
    }
    /**
     * @stability stable
     */
    resetChildHealthchecks() {
        this._childHealthchecks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get childHealthchecksInput() {
        return this._childHealthchecks;
    }
    /**
     * @stability stable
     */
    get cloudwatchAlarmName() {
        return this.getStringAttribute('cloudwatch_alarm_name');
    }
    /**
     * @stability stable
     */
    set cloudwatchAlarmName(value) {
        this._cloudwatchAlarmName = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchAlarmName() {
        this._cloudwatchAlarmName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchAlarmNameInput() {
        return this._cloudwatchAlarmName;
    }
    /**
     * @stability stable
     */
    get cloudwatchAlarmRegion() {
        return this.getStringAttribute('cloudwatch_alarm_region');
    }
    /**
     * @stability stable
     */
    set cloudwatchAlarmRegion(value) {
        this._cloudwatchAlarmRegion = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchAlarmRegion() {
        this._cloudwatchAlarmRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchAlarmRegionInput() {
        return this._cloudwatchAlarmRegion;
    }
    /**
     * @stability stable
     */
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    /**
     * @stability stable
     */
    set disabled(value) {
        this._disabled = value;
    }
    /**
     * @stability stable
     */
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disabledInput() {
        return this._disabled;
    }
    /**
     * @stability stable
     */
    get enableSni() {
        return this.getBooleanAttribute('enable_sni');
    }
    /**
     * @stability stable
     */
    set enableSni(value) {
        this._enableSni = value;
    }
    /**
     * @stability stable
     */
    resetEnableSni() {
        this._enableSni = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableSniInput() {
        return this._enableSni;
    }
    /**
     * @stability stable
     */
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    /**
     * @stability stable
     */
    set failureThreshold(value) {
        this._failureThreshold = value;
    }
    /**
     * @stability stable
     */
    resetFailureThreshold() {
        this._failureThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get failureThresholdInput() {
        return this._failureThreshold;
    }
    /**
     * @stability stable
     */
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    /**
     * @stability stable
     */
    set fqdn(value) {
        this._fqdn = value;
    }
    /**
     * @stability stable
     */
    resetFqdn() {
        this._fqdn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get fqdnInput() {
        return this._fqdn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get insufficientDataHealthStatus() {
        return this.getStringAttribute('insufficient_data_health_status');
    }
    /**
     * @stability stable
     */
    set insufficientDataHealthStatus(value) {
        this._insufficientDataHealthStatus = value;
    }
    /**
     * @stability stable
     */
    resetInsufficientDataHealthStatus() {
        this._insufficientDataHealthStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get insufficientDataHealthStatusInput() {
        return this._insufficientDataHealthStatus;
    }
    /**
     * @stability stable
     */
    get invertHealthcheck() {
        return this.getBooleanAttribute('invert_healthcheck');
    }
    /**
     * @stability stable
     */
    set invertHealthcheck(value) {
        this._invertHealthcheck = value;
    }
    /**
     * @stability stable
     */
    resetInvertHealthcheck() {
        this._invertHealthcheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get invertHealthcheckInput() {
        return this._invertHealthcheck;
    }
    /**
     * @stability stable
     */
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    /**
     * @stability stable
     */
    set ipAddress(value) {
        this._ipAddress = value;
    }
    /**
     * @stability stable
     */
    resetIpAddress() {
        this._ipAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipAddressInput() {
        return this._ipAddress;
    }
    /**
     * @stability stable
     */
    get measureLatency() {
        return this.getBooleanAttribute('measure_latency');
    }
    /**
     * @stability stable
     */
    set measureLatency(value) {
        this._measureLatency = value;
    }
    /**
     * @stability stable
     */
    resetMeasureLatency() {
        this._measureLatency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get measureLatencyInput() {
        return this._measureLatency;
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    /**
     * @stability stable
     */
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    /**
     * @stability stable
     */
    get referenceName() {
        return this.getStringAttribute('reference_name');
    }
    /**
     * @stability stable
     */
    set referenceName(value) {
        this._referenceName = value;
    }
    /**
     * @stability stable
     */
    resetReferenceName() {
        this._referenceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get referenceNameInput() {
        return this._referenceName;
    }
    /**
     * @stability stable
     */
    get regions() {
        return this.getListAttribute('regions');
    }
    /**
     * @stability stable
     */
    set regions(value) {
        this._regions = value;
    }
    /**
     * @stability stable
     */
    resetRegions() {
        this._regions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionsInput() {
        return this._regions;
    }
    /**
     * @stability stable
     */
    get requestInterval() {
        return this.getNumberAttribute('request_interval');
    }
    /**
     * @stability stable
     */
    set requestInterval(value) {
        this._requestInterval = value;
    }
    /**
     * @stability stable
     */
    resetRequestInterval() {
        this._requestInterval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get requestIntervalInput() {
        return this._requestInterval;
    }
    /**
     * @stability stable
     */
    get resourcePath() {
        return this.getStringAttribute('resource_path');
    }
    /**
     * @stability stable
     */
    set resourcePath(value) {
        this._resourcePath = value;
    }
    /**
     * @stability stable
     */
    resetResourcePath() {
        this._resourcePath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourcePathInput() {
        return this._resourcePath;
    }
    /**
     * @stability stable
     */
    get routingControlArn() {
        return this.getStringAttribute('routing_control_arn');
    }
    /**
     * @stability stable
     */
    set routingControlArn(value) {
        this._routingControlArn = value;
    }
    /**
     * @stability stable
     */
    resetRoutingControlArn() {
        this._routingControlArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get routingControlArnInput() {
        return this._routingControlArn;
    }
    /**
     * @stability stable
     */
    get searchString() {
        return this.getStringAttribute('search_string');
    }
    /**
     * @stability stable
     */
    set searchString(value) {
        this._searchString = value;
    }
    /**
     * @stability stable
     */
    resetSearchString() {
        this._searchString = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get searchStringInput() {
        return this._searchString;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            child_health_threshold: cdktf.numberToTerraform(this._childHealthThreshold),
            child_healthchecks: cdktf.listMapper(cdktf.stringToTerraform)(this._childHealthchecks),
            cloudwatch_alarm_name: cdktf.stringToTerraform(this._cloudwatchAlarmName),
            cloudwatch_alarm_region: cdktf.stringToTerraform(this._cloudwatchAlarmRegion),
            disabled: cdktf.booleanToTerraform(this._disabled),
            enable_sni: cdktf.booleanToTerraform(this._enableSni),
            failure_threshold: cdktf.numberToTerraform(this._failureThreshold),
            fqdn: cdktf.stringToTerraform(this._fqdn),
            insufficient_data_health_status: cdktf.stringToTerraform(this._insufficientDataHealthStatus),
            invert_healthcheck: cdktf.booleanToTerraform(this._invertHealthcheck),
            ip_address: cdktf.stringToTerraform(this._ipAddress),
            measure_latency: cdktf.booleanToTerraform(this._measureLatency),
            port: cdktf.numberToTerraform(this._port),
            reference_name: cdktf.stringToTerraform(this._referenceName),
            regions: cdktf.listMapper(cdktf.stringToTerraform)(this._regions),
            request_interval: cdktf.numberToTerraform(this._requestInterval),
            resource_path: cdktf.stringToTerraform(this._resourcePath),
            routing_control_arn: cdktf.stringToTerraform(this._routingControlArn),
            search_string: cdktf.stringToTerraform(this._searchString),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
        };
    }
}
exports.Route53HealthCheck = Route53HealthCheck;
_a = JSII_RTTI_SYMBOL_1;
Route53HealthCheck[_a] = { fqn: "@cdktf/provider-aws.route53.Route53HealthCheck", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Route53HealthCheck.tfResourceType = "aws_route53_health_check";
