// @cdktf/provider-aws/lib/chime/chime-voice-connector-termination.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChimeVoiceConnectorTermination = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/chime_voice_connector_termination.html aws_chime_voice_connector_termination}.
 *
 * @stability stable
 */
class ChimeVoiceConnectorTermination extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/chime_voice_connector_termination.html aws_chime_voice_connector_termination} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_chime_voice_connector_termination',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._callingRegions = config.callingRegions;
        this._cidrAllowList = config.cidrAllowList;
        this._cpsLimit = config.cpsLimit;
        this._defaultPhoneNumber = config.defaultPhoneNumber;
        this._disabled = config.disabled;
        this._voiceConnectorId = config.voiceConnectorId;
    }
    /**
     * @stability stable
     */
    get callingRegions() {
        return this.getListAttribute('calling_regions');
    }
    /**
     * @stability stable
     */
    set callingRegions(value) {
        this._callingRegions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get callingRegionsInput() {
        return this._callingRegions;
    }
    /**
     * @stability stable
     */
    get cidrAllowList() {
        return this.getListAttribute('cidr_allow_list');
    }
    /**
     * @stability stable
     */
    set cidrAllowList(value) {
        this._cidrAllowList = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cidrAllowListInput() {
        return this._cidrAllowList;
    }
    /**
     * @stability stable
     */
    get cpsLimit() {
        return this.getNumberAttribute('cps_limit');
    }
    /**
     * @stability stable
     */
    set cpsLimit(value) {
        this._cpsLimit = value;
    }
    /**
     * @stability stable
     */
    resetCpsLimit() {
        this._cpsLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cpsLimitInput() {
        return this._cpsLimit;
    }
    /**
     * @stability stable
     */
    get defaultPhoneNumber() {
        return this.getStringAttribute('default_phone_number');
    }
    /**
     * @stability stable
     */
    set defaultPhoneNumber(value) {
        this._defaultPhoneNumber = value;
    }
    /**
     * @stability stable
     */
    resetDefaultPhoneNumber() {
        this._defaultPhoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get defaultPhoneNumberInput() {
        return this._defaultPhoneNumber;
    }
    /**
     * @stability stable
     */
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    /**
     * @stability stable
     */
    set disabled(value) {
        this._disabled = value;
    }
    /**
     * @stability stable
     */
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get disabledInput() {
        return this._disabled;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get voiceConnectorId() {
        return this.getStringAttribute('voice_connector_id');
    }
    /**
     * @stability stable
     */
    set voiceConnectorId(value) {
        this._voiceConnectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get voiceConnectorIdInput() {
        return this._voiceConnectorId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            calling_regions: cdktf.listMapper(cdktf.stringToTerraform)(this._callingRegions),
            cidr_allow_list: cdktf.listMapper(cdktf.stringToTerraform)(this._cidrAllowList),
            cps_limit: cdktf.numberToTerraform(this._cpsLimit),
            default_phone_number: cdktf.stringToTerraform(this._defaultPhoneNumber),
            disabled: cdktf.booleanToTerraform(this._disabled),
            voice_connector_id: cdktf.stringToTerraform(this._voiceConnectorId),
        };
    }
}
exports.ChimeVoiceConnectorTermination = ChimeVoiceConnectorTermination;
_a = JSII_RTTI_SYMBOL_1;
ChimeVoiceConnectorTermination[_a] = { fqn: "@cdktf/provider-aws.chime.ChimeVoiceConnectorTermination", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ChimeVoiceConnectorTermination.tfResourceType = "aws_chime_voice_connector_termination";
