// @cdktf/provider-aws/lib/pinpoint/pinpoint-event-stream.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinpointEventStream = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/pinpoint_event_stream.html aws_pinpoint_event_stream}.
 *
 * @stability stable
 */
class PinpointEventStream extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/pinpoint_event_stream.html aws_pinpoint_event_stream} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_pinpoint_event_stream',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._applicationId = config.applicationId;
        this._destinationStreamArn = config.destinationStreamArn;
        this._roleArn = config.roleArn;
    }
    /**
     * @stability stable
     */
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    /**
     * @stability stable
     */
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applicationIdInput() {
        return this._applicationId;
    }
    /**
     * @stability stable
     */
    get destinationStreamArn() {
        return this.getStringAttribute('destination_stream_arn');
    }
    /**
     * @stability stable
     */
    set destinationStreamArn(value) {
        this._destinationStreamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationStreamArnInput() {
        return this._destinationStreamArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    /**
     * @stability stable
     */
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleArnInput() {
        return this._roleArn;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            destination_stream_arn: cdktf.stringToTerraform(this._destinationStreamArn),
            role_arn: cdktf.stringToTerraform(this._roleArn),
        };
    }
}
exports.PinpointEventStream = PinpointEventStream;
_a = JSII_RTTI_SYMBOL_1;
PinpointEventStream[_a] = { fqn: "@cdktf/provider-aws.pinpoint.PinpointEventStream", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
PinpointEventStream.tfResourceType = "aws_pinpoint_event_stream";
