// @cdktf/provider-aws/lib/workspaces/data-aws-workspaces-bundle.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsWorkspacesBundle = exports.DataAwsWorkspacesBundleUserStorage = exports.DataAwsWorkspacesBundleRootStorage = exports.DataAwsWorkspacesBundleComputeType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsWorkspacesBundleComputeType extends cdktf.ComplexComputedList {
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsWorkspacesBundleComputeType = DataAwsWorkspacesBundleComputeType;
_a = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesBundleComputeType[_a] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesBundleComputeType", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsWorkspacesBundleRootStorage extends cdktf.ComplexComputedList {
    // capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get capacity() {
        return this.getStringAttribute('capacity');
    }
}
exports.DataAwsWorkspacesBundleRootStorage = DataAwsWorkspacesBundleRootStorage;
_b = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesBundleRootStorage[_b] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesBundleRootStorage", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsWorkspacesBundleUserStorage extends cdktf.ComplexComputedList {
    // capacity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get capacity() {
        return this.getStringAttribute('capacity');
    }
}
exports.DataAwsWorkspacesBundleUserStorage = DataAwsWorkspacesBundleUserStorage;
_c = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesBundleUserStorage[_c] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesBundleUserStorage", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/workspaces_bundle.html aws_workspaces_bundle}.
 *
 * @stability stable
 */
class DataAwsWorkspacesBundle extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/workspaces_bundle.html aws_workspaces_bundle} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_workspaces_bundle',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._bundleId = config.bundleId;
        this._name = config.name;
        this._owner = config.owner;
    }
    /**
     * @stability stable
     */
    get bundleId() {
        return this.getStringAttribute('bundle_id');
    }
    /**
     * @stability stable
     */
    set bundleId(value) {
        this._bundleId = value;
    }
    /**
     * @stability stable
     */
    resetBundleId() {
        this._bundleId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bundleIdInput() {
        return this._bundleId;
    }
    // compute_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    computeType(index) {
        return new DataAwsWorkspacesBundleComputeType(this, 'compute_type', index);
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get owner() {
        return this.getStringAttribute('owner');
    }
    /**
     * @stability stable
     */
    set owner(value) {
        this._owner = value;
    }
    /**
     * @stability stable
     */
    resetOwner() {
        this._owner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerInput() {
        return this._owner;
    }
    // root_storage - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    rootStorage(index) {
        return new DataAwsWorkspacesBundleRootStorage(this, 'root_storage', index);
    }
    // user_storage - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    userStorage(index) {
        return new DataAwsWorkspacesBundleUserStorage(this, 'user_storage', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            bundle_id: cdktf.stringToTerraform(this._bundleId),
            name: cdktf.stringToTerraform(this._name),
            owner: cdktf.stringToTerraform(this._owner),
        };
    }
}
exports.DataAwsWorkspacesBundle = DataAwsWorkspacesBundle;
_d = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesBundle[_d] = { fqn: "@cdktf/provider-aws.workspaces.DataAwsWorkspacesBundle", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsWorkspacesBundle.tfResourceType = "aws_workspaces_bundle";
