// @cdktf/provider-aws/lib/vpc/vpc-ipam.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcIpam = exports.vpcIpamOperatingRegionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcIpamOperatingRegionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        region_name: cdktf.stringToTerraform(struct.regionName),
    };
}
exports.vpcIpamOperatingRegionsToTerraform = vpcIpamOperatingRegionsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam.html aws_vpc_ipam}.
 *
 * @stability stable
 */
class VpcIpam extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam.html aws_vpc_ipam} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._description = config.description;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._operatingRegions = config.operatingRegions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // private_default_scope_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateDefaultScopeId() {
        return this.getStringAttribute('private_default_scope_id');
    }
    // public_default_scope_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicDefaultScopeId() {
        return this.getStringAttribute('public_default_scope_id');
    }
    // scope_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get scopeCount() {
        return this.getNumberAttribute('scope_count');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get operatingRegions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('operating_regions');
    }
    /**
     * @stability stable
     */
    set operatingRegions(value) {
        this._operatingRegions = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get operatingRegionsInput() {
        return this._operatingRegions;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            operating_regions: cdktf.listMapper(vpcIpamOperatingRegionsToTerraform)(this._operatingRegions),
        };
    }
}
exports.VpcIpam = VpcIpam;
_a = JSII_RTTI_SYMBOL_1;
VpcIpam[_a] = { fqn: "@cdktf/provider-aws.vpc.VpcIpam", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
VpcIpam.tfResourceType = "aws_vpc_ipam";
