// @cdktf/provider-aws/lib/vpc/vpn-connection-route.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpnConnectionRoute = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpn_connection_route.html aws_vpn_connection_route}.
 *
 * @stability stable
 */
class VpnConnectionRoute extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpn_connection_route.html aws_vpn_connection_route} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpn_connection_route',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._destinationCidrBlock = config.destinationCidrBlock;
        this._vpnConnectionId = config.vpnConnectionId;
    }
    /**
     * @stability stable
     */
    get destinationCidrBlock() {
        return this.getStringAttribute('destination_cidr_block');
    }
    /**
     * @stability stable
     */
    set destinationCidrBlock(value) {
        this._destinationCidrBlock = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get destinationCidrBlockInput() {
        return this._destinationCidrBlock;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get vpnConnectionId() {
        return this.getStringAttribute('vpn_connection_id');
    }
    /**
     * @stability stable
     */
    set vpnConnectionId(value) {
        this._vpnConnectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpnConnectionIdInput() {
        return this._vpnConnectionId;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            destination_cidr_block: cdktf.stringToTerraform(this._destinationCidrBlock),
            vpn_connection_id: cdktf.stringToTerraform(this._vpnConnectionId),
        };
    }
}
exports.VpnConnectionRoute = VpnConnectionRoute;
_a = JSII_RTTI_SYMBOL_1;
VpnConnectionRoute[_a] = { fqn: "@cdktf/provider-aws.vpc.VpnConnectionRoute", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
VpnConnectionRoute.tfResourceType = "aws_vpn_connection_route";
