// cdktf-cli/lib/client/react.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLServerProvider = void 0;
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const ws_1 = __importDefault(require("ws"));
const client_1 = require("@apollo/client");
const utilities_1 = require("@apollo/client/utilities");
const ws_2 = require("@apollo/client/link/ws");
const error_1 = require("@apollo/client/link/error");
const logging_1 = require("../logging");
const ink_1 = require("ink");
const server_1 = require("../server");
const react_1 = __importStar(require("react"));
const subscriptions_transport_ws_1 = require("subscriptions-transport-ws");
const useAsyncEffect = (fun, deps) => {
    react_1.useEffect(() => {
        fun();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, deps);
};
exports.GraphQLServerProvider = ({ children, }) => {
    const [client, setClient] = react_1.useState();
    useAsyncEffect(async () => {
        const { port } = await server_1.startServer();
        const webSocketUri = `ws://localhost:${port}/graphql`;
        const wsClient = new subscriptions_transport_ws_1.SubscriptionClient(webSocketUri, {
            reconnect: true,
            connectionCallback: (error) => logging_1.logger.debug(error || "CONNECTED to GraphqlEndpoint"),
        }, ws_1.default);
        wsClient.onError((event) => {
            throw new Error(`Watch could not connect to spawned GraphQL server which handles watching of files and deploying changes: ${event.message}
More debug info can be found in cdktf.log. Enable logging by setting the environment variable CDKTF_DISABLE_LOGGING to false`);
        });
        const wsLink = new ws_2.WebSocketLink(wsClient);
        const httpLink = new client_1.HttpLink({
            uri: `http://localhost:${port}/graphql`,
            fetch: cross_fetch_1.default,
        });
        // Log any GraphQL errors or network error that occurred
        const errorLink = error_1.onError(({ graphQLErrors, networkError }) => {
            if (graphQLErrors)
                graphQLErrors.map(({ message, locations, path }) => logging_1.logger.error(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`));
            if (networkError)
                console.log(`[Network error]: ${networkError}`);
        });
        const networkLink = client_1.split(({ query }) => {
            const definition = utilities_1.getMainDefinition(query);
            return (definition.kind === "OperationDefinition" &&
                definition.operation === "subscription");
        }, wsLink, httpLink);
        const apolloClient = new client_1.ApolloClient({
            link: client_1.from([errorLink, networkLink]),
            cache: new client_1.InMemoryCache(),
        });
        setClient(apolloClient);
    }, []);
    if (!client) {
        return react_1.default.createElement(ink_1.Text, null, "Starting server...");
    }
    return react_1.default.createElement(client_1.ApolloProvider, { client: client }, children);
};
