// @cdktf/provider-aws/lib/glue/glue-partition.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GluePartition = exports.GluePartitionStorageDescriptorOutputReference = exports.gluePartitionStorageDescriptorToTerraform = exports.gluePartitionStorageDescriptorSortColumnsToTerraform = exports.GluePartitionStorageDescriptorSkewedInfoOutputReference = exports.gluePartitionStorageDescriptorSkewedInfoToTerraform = exports.GluePartitionStorageDescriptorSerDeInfoOutputReference = exports.gluePartitionStorageDescriptorSerDeInfoToTerraform = exports.gluePartitionStorageDescriptorColumnsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gluePartitionStorageDescriptorColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comment: cdktf.stringToTerraform(struct.comment),
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.gluePartitionStorageDescriptorColumnsToTerraform = gluePartitionStorageDescriptorColumnsToTerraform;
function gluePartitionStorageDescriptorSerDeInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        parameters: cdktf.hashMapper(cdktf.anyToTerraform)(struct.parameters),
        serialization_library: cdktf.stringToTerraform(struct.serializationLibrary),
    };
}
exports.gluePartitionStorageDescriptorSerDeInfoToTerraform = gluePartitionStorageDescriptorSerDeInfoToTerraform;
/**
 * @stability stable
 */
class GluePartitionStorageDescriptorSerDeInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._parameters) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._serializationLibrary) {
            hasAnyValues = true;
            internalValueResult.serializationLibrary = this._serializationLibrary;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._parameters = undefined;
            this._serializationLibrary = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._parameters = value.parameters;
            this._serializationLibrary = value.serializationLibrary;
        }
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get serializationLibrary() {
        return this.getStringAttribute('serialization_library');
    }
    /**
     * @stability stable
     */
    set serializationLibrary(value) {
        this._serializationLibrary = value;
    }
    /**
     * @stability stable
     */
    resetSerializationLibrary() {
        this._serializationLibrary = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serializationLibraryInput() {
        return this._serializationLibrary;
    }
}
exports.GluePartitionStorageDescriptorSerDeInfoOutputReference = GluePartitionStorageDescriptorSerDeInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GluePartitionStorageDescriptorSerDeInfoOutputReference[_a] = { fqn: "@cdktf/provider-aws.glue.GluePartitionStorageDescriptorSerDeInfoOutputReference", version: "3.0.1" };
function gluePartitionStorageDescriptorSkewedInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        skewed_column_names: cdktf.listMapper(cdktf.stringToTerraform)(struct.skewedColumnNames),
        skewed_column_value_location_maps: cdktf.hashMapper(cdktf.anyToTerraform)(struct.skewedColumnValueLocationMaps),
        skewed_column_values: cdktf.listMapper(cdktf.stringToTerraform)(struct.skewedColumnValues),
    };
}
exports.gluePartitionStorageDescriptorSkewedInfoToTerraform = gluePartitionStorageDescriptorSkewedInfoToTerraform;
/**
 * @stability stable
 */
class GluePartitionStorageDescriptorSkewedInfoOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._skewedColumnNames) {
            hasAnyValues = true;
            internalValueResult.skewedColumnNames = this._skewedColumnNames;
        }
        if (this._skewedColumnValueLocationMaps) {
            hasAnyValues = true;
            internalValueResult.skewedColumnValueLocationMaps = this._skewedColumnValueLocationMaps;
        }
        if (this._skewedColumnValues) {
            hasAnyValues = true;
            internalValueResult.skewedColumnValues = this._skewedColumnValues;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._skewedColumnNames = undefined;
            this._skewedColumnValueLocationMaps = undefined;
            this._skewedColumnValues = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._skewedColumnNames = value.skewedColumnNames;
            this._skewedColumnValueLocationMaps = value.skewedColumnValueLocationMaps;
            this._skewedColumnValues = value.skewedColumnValues;
        }
    }
    /**
     * @stability stable
     */
    get skewedColumnNames() {
        return this.getListAttribute('skewed_column_names');
    }
    /**
     * @stability stable
     */
    set skewedColumnNames(value) {
        this._skewedColumnNames = value;
    }
    /**
     * @stability stable
     */
    resetSkewedColumnNames() {
        this._skewedColumnNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skewedColumnNamesInput() {
        return this._skewedColumnNames;
    }
    /**
     * @stability stable
     */
    get skewedColumnValueLocationMaps() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('skewed_column_value_location_maps');
    }
    /**
     * @stability stable
     */
    set skewedColumnValueLocationMaps(value) {
        this._skewedColumnValueLocationMaps = value;
    }
    /**
     * @stability stable
     */
    resetSkewedColumnValueLocationMaps() {
        this._skewedColumnValueLocationMaps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skewedColumnValueLocationMapsInput() {
        return this._skewedColumnValueLocationMaps;
    }
    /**
     * @stability stable
     */
    get skewedColumnValues() {
        return this.getListAttribute('skewed_column_values');
    }
    /**
     * @stability stable
     */
    set skewedColumnValues(value) {
        this._skewedColumnValues = value;
    }
    /**
     * @stability stable
     */
    resetSkewedColumnValues() {
        this._skewedColumnValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skewedColumnValuesInput() {
        return this._skewedColumnValues;
    }
}
exports.GluePartitionStorageDescriptorSkewedInfoOutputReference = GluePartitionStorageDescriptorSkewedInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GluePartitionStorageDescriptorSkewedInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.glue.GluePartitionStorageDescriptorSkewedInfoOutputReference", version: "3.0.1" };
function gluePartitionStorageDescriptorSortColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        column: cdktf.stringToTerraform(struct.column),
        sort_order: cdktf.numberToTerraform(struct.sortOrder),
    };
}
exports.gluePartitionStorageDescriptorSortColumnsToTerraform = gluePartitionStorageDescriptorSortColumnsToTerraform;
function gluePartitionStorageDescriptorToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_columns: cdktf.listMapper(cdktf.stringToTerraform)(struct.bucketColumns),
        compressed: cdktf.booleanToTerraform(struct.compressed),
        input_format: cdktf.stringToTerraform(struct.inputFormat),
        location: cdktf.stringToTerraform(struct.location),
        number_of_buckets: cdktf.numberToTerraform(struct.numberOfBuckets),
        output_format: cdktf.stringToTerraform(struct.outputFormat),
        parameters: cdktf.hashMapper(cdktf.anyToTerraform)(struct.parameters),
        stored_as_sub_directories: cdktf.booleanToTerraform(struct.storedAsSubDirectories),
        columns: cdktf.listMapper(gluePartitionStorageDescriptorColumnsToTerraform)(struct.columns),
        ser_de_info: gluePartitionStorageDescriptorSerDeInfoToTerraform(struct.serDeInfo),
        skewed_info: gluePartitionStorageDescriptorSkewedInfoToTerraform(struct.skewedInfo),
        sort_columns: cdktf.listMapper(gluePartitionStorageDescriptorSortColumnsToTerraform)(struct.sortColumns),
    };
}
exports.gluePartitionStorageDescriptorToTerraform = gluePartitionStorageDescriptorToTerraform;
/**
 * @stability stable
 */
class GluePartitionStorageDescriptorOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // ser_de_info - computed: false, optional: true, required: false
        this._serDeInfo = new GluePartitionStorageDescriptorSerDeInfoOutputReference(this, "ser_de_info", true);
        // skewed_info - computed: false, optional: true, required: false
        this._skewedInfo = new GluePartitionStorageDescriptorSkewedInfoOutputReference(this, "skewed_info", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _e, _f, _g, _h;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketColumns) {
            hasAnyValues = true;
            internalValueResult.bucketColumns = this._bucketColumns;
        }
        if (this._compressed) {
            hasAnyValues = true;
            internalValueResult.compressed = this._compressed;
        }
        if (this._inputFormat) {
            hasAnyValues = true;
            internalValueResult.inputFormat = this._inputFormat;
        }
        if (this._location) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._numberOfBuckets) {
            hasAnyValues = true;
            internalValueResult.numberOfBuckets = this._numberOfBuckets;
        }
        if (this._outputFormat) {
            hasAnyValues = true;
            internalValueResult.outputFormat = this._outputFormat;
        }
        if (this._parameters) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._storedAsSubDirectories) {
            hasAnyValues = true;
            internalValueResult.storedAsSubDirectories = this._storedAsSubDirectories;
        }
        if (this._columns) {
            hasAnyValues = true;
            internalValueResult.columns = this._columns;
        }
        if ((_e = this._serDeInfo) === null || _e === void 0 ? void 0 : _e.internalValue) {
            hasAnyValues = true;
            internalValueResult.serDeInfo = (_f = this._serDeInfo) === null || _f === void 0 ? void 0 : _f.internalValue;
        }
        if ((_g = this._skewedInfo) === null || _g === void 0 ? void 0 : _g.internalValue) {
            hasAnyValues = true;
            internalValueResult.skewedInfo = (_h = this._skewedInfo) === null || _h === void 0 ? void 0 : _h.internalValue;
        }
        if (this._sortColumns) {
            hasAnyValues = true;
            internalValueResult.sortColumns = this._sortColumns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketColumns = undefined;
            this._compressed = undefined;
            this._inputFormat = undefined;
            this._location = undefined;
            this._numberOfBuckets = undefined;
            this._outputFormat = undefined;
            this._parameters = undefined;
            this._storedAsSubDirectories = undefined;
            this._columns = undefined;
            this._serDeInfo.internalValue = undefined;
            this._skewedInfo.internalValue = undefined;
            this._sortColumns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketColumns = value.bucketColumns;
            this._compressed = value.compressed;
            this._inputFormat = value.inputFormat;
            this._location = value.location;
            this._numberOfBuckets = value.numberOfBuckets;
            this._outputFormat = value.outputFormat;
            this._parameters = value.parameters;
            this._storedAsSubDirectories = value.storedAsSubDirectories;
            this._columns = value.columns;
            this._serDeInfo.internalValue = value.serDeInfo;
            this._skewedInfo.internalValue = value.skewedInfo;
            this._sortColumns = value.sortColumns;
        }
    }
    /**
     * @stability stable
     */
    get bucketColumns() {
        return this.getListAttribute('bucket_columns');
    }
    /**
     * @stability stable
     */
    set bucketColumns(value) {
        this._bucketColumns = value;
    }
    /**
     * @stability stable
     */
    resetBucketColumns() {
        this._bucketColumns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get bucketColumnsInput() {
        return this._bucketColumns;
    }
    /**
     * @stability stable
     */
    get compressed() {
        return this.getBooleanAttribute('compressed');
    }
    /**
     * @stability stable
     */
    set compressed(value) {
        this._compressed = value;
    }
    /**
     * @stability stable
     */
    resetCompressed() {
        this._compressed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get compressedInput() {
        return this._compressed;
    }
    /**
     * @stability stable
     */
    get inputFormat() {
        return this.getStringAttribute('input_format');
    }
    /**
     * @stability stable
     */
    set inputFormat(value) {
        this._inputFormat = value;
    }
    /**
     * @stability stable
     */
    resetInputFormat() {
        this._inputFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputFormatInput() {
        return this._inputFormat;
    }
    /**
     * @stability stable
     */
    get location() {
        return this.getStringAttribute('location');
    }
    /**
     * @stability stable
     */
    set location(value) {
        this._location = value;
    }
    /**
     * @stability stable
     */
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get locationInput() {
        return this._location;
    }
    /**
     * @stability stable
     */
    get numberOfBuckets() {
        return this.getNumberAttribute('number_of_buckets');
    }
    /**
     * @stability stable
     */
    set numberOfBuckets(value) {
        this._numberOfBuckets = value;
    }
    /**
     * @stability stable
     */
    resetNumberOfBuckets() {
        this._numberOfBuckets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get numberOfBucketsInput() {
        return this._numberOfBuckets;
    }
    /**
     * @stability stable
     */
    get outputFormat() {
        return this.getStringAttribute('output_format');
    }
    /**
     * @stability stable
     */
    set outputFormat(value) {
        this._outputFormat = value;
    }
    /**
     * @stability stable
     */
    resetOutputFormat() {
        this._outputFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputFormatInput() {
        return this._outputFormat;
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get storedAsSubDirectories() {
        return this.getBooleanAttribute('stored_as_sub_directories');
    }
    /**
     * @stability stable
     */
    set storedAsSubDirectories(value) {
        this._storedAsSubDirectories = value;
    }
    /**
     * @stability stable
     */
    resetStoredAsSubDirectories() {
        this._storedAsSubDirectories = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storedAsSubDirectoriesInput() {
        return this._storedAsSubDirectories;
    }
    /**
     * @stability stable
     */
    get columns() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('columns');
    }
    /**
     * @stability stable
     */
    set columns(value) {
        this._columns = value;
    }
    /**
     * @stability stable
     */
    resetColumns() {
        this._columns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get columnsInput() {
        return this._columns;
    }
    /**
     * @stability stable
     */
    get serDeInfo() {
        return this._serDeInfo;
    }
    /**
     * @stability stable
     */
    putSerDeInfo(value) {
        this._serDeInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSerDeInfo() {
        this._serDeInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serDeInfoInput() {
        return this._serDeInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get skewedInfo() {
        return this._skewedInfo;
    }
    /**
     * @stability stable
     */
    putSkewedInfo(value) {
        this._skewedInfo.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSkewedInfo() {
        this._skewedInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get skewedInfoInput() {
        return this._skewedInfo.internalValue;
    }
    /**
     * @stability stable
     */
    get sortColumns() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('sort_columns');
    }
    /**
     * @stability stable
     */
    set sortColumns(value) {
        this._sortColumns = value;
    }
    /**
     * @stability stable
     */
    resetSortColumns() {
        this._sortColumns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get sortColumnsInput() {
        return this._sortColumns;
    }
}
exports.GluePartitionStorageDescriptorOutputReference = GluePartitionStorageDescriptorOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GluePartitionStorageDescriptorOutputReference[_c] = { fqn: "@cdktf/provider-aws.glue.GluePartitionStorageDescriptorOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_partition.html aws_glue_partition}.
 *
 * @stability stable
 */
class GluePartition extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_partition.html aws_glue_partition} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_partition',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // storage_descriptor - computed: false, optional: true, required: false
        this._storageDescriptor = new GluePartitionStorageDescriptorOutputReference(this, "storage_descriptor", true);
        this._catalogId = config.catalogId;
        this._databaseName = config.databaseName;
        this._parameters = config.parameters;
        this._partitionValues = config.partitionValues;
        this._tableName = config.tableName;
        this._storageDescriptor.internalValue = config.storageDescriptor;
    }
    /**
     * @stability stable
     */
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    /**
     * @stability stable
     */
    set catalogId(value) {
        this._catalogId = value;
    }
    /**
     * @stability stable
     */
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogIdInput() {
        return this._catalogId;
    }
    // creation_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    /**
     * @stability stable
     */
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseNameInput() {
        return this._databaseName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // last_accessed_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastAccessedTime() {
        return this.getStringAttribute('last_accessed_time');
    }
    // last_analyzed_time - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get lastAnalyzedTime() {
        return this.getStringAttribute('last_analyzed_time');
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get partitionValues() {
        return this.getListAttribute('partition_values');
    }
    /**
     * @stability stable
     */
    set partitionValues(value) {
        this._partitionValues = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get partitionValuesInput() {
        return this._partitionValues;
    }
    /**
     * @stability stable
     */
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    /**
     * @stability stable
     */
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tableNameInput() {
        return this._tableName;
    }
    /**
     * @stability stable
     */
    get storageDescriptor() {
        return this._storageDescriptor;
    }
    /**
     * @stability stable
     */
    putStorageDescriptor(value) {
        this._storageDescriptor.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetStorageDescriptor() {
        this._storageDescriptor.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get storageDescriptorInput() {
        return this._storageDescriptor.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            database_name: cdktf.stringToTerraform(this._databaseName),
            parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._parameters),
            partition_values: cdktf.listMapper(cdktf.stringToTerraform)(this._partitionValues),
            table_name: cdktf.stringToTerraform(this._tableName),
            storage_descriptor: gluePartitionStorageDescriptorToTerraform(this._storageDescriptor.internalValue),
        };
    }
}
exports.GluePartition = GluePartition;
_d = JSII_RTTI_SYMBOL_1;
GluePartition[_d] = { fqn: "@cdktf/provider-aws.glue.GluePartition", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GluePartition.tfResourceType = "aws_glue_partition";
