// @cdktf/provider-aws/lib/glue/glue-crawler.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueCrawler = exports.GlueCrawlerSchemaChangePolicyOutputReference = exports.glueCrawlerSchemaChangePolicyToTerraform = exports.glueCrawlerS3TargetToTerraform = exports.GlueCrawlerRecrawlPolicyOutputReference = exports.glueCrawlerRecrawlPolicyToTerraform = exports.glueCrawlerMongodbTargetToTerraform = exports.GlueCrawlerLineageConfigurationOutputReference = exports.glueCrawlerLineageConfigurationToTerraform = exports.glueCrawlerJdbcTargetToTerraform = exports.glueCrawlerDynamodbTargetToTerraform = exports.glueCrawlerCatalogTargetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueCrawlerCatalogTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
        tables: cdktf.listMapper(cdktf.stringToTerraform)(struct.tables),
    };
}
exports.glueCrawlerCatalogTargetToTerraform = glueCrawlerCatalogTargetToTerraform;
function glueCrawlerDynamodbTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        scan_all: cdktf.booleanToTerraform(struct.scanAll),
        scan_rate: cdktf.numberToTerraform(struct.scanRate),
    };
}
exports.glueCrawlerDynamodbTargetToTerraform = glueCrawlerDynamodbTargetToTerraform;
function glueCrawlerJdbcTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        exclusions: cdktf.listMapper(cdktf.stringToTerraform)(struct.exclusions),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.glueCrawlerJdbcTargetToTerraform = glueCrawlerJdbcTargetToTerraform;
function glueCrawlerLineageConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crawler_lineage_settings: cdktf.stringToTerraform(struct.crawlerLineageSettings),
    };
}
exports.glueCrawlerLineageConfigurationToTerraform = glueCrawlerLineageConfigurationToTerraform;
/**
 * @stability stable
 */
class GlueCrawlerLineageConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._crawlerLineageSettings) {
            hasAnyValues = true;
            internalValueResult.crawlerLineageSettings = this._crawlerLineageSettings;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._crawlerLineageSettings = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._crawlerLineageSettings = value.crawlerLineageSettings;
        }
    }
    /**
     * @stability stable
     */
    get crawlerLineageSettings() {
        return this.getStringAttribute('crawler_lineage_settings');
    }
    /**
     * @stability stable
     */
    set crawlerLineageSettings(value) {
        this._crawlerLineageSettings = value;
    }
    /**
     * @stability stable
     */
    resetCrawlerLineageSettings() {
        this._crawlerLineageSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get crawlerLineageSettingsInput() {
        return this._crawlerLineageSettings;
    }
}
exports.GlueCrawlerLineageConfigurationOutputReference = GlueCrawlerLineageConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueCrawlerLineageConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.glue.GlueCrawlerLineageConfigurationOutputReference", version: "3.0.1" };
function glueCrawlerMongodbTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        path: cdktf.stringToTerraform(struct.path),
        scan_all: cdktf.booleanToTerraform(struct.scanAll),
    };
}
exports.glueCrawlerMongodbTargetToTerraform = glueCrawlerMongodbTargetToTerraform;
function glueCrawlerRecrawlPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recrawl_behavior: cdktf.stringToTerraform(struct.recrawlBehavior),
    };
}
exports.glueCrawlerRecrawlPolicyToTerraform = glueCrawlerRecrawlPolicyToTerraform;
/**
 * @stability stable
 */
class GlueCrawlerRecrawlPolicyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recrawlBehavior) {
            hasAnyValues = true;
            internalValueResult.recrawlBehavior = this._recrawlBehavior;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recrawlBehavior = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recrawlBehavior = value.recrawlBehavior;
        }
    }
    /**
     * @stability stable
     */
    get recrawlBehavior() {
        return this.getStringAttribute('recrawl_behavior');
    }
    /**
     * @stability stable
     */
    set recrawlBehavior(value) {
        this._recrawlBehavior = value;
    }
    /**
     * @stability stable
     */
    resetRecrawlBehavior() {
        this._recrawlBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recrawlBehaviorInput() {
        return this._recrawlBehavior;
    }
}
exports.GlueCrawlerRecrawlPolicyOutputReference = GlueCrawlerRecrawlPolicyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GlueCrawlerRecrawlPolicyOutputReference[_b] = { fqn: "@cdktf/provider-aws.glue.GlueCrawlerRecrawlPolicyOutputReference", version: "3.0.1" };
function glueCrawlerS3TargetToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        dlq_event_queue_arn: cdktf.stringToTerraform(struct.dlqEventQueueArn),
        event_queue_arn: cdktf.stringToTerraform(struct.eventQueueArn),
        exclusions: cdktf.listMapper(cdktf.stringToTerraform)(struct.exclusions),
        path: cdktf.stringToTerraform(struct.path),
        sample_size: cdktf.numberToTerraform(struct.sampleSize),
    };
}
exports.glueCrawlerS3TargetToTerraform = glueCrawlerS3TargetToTerraform;
function glueCrawlerSchemaChangePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_behavior: cdktf.stringToTerraform(struct.deleteBehavior),
        update_behavior: cdktf.stringToTerraform(struct.updateBehavior),
    };
}
exports.glueCrawlerSchemaChangePolicyToTerraform = glueCrawlerSchemaChangePolicyToTerraform;
/**
 * @stability stable
 */
class GlueCrawlerSchemaChangePolicyOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteBehavior) {
            hasAnyValues = true;
            internalValueResult.deleteBehavior = this._deleteBehavior;
        }
        if (this._updateBehavior) {
            hasAnyValues = true;
            internalValueResult.updateBehavior = this._updateBehavior;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteBehavior = undefined;
            this._updateBehavior = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteBehavior = value.deleteBehavior;
            this._updateBehavior = value.updateBehavior;
        }
    }
    /**
     * @stability stable
     */
    get deleteBehavior() {
        return this.getStringAttribute('delete_behavior');
    }
    /**
     * @stability stable
     */
    set deleteBehavior(value) {
        this._deleteBehavior = value;
    }
    /**
     * @stability stable
     */
    resetDeleteBehavior() {
        this._deleteBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteBehaviorInput() {
        return this._deleteBehavior;
    }
    /**
     * @stability stable
     */
    get updateBehavior() {
        return this.getStringAttribute('update_behavior');
    }
    /**
     * @stability stable
     */
    set updateBehavior(value) {
        this._updateBehavior = value;
    }
    /**
     * @stability stable
     */
    resetUpdateBehavior() {
        this._updateBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get updateBehaviorInput() {
        return this._updateBehavior;
    }
}
exports.GlueCrawlerSchemaChangePolicyOutputReference = GlueCrawlerSchemaChangePolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlueCrawlerSchemaChangePolicyOutputReference[_c] = { fqn: "@cdktf/provider-aws.glue.GlueCrawlerSchemaChangePolicyOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_crawler.html aws_glue_crawler}.
 *
 * @stability stable
 */
class GlueCrawler extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_crawler.html aws_glue_crawler} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_crawler',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // lineage_configuration - computed: false, optional: true, required: false
        this._lineageConfiguration = new GlueCrawlerLineageConfigurationOutputReference(this, "lineage_configuration", true);
        // recrawl_policy - computed: false, optional: true, required: false
        this._recrawlPolicy = new GlueCrawlerRecrawlPolicyOutputReference(this, "recrawl_policy", true);
        // schema_change_policy - computed: false, optional: true, required: false
        this._schemaChangePolicy = new GlueCrawlerSchemaChangePolicyOutputReference(this, "schema_change_policy", true);
        this._classifiers = config.classifiers;
        this._configuration = config.configuration;
        this._databaseName = config.databaseName;
        this._description = config.description;
        this._name = config.name;
        this._role = config.role;
        this._schedule = config.schedule;
        this._securityConfiguration = config.securityConfiguration;
        this._tablePrefix = config.tablePrefix;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._catalogTarget = config.catalogTarget;
        this._dynamodbTarget = config.dynamodbTarget;
        this._jdbcTarget = config.jdbcTarget;
        this._lineageConfiguration.internalValue = config.lineageConfiguration;
        this._mongodbTarget = config.mongodbTarget;
        this._recrawlPolicy.internalValue = config.recrawlPolicy;
        this._s3Target = config.s3Target;
        this._schemaChangePolicy.internalValue = config.schemaChangePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get classifiers() {
        return this.getListAttribute('classifiers');
    }
    /**
     * @stability stable
     */
    set classifiers(value) {
        this._classifiers = value;
    }
    /**
     * @stability stable
     */
    resetClassifiers() {
        this._classifiers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get classifiersInput() {
        return this._classifiers;
    }
    /**
     * @stability stable
     */
    get configuration() {
        return this.getStringAttribute('configuration');
    }
    /**
     * @stability stable
     */
    set configuration(value) {
        this._configuration = value;
    }
    /**
     * @stability stable
     */
    resetConfiguration() {
        this._configuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get configurationInput() {
        return this._configuration;
    }
    /**
     * @stability stable
     */
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    /**
     * @stability stable
     */
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get databaseNameInput() {
        return this._databaseName;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get role() {
        return this.getStringAttribute('role');
    }
    /**
     * @stability stable
     */
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get roleInput() {
        return this._role;
    }
    /**
     * @stability stable
     */
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    /**
     * @stability stable
     */
    set schedule(value) {
        this._schedule = value;
    }
    /**
     * @stability stable
     */
    resetSchedule() {
        this._schedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleInput() {
        return this._schedule;
    }
    /**
     * @stability stable
     */
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    /**
     * @stability stable
     */
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    /**
     * @stability stable
     */
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    /**
     * @stability stable
     */
    get tablePrefix() {
        return this.getStringAttribute('table_prefix');
    }
    /**
     * @stability stable
     */
    set tablePrefix(value) {
        this._tablePrefix = value;
    }
    /**
     * @stability stable
     */
    resetTablePrefix() {
        this._tablePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tablePrefixInput() {
        return this._tablePrefix;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get catalogTarget() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('catalog_target');
    }
    /**
     * @stability stable
     */
    set catalogTarget(value) {
        this._catalogTarget = value;
    }
    /**
     * @stability stable
     */
    resetCatalogTarget() {
        this._catalogTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get catalogTargetInput() {
        return this._catalogTarget;
    }
    /**
     * @stability stable
     */
    get dynamodbTarget() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('dynamodb_target');
    }
    /**
     * @stability stable
     */
    set dynamodbTarget(value) {
        this._dynamodbTarget = value;
    }
    /**
     * @stability stable
     */
    resetDynamodbTarget() {
        this._dynamodbTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get dynamodbTargetInput() {
        return this._dynamodbTarget;
    }
    /**
     * @stability stable
     */
    get jdbcTarget() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('jdbc_target');
    }
    /**
     * @stability stable
     */
    set jdbcTarget(value) {
        this._jdbcTarget = value;
    }
    /**
     * @stability stable
     */
    resetJdbcTarget() {
        this._jdbcTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get jdbcTargetInput() {
        return this._jdbcTarget;
    }
    /**
     * @stability stable
     */
    get lineageConfiguration() {
        return this._lineageConfiguration;
    }
    /**
     * @stability stable
     */
    putLineageConfiguration(value) {
        this._lineageConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLineageConfiguration() {
        this._lineageConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lineageConfigurationInput() {
        return this._lineageConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get mongodbTarget() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('mongodb_target');
    }
    /**
     * @stability stable
     */
    set mongodbTarget(value) {
        this._mongodbTarget = value;
    }
    /**
     * @stability stable
     */
    resetMongodbTarget() {
        this._mongodbTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get mongodbTargetInput() {
        return this._mongodbTarget;
    }
    /**
     * @stability stable
     */
    get recrawlPolicy() {
        return this._recrawlPolicy;
    }
    /**
     * @stability stable
     */
    putRecrawlPolicy(value) {
        this._recrawlPolicy.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRecrawlPolicy() {
        this._recrawlPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get recrawlPolicyInput() {
        return this._recrawlPolicy.internalValue;
    }
    /**
     * @stability stable
     */
    get s3Target() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('s3_target');
    }
    /**
     * @stability stable
     */
    set s3Target(value) {
        this._s3Target = value;
    }
    /**
     * @stability stable
     */
    resetS3Target() {
        this._s3Target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3TargetInput() {
        return this._s3Target;
    }
    /**
     * @stability stable
     */
    get schemaChangePolicy() {
        return this._schemaChangePolicy;
    }
    /**
     * @stability stable
     */
    putSchemaChangePolicy(value) {
        this._schemaChangePolicy.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetSchemaChangePolicy() {
        this._schemaChangePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schemaChangePolicyInput() {
        return this._schemaChangePolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            classifiers: cdktf.listMapper(cdktf.stringToTerraform)(this._classifiers),
            configuration: cdktf.stringToTerraform(this._configuration),
            database_name: cdktf.stringToTerraform(this._databaseName),
            description: cdktf.stringToTerraform(this._description),
            name: cdktf.stringToTerraform(this._name),
            role: cdktf.stringToTerraform(this._role),
            schedule: cdktf.stringToTerraform(this._schedule),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            table_prefix: cdktf.stringToTerraform(this._tablePrefix),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            catalog_target: cdktf.listMapper(glueCrawlerCatalogTargetToTerraform)(this._catalogTarget),
            dynamodb_target: cdktf.listMapper(glueCrawlerDynamodbTargetToTerraform)(this._dynamodbTarget),
            jdbc_target: cdktf.listMapper(glueCrawlerJdbcTargetToTerraform)(this._jdbcTarget),
            lineage_configuration: glueCrawlerLineageConfigurationToTerraform(this._lineageConfiguration.internalValue),
            mongodb_target: cdktf.listMapper(glueCrawlerMongodbTargetToTerraform)(this._mongodbTarget),
            recrawl_policy: glueCrawlerRecrawlPolicyToTerraform(this._recrawlPolicy.internalValue),
            s3_target: cdktf.listMapper(glueCrawlerS3TargetToTerraform)(this._s3Target),
            schema_change_policy: glueCrawlerSchemaChangePolicyToTerraform(this._schemaChangePolicy.internalValue),
        };
    }
}
exports.GlueCrawler = GlueCrawler;
_d = JSII_RTTI_SYMBOL_1;
GlueCrawler[_d] = { fqn: "@cdktf/provider-aws.glue.GlueCrawler", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlueCrawler.tfResourceType = "aws_glue_crawler";
