// @cdktf/provider-aws/lib/apigateway/api-gateway-base-path-mapping.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayBasePathMapping = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_base_path_mapping.html aws_api_gateway_base_path_mapping}.
 *
 * @stability stable
 */
class ApiGatewayBasePathMapping extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_base_path_mapping.html aws_api_gateway_base_path_mapping} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_base_path_mapping',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._apiId = config.apiId;
        this._basePath = config.basePath;
        this._domainName = config.domainName;
        this._stageName = config.stageName;
    }
    /**
     * @stability stable
     */
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    /**
     * @stability stable
     */
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get apiIdInput() {
        return this._apiId;
    }
    /**
     * @stability stable
     */
    get basePath() {
        return this.getStringAttribute('base_path');
    }
    /**
     * @stability stable
     */
    set basePath(value) {
        this._basePath = value;
    }
    /**
     * @stability stable
     */
    resetBasePath() {
        this._basePath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get basePathInput() {
        return this._basePath;
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get stageName() {
        return this.getStringAttribute('stage_name');
    }
    /**
     * @stability stable
     */
    set stageName(value) {
        this._stageName = value;
    }
    /**
     * @stability stable
     */
    resetStageName() {
        this._stageName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stageNameInput() {
        return this._stageName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            base_path: cdktf.stringToTerraform(this._basePath),
            domain_name: cdktf.stringToTerraform(this._domainName),
            stage_name: cdktf.stringToTerraform(this._stageName),
        };
    }
}
exports.ApiGatewayBasePathMapping = ApiGatewayBasePathMapping;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayBasePathMapping[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayBasePathMapping", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayBasePathMapping.tfResourceType = "aws_api_gateway_base_path_mapping";
