// @cdktf/provider-aws/lib/apigateway/api-gateway-account.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayAccount = exports.ApiGatewayAccountThrottleSettings = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class ApiGatewayAccountThrottleSettings extends cdktf.ComplexComputedList {
    // burst_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get burstLimit() {
        return this.getNumberAttribute('burst_limit');
    }
    // rate_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get rateLimit() {
        return this.getNumberAttribute('rate_limit');
    }
}
exports.ApiGatewayAccountThrottleSettings = ApiGatewayAccountThrottleSettings;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayAccountThrottleSettings[_a] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayAccountThrottleSettings", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_account.html aws_api_gateway_account}.
 *
 * @stability stable
 */
class ApiGatewayAccount extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_account.html aws_api_gateway_account} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_account',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._cloudwatchRoleArn = config.cloudwatchRoleArn;
    }
    /**
     * @stability stable
     */
    get cloudwatchRoleArn() {
        return this.getStringAttribute('cloudwatch_role_arn');
    }
    /**
     * @stability stable
     */
    set cloudwatchRoleArn(value) {
        this._cloudwatchRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchRoleArn() {
        this._cloudwatchRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchRoleArnInput() {
        return this._cloudwatchRoleArn;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // throttle_settings - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    throttleSettings(index) {
        return new ApiGatewayAccountThrottleSettings(this, 'throttle_settings', index);
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cloudwatch_role_arn: cdktf.stringToTerraform(this._cloudwatchRoleArn),
        };
    }
}
exports.ApiGatewayAccount = ApiGatewayAccount;
_b = JSII_RTTI_SYMBOL_1;
ApiGatewayAccount[_b] = { fqn: "@cdktf/provider-aws.apigateway.ApiGatewayAccount", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
ApiGatewayAccount.tfResourceType = "aws_api_gateway_account";
