// @cdktf/provider-aws/lib/apigateway/data-aws-api-gateway-domain-name.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsApiGatewayDomainName = exports.DataAwsApiGatewayDomainNameEndpointConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsApiGatewayDomainNameEndpointConfiguration extends cdktf.ComplexComputedList {
    // types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get types() {
        return this.getListAttribute('types');
    }
}
exports.DataAwsApiGatewayDomainNameEndpointConfiguration = DataAwsApiGatewayDomainNameEndpointConfiguration;
_a = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayDomainNameEndpointConfiguration[_a] = { fqn: "@cdktf/provider-aws.apigateway.DataAwsApiGatewayDomainNameEndpointConfiguration", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_domain_name.html aws_api_gateway_domain_name}.
 *
 * @stability stable
 */
class DataAwsApiGatewayDomainName extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_domain_name.html aws_api_gateway_domain_name} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_domain_name',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._domainName = config.domainName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    // certificate_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateName() {
        return this.getStringAttribute('certificate_name');
    }
    // certificate_upload_date - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get certificateUploadDate() {
        return this.getStringAttribute('certificate_upload_date');
    }
    // cloudfront_domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudfrontDomainName() {
        return this.getStringAttribute('cloudfront_domain_name');
    }
    // cloudfront_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cloudfrontZoneId() {
        return this.getStringAttribute('cloudfront_zone_id');
    }
    /**
     * @stability stable
     */
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    /**
     * @stability stable
     */
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainNameInput() {
        return this._domainName;
    }
    // endpoint_configuration - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    endpointConfiguration(index) {
        return new DataAwsApiGatewayDomainNameEndpointConfiguration(this, 'endpoint_configuration', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // regional_certificate_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalCertificateArn() {
        return this.getStringAttribute('regional_certificate_arn');
    }
    // regional_certificate_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalCertificateName() {
        return this.getStringAttribute('regional_certificate_name');
    }
    // regional_domain_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalDomainName() {
        return this.getStringAttribute('regional_domain_name');
    }
    // regional_zone_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get regionalZoneId() {
        return this.getStringAttribute('regional_zone_id');
    }
    // security_policy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityPolicy() {
        return this.getStringAttribute('security_policy');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
        };
    }
}
exports.DataAwsApiGatewayDomainName = DataAwsApiGatewayDomainName;
_b = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayDomainName[_b] = { fqn: "@cdktf/provider-aws.apigateway.DataAwsApiGatewayDomainName", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsApiGatewayDomainName.tfResourceType = "aws_api_gateway_domain_name";
