// @cdktf/provider-aws/lib/sagemaker/sagemaker-studio-lifecycle-config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerStudioLifecycleConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_studio_lifecycle_config.html aws_sagemaker_studio_lifecycle_config}.
 *
 * @stability stable
 */
class SagemakerStudioLifecycleConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_studio_lifecycle_config.html aws_sagemaker_studio_lifecycle_config} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_studio_lifecycle_config',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._studioLifecycleConfigAppType = config.studioLifecycleConfigAppType;
        this._studioLifecycleConfigContent = config.studioLifecycleConfigContent;
        this._studioLifecycleConfigName = config.studioLifecycleConfigName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get studioLifecycleConfigAppType() {
        return this.getStringAttribute('studio_lifecycle_config_app_type');
    }
    /**
     * @stability stable
     */
    set studioLifecycleConfigAppType(value) {
        this._studioLifecycleConfigAppType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get studioLifecycleConfigAppTypeInput() {
        return this._studioLifecycleConfigAppType;
    }
    /**
     * @stability stable
     */
    get studioLifecycleConfigContent() {
        return this.getStringAttribute('studio_lifecycle_config_content');
    }
    /**
     * @stability stable
     */
    set studioLifecycleConfigContent(value) {
        this._studioLifecycleConfigContent = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get studioLifecycleConfigContentInput() {
        return this._studioLifecycleConfigContent;
    }
    /**
     * @stability stable
     */
    get studioLifecycleConfigName() {
        return this.getStringAttribute('studio_lifecycle_config_name');
    }
    /**
     * @stability stable
     */
    set studioLifecycleConfigName(value) {
        this._studioLifecycleConfigName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get studioLifecycleConfigNameInput() {
        return this._studioLifecycleConfigName;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            studio_lifecycle_config_app_type: cdktf.stringToTerraform(this._studioLifecycleConfigAppType),
            studio_lifecycle_config_content: cdktf.stringToTerraform(this._studioLifecycleConfigContent),
            studio_lifecycle_config_name: cdktf.stringToTerraform(this._studioLifecycleConfigName),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.SagemakerStudioLifecycleConfig = SagemakerStudioLifecycleConfig;
_a = JSII_RTTI_SYMBOL_1;
SagemakerStudioLifecycleConfig[_a] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerStudioLifecycleConfig", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SagemakerStudioLifecycleConfig.tfResourceType = "aws_sagemaker_studio_lifecycle_config";
