// @cdktf/provider-aws/lib/sagemaker/sagemaker-model-package-group-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerModelPackageGroupPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_model_package_group_policy.html aws_sagemaker_model_package_group_policy}.
 *
 * @stability stable
 */
class SagemakerModelPackageGroupPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_model_package_group_policy.html aws_sagemaker_model_package_group_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_model_package_group_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._modelPackageGroupName = config.modelPackageGroupName;
        this._resourcePolicy = config.resourcePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get modelPackageGroupName() {
        return this.getStringAttribute('model_package_group_name');
    }
    /**
     * @stability stable
     */
    set modelPackageGroupName(value) {
        this._modelPackageGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get modelPackageGroupNameInput() {
        return this._modelPackageGroupName;
    }
    /**
     * @stability stable
     */
    get resourcePolicy() {
        return this.getStringAttribute('resource_policy');
    }
    /**
     * @stability stable
     */
    set resourcePolicy(value) {
        this._resourcePolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourcePolicyInput() {
        return this._resourcePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            model_package_group_name: cdktf.stringToTerraform(this._modelPackageGroupName),
            resource_policy: cdktf.stringToTerraform(this._resourcePolicy),
        };
    }
}
exports.SagemakerModelPackageGroupPolicy = SagemakerModelPackageGroupPolicy;
_a = JSII_RTTI_SYMBOL_1;
SagemakerModelPackageGroupPolicy[_a] = { fqn: "@cdktf/provider-aws.sagemaker.SagemakerModelPackageGroupPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SagemakerModelPackageGroupPolicy.tfResourceType = "aws_sagemaker_model_package_group_policy";
