// @cdktf/provider-aws/lib/fms/fms-policy.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FmsPolicy = exports.FmsPolicySecurityServicePolicyDataOutputReference = exports.fmsPolicySecurityServicePolicyDataToTerraform = exports.FmsPolicyIncludeMapOutputReference = exports.fmsPolicyIncludeMapToTerraform = exports.FmsPolicyExcludeMapOutputReference = exports.fmsPolicyExcludeMapToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fmsPolicyExcludeMapToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account: cdktf.listMapper(cdktf.stringToTerraform)(struct.account),
        orgunit: cdktf.listMapper(cdktf.stringToTerraform)(struct.orgunit),
    };
}
exports.fmsPolicyExcludeMapToTerraform = fmsPolicyExcludeMapToTerraform;
/**
 * @stability stable
 */
class FmsPolicyExcludeMapOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._account) {
            hasAnyValues = true;
            internalValueResult.account = this._account;
        }
        if (this._orgunit) {
            hasAnyValues = true;
            internalValueResult.orgunit = this._orgunit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._account = undefined;
            this._orgunit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._account = value.account;
            this._orgunit = value.orgunit;
        }
    }
    /**
     * @stability stable
     */
    get account() {
        return this.getListAttribute('account');
    }
    /**
     * @stability stable
     */
    set account(value) {
        this._account = value;
    }
    /**
     * @stability stable
     */
    resetAccount() {
        this._account = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountInput() {
        return this._account;
    }
    /**
     * @stability stable
     */
    get orgunit() {
        return this.getListAttribute('orgunit');
    }
    /**
     * @stability stable
     */
    set orgunit(value) {
        this._orgunit = value;
    }
    /**
     * @stability stable
     */
    resetOrgunit() {
        this._orgunit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get orgunitInput() {
        return this._orgunit;
    }
}
exports.FmsPolicyExcludeMapOutputReference = FmsPolicyExcludeMapOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FmsPolicyExcludeMapOutputReference[_a] = { fqn: "@cdktf/provider-aws.fms.FmsPolicyExcludeMapOutputReference", version: "3.0.1" };
function fmsPolicyIncludeMapToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account: cdktf.listMapper(cdktf.stringToTerraform)(struct.account),
        orgunit: cdktf.listMapper(cdktf.stringToTerraform)(struct.orgunit),
    };
}
exports.fmsPolicyIncludeMapToTerraform = fmsPolicyIncludeMapToTerraform;
/**
 * @stability stable
 */
class FmsPolicyIncludeMapOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._account) {
            hasAnyValues = true;
            internalValueResult.account = this._account;
        }
        if (this._orgunit) {
            hasAnyValues = true;
            internalValueResult.orgunit = this._orgunit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._account = undefined;
            this._orgunit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._account = value.account;
            this._orgunit = value.orgunit;
        }
    }
    /**
     * @stability stable
     */
    get account() {
        return this.getListAttribute('account');
    }
    /**
     * @stability stable
     */
    set account(value) {
        this._account = value;
    }
    /**
     * @stability stable
     */
    resetAccount() {
        this._account = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get accountInput() {
        return this._account;
    }
    /**
     * @stability stable
     */
    get orgunit() {
        return this.getListAttribute('orgunit');
    }
    /**
     * @stability stable
     */
    set orgunit(value) {
        this._orgunit = value;
    }
    /**
     * @stability stable
     */
    resetOrgunit() {
        this._orgunit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get orgunitInput() {
        return this._orgunit;
    }
}
exports.FmsPolicyIncludeMapOutputReference = FmsPolicyIncludeMapOutputReference;
_b = JSII_RTTI_SYMBOL_1;
FmsPolicyIncludeMapOutputReference[_b] = { fqn: "@cdktf/provider-aws.fms.FmsPolicyIncludeMapOutputReference", version: "3.0.1" };
function fmsPolicySecurityServicePolicyDataToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        managed_service_data: cdktf.stringToTerraform(struct.managedServiceData),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.fmsPolicySecurityServicePolicyDataToTerraform = fmsPolicySecurityServicePolicyDataToTerraform;
/**
 * @stability stable
 */
class FmsPolicySecurityServicePolicyDataOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._managedServiceData) {
            hasAnyValues = true;
            internalValueResult.managedServiceData = this._managedServiceData;
        }
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._managedServiceData = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._managedServiceData = value.managedServiceData;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get managedServiceData() {
        return this.getStringAttribute('managed_service_data');
    }
    /**
     * @stability stable
     */
    set managedServiceData(value) {
        this._managedServiceData = value;
    }
    /**
     * @stability stable
     */
    resetManagedServiceData() {
        this._managedServiceData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get managedServiceDataInput() {
        return this._managedServiceData;
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.FmsPolicySecurityServicePolicyDataOutputReference = FmsPolicySecurityServicePolicyDataOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FmsPolicySecurityServicePolicyDataOutputReference[_c] = { fqn: "@cdktf/provider-aws.fms.FmsPolicySecurityServicePolicyDataOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/fms_policy.html aws_fms_policy}.
 *
 * @stability stable
 */
class FmsPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/fms_policy.html aws_fms_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fms_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // exclude_map - computed: false, optional: true, required: false
        this._excludeMap = new FmsPolicyExcludeMapOutputReference(this, "exclude_map", true);
        // include_map - computed: false, optional: true, required: false
        this._includeMap = new FmsPolicyIncludeMapOutputReference(this, "include_map", true);
        // security_service_policy_data - computed: false, optional: false, required: true
        this._securityServicePolicyData = new FmsPolicySecurityServicePolicyDataOutputReference(this, "security_service_policy_data", true);
        this._deleteAllPolicyResources = config.deleteAllPolicyResources;
        this._excludeResourceTags = config.excludeResourceTags;
        this._name = config.name;
        this._remediationEnabled = config.remediationEnabled;
        this._resourceTags = config.resourceTags;
        this._resourceType = config.resourceType;
        this._resourceTypeList = config.resourceTypeList;
        this._excludeMap.internalValue = config.excludeMap;
        this._includeMap.internalValue = config.includeMap;
        this._securityServicePolicyData.internalValue = config.securityServicePolicyData;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get deleteAllPolicyResources() {
        return this.getBooleanAttribute('delete_all_policy_resources');
    }
    /**
     * @stability stable
     */
    set deleteAllPolicyResources(value) {
        this._deleteAllPolicyResources = value;
    }
    /**
     * @stability stable
     */
    resetDeleteAllPolicyResources() {
        this._deleteAllPolicyResources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteAllPolicyResourcesInput() {
        return this._deleteAllPolicyResources;
    }
    /**
     * @stability stable
     */
    get excludeResourceTags() {
        return this.getBooleanAttribute('exclude_resource_tags');
    }
    /**
     * @stability stable
     */
    set excludeResourceTags(value) {
        this._excludeResourceTags = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludeResourceTagsInput() {
        return this._excludeResourceTags;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // policy_update_token - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get policyUpdateToken() {
        return this.getStringAttribute('policy_update_token');
    }
    /**
     * @stability stable
     */
    get remediationEnabled() {
        return this.getBooleanAttribute('remediation_enabled');
    }
    /**
     * @stability stable
     */
    set remediationEnabled(value) {
        this._remediationEnabled = value;
    }
    /**
     * @stability stable
     */
    resetRemediationEnabled() {
        this._remediationEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get remediationEnabledInput() {
        return this._remediationEnabled;
    }
    /**
     * @stability stable
     */
    get resourceTags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('resource_tags');
    }
    /**
     * @stability stable
     */
    set resourceTags(value) {
        this._resourceTags = value;
    }
    /**
     * @stability stable
     */
    resetResourceTags() {
        this._resourceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTagsInput() {
        return this._resourceTags;
    }
    /**
     * @stability stable
     */
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    /**
     * @stability stable
     */
    set resourceType(value) {
        this._resourceType = value;
    }
    /**
     * @stability stable
     */
    resetResourceType() {
        this._resourceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTypeInput() {
        return this._resourceType;
    }
    /**
     * @stability stable
     */
    get resourceTypeList() {
        return this.getListAttribute('resource_type_list');
    }
    /**
     * @stability stable
     */
    set resourceTypeList(value) {
        this._resourceTypeList = value;
    }
    /**
     * @stability stable
     */
    resetResourceTypeList() {
        this._resourceTypeList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get resourceTypeListInput() {
        return this._resourceTypeList;
    }
    /**
     * @stability stable
     */
    get excludeMap() {
        return this._excludeMap;
    }
    /**
     * @stability stable
     */
    putExcludeMap(value) {
        this._excludeMap.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetExcludeMap() {
        this._excludeMap.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get excludeMapInput() {
        return this._excludeMap.internalValue;
    }
    /**
     * @stability stable
     */
    get includeMap() {
        return this._includeMap;
    }
    /**
     * @stability stable
     */
    putIncludeMap(value) {
        this._includeMap.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetIncludeMap() {
        this._includeMap.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get includeMapInput() {
        return this._includeMap.internalValue;
    }
    /**
     * @stability stable
     */
    get securityServicePolicyData() {
        return this._securityServicePolicyData;
    }
    /**
     * @stability stable
     */
    putSecurityServicePolicyData(value) {
        this._securityServicePolicyData.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityServicePolicyDataInput() {
        return this._securityServicePolicyData.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            delete_all_policy_resources: cdktf.booleanToTerraform(this._deleteAllPolicyResources),
            exclude_resource_tags: cdktf.booleanToTerraform(this._excludeResourceTags),
            name: cdktf.stringToTerraform(this._name),
            remediation_enabled: cdktf.booleanToTerraform(this._remediationEnabled),
            resource_tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._resourceTags),
            resource_type: cdktf.stringToTerraform(this._resourceType),
            resource_type_list: cdktf.listMapper(cdktf.stringToTerraform)(this._resourceTypeList),
            exclude_map: fmsPolicyExcludeMapToTerraform(this._excludeMap.internalValue),
            include_map: fmsPolicyIncludeMapToTerraform(this._includeMap.internalValue),
            security_service_policy_data: fmsPolicySecurityServicePolicyDataToTerraform(this._securityServicePolicyData.internalValue),
        };
    }
}
exports.FmsPolicy = FmsPolicy;
_d = JSII_RTTI_SYMBOL_1;
FmsPolicy[_d] = { fqn: "@cdktf/provider-aws.fms.FmsPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
FmsPolicy.tfResourceType = "aws_fms_policy";
