// @cdktf/provider-aws/lib/lightsail/lightsail-instance-public-ports.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailInstancePublicPorts = exports.lightsailInstancePublicPortsPortInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lightsailInstancePublicPortsPortInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cidrs: cdktf.listMapper(cdktf.stringToTerraform)(struct.cidrs),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        protocol: cdktf.stringToTerraform(struct.protocol),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.lightsailInstancePublicPortsPortInfoToTerraform = lightsailInstancePublicPortsPortInfoToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/lightsail_instance_public_ports.html aws_lightsail_instance_public_ports}.
 *
 * @stability stable
 */
class LightsailInstancePublicPorts extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lightsail_instance_public_ports.html aws_lightsail_instance_public_ports} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_instance_public_ports',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._instanceName = config.instanceName;
        this._portInfo = config.portInfo;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceName() {
        return this.getStringAttribute('instance_name');
    }
    /**
     * @stability stable
     */
    set instanceName(value) {
        this._instanceName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceNameInput() {
        return this._instanceName;
    }
    /**
     * @stability stable
     */
    get portInfo() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('port_info');
    }
    /**
     * @stability stable
     */
    set portInfo(value) {
        this._portInfo = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInfoInput() {
        return this._portInfo;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            instance_name: cdktf.stringToTerraform(this._instanceName),
            port_info: cdktf.listMapper(lightsailInstancePublicPortsPortInfoToTerraform)(this._portInfo),
        };
    }
}
exports.LightsailInstancePublicPorts = LightsailInstancePublicPorts;
_a = JSII_RTTI_SYMBOL_1;
LightsailInstancePublicPorts[_a] = { fqn: "@cdktf/provider-aws.lightsail.LightsailInstancePublicPorts", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LightsailInstancePublicPorts.tfResourceType = "aws_lightsail_instance_public_ports";
