// @cdktf/provider-aws/lib/glacier/glacier-vault-lock.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlacierVaultLock = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/glacier_vault_lock.html aws_glacier_vault_lock}.
 *
 * @stability stable
 */
class GlacierVaultLock extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glacier_vault_lock.html aws_glacier_vault_lock} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glacier_vault_lock',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._completeLock = config.completeLock;
        this._ignoreDeletionError = config.ignoreDeletionError;
        this._policy = config.policy;
        this._vaultName = config.vaultName;
    }
    /**
     * @stability stable
     */
    get completeLock() {
        return this.getBooleanAttribute('complete_lock');
    }
    /**
     * @stability stable
     */
    set completeLock(value) {
        this._completeLock = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get completeLockInput() {
        return this._completeLock;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ignoreDeletionError() {
        return this.getBooleanAttribute('ignore_deletion_error');
    }
    /**
     * @stability stable
     */
    set ignoreDeletionError(value) {
        this._ignoreDeletionError = value;
    }
    /**
     * @stability stable
     */
    resetIgnoreDeletionError() {
        this._ignoreDeletionError = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ignoreDeletionErrorInput() {
        return this._ignoreDeletionError;
    }
    /**
     * @stability stable
     */
    get policy() {
        return this.getStringAttribute('policy');
    }
    /**
     * @stability stable
     */
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyInput() {
        return this._policy;
    }
    /**
     * @stability stable
     */
    get vaultName() {
        return this.getStringAttribute('vault_name');
    }
    /**
     * @stability stable
     */
    set vaultName(value) {
        this._vaultName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vaultNameInput() {
        return this._vaultName;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            complete_lock: cdktf.booleanToTerraform(this._completeLock),
            ignore_deletion_error: cdktf.booleanToTerraform(this._ignoreDeletionError),
            policy: cdktf.stringToTerraform(this._policy),
            vault_name: cdktf.stringToTerraform(this._vaultName),
        };
    }
}
exports.GlacierVaultLock = GlacierVaultLock;
_a = JSII_RTTI_SYMBOL_1;
GlacierVaultLock[_a] = { fqn: "@cdktf/provider-aws.glacier.GlacierVaultLock", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
GlacierVaultLock.tfResourceType = "aws_glacier_vault_lock";
