// @cdktf/provider-aws/lib/codeartifact/codeartifact-repository.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeartifactRepository = exports.codeartifactRepositoryUpstreamToTerraform = exports.CodeartifactRepositoryExternalConnectionsOutputReference = exports.codeartifactRepositoryExternalConnectionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codeartifactRepositoryExternalConnectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        external_connection_name: cdktf.stringToTerraform(struct.externalConnectionName),
    };
}
exports.codeartifactRepositoryExternalConnectionsToTerraform = codeartifactRepositoryExternalConnectionsToTerraform;
/**
 * @stability stable
 */
class CodeartifactRepositoryExternalConnectionsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._externalConnectionName) {
            hasAnyValues = true;
            internalValueResult.externalConnectionName = this._externalConnectionName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._externalConnectionName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._externalConnectionName = value.externalConnectionName;
        }
    }
    /**
     * @stability stable
     */
    get externalConnectionName() {
        return this.getStringAttribute('external_connection_name');
    }
    /**
     * @stability stable
     */
    set externalConnectionName(value) {
        this._externalConnectionName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get externalConnectionNameInput() {
        return this._externalConnectionName;
    }
}
exports.CodeartifactRepositoryExternalConnectionsOutputReference = CodeartifactRepositoryExternalConnectionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodeartifactRepositoryExternalConnectionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.codeartifact.CodeartifactRepositoryExternalConnectionsOutputReference", version: "3.0.1" };
function codeartifactRepositoryUpstreamToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_name: cdktf.stringToTerraform(struct.repositoryName),
    };
}
exports.codeartifactRepositoryUpstreamToTerraform = codeartifactRepositoryUpstreamToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/codeartifact_repository.html aws_codeartifact_repository}.
 *
 * @stability stable
 */
class CodeartifactRepository extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codeartifact_repository.html aws_codeartifact_repository} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codeartifact_repository',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // external_connections - computed: false, optional: true, required: false
        this._externalConnections = new CodeartifactRepositoryExternalConnectionsOutputReference(this, "external_connections", true);
        this._description = config.description;
        this._domain = config.domain;
        this._domainOwner = config.domainOwner;
        this._repository = config.repository;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._externalConnections.internalValue = config.externalConnections;
        this._upstream = config.upstream;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // administrator_account - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get administratorAccount() {
        return this.getStringAttribute('administrator_account');
    }
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get domain() {
        return this.getStringAttribute('domain');
    }
    /**
     * @stability stable
     */
    set domain(value) {
        this._domain = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainInput() {
        return this._domain;
    }
    /**
     * @stability stable
     */
    get domainOwner() {
        return this.getStringAttribute('domain_owner');
    }
    /**
     * @stability stable
     */
    set domainOwner(value) {
        this._domainOwner = value;
    }
    /**
     * @stability stable
     */
    resetDomainOwner() {
        this._domainOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get domainOwnerInput() {
        return this._domainOwner;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get repository() {
        return this.getStringAttribute('repository');
    }
    /**
     * @stability stable
     */
    set repository(value) {
        this._repository = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get repositoryInput() {
        return this._repository;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get externalConnections() {
        return this._externalConnections;
    }
    /**
     * @stability stable
     */
    putExternalConnections(value) {
        this._externalConnections.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetExternalConnections() {
        this._externalConnections.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get externalConnectionsInput() {
        return this._externalConnections.internalValue;
    }
    /**
     * @stability stable
     */
    get upstream() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('upstream');
    }
    /**
     * @stability stable
     */
    set upstream(value) {
        this._upstream = value;
    }
    /**
     * @stability stable
     */
    resetUpstream() {
        this._upstream = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get upstreamInput() {
        return this._upstream;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            domain: cdktf.stringToTerraform(this._domain),
            domain_owner: cdktf.stringToTerraform(this._domainOwner),
            repository: cdktf.stringToTerraform(this._repository),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            external_connections: codeartifactRepositoryExternalConnectionsToTerraform(this._externalConnections.internalValue),
            upstream: cdktf.listMapper(codeartifactRepositoryUpstreamToTerraform)(this._upstream),
        };
    }
}
exports.CodeartifactRepository = CodeartifactRepository;
_b = JSII_RTTI_SYMBOL_1;
CodeartifactRepository[_b] = { fqn: "@cdktf/provider-aws.codeartifact.CodeartifactRepository", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
CodeartifactRepository.tfResourceType = "aws_codeartifact_repository";
