// @cdktf/provider-aws/lib/elb/app-cookie-stickiness-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppCookieStickinessPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/app_cookie_stickiness_policy.html aws_app_cookie_stickiness_policy}.
 *
 * @stability stable
 */
class AppCookieStickinessPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/app_cookie_stickiness_policy.html aws_app_cookie_stickiness_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_app_cookie_stickiness_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._cookieName = config.cookieName;
        this._lbPort = config.lbPort;
        this._loadBalancer = config.loadBalancer;
        this._name = config.name;
    }
    /**
     * @stability stable
     */
    get cookieName() {
        return this.getStringAttribute('cookie_name');
    }
    /**
     * @stability stable
     */
    set cookieName(value) {
        this._cookieName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cookieNameInput() {
        return this._cookieName;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get lbPort() {
        return this.getNumberAttribute('lb_port');
    }
    /**
     * @stability stable
     */
    set lbPort(value) {
        this._lbPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lbPortInput() {
        return this._lbPort;
    }
    /**
     * @stability stable
     */
    get loadBalancer() {
        return this.getStringAttribute('load_balancer');
    }
    /**
     * @stability stable
     */
    set loadBalancer(value) {
        this._loadBalancer = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerInput() {
        return this._loadBalancer;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cookie_name: cdktf.stringToTerraform(this._cookieName),
            lb_port: cdktf.numberToTerraform(this._lbPort),
            load_balancer: cdktf.stringToTerraform(this._loadBalancer),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.AppCookieStickinessPolicy = AppCookieStickinessPolicy;
_a = JSII_RTTI_SYMBOL_1;
AppCookieStickinessPolicy[_a] = { fqn: "@cdktf/provider-aws.elb.AppCookieStickinessPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
AppCookieStickinessPolicy.tfResourceType = "aws_app_cookie_stickiness_policy";
