// @cdktf/provider-aws/lib/elb/load-balancer-listener-policy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancerListenerPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/load_balancer_listener_policy.html aws_load_balancer_listener_policy}.
 *
 * @stability stable
 */
class LoadBalancerListenerPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/load_balancer_listener_policy.html aws_load_balancer_listener_policy} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_load_balancer_listener_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._loadBalancerName = config.loadBalancerName;
        this._loadBalancerPort = config.loadBalancerPort;
        this._policyNames = config.policyNames;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get loadBalancerName() {
        return this.getStringAttribute('load_balancer_name');
    }
    /**
     * @stability stable
     */
    set loadBalancerName(value) {
        this._loadBalancerName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerNameInput() {
        return this._loadBalancerName;
    }
    /**
     * @stability stable
     */
    get loadBalancerPort() {
        return this.getNumberAttribute('load_balancer_port');
    }
    /**
     * @stability stable
     */
    set loadBalancerPort(value) {
        this._loadBalancerPort = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerPortInput() {
        return this._loadBalancerPort;
    }
    /**
     * @stability stable
     */
    get policyNames() {
        return this.getListAttribute('policy_names');
    }
    /**
     * @stability stable
     */
    set policyNames(value) {
        this._policyNames = value;
    }
    /**
     * @stability stable
     */
    resetPolicyNames() {
        this._policyNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get policyNamesInput() {
        return this._policyNames;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            load_balancer_name: cdktf.stringToTerraform(this._loadBalancerName),
            load_balancer_port: cdktf.numberToTerraform(this._loadBalancerPort),
            policy_names: cdktf.listMapper(cdktf.stringToTerraform)(this._policyNames),
        };
    }
}
exports.LoadBalancerListenerPolicy = LoadBalancerListenerPolicy;
_a = JSII_RTTI_SYMBOL_1;
LoadBalancerListenerPolicy[_a] = { fqn: "@cdktf/provider-aws.elb.LoadBalancerListenerPolicy", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
LoadBalancerListenerPolicy.tfResourceType = "aws_load_balancer_listener_policy";
