// @cdktf/provider-aws/lib/ec2/data-aws-instances.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsInstances = exports.dataAwsInstancesFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsInstancesFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsInstancesFilterToTerraform = dataAwsInstancesFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/instances.html aws_instances}.
 *
 * @stability stable
 */
class DataAwsInstances extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/instances.html aws_instances} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_instances',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._instanceStateNames = config.instanceStateNames;
        this._instanceTags = config.instanceTags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ids() {
        return this.getListAttribute('ids');
    }
    /**
     * @stability stable
     */
    get instanceStateNames() {
        return this.getListAttribute('instance_state_names');
    }
    /**
     * @stability stable
     */
    set instanceStateNames(value) {
        this._instanceStateNames = value;
    }
    /**
     * @stability stable
     */
    resetInstanceStateNames() {
        this._instanceStateNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceStateNamesInput() {
        return this._instanceStateNames;
    }
    /**
     * @stability stable
     */
    get instanceTags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('instance_tags');
    }
    /**
     * @stability stable
     */
    set instanceTags(value) {
        this._instanceTags = value;
    }
    /**
     * @stability stable
     */
    resetInstanceTags() {
        this._instanceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTagsInput() {
        return this._instanceTags;
    }
    // private_ips - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIps() {
        return this.getListAttribute('private_ips');
    }
    // public_ips - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get publicIps() {
        return this.getListAttribute('public_ips');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            instance_state_names: cdktf.listMapper(cdktf.stringToTerraform)(this._instanceStateNames),
            instance_tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._instanceTags),
            filter: cdktf.listMapper(dataAwsInstancesFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsInstances = DataAwsInstances;
_a = JSII_RTTI_SYMBOL_1;
DataAwsInstances[_a] = { fqn: "@cdktf/provider-aws.ec2.DataAwsInstances", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsInstances.tfResourceType = "aws_instances";
