// @cdktf/provider-aws/lib/ec2/data-aws-launch-template.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLaunchTemplate = exports.dataAwsLaunchTemplateFilterToTerraform = exports.DataAwsLaunchTemplateTagSpecifications = exports.DataAwsLaunchTemplatePlacement = exports.DataAwsLaunchTemplateNetworkInterfaces = exports.DataAwsLaunchTemplateMonitoring = exports.DataAwsLaunchTemplateMetadataOptions = exports.DataAwsLaunchTemplateInstanceMarketOptions = exports.DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions = exports.DataAwsLaunchTemplateIamInstanceProfile = exports.DataAwsLaunchTemplateHibernationOptions = exports.DataAwsLaunchTemplateEnclaveOptions = exports.DataAwsLaunchTemplateElasticGpuSpecifications = exports.DataAwsLaunchTemplateCreditSpecification = exports.DataAwsLaunchTemplateBlockDeviceMappings = exports.DataAwsLaunchTemplateBlockDeviceMappingsEbs = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsLaunchTemplateBlockDeviceMappingsEbs extends cdktf.ComplexComputedList {
    // delete_on_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getStringAttribute('delete_on_termination');
    }
    // encrypted - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get encrypted() {
        return this.getStringAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // snapshot_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // throughput - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_size - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsLaunchTemplateBlockDeviceMappingsEbs = DataAwsLaunchTemplateBlockDeviceMappingsEbs;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateBlockDeviceMappingsEbs[_a] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateBlockDeviceMappingsEbs", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateBlockDeviceMappings extends cdktf.ComplexComputedList {
    // device_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // ebs - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ebs() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ebs');
    }
    // no_device - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get noDevice() {
        return this.getStringAttribute('no_device');
    }
    // virtual_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get virtualName() {
        return this.getStringAttribute('virtual_name');
    }
}
exports.DataAwsLaunchTemplateBlockDeviceMappings = DataAwsLaunchTemplateBlockDeviceMappings;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateBlockDeviceMappings[_b] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateBlockDeviceMappings", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateCreditSpecification extends cdktf.ComplexComputedList {
    // cpu_credits - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cpuCredits() {
        return this.getStringAttribute('cpu_credits');
    }
}
exports.DataAwsLaunchTemplateCreditSpecification = DataAwsLaunchTemplateCreditSpecification;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateCreditSpecification[_c] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateCreditSpecification", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateElasticGpuSpecifications extends cdktf.ComplexComputedList {
    // type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsLaunchTemplateElasticGpuSpecifications = DataAwsLaunchTemplateElasticGpuSpecifications;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateElasticGpuSpecifications[_d] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateElasticGpuSpecifications", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateEnclaveOptions extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsLaunchTemplateEnclaveOptions = DataAwsLaunchTemplateEnclaveOptions;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateEnclaveOptions[_e] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateEnclaveOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateHibernationOptions extends cdktf.ComplexComputedList {
    // configured - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get configured() {
        return this.getBooleanAttribute('configured');
    }
}
exports.DataAwsLaunchTemplateHibernationOptions = DataAwsLaunchTemplateHibernationOptions;
_f = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateHibernationOptions[_f] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateHibernationOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateIamInstanceProfile extends cdktf.ComplexComputedList {
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsLaunchTemplateIamInstanceProfile = DataAwsLaunchTemplateIamInstanceProfile;
_g = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateIamInstanceProfile[_g] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateIamInstanceProfile", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions extends cdktf.ComplexComputedList {
    // block_duration_minutes - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get blockDurationMinutes() {
        return this.getNumberAttribute('block_duration_minutes');
    }
    // instance_interruption_behavior - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceInterruptionBehavior() {
        return this.getStringAttribute('instance_interruption_behavior');
    }
    // max_price - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get maxPrice() {
        return this.getStringAttribute('max_price');
    }
    // spot_instance_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get spotInstanceType() {
        return this.getStringAttribute('spot_instance_type');
    }
    // valid_until - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get validUntil() {
        return this.getStringAttribute('valid_until');
    }
}
exports.DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions = DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions;
_h = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions[_h] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateInstanceMarketOptionsSpotOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateInstanceMarketOptions extends cdktf.ComplexComputedList {
    // market_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get marketType() {
        return this.getStringAttribute('market_type');
    }
    // spot_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get spotOptions() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('spot_options');
    }
}
exports.DataAwsLaunchTemplateInstanceMarketOptions = DataAwsLaunchTemplateInstanceMarketOptions;
_j = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateInstanceMarketOptions[_j] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateInstanceMarketOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateMetadataOptions extends cdktf.ComplexComputedList {
    // http_endpoint - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpEndpoint() {
        return this.getStringAttribute('http_endpoint');
    }
    // http_protocol_ipv6 - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpProtocolIpv6() {
        return this.getStringAttribute('http_protocol_ipv6');
    }
    // http_put_response_hop_limit - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpPutResponseHopLimit() {
        return this.getNumberAttribute('http_put_response_hop_limit');
    }
    // http_tokens - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get httpTokens() {
        return this.getStringAttribute('http_tokens');
    }
}
exports.DataAwsLaunchTemplateMetadataOptions = DataAwsLaunchTemplateMetadataOptions;
_k = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateMetadataOptions[_k] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateMetadataOptions", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateMonitoring extends cdktf.ComplexComputedList {
    // enabled - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsLaunchTemplateMonitoring = DataAwsLaunchTemplateMonitoring;
_l = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateMonitoring[_l] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateMonitoring", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateNetworkInterfaces extends cdktf.ComplexComputedList {
    // associate_carrier_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associateCarrierIpAddress() {
        return this.getStringAttribute('associate_carrier_ip_address');
    }
    // associate_public_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associatePublicIpAddress() {
        return this.getStringAttribute('associate_public_ip_address');
    }
    // delete_on_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deleteOnTermination() {
        return this.getStringAttribute('delete_on_termination');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // device_index - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get deviceIndex() {
        return this.getNumberAttribute('device_index');
    }
    // interface_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get interfaceType() {
        return this.getStringAttribute('interface_type');
    }
    // ipv4_address_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv4AddressCount() {
        return this.getNumberAttribute('ipv4_address_count');
    }
    // ipv4_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv4Addresses() {
        return this.getListAttribute('ipv4_addresses');
    }
    // ipv6_address_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6AddressCount() {
        return this.getNumberAttribute('ipv6_address_count');
    }
    // ipv6_addresses - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ipv6Addresses() {
        return this.getListAttribute('ipv6_addresses');
    }
    // network_card_index - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkCardIndex() {
        return this.getNumberAttribute('network_card_index');
    }
    // network_interface_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    // private_ip_address - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // security_groups - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    // subnet_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.DataAwsLaunchTemplateNetworkInterfaces = DataAwsLaunchTemplateNetworkInterfaces;
_m = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateNetworkInterfaces[_m] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateNetworkInterfaces", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplatePlacement extends cdktf.ComplexComputedList {
    // affinity - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get affinity() {
        return this.getStringAttribute('affinity');
    }
    // availability_zone - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // group_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    // host_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostId() {
        return this.getStringAttribute('host_id');
    }
    // host_resource_group_arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get hostResourceGroupArn() {
        return this.getStringAttribute('host_resource_group_arn');
    }
    // partition_number - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get partitionNumber() {
        return this.getNumberAttribute('partition_number');
    }
    // spread_domain - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get spreadDomain() {
        return this.getStringAttribute('spread_domain');
    }
    // tenancy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
}
exports.DataAwsLaunchTemplatePlacement = DataAwsLaunchTemplatePlacement;
_o = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplatePlacement[_o] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplatePlacement", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsLaunchTemplateTagSpecifications extends cdktf.ComplexComputedList {
    // resource_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    // tags - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
}
exports.DataAwsLaunchTemplateTagSpecifications = DataAwsLaunchTemplateTagSpecifications;
_p = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplateTagSpecifications[_p] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplateTagSpecifications", version: "3.0.1" };
function dataAwsLaunchTemplateFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsLaunchTemplateFilterToTerraform = dataAwsLaunchTemplateFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/launch_template.html aws_launch_template}.
 *
 * @stability stable
 */
class DataAwsLaunchTemplate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/launch_template.html aws_launch_template} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_launch_template',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._name = config.name;
        this._tags = config.tags;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    // block_device_mappings - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    blockDeviceMappings(index) {
        return new DataAwsLaunchTemplateBlockDeviceMappings(this, 'block_device_mappings', index);
    }
    // credit_specification - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    creditSpecification(index) {
        return new DataAwsLaunchTemplateCreditSpecification(this, 'credit_specification', index);
    }
    // default_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get defaultVersion() {
        return this.getNumberAttribute('default_version');
    }
    // description - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    // disable_api_termination - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get disableApiTermination() {
        return this.getBooleanAttribute('disable_api_termination');
    }
    // ebs_optimized - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getStringAttribute('ebs_optimized');
    }
    // elastic_gpu_specifications - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    elasticGpuSpecifications(index) {
        return new DataAwsLaunchTemplateElasticGpuSpecifications(this, 'elastic_gpu_specifications', index);
    }
    // enclave_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    enclaveOptions(index) {
        return new DataAwsLaunchTemplateEnclaveOptions(this, 'enclave_options', index);
    }
    // hibernation_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    hibernationOptions(index) {
        return new DataAwsLaunchTemplateHibernationOptions(this, 'hibernation_options', index);
    }
    // iam_instance_profile - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    iamInstanceProfile(index) {
        return new DataAwsLaunchTemplateIamInstanceProfile(this, 'iam_instance_profile', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // image_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    // instance_initiated_shutdown_behavior - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceInitiatedShutdownBehavior() {
        return this.getStringAttribute('instance_initiated_shutdown_behavior');
    }
    // instance_market_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    instanceMarketOptions(index) {
        return new DataAwsLaunchTemplateInstanceMarketOptions(this, 'instance_market_options', index);
    }
    // instance_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // kernel_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get kernelId() {
        return this.getStringAttribute('kernel_id');
    }
    // key_name - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    // latest_version - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get latestVersion() {
        return this.getNumberAttribute('latest_version');
    }
    // metadata_options - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    metadataOptions(index) {
        return new DataAwsLaunchTemplateMetadataOptions(this, 'metadata_options', index);
    }
    // monitoring - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    monitoring(index) {
        return new DataAwsLaunchTemplateMonitoring(this, 'monitoring', index);
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // network_interfaces - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    networkInterfaces(index) {
        return new DataAwsLaunchTemplateNetworkInterfaces(this, 'network_interfaces', index);
    }
    // placement - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    placement(index) {
        return new DataAwsLaunchTemplatePlacement(this, 'placement', index);
    }
    // ram_disk_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ramDiskId() {
        return this.getStringAttribute('ram_disk_id');
    }
    // security_group_names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get securityGroupNames() {
        return this.getListAttribute('security_group_names');
    }
    // tag_specifications - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    tagSpecifications(index) {
        return new DataAwsLaunchTemplateTagSpecifications(this, 'tag_specifications', index);
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    // user_data - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get userData() {
        return this.getStringAttribute('user_data');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            filter: cdktf.listMapper(dataAwsLaunchTemplateFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsLaunchTemplate = DataAwsLaunchTemplate;
_q = JSII_RTTI_SYMBOL_1;
DataAwsLaunchTemplate[_q] = { fqn: "@cdktf/provider-aws.ec2.DataAwsLaunchTemplate", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsLaunchTemplate.tfResourceType = "aws_launch_template";
