// @cdktf/provider-aws/lib/ec2/ec2-capacity-reservation.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2CapacityReservation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_capacity_reservation.html aws_ec2_capacity_reservation}.
 *
 * @stability stable
 */
class Ec2CapacityReservation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_capacity_reservation.html aws_ec2_capacity_reservation} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_capacity_reservation',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._availabilityZone = config.availabilityZone;
        this._ebsOptimized = config.ebsOptimized;
        this._endDate = config.endDate;
        this._endDateType = config.endDateType;
        this._ephemeralStorage = config.ephemeralStorage;
        this._instanceCount = config.instanceCount;
        this._instanceMatchCriteria = config.instanceMatchCriteria;
        this._instancePlatform = config.instancePlatform;
        this._instanceType = config.instanceType;
        this._outpostArn = config.outpostArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._tenancy = config.tenancy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    /**
     * @stability stable
     */
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    /**
     * @stability stable
     */
    get ebsOptimized() {
        return this.getBooleanAttribute('ebs_optimized');
    }
    /**
     * @stability stable
     */
    set ebsOptimized(value) {
        this._ebsOptimized = value;
    }
    /**
     * @stability stable
     */
    resetEbsOptimized() {
        this._ebsOptimized = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ebsOptimizedInput() {
        return this._ebsOptimized;
    }
    /**
     * @stability stable
     */
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    /**
     * @stability stable
     */
    set endDate(value) {
        this._endDate = value;
    }
    /**
     * @stability stable
     */
    resetEndDate() {
        this._endDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endDateInput() {
        return this._endDate;
    }
    /**
     * @stability stable
     */
    get endDateType() {
        return this.getStringAttribute('end_date_type');
    }
    /**
     * @stability stable
     */
    set endDateType(value) {
        this._endDateType = value;
    }
    /**
     * @stability stable
     */
    resetEndDateType() {
        this._endDateType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endDateTypeInput() {
        return this._endDateType;
    }
    /**
     * @stability stable
     */
    get ephemeralStorage() {
        return this.getBooleanAttribute('ephemeral_storage');
    }
    /**
     * @stability stable
     */
    set ephemeralStorage(value) {
        this._ephemeralStorage = value;
    }
    /**
     * @stability stable
     */
    resetEphemeralStorage() {
        this._ephemeralStorage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ephemeralStorageInput() {
        return this._ephemeralStorage;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    /**
     * @stability stable
     */
    set instanceCount(value) {
        this._instanceCount = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceCountInput() {
        return this._instanceCount;
    }
    /**
     * @stability stable
     */
    get instanceMatchCriteria() {
        return this.getStringAttribute('instance_match_criteria');
    }
    /**
     * @stability stable
     */
    set instanceMatchCriteria(value) {
        this._instanceMatchCriteria = value;
    }
    /**
     * @stability stable
     */
    resetInstanceMatchCriteria() {
        this._instanceMatchCriteria = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceMatchCriteriaInput() {
        return this._instanceMatchCriteria;
    }
    /**
     * @stability stable
     */
    get instancePlatform() {
        return this.getStringAttribute('instance_platform');
    }
    /**
     * @stability stable
     */
    set instancePlatform(value) {
        this._instancePlatform = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instancePlatformInput() {
        return this._instancePlatform;
    }
    /**
     * @stability stable
     */
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    /**
     * @stability stable
     */
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceTypeInput() {
        return this._instanceType;
    }
    /**
     * @stability stable
     */
    get outpostArn() {
        return this.getStringAttribute('outpost_arn');
    }
    /**
     * @stability stable
     */
    set outpostArn(value) {
        this._outpostArn = value;
    }
    /**
     * @stability stable
     */
    resetOutpostArn() {
        this._outpostArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outpostArnInput() {
        return this._outpostArn;
    }
    // owner_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get tenancy() {
        return this.getStringAttribute('tenancy');
    }
    /**
     * @stability stable
     */
    set tenancy(value) {
        this._tenancy = value;
    }
    /**
     * @stability stable
     */
    resetTenancy() {
        this._tenancy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tenancyInput() {
        return this._tenancy;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            availability_zone: cdktf.stringToTerraform(this._availabilityZone),
            ebs_optimized: cdktf.booleanToTerraform(this._ebsOptimized),
            end_date: cdktf.stringToTerraform(this._endDate),
            end_date_type: cdktf.stringToTerraform(this._endDateType),
            ephemeral_storage: cdktf.booleanToTerraform(this._ephemeralStorage),
            instance_count: cdktf.numberToTerraform(this._instanceCount),
            instance_match_criteria: cdktf.stringToTerraform(this._instanceMatchCriteria),
            instance_platform: cdktf.stringToTerraform(this._instancePlatform),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            outpost_arn: cdktf.stringToTerraform(this._outpostArn),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            tenancy: cdktf.stringToTerraform(this._tenancy),
        };
    }
}
exports.Ec2CapacityReservation = Ec2CapacityReservation;
_a = JSII_RTTI_SYMBOL_1;
Ec2CapacityReservation[_a] = { fqn: "@cdktf/provider-aws.ec2.Ec2CapacityReservation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
Ec2CapacityReservation.tfResourceType = "aws_ec2_capacity_reservation";
