// @cdktf/provider-aws/lib/ecs/data-aws-ecs-service.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcsService = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecs_service.html aws_ecs_service}.
 *
 * @stability stable
 */
class DataAwsEcsService extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecs_service.html aws_ecs_service} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_service',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._clusterArn = config.clusterArn;
        this._serviceName = config.serviceName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arn() {
        return this.getStringAttribute('arn');
    }
    /**
     * @stability stable
     */
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    /**
     * @stability stable
     */
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterArnInput() {
        return this._clusterArn;
    }
    // desired_count - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get desiredCount() {
        return this.getNumberAttribute('desired_count');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // launch_type - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    // scheduling_strategy - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get schedulingStrategy() {
        return this.getStringAttribute('scheduling_strategy');
    }
    /**
     * @stability stable
     */
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    /**
     * @stability stable
     */
    set serviceName(value) {
        this._serviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceNameInput() {
        return this._serviceName;
    }
    // task_definition - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
            service_name: cdktf.stringToTerraform(this._serviceName),
        };
    }
}
exports.DataAwsEcsService = DataAwsEcsService;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcsService[_a] = { fqn: "@cdktf/provider-aws.ecs.DataAwsEcsService", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsEcsService.tfResourceType = "aws_ecs_service";
