// @cdktf/provider-aws/lib/ecs/ecs-service.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcsService = exports.EcsServiceTimeoutsOutputReference = exports.ecsServiceTimeoutsToTerraform = exports.EcsServiceServiceRegistriesOutputReference = exports.ecsServiceServiceRegistriesToTerraform = exports.ecsServicePlacementConstraintsToTerraform = exports.ecsServiceOrderedPlacementStrategyToTerraform = exports.EcsServiceNetworkConfigurationOutputReference = exports.ecsServiceNetworkConfigurationToTerraform = exports.ecsServiceLoadBalancerToTerraform = exports.EcsServiceDeploymentControllerOutputReference = exports.ecsServiceDeploymentControllerToTerraform = exports.EcsServiceDeploymentCircuitBreakerOutputReference = exports.ecsServiceDeploymentCircuitBreakerToTerraform = exports.ecsServiceCapacityProviderStrategyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecsServiceCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.ecsServiceCapacityProviderStrategyToTerraform = ecsServiceCapacityProviderStrategyToTerraform;
function ecsServiceDeploymentCircuitBreakerToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable: cdktf.booleanToTerraform(struct.enable),
        rollback: cdktf.booleanToTerraform(struct.rollback),
    };
}
exports.ecsServiceDeploymentCircuitBreakerToTerraform = ecsServiceDeploymentCircuitBreakerToTerraform;
/**
 * @stability stable
 */
class EcsServiceDeploymentCircuitBreakerOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enable) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._rollback) {
            hasAnyValues = true;
            internalValueResult.rollback = this._rollback;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enable = undefined;
            this._rollback = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enable = value.enable;
            this._rollback = value.rollback;
        }
    }
    /**
     * @stability stable
     */
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    /**
     * @stability stable
     */
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableInput() {
        return this._enable;
    }
    /**
     * @stability stable
     */
    get rollback() {
        return this.getBooleanAttribute('rollback');
    }
    /**
     * @stability stable
     */
    set rollback(value) {
        this._rollback = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get rollbackInput() {
        return this._rollback;
    }
}
exports.EcsServiceDeploymentCircuitBreakerOutputReference = EcsServiceDeploymentCircuitBreakerOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentCircuitBreakerOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecs.EcsServiceDeploymentCircuitBreakerOutputReference", version: "3.0.1" };
function ecsServiceDeploymentControllerToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceDeploymentControllerToTerraform = ecsServiceDeploymentControllerToTerraform;
/**
 * @stability stable
 */
class EcsServiceDeploymentControllerOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    /**
     * @stability stable
     */
    get type() {
        return this.getStringAttribute('type');
    }
    /**
     * @stability stable
     */
    set type(value) {
        this._type = value;
    }
    /**
     * @stability stable
     */
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get typeInput() {
        return this._type;
    }
}
exports.EcsServiceDeploymentControllerOutputReference = EcsServiceDeploymentControllerOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentControllerOutputReference[_b] = { fqn: "@cdktf/provider-aws.ecs.EcsServiceDeploymentControllerOutputReference", version: "3.0.1" };
function ecsServiceLoadBalancerToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        elb_name: cdktf.stringToTerraform(struct.elbName),
        target_group_arn: cdktf.stringToTerraform(struct.targetGroupArn),
    };
}
exports.ecsServiceLoadBalancerToTerraform = ecsServiceLoadBalancerToTerraform;
function ecsServiceNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform)(struct.subnets),
    };
}
exports.ecsServiceNetworkConfigurationToTerraform = ecsServiceNetworkConfigurationToTerraform;
/**
 * @stability stable
 */
class EcsServiceNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    /**
     * @stability stable
     */
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    /**
     * @stability stable
     */
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    /**
     * @stability stable
     */
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    /**
     * @stability stable
     */
    get securityGroups() {
        return this.getListAttribute('security_groups');
    }
    /**
     * @stability stable
     */
    set securityGroups(value) {
        this._securityGroups = value;
    }
    /**
     * @stability stable
     */
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get securityGroupsInput() {
        return this._securityGroups;
    }
    /**
     * @stability stable
     */
    get subnets() {
        return this.getListAttribute('subnets');
    }
    /**
     * @stability stable
     */
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get subnetsInput() {
        return this._subnets;
    }
}
exports.EcsServiceNetworkConfigurationOutputReference = EcsServiceNetworkConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EcsServiceNetworkConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.ecs.EcsServiceNetworkConfigurationOutputReference", version: "3.0.1" };
function ecsServiceOrderedPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceOrderedPlacementStrategyToTerraform = ecsServiceOrderedPlacementStrategyToTerraform;
function ecsServicePlacementConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServicePlacementConstraintsToTerraform = ecsServicePlacementConstraintsToTerraform;
function ecsServiceServiceRegistriesToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        port: cdktf.numberToTerraform(struct.port),
        registry_arn: cdktf.stringToTerraform(struct.registryArn),
    };
}
exports.ecsServiceServiceRegistriesToTerraform = ecsServiceServiceRegistriesToTerraform;
/**
 * @stability stable
 */
class EcsServiceServiceRegistriesOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerName) {
            hasAnyValues = true;
            internalValueResult.containerName = this._containerName;
        }
        if (this._containerPort) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._port) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._registryArn) {
            hasAnyValues = true;
            internalValueResult.registryArn = this._registryArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerName = undefined;
            this._containerPort = undefined;
            this._port = undefined;
            this._registryArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerName = value.containerName;
            this._containerPort = value.containerPort;
            this._port = value.port;
            this._registryArn = value.registryArn;
        }
    }
    /**
     * @stability stable
     */
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    /**
     * @stability stable
     */
    set containerName(value) {
        this._containerName = value;
    }
    /**
     * @stability stable
     */
    resetContainerName() {
        this._containerName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get containerNameInput() {
        return this._containerName;
    }
    /**
     * @stability stable
     */
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    /**
     * @stability stable
     */
    set containerPort(value) {
        this._containerPort = value;
    }
    /**
     * @stability stable
     */
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get containerPortInput() {
        return this._containerPort;
    }
    /**
     * @stability stable
     */
    get port() {
        return this.getNumberAttribute('port');
    }
    /**
     * @stability stable
     */
    set port(value) {
        this._port = value;
    }
    /**
     * @stability stable
     */
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get portInput() {
        return this._port;
    }
    /**
     * @stability stable
     */
    get registryArn() {
        return this.getStringAttribute('registry_arn');
    }
    /**
     * @stability stable
     */
    set registryArn(value) {
        this._registryArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get registryArnInput() {
        return this._registryArn;
    }
}
exports.EcsServiceServiceRegistriesOutputReference = EcsServiceServiceRegistriesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EcsServiceServiceRegistriesOutputReference[_d] = { fqn: "@cdktf/provider-aws.ecs.EcsServiceServiceRegistriesOutputReference", version: "3.0.1" };
function ecsServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.ecsServiceTimeoutsToTerraform = ecsServiceTimeoutsToTerraform;
/**
 * @stability stable
 */
class EcsServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delete) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._delete = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._delete = value.delete;
        }
    }
    /**
     * @stability stable
     */
    get delete() {
        return this.getStringAttribute('delete');
    }
    /**
     * @stability stable
     */
    set delete(value) {
        this._delete = value;
    }
    /**
     * @stability stable
     */
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deleteInput() {
        return this._delete;
    }
}
exports.EcsServiceTimeoutsOutputReference = EcsServiceTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EcsServiceTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.ecs.EcsServiceTimeoutsOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ecs_service.html aws_ecs_service}.
 *
 * @stability stable
 */
class EcsService extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ecs_service.html aws_ecs_service} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_service',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // deployment_circuit_breaker - computed: false, optional: true, required: false
        this._deploymentCircuitBreaker = new EcsServiceDeploymentCircuitBreakerOutputReference(this, "deployment_circuit_breaker", true);
        // deployment_controller - computed: false, optional: true, required: false
        this._deploymentController = new EcsServiceDeploymentControllerOutputReference(this, "deployment_controller", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new EcsServiceNetworkConfigurationOutputReference(this, "network_configuration", true);
        // service_registries - computed: false, optional: true, required: false
        this._serviceRegistries = new EcsServiceServiceRegistriesOutputReference(this, "service_registries", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EcsServiceTimeoutsOutputReference(this, "timeouts", true);
        this._cluster = config.cluster;
        this._deploymentMaximumPercent = config.deploymentMaximumPercent;
        this._deploymentMinimumHealthyPercent = config.deploymentMinimumHealthyPercent;
        this._desiredCount = config.desiredCount;
        this._enableEcsManagedTags = config.enableEcsManagedTags;
        this._enableExecuteCommand = config.enableExecuteCommand;
        this._forceNewDeployment = config.forceNewDeployment;
        this._healthCheckGracePeriodSeconds = config.healthCheckGracePeriodSeconds;
        this._iamRole = config.iamRole;
        this._launchType = config.launchType;
        this._name = config.name;
        this._platformVersion = config.platformVersion;
        this._propagateTags = config.propagateTags;
        this._schedulingStrategy = config.schedulingStrategy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._taskDefinition = config.taskDefinition;
        this._waitForSteadyState = config.waitForSteadyState;
        this._capacityProviderStrategy = config.capacityProviderStrategy;
        this._deploymentCircuitBreaker.internalValue = config.deploymentCircuitBreaker;
        this._deploymentController.internalValue = config.deploymentController;
        this._loadBalancer = config.loadBalancer;
        this._networkConfiguration.internalValue = config.networkConfiguration;
        this._orderedPlacementStrategy = config.orderedPlacementStrategy;
        this._placementConstraints = config.placementConstraints;
        this._serviceRegistries.internalValue = config.serviceRegistries;
        this._timeouts.internalValue = config.timeouts;
    }
    /**
     * @stability stable
     */
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    /**
     * @stability stable
     */
    set cluster(value) {
        this._cluster = value;
    }
    /**
     * @stability stable
     */
    resetCluster() {
        this._cluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clusterInput() {
        return this._cluster;
    }
    /**
     * @stability stable
     */
    get deploymentMaximumPercent() {
        return this.getNumberAttribute('deployment_maximum_percent');
    }
    /**
     * @stability stable
     */
    set deploymentMaximumPercent(value) {
        this._deploymentMaximumPercent = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentMaximumPercent() {
        this._deploymentMaximumPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentMaximumPercentInput() {
        return this._deploymentMaximumPercent;
    }
    /**
     * @stability stable
     */
    get deploymentMinimumHealthyPercent() {
        return this.getNumberAttribute('deployment_minimum_healthy_percent');
    }
    /**
     * @stability stable
     */
    set deploymentMinimumHealthyPercent(value) {
        this._deploymentMinimumHealthyPercent = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentMinimumHealthyPercent() {
        this._deploymentMinimumHealthyPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentMinimumHealthyPercentInput() {
        return this._deploymentMinimumHealthyPercent;
    }
    /**
     * @stability stable
     */
    get desiredCount() {
        return this.getNumberAttribute('desired_count');
    }
    /**
     * @stability stable
     */
    set desiredCount(value) {
        this._desiredCount = value;
    }
    /**
     * @stability stable
     */
    resetDesiredCount() {
        this._desiredCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get desiredCountInput() {
        return this._desiredCount;
    }
    /**
     * @stability stable
     */
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    /**
     * @stability stable
     */
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    /**
     * @stability stable
     */
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    /**
     * @stability stable
     */
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    /**
     * @stability stable
     */
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    /**
     * @stability stable
     */
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    /**
     * @stability stable
     */
    get forceNewDeployment() {
        return this.getBooleanAttribute('force_new_deployment');
    }
    /**
     * @stability stable
     */
    set forceNewDeployment(value) {
        this._forceNewDeployment = value;
    }
    /**
     * @stability stable
     */
    resetForceNewDeployment() {
        this._forceNewDeployment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get forceNewDeploymentInput() {
        return this._forceNewDeployment;
    }
    /**
     * @stability stable
     */
    get healthCheckGracePeriodSeconds() {
        return this.getNumberAttribute('health_check_grace_period_seconds');
    }
    /**
     * @stability stable
     */
    set healthCheckGracePeriodSeconds(value) {
        this._healthCheckGracePeriodSeconds = value;
    }
    /**
     * @stability stable
     */
    resetHealthCheckGracePeriodSeconds() {
        this._healthCheckGracePeriodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get healthCheckGracePeriodSecondsInput() {
        return this._healthCheckGracePeriodSeconds;
    }
    /**
     * @stability stable
     */
    get iamRole() {
        return this.getStringAttribute('iam_role');
    }
    /**
     * @stability stable
     */
    set iamRole(value) {
        this._iamRole = value;
    }
    /**
     * @stability stable
     */
    resetIamRole() {
        this._iamRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get iamRoleInput() {
        return this._iamRole;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    /**
     * @stability stable
     */
    set launchType(value) {
        this._launchType = value;
    }
    /**
     * @stability stable
     */
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get launchTypeInput() {
        return this._launchType;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    /**
     * @stability stable
     */
    set platformVersion(value) {
        this._platformVersion = value;
    }
    /**
     * @stability stable
     */
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get platformVersionInput() {
        return this._platformVersion;
    }
    /**
     * @stability stable
     */
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    /**
     * @stability stable
     */
    set propagateTags(value) {
        this._propagateTags = value;
    }
    /**
     * @stability stable
     */
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get propagateTagsInput() {
        return this._propagateTags;
    }
    /**
     * @stability stable
     */
    get schedulingStrategy() {
        return this.getStringAttribute('scheduling_strategy');
    }
    /**
     * @stability stable
     */
    set schedulingStrategy(value) {
        this._schedulingStrategy = value;
    }
    /**
     * @stability stable
     */
    resetSchedulingStrategy() {
        this._schedulingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get schedulingStrategyInput() {
        return this._schedulingStrategy;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    /**
     * @stability stable
     */
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    /**
     * @stability stable
     */
    set taskDefinition(value) {
        this._taskDefinition = value;
    }
    /**
     * @stability stable
     */
    resetTaskDefinition() {
        this._taskDefinition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskDefinitionInput() {
        return this._taskDefinition;
    }
    /**
     * @stability stable
     */
    get waitForSteadyState() {
        return this.getBooleanAttribute('wait_for_steady_state');
    }
    /**
     * @stability stable
     */
    set waitForSteadyState(value) {
        this._waitForSteadyState = value;
    }
    /**
     * @stability stable
     */
    resetWaitForSteadyState() {
        this._waitForSteadyState = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get waitForSteadyStateInput() {
        return this._waitForSteadyState;
    }
    /**
     * @stability stable
     */
    get capacityProviderStrategy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('capacity_provider_strategy');
    }
    /**
     * @stability stable
     */
    set capacityProviderStrategy(value) {
        this._capacityProviderStrategy = value;
    }
    /**
     * @stability stable
     */
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy;
    }
    /**
     * @stability stable
     */
    get deploymentCircuitBreaker() {
        return this._deploymentCircuitBreaker;
    }
    /**
     * @stability stable
     */
    putDeploymentCircuitBreaker(value) {
        this._deploymentCircuitBreaker.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentCircuitBreaker() {
        this._deploymentCircuitBreaker.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentCircuitBreakerInput() {
        return this._deploymentCircuitBreaker.internalValue;
    }
    /**
     * @stability stable
     */
    get deploymentController() {
        return this._deploymentController;
    }
    /**
     * @stability stable
     */
    putDeploymentController(value) {
        this._deploymentController.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetDeploymentController() {
        this._deploymentController.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get deploymentControllerInput() {
        return this._deploymentController.internalValue;
    }
    /**
     * @stability stable
     */
    get loadBalancer() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('load_balancer');
    }
    /**
     * @stability stable
     */
    set loadBalancer(value) {
        this._loadBalancer = value;
    }
    /**
     * @stability stable
     */
    resetLoadBalancer() {
        this._loadBalancer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get loadBalancerInput() {
        return this._loadBalancer;
    }
    /**
     * @stability stable
     */
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    /**
     * @stability stable
     */
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    /**
     * @stability stable
     */
    get orderedPlacementStrategy() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('ordered_placement_strategy');
    }
    /**
     * @stability stable
     */
    set orderedPlacementStrategy(value) {
        this._orderedPlacementStrategy = value;
    }
    /**
     * @stability stable
     */
    resetOrderedPlacementStrategy() {
        this._orderedPlacementStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get orderedPlacementStrategyInput() {
        return this._orderedPlacementStrategy;
    }
    /**
     * @stability stable
     */
    get placementConstraints() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('placement_constraints');
    }
    /**
     * @stability stable
     */
    set placementConstraints(value) {
        this._placementConstraints = value;
    }
    /**
     * @stability stable
     */
    resetPlacementConstraints() {
        this._placementConstraints = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get placementConstraintsInput() {
        return this._placementConstraints;
    }
    /**
     * @stability stable
     */
    get serviceRegistries() {
        return this._serviceRegistries;
    }
    /**
     * @stability stable
     */
    putServiceRegistries(value) {
        this._serviceRegistries.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetServiceRegistries() {
        this._serviceRegistries.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceRegistriesInput() {
        return this._serviceRegistries.internalValue;
    }
    /**
     * @stability stable
     */
    get timeouts() {
        return this._timeouts;
    }
    /**
     * @stability stable
     */
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cluster: cdktf.stringToTerraform(this._cluster),
            deployment_maximum_percent: cdktf.numberToTerraform(this._deploymentMaximumPercent),
            deployment_minimum_healthy_percent: cdktf.numberToTerraform(this._deploymentMinimumHealthyPercent),
            desired_count: cdktf.numberToTerraform(this._desiredCount),
            enable_ecs_managed_tags: cdktf.booleanToTerraform(this._enableEcsManagedTags),
            enable_execute_command: cdktf.booleanToTerraform(this._enableExecuteCommand),
            force_new_deployment: cdktf.booleanToTerraform(this._forceNewDeployment),
            health_check_grace_period_seconds: cdktf.numberToTerraform(this._healthCheckGracePeriodSeconds),
            iam_role: cdktf.stringToTerraform(this._iamRole),
            launch_type: cdktf.stringToTerraform(this._launchType),
            name: cdktf.stringToTerraform(this._name),
            platform_version: cdktf.stringToTerraform(this._platformVersion),
            propagate_tags: cdktf.stringToTerraform(this._propagateTags),
            scheduling_strategy: cdktf.stringToTerraform(this._schedulingStrategy),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
            task_definition: cdktf.stringToTerraform(this._taskDefinition),
            wait_for_steady_state: cdktf.booleanToTerraform(this._waitForSteadyState),
            capacity_provider_strategy: cdktf.listMapper(ecsServiceCapacityProviderStrategyToTerraform)(this._capacityProviderStrategy),
            deployment_circuit_breaker: ecsServiceDeploymentCircuitBreakerToTerraform(this._deploymentCircuitBreaker.internalValue),
            deployment_controller: ecsServiceDeploymentControllerToTerraform(this._deploymentController.internalValue),
            load_balancer: cdktf.listMapper(ecsServiceLoadBalancerToTerraform)(this._loadBalancer),
            network_configuration: ecsServiceNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
            ordered_placement_strategy: cdktf.listMapper(ecsServiceOrderedPlacementStrategyToTerraform)(this._orderedPlacementStrategy),
            placement_constraints: cdktf.listMapper(ecsServicePlacementConstraintsToTerraform)(this._placementConstraints),
            service_registries: ecsServiceServiceRegistriesToTerraform(this._serviceRegistries.internalValue),
            timeouts: ecsServiceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.EcsService = EcsService;
_f = JSII_RTTI_SYMBOL_1;
EcsService[_f] = { fqn: "@cdktf/provider-aws.ecs.EcsService", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
EcsService.tfResourceType = "aws_ecs_service";
