// jsii-reflect/lib/docs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Docs = void 0;
const spec_1 = require("@jsii/spec");
class Docs {
    constructor(system, target, spec, parentDocs) {
        this.system = system;
        this.target = target;
        this.parentDocs = parentDocs;
        this.docs = spec !== null && spec !== void 0 ? spec : {};
    }
    /**
     * Returns docstring of summary and remarks
     */
    toString() {
        return [this.docs.summary, this.docs.remarks]
            .filter((txt) => !!txt)
            .join('\n\n');
    }
    get subclassable() {
        return !!this.docs.subclassable;
    }
    /**
     * Return the reason for deprecation of this type
     */
    get deprecationReason() {
        if (this.docs.deprecated !== undefined) {
            return this.docs.deprecated;
        }
        if (this.parentDocs) {
            return this.parentDocs.deprecationReason;
        }
        return undefined;
    }
    /**
     * Return whether this type is deprecated
     */
    get deprecated() {
        return this.deprecationReason !== undefined;
    }
    /**
     * Return the stability of this type
     */
    get stability() {
        var _a;
        return lowestStability(this.docs.stability, (_a = this.parentDocs) === null || _a === void 0 ? void 0 : _a.stability);
    }
    /**
     * Return any custom tags on this type
     */
    customTag(tag) {
        var _a;
        return (_a = this.docs.custom) === null || _a === void 0 ? void 0 : _a[tag];
    }
    /**
     * Return summary of this type
     */
    get summary() {
        var _a;
        return (_a = this.docs.summary) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * Return remarks for this type
     */
    get remarks() {
        var _a;
        return (_a = this.docs.remarks) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * Return examples for this type
     */
    get example() {
        var _a;
        return (_a = this.docs.example) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * Return documentation links for this type
     */
    get link() {
        var _a;
        return (_a = this.docs.see) !== null && _a !== void 0 ? _a : '';
    }
    /**
     * Returns the return type
     */
    get returns() {
        var _a;
        return (_a = this.docs.returns) !== null && _a !== void 0 ? _a : '';
    }
}
exports.Docs = Docs;
const stabilityPrecedence = {
    [spec_1.Stability.Deprecated]: 0,
    [spec_1.Stability.Experimental]: 1,
    [spec_1.Stability.External]: 2,
    [spec_1.Stability.Stable]: 3,
};
function lowestStability(a, b) {
    if (a === undefined) {
        return b;
    }
    if (b === undefined) {
        return a;
    }
    return stabilityPrecedence[a] < stabilityPrecedence[b] ? a : b;
}
