// @cdktf/provider-aws/lib/ssm/data-aws-ssm-parameters-by-path.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSsmParametersByPath = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/ssm_parameters_by_path.html aws_ssm_parameters_by_path}.
 *
 * @stability stable
 */
class DataAwsSsmParametersByPath extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ssm_parameters_by_path.html aws_ssm_parameters_by_path} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_parameters_by_path',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._path = config.path;
        this._withDecryption = config.withDecryption;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arns - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get arns() {
        return this.getListAttribute('arns');
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    // names - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get names() {
        return this.getListAttribute('names');
    }
    /**
     * @stability stable
     */
    get path() {
        return this.getStringAttribute('path');
    }
    /**
     * @stability stable
     */
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get pathInput() {
        return this._path;
    }
    // types - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get types() {
        return this.getListAttribute('types');
    }
    // values - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get values() {
        return this.getListAttribute('values');
    }
    /**
     * @stability stable
     */
    get withDecryption() {
        return this.getBooleanAttribute('with_decryption');
    }
    /**
     * @stability stable
     */
    set withDecryption(value) {
        this._withDecryption = value;
    }
    /**
     * @stability stable
     */
    resetWithDecryption() {
        this._withDecryption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get withDecryptionInput() {
        return this._withDecryption;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            path: cdktf.stringToTerraform(this._path),
            with_decryption: cdktf.booleanToTerraform(this._withDecryption),
        };
    }
}
exports.DataAwsSsmParametersByPath = DataAwsSsmParametersByPath;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSsmParametersByPath[_a] = { fqn: "@cdktf/provider-aws.ssm.DataAwsSsmParametersByPath", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsSsmParametersByPath.tfResourceType = "aws_ssm_parameters_by_path";
