// @cdktf/provider-aws/lib/ssm/ssm-association.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmAssociation = exports.ssmAssociationTargetsToTerraform = exports.SsmAssociationOutputLocationOutputReference = exports.ssmAssociationOutputLocationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssmAssociationOutputLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_key_prefix: cdktf.stringToTerraform(struct.s3KeyPrefix),
        s3_region: cdktf.stringToTerraform(struct.s3Region),
    };
}
exports.ssmAssociationOutputLocationToTerraform = ssmAssociationOutputLocationToTerraform;
/**
 * @stability stable
 */
class SsmAssociationOutputLocationOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3BucketName) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3KeyPrefix) {
            hasAnyValues = true;
            internalValueResult.s3KeyPrefix = this._s3KeyPrefix;
        }
        if (this._s3Region) {
            hasAnyValues = true;
            internalValueResult.s3Region = this._s3Region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3BucketName = undefined;
            this._s3KeyPrefix = undefined;
            this._s3Region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3BucketName = value.s3BucketName;
            this._s3KeyPrefix = value.s3KeyPrefix;
            this._s3Region = value.s3Region;
        }
    }
    /**
     * @stability stable
     */
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    /**
     * @stability stable
     */
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    /**
     * @stability stable
     */
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    /**
     * @stability stable
     */
    set s3KeyPrefix(value) {
        this._s3KeyPrefix = value;
    }
    /**
     * @stability stable
     */
    resetS3KeyPrefix() {
        this._s3KeyPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3KeyPrefixInput() {
        return this._s3KeyPrefix;
    }
    /**
     * @stability stable
     */
    get s3Region() {
        return this.getStringAttribute('s3_region');
    }
    /**
     * @stability stable
     */
    set s3Region(value) {
        this._s3Region = value;
    }
    /**
     * @stability stable
     */
    resetS3Region() {
        this._s3Region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get s3RegionInput() {
        return this._s3Region;
    }
}
exports.SsmAssociationOutputLocationOutputReference = SsmAssociationOutputLocationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsmAssociationOutputLocationOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssm.SsmAssociationOutputLocationOutputReference", version: "3.0.1" };
function ssmAssociationTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.ssmAssociationTargetsToTerraform = ssmAssociationTargetsToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_association.html aws_ssm_association}.
 *
 * @stability stable
 */
class SsmAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_association.html aws_ssm_association} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_association',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // output_location - computed: false, optional: true, required: false
        this._outputLocation = new SsmAssociationOutputLocationOutputReference(this, "output_location", true);
        this._applyOnlyAtCronInterval = config.applyOnlyAtCronInterval;
        this._associationName = config.associationName;
        this._automationTargetParameterName = config.automationTargetParameterName;
        this._complianceSeverity = config.complianceSeverity;
        this._documentVersion = config.documentVersion;
        this._instanceId = config.instanceId;
        this._maxConcurrency = config.maxConcurrency;
        this._maxErrors = config.maxErrors;
        this._name = config.name;
        this._parameters = config.parameters;
        this._scheduleExpression = config.scheduleExpression;
        this._outputLocation.internalValue = config.outputLocation;
        this._targets = config.targets;
    }
    /**
     * @stability stable
     */
    get applyOnlyAtCronInterval() {
        return this.getBooleanAttribute('apply_only_at_cron_interval');
    }
    /**
     * @stability stable
     */
    set applyOnlyAtCronInterval(value) {
        this._applyOnlyAtCronInterval = value;
    }
    /**
     * @stability stable
     */
    resetApplyOnlyAtCronInterval() {
        this._applyOnlyAtCronInterval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get applyOnlyAtCronIntervalInput() {
        return this._applyOnlyAtCronInterval;
    }
    // association_id - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get associationId() {
        return this.getStringAttribute('association_id');
    }
    /**
     * @stability stable
     */
    get associationName() {
        return this.getStringAttribute('association_name');
    }
    /**
     * @stability stable
     */
    set associationName(value) {
        this._associationName = value;
    }
    /**
     * @stability stable
     */
    resetAssociationName() {
        this._associationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get associationNameInput() {
        return this._associationName;
    }
    /**
     * @stability stable
     */
    get automationTargetParameterName() {
        return this.getStringAttribute('automation_target_parameter_name');
    }
    /**
     * @stability stable
     */
    set automationTargetParameterName(value) {
        this._automationTargetParameterName = value;
    }
    /**
     * @stability stable
     */
    resetAutomationTargetParameterName() {
        this._automationTargetParameterName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get automationTargetParameterNameInput() {
        return this._automationTargetParameterName;
    }
    /**
     * @stability stable
     */
    get complianceSeverity() {
        return this.getStringAttribute('compliance_severity');
    }
    /**
     * @stability stable
     */
    set complianceSeverity(value) {
        this._complianceSeverity = value;
    }
    /**
     * @stability stable
     */
    resetComplianceSeverity() {
        this._complianceSeverity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get complianceSeverityInput() {
        return this._complianceSeverity;
    }
    /**
     * @stability stable
     */
    get documentVersion() {
        return this.getStringAttribute('document_version');
    }
    /**
     * @stability stable
     */
    set documentVersion(value) {
        this._documentVersion = value;
    }
    /**
     * @stability stable
     */
    resetDocumentVersion() {
        this._documentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentVersionInput() {
        return this._documentVersion;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    /**
     * @stability stable
     */
    set instanceId(value) {
        this._instanceId = value;
    }
    /**
     * @stability stable
     */
    resetInstanceId() {
        this._instanceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get instanceIdInput() {
        return this._instanceId;
    }
    /**
     * @stability stable
     */
    get maxConcurrency() {
        return this.getStringAttribute('max_concurrency');
    }
    /**
     * @stability stable
     */
    set maxConcurrency(value) {
        this._maxConcurrency = value;
    }
    /**
     * @stability stable
     */
    resetMaxConcurrency() {
        this._maxConcurrency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrencyInput() {
        return this._maxConcurrency;
    }
    /**
     * @stability stable
     */
    get maxErrors() {
        return this.getStringAttribute('max_errors');
    }
    /**
     * @stability stable
     */
    set maxErrors(value) {
        this._maxErrors = value;
    }
    /**
     * @stability stable
     */
    resetMaxErrors() {
        this._maxErrors = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxErrorsInput() {
        return this._maxErrors;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get parameters() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameters');
    }
    /**
     * @stability stable
     */
    set parameters(value) {
        this._parameters = value;
    }
    /**
     * @stability stable
     */
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parametersInput() {
        return this._parameters;
    }
    /**
     * @stability stable
     */
    get scheduleExpression() {
        return this.getStringAttribute('schedule_expression');
    }
    /**
     * @stability stable
     */
    set scheduleExpression(value) {
        this._scheduleExpression = value;
    }
    /**
     * @stability stable
     */
    resetScheduleExpression() {
        this._scheduleExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleExpressionInput() {
        return this._scheduleExpression;
    }
    /**
     * @stability stable
     */
    get outputLocation() {
        return this._outputLocation;
    }
    /**
     * @stability stable
     */
    putOutputLocation(value) {
        this._outputLocation.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetOutputLocation() {
        this._outputLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputLocationInput() {
        return this._outputLocation.internalValue;
    }
    /**
     * @stability stable
     */
    get targets() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('targets');
    }
    /**
     * @stability stable
     */
    set targets(value) {
        this._targets = value;
    }
    /**
     * @stability stable
     */
    resetTargets() {
        this._targets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetsInput() {
        return this._targets;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            apply_only_at_cron_interval: cdktf.booleanToTerraform(this._applyOnlyAtCronInterval),
            association_name: cdktf.stringToTerraform(this._associationName),
            automation_target_parameter_name: cdktf.stringToTerraform(this._automationTargetParameterName),
            compliance_severity: cdktf.stringToTerraform(this._complianceSeverity),
            document_version: cdktf.stringToTerraform(this._documentVersion),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            max_concurrency: cdktf.stringToTerraform(this._maxConcurrency),
            max_errors: cdktf.stringToTerraform(this._maxErrors),
            name: cdktf.stringToTerraform(this._name),
            parameters: cdktf.hashMapper(cdktf.anyToTerraform)(this._parameters),
            schedule_expression: cdktf.stringToTerraform(this._scheduleExpression),
            output_location: ssmAssociationOutputLocationToTerraform(this._outputLocation.internalValue),
            targets: cdktf.listMapper(ssmAssociationTargetsToTerraform)(this._targets),
        };
    }
}
exports.SsmAssociation = SsmAssociation;
_b = JSII_RTTI_SYMBOL_1;
SsmAssociation[_b] = { fqn: "@cdktf/provider-aws.ssm.SsmAssociation", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsmAssociation.tfResourceType = "aws_ssm_association";
