// @cdktf/provider-aws/lib/ssm/ssm-maintenance-window-task.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmMaintenanceWindowTask = exports.SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersToTerraform = exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersParameterToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersToTerraform = exports.SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference = exports.ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersToTerraform = exports.ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersParameterToTerraform = exports.ssmMaintenanceWindowTaskTargetsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssmMaintenanceWindowTaskTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.ssmMaintenanceWindowTaskTargetsToTerraform = ssmMaintenanceWindowTaskTargetsToTerraform;
function ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersParameterToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersParameterToTerraform;
function ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        document_version: cdktf.stringToTerraform(struct.documentVersion),
        parameter: cdktf.listMapper(ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersParameterToTerraform)(struct.parameter),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._documentVersion) {
            hasAnyValues = true;
            internalValueResult.documentVersion = this._documentVersion;
        }
        if (this._parameter) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._documentVersion = undefined;
            this._parameter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._documentVersion = value.documentVersion;
            this._parameter = value.parameter;
        }
    }
    /**
     * @stability stable
     */
    get documentVersion() {
        return this.getStringAttribute('document_version');
    }
    /**
     * @stability stable
     */
    set documentVersion(value) {
        this._documentVersion = value;
    }
    /**
     * @stability stable
     */
    resetDocumentVersion() {
        this._documentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentVersionInput() {
        return this._documentVersion;
    }
    /**
     * @stability stable
     */
    get parameter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameter');
    }
    /**
     * @stability stable
     */
    set parameter(value) {
        this._parameter = value;
    }
    /**
     * @stability stable
     */
    resetParameter() {
        this._parameter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parameterInput() {
        return this._parameter;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_context: cdktf.stringToTerraform(struct.clientContext),
        payload: cdktf.stringToTerraform(struct.payload),
        qualifier: cdktf.stringToTerraform(struct.qualifier),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientContext) {
            hasAnyValues = true;
            internalValueResult.clientContext = this._clientContext;
        }
        if (this._payload) {
            hasAnyValues = true;
            internalValueResult.payload = this._payload;
        }
        if (this._qualifier) {
            hasAnyValues = true;
            internalValueResult.qualifier = this._qualifier;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientContext = undefined;
            this._payload = undefined;
            this._qualifier = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientContext = value.clientContext;
            this._payload = value.payload;
            this._qualifier = value.qualifier;
        }
    }
    /**
     * @stability stable
     */
    get clientContext() {
        return this.getStringAttribute('client_context');
    }
    /**
     * @stability stable
     */
    set clientContext(value) {
        this._clientContext = value;
    }
    /**
     * @stability stable
     */
    resetClientContext() {
        this._clientContext = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get clientContextInput() {
        return this._clientContext;
    }
    /**
     * @stability stable
     */
    get payload() {
        return this.getStringAttribute('payload');
    }
    /**
     * @stability stable
     */
    set payload(value) {
        this._payload = value;
    }
    /**
     * @stability stable
     */
    resetPayload() {
        this._payload = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get payloadInput() {
        return this._payload;
    }
    /**
     * @stability stable
     */
    get qualifier() {
        return this.getStringAttribute('qualifier');
    }
    /**
     * @stability stable
     */
    set qualifier(value) {
        this._qualifier = value;
    }
    /**
     * @stability stable
     */
    resetQualifier() {
        this._qualifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get qualifierInput() {
        return this._qualifier;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference[_b] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_log_group_name: cdktf.stringToTerraform(struct.cloudwatchLogGroupName),
        cloudwatch_output_enabled: cdktf.booleanToTerraform(struct.cloudwatchOutputEnabled),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudwatchLogGroupName) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogGroupName = this._cloudwatchLogGroupName;
        }
        if (this._cloudwatchOutputEnabled) {
            hasAnyValues = true;
            internalValueResult.cloudwatchOutputEnabled = this._cloudwatchOutputEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchLogGroupName = undefined;
            this._cloudwatchOutputEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchLogGroupName = value.cloudwatchLogGroupName;
            this._cloudwatchOutputEnabled = value.cloudwatchOutputEnabled;
        }
    }
    /**
     * @stability stable
     */
    get cloudwatchLogGroupName() {
        return this.getStringAttribute('cloudwatch_log_group_name');
    }
    /**
     * @stability stable
     */
    set cloudwatchLogGroupName(value) {
        this._cloudwatchLogGroupName = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchLogGroupName() {
        this._cloudwatchLogGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchLogGroupNameInput() {
        return this._cloudwatchLogGroupName;
    }
    /**
     * @stability stable
     */
    get cloudwatchOutputEnabled() {
        return this.getBooleanAttribute('cloudwatch_output_enabled');
    }
    /**
     * @stability stable
     */
    set cloudwatchOutputEnabled(value) {
        this._cloudwatchOutputEnabled = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchOutputEnabled() {
        this._cloudwatchOutputEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchOutputEnabledInput() {
        return this._cloudwatchOutputEnabled;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        notification_arn: cdktf.stringToTerraform(struct.notificationArn),
        notification_events: cdktf.listMapper(cdktf.stringToTerraform)(struct.notificationEvents),
        notification_type: cdktf.stringToTerraform(struct.notificationType),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._notificationArn) {
            hasAnyValues = true;
            internalValueResult.notificationArn = this._notificationArn;
        }
        if (this._notificationEvents) {
            hasAnyValues = true;
            internalValueResult.notificationEvents = this._notificationEvents;
        }
        if (this._notificationType) {
            hasAnyValues = true;
            internalValueResult.notificationType = this._notificationType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._notificationArn = undefined;
            this._notificationEvents = undefined;
            this._notificationType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._notificationArn = value.notificationArn;
            this._notificationEvents = value.notificationEvents;
            this._notificationType = value.notificationType;
        }
    }
    /**
     * @stability stable
     */
    get notificationArn() {
        return this.getStringAttribute('notification_arn');
    }
    /**
     * @stability stable
     */
    set notificationArn(value) {
        this._notificationArn = value;
    }
    /**
     * @stability stable
     */
    resetNotificationArn() {
        this._notificationArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationArnInput() {
        return this._notificationArn;
    }
    /**
     * @stability stable
     */
    get notificationEvents() {
        return this.getListAttribute('notification_events');
    }
    /**
     * @stability stable
     */
    set notificationEvents(value) {
        this._notificationEvents = value;
    }
    /**
     * @stability stable
     */
    resetNotificationEvents() {
        this._notificationEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationEventsInput() {
        return this._notificationEvents;
    }
    /**
     * @stability stable
     */
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    /**
     * @stability stable
     */
    set notificationType(value) {
        this._notificationType = value;
    }
    /**
     * @stability stable
     */
    resetNotificationType() {
        this._notificationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationTypeInput() {
        return this._notificationType;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersParameterToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersParameterToTerraform;
function ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        comment: cdktf.stringToTerraform(struct.comment),
        document_hash: cdktf.stringToTerraform(struct.documentHash),
        document_hash_type: cdktf.stringToTerraform(struct.documentHashType),
        document_version: cdktf.stringToTerraform(struct.documentVersion),
        output_s3_bucket: cdktf.stringToTerraform(struct.outputS3Bucket),
        output_s3_key_prefix: cdktf.stringToTerraform(struct.outputS3KeyPrefix),
        service_role_arn: cdktf.stringToTerraform(struct.serviceRoleArn),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
        cloudwatch_config: ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigToTerraform(struct.cloudwatchConfig),
        notification_config: ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigToTerraform(struct.notificationConfig),
        parameter: cdktf.listMapper(ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersParameterToTerraform)(struct.parameter),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // cloudwatch_config - computed: false, optional: true, required: false
        this._cloudwatchConfig = new SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersCloudwatchConfigOutputReference(this, "cloudwatch_config", true);
        // notification_config - computed: false, optional: true, required: false
        this._notificationConfig = new SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersNotificationConfigOutputReference(this, "notification_config", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k, _l, _m;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._comment) {
            hasAnyValues = true;
            internalValueResult.comment = this._comment;
        }
        if (this._documentHash) {
            hasAnyValues = true;
            internalValueResult.documentHash = this._documentHash;
        }
        if (this._documentHashType) {
            hasAnyValues = true;
            internalValueResult.documentHashType = this._documentHashType;
        }
        if (this._documentVersion) {
            hasAnyValues = true;
            internalValueResult.documentVersion = this._documentVersion;
        }
        if (this._outputS3Bucket) {
            hasAnyValues = true;
            internalValueResult.outputS3Bucket = this._outputS3Bucket;
        }
        if (this._outputS3KeyPrefix) {
            hasAnyValues = true;
            internalValueResult.outputS3KeyPrefix = this._outputS3KeyPrefix;
        }
        if (this._serviceRoleArn) {
            hasAnyValues = true;
            internalValueResult.serviceRoleArn = this._serviceRoleArn;
        }
        if (this._timeoutSeconds) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        if ((_j = this._cloudwatchConfig) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.cloudwatchConfig = (_k = this._cloudwatchConfig) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._notificationConfig) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.notificationConfig = (_m = this._notificationConfig) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        if (this._parameter) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._comment = undefined;
            this._documentHash = undefined;
            this._documentHashType = undefined;
            this._documentVersion = undefined;
            this._outputS3Bucket = undefined;
            this._outputS3KeyPrefix = undefined;
            this._serviceRoleArn = undefined;
            this._timeoutSeconds = undefined;
            this._cloudwatchConfig.internalValue = undefined;
            this._notificationConfig.internalValue = undefined;
            this._parameter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._comment = value.comment;
            this._documentHash = value.documentHash;
            this._documentHashType = value.documentHashType;
            this._documentVersion = value.documentVersion;
            this._outputS3Bucket = value.outputS3Bucket;
            this._outputS3KeyPrefix = value.outputS3KeyPrefix;
            this._serviceRoleArn = value.serviceRoleArn;
            this._timeoutSeconds = value.timeoutSeconds;
            this._cloudwatchConfig.internalValue = value.cloudwatchConfig;
            this._notificationConfig.internalValue = value.notificationConfig;
            this._parameter = value.parameter;
        }
    }
    /**
     * @stability stable
     */
    get comment() {
        return this.getStringAttribute('comment');
    }
    /**
     * @stability stable
     */
    set comment(value) {
        this._comment = value;
    }
    /**
     * @stability stable
     */
    resetComment() {
        this._comment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get commentInput() {
        return this._comment;
    }
    /**
     * @stability stable
     */
    get documentHash() {
        return this.getStringAttribute('document_hash');
    }
    /**
     * @stability stable
     */
    set documentHash(value) {
        this._documentHash = value;
    }
    /**
     * @stability stable
     */
    resetDocumentHash() {
        this._documentHash = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentHashInput() {
        return this._documentHash;
    }
    /**
     * @stability stable
     */
    get documentHashType() {
        return this.getStringAttribute('document_hash_type');
    }
    /**
     * @stability stable
     */
    set documentHashType(value) {
        this._documentHashType = value;
    }
    /**
     * @stability stable
     */
    resetDocumentHashType() {
        this._documentHashType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentHashTypeInput() {
        return this._documentHashType;
    }
    /**
     * @stability stable
     */
    get documentVersion() {
        return this.getStringAttribute('document_version');
    }
    /**
     * @stability stable
     */
    set documentVersion(value) {
        this._documentVersion = value;
    }
    /**
     * @stability stable
     */
    resetDocumentVersion() {
        this._documentVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get documentVersionInput() {
        return this._documentVersion;
    }
    /**
     * @stability stable
     */
    get outputS3Bucket() {
        return this.getStringAttribute('output_s3_bucket');
    }
    /**
     * @stability stable
     */
    set outputS3Bucket(value) {
        this._outputS3Bucket = value;
    }
    /**
     * @stability stable
     */
    resetOutputS3Bucket() {
        this._outputS3Bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputS3BucketInput() {
        return this._outputS3Bucket;
    }
    /**
     * @stability stable
     */
    get outputS3KeyPrefix() {
        return this.getStringAttribute('output_s3_key_prefix');
    }
    /**
     * @stability stable
     */
    set outputS3KeyPrefix(value) {
        this._outputS3KeyPrefix = value;
    }
    /**
     * @stability stable
     */
    resetOutputS3KeyPrefix() {
        this._outputS3KeyPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get outputS3KeyPrefixInput() {
        return this._outputS3KeyPrefix;
    }
    /**
     * @stability stable
     */
    get serviceRoleArn() {
        return this.getStringAttribute('service_role_arn');
    }
    /**
     * @stability stable
     */
    set serviceRoleArn(value) {
        this._serviceRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetServiceRoleArn() {
        this._serviceRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceRoleArnInput() {
        return this._serviceRoleArn;
    }
    /**
     * @stability stable
     */
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    /**
     * @stability stable
     */
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    /**
     * @stability stable
     */
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
    /**
     * @stability stable
     */
    get cloudwatchConfig() {
        return this._cloudwatchConfig;
    }
    /**
     * @stability stable
     */
    putCloudwatchConfig(value) {
        this._cloudwatchConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCloudwatchConfig() {
        this._cloudwatchConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cloudwatchConfigInput() {
        return this._cloudwatchConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get notificationConfig() {
        return this._notificationConfig;
    }
    /**
     * @stability stable
     */
    putNotificationConfig(value) {
        this._notificationConfig.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetNotificationConfig() {
        this._notificationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get notificationConfigInput() {
        return this._notificationConfig.internalValue;
    }
    /**
     * @stability stable
     */
    get parameter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('parameter');
    }
    /**
     * @stability stable
     */
    set parameter(value) {
        this._parameter = value;
    }
    /**
     * @stability stable
     */
    resetParameter() {
        this._parameter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get parameterInput() {
        return this._parameter;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference[_e] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input: cdktf.stringToTerraform(struct.input),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._input) {
            hasAnyValues = true;
            internalValueResult.input = this._input;
        }
        if (this._name) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._input = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._input = value.input;
            this._name = value.name;
        }
    }
    /**
     * @stability stable
     */
    get input() {
        return this.getStringAttribute('input');
    }
    /**
     * @stability stable
     */
    set input(value) {
        this._input = value;
    }
    /**
     * @stability stable
     */
    resetInput() {
        this._input = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get inputInput() {
        return this._input;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference[_f] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference", version: "3.0.1" };
function ssmMaintenanceWindowTaskTaskInvocationParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        automation_parameters: ssmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersToTerraform(struct.automationParameters),
        lambda_parameters: ssmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersToTerraform(struct.lambdaParameters),
        run_command_parameters: ssmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersToTerraform(struct.runCommandParameters),
        step_functions_parameters: ssmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersToTerraform(struct.stepFunctionsParameters),
    };
}
exports.ssmMaintenanceWindowTaskTaskInvocationParametersToTerraform = ssmMaintenanceWindowTaskTaskInvocationParametersToTerraform;
/**
 * @stability stable
 */
class SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
        // automation_parameters - computed: false, optional: true, required: false
        this._automationParameters = new SsmMaintenanceWindowTaskTaskInvocationParametersAutomationParametersOutputReference(this, "automation_parameters", true);
        // lambda_parameters - computed: false, optional: true, required: false
        this._lambdaParameters = new SsmMaintenanceWindowTaskTaskInvocationParametersLambdaParametersOutputReference(this, "lambda_parameters", true);
        // run_command_parameters - computed: false, optional: true, required: false
        this._runCommandParameters = new SsmMaintenanceWindowTaskTaskInvocationParametersRunCommandParametersOutputReference(this, "run_command_parameters", true);
        // step_functions_parameters - computed: false, optional: true, required: false
        this._stepFunctionsParameters = new SsmMaintenanceWindowTaskTaskInvocationParametersStepFunctionsParametersOutputReference(this, "step_functions_parameters", true);
    }
    /**
     * @stability stable
     */
    get internalValue() {
        var _j, _k, _l, _m, _o, _p, _q, _r;
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if ((_j = this._automationParameters) === null || _j === void 0 ? void 0 : _j.internalValue) {
            hasAnyValues = true;
            internalValueResult.automationParameters = (_k = this._automationParameters) === null || _k === void 0 ? void 0 : _k.internalValue;
        }
        if ((_l = this._lambdaParameters) === null || _l === void 0 ? void 0 : _l.internalValue) {
            hasAnyValues = true;
            internalValueResult.lambdaParameters = (_m = this._lambdaParameters) === null || _m === void 0 ? void 0 : _m.internalValue;
        }
        if ((_o = this._runCommandParameters) === null || _o === void 0 ? void 0 : _o.internalValue) {
            hasAnyValues = true;
            internalValueResult.runCommandParameters = (_p = this._runCommandParameters) === null || _p === void 0 ? void 0 : _p.internalValue;
        }
        if ((_q = this._stepFunctionsParameters) === null || _q === void 0 ? void 0 : _q.internalValue) {
            hasAnyValues = true;
            internalValueResult.stepFunctionsParameters = (_r = this._stepFunctionsParameters) === null || _r === void 0 ? void 0 : _r.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._automationParameters.internalValue = undefined;
            this._lambdaParameters.internalValue = undefined;
            this._runCommandParameters.internalValue = undefined;
            this._stepFunctionsParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._automationParameters.internalValue = value.automationParameters;
            this._lambdaParameters.internalValue = value.lambdaParameters;
            this._runCommandParameters.internalValue = value.runCommandParameters;
            this._stepFunctionsParameters.internalValue = value.stepFunctionsParameters;
        }
    }
    /**
     * @stability stable
     */
    get automationParameters() {
        return this._automationParameters;
    }
    /**
     * @stability stable
     */
    putAutomationParameters(value) {
        this._automationParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetAutomationParameters() {
        this._automationParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get automationParametersInput() {
        return this._automationParameters.internalValue;
    }
    /**
     * @stability stable
     */
    get lambdaParameters() {
        return this._lambdaParameters;
    }
    /**
     * @stability stable
     */
    putLambdaParameters(value) {
        this._lambdaParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetLambdaParameters() {
        this._lambdaParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get lambdaParametersInput() {
        return this._lambdaParameters.internalValue;
    }
    /**
     * @stability stable
     */
    get runCommandParameters() {
        return this._runCommandParameters;
    }
    /**
     * @stability stable
     */
    putRunCommandParameters(value) {
        this._runCommandParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetRunCommandParameters() {
        this._runCommandParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get runCommandParametersInput() {
        return this._runCommandParameters.internalValue;
    }
    /**
     * @stability stable
     */
    get stepFunctionsParameters() {
        return this._stepFunctionsParameters;
    }
    /**
     * @stability stable
     */
    putStepFunctionsParameters(value) {
        this._stepFunctionsParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetStepFunctionsParameters() {
        this._stepFunctionsParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get stepFunctionsParametersInput() {
        return this._stepFunctionsParameters.internalValue;
    }
}
exports.SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference = SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference[_g] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window_task.html aws_ssm_maintenance_window_task}.
 *
 * @stability stable
 */
class SsmMaintenanceWindowTask extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window_task.html aws_ssm_maintenance_window_task} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_maintenance_window_task',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // task_invocation_parameters - computed: false, optional: true, required: false
        this._taskInvocationParameters = new SsmMaintenanceWindowTaskTaskInvocationParametersOutputReference(this, "task_invocation_parameters", true);
        this._description = config.description;
        this._maxConcurrency = config.maxConcurrency;
        this._maxErrors = config.maxErrors;
        this._name = config.name;
        this._priority = config.priority;
        this._serviceRoleArn = config.serviceRoleArn;
        this._taskArn = config.taskArn;
        this._taskType = config.taskType;
        this._windowId = config.windowId;
        this._targets = config.targets;
        this._taskInvocationParameters.internalValue = config.taskInvocationParameters;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get maxConcurrency() {
        return this.getStringAttribute('max_concurrency');
    }
    /**
     * @stability stable
     */
    set maxConcurrency(value) {
        this._maxConcurrency = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxConcurrencyInput() {
        return this._maxConcurrency;
    }
    /**
     * @stability stable
     */
    get maxErrors() {
        return this.getStringAttribute('max_errors');
    }
    /**
     * @stability stable
     */
    set maxErrors(value) {
        this._maxErrors = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get maxErrorsInput() {
        return this._maxErrors;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    /**
     * @stability stable
     */
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get priority() {
        return this.getNumberAttribute('priority');
    }
    /**
     * @stability stable
     */
    set priority(value) {
        this._priority = value;
    }
    /**
     * @stability stable
     */
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get priorityInput() {
        return this._priority;
    }
    /**
     * @stability stable
     */
    get serviceRoleArn() {
        return this.getStringAttribute('service_role_arn');
    }
    /**
     * @stability stable
     */
    set serviceRoleArn(value) {
        this._serviceRoleArn = value;
    }
    /**
     * @stability stable
     */
    resetServiceRoleArn() {
        this._serviceRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get serviceRoleArnInput() {
        return this._serviceRoleArn;
    }
    /**
     * @stability stable
     */
    get taskArn() {
        return this.getStringAttribute('task_arn');
    }
    /**
     * @stability stable
     */
    set taskArn(value) {
        this._taskArn = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskArnInput() {
        return this._taskArn;
    }
    /**
     * @stability stable
     */
    get taskType() {
        return this.getStringAttribute('task_type');
    }
    /**
     * @stability stable
     */
    set taskType(value) {
        this._taskType = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskTypeInput() {
        return this._taskType;
    }
    /**
     * @stability stable
     */
    get windowId() {
        return this.getStringAttribute('window_id');
    }
    /**
     * @stability stable
     */
    set windowId(value) {
        this._windowId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get windowIdInput() {
        return this._windowId;
    }
    /**
     * @stability stable
     */
    get targets() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('targets');
    }
    /**
     * @stability stable
     */
    set targets(value) {
        this._targets = value;
    }
    /**
     * @stability stable
     */
    resetTargets() {
        this._targets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get targetsInput() {
        return this._targets;
    }
    /**
     * @stability stable
     */
    get taskInvocationParameters() {
        return this._taskInvocationParameters;
    }
    /**
     * @stability stable
     */
    putTaskInvocationParameters(value) {
        this._taskInvocationParameters.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetTaskInvocationParameters() {
        this._taskInvocationParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get taskInvocationParametersInput() {
        return this._taskInvocationParameters.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            max_concurrency: cdktf.stringToTerraform(this._maxConcurrency),
            max_errors: cdktf.stringToTerraform(this._maxErrors),
            name: cdktf.stringToTerraform(this._name),
            priority: cdktf.numberToTerraform(this._priority),
            service_role_arn: cdktf.stringToTerraform(this._serviceRoleArn),
            task_arn: cdktf.stringToTerraform(this._taskArn),
            task_type: cdktf.stringToTerraform(this._taskType),
            window_id: cdktf.stringToTerraform(this._windowId),
            targets: cdktf.listMapper(ssmMaintenanceWindowTaskTargetsToTerraform)(this._targets),
            task_invocation_parameters: ssmMaintenanceWindowTaskTaskInvocationParametersToTerraform(this._taskInvocationParameters.internalValue),
        };
    }
}
exports.SsmMaintenanceWindowTask = SsmMaintenanceWindowTask;
_h = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindowTask[_h] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindowTask", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsmMaintenanceWindowTask.tfResourceType = "aws_ssm_maintenance_window_task";
