// @cdktf/provider-aws/lib/ssm/ssm-maintenance-window.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmMaintenanceWindow = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window.html aws_ssm_maintenance_window}.
 *
 * @stability stable
 */
class SsmMaintenanceWindow extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_maintenance_window.html aws_ssm_maintenance_window} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_maintenance_window',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._allowUnassociatedTargets = config.allowUnassociatedTargets;
        this._cutoff = config.cutoff;
        this._description = config.description;
        this._duration = config.duration;
        this._enabled = config.enabled;
        this._endDate = config.endDate;
        this._name = config.name;
        this._schedule = config.schedule;
        this._scheduleOffset = config.scheduleOffset;
        this._scheduleTimezone = config.scheduleTimezone;
        this._startDate = config.startDate;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    /**
     * @stability stable
     */
    get allowUnassociatedTargets() {
        return this.getBooleanAttribute('allow_unassociated_targets');
    }
    /**
     * @stability stable
     */
    set allowUnassociatedTargets(value) {
        this._allowUnassociatedTargets = value;
    }
    /**
     * @stability stable
     */
    resetAllowUnassociatedTargets() {
        this._allowUnassociatedTargets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get allowUnassociatedTargetsInput() {
        return this._allowUnassociatedTargets;
    }
    /**
     * @stability stable
     */
    get cutoff() {
        return this.getNumberAttribute('cutoff');
    }
    /**
     * @stability stable
     */
    set cutoff(value) {
        this._cutoff = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cutoffInput() {
        return this._cutoff;
    }
    /**
     * @stability stable
     */
    get description() {
        return this.getStringAttribute('description');
    }
    /**
     * @stability stable
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @stability stable
     */
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get descriptionInput() {
        return this._description;
    }
    /**
     * @stability stable
     */
    get duration() {
        return this.getNumberAttribute('duration');
    }
    /**
     * @stability stable
     */
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get durationInput() {
        return this._duration;
    }
    /**
     * @stability stable
     */
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    /**
     * @stability stable
     */
    set enabled(value) {
        this._enabled = value;
    }
    /**
     * @stability stable
     */
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get enabledInput() {
        return this._enabled;
    }
    /**
     * @stability stable
     */
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    /**
     * @stability stable
     */
    set endDate(value) {
        this._endDate = value;
    }
    /**
     * @stability stable
     */
    resetEndDate() {
        this._endDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get endDateInput() {
        return this._endDate;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    /**
     * @stability stable
     */
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    /**
     * @stability stable
     */
    set schedule(value) {
        this._schedule = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleInput() {
        return this._schedule;
    }
    /**
     * @stability stable
     */
    get scheduleOffset() {
        return this.getNumberAttribute('schedule_offset');
    }
    /**
     * @stability stable
     */
    set scheduleOffset(value) {
        this._scheduleOffset = value;
    }
    /**
     * @stability stable
     */
    resetScheduleOffset() {
        this._scheduleOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleOffsetInput() {
        return this._scheduleOffset;
    }
    /**
     * @stability stable
     */
    get scheduleTimezone() {
        return this.getStringAttribute('schedule_timezone');
    }
    /**
     * @stability stable
     */
    set scheduleTimezone(value) {
        this._scheduleTimezone = value;
    }
    /**
     * @stability stable
     */
    resetScheduleTimezone() {
        this._scheduleTimezone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get scheduleTimezoneInput() {
        return this._scheduleTimezone;
    }
    /**
     * @stability stable
     */
    get startDate() {
        return this.getStringAttribute('start_date');
    }
    /**
     * @stability stable
     */
    set startDate(value) {
        this._startDate = value;
    }
    /**
     * @stability stable
     */
    resetStartDate() {
        this._startDate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get startDateInput() {
        return this._startDate;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get tagsAll() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags_all');
    }
    /**
     * @stability stable
     */
    set tagsAll(value) {
        this._tagsAll = value;
    }
    /**
     * @stability stable
     */
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            allow_unassociated_targets: cdktf.booleanToTerraform(this._allowUnassociatedTargets),
            cutoff: cdktf.numberToTerraform(this._cutoff),
            description: cdktf.stringToTerraform(this._description),
            duration: cdktf.numberToTerraform(this._duration),
            enabled: cdktf.booleanToTerraform(this._enabled),
            end_date: cdktf.stringToTerraform(this._endDate),
            name: cdktf.stringToTerraform(this._name),
            schedule: cdktf.stringToTerraform(this._schedule),
            schedule_offset: cdktf.numberToTerraform(this._scheduleOffset),
            schedule_timezone: cdktf.stringToTerraform(this._scheduleTimezone),
            start_date: cdktf.stringToTerraform(this._startDate),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.anyToTerraform)(this._tagsAll),
        };
    }
}
exports.SsmMaintenanceWindow = SsmMaintenanceWindow;
_a = JSII_RTTI_SYMBOL_1;
SsmMaintenanceWindow[_a] = { fqn: "@cdktf/provider-aws.ssm.SsmMaintenanceWindow", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
SsmMaintenanceWindow.tfResourceType = "aws_ssm_maintenance_window";
