// @cdktf/provider-aws/lib/vpc/data-aws-vpc-peering-connection.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpcPeeringConnection = exports.dataAwsVpcPeeringConnectionFilterToTerraform = exports.DataAwsVpcPeeringConnectionPeerCidrBlockSet = exports.DataAwsVpcPeeringConnectionCidrBlockSet = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
 * @stability stable
 */
class DataAwsVpcPeeringConnectionCidrBlockSet extends cdktf.ComplexComputedList {
    // cidr_block - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionCidrBlockSet = DataAwsVpcPeeringConnectionCidrBlockSet;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionCidrBlockSet[_a] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcPeeringConnectionCidrBlockSet", version: "3.0.1" };
/**
 * @stability stable
 */
class DataAwsVpcPeeringConnectionPeerCidrBlockSet extends cdktf.ComplexComputedList {
    // cidr_block - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
}
exports.DataAwsVpcPeeringConnectionPeerCidrBlockSet = DataAwsVpcPeeringConnectionPeerCidrBlockSet;
_b = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnectionPeerCidrBlockSet[_b] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcPeeringConnectionPeerCidrBlockSet", version: "3.0.1" };
function dataAwsVpcPeeringConnectionFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform)(struct.values),
    };
}
exports.dataAwsVpcPeeringConnectionFilterToTerraform = dataAwsVpcPeeringConnectionFilterToTerraform;
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/d/vpc_peering_connection.html aws_vpc_peering_connection}.
 *
 * @stability stable
 */
class DataAwsVpcPeeringConnection extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/d/vpc_peering_connection.html aws_vpc_peering_connection} Data Source.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_peering_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._cidrBlock = config.cidrBlock;
        this._ownerId = config.ownerId;
        this._peerCidrBlock = config.peerCidrBlock;
        this._peerOwnerId = config.peerOwnerId;
        this._peerRegion = config.peerRegion;
        this._peerVpcId = config.peerVpcId;
        this._region = config.region;
        this._status = config.status;
        this._tags = config.tags;
        this._vpcId = config.vpcId;
        this._filter = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // accepter - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    accepter(key) {
        return new cdktf.BooleanMap(this, 'accepter').lookup(key);
    }
    /**
     * @stability stable
     */
    get cidrBlock() {
        return this.getStringAttribute('cidr_block');
    }
    /**
     * @stability stable
     */
    set cidrBlock(value) {
        this._cidrBlock = value;
    }
    /**
     * @stability stable
     */
    resetCidrBlock() {
        this._cidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cidrBlockInput() {
        return this._cidrBlock;
    }
    // cidr_block_set - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    cidrBlockSet(index) {
        return new DataAwsVpcPeeringConnectionCidrBlockSet(this, 'cidr_block_set', index);
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    /**
     * @stability stable
     */
    set ownerId(value) {
        this._ownerId = value;
    }
    /**
     * @stability stable
     */
    resetOwnerId() {
        this._ownerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ownerIdInput() {
        return this._ownerId;
    }
    /**
     * @stability stable
     */
    get peerCidrBlock() {
        return this.getStringAttribute('peer_cidr_block');
    }
    /**
     * @stability stable
     */
    set peerCidrBlock(value) {
        this._peerCidrBlock = value;
    }
    /**
     * @stability stable
     */
    resetPeerCidrBlock() {
        this._peerCidrBlock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerCidrBlockInput() {
        return this._peerCidrBlock;
    }
    // peer_cidr_block_set - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    peerCidrBlockSet(index) {
        return new DataAwsVpcPeeringConnectionPeerCidrBlockSet(this, 'peer_cidr_block_set', index);
    }
    /**
     * @stability stable
     */
    get peerOwnerId() {
        return this.getStringAttribute('peer_owner_id');
    }
    /**
     * @stability stable
     */
    set peerOwnerId(value) {
        this._peerOwnerId = value;
    }
    /**
     * @stability stable
     */
    resetPeerOwnerId() {
        this._peerOwnerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerOwnerIdInput() {
        return this._peerOwnerId;
    }
    /**
     * @stability stable
     */
    get peerRegion() {
        return this.getStringAttribute('peer_region');
    }
    /**
     * @stability stable
     */
    set peerRegion(value) {
        this._peerRegion = value;
    }
    /**
     * @stability stable
     */
    resetPeerRegion() {
        this._peerRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerRegionInput() {
        return this._peerRegion;
    }
    /**
     * @stability stable
     */
    get peerVpcId() {
        return this.getStringAttribute('peer_vpc_id');
    }
    /**
     * @stability stable
     */
    set peerVpcId(value) {
        this._peerVpcId = value;
    }
    /**
     * @stability stable
     */
    resetPeerVpcId() {
        this._peerVpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get peerVpcIdInput() {
        return this._peerVpcId;
    }
    /**
     * @stability stable
     */
    get region() {
        return this.getStringAttribute('region');
    }
    /**
     * @stability stable
     */
    set region(value) {
        this._region = value;
    }
    /**
     * @stability stable
     */
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get regionInput() {
        return this._region;
    }
    // requester - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    requester(key) {
        return new cdktf.BooleanMap(this, 'requester').lookup(key);
    }
    /**
     * @stability stable
     */
    get status() {
        return this.getStringAttribute('status');
    }
    /**
     * @stability stable
     */
    set status(value) {
        this._status = value;
    }
    /**
     * @stability stable
     */
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get statusInput() {
        return this._status;
    }
    /**
     * @stability stable
     */
    get tags() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('tags');
    }
    /**
     * @stability stable
     */
    set tags(value) {
        this._tags = value;
    }
    /**
     * @stability stable
     */
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get tagsInput() {
        return this._tags;
    }
    /**
     * @stability stable
     */
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    /**
     * @stability stable
     */
    set vpcId(value) {
        this._vpcId = value;
    }
    /**
     * @stability stable
     */
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get vpcIdInput() {
        return this._vpcId;
    }
    /**
     * @stability stable
     */
    get filter() {
        // Getting the computed value is not yet implemented
        return this.interpolationForAttribute('filter');
    }
    /**
     * @stability stable
     */
    set filter(value) {
        this._filter = value;
    }
    /**
     * @stability stable
     */
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get filterInput() {
        return this._filter;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cidr_block: cdktf.stringToTerraform(this._cidrBlock),
            owner_id: cdktf.stringToTerraform(this._ownerId),
            peer_cidr_block: cdktf.stringToTerraform(this._peerCidrBlock),
            peer_owner_id: cdktf.stringToTerraform(this._peerOwnerId),
            peer_region: cdktf.stringToTerraform(this._peerRegion),
            peer_vpc_id: cdktf.stringToTerraform(this._peerVpcId),
            region: cdktf.stringToTerraform(this._region),
            status: cdktf.stringToTerraform(this._status),
            tags: cdktf.hashMapper(cdktf.anyToTerraform)(this._tags),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            filter: cdktf.listMapper(dataAwsVpcPeeringConnectionFilterToTerraform)(this._filter),
        };
    }
}
exports.DataAwsVpcPeeringConnection = DataAwsVpcPeeringConnection;
_c = JSII_RTTI_SYMBOL_1;
DataAwsVpcPeeringConnection[_c] = { fqn: "@cdktf/provider-aws.vpc.DataAwsVpcPeeringConnection", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
DataAwsVpcPeeringConnection.tfResourceType = "aws_vpc_peering_connection";
