// @cdktf/provider-aws/lib/vpc/vpc-ipam-pool-cidr.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcIpamPoolCidr = exports.VpcIpamPoolCidrCidrAuthorizationContextOutputReference = exports.vpcIpamPoolCidrCidrAuthorizationContextToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcIpamPoolCidrCidrAuthorizationContextToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        message: cdktf.stringToTerraform(struct.message),
        signature: cdktf.stringToTerraform(struct.signature),
    };
}
exports.vpcIpamPoolCidrCidrAuthorizationContextToTerraform = vpcIpamPoolCidrCidrAuthorizationContextToTerraform;
/**
 * @stability stable
 */
class VpcIpamPoolCidrCidrAuthorizationContextOutputReference extends cdktf.ComplexObject {
    /**
     * @param terraformResource The parent resource.
     * @param terraformAttribute The attribute on the parent resource this class is referencing.
     * @param isSingleItem True if this is a block, false if it's a list.
     * @stability stable
     */
    constructor(terraformResource, terraformAttribute, isSingleItem) {
        super(terraformResource, terraformAttribute, isSingleItem);
        this.isEmptyObject = false;
    }
    /**
     * @stability stable
     */
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._message) {
            hasAnyValues = true;
            internalValueResult.message = this._message;
        }
        if (this._signature) {
            hasAnyValues = true;
            internalValueResult.signature = this._signature;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    /**
     * @stability stable
     */
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._message = undefined;
            this._signature = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._message = value.message;
            this._signature = value.signature;
        }
    }
    /**
     * @stability stable
     */
    get message() {
        return this.getStringAttribute('message');
    }
    /**
     * @stability stable
     */
    set message(value) {
        this._message = value;
    }
    /**
     * @stability stable
     */
    resetMessage() {
        this._message = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get messageInput() {
        return this._message;
    }
    /**
     * @stability stable
     */
    get signature() {
        return this.getStringAttribute('signature');
    }
    /**
     * @stability stable
     */
    set signature(value) {
        this._signature = value;
    }
    /**
     * @stability stable
     */
    resetSignature() {
        this._signature = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get signatureInput() {
        return this._signature;
    }
}
exports.VpcIpamPoolCidrCidrAuthorizationContextOutputReference = VpcIpamPoolCidrCidrAuthorizationContextOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcIpamPoolCidrCidrAuthorizationContextOutputReference[_a] = { fqn: "@cdktf/provider-aws.vpc.VpcIpamPoolCidrCidrAuthorizationContextOutputReference", version: "3.0.1" };
/**
 * Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam_pool_cidr.html aws_vpc_ipam_pool_cidr}.
 *
 * @stability stable
 */
class VpcIpamPoolCidr extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam_pool_cidr.html aws_vpc_ipam_pool_cidr} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam_pool_cidr',
            terraformGeneratorMetadata: {
                providerName: 'aws'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        // cidr_authorization_context - computed: false, optional: true, required: false
        this._cidrAuthorizationContext = new VpcIpamPoolCidrCidrAuthorizationContextOutputReference(this, "cidr_authorization_context", true);
        this._cidr = config.cidr;
        this._ipamPoolId = config.ipamPoolId;
        this._cidrAuthorizationContext.internalValue = config.cidrAuthorizationContext;
    }
    /**
     * @stability stable
     */
    get cidr() {
        return this.getStringAttribute('cidr');
    }
    /**
     * @stability stable
     */
    set cidr(value) {
        this._cidr = value;
    }
    /**
     * @stability stable
     */
    resetCidr() {
        this._cidr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cidrInput() {
        return this._cidr;
    }
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get ipamPoolId() {
        return this.getStringAttribute('ipam_pool_id');
    }
    /**
     * @stability stable
     */
    set ipamPoolId(value) {
        this._ipamPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get ipamPoolIdInput() {
        return this._ipamPoolId;
    }
    /**
     * @stability stable
     */
    get cidrAuthorizationContext() {
        return this._cidrAuthorizationContext;
    }
    /**
     * @stability stable
     */
    putCidrAuthorizationContext(value) {
        this._cidrAuthorizationContext.internalValue = value;
    }
    /**
     * @stability stable
     */
    resetCidrAuthorizationContext() {
        this._cidrAuthorizationContext.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get cidrAuthorizationContextInput() {
        return this._cidrAuthorizationContext.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            cidr: cdktf.stringToTerraform(this._cidr),
            ipam_pool_id: cdktf.stringToTerraform(this._ipamPoolId),
            cidr_authorization_context: vpcIpamPoolCidrCidrAuthorizationContextToTerraform(this._cidrAuthorizationContext.internalValue),
        };
    }
}
exports.VpcIpamPoolCidr = VpcIpamPoolCidr;
_b = JSII_RTTI_SYMBOL_1;
VpcIpamPoolCidr[_b] = { fqn: "@cdktf/provider-aws.vpc.VpcIpamPoolCidr", version: "3.0.1" };
// =================
// STATIC PROPERTIES
// =================
/**
 * @stability stable
 */
VpcIpamPoolCidr.tfResourceType = "aws_vpc_ipam_pool_cidr";
